// lightning-base-components/src/lightning/formattedRichText/formattedRichText.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import LightningFormattedRichTextRenderer from 'lightning/formattedRichTextRenderer';
import { privateContext } from 'lightning/utilsInternal';
import { features } from 'lightning/features';
const { setContext, assertContext } = privateContext;
export default class LightningFormattedRichText extends LightningFormattedRichTextRenderer {
    csrOnly = false;
    get enableFormattedRichTextParts() {
        return features.enableFormattedRichTextParts;
    }
    constructor() {
        super();
        if (features.enableAssertContext260) {
            setContext(this);
        }
    }
    connectedCallback() {
        super.connectedCallback();
        this.csrOnly =
            !import.meta.env.SSR &&
                !this.template.querySelector('lightning-formatted-rich-text-renderer');
    }
    renderedCallback() {
        if (this.csrOnly) {
            super.renderedCallback();
        }
    }
    renderRichText() {
        if (features.enableAssertContext260) {
            assertContext(this);
        }
        if (this.csrOnly) {
            super.renderRichText();
            const container = this.container;
            if (this.rawTextIfSanitizerThrewError) {
                const textNode = document.createTextNode(this.rawTextIfSanitizerThrewError);
                while (container.hasChildNodes()) {
                    container.removeChild(container.lastChild);
                }
                container.appendChild(textNode);
            }
            else {
                container.innerHTML = this.richText;
                if (this.linkingRequired) {
                    this.linkRichText();
                }
            }
        }
    }
    disconnectedCallback() {
        if (this.csrOnly) {
            super.disconnectedCallback();
        }
    }
}
