// lightning-base-components/src/lightning/button/kinetics.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
export const CLICK_CLASS = 'slds-kx-is-animating-from-click';
export const KineticsType = {
    Underline: 'underline',
    Ripple: 'ripple',
};
const variantKinectAttributes = {
    base: {
        'kx-scope': 'button',
        'kx-type': KineticsType.Underline,
    },
    neutral: {
        'kx-scope': 'button-neutral',
        'kx-type': KineticsType.Ripple,
    },
    brand: {
        'kx-scope': 'button-brand',
        'kx-type': KineticsType.Ripple,
    },
    'brand-outline': {
        'kx-scope': 'button-outline',
        'kx-type': KineticsType.Ripple,
    },
    destructive: {
        'kx-scope': 'button-filled',
        'kx-type': KineticsType.Ripple,
    },
    'destructive-text': {
        'kx-scope': 'button-outline',
        'kx-type': KineticsType.Ripple,
    },
    inverse: {
        'kx-scope': 'button-outline',
        'kx-type': KineticsType.Ripple,
    },
    success: {
        'kx-scope': 'button-filled',
        'kx-type': KineticsType.Ripple,
    },
};
export function getKineticsAttributes(variant) {
    if (!variant) {
        console.warn('getKineticsAttributes: variant is required');
        return [];
    }
    const attributes = [];
    const map = variantKinectAttributes[variant];
    Object.keys(map).forEach((attribute) => {
        attributes.push({ name: attribute, value: map[attribute] });
    });
    return attributes;
}
const previousTimeRef = {};
const requestRef = {};
const previousPointerRef = {};
const pointerRef = {};
export function getComponentName(element) {
    const kxScope = element.getAttribute('kx-scope');
    const match = kxScope ? kxScope.match(/^(\w+)/) : null;
    return match ? match[1] : '';
}
export function setCssVariables(element) {
    if (!pointerRef.current ||
        pointerRef.current === previousPointerRef.current) {
        return;
    }
    const { offsetX, offsetY } = pointerRef.current;
    const component = getComponentName(element);
    element.style.setProperty(`--slds-kx-${component}-pointer-position-x`, `${offsetX}px`);
    const kxType = element.getAttribute('kx-type');
    if (kxType !== KineticsType.Underline) {
        element.style.setProperty(`--slds-kx-${component}-pointer-position-y`, `${offsetY}px`);
    }
}
export function removeCssVariables(element) {
    if (!element) {
        return;
    }
    const component = getComponentName(element);
    element.style.removeProperty(`--slds-kx-${component}-pointer-position-x`);
    element.style.removeProperty(`--slds-kx-${component}-pointer-position-y`);
}
export function handleMouseMove(event) {
    if (!event) {
        return;
    }
    const { offsetX, offsetY } = event;
    pointerRef.current = { offsetX, offsetY };
}
function handleMouseEnter(event) {
    const { target } = event;
    enterAnimate(target);
}
export function animate(element) {
    if (!element) {
        return;
    }
    element.addEventListener('mouseenter', handleMouseEnter);
}
function enterAnimate(element, timestamp) {
    const kxType = element.getAttribute('kx-type');
    function handleClick() {
        clickAnimate(element);
    }
    function handleMouseLeave() {
        window.cancelAnimationFrame(requestRef.current);
        element.removeEventListener('mousemove', handleMouseMove);
        element.removeEventListener('click', handleClick);
        element.removeEventListener('mouseleave', handleMouseLeave);
        setTimeout(() => {
            removeCssVariables(element);
        }, 250);
    }
    if (!timestamp) {
        if (kxType !== KineticsType.Underline) {
            element.addEventListener('click', handleClick);
        }
        element.addEventListener('mouseleave', handleMouseLeave);
        element.addEventListener('mousemove', handleMouseMove);
    }
    if (previousTimeRef.current !== undefined) {
        setCssVariables(element);
        previousPointerRef.current = pointerRef.current;
    }
    previousTimeRef.current = timestamp;
    requestRef.current = window.requestAnimationFrame((timestamp2) => {
        enterAnimate(element, timestamp2);
    });
}
let cacheHandleAnimationEnd;
function getSingletonHandleAnimationEnd(element) {
    if (cacheHandleAnimationEnd) {
        return cacheHandleAnimationEnd;
    }
    cacheHandleAnimationEnd = function () {
        element.classList.remove(CLICK_CLASS);
        element.removeEventListener('animationend', cacheHandleAnimationEnd);
    };
    return cacheHandleAnimationEnd;
}
function clickAnimate(element) {
    const handleAnimationEnd = getSingletonHandleAnimationEnd(element);
    element.removeEventListener('animationend', handleAnimationEnd);
    element.classList.remove(CLICK_CLASS);
    void element.offsetWidth;
    element.addEventListener('animationend', handleAnimationEnd);
    element.classList.add(CLICK_CLASS);
}
