// lightning-base-components/src/lightning/primitiveHeaderActions/primitiveHeaderActions.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import labelClipText from '@salesforce/label/LightningDatatable.clipText';
import labelShowActions from '@salesforce/label/LightningDatatable.showActions';
import labelWrapText from '@salesforce/label/LightningDatatable.wrapText';
import { api, track } from 'lwc';
import { deepCopy } from 'lightning/utilsPrivate';
import { formatLabel } from 'lightning/utils';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
const i18n = {
    clipText: labelClipText,
    showActions: labelShowActions,
    wrapText: labelWrapText,
};
export default class PrimitiveHeaderActions extends LightningShadowBaseClass {
    @track
    containerRect;
    @track
    _internalActions = [];
    @track
    _customerActions = [];
    static delegatesFocus = true;
    _actionMenuAlignment;
    @api
    colKeyValue;
    @api
    get actions() {
        return this._actions;
    }
    set actions(value) {
        this._actions = value;
        this.updateActions();
    }
    @api
    columnHeader;
    @api
    focus() {
        const btnMenu = this.refs.buttonMenu;
        if (btnMenu) {
            btnMenu.focus();
        }
    }
    get i18n() {
        return i18n;
    }
    get hasActions() {
        return (this._internalActions.length > 0 || this._customerActions.length > 0);
    }
    get hasActionsDivider() {
        return (this._internalActions.length > 0 && this._customerActions.length > 0);
    }
    handleMenuOpen(event) {
        event.preventDefault();
        event.stopPropagation();
        this.elementRect = this.refs.buttonMenu.getBoundingClientRect();
        this.dispatchEvent(new CustomEvent('privatecellheaderactionmenuopening', {
            bubbles: true,
            composed: true,
            cancelable: true,
            detail: {
                actionsCount: this._internalActions.length +
                    this._customerActions.length,
                dividersCount: this.hasActionsDivider ? 1 : 0,
                saveContainerPosition: (containerRect) => {
                    this.containerRect = containerRect;
                },
            },
        }));
    }
    handleMenuClose() {
        this.dispatchEvent(new CustomEvent('privatecellheaderactionmenuclosed', {
            bubbles: true,
            composed: true,
            cancelable: true,
        }));
    }
    handleActionSelect(event) {
        const action = event.detail.value;
        this.dispatchEvent(new CustomEvent('privatecellheaderactiontriggered', {
            composed: true,
            bubbles: true,
            cancelable: true,
            detail: {
                action: deepCopy(action._action),
                actionType: action._type,
                colKeyValue: this.colKeyValue,
            },
        }));
    }
    updateActions() {
        const actionTypeReducer = (type) => (actions, action) => {
            const overrides = { _type: type, _action: action };
            actions.push(Object.assign({}, action, overrides));
            return actions;
        };
        this._internalActions = this.getActionsByType('internalActions').reduce(actionTypeReducer('internal'), []);
        this._customerActions = this.getActionsByType('customerActions').reduce(actionTypeReducer('customer'), []);
        this._actionMenuAlignment =
            this._actions.menuAlignment &&
                this._actions.menuAlignment.replace('auto-', '');
    }
    getActionsByType(type) {
        return Array.isArray(this._actions[type]) ? this._actions[type] : [];
    }
    get alternativeText() {
        return formatLabel(i18n.showActions, this.columnHeader || '');
    }
}
