// lightning-base-components/src/lightning/internationalizationLibrary/datetime/dateTimeUtils.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { assert } from 'lightning/utilsPrivate';
import locale from '@salesforce/i18n/locale';
import shortDateFormat from '@salesforce/i18n/dateTime.shortDateFormat';
import mediumDateFormat from '@salesforce/i18n/dateTime.mediumDateFormat';
import longDateFormat from '@salesforce/i18n/dateTime.longDateFormat';
import shortTimeFormat from '@salesforce/i18n/dateTime.shortTimeFormat';
import mediumTimeFormat from '@salesforce/i18n/dateTime.mediumTimeFormat';
import getFormattingOptions from 'lightning/i18nCldrOptions';
import { getDateTimeFormat, getDateTimeISO8601Parser, getDateTimeCLDRParser, } from 'lightning/i18nService';
import { isValidISODateTimeString, isValidISOTimeString, padMilliseconds, removeTimeZoneSuffix, STANDARD_DATE_FORMAT, TIME_SEPARATOR, } from 'lightning/iso8601Utils';
import { formatDate, formatTime, formatDateTimeUTC, parseDateTimeUTC, parseDateTimeISO8601, syncUTCToWallTime, syncWallTimeToUTC, parseTime, parseDateTime, toOtherCalendar, fromOtherCalendar, } from '../localizationService';
import { isAuraL10NAvailable, getDateTimeErrorMessage } from '../utils';
import { features } from 'lightning/features';
const ISO_FORMAT = "yyyy-MM-dd'T'hh:mm:ss.SSS'Z'";
const LATN = 'latn';
const utcTimeZone = { timeZone: 'UTC' };
const hasTimeZoneOffset = (value) => {
    if (typeof value === 'string') {
        return value.indexOf('+') !== -1 || value.indexOf('-') !== -1;
    }
    return false;
};
const getTrimmedValue = (value) => {
    return typeof value === 'string' ? value.trim() : value;
};
const getNormalizedValue = (value) => {
    const trimmedValue = getTrimmedValue(value);
    return padMilliseconds(trimmedValue);
};
const getDateOnly = (value) => {
    return typeof value === 'string' ? value.split(TIME_SEPARATOR)[0] : value;
};
export function isDate(value) {
    return (Object.prototype.toString.call(value) === '[object Date]' &&
        !isNaN(value.getTime()));
}
export function getCurrentDateString() {
    const today = getTodayBasedOnTimezone();
    return getISODateString(today);
}
export function getISODateString(date) {
    if (!isDate(date)) {
        return null;
    }
    return `${date.getFullYear()}-${pad(date.getMonth() + 1)}-${pad(date.getDate())}`;
}
export function getISOTimeString(date) {
    if (!isDate(date)) {
        return null;
    }
    return `${pad(date.getHours())}:${pad(date.getMinutes())}:${pad(date.getSeconds())}.${doublePad(date.getMilliseconds())}`;
}
function getISOTimeStringUTC(date) {
    if (!isDate(date)) {
        return null;
    }
    return `${pad(date.getUTCHours())}:${pad(date.getUTCMinutes())}:${pad(date.getUTCSeconds())}.${doublePad(date.getUTCMilliseconds())}`;
}
export function getCurrentTimeString(timezone) {
    const today = getTodayBasedOnTimezone(timezone);
    return `${pad(today.getHours())}:${pad(today.getMinutes())}`;
}
function getOffsetForTimezone(date, timeZone, shouldAddUTCOffset) {
    const time1 = getDateTimeCLDRParser({
        pattern: ISO_FORMAT,
        timeZone,
        numberingSystem: LATN,
    }).parse(date.toISOString());
    const extraOffset = shouldAddUTCOffset ? date.getTimezoneOffset() : 0;
    const time2 = (date.getTime() - date.getMilliseconds()) + (extraOffset * 60 * 1000);
    return time2 - time1;
}
function adjustDateByOffset(date, timeZone, shouldAddUTCOffset) {
    const offset = getOffsetForTimezone(date, timeZone, shouldAddUTCOffset);
    const timestamp = date.getTime() + offset;
    return new Date(timestamp);
}
function getTodayBasedOnTimezone(timeZone) {
    const today = new Date();
    if (isAuraL10NAvailable) {
        today.setTime(today.getTime() + (today.getTimezoneOffset() * 60 * 1000));
        return syncUTCToWallTime(today, timeZone);
    }
    return adjustDateByOffset(today, timeZone, true);
}
function pad(n) {
    return Number(n) < 10 ? '0' + n : n;
}
function doublePad(n) {
    const number = Number(n);
    if (number < 10) {
        return '00' + n;
    }
    else if (number < 100) {
        return '0' + n;
    }
    return n;
}
function hasDateRolledOver(originalValue, parsedDate, format) {
    const formatOptions = getFormattingOptions(format);
    const utcOptions = Object.assign({}, utcTimeZone, formatOptions);
    const formattedDate = getDateTimeFormat(locale, utcOptions).format(parsedDate);
    const originalNumbers = originalValue.match(/\d+/g) || [];
    const formattedNumbers = formattedDate.match(/\d+/g) || [];
    return originalNumbers.join(',') !== formattedNumbers.join(',');
}
const DATE = 'date';
const TIME = 'time';
const DATETIME_STYLES = {
    SHORT: 'short',
    MEDIUM: 'medium',
    LONG: 'long',
};
const DATETIME_FORMATS = {
    [DATE]: {
        [DATETIME_STYLES.SHORT]: shortDateFormat,
        [DATETIME_STYLES.MEDIUM]: mediumDateFormat,
        [DATETIME_STYLES.LONG]: longDateFormat,
        default: mediumDateFormat,
    },
    [TIME]: {
        [DATETIME_STYLES.SHORT]: shortTimeFormat,
        [DATETIME_STYLES.MEDIUM]: mediumTimeFormat,
        [DATETIME_STYLES.LONG]: mediumTimeFormat,
        default: shortTimeFormat,
    },
};
const getFormatFromStyle = (type, style, isCldr) => {
    let cldrString = DATETIME_FORMATS[type][style];
    if (!cldrString) {
        cldrString = DATETIME_FORMATS[type].default;
    }
    if (isAuraL10NAvailable || isCldr) {
        return cldrString;
    }
    return getFormattingOptions(cldrString);
};
export const getDateFormatFromStyle = (dateStyle, isCldr) => {
    return getFormatFromStyle(DATE, dateStyle, isCldr);
};
export const getTimeFormatFromStyle = (timeStyle, isCldr) => {
    return getFormatFromStyle(TIME, timeStyle, isCldr);
};
export function normalizeISODate(value, dateStyle) {
    const format = getDateFormatFromStyle(dateStyle);
    const isoValue = getDateOnly(getNormalizedValue(value));
    let parsedDate, displayValue;
    if (isoValue) {
        const isValid = isValidISODateTimeString(isoValue);
        if (!import.meta.env.SSR && !isValid) {
            if (features.enableDateInvalidYearHandling) {
                console.error(getDateTimeErrorMessage(value));
            }
            else {
                assert(isValid, getDateTimeErrorMessage(value));
            }
        }
        else if (!isValid) {
            return {
                isoValue: null,
                displayValue: '',
            };
        }
        try {
            if (isAuraL10NAvailable) {
                parsedDate = parseDateTime(isoValue, STANDARD_DATE_FORMAT);
                const civilDate = toOtherCalendar(parsedDate);
                displayValue = formatDate(civilDate, format);
            }
            else {
                parsedDate = getDateTimeISO8601Parser().parse(isoValue);
                const formatOptions = Object.assign({}, utcTimeZone, format);
                displayValue = getDateTimeFormat(locale, formatOptions).format(parsedDate);
            }
        }
        catch (error) {
            console.warn(error.message);
            parsedDate = null;
        }
    }
    if (!parsedDate) {
        return {
            isoValue: null,
            displayValue: value || '',
        };
    }
    return { isoValue, displayValue };
}
export function normalizeISOTime(value, timeStyle) {
    const format = getTimeFormatFromStyle(timeStyle);
    const timeValue = getNormalizedValue(removeTimeZoneSuffix(value));
    let parsedValue, displayValue, getISO;
    if (timeValue) {
        const isValid = isValidISOTimeString(timeValue);
        if (!import.meta.env.SSR && !isValid) {
            if (features.enableDateInvalidYearHandling) {
                console.error(getDateTimeErrorMessage(value));
            }
            else {
                assert(isValid, getDateTimeErrorMessage(value));
            }
        }
        else if (!isValid) {
            return {
                isoValue: null,
                parsedValue: null,
                displayValue: '',
            };
        }
        try {
            if (isAuraL10NAvailable) {
                parsedValue = parseTime(timeValue);
                displayValue = formatTime(parsedValue, format);
                getISO = getISOTimeString;
            }
            else {
                parsedValue = getDateTimeISO8601Parser().parse(timeValue);
                const formatOptions = Object.assign({}, utcTimeZone, format);
                displayValue = getDateTimeFormat(locale, formatOptions).format(parsedValue);
                getISO = getISOTimeStringUTC;
            }
        }
        catch (error) {
            console.warn(error.message);
            parsedValue = null;
        }
    }
    if (!parsedValue) {
        return {
            isoValue: null,
            parsedValue: null,
            displayValue: value || '',
        };
    }
    else if (hasTimeZoneOffset(value)) {
        console.warn('lightning-formatted-time does not support non-UTC time zones and will ignore any time zone offsets.');
    }
    const isoValue = getISO(parsedValue);
    return { isoValue, parsedValue, displayValue };
}
export function normalizeISODateTime(value, timeZone) {
    const dateTimeValue = getNormalizedValue(value);
    let parsedDate, displayValue;
    if (dateTimeValue) {
        const isValid = isValidISODateTimeString(dateTimeValue);
        if (!import.meta.env.SSR && !isValid) {
            if (features.enableDateInvalidYearHandling) {
                console.error(getDateTimeErrorMessage(value));
            }
            else {
                assert(isValid, getDateTimeErrorMessage(value));
            }
        }
        else if (!isValid) {
            return {
                isoValue: null,
                displayValue: '',
            };
        }
        try {
            if (isAuraL10NAvailable) {
                const utcDate = parseDateTimeISO8601(dateTimeValue);
                if (utcDate) {
                    parsedDate = syncUTCToWallTime(utcDate, timeZone);
                    displayValue = formatDateTimeUTC(parsedDate);
                }
            }
            else {
                const utcDate = getDateTimeISO8601Parser().parse(dateTimeValue);
                parsedDate = adjustDateByOffset(utcDate, timeZone);
                const dateFormat = getFormattingOptions(mediumDateFormat);
                const timeFormat = getFormattingOptions(mediumTimeFormat);
                const formatOptions = Object.assign({}, utcTimeZone, dateFormat, timeFormat);
                displayValue = getDateTimeFormat(locale, formatOptions).format(parsedDate);
            }
        }
        catch (error) {
            console.warn(error.message);
            parsedDate = null;
        }
    }
    if (!parsedDate) {
        return {
            isoValue: null,
            displayValue: value || '',
        };
    }
    return {
        isoValue: removeTimeZoneSuffix(parsedDate.toISOString()),
        displayValue,
    };
}
export function normalizeFormattedDate(value, dateStyle) {
    const dateValue = getTrimmedValue(value);
    const format = getDateFormatFromStyle(dateStyle, true);
    let parsedDate;
    try {
        if (dateValue) {
            if (isAuraL10NAvailable) {
                parsedDate = fromOtherCalendar(parseDateTime(dateValue, format, true));
            }
            else {
                parsedDate = getDateTimeCLDRParser({
                    pattern: format,
                }).parse(dateValue);
                if (parsedDate && hasDateRolledOver(dateValue, parsedDate, format)) {
                    return null;
                }
            }
        }
    }
    catch (error) {
        return null;
    }
    if (!parsedDate ||
        (features.enableDateInvalidYearHandling &&
            parsedDate.getFullYear().toString().length !== 4)) {
        return null;
    }
    return getISODateString(parsedDate);
}
export function normalizeFormattedTime(value, timeStyle) {
    const timeValue = getTrimmedValue(value);
    const format = getTimeFormatFromStyle(timeStyle, true);
    let parsedTime;
    try {
        if (timeValue) {
            if (isAuraL10NAvailable) {
                parsedTime = parseTime(timeValue, format || mediumTimeFormat, true);
            }
            else {
                parsedTime = getDateTimeCLDRParser({
                    pattern: format,
                }).parse(timeValue);
            }
        }
    }
    catch (error) {
        return null;
    }
    if (!parsedTime) {
        return null;
    }
    return getISOTimeString(parsedTime);
}
export function normalizeFormattedDateTime(value, timeZone) {
    const datetimeValue = getTrimmedValue(value);
    let parsedDate;
    try {
        if (datetimeValue) {
            if (isAuraL10NAvailable) {
                const utcDate = parseDateTimeUTC(datetimeValue);
                if (utcDate) {
                    parsedDate = syncWallTimeToUTC(utcDate, timeZone);
                }
            }
            else {
                const utcDate = getDateTimeISO8601Parser().parse(datetimeValue);
                parsedDate = getDateTimeCLDRParser({
                    pattern: ISO_FORMAT,
                    timeZone,
                    numberingSystem: LATN,
                }).parse(utcDate.toISOString());
            }
        }
    }
    catch (error) {
        return null;
    }
    if (!parsedDate ||
        (features.enableDateInvalidYearHandling &&
            parsedDate.getFullYear().toString().length !== 4)) {
        return null;
    }
    return parsedDate.toISOString();
}
const parseFormattedValue = (normalizeFormattedValue, displayValue) => {
    const allowedStyles = Object.values(DATETIME_STYLES);
    let value = null, style = null;
    for (let i = 0; i < allowedStyles.length; i++) {
        value = normalizeFormattedValue(displayValue, allowedStyles[i]);
        if (value) {
            style = allowedStyles[i];
            break;
        }
    }
    return { value, style };
};
export const parseFormattedTime = (value) => {
    return parseFormattedValue(normalizeFormattedTime, value);
};
export const parseFormattedDate = (value) => {
    return parseFormattedValue(normalizeFormattedDate, value);
};
