// lightning-base-components/src/lightning/internationalizationLibrary/datetime/intlFormat.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import locale from '@salesforce/i18n/locale';
import shortDateFormat from '@salesforce/i18n/dateTime.shortDateFormat';
import mediumDateFormat from '@salesforce/i18n/dateTime.mediumDateFormat';
import longDateFormat from '@salesforce/i18n/dateTime.longDateFormat';
import getFormattingOptions from 'lightning/i18nCldrOptions';
import { getSkeleton } from './dateTimeOptions';
import { isDate } from './dateTimeUtils';
import { formatDateUTC, formatDate, toOtherCalendar, syncUTCToWallTime, } from '../localizationService';
import { isAuraL10NAvailable } from '../utils';
import { isValidISODateTimeString, padMilliseconds, TIME_SEPARATOR, } from 'lightning/iso8601Utils';
import { getDateTimeFormat, getDateTimeISO8601Parser, } from 'lightning/i18nService';
function convertAndFormatDate(date, format, timeZone) {
    const translatedDate = toOtherCalendar(date);
    const converted = syncUTCToWallTime(translatedDate, timeZone);
    return formatDateUTC(converted, format);
}
const dateFormatMap = {
    short: shortDateFormat,
    medium: mediumDateFormat,
    long: longDateFormat,
};
export function toDate(value) {
    try {
        return toDateStrict(value);
    }
    catch (err) {
        console.warn(`'${value}' does not follow a supported date format. Please use either a timestamp, ISO8601 string, or Date object to avoid risking inconsistencies and breakages.`);
        return new Date(value);
    }
}
function toDateStrict(value) {
    if (value && !isDate(value) && isFinite(value)) {
        return new Date(parseInt(value, 10));
    }
    else if (value && !isDate(value)) {
        const strictString = padMilliseconds(value);
        return getDateTimeISO8601Parser().parse(strictString);
    }
    return value;
}
const isNonUTCSupported = (function () {
    try {
        Intl.DateTimeFormat('en-US', { timeZone: 'America/Los_Angeles' });
    }
    catch (err) {
        return false;
    }
    return true;
})();
function dateTimeFormatFallback(options, dateStyle) {
    const defaultFormat = dateFormatMap[dateStyle];
    const format = !useDefaultFormat(options)
        ? getSkeleton(options)
        : defaultFormat;
    const { timeZone } = options;
    return {
        format: (value) => {
            if (isValidISODateTimeString(value) &&
                value.indexOf(TIME_SEPARATOR) < 0) {
                return formatDate(value, dateFormatMap[dateStyle]);
            }
            const dateObj = toDate(value);
            if (isDate(dateObj)) {
                if (timeZone === 'UTC') {
                    dateObj.setTime(dateObj.getTime() +
                        (dateObj.getTimezoneOffset() * 60 * 1000));
                }
                return convertAndFormatDate(dateObj, format, timeZone);
            }
            return '';
        },
    };
}
function useAuraFallbackFormatting(options) {
    const { timeZone } = options;
    const isNonUTCTimeZone = timeZone && timeZone !== 'UTC';
    if (isAuraL10NAvailable &&
        (useDefaultFormat(options) || (!isNonUTCSupported && isNonUTCTimeZone))) {
        return true;
    }
    return false;
}
function useDefaultFormat(options) {
    const excludedOptions = {
        hour12: true,
        hourCycle: true,
        timeZone: true,
        timeZoneName: true,
    };
    const hasFormattingOption = Object.keys(options).some((opt) => options[opt] !== undefined && !excludedOptions[opt]);
    return !hasFormattingOption;
}
export function dateTimeFormat(opts, dateStyle = 'medium') {
    const options = opts || {};
    if (useAuraFallbackFormatting(options)) {
        return dateTimeFormatFallback(options, dateStyle);
    }
    else if (useDefaultFormat(options)) {
        const dateFormat = dateFormatMap[dateStyle];
        opts = Object.assign(options, getFormattingOptions(dateFormat));
    }
    return {
        format: (value) => {
            const formatter = getDateTimeFormat(locale, opts);
            return formatter.format(toDate(value));
        },
    };
}
