// lightning-base-components/src/lightning/utilsPrivate/eventEmitter.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
export class EventEmitter {
    registry;
    constructor() {
        this.registry = {};
    }
    on(name, listener) {
        this.registry[name] = this.registry[name] || [];
        this.registry[name].push(listener);
        return this;
    }
    once(name, listener) {
        const doOnce = (...args) => {
            listener(...args);
            this.removeListener(name, doOnce);
        };
        this.on(name, doOnce);
        return this;
    }
    emit(name, ...args) {
        const listeners = this.registry[name];
        let count = 0;
        if (listeners) {
            for (let i = 0, { length } = listeners; i < length; i += 1) {
                count += 1;
                listeners[i](...args);
            }
        }
        return count > 0;
    }
    removeListener(name, listener) {
        const listeners = this.registry[name];
        if (listeners) {
            for (let i = 0, len = listeners.length; i < len; i += 1) {
                if (listeners[i] === listener) {
                    listeners.splice(i, 1);
                    return this;
                }
            }
        }
        return this;
    }
}
