// lightning-base-components/src/lightning/utilsPrivate/formatUtils.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { isValidDate, isValidLatitude, isValidLongitude, } from './validationUtils';
export function toDateTimeFormatOptions(instance) {
    const options = {
        weekday: instance.weekday,
        era: instance.era,
        year: instance.year,
        month: instance.month,
        day: instance.day,
        hour: instance.hour,
        hourCycle: undefined,
        hour12: undefined,
        minute: instance.minute,
        second: instance.second,
        timeZoneName: instance.timeZoneName,
        timeZone: instance.timeZone,
    };
    const { hour12 } = instance;
    if (instance._hour12Set) {
        if (hour12 === false) {
            options.hourCycle = 'h23';
        }
        else {
            options.hour12 = hour12;
        }
    }
    return options;
}
export function toFormattedDate(instance, dateTimeFormat) {
    const { value } = instance;
    if (isValidDate(value)) {
        const formatted = dateTimeFormat(toDateTimeFormatOptions(instance), instance.dateStyle).format(value);
        if (formatted) {
            return formatted;
        }
    }
    if (value) {
        console.warn(`<lightning-formatted-date-time> The value attribute accepts either a Date object, a timestamp, or a valid ISO8601 formatted string ` +
            `with timezone offset. but we are getting the ${typeof value} value "${value}" instead.`);
    }
    return '';
}
export function toFormattedLocation(instance) {
    const { latitude, longitude } = instance;
    if (isValidLatitude(latitude) && isValidLongitude(longitude)) {
        return `${latitude}, ${longitude}`;
    }
    console.warn(`<lightning-formatted-location> expects latitude in range [-90.0, 90.0], longitude in range [-180.0, 180.0].`);
    return '';
}
export function toFormattedNumber(instance, numberFormat) {
    const { value } = instance;
    if (value !== null &&
        value !== undefined &&
        value !== '' &&
        isFinite(value)) {
        const { formatStyle } = instance;
        const options = {
            style: formatStyle,
            currency: instance.currencyCode,
            currencyDisplay: instance.currencyDisplayAs,
            minimumIntegerDigits: instance.minimumIntegerDigits,
            minimumFractionDigits: instance.minimumFractionDigits,
            maximumFractionDigits: instance.maximumFractionDigits,
            minimumSignificantDigits: instance.minimumSignificantDigits,
            maximumSignificantDigits: instance.maximumSignificantDigits,
        };
        const valueAsString = String(value);
        let valueToFormat = valueAsString;
        if (formatStyle === 'percent-fixed') {
            options.style = 'percent';
            valueToFormat = parseFloat(value) / 100;
            if (valueAsString.indexOf('.') > 0 &&
                valueAsString.indexOf('e') < 0) {
                valueToFormat = valueToFormat.toFixed(valueAsString.split('.')[1].length + 2);
            }
        }
        return numberFormat(options).format(valueToFormat);
    }
    return '';
}
