// lightning-base-components/src/lightning/utilsPrivate/utilsPrivate.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import formFactor from '@salesforce/client/formFactor';
export * from './language';
export { assert } from './assert';
export { ARIA, ARIA_TO_CAMEL, isAriaDescriptionSupported, updateAriaInvalidOnElement, computeAriaInvalid, } from './aria';
export { EventEmitter } from './eventEmitter';
export { toNorthAmericanPhoneNumber } from './phonify';
export { sanitizeURL, FALLBACK_URL } from './url';
export { deepCopy, arraysEqual, ArraySlice } from './utility';
export { guid } from './guid';
export { classListMutation } from './classListMutation';
export { classSetToString } from './classSet';
export { makeEverythingExceptElementInert, restoreInertness } from './inert';
export { hasAnimation } from './animation';
export { normalizeBoolean, normalizeNumber, normalizeString, normalizeArray, normalizeAriaAttribute, } from './normalize';
export { keyCodes, runActionOnBufferedTypedCharacters, normalizeKeyValue, isShiftMetaOrControlKey, } from './keyboard';
export { raf } from './scroll';
export { isChrome, isIE11, isSafari } from './browser';
export { observePosition } from './observers';
export { hasOnlyAllowedVideoIframes } from './videoUtils';
export { parseToFormattedLinkifiedParts, parseToFormattedParts, } from './linkify';
export { isValidPageReference } from './pageReference';
export { isMacOS, isWindowsOS, isiOS, isAndroidOS } from './os';
export { VALID_HEADING_LEVELS, isHeadingLevelValid, } from './ariaLevelHeadingUtils';
export { toDateTimeFormatOptions, toFormattedDate, toFormattedLocation, toFormattedNumber, } from './formatUtils';
export { isTextIgnoreRTL } from './textUtils';
export { getActiveTooltip, hasActiveTooltip, setActiveTooltip, } from './tooltip';
export { isValidDate, isValidLatitude, isValidLongitude, isValidPhone, } from './validationUtils';
const LIGHTNING_TAG_REGEXP = /^LIGHTNING/i;
const LIGHTNING_DASH_NAME_REGEXP = /-\w/g;
function dashWordCharReplacement(dashWordCharMatch) {
    return dashWordCharMatch[1].toUpperCase();
}
export const isProductionEnv = () => process.env.NODE_ENV === 'production';
function synchronizeHTMLElementAttrs(element, attrs) {
    const attrNames = Object.keys(attrs);
    for (let i = 0, { length } = attrNames; i < length; i += 1) {
        const attrName = attrNames[i];
        const attrValue = attrs[attrName];
        if (attrValue) {
            element.setAttribute(attrName, attrValue);
        }
        else {
            element.removeAttribute(attrName);
        }
    }
}
function synchronizeLightningElementAttrs(element, attrs) {
    const attrNames = Object.keys(attrs);
    for (let i = 0, { length } = attrNames; i < length; i += 1) {
        const attrName = attrNames[i];
        const attrValue = attrs[attrName];
        const normalizedName = attrName.replace(LIGHTNING_DASH_NAME_REGEXP, dashWordCharReplacement);
        element[normalizedName] = attrValue || null;
    }
}
export function synchronizeAttrs(element, attrs) {
    if (element) {
        if (LIGHTNING_TAG_REGEXP.test(element.tagName)) {
            synchronizeLightningElementAttrs(element, attrs);
        }
        else {
            synchronizeHTMLElementAttrs(element, attrs);
        }
    }
}
export function reflectAttribute(element, attrName, value) {
    if (!element) {
        return;
    }
    if (typeof value === 'string') {
        element.setAttribute(attrName, value);
    }
    else if (value === true) {
        element.setAttribute(attrName, '');
    }
    else if (!value) {
        element.removeAttribute(attrName);
    }
    else {
        console.warn(`Invalid attribute value for "${attrName}": ${value}`);
    }
}
export function valueSetter(element, validNodeCtorArr = [], value) {
    if (!element || typeof element.tagName !== 'string') {
        return;
    }
    const isCustomElementTagName = (tagName) => /^[a-z][-_.\w]*-[-.0-9_a-z\xB7\xC0-\xD6\xD8-\xF6\xF8-\u37D0\u037F-\u1FFF\u200C-\u200D\u203F-\u2040\u2070-\u218F\u2C00-\u2FEF\u3001-\uD7FF\uF900-\uFDCF\uFDF0-\uFFFD\u10000-\uEFFFF]*$/.test(tagName);
    if (isCustomElementTagName(element.tagName.toLowerCase())) {
        element.value = value;
        return;
    }
    if (isCustomElementTagName(element.tagName.toLowerCase())) {
        element.value = value;
        return;
    }
    function isSetterOverwritten(obj, prop) {
        const originalDescriptor = Object.getOwnPropertyDescriptor(Object.getPrototypeOf(obj), prop);
        const currentDescriptor = Object.getOwnPropertyDescriptor(obj, prop);
        return (currentDescriptor &&
            currentDescriptor.set !== originalDescriptor?.set);
    }
    for (const validCtor of validNodeCtorArr) {
        if (element instanceof validCtor &&
            prototypeMatches(element, validCtor) &&
            !isSetterOverwritten(element, 'value')) {
            element.value = value;
            break;
        }
    }
}
export function prototypeMatches(obj, ctor) {
    return Object.getPrototypeOf(obj) === ctor.prototype;
}
export function getRealDOMId(el) {
    if (typeof el === 'string') {
        return el.length > 0 ? el : null;
    }
    return typeof el === 'object' && el !== null ? el.getAttribute('id') : null;
}
export function getShadowActiveElement() {
    let activeElement = document.activeElement;
    while (activeElement?.shadowRoot &&
        activeElement.shadowRoot.activeElement) {
        activeElement = activeElement.shadowRoot.activeElement;
    }
    return activeElement;
}
export function getShadowActiveElements() {
    let activeElement = document.activeElement;
    const shadowActiveElements = [];
    while (activeElement &&
        activeElement.shadowRoot &&
        activeElement.shadowRoot.activeElement) {
        shadowActiveElements.push(activeElement);
        activeElement = activeElement.shadowRoot.activeElement;
    }
    if (activeElement) {
        shadowActiveElements.push(activeElement);
    }
    return shadowActiveElements;
}
export function isRTL() {
    return !import.meta.env.SSR && document.dir === 'rtl';
}
export function isUndefinedOrNull(value) {
    return value === null || value === undefined;
}
export function isNotUndefinedOrNull(value) {
    return !isUndefinedOrNull(value);
}
const DEFAULT_MODAL_ZINDEX = 9000;
const DEFAULT_ZINDEX_OFFSET = 100;
const DEFAULT_ZINDEX_BASELINE = DEFAULT_MODAL_ZINDEX + DEFAULT_ZINDEX_OFFSET;
export function getZIndexBaseline() {
    const modalZindexValueLwc = !import.meta.env.SSR
        ? (window.getComputedStyle(document.documentElement) ||
            document.documentElement.style).getPropertyValue('--lwc-zIndexModal')
        : DEFAULT_MODAL_ZINDEX.toString();
    const baseZindexModalLwc = parseInt(modalZindexValueLwc, 10);
    return isNaN(baseZindexModalLwc)
        ? DEFAULT_ZINDEX_BASELINE
        : baseZindexModalLwc + DEFAULT_ZINDEX_OFFSET;
}
export function timeout(interval) {
    return new Promise((resolve) => {
        setTimeout(resolve, interval);
    });
}
export function animationFrame() {
    return new Promise((resolve) => {
        window.requestAnimationFrame(resolve);
    });
}
export function decorateInputForDragon(element) {
    const valuePropertyDescriptor = getInputValuePropertyDescriptor(element);
    Object.defineProperty(element, 'value', {
        set(value) {
            valuePropertyDescriptor?.set?.call(this, value);
            this.dispatchEvent(new CustomEvent('input'));
        },
        get: valuePropertyDescriptor.get,
        enumerable: true,
        configurable: true,
    });
}
function getInputValuePropertyDescriptor(element) {
    return Object.getOwnPropertyDescriptor(Object.getPrototypeOf(element), 'value');
}
export function setDecoratedDragonInputValueWithoutEvent(element, value) {
    const valuePropertyDescriptor = getInputValuePropertyDescriptor(element);
    valuePropertyDescriptor.set.call(element, value);
}
export function escapeHTML(html) {
    return html
        .replace(/&/g, '&amp;')
        .replace(/</g, '&lt;')
        .replace(/>/g, '&gt;')
        .replace(/"/g, '&quot;')
        .replace(/'/g, '&#039;');
}
export const BUTTON_GROUP_ORDER = {
    FIRST: 'first',
    MIDDLE: 'middle',
    LAST: 'last',
    ONLY: 'only',
};
export function buttonGroupOrderClass(groupOrder) {
    return {
        [BUTTON_GROUP_ORDER.FIRST]: 'slds-button_first',
        [BUTTON_GROUP_ORDER.MIDDLE]: 'slds-button_middle',
        [BUTTON_GROUP_ORDER.LAST]: 'slds-button_last',
        [BUTTON_GROUP_ORDER.ONLY]: 'single-button',
    }[groupOrder];
}
export function isNativeComponent(cmp) {
    return cmp?.template && !cmp.template.synthetic;
}
export function isBubblingFocusEvent(event) {
    const focusEvents = ['focus', 'blur'];
    return event instanceof FocusEvent && focusEvents.includes(event.type);
}
export function isDesktopBrowser() {
    return formFactor === 'Large';
}
const { hasOwn: ObjectHasOwn } = Object;
const { hasOwnProperty: ObjectProtoHasOwnProperty } = Object.prototype;
export const hasOwn = typeof ObjectHasOwn === 'function'
    ? ObjectHasOwn
    : function hasOwn(object, key) {
        return ObjectProtoHasOwnProperty.call(object, key);
    };
export function hasOwnProperties(object) {
    for (const key in object) {
        if (hasOwn(object, key)) {
            return true;
        }
    }
    return false;
}
export { default as Observable } from './observable';
