// vite-plugin-lwc/dist/index.mjs
import { createFilter } from 'vite';
import lwc from '@lwc/rollup-plugin';
import path from 'node:path';

function patchPlugins(patches) {
  return {
    name: "lwc:patch-plugins",
    enforce: "pre",
    configResolved(config) {
      for (const [name, patch] of Object.entries(patches)) {
        const plugin = config.plugins.find((plugin2) => plugin2.name === name);
        if (!plugin) {
          throw new Error(`Could not find plugin "${name}" to patch.`);
        }
        patch(plugin);
      }
    }
  };
}

function createRollupPlugin(options) {
  const plugin = lwc(options);
  const buildStart = plugin.buildStart ? "handler" in plugin.buildStart ? plugin.buildStart.handler : plugin.buildStart : () => {
  };
  const resolveId = plugin.resolveId ? "handler" in plugin.resolveId ? plugin.resolveId.handler : plugin.resolveId : () => {
  };
  const load = plugin.load ? "handler" in plugin.load ? plugin.load.handler : plugin.load : () => {
  };
  const transform = plugin.transform ? "handler" in plugin.transform ? plugin.transform.handler : plugin.transform : () => {
  };
  return {
    version: plugin.version,
    buildStart,
    resolveId,
    load,
    transform
  };
}
function lwcVite(config) {
  const csr = createRollupPlugin(config);
  const ssr = createRollupPlugin({ ...config, targetSSR: true });
  const exclude = [config.exclude].flat().filter((e) => e !== void 0 && e !== null);
  const filter = createFilter(config.include, [
    "**/vite/**",
    "**/@vitest/**",
    "**/.vite/**",
    "index.html",
    "/__vitest_test__/**",
    ...exclude
  ]);
  return {
    name: "lwc:vite-plugin",
    config() {
      return {
        define: {
          "process.env.SKIP_LWC_VERSION_MISMATCH_CHECK": "false"
        }
      };
    },
    async buildStart(options) {
      try {
        await csr.buildStart.call(this, options);
        await ssr.buildStart.call(this, options);
      } catch (e) {
        this.error(getError(e));
      }
    },
    async resolveId(source, importer, options) {
      if (!filter(source)) {
        return;
      }
      if (importer && path.extname(importer) === ".html" && path.isAbsolute(importer) && path.extname(source) !== "" && path.isAbsolute(source)) {
        const dir = path.dirname(importer);
        return path.join(dir, source);
      }
      try {
        const id = await (options.ssr ? ssr : csr).resolveId.call(
          this,
          source,
          importer,
          options
        );
        if (!id) {
          return;
        }
        return id;
      } catch (e) {
        this.error(getError(e, source));
      }
    },
    load(id, options) {
      if (!filter(id)) {
        return;
      }
      try {
        return (options?.ssr ? ssr : csr).load.call(this, id);
      } catch (e) {
        this.error(getError(e, id));
      }
    },
    transform(code, id, options) {
      if (!filter(id)) {
        return;
      }
      try {
        return (options?.ssr ? ssr : csr).transform.call(
          this,
          code,
          id
        );
      } catch (e) {
        this.error(getError(e, id, code));
      }
    }
  };
}
function getError(error, id, src) {
  if (typeof error === "string") {
    return error;
  }
  if (typeof error !== "object" || error === null) {
    return String(error);
  }
  const rollupError = {
    message: "An unknown error occurred."
  };
  addErrorCode(error, rollupError);
  addErrorMessage(error, rollupError);
  addErrorLocation(error, rollupError, src);
  addErrorId(id, rollupError);
  return rollupError;
}
function addErrorCode(error, rollupError) {
  if ("code" in error && typeof error.code === "number") {
    rollupError.pluginCode = error.code;
  }
}
function addErrorMessage(error, rollupError) {
  if ("message" in error && typeof error.message === "string") {
    rollupError.message = error.message;
  }
}
function addErrorLocation(error, rollupError, src) {
  if ("filename" in error && typeof error.filename === "string") {
    rollupError.loc = {
      file: error.filename,
      line: 1,
      column: 1
    };
  }
  if ("location" in error && typeof error.location === "object" && error.location !== null) {
    rollupError.loc = {
      ...rollupError.loc,
      line: "line" in error.location && typeof error.location.line === "number" ? error.location.line : 1,
      column: "column" in error.location && typeof error.location.column === "number" ? error.location.column : 1
    };
    if ("start" in error.location && typeof error.location.start === "number" && "length" in error.location && typeof error.location.length === "number") {
      rollupError.frame = src?.substring(
        error.location.start,
        error.location.start + error.location.length
      );
    }
  }
}
function addErrorId(id, rollupError) {
  if (typeof id === "string") {
    rollupError.id = id;
  }
}

const alias = [
  {
    // Find all html imports and add ?import to the end of the import
    find: /^(?!.*index)(.*)\.html$/,
    replacement: "$1.html?import"
  }
];
function lwcAliasPlugin() {
  return {
    name: "lwc:alias",
    enforce: "pre",
    apply: "build",
    config() {
      return {
        resolve: {
          alias
        }
      };
    }
  };
}

function lwcHmrPlugin() {
  return {
    name: "lwc:hmr",
    enforce: "post",
    apply: "serve",
    transform(code, id, options) {
      if (options?.ssr) {
        return;
      }
      if (!id) {
        return;
      }
      if (!code.includes("export default __lwc_component_class_internal")) {
        return;
      }
      return code + `
if (import.meta.hot) {
  const { swapComponent, isComponentConstructor } = await import('lwc');
    import.meta.hot.accept((jsModule) => {
    if (!jsModule) {
      import.meta.hot?.invalidate("new module is undefined");
      return;
    }

    if (!isComponentConstructor(jsModule.default)) {
      import.meta.hot?.invalidate("not a component constructor");
      return;
    }

    const success = swapComponent(__lwc_component_class_internal, jsModule.default);

    if (!success) {
      import.meta.hot?.invalidate("swapComponent returned false")
      return;
    }
  })
}`;
    }
  };
}

function normalizeOptions(rawConfig) {
  const config = rawConfig;
  config.rootDir ??= ".";
  config.defaultModules ??= [];
  return config;
}

const index = (options = {}) => {
  options = normalizeOptions(options);
  return [
    patchPlugins({
      "vite:css": (p) => {
        p.transform = void 0;
      },
      "vite:css-post": (p) => {
        p.transform = void 0;
      }
    }),
    lwcAliasPlugin(),
    {
      ...lwcVite(options),
      apply: "build"
    },
    {
      ...lwcVite(options),
      enforce: "post",
      apply: "serve"
    },
    lwcHmrPlugin()
  ];
};

export { index as default };
