// lightning-base-components/src/lightning/treeGrid/__examples__/lazyLoading/lazyLoading.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { LightningElement } from 'lwc';
import { EXAMPLES_COLUMNS_DEFINITION_BASIC, EXAMPLES_DATA_LAZY_LOADING, } from './sampleData';
export default class TreeGridLazyLoading extends LightningElement {
    gridLoadingState = false;
    gridColumns = EXAMPLES_COLUMNS_DEFINITION_BASIC;
    gridData = EXAMPLES_DATA_LAZY_LOADING;
    gridExpandedRows = ['123556', '123556-B', '123557'];
    handleRowToggle(event) {
        const rowName = event.detail.name;
        const hasChildrenContent = event.detail.hasChildrenContent;
        if (hasChildrenContent === false) {
            this.gridLoadingState = true;
            this.retrieveUpdatedData(rowName).then((newData) => {
                this.gridData = newData;
                this.gridLoadingState = false;
            });
        }
    }
    retrieveUpdatedData(rowName) {
        return new Promise((resolve) => {
            window.setTimeout(() => {
                const updatedData = this.addChildrenToRow(this.gridData, rowName, this.childrenData[rowName]);
                resolve(updatedData);
            }, 2000);
        });
    }
    addChildrenToRow(data, rowName, children) {
        const newData = data.map((row) => {
            let hasChildrenContent = false;
            if (Object.prototype.hasOwnProperty.call(row, '_children') &&
                Array.isArray(row._children) &&
                row._children.length > 0) {
                hasChildrenContent = true;
            }
            if (row.name === rowName) {
                row._children = children;
            }
            else if (hasChildrenContent) {
                this.addChildrenToRow(row._children, rowName, children);
            }
            return row;
        });
        return newData;
    }
    childrenData = {
        '123556-A': [
            {
                name: '123556-A-A',
                accountName: 'Acme Corporation (Oakland)',
                employees: 745,
                phone: '837-555-1212',
                accountOwner: 'http://example.com/john-doe',
                accountOwnerName: 'John Doe',
                billingCity: 'New York, NY',
            },
            {
                name: '123556-A-B',
                accountName: 'Acme Corporation (San Francisco)',
                employees: 578,
                phone: '837-555-1212',
                accountOwner: 'http://example.com/jane-doe',
                accountOwnerName: 'Jane Doe',
                billingCity: 'Los Angeles, CA',
            },
        ],
        '123556-B-B': [
            {
                name: '123556-B-B-A',
                accountName: 'Allied Technologies',
                employees: 390,
                phone: '837-555-1212',
                accountOwner: 'http://example.com/jane-doe',
                accountOwnerName: 'Jane Doe',
                billingCity: 'Los Angeles, CA',
                _children: [
                    {
                        name: '123556-B-B-A-A',
                        accountName: 'Allied Technologies (UV)',
                        employees: 270,
                        phone: '837-555-1212',
                        accountOwner: 'http://example.com/john-doe',
                        accountOwnerName: 'John Doe',
                        billingCity: 'San Francisco, CA',
                    },
                ],
            },
        ],
    };
}
