// lightning-base-components/src/lightning/menuItem/menuItem.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { api, track } from 'lwc';
import { features } from 'lightning/features';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import { classSet } from 'lightning/utils';
import { normalizeBoolean, normalizeString, keyCodes, sanitizeURL, FALLBACK_URL, } from 'lightning/utilsPrivate';
import { getCategory, isValidName, computeSldsClass } from 'lightning/iconUtils';
const selectedClass = 'slds-is-selected';
export default class LightningMenuItem extends LightningShadowBaseClass {
    static validationOptOut = ['class'];
    @api
    value;
    @api
    label;
    @api
    download;
    @track
    _accesskey = null;
    @track
    _disabled = false;
    @track
    _tabindex = '-1';
    @track
    _checked = undefined;
    @track
    _isDraft = false;
    @track
    _target = null;
    _iconName;
    suffixSVGClass;
    suffixIconVariant;
    _prefixIconName;
    prefixSVGClass;
    prefixIconVariant;
    _iconType = 'standard';
    @api
    href;
    @api
    draftAlternativeText;
    connectedCallback() {
        super.connectedCallback();
        this.setAttribute('role', 'presentation');
    }
    @api
    get iconName() {
        return this._iconName;
    }
    set iconName(newValue) {
        if (newValue !== this._iconName) {
            this._iconName = newValue;
            this.updateIconProperties('suffix', this._iconName);
        }
    }
    @api
    get prefixIconName() {
        return this._prefixIconName;
    }
    set prefixIconName(newValue) {
        if (newValue !== this._prefixIconName) {
            this._prefixIconName = newValue;
            this.updateIconProperties('prefix', this._prefixIconName);
        }
    }
    @api
    get iconType() {
        return this._iconType;
    }
    set iconType(newValue) {
        if (!features.ENABLE_COLOR_ICONS_MENU_ITEM) {
            return;
        }
        this._iconType = normalizeString(newValue, {
            fallbackValue: 'standard',
            validValues: ['standard', 'color'],
        });
        this.updateIconProperties('suffix', this._iconName);
        this.updateIconProperties('prefix', this._prefixIconName);
    }
    @api
    get isDraft() {
        return this._isDraft;
    }
    set isDraft(value) {
        this._isDraft = normalizeBoolean(value);
    }
    @api
    get accessKey() {
        return this._accesskey;
    }
    set accessKey(newValue) {
        this._accesskey = newValue;
        this.handleAccessKeyChange(newValue);
    }
    @api
    get tabIndex() {
        return this._tabindex;
    }
    set tabIndex(newValue) {
        this._tabindex = newValue;
        this.handleTabIndexChange(newValue);
    }
    @api
    get target() {
        return this._target;
    }
    set target(newValue) {
        this._target = normalizeString(newValue, {
            fallbackValue: null,
            validValues: ['_self', '_blank'],
        });
    }
    handleAccessKeyChange(value) {
        this._accesskey = value;
    }
    handleTabIndexChange(value) {
        this._tabindex = value;
    }
    get computedAccessKey() {
        return this._accesskey;
    }
    get computedTabIndex() {
        return this._tabindex;
    }
    @api
    get disabled() {
        return this._disabled;
    }
    set disabled(value) {
        this._disabled = normalizeBoolean(value);
    }
    @api
    get checked() {
        return this._checked;
    }
    set checked(value) {
        if (typeof value === 'string') {
            value = normalizeString(value, {
                fallbackValue: undefined,
                validValues: ['true', 'false'],
            });
            if (value !== undefined) {
                value = value === 'true';
            }
        }
        if (value !== undefined) {
            value = normalizeBoolean(value);
        }
        this._checked = value;
        if (this.checked === true) {
            this.classList.add(selectedClass);
        }
        else {
            this.classList.remove(selectedClass);
        }
    }
    get computedCheckedIconClass() {
        return classSet('slds-icon slds-icon_x-small slds-icon-text-default slds-m-right_x-small')
            .add({ 'slds-icon_selected': !this.checked })
            .toString();
    }
    updateIconProperties(type, iconName) {
        const category = getCategory(iconName);
        const isStandard = this._iconType === 'standard' ||
            category === 'utility' ||
            category === 'doctype';
        const isActionIcon = category === 'action';
        const svgClassSet = classSet('slds-icon').add({
            'slds-icon-text-default': isStandard,
            'slds-icon_container': !isStandard,
            'slds-p-around_xxx-small slds-icon_container_circle': !isStandard && isActionIcon,
        });
        if (!isStandard && isValidName(iconName)) {
            svgClassSet.add(computeSldsClass(iconName));
        }
        if (type === 'suffix') {
            svgClassSet.add('slds-shrink-none');
        }
        this[`${type}SVGClass`] = svgClassSet.toString();
        this[`${type}IconVariant`] = isStandard ? 'bare' : 'inverse';
    }
    get computedHref() {
        return sanitizeURL(this.href);
    }
    get computedAriaChecked() {
        return this.isMenuItemCheckbox ? this.checked + '' : null;
    }
    get computedAriaDisabled() {
        return this.disabled ? 'true' : 'false';
    }
    get isMenuItemCheckbox() {
        return this.checked !== undefined;
    }
    get computedRole() {
        return this.isMenuItemCheckbox ? 'menuitemcheckbox' : 'menuitem';
    }
    handleBlur(event) {
        this.dispatchEvent(new CustomEvent('blur'));
        this.dispatchEvent(new CustomEvent('privateblur', {
            composed: true,
            bubbles: true,
            cancelable: true,
            detail: {
                relatedTarget: event.relatedTarget,
            },
        }));
    }
    handleFocus() {
        this.dispatchEvent(new CustomEvent('privatefocus', {
            bubbles: true,
            cancelable: true,
        }));
    }
    handleClick(event) {
        if (this.disabled) {
            event.preventDefault();
            event.stopPropagation();
            return;
        }
        if (this.href) {
            if (this.computedHref === FALLBACK_URL) {
                event.preventDefault();
            }
            else if (this.value) {
                const cancelled = !this.dispatchSelect();
                if (cancelled) {
                    event.preventDefault();
                }
            }
        }
        else {
            event.preventDefault();
            this.dispatchSelect();
        }
    }
    handleKeyDown(event) {
        if (this.disabled) {
            event.preventDefault();
            return;
        }
        if (event.keyCode === keyCodes.space) {
            if (this.href) {
                if (this.value) {
                    const cancelled = !this.dispatchSelect();
                    if (!cancelled) {
                        this.template.querySelector('a').click();
                    }
                }
                else {
                    this.template.querySelector('a').click();
                }
            }
            else {
                this.dispatchSelect();
            }
        }
    }
    dispatchSelect() {
        if (!this.disabled) {
            return this.dispatchEvent(new CustomEvent('privateselect', {
                bubbles: true,
                cancelable: true,
                detail: {
                    value: this.value,
                },
            }));
        }
        return true;
    }
    @api
    focus() {
        this.template.querySelector('a').focus();
        this.dispatchEvent(new CustomEvent('focus'));
    }
    @api
    click() {
        const anchor = this.template.querySelector('a');
        if (anchor) {
            anchor.click();
        }
    }
}
