// lightning-base-components/src/lightning/touchScrollLibrary/touchScrollLibrary.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { assert } from 'lightning/utilsPrivate';
import formFactor from '@salesforce/client/formFactor';
function getPositionY(event) {
    return event.touches ? event.touches[0].screenY : event.screenY;
}
function canScroll(el) {
    const canScrollY = el.scrollHeight > el.offsetHeight;
    const canScrollX = el.scrollWidth > el.offsetWidth;
    return canScrollY || canScrollX;
}
function alwaysScrolls(el) {
    const isInputRange = el.tagName === 'INPUT' && el.type === 'range';
    return isInputRange;
}
function nativeScroll(el) {
    const isTextarea = el.tagName === 'TEXTAREA';
    return isTextarea;
}
function isAtTop(el) {
    return el.scrollTop === 0;
}
function isAtBottom(el) {
    return (el.scrollHeight - el.scrollTop === el.offsetHeight ||
        el.scrollHeight - el.scrollTop === el.clientHeight);
}
function skipUiScroller(event) {
    event.cancelScrolling = true;
    event.preventBounce = false;
}
function enableUiScroller(event) {
    event.cancelScrolling = false;
    event.preventBounce = true;
}
export class TouchScroller {
    _initialized = false;
    _startY = null;
    _target;
    static isMobile() {
        return formFactor !== 'Large';
    }
    constructor(target) {
        assert(!!target, 'a non-empty target is required for TouchScroller to add touch listeners to');
        this._target = target;
        this.initialize();
    }
    initialize() {
        if (!this._initialized && this._target) {
            if (TouchScroller.isMobile()) {
                this._target.addEventListener('touchstart', (event) => this.handleTouchStart(event));
                this._target.addEventListener('touchmove', (event) => this.handleTouchMove(event));
            }
            this._initialized = true;
        }
    }
    handleTouchStart(event) {
        this._startY = getPositionY(event);
    }
    handleTouchMove(event) {
        const el = event.currentTarget;
        const target = event.target;
        if (alwaysScrolls(target)) {
            skipUiScroller(event);
            return;
        }
        const canScrollCurrent = canScroll(el);
        const canScrollTarget = nativeScroll(target) && canScroll(target);
        if (canScrollTarget || canScrollCurrent) {
            const curY = getPositionY(event);
            let atTop = this._startY !== null && this._startY < curY && isAtTop(el);
            let atBottom = this._startY !== null && this._startY > curY && isAtBottom(el);
            if (canScrollTarget && !event.preventBounce) {
                atTop = atTop && isAtTop(target);
                atBottom = atBottom && isAtBottom(target);
            }
            if (!atTop && !atBottom) {
                skipUiScroller(event);
            }
            else {
                enableUiScroller(event);
            }
        }
    }
}
