// lightning-base-components/src/lightning/helptext/helptext.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import labelButtonAlternativeText from '@salesforce/label/LightningHelptext.buttonAlternativeText';
import { api, track } from 'lwc';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import { isValidName } from 'lightning/iconUtils';
import { normalizeString } from 'lightning/utilsPrivate';
const DEFAULT_BUTTON_ALT_TEXT = labelButtonAlternativeText;
const DEFAULT_ICON_NAME = 'utility:info';
const DEFAULT_ICON_VARIANT = 'bare';
export default class LightningHelptext extends LightningShadowBaseClass {
    @track
    state = {
        iconName: DEFAULT_ICON_NAME,
        iconVariant: DEFAULT_ICON_VARIANT,
        alternativeText: DEFAULT_BUTTON_ALT_TEXT,
    };
    @api
    content;
    @api
    tabIndex;
    @api
    get iconName() {
        if (isValidName(this.state.iconName)) {
            return this.state.iconName;
        }
        return DEFAULT_ICON_NAME;
    }
    set iconName(value) {
        this.state.iconName = value;
    }
    @api
    get iconVariant() {
        return normalizeString(this.state.iconVariant, {
            fallbackValue: DEFAULT_ICON_VARIANT,
            validValues: ['bare', 'error', 'inverse', 'warning'],
        });
    }
    set iconVariant(value) {
        this.state.iconVariant = value;
    }
    @api
    get alternativeText() {
        return this.state.alternativeText;
    }
    set alternativeText(value) {
        if (value && typeof value === 'string' && value.trim() !== '') {
            this.state.alternativeText = value;
        }
        else {
            console.warn(`<lightning-helptext> Invalid alternativeText value: ${value}`);
        }
    }
    @api
    focus() {
        this.template.querySelector('lightning-button-icon').focus();
    }
    get computedButtonVariant() {
        switch (this.iconVariant) {
            case 'inverse':
                return 'bare-inverse';
            default:
                return 'bare';
        }
    }
    get computedIconClass() {
        switch (this.iconVariant) {
            case 'error':
                return 'slds-icon-text-error';
            case 'warning':
                return 'slds-icon-text-warning';
            default:
                return '';
        }
    }
}
