// lightning-base-components/src/lightning/modal/modal.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { api } from 'lwc';
import LightningPrimitiveOverlay from 'lightning/primitiveOverlay';
import { parent, instanceName, secure } from 'lightning/overlayUtils';
import LightningModalBase from 'lightning/modalBase';
import { normalizeBoolean } from 'lightning/utilsPrivate';
export default class LightningModal extends LightningPrimitiveOverlay {
    static [parent] = LightningModalBase;
    static [instanceName] = 'lightning-modal';
    _disableClose = false;
    @api
    size = 'medium';
    @api
    label = '';
    @api
    description = '';
    @api
    get disableClose() {
        return this._disableClose;
    }
    set disableClose(value) {
        const currentDisableClose = this._disableClose;
        this._disableClose = normalizeBoolean(value);
        if (currentDisableClose !== this._disableClose) {
            this.dispatchEvent(new CustomEvent('privatedisableclosebutton', {
                detail: {
                    disableClose: this._disableClose,
                    [secure]: true,
                },
                bubbles: true,
            }));
        }
    }
    @api
    close(result) {
        if (!this.disableClose) {
            const promise = new Promise((resolve) => {
                this.dispatchEvent(new CustomEvent('privateclose', {
                    detail: {
                        resolve,
                        [secure]: true,
                    },
                    bubbles: true,
                }));
            });
            super.close(result, promise);
        }
        else {
            let errorMsg = 'LightningModal - Any interactions (buttons, processes, etc) that call modal.close() should be ';
            errorMsg += '(a) disabled, while disableClose api set true, and ';
            errorMsg += '(b) re-enabled, when disableClose set false';
            console.error(errorMsg);
        }
    }
    connectedCallback() {
        if (this.constructor.name === 'LightningModal') {
            throw new Error('Use of <lightning-modal> as an element is not supported. See developer documentation for extending the LightningModal class.');
        }
    }
}
