// lightning-base-components/src/lightning/menuDivider/menuDivider.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { api, track } from 'lwc';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import { classSet } from 'lightning/utils';
import { normalizeString as normalize } from 'lightning/utilsPrivate';
export default class LightningMenuDivider extends LightningShadowBaseClass {
    @track
    _variant = 'standard';
    connectedCallback() {
        super.connectedCallback();
        this.setAttribute('role', 'separator');
    }
    @api
    get variant() {
        return this._variant;
    }
    set variant(value) {
        this._variant = normalize(value, {
            fallbackValue: 'standard',
            validValues: ['standard', 'compact'],
        });
    }
    get computedClass() {
        return classSet({
            'slds-has-divider_top-space': this.variant === 'standard',
            'slds-has-divider_top': this.variant === 'compact',
        }).toString();
    }
}
