// lightning-base-components/src/lightning/modalUtils/modalUtils.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { getElementWithFocus, returnFocusToElement, } from 'lightning/focusUtils';
import { isLwcModalActive } from 'lightning/overlayManager';
import { secure } from 'lightning/overlayUtils';
const MODAL_SELECTORS = [
    'lightning-modal-header',
    'lightning-modal-body',
    'lightning-modal-footer',
    'lightning-focus-trap',
    'lightning-overlay-container',
];
export function modalFocusinEventHandler(e) {
    const { isModalClosing, lastFocusedElement } = this;
    if (!isLwcModalActive() || isModalClosing) {
        return;
    }
    let isTargetRelatedToModal = null;
    if (e.target && typeof e.target.closest === 'function') {
        isTargetRelatedToModal = e.target.closest(MODAL_SELECTORS.join(','));
    }
    else if (e.relatedTarget &&
        typeof e.relatedTarget.closest === 'function') {
        isTargetRelatedToModal = e.relatedTarget.closest(MODAL_SELECTORS.join(','));
    }
    if (isTargetRelatedToModal) {
        this.lastFocusedElement = getElementWithFocus();
        this.dispatchEvent(new CustomEvent('privatelightningmodallastfocus', {
            composed: true,
            bubbles: true,
            detail: {
                [secure]: true,
                privatelightningmodallastfocus: this.lastFocusedElement,
            },
        }));
    }
    else if (lastFocusedElement) {
        if (!lastFocusedElement.ariaHasPopup) {
            returnFocusToElement(lastFocusedElement);
        }
    }
}
