// lightning-base-components/src/lightning/primitiveButton/primitiveButton.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { api, track } from 'lwc';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import { isIE11, normalizeBoolean, normalizeString, buttonGroupOrderClass, } from 'lightning/utilsPrivate';
import { classSet } from 'lightning/utils';
import AriaObserver from 'lightning/ariaObserver';
export default class LightningPrimitiveButton extends LightningShadowBaseClass {
    _initialized = false;
    @track
    state = {
        ariaAtomic: null,
        ariaBusy: null,
        ariaControls: null,
        ariaDetails: null,
        ariaDescribedBy: null,
        ariaExpanded: null,
        ariaFlowTo: null,
        ariaHasPopup: null,
        ariaHidden: null,
        ariaLabelledBy: null,
        ariaLive: null,
        ariaOwns: null,
        ariaPressed: null,
        ariaRelevant: null,
        disabled: false,
    };
    ariaObserver;
    originalAriaAtomic;
    originalAriaBusy;
    originalAriaControls;
    originalAriaDetails;
    originalAriaDescribedBy;
    originalAriaExpanded;
    originalAriaFlowTo;
    originalAriaHasPopup;
    originalAriaHidden;
    originalAriaLabelledBy;
    originalAriaLive;
    originalAriaOwns;
    originalAriaPressed;
    originalAriaRelevant;
    originalDisabledValue;
    @api
    set disabled(value) {
        this.originalDisabledValue = value;
        this.state.disabled = normalizeBoolean(value);
    }
    get disabled() {
        return this.state.disabled;
    }
    @api
    accessKey;
    @api
    title;
    @api
    ariaLabel;
    @api
    set ariaLabelledBy(value) {
        this.originalAriaLabelledBy = value;
        this.ariaObserver.connect({
            targetSelector: 'button',
            attribute: 'aria-labelledby',
            relatedNodeIds: value,
        });
    }
    get ariaLabelledBy() {
        return this.originalAriaLabelledBy;
    }
    @api
    set ariaDescribedBy(value) {
        this.originalAriaDescribedBy = value;
        this.ariaObserver.connect({
            targetSelector: 'button',
            attribute: 'aria-describedby',
            relatedNodeIds: value,
        });
    }
    get ariaDescribedBy() {
        return this.originalAriaDescribedBy;
    }
    @api
    set ariaControls(value) {
        this.originalAriaControls = value;
        this.setAttribute('aria-controls', value);
    }
    get ariaControls() {
        return this.originalAriaControls;
    }
    @api
    set ariaOwns(value) {
        this.originalAriaOwns = value;
        this.setAttribute('aria-owns', value);
    }
    get ariaOwns() {
        return this.originalAriaOwns;
    }
    @api
    set ariaDetails(value) {
        this.originalAriaDetails = value;
        this.setAttribute('aria-details', value);
    }
    get ariaDetails() {
        return this.originalAriaDetails;
    }
    @api
    set ariaFlowTo(value) {
        this.originalAriaFlowTo = value;
        this.setAttribute('aria-flowto', value);
    }
    get ariaFlowTo() {
        return this.originalAriaFlowTo;
    }
    @api
    get ariaExpanded() {
        return this.originalAriaExpanded !== undefined
            ? String(this.originalAriaExpanded)
            : null;
    }
    set ariaExpanded(value) {
        this.originalAriaExpanded = value;
        if (typeof value === 'string') {
            this.state.ariaExpanded = normalizeString(value, {
                fallbackValue: null,
                validValues: ['true', 'false'],
            });
        }
        else if (typeof value === 'boolean') {
            this.state.ariaExpanded = normalizeBoolean(value);
        }
    }
    get computedAriaExpanded() {
        return this.state.ariaExpanded;
    }
    @api
    set ariaPressed(value) {
        this.originalAriaPressed = value;
        this.state.ariaPressed = normalizeString(value, {
            fallbackValue: null,
            validValues: ['true', 'false'],
        });
    }
    get ariaPressed() {
        return this.originalAriaPressed;
    }
    get computedAriaPressed() {
        return this.state.ariaPressed;
    }
    @api
    set ariaHidden(value) {
        this.originalAriaHidden = value;
        this.state.ariaHidden = normalizeString(value, {
            fallbackValue: null,
            validValues: ['true', 'false'],
        });
    }
    get ariaHidden() {
        return this.originalAriaHidden;
    }
    get computedAriaHidden() {
        return this.state.ariaHidden;
    }
    @api
    ariaCurrent;
    @api
    ariaKeyShortcuts;
    @api
    set ariaHasPopup(value) {
        this.originalAriaHasPopup = value;
        this.state.ariaHasPopup = normalizeString(value, {
            fallbackValue: null,
            validValues: ['true', 'dialog', 'menu', 'listbox', 'tree', 'grid'],
        });
    }
    get ariaHasPopup() {
        return this.originalAriaHasPopup;
    }
    get computedAriaHasPopup() {
        return this.state.ariaHasPopup;
    }
    @api
    set ariaRelevant(value) {
        this.originalAriaRelevant = value;
        this.state.ariaRelevant = normalizeString(value, {
            fallbackValue: null,
            validValues: ['additions', 'removals', 'text', 'all'],
        });
    }
    get ariaRelevant() {
        return this.originalAriaRelevant;
    }
    get computedAriaRelevant() {
        return this.state.ariaRelevant;
    }
    @api
    set ariaLive(value) {
        this.originalAriaLive = value;
        this.state.ariaLive = normalizeString(value, {
            fallbackValue: null,
            validValues: ['polite', 'assertive', 'off'],
        });
    }
    get ariaLive() {
        return this.originalAriaLive;
    }
    get computedAriaLive() {
        return this.state.ariaLive;
    }
    @api
    set ariaAtomic(value) {
        this.originalAriaAtomic = value;
        this.state.ariaAtomic = normalizeString(value, {
            fallbackValue: null,
            validValues: ['true', 'false'],
        });
    }
    get ariaAtomic() {
        return this.originalAriaAtomic;
    }
    get computedAriaAtomic() {
        return this.state.ariaAtomic;
    }
    @api
    set ariaBusy(value) {
        this.originalAriaBusy = value;
        this.state.ariaBusy = normalizeString(value, {
            fallbackValue: null,
            validValues: ['true', 'false'],
        });
    }
    get ariaBusy() {
        return this.originalAriaBusy;
    }
    get computedAriaBusy() {
        return this.state.ariaBusy;
    }
    get computedButtonClass() {
        const classes = classSet('slds-button');
        classes.add(buttonGroupOrderClass(this.groupOrder));
        return classes.toString();
    }
    @api
    focus() { }
    @api
    groupOrder = '';
    constructor() {
        super();
        this.ariaObserver = new AriaObserver(this);
        if (isIE11 && !import.meta.env.SSR) {
            this.template.addEventListener('click', (event) => {
                if (this.disabled) {
                    event.stopImmediatePropagation();
                }
            });
        }
    }
    connectedCallback() {
        super.connectedCallback();
        if (!this.ariaObserver) {
            this.ariaObserver = new AriaObserver(this);
        }
    }
    renderedCallback() {
        if (this.isConnected) {
            this.ariaObserver.sync();
        }
    }
    disconnectedCallback() {
        if (this.ariaObserver) {
            this.ariaObserver.disconnect();
            this.ariaObserver = undefined;
        }
    }
}
