// lightning-base-components/src/lightning/context/context.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { createContextProvider, readonly } from 'lwc';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import { createV2ContextAdapter } from './createContextWireAdapter';
const ContextProviderMetaMap = new Map();
const ContextDataMap = new WeakMap();
function getWireAdapterMeta(ctor) {
    let adapterMeta = ContextProviderMetaMap.get(ctor);
    if (adapterMeta === undefined) {
        const wireAdapter = createV2ContextAdapter(ctor);
        const contextualizer = createContextProvider(wireAdapter);
        adapterMeta = {
            wireAdapter,
            contextualizer,
        };
        ContextProviderMetaMap.set(ctor, adapterMeta);
    }
    return adapterMeta;
}
function getContextData(eventTarget) {
    let contextData = ContextDataMap.get(eventTarget);
    if (contextData === undefined) {
        contextData = {
            consumers: [],
            value: null,
        };
        ContextDataMap.set(eventTarget, contextData);
    }
    return contextData;
}
function setupNewContextProvider(target) {
    const { contextualizer } = getWireAdapterMeta(target.constructor);
    contextualizer(target, {
        consumerConnectedCallback(consumer) {
            const contextData = getContextData(target);
            contextData.consumers.push(consumer);
            consumer.provide({ value: contextData.value });
        },
        consumerDisconnectedCallback(consumer) {
            const contextData = getContextData(target);
            const i = contextData.consumers.indexOf(consumer);
            if (i >= 0) {
                contextData.consumers.splice(i, 1);
            }
            else {
                throw new TypeError(`Invalid context operation in ${target}.`);
            }
        },
    });
}
function emitNewContextValue(target, newValue) {
    const contextData = getContextData(target);
    contextData.value = readonly(newValue);
    contextData.consumers.forEach((consumer) => consumer.provide({ value: readonly(newValue) }));
}
export default class LightningContext extends LightningShadowBaseClass {
    constructor() {
        super();
        setupNewContextProvider(this);
    }
    setContext(newValue) {
        emitNewContextValue(this, newValue);
    }
    static get Provider() {
        return getWireAdapterMeta(this).wireAdapter;
    }
    static getDefaultContext() {
        return undefined;
    }
}
