// lightning-base-components/src/lightning/numberUtils/numberUtils.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import groupingSeparator from '@salesforce/i18n/number.groupingSeparator';
import decimalSeparator from '@salesforce/i18n/number.decimalSeparator';
import { fromLocalizedDigits, toLocalizedDigits, numberFormat, } from 'lightning/internationalizationLibrary';
import { isEmptyString } from 'lightning/inputUtils';
import { normalizeNumber } from 'lightning/utilsPrivate';
const VALID_NUMBER_CHARACTERS_EXPRESSION = new RegExp('^[-+0-9kKmMbBtTeE.,\u0660-\u0669\u06F0-\u06F9\u07C0-\u07C9' +
    '\u0966-\u096F\u09E6-\u09EF\u0A66-\u0A6F\u0AE6-\u0AEF' +
    '\u0B66-\u0B6F\u0BE6-\u0BEF\u0C66-\u0C6F\u0CE6-\u0CEF' +
    '\u0D66-\u0D6F\u0DE6-\u0DEF\u0E50-\u0E59\u0ED0-\u0ED9' +
    '\u0F20-\u0F29\u1040-\u1049\u1090-\u1099\u17E0-\u17E9' +
    '\u1810-\u1819\u1946-\u194F\u19D0-\u19D9\u1A80-\u1A89' +
    '\u1A90-\u1A99\u1B50-\u1B59\u1BB0-\u1BB9\u1C40-\u1C49' +
    '\u1C50-\u1C59\uA620-\uA629\uA8D0-\uA8D9\uA900-\uA909' +
    '\uA9D0-\uA9D9\uA9F0-\uA9F9\uAA50-\uAA59\uABF0-\uABF9' +
    ']$');
const SHORTCUT_FACTORS = {
    k: 3,
    m: 6,
    b: 9,
    t: 12,
};
const SHORTCUTS = ['k', 'm', 'b', 't'];
const NUMBER_SYMBOLS = ['+', '-'];
export function toIsoDecimal(numberAsString) {
    const result = transformLocalizedNumberToIsoDecimal(numberAsString);
    if (isNaN(result)) {
        return '';
    }
    return result;
}
export function isValidNumber(numberAsString) {
    return !isNaN(transformLocalizedNumberToIsoDecimal(numberAsString));
}
export function fromIsoDecimal(numberAsString) {
    return toLocalizedDigits(numberAsString.replace('.', decimalSeparator));
}
export function increaseNumberByStep({ value, increment, step, fractionDigits, }) {
    const startingValue = value === '' || value == null ? '0' : value;
    const stepAsFloat = parseFloat(step);
    let result;
    if (isNaN(stepAsFloat)) {
        result = parseFloat(startingValue) + increment;
    }
    else {
        const increaseBy = increment * stepAsFloat;
        result = parseFloat(startingValue) + increaseBy;
    }
    return result.toFixed(fractionDigits);
}
export function calculateFractionDigitsFromStep(step) {
    let calculatedFractionDigits;
    if (step) {
        const stepAsString = String(step).toLowerCase();
        if (stepAsString !== 'any') {
            let fractionDigits = 0;
            if (stepAsString.indexOf('.') >= 0 &&
                stepAsString.indexOf('e') < 0) {
                const fractionalPart = stepAsString.split('.')[1];
                if (parseInt(fractionalPart, 10) > 0) {
                    fractionDigits = fractionalPart.length;
                }
            }
            else if (stepAsString.indexOf('e-') > 0) {
                const splitOnExponent = stepAsString.split('e-');
                const fractionalPart = splitOnExponent[0].split('.')[1];
                const exponentPart = splitOnExponent[1];
                const fractionalPartLength = fractionalPart
                    ? fractionalPart.length
                    : 0;
                fractionDigits =
                    parseInt(exponentPart, 10) + fractionalPartLength;
            }
            calculatedFractionDigits = fractionDigits;
        }
    }
    return calculatedFractionDigits;
}
export function formatNumber(numberAsString, options) {
    if (isEmptyString(numberAsString)) {
        return '';
    }
    let formattedValue = numberAsString;
    let inputValue = numberAsString;
    const formatStyle = options.style;
    const formatOptions = { style: formatStyle };
    formatOptions.minimumFractionDigits = options.minimumFractionDigits;
    formatOptions.maximumFractionDigits = options.maximumFractionDigits;
    if (formatStyle === 'percent-fixed') {
        formatOptions.style = 'percent';
        const inputValueAsString = inputValue.toString();
        const normalisedNumberInPercent = parseFloat(inputValue) / 100;
        if (inputValueAsString.indexOf('.') > 0 &&
            inputValueAsString.indexOf('e') < 0) {
            inputValue = normalisedNumberInPercent.toFixed(inputValueAsString.split('.')[1].length + 2);
        }
        else {
            inputValue = normalisedNumberInPercent;
        }
    }
    try {
        formattedValue = numberFormat(formatOptions).format(inputValue) || '';
    }
    catch (ignore) {
    }
    return formattedValue;
}
export function transformLocalizedNumberToIsoDecimal(numberAsString, separators) {
    if (numberAsString == null || numberAsString.length === 0) {
        return '';
    }
    const decimalSymbol = separators
        ? separators.decimalSeparator
        : decimalSeparator;
    const groupingSymbol = separators
        ? separators.groupSeparator
        : groupingSeparator;
    let result = numberAsString.split(groupingSymbol).join('');
    if (decimalSymbol !== '.') {
        result = result.replace(decimalSymbol, '.');
    }
    return expandShortcuts(addLeadingZeroIfNeeded(fromLocalizedDigits(result)));
}
export function isValidNumberCharacter(character) {
    return VALID_NUMBER_CHARACTERS_EXPRESSION.test(character);
}
export function stringifyNumber(value) {
    const numericValue = typeof value === 'string' ? Number(value) : value;
    const normalizedNumber = normalizeNumber(numericValue);
    return normalizedNumber === undefined ? '' : String(value);
}
export function hasValidNumberSymbol(value) {
    const validSymbols = NUMBER_SYMBOLS.join('');
    const matchSymbols = new RegExp(`[${validSymbols}]`);
    return value.match(matchSymbols) ? true : false;
}
export function hasValidNumberShortcut(value) {
    const result = value.toLowerCase().trim();
    const kmb = SHORTCUTS.join('');
    const matchShortcuts = new RegExp(`([${kmb}])`, 'g');
    const shortcutMatch = result.match(matchShortcuts);
    if (shortcutMatch && shortcutMatch.length > 1) {
        return false;
    }
    const matchEndsWith = new RegExp(`[${kmb}]$`);
    const endsWithShortcut = result.match(matchEndsWith) !== null;
    return endsWithShortcut && result.length > 1;
}
export function expandShortcuts(isoValue) {
    if (!hasValidNumberShortcut(isoValue)) {
        return isoValue;
    }
    let result = isoValue.toLowerCase().trim();
    const shortcut = result.charAt(result.length - 1);
    result = result.substring(0, result.length - 1);
    if (isNaN(result)) {
        return isoValue;
    }
    const parts = result.split('.');
    let fractionDigits = 0;
    const hasDecimalPart = parts.length > 1;
    if (hasDecimalPart) {
        fractionDigits = parts[1].length;
    }
    const exponent = SHORTCUT_FACTORS[shortcut];
    const newFractionDigits = Math.max(0, fractionDigits - exponent);
    return parseFloat(result * Math.pow(10, exponent)).toFixed(newFractionDigits);
}
function addLeadingZeroIfNeeded(result) {
    const decimalSeparatorLocation = result.indexOf('.');
    if (decimalSeparatorLocation === 0 || decimalSeparatorLocation === 1) {
        const firstCharacter = result.charAt(0);
        if (firstCharacter === '+' ||
            firstCharacter === '-' ||
            firstCharacter === '.') {
            result =
                result.substring(0, decimalSeparatorLocation) +
                    '0' +
                    result.substring(decimalSeparatorLocation);
        }
    }
    return result;
}
export function getDecimalPlaces(num) {
    const match = ('' + num).match(/(?:\.(\d+))?(?:[eE]([+-]?\d+))?$/);
    if (!match) {
        return 0;
    }
    return Math.max(0, (match[1] ? match[1].length : 0) -
        (match[2] ? +match[2] : 0));
}
