// lightning-base-components/src/lightning/formattedDateTime/formattedDateTime.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { api, track } from 'lwc';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import { dateTimeFormat } from 'lightning/internationalizationLibrary';
import { normalizeBoolean, normalizeString, toDateTimeFormatOptions, toFormattedDate, } from 'lightning/utilsPrivate';
const DATE_STYLE_DEFAULT = 'medium';
export default class LightningFormattedDateTime extends LightningShadowBaseClass {
    @api
    value;
    @api
    weekday;
    @api
    era;
    @api
    year;
    @api
    month;
    @api
    day;
    @api
    hour;
    @api
    minute;
    @api
    second;
    @api
    timeZoneName;
    @api
    timeZone;
    _dateStyle = DATE_STYLE_DEFAULT;
    @track
    _hour12 = false;
    @track
    _hour12Set = false;
    @api
    get hour12() {
        return this._hour12;
    }
    set hour12(value) {
        if (value === undefined) {
            this._hour12Set = false;
            this._hour12 = value;
        }
        else {
            this._hour12Set = true;
            this._hour12 = normalizeBoolean(value);
        }
    }
    @api
    get dateStyle() {
        return this._dateStyle;
    }
    set dateStyle(value) {
        const normalizedValue = normalizeString(value, {
            fallbackValue: 'medium',
            validValues: ['short', 'medium', 'long'],
        });
        this._dateStyle = normalizedValue;
    }
    get formattedValue() {
        return toFormattedDate(this, dateTimeFormat);
    }
    getOptions() {
        return toDateTimeFormatOptions(this);
    }
}
