// lightning-base-components/src/lightning/popup/__examples__/walkthrough/walkthrough.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { LightningElement } from 'lwc';
export default class PopupExample extends LightningElement {
    openPopup() {
        const referenceElement = this.template.querySelector('lightning-button');
        this.popup.show(referenceElement, {
            reference: { horizontal: 'right', vertical: 'top' },
            popup: { horizontal: 'left', vertical: 'top' },
            padding: 1,
            offset: 1.25,
        });
    }
    handleAlignmentUpdate(event) {
        this._alignment = event.target.alignment;
    }
    handleAutoFocus(event) {
        event.preventDefault();
        this.template.querySelector('lightning-input').focus();
    }
    get popupClasses() {
        const { horizontal, vertical } = this._alignment;
        const nubbinClass = `slds-nubbin_${horizontal}-${vertical}`;
        return `slds-popover slds-popover_walkthrough ${nubbinClass}`;
    }
    get popup() {
        return this.template.querySelector('lightning-popup');
    }
}
