// lightning-base-components/src/lightning/modal/__modalUtils__/modalContainerTestMethods.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
const { OVERLAY_CONTAINER, MODAL_BASE, MODAL, MODAL_DIV_SLOT, MODAL_CONTAINER_DIV, MODAL_HEADER, MODAL_BODY, MODAL_BODY_DIV, MODAL_BODY_BACKDROP, MODAL_BODY_SLOT, MODAL_FOOTER, FOCUS_TRAP, MODAL_CLOSE_BTN, MODAL_CLOSE_BTN_CLASS, MODAL_CLOSE_BTN_FULL_CLASS, MODAL_CLOSE_BUTTON_NORMAL_VARIANT, SCREEN_SIZE_LARGE, MODAL_SIZE_FULL, MODAL_FULL_SCREEN_SMALL_BREAKPOINT, MODAL_DEFAULT_PX_OFFSET_X, MODAL_ELEM_FULL_PX_OFFSET_X, MODAL_DEFAULT_SCREEN_USE_PERCENT, MODAL_FULL_SCREEN_USE_PERCENT, MAX_HEIGHT, MIN_HEIGHT, MODAL_BODY_MIN_HEIGHT_PX, SPEC_TO_TABS_TO_CLOSE_BTN, BROWSER_RESIZE_PAUSE, MODAL_RENDER_PAUSE, PAUSE_MICRO, SELECTORS, NAME_TO_SIZE, SCREEN_SIZE, KEY, MODAL_DESCRIPTION_SELECTOR, MODAL_ARIALIVE_SELECTOR, } = require('./modalContainerTestConstants.js');
let wrapper = null;
async function clickButton(btnSelector) {
    wrapper = await kontajner.getWrapper();
    await wrapper.waitForDisplayed();
    const button = await wrapper.shadow$(btnSelector);
    await button.click();
}
async function getOverlayContainer() {
    const htmlElem = await $('html');
    const bodyElem = await $('body');
    const headElem = await $('head');
    const overlayContainerElem = await $(OVERLAY_CONTAINER);
    return { htmlElem, bodyElem, headElem, overlayContainerElem };
}
async function setupStrategy() {
    await browser.addLocatorStrategy('getFirstSlotElement', (slotElem) => slotElem.assignedNodes()[0]);
    await browser.addLocatorStrategy('getSlotElements', (slotElem) => slotElem.assignedNodes());
}
function isModalOfVariantType(modalVariantType, type) {
    if (!modalVariantType || !type) {
        return null;
    }
    const lowerCaseVariantType = modalVariantType.toLowerCase();
    return lowerCaseVariantType.includes(type);
}
async function getModalInternals(config, modalIndex = 0) {
    const { modalVariantType } = config;
    expect(modalVariantType).not.toBeFalsy();
    let modalBaseElem = null;
    let modalBaseElems = null;
    let modalBaseBackdropElem = null;
    let modalSectionElem = null;
    let modalAriaDescribedBySpanElem = null;
    let modalAriaLiveMessageSpanElem = null;
    let modalCloseButton = null;
    let modalDataSlot = null;
    let modalContainerElem = null;
    let modalElem = null;
    let modalHeaderElem = null;
    let modalBodyElem = null;
    let modalFooterElem = null;
    let focusTrapElem = null;
    let focusTrapSlotElem = null;
    const { overlayContainerElem } = await getOverlayContainer();
    if (modalIndex === 0) {
        modalBaseElem = await overlayContainerElem.shadow$(MODAL_BASE);
    }
    else {
        modalBaseElems = await overlayContainerElem.shadow$$(MODAL_BASE);
        const modalBaseElemRequested = await modalBaseElems[modalIndex];
        modalBaseElem = modalBaseElemRequested ? modalBaseElemRequested : null;
    }
    const modalBaseExists = await modalBaseElem.isExisting();
    if (modalBaseExists) {
        modalBaseBackdropElem =
            await modalBaseElem.shadow$(MODAL_BODY_BACKDROP);
        focusTrapElem = await modalBaseElem.shadow$(FOCUS_TRAP);
        focusTrapSlotElem = await focusTrapElem.shadow$('slot');
        modalSectionElem = await browser.custom$('getFirstSlotElement', focusTrapSlotElem);
        modalAriaDescribedBySpanElem = await modalSectionElem.$(MODAL_DESCRIPTION_SELECTOR);
        modalAriaLiveMessageSpanElem = await modalSectionElem.$(MODAL_ARIALIVE_SELECTOR);
        modalCloseButton = await modalSectionElem.$(MODAL_CLOSE_BTN);
        modalDataSlot = await modalSectionElem.$(MODAL_DIV_SLOT);
        modalContainerElem = await modalSectionElem.$(MODAL_CONTAINER_DIV);
        modalElem = await modalSectionElem.$(MODAL);
        modalBodyElem = await modalElem.shadow$(MODAL_BODY);
        if (!isModalOfVariantType(modalVariantType, 'headless')) {
            modalHeaderElem = await modalElem.shadow$(MODAL_HEADER);
        }
        if (!isModalOfVariantType(modalVariantType, 'footless')) {
            modalFooterElem = await modalElem.shadow$(MODAL_FOOTER);
        }
    }
    return {
        modalBaseExists,
        modalBaseElem,
        modalBaseBackdropElem,
        modalSectionElem,
        modalAriaDescribedBySpanElem,
        modalAriaLiveMessageSpanElem,
        modalCloseButton,
        modalDataSlot,
        modalContainerElem,
        modalElem,
        modalHeaderElem,
        modalBodyElem,
        modalFooterElem,
        focusTrapElem,
        focusTrapSlotElem,
    };
}
async function getModalHeaderInternals(headerElem) {
    if (!headerElem) {
        return null;
    }
    let headerOuterDiv = null;
    let headerHeading = null;
    let headerSlot = null;
    if (headerElem) {
        headerOuterDiv = await headerElem.shadow$('.slds-modal__header');
        headerHeading = await headerElem.shadow$('[data-label]');
        headerSlot = await headerElem.shadow$('[data-default-slot]');
    }
    return {
        headerOuterDiv,
        headerHeading,
        headerSlot,
    };
}
async function getModalBodyInternals(bodyElem) {
    if (!bodyElem) {
        return null;
    }
    let bodyOuterDiv = null;
    let bodySlotContents = null;
    if (bodyElem) {
        bodyOuterDiv = await bodyElem.shadow$(MODAL_BODY_DIV);
        const bodySlot = await bodyElem.shadow$(MODAL_BODY_SLOT);
        bodySlotContents = await browser.custom$('getSlotElements', bodySlot);
    }
    return {
        bodyOuterDiv,
        bodySlotContents,
    };
}
async function getModalFooterInternals(footerElem) {
    if (!footerElem) {
        return null;
    }
    let footerOuterDiv = null;
    let footerSlotContents = null;
    if (footerElem) {
        footerOuterDiv = await footerElem.shadow$('.slds-modal__footer');
        const footerSlot = await footerElem.shadow$('[data-footer-slot]');
        footerSlotContents = await browser.custom$('getSlotElements', footerSlot);
    }
    return {
        footerOuterDiv,
        footerSlotContents,
    };
}
async function getElementInModalFooter(footerSlotContents, selector, elemIndex = 0) {
    let element = null;
    const elements = await footerSlotContents.$$(selector);
    if (elements) {
        element = await elements[elemIndex];
    }
    return element;
}
async function getNumModals() {
    let numModals = 0;
    const { overlayContainerElem } = await getOverlayContainer();
    const modalBaseElems = await overlayContainerElem.shadow$$(MODAL_BASE);
    if (modalBaseElems) {
        numModals = modalBaseElems.length;
    }
    return numModals;
}
async function validateModalCloseButtonAttributes(config) {
    const { modalSize } = config;
    const { width: windowWidth } = await browser.getWindowSize();
    let modalCloseButtonVariant = null;
    let modalCloseButtonCssClass = null;
    const { modalCloseButton } = await getModalInternals(config);
    if (modalCloseButton) {
        modalCloseButtonVariant =
            await modalCloseButton.getAttribute('variant');
        modalCloseButtonCssClass = await modalCloseButton.getAttribute('class');
    }
    expect(modalCloseButtonVariant).not.toBeNull();
    expect(modalCloseButtonCssClass).toContain(MODAL_CLOSE_BTN_CLASS);
    expect(modalCloseButtonVariant).toEqual(MODAL_CLOSE_BUTTON_NORMAL_VARIANT);
    if (modalSize === MODAL_SIZE_FULL &&
        windowWidth <= MODAL_FULL_SCREEN_SMALL_BREAKPOINT) {
        expect(modalCloseButtonCssClass).toContain(MODAL_CLOSE_BTN_FULL_CLASS);
    }
    else {
        expect(modalCloseButtonCssClass).not.toContain(MODAL_CLOSE_BTN_FULL_CLASS);
    }
}
async function getModalInternalsDomBased() {
    const result = await browser.execute(() => {
        const loc = document.querySelector('lightning-overlay-container');
        const modalBase = loc.shadowRoot.querySelector('lightning-modal-base');
        const modalBackdrop = modalBase.shadowRoot.querySelector('.slds-backdrop');
        const focusTrap = modalBase.shadowRoot.querySelector('lightning-focus-trap');
        const focusTrapSlot = focusTrap.shadowRoot.querySelector('slot');
        const modalSectionElem = focusTrapSlot.assignedNodes()[0];
        const modalContainerElem = modalSectionElem.querySelector('.slds-modal__container');
        const modalCloseButton = modalSectionElem.querySelector('lightning-button-icon');
        const modalElem = modalSectionElem.querySelector('lightning-modal');
        return {
            window: {
                width: window.innerWidth,
                height: window.innerHeight,
            },
            modalBase: {
                offset: modalBase.getBoundingClientRect(),
            },
            modalBackdrop: {
                offset: modalBackdrop.getBoundingClientRect(),
            },
            modalSectionElem: {
                class: modalSectionElem.getAttribute('class'),
                offset: modalSectionElem.getBoundingClientRect(),
            },
            modalContainerElem: {
                class: modalContainerElem.getAttribute('class'),
                offset: modalContainerElem.getBoundingClientRect(),
            },
            modalCloseButton: {
                class: modalCloseButton.getAttribute('class'),
            },
            modalElem: {
                offset: modalElem.getBoundingClientRect(),
            },
        };
    });
    return result;
}
async function validateModalHeightBehavior(config) {
    const { modalSize, modalVariantType } = config;
    const { height: windowHeight, width: windowWidth } = await browser.getWindowSize();
    const { modalHeaderElem, modalBodyElem, modalFooterElem } = await getModalInternals(config);
    const { bodyOuterDiv } = await getModalBodyInternals(modalBodyElem);
    const modalBodyOuterDivStyle = await bodyOuterDiv.getAttribute('style');
    const modalBodyStyleProps = parseStyleAttributes(modalBodyOuterDivStyle);
    const isVariantHeadless = isModalOfVariantType(modalVariantType, 'headless');
    const isVariantFootless = isModalOfVariantType(modalVariantType, 'footless');
    const heightBody = (await bodyOuterDiv.getSize('height')) || 0;
    let totalModalElemHeight = heightBody;
    if (!isVariantHeadless) {
        const { headerOuterDiv } = await getModalHeaderInternals(modalHeaderElem);
        const heightHdr = await headerOuterDiv.getSize('height');
        if (heightHdr > 0) {
            totalModalElemHeight += heightHdr;
        }
    }
    if (!isVariantFootless) {
        const { footerOuterDiv } = await getModalFooterInternals(modalFooterElem);
        const heightFtr = await footerOuterDiv.getSize('height');
        if (heightFtr > 0) {
            totalModalElemHeight += heightFtr;
        }
    }
    const percentPageHeightUtilized = totalModalElemHeight / windowHeight;
    if (!isVariantFootless) {
        const { footerOuterDiv } = await getModalFooterInternals(modalFooterElem);
        const heightFtr = await footerOuterDiv.getSize('height');
        if (heightFtr > 0) {
            totalModalElemHeight += heightFtr;
        }
    }
    if (modalSize === MODAL_SIZE_FULL &&
        windowWidth <= MODAL_FULL_SCREEN_SMALL_BREAKPOINT) {
        expect(percentPageHeightUtilized).toBeGreaterThan(MODAL_FULL_SCREEN_USE_PERCENT);
        expect(modalBodyStyleProps[MAX_HEIGHT]).toBeUndefined();
        expect(modalBodyStyleProps[MIN_HEIGHT]).toBeUndefined();
    }
    else {
        expect(percentPageHeightUtilized).toBeLessThan(MODAL_DEFAULT_SCREEN_USE_PERCENT);
        expect(modalBodyStyleProps[MAX_HEIGHT]).toBeDefined();
        expect(modalBodyStyleProps[MIN_HEIGHT]).toBeDefined();
        expect(modalBodyStyleProps[MAX_HEIGHT]).not.toEqual(modalBodyStyleProps[MIN_HEIGHT]);
        expect(modalBodyStyleProps[MIN_HEIGHT]).toEqual(MODAL_BODY_MIN_HEIGHT_PX);
    }
}
async function validateModalWidthBehavior(config) {
    const { modalSize, modalVariantType } = config;
    const { width: windowWidth } = await browser.getWindowSize();
    const { modalHeaderElem, modalBodyElem, modalFooterElem } = await getModalInternals(config);
    const { bodyOuterDiv } = await getModalBodyInternals(modalBodyElem);
    const isVariantHeadless = isModalOfVariantType(modalVariantType, 'headless');
    const isVariantFootless = isModalOfVariantType(modalVariantType, 'footless');
    let locHeaderX = null, locFooterX = null;
    const { x: locBodyX } = await bodyOuterDiv.getLocation();
    if (!isVariantHeadless) {
        const { headerOuterDiv } = await getModalHeaderInternals(modalHeaderElem);
        const locHeader = await headerOuterDiv.getLocation();
        if (locHeader) {
            locHeaderX = locHeader.x;
        }
    }
    if (!isVariantFootless) {
        const { footerOuterDiv } = await getModalFooterInternals(modalFooterElem);
        const locFooter = await footerOuterDiv.getLocation();
        if (locFooter) {
            locFooterX = locFooter.x;
        }
    }
    if (modalSize === MODAL_SIZE_FULL &&
        windowWidth <= MODAL_FULL_SCREEN_SMALL_BREAKPOINT) {
        if (!isVariantFootless) {
            expect(locFooterX).toEqual(MODAL_ELEM_FULL_PX_OFFSET_X);
        }
        if (!isVariantHeadless) {
            expect(locHeaderX).toEqual(MODAL_ELEM_FULL_PX_OFFSET_X);
        }
        expect(locBodyX).toEqual(MODAL_ELEM_FULL_PX_OFFSET_X);
    }
    else {
        if (!isVariantFootless) {
            expect(locFooterX).toBeGreaterThan(MODAL_DEFAULT_PX_OFFSET_X);
        }
        if (!isVariantHeadless) {
            expect(locHeaderX).toBeGreaterThan(MODAL_DEFAULT_PX_OFFSET_X);
        }
        expect(locBodyX).toBeGreaterThan(MODAL_DEFAULT_PX_OFFSET_X);
    }
}
const getScreenSizeValues = (size = SCREEN_SIZE_LARGE) => {
    if (!size) {
        return null;
    }
    return SCREEN_SIZE[size];
};
const getModalSizeFromName = (name = 'all') => {
    if (!name) {
        return null;
    }
    return NAME_TO_SIZE[name];
};
async function setWindowSize(screenSizeToSet) {
    const { width, height } = getScreenSizeValues(screenSizeToSet);
    await browser.setWindowSize(width, height);
    await browser.pause(BROWSER_RESIZE_PAUSE);
}
async function openModal({ modalVariantType = '', screenSizeToSet = SCREEN_SIZE_LARGE, runMockMatchMedia = false, }) {
    if (!modalVariantType) {
        expect(modalVariantType).not.toEqual('');
        return false;
    }
    if (runMockMatchMedia) {
        await mockMatchMedia();
    }
    await setWindowSize(screenSizeToSet);
    const modalBtnSelector = SELECTORS[modalVariantType];
    await clickButton(modalBtnSelector);
    await browser.pause(MODAL_RENDER_PAUSE);
    return true;
}
function isModalCloseButton(info) {
    if (!info) {
        expect(info).not.toBeFalsy();
    }
    const { html, type } = info;
    const isTypeButton = type === 'button';
    const hasCorrectSldsClasses = html.indexOf('slds-button') >= 0 &&
        html.indexOf('slds-button_icon') >= 0;
    const hasCorrectText = html.indexOf('Cancel and close') >= 0;
    return (isTypeButton && hasCorrectSldsClasses && hasCorrectText) || false;
}
async function closeModal(config, modalIndex = 0, closeMethod = 'click') {
    if (typeof modalIndex !== 'number' && modalIndex >= 0) {
        console.error('closeModal :: requires modalIndex value to correctly function');
    }
    if (closeMethod === 'click') {
        const { modalCloseButton } = await getModalInternals(config, modalIndex);
        if (modalCloseButton) {
            await modalCloseButton.click();
            await browser.pause(MODAL_RENDER_PAUSE);
        }
    }
    else if (closeMethod === KEY.ENTER || closeMethod === KEY.SPACE) {
        const activeElement = await getActiveShadowElement();
        const closeButtonIsFocused = isModalCloseButton(activeElement);
        expect(closeButtonIsFocused).toBeTrue();
        if (closeButtonIsFocused) {
            await browser.keys(closeMethod);
            await browser.pause(MODAL_RENDER_PAUSE);
        }
    }
    else if (closeMethod === KEY.ESC) {
        await browser.keys(closeMethod);
        await browser.pause(MODAL_RENDER_PAUSE);
    }
}
async function mockMatchMedia() {
    await browser.execute(() => {
        Object.defineProperty(window, 'matchMedia', {
            writable: true,
            value: () => ({
                matches: true,
                media: '(prefers-reduced-motion: reduce)',
            }),
        });
    });
}
async function checkBackgroundElemForInertness(elems, expectedValue = true) {
    return Promise.all(elems.map(async (elem) => {
        const elemType = await elem.getTagName();
        const initAriaHiddenValue = await elem.getAttribute('aria-hidden');
        const ariaHiddenValue = initAriaHiddenValue === 'true' ? true : null;
        if (elemType === 'lightning-overlay-container' ||
            elemType === 'lightning-primitive-bubble') {
            return;
        }
        const messageIfFails = `${elemType} should have aria-hidden set ${expectedValue}, but was ${ariaHiddenValue}`;
        expect(ariaHiddenValue).toEqual(expectedValue, messageIfFails);
    }));
}
async function validateInertness(bgInert = true) {
    const ariaHiddenValue = bgInert ? true : null;
    const htmlElem = await $('html');
    const rootElemsToCheck = await htmlElem.$$('body > *');
    const rootElemsCleaned = await rootElemsToCheck.filter((elem) => Boolean(elem));
    const headElem = await $('head');
    const headAriaHiddenValueCheck = await headElem.getAttribute('aria-hidden');
    expect(headAriaHiddenValueCheck).toBeNull();
    await checkBackgroundElemForInertness(rootElemsCleaned, ariaHiddenValue);
}
async function repeatTab(tabCount = 0) {
    if (tabCount !== 0) {
        const shouldShiftTab = tabCount < 0;
        const numTabs = Math.abs(tabCount);
        for (let count = numTabs; count > 0; count--) {
            if (shouldShiftTab) {
                await browser.keys([KEY.SHIFT, KEY.TAB, 'NULL']);
            }
            else {
                await browser.keys(KEY.TAB);
            }
        }
        if (shouldShiftTab) {
            await browser.keys(KEY.SHIFT);
        }
    }
}
async function repeatKey(repeatCount = 0, keyName) {
    if (repeatCount > 0 || !keyName) {
        for (let count = Math.abs(repeatCount); count > 0; count--) {
            await browser.keys(keyName);
            await browser.pause(PAUSE_MICRO);
        }
    }
}
async function typeLetters(word) {
    if (word && word.length > 0) {
        const arrOfLetters = word.split('');
        for (let count = 0; count < arrOfLetters.length; count++) {
            const letterToType = arrOfLetters[count];
            await browser.keys(letterToType);
            await browser.pause(PAUSE_MICRO);
        }
    }
}
async function fillForm({ data = [], tabToFormAction = 0 }) {
    if (!data || data.length === 0) {
        console.error('fillForm :: requires data array of fields');
    }
    for (let count = 0; count < data.length; count++) {
        const { name, value, type, numTabsToNext } = data[count];
        if (type === 'text') {
            await browser.keys(value);
        }
        else if (type === 'combobox') {
            if (typeof value === 'number') {
                await browser.keys('Enter');
                await repeatKey(value, 'ArrowDown');
                await browser.keys('Enter');
            }
            else {
                await browser.keys('Enter');
                await typeLetters(value);
                await browser.keys('Enter');
            }
        }
        else if (type === 'datepicker') {
            await browser.keys(value);
            await browser.keys('Enter');
        }
        else if (type === 'timepicker') {
            await browser.keys('Enter');
            await repeatKey(value, 'ArrowDown');
            await browser.keys('Enter');
        }
        else if (type === 'heading') {
        }
        else {
            console.error("fillForm :: Ooops, shouldn't happen: ", name, ' field');
        }
        if (numTabsToNext !== 0) {
            await repeatTab(numTabsToNext);
        }
    }
    if (tabToFormAction > 0) {
        await repeatTab(tabToFormAction);
        await browser.keys('Enter');
    }
}
function validateData(verifyData, origData) {
    if (!verifyData || !origData) {
        console.error('validateData :: missing data');
    }
    const dataValidated = {};
    let countInvalidData = 0;
    let cleanedStr;
    let cleanedData;
    if (typeof verifyData === 'string') {
        cleanedStr = verifyData.replace(/\\/g, '');
        const unwantedChar = '"';
        if (cleanedStr.charAt(0) === unwantedChar) {
            cleanedStr = cleanedStr.substring(1, cleanedStr.length - 1);
        }
        if (cleanedStr.charAt(cleanedStr.length - 1) === unwantedChar) {
            cleanedStr = cleanedStr.substring(0, cleanedStr.length);
        }
        cleanedData = JSON.parse(cleanedStr);
    }
    const data = cleanedData ? cleanedData : verifyData;
    origData.forEach((item) => {
        const { name, value, result } = item;
        const hasKeys = name ? name.indexOf('.') >= 0 : false;
        const keys = hasKeys ? name.split('.') : [name];
        const originalValue = String(result ? result : value);
        const receivedValue = keys.length === 2
            ? String(data[keys[0]][keys[1]])
            : String(data[keys[0]]);
        const valuesMatch = receivedValue === originalValue;
        if (valuesMatch === false) {
            countInvalidData++;
        }
        dataValidated[name] = receivedValue === originalValue;
    });
    return {
        incorrectValues: countInvalidData,
        dataValidated,
    };
}
async function getActiveShadowElements() {
    return browser.execute(() => {
        let activeElement = document.activeElement;
        const activeElements = [];
        while (activeElement &&
            activeElement.shadowRoot &&
            activeElement.shadowRoot.activeElement) {
            const elem = {
                type: activeElement.localName,
                html: null,
            };
            if (activeElement.outerHTML) {
                elem.html = activeElement.outerHTML;
            }
            activeElements.push(elem);
            activeElement = activeElement.shadowRoot.activeElement;
        }
        if (activeElement) {
            const elem = {
                type: activeElement.localName,
                html: null,
            };
            if (activeElement.outerHTML) {
                elem.html = activeElement.outerHTML;
            }
            activeElements.push(elem);
        }
        return activeElements;
    });
}
async function getActiveShadowElement() {
    const allActiveShadowElements = await getActiveShadowElements();
    if (allActiveShadowElements) {
        const activeElemsLength = allActiveShadowElements.length;
        return allActiveShadowElements[activeElemsLength - 1];
    }
    return [];
}
async function getCalendarDetail() {
    return browser.execute(function () {
        let activeElement = document.activeElement;
        let details;
        let nodes;
        let calendarStyles = {};
        const activeElements = [];
        const CALENDAR = 'lightning-calendar';
        function getStylesObj(str = '') {
            const styleObj = {};
            const stylesStr = str.replace(/ /g, '');
            const styleArr = stylesStr.split(';');
            styleArr.forEach((string) => {
                const [key, value] = string.split(':');
                if (key && value && !styleObj[key]) {
                    let num = null;
                    if (value.indexOf('px') > 0) {
                        num = value.substring(0, value.length - 2);
                    }
                    if (key === 'z-index') {
                        num = value;
                    }
                    styleObj[key] = num ? Number(num) : value;
                }
            });
            return styleObj;
        }
        while (activeElement &&
            activeElement.shadowRoot &&
            activeElement.shadowRoot.activeElement) {
            const type = activeElement.localName;
            activeElements.push(type);
            if (type === CALENDAR) {
                const shadowRoot = activeElement.shadowRoot;
                nodes = Array.from(shadowRoot.childNodes);
                const nodeObj = nodes[0];
                const nodeObjStyleAttr = nodeObj.getAttribute('style');
                calendarStyles = getStylesObj(nodeObjStyleAttr);
                details = {
                    calendarType: true,
                    calendarStyles,
                };
            }
            activeElement = activeElement.shadowRoot.activeElement;
        }
        return details;
    });
}
function parseStyleAttributes(styleString) {
    if (!styleString) {
        return {};
    }
    const styleProps = {};
    let stylesSplit = styleString.trim().split('; ');
    if (stylesSplit && stylesSplit.length > 0) {
        stylesSplit = stylesSplit.forEach((style) => {
            const updatedStyle = style.replace(';', '').replace(':', '').trim();
            const propSplit = updatedStyle.split(' ');
            if (propSplit) {
                const name = propSplit[0];
                let value = propSplit[1];
                if (value.includes('px')) {
                    value = Math.floor(Number(value.split('px')[0]));
                }
                styleProps[name] = value;
            }
        });
    }
    return styleProps;
}
async function validateModalFocus({ html = '', type = '', whichArrayElem = '', }) {
    expect(html).not.toBeFalsy();
    expect(type).not.toBeFalsy();
    const activeElemArr = await getActiveShadowElements();
    let penultimateElem;
    const len = activeElemArr.length;
    const lastElem = activeElemArr[len - 1];
    let actualHtml = lastElem.html;
    let actualType = lastElem.type;
    if (len > 1 &&
        actualType &&
        actualType === 'button' &&
        whichArrayElem === 'penultimate') {
        penultimateElem = activeElemArr[len - 2];
        const { type: penultimateElemType, html: penultimateElemHtml } = penultimateElem;
        if (penultimateElemType &&
            [
                'lightning-button',
                'lightning-button-icon',
                'lightning-helptext',
            ].some((val) => penultimateElemType === val)) {
            actualHtml = penultimateElemHtml;
            actualType = penultimateElemType;
        }
    }
    expect(actualHtml).toContain(html);
    expect(actualType).toEqual(type);
    return { html: actualHtml, type: actualType };
}
async function validateNumModals(values, checkNow = false) {
    expect(values).not.toBeFalsy();
    let numCurrentModals = null;
    if (checkNow) {
        numCurrentModals = await getNumModals();
    }
    if (Array.isArray(values)) {
        values.forEach(({ actual = null, expected = null }) => {
            if (actual === null || actual === undefined) {
                expect(numCurrentModals).toEqual(expected);
            }
            else {
                expect(actual).toEqual(expected);
            }
        });
    }
    else if (typeof values === 'object') {
        const { actual, expected } = values;
        if (actual === null || actual === undefined) {
            expect(numCurrentModals).toEqual(expected);
        }
        else {
            expect(actual).toEqual(expected);
        }
    }
    else {
        console.error('validateNumModals expects an array of objects, or an object');
    }
}
function getNumTabsToCloseButton(spec = '', modalVariantType = '') {
    expect(spec).toBeTruthy();
    expect(modalVariantType).toBeTruthy();
    if (spec && modalVariantType) {
        return SPEC_TO_TABS_TO_CLOSE_BTN[spec][modalVariantType] || 0;
    }
    return 0;
}
async function validateModalOpenCloseFocusBehavior(config, expectedAfterOpen, expectedAfterClose, closeMethod = 'click', spec = 'accessibility') {
    expect(config).not.toBeFalsy();
    expect(expectedAfterOpen).not.toBeFalsy();
    expect(expectedAfterClose).not.toBeFalsy();
    await openModal(config);
    await validateModalFocus(expectedAfterOpen);
    const numModalsAtStart = await getNumModals();
    if (closeMethod === KEY.ENTER || closeMethod === KEY.SPACE) {
        const { modalVariantType } = config;
        const tabNavValue = getNumTabsToCloseButton(spec, modalVariantType);
        if (tabNavValue !== 0 && typeof tabNavValue === 'number') {
            await repeatTab(tabNavValue);
        }
    }
    await closeModal(config, 0, closeMethod);
    const numModalsAfterClose = await getNumModals();
    await validateModalFocus(expectedAfterClose);
    const expectedNumModals = [
        { actual: numModalsAtStart, expected: 1 },
        { actual: numModalsAfterClose, expected: 0 },
    ];
    validateNumModals(expectedNumModals);
}
async function validateModalOpenTabNavFocusBehavior(config, expectedAfterOpen, expectedAfterTab, numOfTabsToNext) {
    expect(config).not.toBeFalsy();
    expect(expectedAfterOpen).not.toBeFalsy();
    expect(expectedAfterTab).not.toBeFalsy();
    expect(numOfTabsToNext).not.toBeFalsy();
    await openModal(config);
    await validateModalFocus(expectedAfterOpen);
    await repeatTab(numOfTabsToNext);
    await validateModalFocus(expectedAfterTab);
}
module.exports = {
    getModalInternalsDomBased,
    setupStrategy,
    isModalOfVariantType,
    parseStyleAttributes,
    mockMatchMedia,
    openModal,
    closeModal,
    clickButton,
    getModalInternals,
    getOverlayContainer,
    getScreenSizeValues,
    getModalSizeFromName,
    getModalHeaderInternals,
    getModalBodyInternals,
    getModalFooterInternals,
    setWindowSize,
    validateModalCloseButtonAttributes,
    validateModalHeightBehavior,
    validateModalWidthBehavior,
    validateInertness,
    checkBackgroundElemForInertness,
    repeatTab,
    repeatKey,
    typeLetters,
    fillForm,
    validateData,
    getActiveShadowElements,
    getActiveShadowElement,
    getNumModals,
    getElementInModalFooter,
    getCalendarDetail,
    validateNumModals,
    validateModalOpenCloseFocusBehavior,
    validateModalOpenTabNavFocusBehavior,
    validateModalFocus,
};
