// lightning-base-components/src/lightning/primitiveFigure/primitiveFigure.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { LightningElement, api } from 'lwc';
import { reflectAttribute } from 'lightning/primitiveUtils';
const sdsFigureParser = new DOMParser();
export default class LightningPrimitiveFigure extends LightningElement {
    static shadowSupportMode = 'native';
    _set;
    _symbol;
    _figureSets = new Map();
    _ariaHidden;
    _generatedFigure = false;
    @api
    get figureSets() {
        return this._figureSets;
    }
    set figureSets(value) {
        this._figureSets = new Map(value.map((item) => [item.name, item]));
    }
    @api
    get set() {
        return this._set;
    }
    set set(value) {
        this._set = value;
        reflectAttribute(this, 'set', this._set);
    }
    @api
    get symbol() {
        return this._symbol;
    }
    set symbol(value) {
        this._symbol = value;
        reflectAttribute(this, 'symbol', this._symbol);
    }
    getFigureSet(name) {
        return this._figureSets.get(name);
    }
    getFigureUrl(figureSet, set, symbol) {
        if (figureSet && symbol) {
            return figureSet.resolver(set, symbol);
        }
    }
    getSymbolParams(set, symbol) {
        const figureSet = this.getFigureSet(set);
        const url = this.getFigureUrl(figureSet, set, symbol);
        return {
            figureSet,
            url,
        };
    }
    hasFigureConfigured() {
        return Boolean(this._set && this._symbol);
    }
    isValidSvgUrl(url) {
        const svgUrlPattern = /\.svg[?]?/;
        return svgUrlPattern.test(url);
    }
    isValidContentType(contentType) {
        return contentType && contentType.includes('image/svg+xml');
    }
    async fetchFigure(url) {
        try {
            if (!this.isValidSvgUrl(url)) {
                throw new Error(`Invalid URL: ${url}. Only SVG files are allowed.`);
            }
            const response = await fetch(url);
            if (response.ok) {
                const contentType = response.headers.get('content-type');
                if (!this.isValidContentType(contentType)) {
                    throw new Error(`Invalid content type: ${contentType}. Only SVG files are allowed.`);
                }
                const fetchText = await response.text();
                const parsedFetchDoc = sdsFigureParser.parseFromString(fetchText, 'text/html');
                const parsedSvgEl = parsedFetchDoc.body.querySelector('svg');
                const result = {
                    url: url,
                    svg: parsedSvgEl &&
                        parsedSvgEl.tagName.toLowerCase() === 'svg'
                        ? parsedSvgEl
                        : '',
                };
                return result;
            }
            else {
                throw new Error(`There's an error: Fetch response.ok is false.`);
            }
        }
        catch (error) {
            throw new Error(error.message ||
                `Network error: There's an error in the fetch API.`);
        }
    }
    async createFigure() {
        const { figureSet, url } = this.getSymbolParams(this._set, this._symbol);
        if (figureSet && url) {
            const slotContainer = !this._generatedFigure
                ? this.refs.illustration
                : this.template.querySelector(`svg`);
            const figure = await this.fetchFigure(url);
            !this._generatedFigure
                ? slotContainer.appendChild(figure.svg)
                : slotContainer.replaceWith(figure.svg);
            this._generatedFigure = true;
        }
        else {
            throw new Error(`Unable to create a figure using set: ${this._set}, symbol: ${this._symbol} and URL: ${url}. Double check your registered figureSet name(s) and path(s)`);
        }
    }
    hasSlottedChildren() {
        const slot = this.template.querySelector('slot');
        return slot.assignedNodes({ flatten: true }).length !== 0;
    }
    async handleSlotChange() {
        if (this._generatedFigure) {
            this.template.querySelector('svg').remove();
            this._generatedFigure = false;
        }
        try {
            if (!this.hasSlottedChildren() &&
                !this._generatedFigure &&
                this.hasFigureConfigured()) {
                await this.createFigure();
            }
        }
        catch (error) {
            console.error(error.message ||
                'Error: Unable to create sds-figure. There is no slotted figure. Check all required APIs.');
        }
    }
    async renderedCallback() {
        if (!this.hasSlottedChildren() && this.hasFigureConfigured()) {
            try {
                await this.createFigure();
            }
            catch (error) {
                this.template
                    .querySelectorAll('.slds-figure_error')
                    .forEach((el) => el.remove());
                const div = document.createElement('div');
                div.classList.add('slds-figure_error');
                div.textContent = `Error: ${error.message || 'Could not create figure'}`;
                this.template.appendChild(div);
            }
        }
        else if (!this.hasSlottedChildren() && !this.hasFigureConfigured()) {
            console.error('Error: Unable to create sds-figure. Check all required APIs.');
        }
        this.ariaHidden = this.ariaLabel ? 'false' : 'true';
        this.role = 'img';
    }
}
