// lightning-base-components/src/lightning/textarea/textarea.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import labelRequired from '@salesforce/label/LightningControl.required';
import labelHelpTextAlternativeText from '@salesforce/label/LightningInput.helptextAlternativeText';
import { api, track } from 'lwc';
import { features } from 'lightning/features';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import { classSet, formatLabel } from 'lightning/utils';
import { normalizeBoolean, classListMutation, decorateInputForDragon, reflectAttribute, setDecoratedDragonInputValueWithoutEvent, computeAriaInvalid, isNativeComponent, setRangeText, } from 'lightning/utilsPrivate';
import { createPrivateContext } from 'lightning/utilsInternal';
import { InteractingState, normalizeVariant, FieldConstraintApiWithProxyInput, isEmptyString, VARIANT, } from 'lightning/inputUtils';
import { TouchScroller } from 'lightning/touchScrollLibrary';
import AriaObserver from 'lightning/ariaObserver';
import emptyStylesheet from './textarea.lbc.empty.css';
import readonlyTextareaStyles from './textarea.lbc.readonly.css';
import { LightningResizeObserver } from 'lightning/resizeObserver';
const i18n = {
    required: labelRequired,
    helpTextAlternativeText: labelHelpTextAlternativeText,
};
const { setContext, assertContext, getContext } = createPrivateContext();
export default class LightningTextarea extends LightningShadowBaseClass {
    static delegatesFocus = true;
    static validationOptOut = ['class'];
    static stylesheets = [
        (features.enableTextareaReadOnlyStatic && readonlyTextareaStyles) ||
            emptyStylesheet,
    ];
    @track
    _maxLength;
    @track
    _minLength;
    @track
    _defaultValue = '';
    @track
    _disabled = false;
    @track
    _required = false;
    @track
    _readOnly = false;
    @track
    _variant;
    @track
    _helpMessage;
    @track
    _fieldLevelHelp;
    _connected;
    _rendered;
    _isNativeShadow = false;
    _value;
    _resizeObserver = null;
    _helpMessageChanged = false;
    _constraintApi;
    _constraintApiProxyInputUpdater;
    @track
    _ariaDescribedBy = '';
    @track
    _ariaLabelledBy = '';
    ariaObserver = null;
    interactingState;
    touchScroller;
    @api
    autocomplete;
    @api
    label;
    @api
    placeholder;
    @api
    name;
    @api
    messageWhenBadInput;
    @api
    messageWhenTooShort;
    @api
    messageWhenTooLong;
    @api
    messageWhenValueMissing;
    @api
    accessKey;
    @api
    get ariaDescribedBy() {
        return this._ariaDescribedBy;
    }
    set ariaDescribedBy(elements) {
        this._ariaDescribedBy = elements;
        this.connectAriaDescribedBy();
    }
    @api
    get ariaLabelledBy() {
        return this._ariaLabelledBy;
    }
    set ariaLabelledBy(references) {
        this._ariaLabelledBy = references;
        if (this._rendered) {
            this.connectAriaLabelledBy();
        }
    }
    @api
    get maxLength() {
        return this._maxLength;
    }
    set maxLength(value) {
        this._maxLength = value;
        this._updateProxyInputAttributes('maxlength');
    }
    @api
    get minLength() {
        return this._minLength;
    }
    set minLength(value) {
        this._minLength = value;
        this._updateProxyInputAttributes('minlength');
    }
    @api
    get value() {
        return this._value;
    }
    set value(value) {
        if (this._value !== value) {
            this._value = value || '';
            if (this._connected) {
                this._setInputValue(this._value);
                this.updateTextareaHeightWhenReadonly();
            }
            else {
                this._defaultValue = this._value;
            }
        }
        this._updateProxyInputAttributes('value');
    }
    @api
    get disabled() {
        return this._disabled;
    }
    set disabled(value) {
        this._disabled = normalizeBoolean(value);
        this._updateProxyInputAttributes('disabled');
    }
    @api
    get fieldLevelHelp() {
        return this._fieldLevelHelp;
    }
    set fieldLevelHelp(value) {
        this._fieldLevelHelp = value;
    }
    @api
    get readOnly() {
        return this._readOnly;
    }
    set readOnly(value) {
        const { _readOnly: wasReadOnly } = this;
        this._readOnly = normalizeBoolean(value);
        this._updateProxyInputAttributes('readonly');
        this.toggleResizeObserver(wasReadOnly);
    }
    @api
    get required() {
        return this._required;
    }
    set required(value) {
        this._required = normalizeBoolean(value);
        this._updateProxyInputAttributes('required');
    }
    @api
    get variant() {
        return this._variant || VARIANT.STANDARD;
    }
    set variant(value) {
        this._variant = normalizeVariant(value);
        reflectAttribute(this, 'variant', this._variant);
        this.updateHostClassList();
    }
    @api
    get validity() {
        return this._constraint.validity;
    }
    @api
    checkValidity() {
        return this._constraint.checkValidity();
    }
    @api
    reportValidity() {
        return this._constraint.reportValidity((message) => {
            this._helpMessage = message;
            this._helpMessageChanged = true;
        });
    }
    @api
    setCustomValidity(message) {
        this._constraint.setCustomValidity(message);
    }
    @api
    showHelpMessageIfInvalid() {
        this.reportValidity();
    }
    @api
    focus() {
        if (this._connected) {
            this.inputElement.focus();
        }
    }
    @api
    blur() {
        if (this._connected) {
            this.inputElement.blur();
        }
    }
    @api
    setRangeText(...args) {
        assertContext(this);
        if (this._connected) {
            setRangeText.apply(this.inputElement, args);
            this.value = this.inputElement.value;
        }
    }
    get isLabelHidden() {
        return this.variant === VARIANT.LABEL_HIDDEN;
    }
    get i18n() {
        return i18n;
    }
    get computedLabelClass() {
        return classSet('slds-form-element__label')
            .add({ 'slds-assistive-text': this.isLabelHidden })
            .toString();
    }
    get inputElement() {
        assertContext(this);
        const privateStore = getContext(this);
        if (!privateStore.textarea) {
            privateStore.textarea = !import.meta.env.SSR
                ? privateStore.template.querySelector('textarea')
                : null;
        }
        const element = privateStore.textarea;
        decorateInputForDragon(element);
        return element;
    }
    get _constraint() {
        if (!this._constraintApi) {
            this._constraintApi = new FieldConstraintApiWithProxyInput(() => this, {
                valueMissing: () => this._required && isEmptyString(this._value),
                tooShort: () => this._connected && this.inputElement.validity.tooShort,
                tooLong: () => this._connected && this.inputElement.validity.tooLong,
            }, 'textarea');
            this._constraintApiProxyInputUpdater =
                this._constraint.setInputAttributes({
                    value: () => this.value,
                    maxlength: () => this.maxLength,
                    minlength: () => this.minLength,
                    disabled: () => this.disabled,
                    readonly: () => this.readOnly,
                    required: () => this.required,
                });
        }
        return this._constraintApi;
    }
    get computedAriaInvalid() {
        return computeAriaInvalid(this._helpMessage, this.value);
    }
    get helptextAlternativeText() {
        return formatLabel(i18n.helpTextAlternativeText, this.label);
    }
    get shouldRenderTextareaReadonly() {
        return features.enableTextareaReadOnlyStatic && this.readOnly && !this.disabled;
    }
    get shouldAddDataTextareaReadonlyAttr() {
        return this.shouldRenderTextareaReadonly ? '' : null;
    }
    get computedDivClass() {
        const { shouldRenderTextareaReadonly: isReadOnly } = this;
        return {
            'slds-form-element__control': true,
            'textarea-container': true,
            'slds-form-element__static': isReadOnly,
            'fix-slds-form-element__static': isReadOnly,
            'slds-grow': !isReadOnly,
        };
    }
    get computedTextareaClass() {
        const { shouldRenderTextareaReadonly: isReadOnly, disabled } = this;
        return {
            'slds-textarea': true,
            'fix-slds-textarea': isReadOnly,
            'slds-border_bottom': !disabled && isReadOnly,
        };
    }
    calculateTextareaHeight(inputElement, style) {
        const tempParaElem = document.createElement('p');
        Object.assign(tempParaElem.style, {
            visibility: 'hidden',
            position: 'absolute',
            whiteSpace: 'pre-wrap',
            wordBreak: 'break-word',
            padding: style.padding,
            font: style.font,
            letterSpacing: style.letterSpacing,
            lineHeight: style.lineHeight,
            width: style.width,
            boxSizing: style.boxSizing,
            border: style.border,
        });
        tempParaElem.textContent = inputElement.value || ' ';
        document.body.appendChild(tempParaElem);
        const height = tempParaElem.offsetHeight;
        document.body.removeChild(tempParaElem);
        return height;
    }
    setTextareaHeight(inputElement) {
        const LINE_HEIGHT = 21;
        const PAD_TOP = 4;
        const PAD_BTM = 2;
        const SINGLE_LINE = LINE_HEIGHT + PAD_TOP + PAD_BTM;
        let style = null;
        try {
            style = window.getComputedStyle(inputElement);
        }
        catch (e) {
            style = null;
        }
        if (!style) {
            inputElement.style.height = `${SINGLE_LINE}px`;
            return;
        }
        const value = inputElement.value || '';
        const measuredPixelHeight = this.calculateTextareaHeight(inputElement, style);
        const lineHeight = parseFloat(style.lineHeight) || LINE_HEIGHT;
        const padTop = parseFloat(style.paddingTop) || PAD_TOP;
        const padBottom = parseFloat(style.paddingBottom) || PAD_BTM;
        const heightSingleLine = lineHeight + padTop + padBottom;
        const heightToSet = !value || measuredPixelHeight <= heightSingleLine
            ? heightSingleLine
            : measuredPixelHeight;
        inputElement.style.height = `${heightToSet}px`;
    }
    get fieldSizingSupport() {
        if (import.meta.env.SSR) {
            return false;
        }
        try {
            return ((CSS &&
                CSS.supports &&
                CSS.supports('field-sizing', 'content')) ||
                false);
        }
        catch (e) {
            return false;
        }
    }
    updateTextareaHeightWhenReadonly() {
        const { _rendered, shouldRenderTextareaReadonly: isReadOnly, inputElement } = this;
        if (!_rendered ||
            !features.enableTextareaReadOnlyStatic ||
            !inputElement) {
            return;
        }
        const { fieldSizingSupport, value } = this;
        if (!fieldSizingSupport && isReadOnly) {
            this.setTextareaHeight(inputElement);
        }
        else if (fieldSizingSupport) {
            inputElement.classList.remove('fix-slds-textarea-height');
            if (isReadOnly && !value) {
                inputElement.classList.add('fix-slds-textarea-height');
            }
        }
    }
    toggleResizeObserver(wasReadOnly) {
        if (import.meta.env.SSR || !features.enableTextareaReadOnlyStatic) {
            return;
        }
        const { shouldRenderTextareaReadonly: isReadOnly, _resizeObserver, _rendered } = this;
        if (!isReadOnly && wasReadOnly && _rendered && _resizeObserver) {
            this.inputElement.style.height = null;
            this.removeResizeObserver();
            return;
        }
        if (isReadOnly && !wasReadOnly && !this.fieldSizingSupport) {
            this.setupResizeObserver();
        }
    }
    shouldSetupResizeObserver() {
        return (this._rendered &&
            this.shouldRenderTextareaReadonly &&
            features.enableTextareaReadOnlyStatic &&
            !this.fieldSizingSupport &&
            !this._resizeObserver &&
            this.inputElement);
    }
    setupResizeObserver() {
        if (!this.shouldSetupResizeObserver()) {
            return;
        }
        this._resizeObserver = new LightningResizeObserver(() => {
            this.updateTextareaHeightWhenReadonly();
        });
        this._resizeObserver.observe(this.inputElement);
    }
    removeResizeObserver() {
        if (!features.enableTextareaReadOnlyStatic) {
            return;
        }
        const { _resizeObserver } = this;
        if (_resizeObserver) {
            _resizeObserver.disconnect();
            this._resizeObserver = null;
        }
    }
    constructor() {
        super();
        this.ariaObserver = new AriaObserver(this);
        setContext(this, {
            textarea: null,
            template: super.template,
        });
    }
    connectedCallback() {
        assertContext(this);
        super.connectedCallback();
        if (!this.ariaObserver) {
            this.ariaObserver = new AriaObserver(this);
        }
        this.updateHostClassList();
        this._connected = true;
        this.interactingState = new InteractingState();
        this.interactingState.onleave(() => this.showHelpMessageIfInvalid());
        this._isNativeShadow = isNativeComponent(this);
    }
    renderedCallback() {
        if (this.isConnected) {
            this.ariaObserver.sync(this._isNativeShadow);
        }
        if (!this._rendered) {
            this.connectAriaDescribedBy();
            this.connectAriaLabelledBy();
            this._rendered = true;
            this._setInputValue(this._defaultValue);
            const scrollTarget = this.template.querySelector('.textarea-container');
            this.touchScroller = new TouchScroller(scrollTarget);
        }
        this.setupResizeObserver();
        this.updateTextareaHeightWhenReadonly();
        if (this._helpMessageChanged) {
            this.connectAriaDescribedBy();
            this._helpMessageChanged = false;
        }
    }
    disconnectedCallback() {
        this._connected = false;
        if (this.ariaObserver && this.ariaObserver.disconnect) {
            this.ariaObserver.disconnect();
            this.ariaObserver = null;
        }
        this.removeResizeObserver();
    }
    handleFocus() {
        this.interactingState.enter();
        this.dispatchEvent(new CustomEvent('focus'));
    }
    handleBlur() {
        this.interactingState.leave();
        this.dispatchEvent(new CustomEvent('blur'));
    }
    handleChange(event) {
        event.stopPropagation();
    }
    handleInput(event) {
        event.stopPropagation();
        const inputElement = event.target;
        const inputLength = inputElement.value.length;
        const warningId = 'char-limit-warning';
        let warningEl = this.template.querySelector(`#${warningId}`);
        if (inputLength >= this.maxLength) {
            if (!warningEl) {
                warningEl = document.createElement('p');
                warningEl.textContent = `Limit reached. You can only use ${this.maxLength} characters in this field.`;
                warningEl.setAttribute('role', 'alert');
                warningEl.id = warningId;
                warningEl.classList.add('slds-assistive-text');
                inputElement.after(warningEl);
            }
        }
        else if (warningEl) {
            warningEl.remove();
        }
        if (!this._connected ||
            this._value === event.target.value) {
            return;
        }
        this.interactingState.interacting();
        this._value = this.inputElement.value;
        this._updateProxyInputAttributes('value');
        this.dispatchEvent(new CustomEvent('change', {
            bubbles: true,
            composed: true,
            detail: {
                value: this._value,
            },
        }));
    }
    connectAriaDescribedBy() {
        const helpMessage = !import.meta.env.SSR
            ? this.template.querySelector('[data-help-message]')
            : null;
        this.ariaObserver.connect({
            targetSelector: 'textarea',
            attribute: 'aria-describedby',
            relatedNodeIds: this.ariaDescribedBy,
            relatedNodes: helpMessage,
        });
    }
    connectAriaLabelledBy() {
        this.ariaObserver.connect({
            targetSelector: 'textarea',
            attribute: 'aria-labelledby',
            relatedNodeIds: this.ariaLabelledBy,
        });
    }
    _setInputValue(value) {
        assertContext(this);
        setDecoratedDragonInputValueWithoutEvent(this.inputElement, value);
    }
    _updateProxyInputAttributes(attributes) {
        if (this._constraintApiProxyInputUpdater) {
            this._constraintApiProxyInputUpdater(attributes);
        }
    }
    updateHostClassList() {
        classListMutation(this.classList, {
            'slds-form-element': true,
            'slds-form-element_stacked': this.variant === VARIANT.LABEL_STACKED,
            'slds-form-element_horizontal': this.variant === VARIANT.LABEL_INLINE,
        });
    }
}
LightningTextarea['interopMap'] = {
    exposeNativeEvent: {
        change: true,
        focus: true,
        blur: true,
    },
};
