// lightning-base-components/src/lightning/positionLibrary/direction.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { normalizeString } from 'lightning/utilsPrivate';
import { WindowManager } from './util';
import { features } from 'lightning/features';
const ALIGN_REGEX = /^(left|right|center)\s(top|bottom|center)$/;
export const Direction = {
    Center: 'center',
    Middle: 'middle',
    Right: 'right',
    Left: 'left',
    Bottom: 'bottom',
    Top: 'top',
    Default: 'default',
};
const VerticalMap = {
    top: Direction.Top,
    bottom: Direction.Bottom,
    center: Direction.Middle,
};
const HorizontalMap = {
    left: Direction.Left,
    right: Direction.Right,
    center: Direction.Center,
};
const FlipMap = {
    left: Direction.Right,
    right: Direction.Left,
    top: Direction.Bottom,
    bottom: Direction.Top,
    center: Direction.Center,
    default: Direction.Right,
};
function getContainerSize(parent) {
    if (parent) {
        return parent.getBoundingClientRect();
    }
    const rect = {
        width: WindowManager.window.innerWidth || document.body.clientWidth || 0,
        height: WindowManager.window.innerHeight || document.body.clientHeight || 0,
        top: 0,
        left: 0,
    };
    rect.bottom = rect.height;
    rect.right = rect.width;
    return rect;
}
export function normalizeDirection(direction, defaultValue) {
    return normalizeString(direction, {
        fallbackValue: defaultValue || Direction.Default,
        validValues: [
            Direction.Center,
            Direction.Right,
            Direction.Left,
            Direction.Bottom,
            Direction.Top,
            Direction.Middle,
            Direction.Default,
        ],
    });
}
export function mapToHorizontal(value) {
    value = normalizeDirection(value, Direction.Left);
    return HorizontalMap[value];
}
export function mapToVertical(value) {
    value = normalizeDirection(value, Direction.Left);
    return VerticalMap[value];
}
export function flipDirection(value) {
    value = normalizeDirection(value, Direction.Left);
    return FlipMap[value];
}
export function isValidDirection(value) {
    return value && value.match(ALIGN_REGEX);
}
export function checkFlipPossibility(parent, element, target, leftAsBoundary) {
    const viewPort = getContainerSize(parent);
    const windowViewPort = getContainerSize();
    const elemRect = element.getBoundingClientRect();
    const referenceElemRect = target.getBoundingClientRect();
    const height = typeof elemRect.height !== 'undefined'
        ? elemRect.height
        : elemRect.bottom - elemRect.top;
    const width = typeof elemRect.width !== 'undefined'
        ? elemRect.width
        : elemRect.right - elemRect.left;
    let rightAsBoundary = false;
    let isRTL = false;
    if (document.dir === 'rtl') {
        rightAsBoundary = leftAsBoundary;
        leftAsBoundary = false;
        isRTL = true;
    }
    const aboveSpace = referenceElemRect.top - viewPort.top - height;
    const belowSpace = viewPort.height - referenceElemRect.bottom - height;
    let hasSpaceAbove = aboveSpace >= 0 ||
        (parent == null && belowSpace < 0 && aboveSpace > belowSpace);
    let hasSpaceBelow = belowSpace >= 0 || (aboveSpace < 0 && belowSpace > aboveSpace);
    let shouldAlignToRight = false;
    if (parent) {
        if (!features.enablePositionLibFlipCalcWithViewportRight) {
            shouldAlignToRight =
                referenceElemRect.right - viewPort.left >= width &&
                    referenceElemRect.left + width >
                        (rightAsBoundary ? referenceElemRect.right : viewPort.width);
        }
        else {
            shouldAlignToRight =
                referenceElemRect.right - viewPort.left >= width &&
                    referenceElemRect.left + width >
                        (rightAsBoundary ? referenceElemRect.right : viewPort.right);
        }
    }
    else {
        shouldAlignToRight =
            referenceElemRect.right >= width &&
                referenceElemRect.left + width >
                    (rightAsBoundary ? referenceElemRect.right : viewPort.width);
    }
    let shouldAlignToLeft = false;
    if (parent) {
        shouldAlignToLeft =
            referenceElemRect.left + width <= viewPort.right &&
                referenceElemRect.right - viewPort.left - width <
                    (leftAsBoundary ? referenceElemRect.left : 0);
    }
    else {
        shouldAlignToLeft =
            referenceElemRect.left + width <= viewPort.width &&
                referenceElemRect.right - width <
                    (leftAsBoundary ? referenceElemRect.left : 0);
    }
    const spaceRemainingRight = viewPort.width - referenceElemRect.left;
    const spaceRemainingLeft = referenceElemRect.right;
    if (!shouldAlignToLeft && !shouldAlignToRight && parent) {
        shouldAlignToRight =
            (referenceElemRect.right - viewPort.left >= width ||
                referenceElemRect.right >= width) &&
                referenceElemRect.left + width >
                    (rightAsBoundary ? referenceElemRect.right : viewPort.width);
        shouldAlignToLeft =
            (referenceElemRect.left + width <= viewPort.right ||
                referenceElemRect.left + width <= windowViewPort.width) &&
                referenceElemRect.right - width <
                    (leftAsBoundary ? referenceElemRect.left : 0);
    }
    else if (!shouldAlignToLeft &&
        !shouldAlignToRight &&
        spaceRemainingRight < width &&
        spaceRemainingLeft < width) {
        shouldAlignToRight = spaceRemainingLeft > spaceRemainingRight;
        shouldAlignToLeft = spaceRemainingLeft < spaceRemainingRight;
    }
    if (isRTL &&
        !shouldAlignToLeft &&
        !shouldAlignToRight &&
        element.tagName === 'LIGHTNING-PRIMITIVE-BUBBLE') {
        shouldAlignToRight = spaceRemainingLeft > spaceRemainingRight;
        shouldAlignToLeft = spaceRemainingLeft < spaceRemainingRight;
    }
    if (!hasSpaceAbove &&
        !hasSpaceBelow &&
        parent &&
        element.style &&
        element.style.position === 'fixed') {
        hasSpaceAbove = referenceElemRect.top - windowViewPort.top - height > 0;
        hasSpaceBelow =
            windowViewPort.height - referenceElemRect.bottom - height > 0;
    }
    const centerOverflow = {
        left: referenceElemRect.left - (width * 0.5) < 0,
        right: referenceElemRect.right + (width * 0.5) > viewPort.width,
        top: referenceElemRect.top - (height * 0.5) < 0,
        bottom: referenceElemRect.bottom + (height * 0.5) > viewPort.height,
    };
    return {
        shouldAlignToLeft,
        shouldAlignToRight,
        hasSpaceAbove,
        hasSpaceBelow,
        centerOverflow,
    };
}
