// lightning-base-components/src/lightning/positionLibrary/elementProxy.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { isInDom, WindowManager, pxToFloat } from './util';
import { isSafari } from 'lightning/utilsPrivate';
import { features } from 'lightning/features';
import { privateContext } from 'lightning/utilsInternal';
const { setContext, assertContext } = privateContext;
export class ElementProxy {
    constructor(el, id) {
        setContext(this);
        this.id = id;
        this.width = 0;
        this.height = 0;
        this.left = 0;
        this.top = 0;
        this.right = 0;
        this.bottom = 0;
        this._dirty = false;
        this._node = null;
        this._releaseCb = null;
        if (!el) {
            throw new Error('Element missing');
        }
        if (WindowManager.isWindow(el)) {
            el = WindowManager.window;
        }
        this._node = el;
        this.setupObserver();
        this.refresh();
    }
    setupObserver() {
        if (WindowManager.MutationObserver && !this._node.isObserved) {
            this._observer = new WindowManager.MutationObserver(this.refresh.bind(this));
            if (!WindowManager.isWindow(this._node)) {
                this._observer.observe(this._node, {
                    attributes: true,
                    childList: true,
                    characterData: true,
                    subtree: true,
                });
                this._node.isObserved = true;
            }
        }
    }
    setReleaseCallback(cb, scope) {
        if (features.enableAssertContext260Round2) {
            assertContext(this);
        }
        const scopeObj = scope || this;
        this._releaseCb = cb.bind(scopeObj);
    }
    checkNodeIsInDom() {
        if (!isInDom(this._node)) {
            return false;
        }
        return true;
    }
    refresh() {
        const w = WindowManager.window;
        if (!this.isDirty()) {
            if (!this.checkNodeIsInDom()) {
                return this.release();
            }
            let box, x, scrollTop, scrollLeft;
            if (typeof w.pageYOffset !== 'undefined') {
                scrollTop = w.pageYOffset;
                scrollLeft = w.pageXOffset;
            }
            else {
                scrollTop = w.scrollY;
                scrollLeft = w.scrollX;
            }
            if (!WindowManager.isWindow(this._node)) {
                const offsetHeight = this._node.offsetHeight;
                box = this._node.getBoundingClientRect();
                for (x in box) {
                    this[x] = Math.floor(box[x]);
                }
                this.setAdjustedBoundingRect(this, box);
                this.top = Math.floor(this.top + scrollTop);
                this.bottom = Math.floor(this.top + box.height);
                this.left = Math.floor(this.left + scrollLeft);
                this.right = Math.floor(this.left + box.width);
            }
            else {
                box = {};
                this.width = WindowManager.documentElement.clientWidth;
                this.height = WindowManager.documentElement.clientHeight;
                this.left = scrollLeft;
                this.top = scrollTop;
                this.right =
                    WindowManager.documentElement.clientWidth + scrollLeft;
                this.bottom = WindowManager.documentElement.clientHeight;
            }
            this._dirty = false;
        }
        return this._dirty;
    }
    getNode() {
        return this._node;
    }
    isDirty() {
        return this._dirty;
    }
    getElementZoomRatio(style) {
        const element = this._node;
        if (features.enablePositionLibraryTransformScaleFix) {
            if (!import.meta.env.SSR && 'currentCSSZoom' in Element.prototype) {
                return element.currentCSSZoom;
            }
            return parseFloat(style.zoom) || 1;
        }
        const rect = element.getBoundingClientRect();
        const offsetWidth = element.offsetWidth || Math.round(rect.width) || 1;
        const offsetHeight = element.offsetHeight || Math.round(rect.height) || 1;
        const actualWidth = Math.round(rect.width);
        const actualHeight = Math.round(rect.height);
        const widthRatio = Number((actualWidth / offsetWidth).toFixed(2));
        const heightRatio = Number((actualHeight / offsetHeight).toFixed(2));
        return Math.min(widthRatio, heightRatio);
    }
    setAdjustedBoundingRect(varToSet, boundingRect, computedStyle, zoomLevel) {
        if (features.enablePositionLibraryTransformScaleFix &&
            !import.meta.env.SSR &&
            isSafari) {
            const style = computedStyle || getComputedStyle(this._node);
            const zoom = zoomLevel || this.getElementZoomRatio(style);
            if (zoom !== 1) {
                const rect = boundingRect || this._node.getBoundingClientRect();
                for (const x in rect) {
                    varToSet[x] = rect[x] * zoom;
                }
                return true;
            }
        }
        return false;
    }
    bake() {
        const w = WindowManager.window;
        let absPos = {};
        const style = w.getComputedStyle(this._node) || this._node.style;
        const zoomLevel = this.getElementZoomRatio(style);
        const setAbsPos = this.setAdjustedBoundingRect(absPos, null, style, zoomLevel);
        if (!setAbsPos) {
            absPos = this._node.getBoundingClientRect();
        }
        const hasPageOffset = typeof w.pageYOffset !== 'undefined';
        const scrollTop = hasPageOffset ? w.pageYOffset : w.scrollY;
        const scrollLeft = hasPageOffset ? w.pageXOffset : w.scrollX;
        const originalLeft = style.left.match(/auto|fixed/)
            ? '0'
            : pxToFloat(style.left);
        const originalTop = style.top.match(/auto|fixed/)
            ? '0'
            : pxToFloat(style.top);
        const adjustForZoom = (value) => value / zoomLevel;
        const marginTop = adjustForZoom(pxToFloat(style.marginTop));
        const adjustedAbsPosTop = adjustForZoom(Math.round(absPos.top)) +
            scrollTop -
            (adjustForZoom(this.top) <
                adjustForZoom(Math.round(absPos.top)) + scrollTop
                ? marginTop
                : -marginTop);
        const marginAdjustment = adjustForZoom(this.top) < adjustForZoom(adjustedAbsPosTop)
            ? marginTop
            : -marginTop;
        const zoomAdjustedAbsLeft = adjustForZoom(Math.round(absPos.left) + scrollLeft);
        const zoomAdjustedAbsTop = adjustForZoom(Math.round(absPos.top) + scrollTop);
        let leftDif = Math.round(adjustForZoom(this.left) - zoomAdjustedAbsLeft);
        const topDif = Math.round(adjustForZoom(this.top) - (zoomAdjustedAbsTop + marginAdjustment));
        const viewPortWidth = w.innerWidth || document.body.clientWidth;
        let rightPos = parseInt(style.right, 10);
        const elementWidth = this._node.offsetWidth;
        const zoomAdjustedWidth = elementWidth * zoomLevel;
        if (originalLeft + leftDif + zoomAdjustedWidth > viewPortWidth &&
            rightPos <= 0) {
            rightPos = rightPos < 0 ? -rightPos : leftDif;
            this._node.style.right = `${-rightPos}px`;
            leftDif = 0;
        }
        this._node.style.left = `${originalLeft + leftDif}px`;
        this._node.style.top = `${originalTop + topDif}px`;
        if (this._restoreSize) {
            if (this.originalHeight === undefined) {
                this.originalHeight = this._node.style.height;
            }
            if (this.originalWidth === undefined) {
                this.originalWidth = this._node.style.width;
            }
            const adjustedWidth = adjustForZoom(this.width);
            const adjustedHeight = adjustForZoom(this.height);
            this._node.style.width = `${adjustedWidth}px`;
            this._node.style.height = `${adjustedHeight}px`;
        }
        this._dirty = false;
    }
    setDirection(direction, val) {
        this[direction] = val;
        this._dirty = true;
        if (direction === 'height' || direction === 'width') {
            this._restoreSize = true;
        }
    }
    release() {
        if (this._restoreSize) {
            this._node.style.width = this.originalWidth;
            this._node.style.height = this.originalHeight;
            if (this._removeMinHeight) {
                this._node.style.minHeight = '';
            }
        }
        if (this._releaseCb) {
            this._releaseCb(this);
        }
        if (this._observer) {
            this._observer.disconnect();
            this._observer = null;
        }
    }
    querySelectorAll(selector) {
        return this._node.querySelectorAll(selector);
    }
}
