// lightning-base-components/src/lightning/primitiveInputColor/primitiveInputColor.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { api } from 'lwc';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import { classSet } from 'lightning/utils';
import { VARIANT } from 'lightning/inputUtils';
import { reflectAttribute, isNativeComponent } from 'lightning/utilsPrivate';
import { privateContext } from 'lightning/utilsInternal';
import { features } from 'lightning/features';
const { setContext, assertContext } = privateContext;
export default class LightningPrimitiveInputColor extends LightningShadowBaseClass {
    @api
    hasExternalLabel;
    @api
    required;
    @api
    accessKey;
    @api
    label;
    @api
    fieldLevelHelp;
    @api
    helptextAlternativeText;
    @api
    helpMessage;
    @api
    disabled;
    @api
    name;
    @api
    autocomplete;
    @api
    ariaLabel;
    @api
    ariaInvalid;
    @api
    ariaDisabled;
    @api
    ariaKeyShortcuts;
    @api
    placeholder;
    @api
    pattern;
    @api
    variant;
    _value = '';
    rendered = false;
    initialValueSet = false;
    constructor() {
        super();
        if (features.enableAssertContext260) {
            setContext(this);
        }
    }
    @api
    focus() {
        if (this.rendered) {
            this.inputElement.focus();
        }
    }
    @api
    get value() {
        return this._value;
    }
    set value(value) {
        this._value = this.normalizeInput(value);
        if (this.rendered && this.inputElement.value !== this._value) {
            this.setInputValue(this._value);
        }
    }
    @api
    get inputElement() {
        if (!this.cachedInputElement) {
            let inputElement = this.template.querySelector('input');
            this.cachedInputElement = inputElement;
        }
        return this.cachedInputElement;
    }
    @api
    get ariaDescribedByElements() {
        return this.template.querySelector('[data-help-message]');
    }
    @api
    get ariaErrorMessageElement() {
        return this.template.querySelector('[data-help-message]');
    }
    @api
    get isNativeShadow() {
        return this._isNativeShadow;
    }
    connectedCallback() {
        super.connectedCallback();
        this._isNativeShadow = isNativeComponent(this);
    }
    disconnectedCallback() {
        this.rendered = false;
        this.initialValueSet = false;
        this.cachedInputElement = undefined;
    }
    renderedCallback() {
        this.rendered = true;
        if (!this.initialValueSet) {
            this.inputElement.value = this._value;
            this.initialValueSet = true;
        }
        reflectAttribute(this, 'disabled', this.disabled);
        reflectAttribute(this, 'invalid', !!this.helpMessage);
        reflectAttribute(this, 'variant', this.variant);
    }
    get computedColorLabelClass() {
        return classSet('slds-form-element__label slds-color-picker__summary-label').add({ 'slds-assistive-text': this.isLabelHidden });
    }
    get isLabelHidden() {
        return this.variant === VARIANT.LABEL_HIDDEN;
    }
    handleBlur() {
        this.dispatchEvent(new CustomEvent('blur'));
    }
    handleFocus() {
        this.dispatchEvent(new CustomEvent('focus'));
    }
    handleChange(event) {
        event.stopPropagation();
        this.dispatchCommitEvent();
        this.dispatchChangeEvent();
    }
    handleColorChange(event) {
        const selectedColor = event.detail.color;
        if (selectedColor !== this.inputElement.value) {
            this.setInputValue(selectedColor);
            this._value = selectedColor;
            this.focus();
            this.dispatchChangeEventWithDetail({ value: selectedColor });
            this.dispatchCommitEvent();
        }
        this.template
            .querySelector('lightning-primitive-colorpicker-button')
            .focus();
    }
    handleInput(event) {
        event.stopPropagation();
        this.dispatchChangeEvent();
    }
    dispatchChangeEventWithDetail(detail) {
        this.dispatchEvent(new CustomEvent('change', {
            composed: true,
            bubbles: true,
            detail,
        }));
    }
    dispatchChangeEvent() {
        const detail = { value: this.inputElement.value };
        this._value = detail.value;
        this.dispatchChangeEventWithDetail(detail);
    }
    dispatchCommitEvent() {
        this.dispatchEvent(new CustomEvent('commit'));
    }
    setInputValue(value) {
        if (features.enableAssertContext260) {
            assertContext(this);
        }
        this.inputElement.value = value;
    }
    normalizeInput(value) {
        if (typeof value === 'number' || typeof value === 'string') {
            return String(value);
        }
        return '';
    }
}
