// lightning-base-components/src/lightning/overlay/overlay.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { api } from 'lwc';
import LightningPrimitiveOverlay from 'lightning/primitiveOverlay';
import { parent, instanceName, secure } from 'lightning/overlayUtils';
import LightningOverlayBase from 'lightning/overlayBase';
import { normalizeString, normalizeBoolean } from 'lightning/utilsPrivate';
import { DEFAULT_POPOVER, VALID_POPOVER_VALUES, DEFAULT_POSITION, VALID_POSITION_VALUES, } from 'lightning/overlayPositionUtils';
export default class LightningOverlay extends LightningPrimitiveOverlay {
    static [parent] = LightningOverlayBase;
    static [instanceName] = 'lightning-overlay';
    @api
    source = null;
    @api
    minWidth = '1rem';
    @api
    maxWidth = 'auto';
    @api
    minHeight = '1rem';
    @api
    maxHeight = 'auto';
    @api
    offsetInline = 0;
    @api
    offsetBlock = 0;
    _popover = DEFAULT_POPOVER;
    _position = DEFAULT_POSITION;
    _enableFocusTrap = false;
    @api
    get popover() {
        return this._popover;
    }
    set popover(value) {
        this._popover = normalizeString(value, {
            fallbackValue: DEFAULT_POPOVER,
            validValues: VALID_POPOVER_VALUES,
        });
    }
    @api
    get position() {
        return this._position;
    }
    set position(value) {
        this._position = normalizeString(value, {
            fallbackValue: DEFAULT_POSITION,
            validValues: VALID_POSITION_VALUES,
        });
    }
    @api
    get enableFocusTrap() {
        return this._enableFocusTrap;
    }
    set enableFocusTrap(value) {
        this._enableFocusTrap = normalizeBoolean(value);
    }
    @api
    close(result, maybeSecure) {
        if (maybeSecure === secure) {
            super.close(result);
        }
        else {
            const rootNode = this.template.host.getRootNode();
            const base = rootNode.host;
            base.hideSurfacePopover(result);
        }
    }
    connectedCallback() {
        if (this.constructor.name === 'LightningOverlay') {
            throw new Error('Use of <lightning-overlay> as an element is not supported. See developer documentation for extending the LightningOverlay class.');
        }
    }
}
