// lightning-base-components/src/lightning/dualListbox/keyboard.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { keyCodes, runActionOnBufferedTypedCharacters, } from 'lightning/utilsPrivate';
import { features } from 'lightning/features';
function isOptionDisabled(optionEl) {
    return !!(optionEl?.getAttribute('aria-disabled') === 'true');
}
function preventDefaultAndStopPropagation(event) {
    event.preventDefault();
    event.stopPropagation();
}
function setFocusOnNextOption(option, moveUp, intf) {
    const index = parseInt(option.getAttribute('data-index'), 10);
    const i = index + (moveUp ? -1 : 1);
    const options = intf.getElementsOfList(option.getAttribute('data-type'));
    const next = options[i];
    if (next) {
        next.focus();
    }
}
function selectNextOption(option, moveUp, intf) {
    const selected = option.getAttribute('aria-selected') === 'true';
    const isLockedOption = option.getAttribute('aria-disabled') === 'true';
    const index = parseInt(option.getAttribute('data-index'), 10);
    const shouldSkipLocked = !features.enableDisabledOptDualListbox && isLockedOption;
    const shouldMoveFromLockedOption = features.enableDisabledOptDualListbox && isLockedOption;
    const i = index +
        (selected || shouldSkipLocked || shouldMoveFromLockedOption
            ? moveUp
                ? -1
                : 1
            : 0);
    const options = intf.getElementsOfList(option.getAttribute('data-type'));
    const next = options[i];
    if (!features.enableDisabledOptDualListbox && isOptionDisabled(next)) {
        selectNextOption(next, moveUp, intf);
        return;
    }
    if (next) {
        intf.updateSelectedOptions(next, true, false);
    }
}
function selectNextOptionFromShift(option, moveUp, isMultiple, intf) {
    const curr = parseInt(option.getAttribute('data-index'), 10);
    if (intf.getShiftIndex() < 0) {
        intf.setShiftIndex(curr);
        intf.setLastShift(moveUp);
    }
    const next = curr + (intf.getLastShift() !== moveUp ? 0 : moveUp ? -1 : 1);
    const pos = next < intf.getShiftIndex();
    const shiftAdd = pos === moveUp || intf.getShiftIndex() === next;
    const options = intf.getElementsOfList(option.getAttribute('data-type'));
    const nextOption = options[next];
    if (!features.enableDisabledOptDualListbox && isOptionDisabled(nextOption)) {
        selectNextOptionFromShift(nextOption, moveUp, isMultiple, intf);
        return;
    }
    if (nextOption) {
        intf.updateSelectedOptions(nextOption, shiftAdd, true);
        intf.setLastShift(moveUp);
    }
}
function isTypeaheadEvent(event, isShiftKeyDown) {
    const { key } = event;
    if (!features.enableDualListboxSingleCharTypeahead ||
        !key ||
        key.length > 1) {
        return false;
    }
    const result = isShiftKeyDown
        ? /^\p{L}$/u.test(key)
        : /^[\p{L}\p{N}]$/u.test(key);
    return result;
}
function handleTypeaheadChar({ event, keyboardInterface }) {
    if (import.meta.env.SSR) {
        return;
    }
    const { target, key } = event;
    requestAnimationFrame(() => runActionOnBufferedTypedCharacters(event, keyboardInterface.focusOptionByFirstChar.bind(this, target, key)));
}
export function handleKeyDownOnOption(event, keyboardInterface) {
    const { keyCode, metaKey, ctrlKey, shiftKey, target } = event;
    if (metaKey || ctrlKey) {
        keyboardInterface.setShiftIndex(-1);
        const keyCodesA = 'A'.charCodeAt(0);
        const selected = target.getAttribute('aria-selected') === 'true';
        switch (keyCode) {
            case keyCodes.up:
                preventDefaultAndStopPropagation(event);
                setFocusOnNextOption(target, true, keyboardInterface);
                break;
            case keyCodes.down:
                preventDefaultAndStopPropagation(event);
                setFocusOnNextOption(target, false, keyboardInterface);
                break;
            case keyCodes.right:
                preventDefaultAndStopPropagation(event);
                keyboardInterface.moveOptionsBetweenLists(true);
                break;
            case keyCodes.left:
                preventDefaultAndStopPropagation(event);
                keyboardInterface.moveOptionsBetweenLists(false);
                break;
            case keyCodes.space:
                preventDefaultAndStopPropagation(event);
                keyboardInterface.updateSelectedOptions(target, !selected, true);
                break;
            case keyCodesA:
                preventDefaultAndStopPropagation(event);
                keyboardInterface.selectAllOptions(target);
                break;
            default:
        }
    }
    else if (shiftKey) {
        if (isTypeaheadEvent(event, true)) {
            handleTypeaheadChar({ event, keyboardInterface });
            return;
        }
        switch (keyCode) {
            case keyCodes.up:
                preventDefaultAndStopPropagation(event);
                selectNextOptionFromShift(target, true, true, keyboardInterface);
                break;
            case keyCodes.down:
                preventDefaultAndStopPropagation(event);
                selectNextOptionFromShift(target, false, true, keyboardInterface);
                break;
            default:
        }
    }
    else {
        if (isTypeaheadEvent(event, false)) {
            handleTypeaheadChar({ event, keyboardInterface });
            return;
        }
        keyboardInterface.setShiftIndex(-1);
        switch (keyCode) {
            case keyCodes.up:
                preventDefaultAndStopPropagation(event);
                selectNextOption(target, true, keyboardInterface);
                break;
            case keyCodes.down:
                preventDefaultAndStopPropagation(event);
                selectNextOption(target, false, keyboardInterface);
                break;
            default:
        }
    }
}
