// lightning-base-components/src/lightning/combobox/combobox.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import labelRequired from '@salesforce/label/LightningControl.required';
import labelPlaceholder from '@salesforce/label/LightningCombobox.placeholder';
import labelHelpTextAlternativeText from '@salesforce/label/LightningInput.helptextAlternativeText';
import { api, track } from 'lwc';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import { classSet, formatLabel } from 'lightning/utils';
import { classListMutation, normalizeArray, normalizeBoolean, reflectAttribute, synchronizeAttrs, isNativeComponent, computeAriaInvalid, } from 'lightning/utilsPrivate';
import { privateContext } from 'lightning/utilsInternal';
import { isEmptyString, InteractingState, FieldConstraintApi, normalizeVariant, VARIANT, } from 'lightning/inputUtils';
import { features } from 'lightning/features';
const { setContext, assertContext } = privateContext;
const i18n = {
    required: labelRequired,
    placeholder: labelPlaceholder,
    helpTextAlternativeText: labelHelpTextAlternativeText,
};
export default class LightningCombobox extends LightningShadowBaseClass {
    static delegatesFocus = true;
    static validationOptOut = ['class'];
    _ariaLabelledBy = '';
    _ariaDescribedBy = '';
    _fieldLevelHelp = '';
    _selectedLabel = '';
    _disabled = false;
    _readOnly = false;
    _spinnerActive = false;
    _required = false;
    @track
    _variant;
    @track
    _helpMessage;
    @track
    _labelForId;
    @track
    _ariaInvalid;
    @track
    _items = [];
    _options;
    _selectedItem;
    _constraintApi;
    selectedValue;
    connected;
    rendered;
    _isNativeShadow;
    interactingState;
    @api
    autocomplete;
    @api
    get ariaInvalid() {
        return this._ariaInvalid;
    }
    set ariaInvalid(value) {
        this._ariaInvalid =
            typeof value == 'undefined' ? undefined : normalizeBoolean(value);
    }
    get computedAriaInvalid() {
        return computeAriaInvalid(this._helpMessage, this.value, this.ariaInvalid);
    }
    @api
    label;
    @api
    dropdownAlignment = 'left';
    @api
    placeholder = i18n.placeholder;
    @api
    messageWhenValueMissing;
    @api
    name;
    constructor() {
        super();
        setContext(this);
    }
    renderedCallback() {
        if (!this.rendered) {
            this.rendered = true;
            this.updateValidity();
        }
        this.synchronizeA11y();
    }
    connectedCallback() {
        assertContext(this);
        super.connectedCallback();
        this._isNativeShadow = isNativeComponent(this);
        this.classList.add('slds-form-element');
        this.updateClassList();
        this.interactingState = new InteractingState();
        this.interactingState.onleave(() => this.showHelpMessageIfInvalid());
        this.connected = true;
        this._items = this.generateItems(this.options);
        if (this.options && this.selectedValue !== undefined) {
            this.updateSelectedOptions();
        }
    }
    updateClassList() {
        classListMutation(this.classList, {
            'slds-form-element_stacked': this.variant === VARIANT.LABEL_STACKED,
            'slds-form-element_horizontal': this.variant === VARIANT.LABEL_INLINE,
        });
    }
    disconnectedCallback() {
        this.connected = false;
    }
    @api
    get ariaLabelledBy() {
        return this._ariaLabelledBy;
    }
    set ariaLabelledBy(labelledBy) {
        this._ariaLabelledBy = labelledBy;
    }
    get helpTextElement() {
        return !import.meta.env.SSR
            ? this.template.querySelector('[data-help-text]')
            : null;
    }
    @api
    get ariaDescribedBy() {
        return this._ariaDescribedBy;
    }
    set ariaDescribedBy(describedBy) {
        this._ariaDescribedBy = describedBy;
    }
    @api
    get fieldLevelHelp() {
        return this._fieldLevelHelp;
    }
    set fieldLevelHelp(value) {
        this._fieldLevelHelp = value;
    }
    @api
    get variant() {
        return this._variant || VARIANT.STANDARD;
    }
    set variant(value) {
        this._variant = normalizeVariant(value);
        this.updateClassList();
        reflectAttribute(this, 'variant', this._variant);
    }
    @api
    get value() {
        return this.selectedValue;
    }
    set value(newValue) {
        if (newValue !== this.selectedValue) {
            this.selectedValue = newValue;
            if (this.connected && this.options) {
                this.updateSelectedOptions();
            }
        }
    }
    @api
    get options() {
        return this._options || [];
    }
    set options(newValue) {
        this._options = normalizeArray(newValue);
        if (this.connected) {
            this._items = this.generateItems(this._options);
            this.updateSelectedOptions();
        }
    }
    @api
    get disabled() {
        return this._disabled || this._readOnly || false;
    }
    set disabled(value) {
        this._disabled = normalizeBoolean(value);
    }
    @api
    get readOnly() {
        return this.disabled;
    }
    set readOnly(value) {
        this._readOnly = normalizeBoolean(value);
    }
    @api
    get required() {
        return this._required;
    }
    set required(value) {
        this._required = normalizeBoolean(value);
        if (this.connected) {
            this.updateValidity();
        }
    }
    @api
    get spinnerActive() {
        return this._spinnerActive;
    }
    set spinnerActive(value) {
        this._spinnerActive = normalizeBoolean(value);
    }
    @api
    focus() {
        if (this.connected) {
            this.getBaseComboboxElement().focus();
        }
    }
    @api
    blur() {
        if (this.connected) {
            this.getBaseComboboxElement().blur();
        }
    }
    @api
    get validity() {
        return this._constraint.validity;
    }
    @api
    checkValidity() {
        return this._constraint.checkValidity();
    }
    @api
    reportValidity() {
        return this._constraint.reportValidity((message) => {
            reflectAttribute(this.template.querySelector('lightning-base-combobox'), 'invalid', !this.checkValidity());
            this._helpMessage = message;
        });
    }
    @api
    setCustomValidity(message) {
        this._constraint.setCustomValidity(message);
        this.updateValidity();
    }
    @api
    showHelpMessageIfInvalid() {
        this.reportValidity();
    }
    handleComboboxReady(e) {
        this._labelForId = e.detail.id;
    }
    handleLabelClicked() {
        if (this._isNativeShadow) {
            const combobox = this.getBaseComboboxElement();
            combobox.blur();
            setTimeout(() => {
                combobox.focusAndOpenDropdownIfNotEmpty();
            }, 50);
        }
    }
    synchronizeA11y() {
        synchronizeAttrs(this.template.querySelector('label'), {
            for: this._labelForId,
        });
        const baseCombobox = this.template.querySelector('lightning-base-combobox');
        baseCombobox.inputLabelledByElement = this.ariaLabelledBy;
        baseCombobox.inputDescribedByElements = this.ariaDescribedBy;
    }
    get templateRootNode() {
        return !import.meta.env.SSR ? this.template.host.getRootNode() : null;
    }
    get i18n() {
        return i18n;
    }
    get isLabelHidden() {
        return this.variant === VARIANT.LABEL_HIDDEN;
    }
    get computedLabelClass() {
        return classSet('slds-form-element__label')
            .add({ 'slds-assistive-text': this.isLabelHidden })
            .toString();
    }
    handleSelect(event) {
        if (event.detail.value === this.selectedValue) {
            return;
        }
        this.selectedValue = event.detail.value;
        this.updateSelectedOptions();
        this.dispatchEvent(new CustomEvent('change', {
            composed: true,
            bubbles: true,
            detail: {
                value: this.selectedValue,
            },
        }));
    }
    handleFocus() {
        this.interactingState.enter();
        this.dispatchEvent(new CustomEvent('focus'));
    }
    handleBlur() {
        this.interactingState.leave();
        this.dispatchEvent(new CustomEvent('blur'));
    }
    handleDropdownOpen() {
        this.dispatchEvent(new CustomEvent('open'));
    }
    updateSelectedOptions() {
        assertContext(this);
        this.updateSelectedLabelFromValue(this.selectedValue);
        this.markOptionSelectedFromValue(this.selectedValue);
        this.updateValidity();
    }
    updateValidity() {
        if (!features.enableComboboxElementInternals || import.meta.env.SSR) {
            return;
        }
        const baseCombobox = this.getBaseComboboxElement();
        if (baseCombobox) {
            const { _constraint } = this;
            baseCombobox.setValidity(_constraint.validity, _constraint.validationMessage);
        }
    }
    markOptionSelectedFromValue(value) {
        assertContext(this);
        if (this._items) {
            const selectedItem = this._items.find((item) => item.value === value);
            if (this._selectedItem) {
                this._selectedItem.checked = false;
                this._selectedItem.iconName = undefined;
                this._selectedItem.highlight = false;
            }
            this._selectedItem = selectedItem;
            if (selectedItem) {
                selectedItem.iconName = 'utility:check';
                this._selectedItem.highlight = true;
                this._selectedItem.checked = true;
            }
            this._items = this._items.slice();
        }
    }
    updateSelectedLabelFromValue(newValue) {
        this._selectedLabel = this.getOptionLabelByValue(newValue);
    }
    getOptionLabelByValue(value) {
        const foundOption = this.options.find((option) => option.value === value);
        if (foundOption) {
            return foundOption.label;
        }
        return '';
    }
    generateItems(options) {
        return options.map((option) => {
            const type = option.description ? 'option-card' : 'option-inline';
            return {
                type,
                text: option.label,
                subText: option.description,
                highlight: this.value === option.value,
                iconSize: 'x-small',
                value: option.value,
            };
        });
    }
    getBaseComboboxElement() {
        return this.template.querySelector('lightning-base-combobox');
    }
    get _constraint() {
        if (!this._constraintApi) {
            this._constraintApi = new FieldConstraintApi(() => this, {
                valueMissing: () => !this.disabled &&
                    this.required &&
                    isEmptyString(this.selectedValue),
            });
        }
        return this._constraintApi;
    }
    get helptextAlternativeText() {
        return formatLabel(i18n.helpTextAlternativeText, this.label);
    }
}
