// lightning-base-components/src/lightning/overlayBase/overlayBase.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { api } from 'lwc';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import { instanceName, secure } from 'lightning/overlayUtils';
import { AnchorPositioningAdapter, DEFAULT_POPOVER, } from 'lightning/overlayPositionUtils';
import { normalizeBoolean } from 'lightning/utilsPrivate';
export default class LightningOverlayBase extends LightningShadowBaseClass {
    static [instanceName] = 'lightning-overlay-base';
    _focusTrapEnabled = false;
    _popover = DEFAULT_POPOVER;
    _initialized = false;
    _anchorAdapter = null;
    _focusTrapChanged = false;
    isPanelOpen = false;
    _closeResult;
    disconnectedCallback() {
        if (this._anchorAdapter) {
            this._anchorAdapter.destroy();
            this._anchorAdapter = null;
        }
    }
    @api
    get focusTrapEnabled() {
        return this._focusTrapEnabled;
    }
    set focusTrapEnabled(value) {
        this._focusTrapEnabled = normalizeBoolean(value);
    }
    get popover() {
        return this._popover;
    }
    renderedCallback() {
        this._focusTrapChanged = false;
        if (!this.isPanelOpen) {
            this.openPanel();
        }
        else {
            this.applyDynamicUpdates();
        }
    }
    openPanel() {
        if (!this.isPanelOpen) {
            this.isPanelOpen = true;
        }
    }
    applyDynamicUpdates() {
        if (!this._initialized) {
            this.initialize();
            this._initialized = true;
        }
        if (!this._focusTrapChanged) {
            const overlay = this.overlay;
            const surface = this.refs && this.refs.surface;
            if (!overlay || !surface) {
                return;
            }
            if (surface.popover === this._popover) {
                const source = overlay.source;
                if (source) {
                    surface.showPopover({ source });
                }
                else {
                    surface.showPopover();
                }
            }
        }
    }
    initialize() {
        if (this._popover !== this.overlay.popover) {
            this._popover = this.overlay.popover;
        }
        if (this._focusTrapEnabled !== this.overlay.enableFocusTrap) {
            this._focusTrapEnabled = this.overlay.enableFocusTrap;
            this._focusTrapChanged = true;
        }
        this.setupAnchorPositioning();
    }
    setupAnchorPositioning() {
        if (!this.overlay ||
            !this.overlay.source ||
            !this.refs.surface ||
            this._anchorAdapter) {
            return;
        }
        const { source, position, minWidth, maxWidth, minHeight, maxHeight, offsetInline, offsetBlock, } = this.overlay;
        this._anchorAdapter = new AnchorPositioningAdapter(this.refs?.surface, {
            source,
            position,
            minWidth,
            maxWidth,
            minHeight,
            maxHeight,
            offsetInline,
            offsetBlock,
        });
        this._anchorAdapter.apply();
    }
    handleToggle(event) {
        if (event.newState === 'closed') {
            this.overlay.close(this._closeResult, secure);
        }
    }
    get overlay() {
        return ((!import.meta.env.SSR &&
            this.defaultSlot &&
            this.defaultSlot.childNodes[0]) ||
            null);
    }
    @api
    get defaultSlot() {
        return this.template.querySelector('[data-slot]');
    }
    @api
    hideSurfacePopover(result) {
        this._closeResult = result;
        const surface = this.refs?.surface;
        if (surface) {
            surface?.hidePopover();
        }
    }
}
