// lightning-base-components/src/lightning/routingService/routingService.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import formFactor from '@salesforce/client/formFactor';
import { shouldDispatchToApp } from './routingServiceUtils';
const GET_LINK_INFO_EVENT = 'lightningroutingservicegetlinkinfo';
const LINK_PROVIDERS = new WeakMap();
export const urlTypes = {
    standard: 'standard_webPage',
};
export class LinkInfo {
    url;
    dispatcher;
    constructor(url, dispatcher) {
        this.url = url;
        this.dispatcher = dispatcher;
        Object.freeze(this);
    }
}
export function hasLinkProvider(element) {
    let { parentNode } = element;
    while (parentNode) {
        if (LINK_PROVIDERS.has(parentNode)) {
            return true;
        }
        parentNode = parentNode.parentNode;
    }
    return false;
}
export function isLinkProvider(element) {
    return LINK_PROVIDERS.has(element);
}
export function registerLinkProvider(element, providerFn) {
    let listeners = LINK_PROVIDERS.get(element);
    if (listeners === undefined) {
        listeners = new Set();
        LINK_PROVIDERS.set(element, listeners);
    }
    listeners.add(providerFn);
    element.addEventListener(GET_LINK_INFO_EVENT, providerFn);
}
export function unregisterLinkProvider(element, providerFn) {
    const listeners = LINK_PROVIDERS.get(element);
    if (listeners) {
        listeners.delete(providerFn);
        if (!listeners.size) {
            LINK_PROVIDERS.delete(element);
        }
    }
    element.removeEventListener(GET_LINK_INFO_EVENT, providerFn);
}
export function getLinkInfo(element, stateRef) {
    return new Promise((resolve, reject) => {
        const getLinkInfoEvent = new CustomEvent(GET_LINK_INFO_EVENT, {
            detail: {
                stateRef,
                callback: (err, linkInfo) => {
                    if (err) {
                        reject(err);
                    }
                    else {
                        resolve(linkInfo);
                    }
                },
            },
            bubbles: true,
            composed: true,
            cancelable: true,
        });
        element.dispatchEvent(getLinkInfoEvent);
    });
}
export function updateRawLinkInfo(element, { url, target }) {
    if (url === null || url === undefined) {
        console.error('url must be specified');
    }
    if (!import.meta.env.SSR && shouldDispatchToApp(target, formFactor)) {
        return getLinkInfo(element, {
            stateType: urlTypes.standard,
            attributes: {
                url,
                target,
            },
        });
    }
    return new Promise((resolve) => {
        resolve(new LinkInfo(url, () => { }));
    });
}
