// lightning-base-components/src/lightning/progressRing/progressRing.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { api } from 'lwc';
import { classSet } from 'lightning/utils';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import { isUndefinedOrNull } from 'lightning/utilsPrivate';
import labelProgressRing from '@salesforce/label/LightningProgressRing.progressRing';
import labelWarningAltText from '@salesforce/label/LightningProgressRing.warning';
import labelExpiredAltText from '@salesforce/label/LightningProgressRing.expired';
import labelCompleteAltText from '@salesforce/label/LightningProgressRing.complete';
const i18n = {
    progress: labelProgressRing,
    warning: labelWarningAltText,
    expired: labelExpiredAltText,
    complete: labelCompleteAltText,
};
export const PROGRESS_HEAD_VIEWBOX_HEIGHT = 2;
export default class progressRing extends LightningShadowBaseClass {
    @api
    value = 0;
    @api
    variant = 'base';
    @api
    direction = 'fill';
    @api
    ariaLabel;
    @api
    size = 'medium';
    arcY = 0;
    arcX = 0;
    radiusOfCircularMotionForProgressHead = 0;
    isRendered = false;
    calculateRadiusOfCircularMotionForProgressHead() {
        const progressHeadContainerHeight = parseFloat(getComputedStyle(this.template.querySelector('.slds-progress-ring__progress-head')).height);
        const progressArcContainerHeight = parseFloat(getComputedStyle(this.template.querySelector('.slds-progress-ring__progress')).height);
        const progressContentContainerHeight = parseFloat(getComputedStyle(this.template.querySelector('.slds-progress-ring__content')).height);
        const lengthScalingFactorInProgressHeadViewBox = PROGRESS_HEAD_VIEWBOX_HEIGHT / progressHeadContainerHeight;
        const radiusOfProgressArcInPixels = progressArcContainerHeight / 2;
        const radiusOfProgressArcInProgressHeadViewBox = radiusOfProgressArcInPixels *
            lengthScalingFactorInProgressHeadViewBox;
        const radiusOfProgressContentInPixels = progressContentContainerHeight / 2;
        const widthOfArcInPixels = radiusOfProgressArcInPixels - radiusOfProgressContentInPixels;
        const widthOfArcInProgressHeadViewBox = widthOfArcInPixels * lengthScalingFactorInProgressHeadViewBox;
        this.radiusOfCircularMotionForProgressHead =
            radiusOfProgressArcInProgressHeadViewBox -
                (widthOfArcInProgressHeadViewBox / 2);
    }
    renderedCallback() {
        if (this.isRendered) {
            return;
        }
        this.calculateRadiusOfCircularMotionForProgressHead();
        this.isRendered = true;
    }
    get progressHeadCenter() {
        const x = this.arcX * this.radiusOfCircularMotionForProgressHead;
        const y = this.arcY * this.radiusOfCircularMotionForProgressHead;
        return {
            x,
            y,
        };
    }
    get isProgressHeadVisible() {
        if (this.value > 0 && this.value < 100) {
            return true;
        }
        return false;
    }
    get d() {
        const fillPercent = this.value / 100;
        const filldrain = this.direction === 'drain' ? 1 : 0;
        const inverter = this.direction === 'drain' ? 1 : -1;
        const islong = fillPercent > 0.5 ? 1 : 0;
        const subCalc = 2 * Math.PI * fillPercent;
        this.arcX = Math.cos(subCalc);
        this.arcY = Math.sin(subCalc) * inverter;
        return `M 1 0 A 1 1 0 ${islong} ${filldrain} ${this.arcX} ${this.arcY} L 0 0`;
    }
    get iconSvg() {
        if (this.variant === 'warning') {
            return 'M23.7 19.6L13.2 2.5c-.6-.9-1.8-.9-2.4 0L.3 19.6c-.7 1.1 0 2.6 1.1 2.6h21.2c1.1 0 1.8-1.5 1.1-2.6zM12 18.5c-.8 0-1.4-.6-1.4-1.4s.6-1.4 1.4-1.4 1.4.6 1.4 1.4-.6 1.4-1.4 1.4zm1.4-4.2c0 .3-.2.5-.5.5h-1.8c-.3 0-.5-.2-.5-.5v-6c0-.3.2-.5.5-.5h1.8c.3 0 .5.2.5.5v6z';
        }
        if (this.variant === 'expired') {
            return 'M12 .9C5.9.9.9 5.9.9 12s5 11.1 11.1 11.1 11.1-5 11.1-11.1S18.1.9 12 .9zM3.7 12c0-4.6 3.7-8.3 8.3-8.3 1.8 0 3.5.5 4.8 1.5L5.2 16.8c-1-1.3-1.5-3-1.5-4.8zm8.3 8.3c-1.8 0-3.5-.5-4.8-1.5L18.8 7.2c1 1.3 1.5 3 1.5 4.8 0 4.6-3.7 8.3-8.3 8.3z';
        }
        if (this.isComplete) {
            return 'M8.8 19.6L1.2 12c-.3-.3-.3-.8 0-1.1l1-1c.3-.3.8-.3 1 0L9 15.7c.1.2.5.2.6 0L20.9 4.4c.2-.3.7-.3 1 0l1 1c.3.3.3.7 0 1L9.8 19.6c-.2.3-.7.3-1 0z';
        }
        return undefined;
    }
    get i18n() {
        return i18n;
    }
    get computedAltText() {
        if (this.variant === 'warning') {
            return this.i18n.warning;
        }
        if (this.variant === 'expired') {
            return this.i18n.expired;
        }
        if (this.isComplete) {
            return this.i18n.complete;
        }
        return undefined;
    }
    get computedOuterClass() {
        return classSet('slds-progress-ring')
            .add({
            'slds-progress-ring_large': this.size === 'large',
            'slds-progress-ring_warning': this.variant === 'warning',
            'slds-progress-ring_expired': this.variant === 'expired',
            'slds-progress-ring_active-step': this.variant === 'active-step',
            'slds-progress-ring_complete': this.isComplete,
        });
    }
    get computedIconTheme() {
        return classSet('slds-icon_container')
            .add({
            'slds-icon-utility-warning': this.variant === 'warning',
            'slds-icon-utility-error': this.variant === 'expired',
            'slds-icon-utility-success': this.isComplete,
        });
    }
    get isComplete() {
        return (this.variant === 'base-autocomplete' && Number(this.value) === 100);
    }
    get computeAriaLabel() {
        return isUndefinedOrNull(this.ariaLabel)
            ? i18n.progress
            : this.ariaLabel;
    }
}
