// lightning-base-components/src/lightning/f6Controller/f6Controller.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { debounce } from 'lightning/inputUtils';
import { findAllTabbableElements } from 'lightning/focusUtils';
const DEBOUNCE_KEY_DOWN = 300;
export const DEFAULT_CONFIG = {
    navKey: 'F6',
    f6RegionAttribute: 'data-f6-region',
    f6RegionHighlightClass: 'f6-highlight',
};
export const getActiveElement = (element) => {
    if (!element) {
        return element;
    }
    if (!element.shadowRoot) {
        if (element.activeElement) {
            return getActiveElement(element.activeElement);
        }
        return element;
    }
    if (!element.shadowRoot.activeElement) {
        return element;
    }
    return getActiveElement(element.shadowRoot.activeElement);
};
export class F6Controller {
    regions = [];
    config = DEFAULT_CONFIG;
    _debounceKeyDownHandler;
    _styleElement;
    constructor() {
        this._debounceKeyDownHandler = debounce(this.handleKeyDown.bind(this), DEBOUNCE_KEY_DOWN);
        this.initialize();
    }
    appendStyleElement() {
        this._styleElement = document.createElement('style');
        const selector = `[${this.config.f6RegionAttribute}].${this.config.f6RegionHighlightClass}`;
        this._styleElement.innerText = `${selector} { position: relative; } ${selector}::after { width: 100%; height: 100%; content: ''; outline: rgb(94, 158, 214) 3px solid; outline-offset: -3px; position: absolute; top: 0; left: 0; z-index: 9999; }`;
        this._styleElement.setAttribute('type', 'text/css');
        document.head.appendChild(this._styleElement);
    }
    clearRegionHighlights() {
        if (!this.regions || !this.regions.length) {
            return;
        }
        this.regions.forEach((region) => {
            region.classList.remove(this.config.f6RegionHighlightClass);
        });
    }
    addRegionHighlight(element) {
        element.classList.add(this.config.f6RegionHighlightClass);
    }
    findFocusableElement(element) {
        if (!element || !element.isConnected) {
            return undefined;
        }
        const tabbableElements = findAllTabbableElements(element);
        if (tabbableElements && tabbableElements.length) {
            return tabbableElements[0];
        }
        element.tabIndex = '-1';
        return element;
    }
    focusIn(region) {
        const element = this.findFocusableElement(region);
        if (element) {
            element.focus();
        }
    }
    handleNavigation(event) {
        event.preventDefault();
        if (!this.regions.length) {
            return;
        }
        const isPrevious = event.shiftKey;
        const currentRegionIndex = this.getElementRegionIndex(getActiveElement(event.target));
        const adjacentRegionIndex = this.getAdjacentRegionIndex(currentRegionIndex, isPrevious);
        if (adjacentRegionIndex === -1) {
            return;
        }
        const regionToFocus = this.regions[adjacentRegionIndex];
        this.focusIn(regionToFocus);
        this.addRegionHighlight(regionToFocus);
    }
    shadowContains(container, element) {
        if (container === element || container.contains(element)) {
            return true;
        }
        if (container.shadowRoot) {
            if (this.isElementInContainerElements(container.shadowRoot.children, element)) {
                return true;
            }
        }
        if (container.tagName === 'SLOT') {
            if (this.isElementInContainerElements(container.assignedElements(), element)) {
                return true;
            }
        }
        return this.isElementInContainerElements(container.children, element);
    }
    isElementInContainerElements(containerElements, searchElement) {
        if (!containerElements || !containerElements.length) {
            return false;
        }
        const numElements = containerElements.length;
        for (let index = 0; index < numElements; index++) {
            if (this.shadowContains(containerElements[index], searchElement)) {
                return true;
            }
        }
        return false;
    }
    getElementRegionIndex(element) {
        if (!this.regions || !this.regions.length) {
            return -1;
        }
        return this.regions.findIndex((container) => {
            return this.shadowContains(container, element);
        });
    }
    getAdjacentRegionIndex(currentRegionIndex, isPrevious) {
        const lastRegionIndex = this.regions.length - 1;
        if (lastRegionIndex < 0) {
            return -1;
        }
        if (isPrevious) {
            if (currentRegionIndex <= 0) {
                return lastRegionIndex;
            }
            return currentRegionIndex - 1;
        }
        if (currentRegionIndex === lastRegionIndex) {
            return 0;
        }
        return currentRegionIndex + 1;
    }
    isVisible(element) {
        if (element === document.body || !(element instanceof Element)) {
            return true;
        }
        try {
            const computedStyle = window.getComputedStyle(element) || element.style;
            if (!computedStyle) {
                return false;
            }
            const { display, visibility } = computedStyle;
            if ((display && display.toLowerCase() === 'none') ||
                (visibility && visibility.toLowerCase() === 'hidden')) {
                return false;
            }
            return this.isVisible(element.parentNode);
        }
        catch (error) {
            return false;
        }
    }
    isEmpty(element) {
        if (element.tagName === 'SLOT') {
            if (!this.isEmptyChildren(element.assignedElements())) {
                return false;
            }
        }
        let children = element.children;
        if ((!children || !children.length) && element.shadowRoot) {
            children = element.shadowRoot.children;
        }
        return this.isEmptyChildren(children);
    }
    isEmptyChildren(children) {
        if (!children || !children.length) {
            return true;
        }
        const numChildren = children.length;
        for (let index = 0; index < numChildren; index++) {
            const child = children[index];
            if (child.tagName !== 'SLOT' || !this.isEmpty(child)) {
                return false;
            }
        }
        return true;
    }
    populateRegions() {
        this.regions = [];
        this.shadowTreeWalker(document.body);
    }
    shadowTreeWalker(root) {
        const treeWalker = document.createTreeWalker(root, NodeFilter.SHOW_ELEMENT, this.elementFilter);
        while (treeWalker.nextNode()) {
            if (this.isF6Region(treeWalker.currentNode)) {
                this.regions.push(treeWalker.currentNode);
            }
            else if (treeWalker.currentNode.shadowRoot) {
                this.shadowTreeWalker(treeWalker.currentNode.shadowRoot);
            }
        }
    }
    initialize() {
        document.addEventListener('keydown', this._debounceKeyDownHandler);
        document.addEventListener('click', this.handleClick);
        this.regions = [];
        this.appendStyleElement();
    }
    cleanUp() {
        document.removeEventListener('keydown', this._debounceKeyDownHandler);
        document.removeEventListener('click', this.handleClick);
        this._regions = [];
        this._debounceKeyDownHandler = null;
        if (this._styleElement) {
            document.head.removeChild(this._styleElement);
            this._styleElement = null;
        }
    }
    handleClick = () => {
        this.clearRegionHighlights();
    };
    handleKeyDown = (event) => {
        this.clearRegionHighlights();
        const { key, ctrlKey, metaKey } = event;
        if (key === this.config.navKey && (ctrlKey || metaKey)) {
            this.populateRegions();
            this.handleNavigation(event);
        }
    };
    isF6Region = (element) => element.matches(`*[${this.config.f6RegionAttribute}]`) &&
        this.isVisible(element) &&
        !this.isEmpty(element);
    elementFilter = (element) => {
        if (element.parentElement &&
            element.parentElement.matches(`*[${this.config.f6RegionAttribute}]`)) {
            return NodeFilter.FILTER_REJECT;
        }
        if (this.isF6Region(element) || element.shadowRoot) {
            return NodeFilter.FILTER_ACCEPT;
        }
        return NodeFilter.FILTER_SKIP;
    };
    getConfig() {
        return Object.freeze(this.config);
    }
}
let f6Controller;
export const createF6Controller = () => {
    if (!f6Controller) {
        f6Controller = new F6Controller();
    }
    return f6Controller;
};
export const getCurrentRegionAttributeName = () => {
    if (f6Controller) {
        const config = f6Controller.getConfig();
        if (config) {
            return config.f6RegionAttribute;
        }
    }
    return undefined;
};
