// lightning-base-components/src/lightning/verticalNavigationItem/verticalNavigationItem.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { api, track } from 'lwc';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import { sanitizeURL } from 'lightning/utilsPrivate';
import { privateContext } from 'lightning/utilsInternal';
const DEFAULT_HREF = 'javascript:void(0)';
const { setContext, assertContext } = privateContext;
export default class LightningVerticalNavigationItem extends LightningShadowBaseClass {
    static validationOptOut = ['class'];
    _rendered = false;
    @api
    label;
    @api
    name;
    @api
    get href() {
        return this._href || DEFAULT_HREF;
    }
    set href(value) {
        this._href = sanitizeURL(value);
    }
    @track
    _href;
    @track
    _selected = false;
    constructor() {
        super();
        setContext(this);
    }
    connectedCallback() {
        super.connectedCallback();
        assertContext(this);
        this.setAttribute('role', 'listitem');
        this.classList.add('slds-nav-vertical__item');
        if (!import.meta.env.SSR) {
            this.dispatchEvent(new CustomEvent('privateitemregister', {
                bubbles: true,
                cancelable: true,
                detail: {
                    callbacks: {
                        select: this.select.bind(this),
                        deselect: this.deselect.bind(this),
                    },
                    name: this.name,
                },
            }));
        }
    }
    renderedCallback() {
        if (!this._rendered) {
            this._rendered = true;
        }
    }
    select() {
        this._selected = true;
        this.classList.add('slds-is-active');
    }
    deselect() {
        this._selected = false;
        this.classList.remove('slds-is-active');
    }
    get ariaCurrent() {
        return this._selected && this._rendered ? 'page' : null;
    }
    handleClick(event) {
        this.dispatchEvent(new CustomEvent('privateitemselect', {
            bubbles: true,
            cancelable: true,
            composed: true,
            detail: {
                name: this.name,
            },
        }));
        if (this.href === DEFAULT_HREF) {
            event.preventDefault();
        }
    }
}
