// lightning-base-components/src/lightning/internationalizationLibrary/address/AddressFormat.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
function _toPrimitive(t, r) {
    if ('object' != typeof t || !t)
        return t;
    let e = t[Symbol.toPrimitive];
    if (void 0 !== e) {
        let i = e.call(t, r || 'default');
        if ('object' != typeof i)
            return i;
        throw new TypeError('@@toPrimitive must return a primitive value.');
    }
    return ('string' === r ? String : Number)(t);
}
function _toPropertyKey(t) {
    let i = _toPrimitive(t, 'string');
    return 'symbol' == typeof i ? i : String(i);
}
function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError('Cannot call a class as a function');
    }
}
function _defineProperties(target, props) {
    for (let i = 0; i < props.length; i++) {
        let descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ('value' in descriptor)
            descriptor.writable = true;
        Object.defineProperty(target, _toPropertyKey(descriptor.key), descriptor);
    }
}
function _createClass(Constructor, protoProps, staticProps) {
    if (protoProps)
        _defineProperties(Constructor.prototype, protoProps);
    if (staticProps)
        _defineProperties(Constructor, staticProps);
    Object.defineProperty(Constructor, 'prototype', {
        writable: false,
    });
    return Constructor;
}
function _unsupportedIterableToArray(o, minLen) {
    if (!o)
        return;
    if (typeof o === 'string')
        return _arrayLikeToArray(o, minLen);
    let n = Object.prototype.toString.call(o).slice(8, -1);
    if (n === 'Object' && o.constructor)
        n = o.constructor.name;
    if (n === 'Map' || n === 'Set')
        return Array.from(o);
    if (n === 'Arguments' || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n))
        return _arrayLikeToArray(o, minLen);
}
function _arrayLikeToArray(arr, len) {
    if (len == null || len > arr.length)
        len = arr.length;
    for (var i = 0, arr2 = new Array(len); i < len; i++)
        arr2[i] = arr[i];
    return arr2;
}
function _createForOfIteratorHelper(o, allowArrayLike) {
    let it = typeof Symbol !== 'undefined' && o[Symbol.iterator] || o['@@iterator'];
    if (!it) {
        if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === 'number') {
            if (it)
                o = it;
            let i = 0;
            let F = function () { };
            return {
                s: F,
                n: function () {
                    if (i >= o.length) {
                        return {
                            done: true,
                        };
                    }
                    return {
                        done: false,
                        value: o[i++],
                    };
                },
                e: function (e) {
                    throw e;
                },
                f: F,
            };
        }
        throw new TypeError('Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.');
    }
    let normalCompletion = true;
    let didErr = false;
    let err;
    return {
        s: function () {
            it = it.call(o);
        },
        n: function () {
            let step = it.next();
            normalCompletion = step.done;
            return step;
        },
        e: function (e) {
            didErr = true;
            err = e;
        },
        f: function () {
            try {
                if (!normalCompletion && it.return != null)
                    it.return();
            }
            finally {
                if (didErr)
                    throw err;
            }
        },
    };
}
let data = {
    AE: {
        fmt: '%A%n%S%n%K',
        require: 'ASK',
        input: 'ASK',
    },
    AL: {
        fmt: '%A%n%Z%n%C%n%K',
        require: 'AZCK',
        input: 'AZCK',
    },
    EC: {
        _ref: 'AL',
    },
    MU: {
        _ref: 'AL',
    },
    OM: {
        _ref: 'AL',
    },
    AM: {
        fmt: '%A%n%Z%n%C%n%S%n%K',
        require: 'AZK',
        input: 'AZK',
    },
    AR: {
        fmt: '%A%n%Z %C%n%S%n%K',
        require: 'AZCSK',
        input: 'AZCSK',
    },
    CL: {
        _ref: 'AR',
    },
    CV: {
        _ref: 'AR',
    },
    MY: {
        _ref: 'AR',
    },
    UZ: {
        _ref: 'AR',
    },
    AT: {
        fmt: '%A%n%Z %C%n%K',
        require: 'AZCK',
        input: 'AZCK',
    },
    BA: {
        _ref: 'AT',
    },
    BG: {
        _ref: 'AT',
    },
    CH: {
        _ref: 'AT',
    },
    CY: {
        _ref: 'AT',
    },
    DE: {
        _ref: 'AT',
    },
    DK: {
        _ref: 'AT',
    },
    DO: {
        _ref: 'AT',
    },
    DZ: {
        _ref: 'AT',
    },
    EE: {
        _ref: 'AT',
    },
    ET: {
        _ref: 'AT',
    },
    FR: {
        _ref: 'AT',
    },
    GE: {
        _ref: 'AT',
    },
    GL: {
        _ref: 'AT',
    },
    GR: {
        _ref: 'AT',
    },
    IL: {
        _ref: 'AT',
    },
    KW: {
        _ref: 'AT',
    },
    LA: {
        _ref: 'AT',
    },
    LR: {
        _ref: 'AT',
    },
    IS: {
        _ref: 'AT',
    },
    MA: {
        _ref: 'AT',
    },
    MG: {
        _ref: 'AT',
    },
    MK: {
        _ref: 'AT',
    },
    MZ: {
        _ref: 'AT',
    },
    NL: {
        _ref: 'AT',
    },
    NO: {
        _ref: 'AT',
    },
    PL: {
        _ref: 'AT',
    },
    PT: {
        _ref: 'AT',
    },
    PY: {
        _ref: 'AT',
    },
    RO: {
        _ref: 'AT',
    },
    RS: {
        _ref: 'AT',
    },
    SK: {
        _ref: 'AT',
    },
    TJ: {
        _ref: 'AT',
    },
    TN: {
        _ref: 'AT',
    },
    TZ: {
        _ref: 'AT',
    },
    WF: {
        _ref: 'AT',
    },
    AU: {
        fmt: '%A%n%C %S %Z%n%K',
        require: 'ACSZK',
        input: 'ACSZK',
    },
    CA: {
        _ref: 'AU',
    },
    AZ: {
        fmt: '%A%n%Z %C%n%K',
        require: 'AZCK',
        input: 'AZCK',
    },
    AF: {
        fmt: '%A%n%C%n%Z%n%K',
        require: 'ACZK',
        input: 'ACZK',
    },
    FK: {
        _ref: 'AF',
    },
    GB: {
        fmt: '%A%n%C%n%S%n%Z%n%K',
        require: 'ACZK',
        input: 'ACSZK',
    },
    KE: {
        _ref: 'AF',
    },
    LK: {
        _ref: 'AF',
    },
    ZA: {
        _ref: 'AF',
    },
    SH: {
        _ref: 'AF',
    },
    SZ: {
        _ref: 'AF',
    },
    US: {
        fmt: '%A%n%C, %S %Z%n%K',
        require: 'ACSZK',
        input: 'ACSZK',
    },
    BB: {
        _ref: 'US',
    },
    BS: {
        _ref: 'US',
    },
    SO: {
        _ref: 'US',
    },
    ES: {
        fmt: '%A%n%Z %C %S%n%K',
        require: 'AZCSK',
        input: 'AZCSK',
    },
    IT: {
        _ref: 'ES',
    },
    UY: {
        _ref: 'ES',
    },
    ID: {
        fmt: '%A%n%C%n%S %Z%n%K',
        require: 'ACSZK',
        input: 'ACSZK',
    },
    IE: {
        _ref: 'ID',
    },
    TH: {
        _ref: 'ID',
    },
    VN: {
        _ref: 'ID',
    },
    HU: {
        fmt: '%C%n%A%n%Z%n%K',
        require: 'CAZK',
        input: 'CAZK',
    },
    BH: {
        fmt: '%A%n%C %Z%n%K',
        require: 'ACZK',
        input: 'ACZK',
    },
    BM: {
        _ref: 'BH',
    },
    BN: {
        _ref: 'BH',
    },
    BT: {
        _ref: 'BH',
    },
    KH: {
        _ref: 'BH',
    },
    LB: {
        _ref: 'BH',
    },
    JO: {
        _ref: 'BH',
    },
    MT: {
        _ref: 'BH',
    },
    NP: {
        _ref: 'BH',
    },
    NZ: {
        _ref: 'BH',
    },
    SA: {
        _ref: 'BH',
    },
    BD: {
        fmt: '%A%n%C - %Z%n%K',
        require: 'ACZK',
        input: 'ACZK',
    },
    BR: {
        fmt: '%A%n%C-%S%n%Z%n%K',
        require: 'ACSZK',
        input: 'ACSZK',
    },
    CN: {
        fmt: '%K%n%S %C%n%A%n%Z',
        require: 'KCAZ',
        input: 'KSCAZ',
    },
    HK: {
        fmt: '%K%S%C%n%A%n%Z',
        require: 'KCA',
        input: 'KSCAZ',
    },
    CO: {
        fmt: '%A%n%C, %S, %Z%n%K',
        require: 'ACSZK',
        input: 'ACSZK',
    },
    CR: {
        fmt: '%A%n%S, %C%n%Z%n%K',
        require: 'ACSZK',
        input: 'ASCZK',
    },
    EG: {
        fmt: '%A%n%C%n%S%n%Z%n%K',
        require: 'ACSZK',
        input: 'ACSZK',
    },
    RU: {
        _ref: 'EG',
    },
    UA: {
        _ref: 'EG',
    },
    FI: {
        fmt: '%A%n%Z %C%n%K',
        require: 'AZCK',
        input: 'AZCK',
    },
    GT: {
        fmt: '%A%n%Z-%C%n%K',
        require: 'AZCK',
        input: 'AZCK',
    },
    HN: {
        fmt: '%A%n%C, %S%n%Z%n%K',
        require: 'ACSZK',
        input: 'ACSZK',
    },
    IQ: {
        _ref: 'HN',
    },
    HR: {
        fmt: '%A%n%Z %C%n%K',
        require: 'AZCK',
        input: 'AZCK',
    },
    HT: {
        fmt: '%A%n%Z %C%n%K',
        require: 'AZCK',
        input: 'AZCK',
    },
    IN: {
        fmt: '%A%n%C %Z%n%S%n%K',
        require: 'ACZSK',
        input: 'ACZSK',
    },
    NG: {
        _ref: 'IN',
    },
    PE: {
        _ref: 'IN',
    },
    IR: {
        fmt: '%S%n%C%n%A%n%Z%n%K',
        require: 'SCAZK',
        input: 'SCAZK',
    },
    JM: {
        fmt: '%A%n%C%n%S%n%K',
        require: 'ACSK',
        input: 'ACSK',
    },
    PA: {
        _ref: 'JM',
    },
    SC: {
        _ref: 'JM',
    },
    SR: {
        _ref: 'JM',
    },
    JP: {
        fmt: '%K%n' + String.fromCharCode(12306) + '%Z%n%S %C%n%A',
        require: 'KZCA',
        input: 'KZSCA',
    },
    EN_JP: {
        fmt: '%A%n%C %S%n%Z %K',
        require: 'ACSZK',
        input: 'ACSZK',
    },
    KG: {
        fmt: '%Z %C%n%A%n%S%n%K',
        require: 'ZCAK',
        input: 'ZCAK',
    },
    KR: {
        fmt: '%S %C%n%A%n%Z%n%K',
        require: 'SCAZK',
        input: 'SCAZK',
    },
    KY: {
        fmt: '%A%n%S %Z%n%K',
        require: 'ASZK',
        input: 'ASZK',
    },
    KZ: {
        fmt: '%Z%n%S%n%C%n%A%n%K',
        require: 'ZSCAK',
        input: 'ZSCAK',
    },
    LT: {
        fmt: '%A%n%Z %C%n%K',
        require: 'AZCK',
        input: 'AZCK',
    },
    LV: {
        fmt: '%A%n%C, %Z%n%K',
        require: 'ACZK',
        input: 'ACZK',
    },
    MM: {
        _ref: 'LV',
    },
    MC: {
        fmt: '%A%n%Z %C%n%K',
        require: 'AZCK',
        input: 'AZCK',
    },
    MD: {
        fmt: '%A%n%Z %C%n%K',
        require: 'AZCK',
        input: 'AZCK',
    },
    MW: {
        fmt: '%A%n%C%n%K',
        require: 'ACK',
        input: 'ACK',
    },
    MX: {
        fmt: '%A%n%Z %C, %S%n%K',
        require: 'AZCSK',
        input: 'AZCSK',
    },
    NI: {
        fmt: '%A%n%Z%n%C, %S%n%K',
        require: 'AZCSK',
        input: 'AZCSK',
    },
    PG: {
        fmt: '%A%n%C %Z %S%n%K',
        require: 'ACZSK',
        input: 'ACZSK',
    },
    PH: {
        fmt: '%A, %C%n%Z %S%n%K',
        require: 'ACZSK',
        input: 'ACZSK',
    },
    PK: {
        fmt: '%A%n%C-%Z%n%K',
        require: 'ACZK',
        input: 'ACZK',
    },
    PR: {
        fmt: '%A%n%C %Z%n%K',
        require: 'ACZK',
        input: 'ACZK',
    },
    SE: {
        fmt: '%A%n%Z %C%n%K',
        require: 'AZCK',
        input: 'AZCK',
    },
    SG: {
        fmt: '%A%n%C %Z%n%S%n%K',
        require: 'AZK',
        input: 'AZK',
    },
    SI: {
        fmt: '%A%n%Z %C%n%K',
        require: 'AZCK',
        input: 'AZCK',
    },
    SV: {
        fmt: '%A%n%Z-%C%n%S%n%K',
        require: 'AZCSK',
        input: 'AZCSK',
    },
    TR: {
        fmt: '%A%n%Z %C/%S%n%K',
        require: 'AZCK',
        input: 'AZCK',
    },
    TW: {
        fmt: '%K%n%Z%n%S %C%n%A',
        require: 'KZSCA',
        input: 'KZSCA',
    },
    VE: {
        fmt: '%A%n%C %Z, %S%n%K',
        require: 'ACZSK',
        input: 'ACZSK',
    },
};
let languageCodeToCountry = {
    languageCode: {
        'af': 'ZA',
        'am': 'ET',
        'ar': 'AE',
        'bg': 'BG',
        'bn': 'BD',
        'bs': 'BA',
        'ca': 'ES',
        'cs': 'CZ',
        'cy': 'GB',
        'da': 'DK',
        'de': 'DE',
        'el': 'GR',
        'es': 'ES',
        'et': 'EE',
        'eu': 'ES',
        'fa': 'IR',
        'fi': 'FI',
        'fr': 'FR',
        'ga': 'IE',
        'gu': 'IN',
        'haw': 'US',
        'hi': 'IN',
        'hmn': 'US',
        'hr': 'HR',
        'ht': 'HT',
        'hu': 'HU',
        'hy': 'AM',
        'in': 'ID',
        'is': 'IS',
        'it': 'IT',
        'iw': 'IL',
        'ja': 'JP',
        'ji': 'IL',
        'ka': 'GE',
        'kk': 'KZ',
        'kl': 'GL',
        'km': 'KH',
        'kn': 'IN',
        'ko': 'KR',
        'lb': 'LU',
        'lt': 'LT',
        'lv': 'LV',
        'mi': 'NZ',
        'mk': 'MK',
        'ml': 'IN',
        'mr': 'IN',
        'ms': 'MY',
        'mt': 'MT',
        'my': 'MM',
        'nl': 'NL',
        'no': 'NO',
        'pa': 'IN',
        'pl': 'PL',
        'pt': 'PT',
        'rm': 'CH',
        'ro': 'RO',
        'ru': 'RU',
        'sh': 'RS',
        'sk': 'SK',
        'sl': 'SL',
        'sm': 'WS',
        'sq': 'AL',
        'sr': 'RS',
        'sv': 'SE',
        'sw': 'ZA',
        'ta': 'IN',
        'te': 'IN',
        'th': 'TH',
        'tl': 'PH',
        'tr': 'TR',
        'uk': 'UA',
        'ur': 'IN',
        'vi': 'VN',
        'xh': 'ZA',
        'zh': 'CN',
        'zu': 'ZA',
    },
};
var AddressFormatPattern = Object.freeze({
    A: Symbol('Address Lines'),
    C: Symbol('City'),
    S: Symbol('State'),
    K: Symbol('Country'),
    Z: Symbol('Zip Code'),
    n: Symbol('New Line'),
    fromPlaceHolder: function fromPlaceHolder(placeHolder) {
        switch (placeHolder) {
            case 'A':
                return AddressFormatPattern.A;
            case 'C':
                return AddressFormatPattern.C;
            case 'S':
                return AddressFormatPattern.S;
            case 'K':
                return AddressFormatPattern.K;
            case 'Z':
                return AddressFormatPattern.Z;
            case 'n':
                return AddressFormatPattern.n;
        }
        return null;
    },
    getPlaceHolder: function getPlaceHolder(pattern) {
        switch (pattern) {
            case AddressFormatPattern.A:
                return 'A';
            case AddressFormatPattern.C:
                return 'C';
            case AddressFormatPattern.S:
                return 'S';
            case AddressFormatPattern.K:
                return 'K';
            case AddressFormatPattern.Z:
                return 'Z';
            case AddressFormatPattern.n:
                return 'n';
        }
        return null;
    },
    getData: function getData(pattern, data) {
        if (data) {
            switch (pattern) {
                case AddressFormatPattern.A:
                    return data.address;
                case AddressFormatPattern.C:
                    return data.city;
                case AddressFormatPattern.S:
                    return data.state;
                case AddressFormatPattern.K:
                    return data.country;
                case AddressFormatPattern.Z:
                    return data.zipCode;
                case AddressFormatPattern.n:
                    return data.newLine;
            }
        }
        return null;
    },
});
let AddressTokenTypes = Object.freeze({
    DATA: Symbol('data'),
    STRING: Symbol('string'),
    NEWLINE: Symbol('newline'),
    GROUP: Symbol('group'),
});
let AddressToken = function () {
    function AddressToken(type, string, pattern) {
        _classCallCheck(this, AddressToken);
        this.type = type;
        this.string = string;
        this.pattern = pattern;
    }
    _createClass(AddressToken, null, [{
            key: 'string',
            value: function string(_string) {
                return new AddressToken(AddressTokenTypes.STRING, _string);
            },
        }, {
            key: 'data',
            value: function data(pattern) {
                return new AddressToken(AddressTokenTypes.DATA, undefined, pattern);
            },
        }, {
            key: 'newLine',
            value: function newLine() {
                return new AddressToken(AddressTokenTypes.NEWLINE);
            },
        }]);
    return AddressToken;
}();
let TokenizerState = _createClass(function TokenizerState(pattern, start) {
    _classCallCheck(this, TokenizerState);
    this.pattern = pattern;
    this.start = start;
});
function tokenize(state, tokens) {
    let nextIndex = state.start;
    if (state.pattern) {
        let len = state.pattern.length;
        while (state.start < len) {
            nextIndex = state.pattern.indexOf('%', nextIndex);
            if (nextIndex >= 0 && nextIndex + 1 < len) {
                let placeHolder = state.pattern.substring(nextIndex + 1, nextIndex + 2);
                switch (placeHolder) {
                    case 'n':
                        {
                            if (nextIndex - state.start > 0) {
                                tokens.push(AddressToken.string(state.pattern.substring(state.start, nextIndex)));
                            }
                            tokens.push(AddressToken.newLine());
                            state.start = nextIndex + 2;
                            nextIndex = state.start;
                            break;
                        }
                    default:
                        {
                            let p = AddressFormatPattern.fromPlaceHolder(placeHolder);
                            if (p) {
                                if (nextIndex - state.start > 0) {
                                    tokens.push(AddressToken.string(state.pattern.substring(state.start, nextIndex)));
                                }
                                tokens.push(AddressToken.data(p));
                                state.start = nextIndex + 2;
                                nextIndex = state.start;
                            }
                            else {
                                state.start = nextIndex + 2;
                                nextIndex = state.start;
                            }
                            break;
                        }
                }
            }
            else {
                if (state.start < len) {
                    tokens.push(AddressToken.string(state.pattern.substring(state.start)));
                }
                state.start = len;
            }
        }
    }
    return state;
}
function formatLineTokens(tokens, data, ignoreEmptyLines, firstIndex, lastIndex) {
    let parts = [];
    for (let index = firstIndex; index <= lastIndex; index++) {
        let token = tokens[index];
        if (!token) {
            continue;
        }
        else if (token.type == AddressTokenTypes.DATA) {
            let dataBuffer = '';
            let lastDataIndex = index;
            for (let dataIndex = index; dataIndex <= lastIndex; dataIndex++) {
                let dataToken = tokens[dataIndex];
                if (!dataToken || dataToken.type != AddressTokenTypes.DATA) {
                    break;
                }
                let fieldData = AddressFormatPattern.getData(dataToken.pattern, data);
                if (fieldData) {
                    dataBuffer += fieldData;
                    lastDataIndex = dataIndex;
                }
            }
            let hasData = dataBuffer && dataBuffer.length > 0;
            let hasPreviousData = false;
            if (index - 1 >= firstIndex) {
                let stringToken = tokens[index - 1];
                if (stringToken && stringToken.type == AddressTokenTypes.STRING && stringToken.string) {
                    for (let prevIndex = index - 2; prevIndex >= firstIndex; prevIndex--) {
                        let prevToken = tokens[prevIndex];
                        if (prevToken && prevToken.type == AddressTokenTypes.DATA) {
                            let _fieldData = AddressFormatPattern.getData(prevToken.pattern, data);
                            if (_fieldData) {
                                hasPreviousData = true;
                                break;
                            }
                        }
                        else if (prevToken && prevToken.type == AddressTokenTypes.STRING) {
                            stringToken = prevToken;
                        }
                    }
                    if (!ignoreEmptyLines || hasPreviousData && hasData || index - 1 == firstIndex && hasData) {
                        parts.push(stringToken.string);
                    }
                }
            }
            if (hasData) {
                parts.push(dataBuffer);
            }
            index = lastDataIndex;
            if (index + 1 == lastIndex) {
                let _stringToken = tokens[index + 1];
                if (_stringToken && _stringToken.type == AddressTokenTypes.STRING && _stringToken.string) {
                    if (!ignoreEmptyLines || hasData || hasPreviousData) {
                        parts.push(_stringToken.string);
                    }
                }
                index = index + 1;
            }
        }
        else
            ;
    }
    return parts.join('').trim();
}
function formatTokens(tokens, data, lineBreak, ignoreEmptyLines) {
    let lines = [];
    let lineIndex = -1;
    for (let index = 0; index < tokens.length; index++) {
        let doFormat = false;
        let endWithNewLine = false;
        let token = tokens[index];
        switch (token.type) {
            case AddressTokenTypes.NEWLINE:
                {
                    if (lineIndex >= 0) {
                        doFormat = true;
                        endWithNewLine = true;
                    }
                    else if (!ignoreEmptyLines) {
                        lines.push('');
                        if (index + 1 == tokens.length) {
                            lines.push('');
                        }
                    }
                    break;
                }
            default:
                {
                    lineIndex = lineIndex < 0 ? index : lineIndex;
                    doFormat = index + 1 == tokens.length ? true : doFormat;
                    break;
                }
        }
        if (doFormat) {
            let line = formatLineTokens(tokens, data, ignoreEmptyLines, lineIndex, endWithNewLine ? index - 1 : index);
            if (!ignoreEmptyLines || line) {
                lines.push(line);
            }
            if (!ignoreEmptyLines && endWithNewLine && index + 1 == tokens.length) {
                lines.push('');
            }
            lineIndex = -1;
        }
    }
    return lines.join(lineBreak);
}
function format(data, pattern, lineBreak, ignoreEmptyLines) {
    ignoreEmptyLines = ignoreEmptyLines === false ? false : true;
    lineBreak = lineBreak || '\n';
    let tokens = [];
    tokenize(new TokenizerState(pattern, 0), tokens);
    return formatTokens(tokens, data, lineBreak, ignoreEmptyLines);
}
let addressFormatter = {
    format: format,
};
let CJK_COUNTRIES = ['CN', 'HK', 'TW', 'JP', 'KR', 'KP'];
let CJK_LANGUAGES = ['zh', 'ja', 'ko'];
let address = {
    getAddressInfoForCountry: function getAddressInfoForCountry(langCode, countryCode) {
        let code = this.getCountryFromLocale(langCode, countryCode);
        if (data[code]) {
            let cloneAddressRep = Object.freeze(Object.assign({}, data[code]));
            return Object.freeze({
                address: cloneAddressRep,
            });
        }
        return {};
    },
    getAddressFormat: function getAddressFormat(langCode, countryCode) {
        let code = this.getCountryFromLocale(langCode, countryCode);
        if (data[code]) {
            return data[code].fmt;
        }
        return '';
    },
    getAddressInputOrder: function getAddressInputOrder(langCode, countryCode) {
        if (langCode && langCode.toLowerCase() == 'en' && countryCode && countryCode.toUpperCase() == 'HK') {
            langCode = 'en';
            countryCode = 'US';
        }
        let code = this.getCountryFromLocale(langCode, countryCode);
        if (data[code]) {
            return data[code].input;
        }
        return '';
    },
    getAddressInputOrderAllField: function getAddressInputOrderAllField(langCode, countryCode) {
        if (langCode && langCode.toLowerCase() == 'en' && countryCode && countryCode.toUpperCase() == 'HK') {
            langCode = 'en';
            countryCode = 'US';
        }
        let code = this.getCountryFromLocale(langCode, countryCode);
        if (data[code]) {
            let input = data[code].input;
            if (input.indexOf('S') === -1) {
                input = input.replace('K', 'SK');
            }
            if (input.indexOf('C') === -1) {
                input = input.replace('S', 'CS');
            }
            if (input.indexOf('Z') === -1) {
                input = input.replace('C', 'ZC');
            }
            return input;
        }
        return '';
    },
    getAddressRequireFields: function getAddressRequireFields(langCode, countryCode) {
        let code = this.getCountryFromLocale(langCode, countryCode);
        if (data[code]) {
            return data[code].require;
        }
        return '';
    },
    formatAddressAllFields: function formatAddressAllFields(langCode, countryCode, values, lineBreak) {
        let code = this.getCountryFromLocale(langCode, countryCode, values);
        if (data[code]) {
            let pattern = data[code].fmt;
            if (values.zipCode && pattern.indexOf('%Z') === -1) {
                pattern = pattern.replace('%K', '%Z %K');
            }
            if (values.city && pattern.indexOf('%C') === -1) {
                pattern = pattern.replace('%K', '%C %K');
            }
            if (values.state && pattern.indexOf('%S') === -1) {
                pattern = pattern.replace('%K', '%S %K');
            }
            return this.buildAddressLines(pattern, values, lineBreak, true);
        }
        return '';
    },
    formatAddress: function formatAddress(langCode, countryCode, values, lineBreak) {
        let code = this.getCountryFromLocale(langCode, countryCode, values);
        if (data[code]) {
            return this.buildAddressLines(data[code].fmt, values, lineBreak, true);
        }
        return '';
    },
    buildAddressLines: function buildAddressLines(pattern, values, lineBreak, ignoreEmptyLines) {
        return addressFormatter.format(values, pattern, lineBreak, ignoreEmptyLines);
    },
    followReferences: function followReferences(data, countryCode) {
        if (data[countryCode] && data[countryCode]._ref) {
            return this.followReferences(data, data[countryCode]._ref);
        }
        return countryCode;
    },
    containsHanScript: function containsHanScript() {
        for (var _len = arguments.length, values = new Array(_len), _key = 0; _key < _len; _key++) {
            values[_key] = arguments[_key];
        }
        if (!values || !Array.isArray(values))
            return false;
        return values.some(function (value) {
            if (!value)
                return false;
            let _iterator = _createForOfIteratorHelper(value);
            let _step;
            try {
                for (_iterator.s(); !(_step = _iterator.n()).done;) {
                    let singleChar = _step.value;
                    let codePoint = singleChar.codePointAt(0);
                    if (0x2e80 <= codePoint && codePoint <= 0x2eff ||
                        0x3300 <= codePoint && codePoint <= 0x33ff ||
                        0xfe30 <= codePoint && codePoint <= 0xfe4f ||
                        0xf900 <= codePoint && codePoint <= 0xfaff ||
                        0x2f800 <= codePoint && codePoint <= 0x2fa1f ||
                        0x3000 <= codePoint && codePoint <= 0x303f ||
                        0x4e00 <= codePoint && codePoint <= 0x9fff ||
                        0x3400 <= codePoint && codePoint <= 0x4dbf ||
                        0x20000 <= codePoint && codePoint <= 0x2a6df ||
                        0x2a700 <= codePoint && codePoint <= 0x2b73f ||
                        0x2b740 <= codePoint && codePoint <= 0x2b81f ||
                        0x2b820 <= codePoint && codePoint <= 0x2ceaf ||
                        0x2ceb0 <= codePoint && codePoint <= 0x2ebef ||
                        0x3200 <= codePoint && codePoint <= 0x32ff ||
                        0x31c0 <= codePoint && codePoint <= 0x31ef ||
                        0x3100 <= codePoint && codePoint <= 0x312f ||
                        0x31a0 <= codePoint && codePoint <= 0x31bf ||
                        0x2f00 <= codePoint && codePoint <= 0x2fdf ||
                        0x2ff0 <= codePoint && codePoint <= 0x2fff ||
                        0xff00 <= codePoint && codePoint <= 0xffef ||
                        0x3040 <= codePoint && codePoint <= 0x309f ||
                        0x30a0 <= codePoint && codePoint <= 0x30ff ||
                        0x31f0 <= codePoint && codePoint <= 0x31ff ||
                        0x1b000 <= codePoint && codePoint <= 0x1b0ff ||
                        0x1b100 <= codePoint && codePoint <= 0x1b12f ||
                        0x1100 <= codePoint && codePoint <= 0x11ff ||
                        0xac00 <= codePoint && codePoint <= 0xd7af ||
                        0x3130 <= codePoint && codePoint <= 0x318f ||
                        0xa960 <= codePoint && codePoint <= 0xa97f ||
                        0xd7b0 <= codePoint && codePoint <= 0xd7ff) {
                        return true;
                    }
                }
            }
            catch (err) {
                _iterator.e(err);
            }
            finally {
                _iterator.f();
            }
            return false;
        });
    },
    getCountryFromLocale: function getCountryFromLocale(langCode, countryCode, values) {
        if (values) {
            let isCJK = !countryCode && CJK_LANGUAGES.indexOf(langCode.toLowerCase()) >= 0 || countryCode && CJK_COUNTRIES.indexOf(countryCode.toUpperCase()) >= 0;
            let isJA = !countryCode && 'ja' == langCode.toLowerCase() || countryCode && 'JP' == countryCode.toUpperCase();
            if (!(isJA && this.containsHanScript(values.address, values.city, values.state, values.country)) && isCJK && !this.containsHanScript(values.address)) {
                return this.getCountryFromLocale(langCode, 'EN_' + countryCode);
            }
        }
        let country = countryCode;
        if (langCode == 'uz' && country && country.toLowerCase() == 'latn') {
            country = 'UZ';
        }
        if (!country && languageCodeToCountry.languageCode[langCode]) {
            country = languageCodeToCountry.languageCode[langCode];
        }
        country = this.followReferences(data, country);
        if (!country || !data[country]) {
            return 'US';
        }
        return country;
    },
    getFallback: function getFallback(langCode, countryCode, address) {
        return this.getCountryFromLocale(langCode, countryCode);
    },
};
export { address };
