// lightning-base-components/src/lightning/alert/alert.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { api } from 'lwc';
import LightningInteractiveDialogBase from 'lightning/interactiveDialogBase';
import LightningPrimitiveOverlay from 'lightning/primitiveOverlay';
import { parent, instanceName, secure } from 'lightning/overlayUtils';
import labelDefault from '@salesforce/label/LightningAlert.defaultLabel';
export default class LightningAlert extends LightningPrimitiveOverlay {
    static [instanceName] = 'lightning-alert';
    static [parent] = LightningInteractiveDialogBase;
    @api
    label = labelDefault;
    @api
    message = '';
    @api
    variant = 'header';
    @api
    theme = 'default';
    close(result) {
        const promise = new Promise((resolve) => {
            if (!import.meta.env.SSR) {
                this.dispatchEvent(new CustomEvent('privateclose', {
                    detail: {
                        resolve,
                        [secure]: true,
                    },
                    bubbles: true,
                }));
            }
        });
        super.close(result, promise);
    }
    renderedCallback() {
        const evt = new CustomEvent('privatechildrender', {
            bubbles: true,
            detail: {
                message: this.message,
                label: this.label,
                hideCancel: true,
                showDescription: true,
                role: 'alertdialog',
            },
        });
        this.dispatchEvent(evt);
    }
}
