// lightning-base-components/src/lightning/focusUtils/focus.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { BUBBLE_PREFIX } from 'lightning/tooltipLibrary';
const IGNORE_AUTOFOCUS = 'ignore-autofocus';
export function findAllTabbableElements(container) {
    const result = [];
    traverseActiveTreeRecursively(container, (element) => {
        try {
            if (isTabbable({
                element: element,
                rootContainer: container,
            })) {
                result.push(element);
            }
        }
        catch (e) {
            console.warn(e);
        }
    });
    return result;
}
const FOCUSABLE_NODES = /^input$|^select$|^textarea$|^a$|^button$/;
export function findAllFocusableNodes(container) {
    const result = [];
    traverseActiveTreeRecursively(container, (element) => {
        if (FOCUSABLE_NODES.test(element.tagName.toLowerCase())) {
            result.push(element);
        }
    });
    return result;
}
export function getElementWithFocus() {
    let currentFocusedElement = document.activeElement;
    while (currentFocusedElement) {
        if (currentFocusedElement.shadowRoot) {
            const nextFocusedElement = currentFocusedElement.shadowRoot.activeElement;
            if (nextFocusedElement) {
                currentFocusedElement = nextFocusedElement;
            }
            else {
                return currentFocusedElement;
            }
        }
        else if (currentFocusedElement.contentDocument) {
            const nextFocusedElement = currentFocusedElement.contentDocument.activeElement;
            if (nextFocusedElement) {
                currentFocusedElement = nextFocusedElement;
            }
            else {
                return currentFocusedElement;
            }
        }
        else {
            return currentFocusedElement;
        }
    }
    return undefined;
}
export function returnFocusToElement(element) {
    const isSavedElemInDOM = document.body.contains(element);
    if (element) {
        if (isSavedElemInDOM) {
            element.focus();
            return true;
        }
        const isElementInDOM = isElementInDocument(element);
        if (isElementInDOM) {
            element.focus();
            return true;
        }
    }
    return false;
}
function isElementInDocument(element) {
    let currentElement = element;
    while (currentElement && currentElement.parentNode) {
        if (currentElement.parentNode === document) {
            return true;
        }
        else if (currentElement.parentNode instanceof DocumentFragment) {
            currentElement = currentElement.parentNode.host;
        }
        else {
            currentElement = currentElement.parentNode;
        }
    }
    return false;
}
function traverseActiveTreeRecursively(node, callback) {
    if (!node) {
        return;
    }
    if (node.nodeType === Node.ELEMENT_NODE) {
        if (node.hasAttribute('inert')) {
            return;
        }
        if (isIframe(node)) {
            if (isIframeOfSameOrigin(node)) {
                traverseActiveTreeRecursively(node.contentDocument, callback);
            }
            else if (callback) {
                callback(node);
            }
            return;
        }
        if (callback) {
            callback(node);
        }
        if (node.shadowRoot) {
            traverseActiveTreeRecursively(node.shadowRoot, callback);
            return;
        }
        if (node.localName === 'slot') {
            const slottedNodes = node.assignedNodes({
                flatten: true,
            });
            for (let i = 0; i < slottedNodes.length; i++) {
                traverseActiveTreeRecursively(slottedNodes[i], callback);
            }
            return;
        }
    }
    let child = node.firstChild;
    while (child !== null) {
        traverseActiveTreeRecursively(child, callback);
        child = child.nextSibling;
    }
}
function isIframe(node) {
    return node.tagName === 'IFRAME' || node instanceof HTMLIFrameElement;
}
function isIframeOfSameOrigin(iframe) {
    return !!iframe.contentDocument;
}
const ELEMENTS_WITH_DISABLED_ATTRIBUTE = [
    'button',
    'select',
    'textarea',
    'input',
];
const ELEMENTS_WITH_TABINDEX_ZERO_BY_DEFAULT = [
    'a',
    'select',
    'textarea',
    'input',
    'button',
    'iframe',
    'object',
    'area',
    'frame',
];
function isTabbable({ element, rootContainer, }) {
    const elementLocalName = element.localName;
    if (elementLocalName === 'input' && elementLocalName.type === 'hidden') {
        return false;
    }
    const tabIndexAttribute = element.getAttribute('tabindex');
    if (tabIndexAttribute === '-1') {
        return false;
    }
    if (element.disabled &&
        ELEMENTS_WITH_DISABLED_ATTRIBUTE.includes(element.localName)) {
        return false;
    }
    const hasTabIndexZero = tabIndexAttribute === '0' ||
        (element.tabIndex === 0 &&
            ELEMENTS_WITH_TABINDEX_ZERO_BY_DEFAULT.includes(element.localName));
    return (hasTabIndexZero &&
        isElementVisible(element) &&
        isParentCustomElementTabbable({ element, rootContainer }));
}
const elemHasRoleTooltip = (elem) => {
    return elem.getAttribute('role') === 'tooltip';
};
const elemHasIgnoreAutofocus = (elem) => {
    return elem.classList.contains(IGNORE_AUTOFOCUS);
};
const elemIsHelpTextComponent = (elem) => {
    const tagNameIsButton = elem.tagName.toLowerCase() === 'button';
    const ariaDescValue = elem.getAttribute('aria-describedby');
    const ariaValueMatchesPrefix = ariaDescValue
        ? String(ariaDescValue).indexOf(BUBBLE_PREFIX) === 0
        : false;
    return tagNameIsButton && ariaValueMatchesPrefix;
};
const elemIsNotTooltip = (elem) => {
    if (!elem) {
        return false;
    }
    const elemIsTooltip = [
        elemHasRoleTooltip(elem),
        elemIsHelpTextComponent(elem),
        elemHasIgnoreAutofocus(elem),
    ].some((val) => val === true);
    return !elemIsTooltip;
};
export function filterTooltips(elemsArray) {
    return elemsArray && Array.isArray(elemsArray) && elemsArray.length > 0
        ? elemsArray.filter(elemIsNotTooltip)
        : [];
}
function isElementVisible(element) {
    const { width, height } = element.getBoundingClientRect();
    const nonZeroSize = width > 0 || height > 0;
    return nonZeroSize && getComputedStyle(element).visibility !== 'hidden';
}
function isParentCustomElementTabbable({ element, rootContainer, }) {
    const parentRoot = rootContainer.getRootNode();
    const ownerDocument = element.ownerDocument;
    let root = element.getRootNode();
    while (root !== parentRoot && root !== ownerDocument) {
        const host = root.host;
        if (host.getAttribute('tabindex') === '-1') {
            return false;
        }
        root = host && host.getRootNode();
    }
    return true;
}
