// lightning-base-components/src/lightning/formattedPhone/formattedPhone.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { api } from 'lwc';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import { toNorthAmericanPhoneNumber } from 'lightning/utilsPrivate';
export default class LightningFormattedPhone extends LightningShadowBaseClass {
    @api
    value;
    @api
    disabled = false;
    @api
    tabIndex;
    _connected = false;
    connectedCallback() {
        super.connectedCallback();
        this._connected = true;
    }
    disconnectedCallback() {
        this._connected = false;
    }
    @api
    focus() {
        if (this.phoneAnchor) {
            this.phoneAnchor.focus();
        }
    }
    @api
    blur() {
        if (this.phoneAnchor) {
            this.phoneAnchor.blur();
        }
    }
    @api
    click() {
        const anchor = this.phoneAnchor;
        if (anchor && anchor.click) {
            anchor.click();
        }
    }
    get phoneAnchor() {
        if (this._connected && this.showLink && !this.disabled) {
            return this.template.querySelector('a');
        }
        return undefined;
    }
    get showLink() {
        return this.value != null && this.value !== '';
    }
    get formattedPhoneNumber() {
        return toNorthAmericanPhoneNumber(this.value);
    }
    get link() {
        return `tel:${this.value}`;
    }
}
