// lightning-base-components/src/lightning/tree/tree.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { api, track } from 'lwc';
import { TreeData } from './treeData';
import { keyCodes, deepCopy, isHeadingLevelValid, } from 'lightning/utilsPrivate';
import { privateContext } from 'lightning/utilsInternal';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import { features } from 'lightning/features';
const { setContext, assertContext } = privateContext;
export default class LightningTree extends LightningShadowBaseClass {
    @api
    header;
    @track
    _currentFocusedItem = null;
    @track
    _childNodes;
    @track
    _key;
    @track
    _focusedChild = null;
    @track
    _items = [];
    _defaultFocused = { key: '1', parent: '0' };
    _selected = null;
    @track
    _selectedItem = null;
    hasDetachedListeners = true;
    callbackMap;
    treedata;
    _lastFocusedIndex;
    constructor() {
        super();
        setContext(this);
        this.callbackMap = {};
        this.treedata = null;
        this.addEventListener('privateitemkeydown', this.handleKeydown);
        this.addEventListener('privateitemclick', this.handleClick);
        this.addEventListener('privateregisteritem', this.handleRegistration);
    }
    @track
    _privateHeadingAriaLevel;
    @api
    get headingLevel() {
        return this._privateHeadingAriaLevel;
    }
    set headingLevel(value) {
        if (isHeadingLevelValid(value)) {
            this._privateHeadingAriaLevel = value;
        }
    }
    @api
    get items() {
        return this._items || [];
    }
    set items(value) {
        this.normalizeData(value);
    }
    @api
    get selectedItem() {
        return this._selected;
    }
    set selectedItem(value) {
        this._selected = value;
        this.syncSelected();
    }
    get children() {
        return this._childNodes;
    }
    get rootElement() {
        return this._key;
    }
    get focusedChild() {
        return this._focusedChild;
    }
    get computedLabelledBy() {
        if (this.header) {
            return 'treeheading';
        }
        return null;
    }
    syncSelected() {
        if (this.treedata && this._childNodes.length > 0) {
            const treeRoot = this.treedata.parse(this.items, this._selected);
            this._childNodes = treeRoot ? treeRoot.children : [];
            this._selectedItem = this.treedata.syncSelectedToData(this._selected);
            this.syncCurrentFocused();
            if (this._selectedItem === null) {
                this.setFocusToItem(this._currentFocusedItem, false, false);
            }
        }
    }
    normalizeData(items) {
        if (this.treedata) {
            this._lastFocusedIndex = this.treedata.currentFocusedItemIndex;
        }
        if (items) {
            this.treedata = new TreeData();
            this._items = items.map((item) => {
                return this.treedata.cloneItems(item);
            });
            const treeRoot = this.treedata.parse(this.items, this.selectedItem);
            this._childNodes = treeRoot ? treeRoot.children : [];
            this._selectedItem = treeRoot.selectedItem;
            this.treedata.updateCurrentFocusedItemIndex(this._lastFocusedIndex);
            this._key = this._childNodes.length > 0 ? treeRoot.key : null;
            if (this._key) {
                this.syncCurrentFocused();
            }
        }
    }
    syncCurrentFocused() {
        if (this._selectedItem) {
            this._currentFocusedItem = this._selectedItem;
        }
        else {
            this._currentFocusedItem = this._defaultFocused;
        }
        this.updateCurrentFocusedChild();
    }
    updateCurrentFocusedChild() {
        if (this._key === this._currentFocusedItem.parent) {
            this._focusedChild = this.treedata.getChildNum(this._currentFocusedItem.key);
        }
        else {
            this._focusedChild = this._currentFocusedItem.key;
            this.treedata.updateCurrentFocusedChild(this._currentFocusedItem.key);
        }
    }
    handleTreeFocusIn(event) {
        const relatedTarget = event.relatedTarget;
        if (this._currentFocusedItem &&
            relatedTarget &&
            relatedTarget.tagName !== 'LIGHTNING-TREE-ITEM') {
            this.setFocusToItem(this._currentFocusedItem, false);
        }
    }
    renderedCallback() {
        assertContext(this);
        if (this._selectedItem) {
            this.setFocusToItem(this._currentFocusedItem, false);
        }
        if (this.hasDetachedListeners) {
            const container = this.template.querySelector('.slds-tree_container');
            container.addEventListener('focus', this.handleTreeFocusIn.bind(this));
            this.hasDetachedListeners = false;
        }
    }
    disconnectedCallback() {
        this.hasDetachedListeners = true;
    }
    handleClick = (event) => {
        assertContext(this);
        const key = event.detail.key;
        const target = event.detail.target;
        const item = this.treedata.getItem(key);
        if (item) {
            if (target === 'chevron') {
                if (item.treeNode.nodeRef.expanded) {
                    this.collapseBranch(item.treeNode);
                }
                else {
                    this.expandBranch(item.treeNode);
                }
            }
            else {
                this._selectedItem = item;
                this.dispatchSelectEvent(item.treeNode);
                this.setFocusToItem(item);
            }
        }
    };
    expandBranch(node) {
        assertContext(this);
        if (!node.isLeaf && !node.isDisabled) {
            node.nodeRef.expanded = true;
            if (this._selectedItem?.key.startsWith(node.key)) {
                setTimeout(() => {
                    this.setFocusToItem(this._selectedItem);
                }, 0);
            }
            this.dispatchEvent(new CustomEvent('change', {
                detail: {
                    items: deepCopy(this._items),
                    key: node.nodeRef.name,
                    expand: node.nodeRef.expanded,
                },
            }));
        }
    }
    collapseBranch(node) {
        assertContext(this);
        if (!node.isLeaf && !node.isDisabled) {
            node.nodeRef.expanded = false;
            this.treedata.updateVisibleTreeItemsOnCollapse(node.key);
            this.dispatchEvent(new CustomEvent('change', {
                detail: {
                    items: deepCopy(this._items),
                    key: node.nodeRef.name,
                    expand: node.nodeRef.expanded,
                },
            }));
        }
    }
    dispatchSelectEvent(node) {
        if (!node.isDisabled) {
            const customEvent = new CustomEvent('select', {
                bubbles: true,
                composed: true,
                cancelable: true,
                detail: { name: node.name },
            });
            this.dispatchEvent(customEvent);
        }
    }
    handleKeydown = (event) => {
        assertContext(this);
        event.preventDefault();
        event.stopPropagation();
        const item = this.treedata.getItem(event.detail.key);
        switch (event.detail.keyCode) {
            case keyCodes.up:
                this.setFocusToPrevItem();
                break;
            case keyCodes.down:
                this.setFocusToNextItem();
                break;
            case keyCodes.home:
                this.setFocusToFirstItem();
                break;
            case keyCodes.end:
                this.setFocusToLastItem();
                break;
            case keyCodes.right:
                this.expandBranch(item.treeNode);
                break;
            case keyCodes.left:
                if (item.treeNode.nodeRef.expanded && !item.treeNode.isLeaf) {
                    this.collapseBranch(item.treeNode);
                }
                else {
                    this.handleParentCollapse(event.detail.key);
                }
                break;
            default:
                break;
        }
    };
    setFocusToItem(item, shouldFocus = true, shouldSelect = true) {
        const currentFocused = this.treedata.getItemAtIndex(this.treedata.currentFocusedItemIndex);
        if ((!features.enableTreeGackFix && this.callbackMap[currentFocused.parent]) ||
            (features.enableTreeGackFix && this.callbackMap[currentFocused.key])) {
            this.callbackMap[currentFocused.key].unfocusCallback();
        }
        if (item) {
            this._currentFocusedItem =
                this.treedata.updateCurrentFocusedItemIndex(item.index);
            if (this.callbackMap[item.parent]) {
                this.callbackMap[item.parent].focusCallback(item.key, shouldFocus, shouldSelect);
            }
        }
    }
    setFocusToNextItem() {
        const nextNode = this.treedata.findNextNodeToFocus();
        if (nextNode && nextNode.index !== -1) {
            this.setFocusToItem(nextNode);
        }
    }
    setFocusToPrevItem() {
        const prevNode = this.treedata.findPrevNodeToFocus();
        if (prevNode && prevNode.index !== -1) {
            this.setFocusToItem(prevNode);
        }
    }
    setFocusToFirstItem() {
        const node = this.treedata.findFirstNodeToFocus();
        if (node && node.index !== -1) {
            this.setFocusToItem(node);
        }
    }
    setFocusToLastItem() {
        const lastNode = this.treedata.findLastNodeToFocus();
        if (lastNode && lastNode.index !== -1) {
            this.setFocusToItem(lastNode);
        }
    }
    handleFocusFirst(event) {
        event.stopPropagation();
        this.setFocusToFirstItem();
    }
    handleFocusLast(event) {
        event.stopPropagation();
        this.setFocusToLastItem();
    }
    handleFocusNext(event) {
        event.stopPropagation();
        this.setFocusToNextItem();
    }
    handleFocusPrev(event) {
        event.stopPropagation();
        this.setFocusToPrevItem();
    }
    handleChildBranchCollapse(event) {
        event.stopPropagation();
        this.treedata.updateVisibleTreeItemsOnCollapse(event.detail.key);
    }
    handleParentCollapse(key) {
        const item = this.treedata.getItem(key);
        if (item && item.level > 1) {
            const parent = this.treedata.getItem(item.parent);
            this.collapseBranch(parent.treeNode);
            this.setFocusToItem(parent);
        }
    }
    handleRegistration = (event) => {
        assertContext(this);
        const itemKey = event.detail.key;
        this.callbackMap[itemKey] = {
            focusCallback: event.detail.focusCallback,
            unfocusCallback: event.detail.unfocusCallback,
        };
        this.treedata.addVisible(itemKey);
        event.stopPropagation();
    };
    get hasChildren() {
        return this._items && this._items.length > 0;
    }
}
