// lightning-base-components/src/lightning/primitiveInputSimple/primitiveInputSimple.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { api } from 'lwc';
import { features } from 'lightning/features';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import { privateContext } from 'lightning/utilsInternal';
import labelRequired from '@salesforce/label/LightningControl.required';
import labelClearInput from '@salesforce/label/LightningControl.clear';
import labelLoadingIndicator from '@salesforce/label/LightningControl.loading';
import formFactor from '@salesforce/client/formFactor';
const { setContext, assertContext } = privateContext;
import { normalizeTime, normalizeDate, normalizeUTCDateTime, normalizeDateTimeToUTC, } from './mobileInputUtils';
import { calculateFractionDigitsFromStep, formatNumber, fromIsoDecimal, hasValidNumberShortcut, hasValidNumberSymbol, increaseNumberByStep, isValidNumber, isValidNumberCharacter, stringifyNumber, toIsoDecimal, } from 'lightning/numberUtils';
import { isEmptyString, VARIANT } from 'lightning/inputUtils';
import { reflectAttribute, isSafari, isNativeComponent, isNotUndefinedOrNull, normalizeKeyValue, normalizeString, setDecoratedDragonInputValueWithoutEvent, isDesktopBrowser, } from 'lightning/utilsPrivate';
import { InputSelectionCache } from './selection';
import { normalizeInput } from './normalize';
const i18n = {
    required: labelRequired,
    clear: labelClearInput,
    loading: labelLoadingIndicator,
};
const VALID_NUMBER_FORMATTERS = [
    'decimal',
    'percent',
    'percent-fixed',
    'currency',
];
const DEFAULT_FORMATTER = VALID_NUMBER_FORMATTERS[0];
export default class LightningPrimitiveInputSimple extends LightningShadowBaseClass {
    _type;
    @api
    hasExternalLabel;
    @api
    computedLabelClass;
    @api
    required;
    @api
    label;
    @api
    fieldLevelHelp;
    @api
    helptextAlternativeText;
    @api
    ariaHasPopup;
    @api
    accessKey;
    @api
    autocomplete;
    @api
    isLoading;
    @api
    max;
    @api
    min;
    @api
    maxLength;
    @api
    minLength;
    @api
    pattern;
    @api
    placeholder;
    @api
    name;
    @api
    readOnly;
    @api
    disabled;
    @api
    variant;
    @api
    validity;
    @api
    ariaLabel;
    @api
    ariaInvalid;
    @api
    timezone;
    @api
    ariaKeyShortcuts;
    @api
    ariaDisabled;
    @api
    ariaRoleDescription;
    @api
    ariaExpanded;
    @api
    ariaAutoComplete;
    @api
    get role() {
        if (this.isTypeNumber && formFactor === 'Large') {
            this._role = 'spinbutton';
        }
        return this._role;
    }
    set role(value) {
        this._role = value;
    }
    getHelpMessageElement() {
        return this.template.querySelector('[data-help-message]');
    }
    @api
    get ariaDescribedByElements() {
        return this.getHelpMessageElement();
    }
    @api
    get ariaErrorMessageElement() {
        return this.getHelpMessageElement();
    }
    @api
    get isNativeShadow() {
        return this._isNativeShadow;
    }
    @api
    get type() {
        return this._type;
    }
    set type(value) {
        const normalizedValue = normalizeString(value);
        this._type = normalizedValue;
        if (this._rendered) {
            this._initialValueSet = false;
            if (this.isTypeNumber) {
                this.updateNumberValue(this._value);
            }
        }
    }
    @api
    get formatter() {
        return this._formatter;
    }
    set formatter(value) {
        this._formatter = normalizeString(value, {
            fallbackValue: DEFAULT_FORMATTER,
            validValues: VALID_NUMBER_FORMATTERS,
        });
        this.updateInputDisplayValueIfTypeNumber();
    }
    @api
    get formatFractionDigits() {
        return this._formatFractionDigits;
    }
    set formatFractionDigits(value) {
        this._formatFractionDigits = value;
        if (this._rendered && this.isTypeNumber) {
            this.setInputValue(this.displayedValue);
        }
    }
    @api
    get step() {
        return this._step;
    }
    set step(value) {
        this._step = value;
        if (this._rendered &&
            this.isTypeNumber &&
            !this._showRawNumber &&
            this.inputElement) {
            this.setInputValue(this.displayedValue);
        }
    }
    @api
    get helpMessage() {
        return this._helpMessage;
    }
    set helpMessage(message) {
        this._helpMessage = message;
        reflectAttribute(this, 'invalid', !!message);
    }
    @api
    get value() {
        return this._value;
    }
    set value(value) {
        const previousValue = this._value;
        this._value = normalizeInput(value);
        if (this._rendered) {
            if (this.isTypeNumber) {
                this._value = stringifyNumber(value);
                if (this._value !== previousValue) {
                    this.updateNumberValue(value);
                }
            }
            this.reflectPopulatedState();
        }
    }
    @api
    get inputElement() {
        if (!this.cachedInputElement) {
            let inputElement = this.template.querySelector('input');
            this.cachedInputElement = inputElement;
        }
        return this.cachedInputElement;
    }
    @api
    get inputmode() {
        return this._inputmode || null;
    }
    set inputmode(value) {
        this._inputmode = value;
    }
    @api
    getDisplayedValue() {
        if (this.isTypeNumber) {
            if (this.inputElement.value.length === 1 &&
                hasValidNumberSymbol(this.inputElement.value)) {
                return this.inputElement.value;
            }
            if (this._showRawNumber || this.validity?.valid === false) {
                if (hasValidNumberShortcut(this._numberRawValue) &&
                    isValidNumber(this._numberRawValue)) {
                    this._numberRawValue = fromIsoDecimal(this._value);
                }
                return this._numberRawValue;
            }
            return formatNumber(this._value, this.buildFormatNumberOptions(this.formatter));
        }
        if (this.isTypeMobileDate ||
            this.isTypeMobileDateTime ||
            this.isTypeMobileTime) {
            return this.normalizeDateTimeString(this._value);
        }
        return this._value;
    }
    @api
    getNumberRawValue() {
        return this._numberRawValue;
    }
    _formatter = DEFAULT_FORMATTER;
    _showRawNumber = false;
    _initialValueSet = false;
    _rendered;
    _selectionCache;
    _helpMessage;
    _step;
    _numberRawValue = '';
    _value = '';
    _committedValue;
    _focusOutListenerAttached = false;
    handleFocus(event) {
        if (features.enableAssertContext260) {
            assertContext(this);
        }
        event.stopPropagation();
        if (this._rendered && this.isTypeNumber) {
            this._showRawNumber = true;
            this._selectionCache.preserve(this.inputElement);
            this.inputElement.value = this.displayedValue;
            this._selectionCache.restore(this.inputElement);
        }
        this.dispatchEvent(new CustomEvent('focus'));
    }
    handleBlur(event) {
        event.stopPropagation();
        if (this.validity.valid) {
            this._showRawNumber = false;
        }
        this.dispatchEvent(new CustomEvent('blur'));
    }
    handleKeyDown(event) {
        if (this.isTypeNumber) {
            const hasMetaOrCtrlModifier = event.metaKey || event.ctrlKey;
            if (!hasMetaOrCtrlModifier && !this.readOnly && event.key) {
                const key = normalizeKeyValue(event.key);
                if (key.length === 1 && !isValidNumberCharacter(key)) {
                    event.preventDefault();
                }
                if (key === 'ArrowUp') {
                    event.preventDefault();
                    this.numberStepUpAndDispatchEvents(1);
                }
                else if (key === 'ArrowDown') {
                    event.preventDefault();
                    this.numberStepUpAndDispatchEvents(-1);
                }
            }
        }
        if (this.isTypeSearch && event.key === 'Escape') {
            event.preventDefault();
        }
        if (features.enableInputSearchCommitEvent &&
            this.isTypeSearch &&
            event.key === 'Enter') {
            this.dispatchCommitEvent();
        }
    }
    handleInput(event) {
        event.stopPropagation();
        const inputElement = event.target;
        const inputLength = inputElement.value.length;
        const warningId = 'char-limit-warning';
        let warningEl = this.template.querySelector(`#${warningId}`);
        if (inputLength >= this.maxLength) {
            if (!warningEl) {
                warningEl = document.createElement('p');
                warningEl.textContent = `Limit reached. You can only use ${this.maxLength} characters in this field.`;
                warningEl.setAttribute('role', 'alert');
                warningEl.id = warningId;
                warningEl.classList.add('slds-assistive-text');
                inputElement.after(warningEl);
            }
        }
        else {
            if (warningEl) {
                warningEl.remove();
            }
        }
        if (this.isTypeNumber) {
            this._numberRawValue = this.inputElement.value;
        }
        if (this.value === event.target.value) {
            return;
        }
        this.dispatchChangeEvent();
        this.reflectPopulatedState();
    }
    handleFocusOut(event) {
        const { relatedTarget } = event;
        const focusLeftComponent = !(relatedTarget instanceof Node &&
            this.template.contains(relatedTarget));
        if (focusLeftComponent) {
            this.dispatchCommitEvent();
        }
    }
    handleChange(event) {
        event.stopPropagation();
        if (features.enableInputSearchCommitEvent) {
            if (!this.isTypeSearch) {
                this.dispatchCommitEvent();
            }
        }
        else {
            this.dispatchCommitEvent();
        }
        if (this.isTypeMobileDateTime) {
            if (this.value ===
                normalizeDateTimeToUTC(event.target.value, this.timezone))
                return;
        }
        if (this.value === event.target.value) {
            return;
        }
        this.dispatchChangeEvent();
    }
    handleSelect() {
        if (isSafari) {
            this._selectionCache.preserve(this.inputElement);
        }
    }
    dispatchCommitEvent() {
        if (features.enableInputSearchCommitEvent) {
            if (this._committedValue !== this.value) {
                this._committedValue = this.value;
                this.dispatchEvent(new CustomEvent('commit'));
            }
        }
        else {
            this.dispatchEvent(new CustomEvent('commit'));
        }
    }
    dispatchChangeEvent() {
        const detail = {};
        if (this.isTypeNumber) {
            this._numberRawValue = this.inputElement.value;
            detail.value = toIsoDecimal(this.inputElement.value);
        }
        else if (this.isTypeMobileDateTime) {
            detail.value = normalizeDateTimeToUTC(this.inputElement.value, this.timezone);
        }
        else if (this.isTypeMobileTime) {
            detail.value = normalizeTime(this.inputElement.value);
        }
        else {
            detail.value = this.inputElement.value;
        }
        this._value = detail.value;
        this.dispatchChangeEventWithDetail(detail);
    }
    dispatchChangeEventWithDetail(detail) {
        this.dispatchEvent(new CustomEvent('change', {
            composed: true,
            bubbles: true,
            detail,
        }));
    }
    get internalType() {
        if (this.isTypeNumber || this.isTypeEmail) {
            return 'text';
        }
        return this.type;
    }
    get isTypeNumber() {
        return this.type === 'number';
    }
    get isTypeEmail() {
        return this.type === 'email';
    }
    get isTypeSearch() {
        return this.type === 'search';
    }
    get isTypeText() {
        return this.type === 'text';
    }
    get isTypeDate() {
        return this.type === 'date';
    }
    get isTypeDateTime() {
        return this.type === 'datetime' || this.type === 'datetime-local';
    }
    get isTypeTime() {
        return this.type === 'time';
    }
    get isTypeMobileDate() {
        return this.isTypeDate && !isDesktopBrowser();
    }
    get isTypeMobileTime() {
        return this.isTypeTime && !isDesktopBrowser();
    }
    get isTypeMobileDateTime() {
        return this.isTypeDateTime && !isDesktopBrowser();
    }
    get computedHelpMessageClass() {
        return this.variant === VARIANT.LABEL_INLINE
            ? 'slds-form-element__help'
            : 'slds-form-element__help slds-m-left_none';
    }
    get showClearButton() {
        return (this.isTypeSearch &&
            isNotUndefinedOrNull(this._value) &&
            this._value !== '');
    }
    get displayedValue() {
        if (this.isTypeNumber) {
            return this.getDisplayedValue();
        }
        if (this.isTypeMobileTime ||
            this.isTypeMobileDate ||
            this.isTypeMobileDateTime) {
            return this.normalizeDateTimeString(this._value, this.timezone);
        }
        return this._value;
    }
    normalizeDateTimeString(value) {
        if (this.isTypeDate) {
            return normalizeDate(value);
        }
        else if (this.isTypeTime) {
            return normalizeTime(value);
        }
        else if (this.isTypeDateTime) {
            return normalizeUTCDateTime(value, this.timezone);
        }
        return value;
    }
    get i18n() {
        return i18n;
    }
    get showErrorIcon() {
        return (this._helpMessage &&
            !this.isTypeSearch);
    }
    get computedFormElementClass() {
        return {
            'slds-form-element__control': true,
            'slds-grow': true,
            'slds-input-has-icon': this.isTypeSearch || this.showErrorIcon,
            'slds-input-has-icon_left-right': this.isTypeSearch,
            'slds-input-has-icon_left': this.showErrorIcon,
        };
    }
    updateNumberValue(value) {
        const newValue = stringifyNumber(value);
        this._value = newValue;
        this._numberRawValue = fromIsoDecimal(newValue);
    }
    numberStepUpAndDispatchEvents(increment) {
        if (this._readOnly || this._disabled) {
            return;
        }
        this._value = increaseNumberByStep({
            value: this._value,
            step: this.step,
            increment,
            fractionDigits: this.buildFormatNumberOptions(this.formatter)
                .minimumFractionDigits,
        });
        this._numberRawValue = fromIsoDecimal(this._value);
        this.setInputValue(this.displayedValue);
        this.dispatchChangeEvent();
        this.dispatchCommitEvent();
    }
    buildFormatNumberOptions(formatter) {
        const options = {
            style: formatter,
        };
        if (this._formatFractionDigits !== undefined) {
            options.minimumFractionDigits = this._formatFractionDigits;
            options.maximumFractionDigits = this._formatFractionDigits;
        }
        else {
            let digitsFromStep = calculateFractionDigitsFromStep(this._step);
            if (formatter === 'percent' && typeof digitsFromStep === 'number') {
                digitsFromStep -= 2;
                if (digitsFromStep < 0) {
                    digitsFromStep = 0;
                }
            }
            options.minimumFractionDigits = digitsFromStep;
            options.maximumFractionDigits = digitsFromStep;
        }
        return options;
    }
    updateInputDisplayValueIfTypeNumber() {
        if (this._rendered &&
            this.isTypeNumber &&
            !this._showRawNumber &&
            this.inputElement) {
            this.setInputValue(this.displayedValue);
        }
    }
    clearAndSetFocusOnInput(event) {
        if (this.disabled)
            return;
        event.stopPropagation();
        this.setInputValue('');
        this._value = '';
        this.inputElement.focus();
        this.dispatchChangeEventWithDetail({
            value: this._value,
        });
        this.dispatchCommitEvent();
    }
    setInputValue(value) {
        setDecoratedDragonInputValueWithoutEvent(this.inputElement, value);
        this.reflectPopulatedState();
    }
    reflectPopulatedState() {
        if (this.refs) {
            reflectAttribute(this.refs.container, 'populated', !isEmptyString(this._value) ||
                !isEmptyString(this._numberRawValue));
        }
    }
    connectedCallback() {
        super.connectedCallback();
        if (this.isTypeNumber) {
            this.updateNumberValue(this._value);
        }
        this._isNativeShadow = isNativeComponent(this);
    }
    renderedCallback() {
        if (isSafari) {
            this.inputElement.addEventListener('select', this.handleSelect.bind(this));
        }
        if (!this._initialValueSet && this.inputElement) {
            this._rendered = true;
            if (this.isTypeNumber) {
                this._numberRawValue = fromIsoDecimal(this._value);
            }
            this.setInputValue(this.displayedValue);
            this._initialValueSet = true;
        }
        reflectAttribute(this, 'variant', this.variant);
        reflectAttribute(this, 'disabled', this.disabled);
        reflectAttribute(this, 'readonly', this.readOnly);
        if (this.isTypeSearch &&
            !this._focusOutListenerAttached &&
            features.enableInputSearchCommitEvent) {
            this.focusOutEventHandler = this.handleFocusOut.bind(this);
            this.refs.container.addEventListener('focusout', this.focusOutEventHandler);
            this._focusOutListenerAttached = true;
        }
    }
    disconnectedCallback() {
        if (this.isTypeSearch && features.enableInputSearchCommitEvent) {
            this.refs.container.removeEventListener('focusout', this.focusOutEventHandler);
            this._focusOutListenerAttached = false;
        }
    }
    constructor() {
        super();
        if (features.enableAssertContext260) {
            setContext(this);
        }
        this._selectionCache = new InputSelectionCache();
    }
}
