// lightning-base-components/src/lightning/primitiveInputSimple/selection.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { isSafari, isIE11, timeout } from 'lightning/utilsPrivate';
function restoreCacheToInput(input, cache) {
    input.setSelectionRange(cache._selectionStartCached, cache._selectionEndCached);
    cache.clearCache();
}
export class InputSelectionCache {
    _selectionStartCached = null;
    _selectionEndCached = null;
    _cacheSelectionRange(start, end) {
        if (typeof start === 'number' && typeof end === 'number') {
            this._selectionStartCached = start;
            this._selectionEndCached = end;
        }
    }
    clearCache() {
        this._selectionStartCached = null;
        this._selectionEndCached = null;
    }
    hasSelection(input) {
        return (input.selectionStart !== null &&
            input.selectionStart !== input.selectionEnd);
    }
    isCached() {
        return (this._selectionStartCached !== null &&
            this._selectionEndCached !== null);
    }
    preserve(input) {
        if (!this.hasSelection(input) && this.isCached()) {
            return;
        }
        if (this.hasSelection(input)) {
            this._cacheSelectionRange(input.selectionStart, input.selectionEnd);
        }
    }
    restore(input) {
        if (!this.isCached()) {
            return Promise.resolve();
        }
        if (isSafari) {
            return timeout(0).then(() => {
                if (this.isCached() && input) {
                    restoreCacheToInput(input, this);
                }
            });
        }
        if (isIE11) {
            const length = input.value.length;
            input.setSelectionRange(length, length);
            return Promise.resolve();
        }
        restoreCacheToInput(input, this);
        return Promise.resolve();
    }
}
