// lightning-base-components/src/lightning/tile/tile.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import labelShowActions from '@salesforce/label/LightningPrimitiveCellActions.showActions';
import { assert, classListMutation, normalizeString, sanitizeURL, } from 'lightning/utilsPrivate';
import { api, track } from 'lwc';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import mediaTile from './mediaTile.html';
import standardTile from './standardTile.html';
const VALID_TYPE_VALUES = ['standard', 'media'];
const i18n = {
    showActions: labelShowActions,
};
export default class LightningTile extends LightningShadowBaseClass {
    static validationOptOut = ['class'];
    @api
    label;
    @track
    _type = 'standard';
    @track
    _actions = [];
    _href;
    @api
    get actions() {
        return this._actions;
    }
    set actions(actions) {
        this._actions = actions;
    }
    @api
    get href() {
        return this._href;
    }
    set href(value) {
        this._href = sanitizeURL(value);
    }
    @api
    get type() {
        return this._type;
    }
    set type(value) {
        assert(VALID_TYPE_VALUES.indexOf(value) !== -1, `Invalid type attribute value of ${value}. Must be one of ${VALID_TYPE_VALUES}.`);
        this._type = normalizeString(value, {
            fallbackValue: 'standard',
            validValues: VALID_TYPE_VALUES,
        });
        if (this._connected) {
            this.setClassesAndTemplate();
        }
    }
    handleActionSelect(event) {
        this.dispatchEvent(new CustomEvent('actiontriggered', {
            detail: {
                action: event.detail.value,
            },
        }));
    }
    setClassesAndTemplate() {
        classListMutation(this.classList, {
            'slds-media': this.isMedia,
            'slds-hint-parent': this.hasActions,
        });
    }
    connectedCallback() {
        super.connectedCallback();
        this._connected = true;
        this.classList.add('slds-tile');
        this.setClassesAndTemplate();
    }
    render() {
        return this.isMedia ? mediaTile : standardTile;
    }
    get isMedia() {
        return this.type === 'media';
    }
    get hasActions() {
        return Object.keys(this.actions).length > 0;
    }
    get buttonAlternateText() {
        return `${i18n.showActions}`;
    }
}
