// lightning-base-components/src/lightning/treeGrid/__examples__/sorting/sorting.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { LightningElement } from 'lwc';
import { EXAMPLES_COLUMNS_DEFINITION_SORTING, EXAMPLES_DATA_SORTING, } from './sampleData';
export default class TreeGridSorting extends LightningElement {
    gridColumns = EXAMPLES_COLUMNS_DEFINITION_SORTING;
    gridData = EXAMPLES_DATA_SORTING;
    defaultSortDirection = 'asc';
    sortDirection = 'asc';
    sortedBy;
    sortBy(field, reverse, primer) {
        const key = primer
            ? function (x) {
                return primer(x[field]);
            }
            : function (x) {
                return x[field];
            };
        return function (a, b) {
            a = key(a);
            b = key(b);
            return reverse * ((a > b) - (b > a));
        };
    }
    recursiveSort(arr, sortedBy, sortDirection) {
        arr.sort(this.sortBy(sortedBy, sortDirection === 'asc' ? 1 : -1));
        for (let i = 0; i < arr.length; i += 1) {
            if (Array.isArray(arr[i]?._children)) {
                this.recursiveSort(arr[i]._children, sortedBy, sortDirection);
            }
        }
    }
    onHandleSort(event) {
        const { fieldName: sortedBy, sortDirection } = event.detail;
        const cloneData = [...this.gridData];
        this.recursiveSort(cloneData, sortedBy, sortDirection);
        this.gridData = cloneData;
        this.sortDirection = sortDirection;
        this.sortedBy = sortedBy;
    }
}
