// lightning-base-components/src/lightning/progressBar/progressBar.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import labelProgress from '@salesforce/label/LightningProgressBar.progress';
import labelProgressBar from '@salesforce/label/LightningProgressBar.progressBar';
import { track, api } from 'lwc';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import { classSet } from 'lightning/utils';
import { normalizeString as normalize, isUndefinedOrNull, } from 'lightning/utilsPrivate';
import { numberFormat } from 'lightning/internationalizationLibrary';
const i18n = {
    progress: labelProgress,
    progressBar: labelProgressBar,
};
const DEFAULT_SIZE = 'medium';
const DEFAULT_VARIANT = 'base';
export default class LightningProgressBar extends LightningShadowBaseClass {
    @track
    privateVariant = DEFAULT_VARIANT;
    @track
    privateSize = DEFAULT_SIZE;
    @api
    value = 0;
    @api
    ariaLabel;
    @api
    get variant() {
        return this.privateVariant;
    }
    set variant(value) {
        this.privateVariant = normalize(value, {
            fallbackValue: DEFAULT_VARIANT,
            validValues: ['base', 'circular'],
        });
    }
    @api
    get size() {
        return this.privateSize;
    }
    set size(value) {
        this.privateSize = normalize(value, {
            fallbackValue: DEFAULT_SIZE,
            validValues: ['x-small', 'small', 'medium', 'large'],
        });
    }
    get ariaBusy() {
        const value = this.percentValue;
        if (value > 0 && value < 100) {
            return 'true';
        }
        return null;
    }
    get computedClass() {
        const { size, variant } = this;
        const classes = classSet('slds-progress-bar');
        classes.add(`slds-progress-bar_${size}`);
        if (variant === 'circular') {
            classes.add('slds-progress-bar_circular');
        }
        return classes.toString();
    }
    get percentValue() {
        const { value } = this;
        if (!value || value <= 0) {
            return 0;
        }
        if (value >= 100) {
            return 100;
        }
        return Math.round(value);
    }
    get computedStyle() {
        return `width: ${this.percentValue}%;`;
    }
    get assistiveText() {
        const formattedPercent = numberFormat({ style: 'percent' }).format(this.percentValue / 100);
        return `${i18n.progress} ${formattedPercent}`;
    }
    get computeAriaLabel() {
        return isUndefinedOrNull(this.ariaLabel)
            ? i18n.progressBar
            : this.ariaLabel;
    }
}
