// lightning-base-components/src/lightning/overlayPositionUtils/anchorPositioningAdapter.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { supportsCSSAnchorPositioning, CSSAnchorPositioning, } from './cssAnchorPositioning';
import { PositionLibraryFallback } from './positionLibraryFallback';
export class AnchorPositioningAdapter {
    element;
    config;
    usesNativeAnchor;
    cssAnchorPositioning;
    positionLibraryFallback;
    constructor(element, config = {}) {
        if (!element) {
            console.error('AnchorPositioningAdapter requires an element');
            return;
        }
        this.element = element;
        this.config = config;
        this.usesNativeAnchor = supportsCSSAnchorPositioning();
        this.cssAnchorPositioning = null;
        this.positionLibraryFallback = null;
    }
    apply() {
        if (this.usesNativeAnchor) {
            if (!this.cssAnchorPositioning) {
                this.cssAnchorPositioning = new CSSAnchorPositioning(this.element, this.config);
            }
            this.cssAnchorPositioning.apply();
        }
        else {
            if (!this.positionLibraryFallback) {
                this.positionLibraryFallback = new PositionLibraryFallback(this.element, this.config);
            }
            this.positionLibraryFallback.apply();
        }
    }
    destroy() {
        if (this.cssAnchorPositioning) {
            this.cssAnchorPositioning.cleanup();
            this.cssAnchorPositioning = null;
        }
        if (this.positionLibraryFallback) {
            this.positionLibraryFallback.cleanup();
            this.positionLibraryFallback = null;
        }
    }
}
