// lightning-base-components/src/lightning/overlayPositionUtils/cssAnchorPositioning.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { convertLogicalToDirectionalInsets, normalizeSizingValue, } from './overlayPositionUtils';
export function supportsCSSAnchorPositioning() {
    if (import.meta.env.SSR) {
        return false;
    }
    const supportsAnchorName = 'anchorName' in document.documentElement.style;
    const supportsPositionAnchor = typeof CSS !== 'undefined' &&
        CSS.supports &&
        CSS.supports('position-anchor', '--test');
    return supportsAnchorName && supportsPositionAnchor;
}
export class CSSAnchorPositioning {
    element;
    config;
    constructor(element, config = {}) {
        this.element = element;
        this.config = config;
    }
    apply() {
        this.applyPanelStyles();
        this.setPositionArea();
    }
    setPositionArea() {
        this.element?.style?.setProperty('position-area', this.config.position);
    }
    applyPanelStyles() {
        const { minWidth, maxWidth, minHeight, maxHeight, position, offsetInline, offsetBlock, } = this.config;
        const elem = this.element;
        const normalizedMinWidth = normalizeSizingValue(minWidth);
        const normalizedMaxWidth = normalizeSizingValue(maxWidth);
        const normalizedMinHeight = normalizeSizingValue(minHeight);
        const normalizedMaxHeight = normalizeSizingValue(maxHeight);
        if (normalizedMinWidth) {
            elem.style.minWidth = normalizedMinWidth;
        }
        if (normalizedMaxWidth) {
            elem.style.maxWidth = normalizedMaxWidth;
        }
        if (normalizedMinHeight) {
            elem.style.minHeight = normalizedMinHeight;
        }
        if (normalizedMaxHeight) {
            elem.style.maxHeight = normalizedMaxHeight;
        }
        this.applyLogicalInsets(position, offsetInline, offsetBlock);
    }
    applyLogicalInsets(position, offsetInline, offsetBlock) {
        const elem = this.element;
        const { top, bottom, left, right } = convertLogicalToDirectionalInsets(position, offsetInline, offsetBlock);
        if (left) {
            elem.style.setProperty('left', `${left}px`);
        }
        if (right) {
            elem.style.setProperty('right', `${right}px`);
        }
        if (top) {
            elem.style.setProperty('top', `${top}px`);
        }
        if (bottom) {
            elem.style.setProperty('bottom', `${bottom}px`);
        }
    }
    cleanup() {
        if (this.element) {
            const propertiesToRemove = [
                'position-anchor',
                'position-area',
                'left',
                'right',
                'top',
                'bottom',
                'min-width',
                'max-width',
                'min-height',
                'max-height',
                'margin-inline-start',
                'margin-inline-end',
                'margin-block-start',
                'margin-block-end',
            ];
            propertiesToRemove.forEach((prop) => {
                this.element?.style?.removeProperty(prop);
            });
            if (this.element?.style?.position === 'absolute') {
                this.element?.style?.removeProperty('position');
            }
        }
    }
}
