// lightning-base-components/src/lightning/overlayPositionUtils/overlayPositionUtils.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
export { AnchorPositioningAdapter } from './anchorPositioningAdapter';
export const OverlayPosition = {
    TopLeft: 'top left',
    TopCenter: 'top center',
    TopRight: 'top right',
    TopSpanLeft: 'top span-left',
    TopSpanAll: 'top span-all',
    TopSpanRight: 'top span-right',
    CenterLeft: 'center left',
    CenterCenter: 'center center',
    CenterRight: 'center right',
    CenterSpanLeft: 'center span-left',
    CenterSpanRight: 'center span-right',
    CenterSpanAll: 'center span-all',
    BottomLeft: 'bottom left',
    BottomCenter: 'bottom center',
    BottomRight: 'bottom right',
    BottomSpanLeft: 'bottom span-left',
    BottomSpanAll: 'bottom span-all',
    BottomSpanRight: 'bottom span-right',
    SpanTopLeft: 'span-top left',
    SpanTopCenter: 'span-top center',
    SpanTopRight: 'span-top right',
    SpanTopSpanLeft: 'span-top span-left',
    SpanTopSpanAll: 'span-top span-all',
    SpanTopSpanRight: 'span-top span-right',
    SpanAllLeft: 'span-all left',
    SpanAllCenter: 'span-all center',
    SpanAllRight: 'span-all right',
    SpanAllSpanLeft: 'span-all span-left',
    SpanAllSpanAll: 'span-all span-all',
    SpanAllSpanRight: 'span-all span-right',
    SpanBottomLeft: 'span-bottom left',
    SpanBottomCenter: 'span-bottom center',
    SpanBottomRight: 'span-bottom right',
    SpanBottomSpanLeft: 'span-bottom span-left',
    SpanBottomSpanAll: 'span-bottom span-all',
    SpanBottomSpanRight: 'span-bottom span-right',
};
export const VALID_POSITION_VALUES = Object.values(OverlayPosition);
export const DEFAULT_POSITION = OverlayPosition.TopRight;
export const VALID_POPOVER_VALUES = ['auto', 'manual', 'hint'];
export const DEFAULT_POPOVER = 'auto';
export function normalizeSizingValue(value) {
    if (value === null || value === undefined || value === '') {
        return 'auto';
    }
    const strValue = String(value).trim();
    const cssUnitPattern = /^-?\d*\.?\d+(px|rem|em|%|vh|vw|vmin|vmax|ch|ex|cm|mm|in|pt|pc)$/i;
    if (cssUnitPattern.test(strValue)) {
        return strValue;
    }
    const plainNumberPattern = /^-?\d*\.?\d+$/;
    if (plainNumberPattern.test(strValue)) {
        return `${strValue}rem`;
    }
    const validKeywords = [
        'auto',
        'max-content',
        'min-content',
        'fit-content',
        'inherit',
        'initial',
        'unset',
        'revert',
    ];
    if (validKeywords.includes(strValue.toLowerCase())) {
        return strValue.toLowerCase();
    }
    if (strValue.includes('calc(') ||
        strValue.includes('var(') ||
        strValue.includes('min(') ||
        strValue.includes('max(') ||
        strValue.includes('clamp(')) {
        return strValue;
    }
    return 'auto';
}
export function convertLogicalToDirectionalInsets(position, offsetInline, offsetBlock) {
    let top = 0;
    let bottom = 0;
    let left = 0;
    let right = 0;
    const [vertical, horizontal] = position.toLowerCase().split(/\s+/);
    if (horizontal) {
        if (horizontal.includes('right')) {
            left = offsetInline;
        }
        else if (horizontal.includes('left')) {
            right = offsetInline;
        }
    }
    if (vertical) {
        if (vertical.includes('top')) {
            bottom = offsetBlock;
        }
        else if (vertical.includes('bottom')) {
            top = offsetBlock;
        }
    }
    return { top, bottom, left, right };
}
