// lightning-base-components/src/lightning/overlayPositionUtils/positionLibraryFallback.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { AutoPosition, Direction } from 'lightning/positionLibrary';
import { normalizeSizingValue } from './overlayPositionUtils';
import { isRTL } from 'lightning/utilsPrivate';
export class PositionLibraryFallback {
    element;
    config;
    autoPosition;
    positionUpdater;
    constructor(element, config = {}) {
        this.element = element;
        this.config = config;
        this.autoPosition = null;
        this.positionUpdater = null;
    }
    getPhysicalDirection(direction) {
        if (isRTL()) {
            if (direction === Direction.Left) {
                return Direction.Right;
            }
            if (direction === Direction.Right) {
                return Direction.Left;
            }
        }
        return direction;
    }
    shouldAlignWidth(position) {
        const horizontal = position.toLowerCase().split(/\s+/)[1];
        return horizontal === 'center';
    }
    apply() {
        const { source, position, offsetInline = 0, offsetBlock = 0, } = this.config;
        if (this.element?.style?.display === 'none') {
            this.element.style.display = '';
        }
        this.applySizingStyles();
        if (!this.autoPosition) {
            this.autoPosition = new AutoPosition(this.element);
        }
        let alignment = this.convertPositionToAlignment(position);
        alignment = {
            align: {
                horizontal: this.getPhysicalDirection(alignment.align.horizontal),
                vertical: alignment.align.vertical,
            },
            targetAlign: {
                horizontal: this.getPhysicalDirection(alignment.targetAlign.horizontal),
                vertical: alignment.targetAlign.vertical,
            },
        };
        const padding = {
            padLeft: offsetInline,
            padTop: offsetBlock,
        };
        const positionConfig = {
            target: () => source,
            element: () => this.element,
            align: alignment.align,
            targetAlign: alignment.targetAlign,
            alignWidth: this.shouldAlignWidth(position),
            autoFlip: true,
            considerPadForFlip: true,
            autoShrink: true,
            ...padding,
        };
        try {
            const autoPositionUpdater = this.autoPosition.start(positionConfig);
            if (autoPositionUpdater) {
                this.positionUpdater = autoPositionUpdater;
            }
        }
        catch (error) {
            console.error('Error positioning element with positionLibrary:', error);
        }
    }
    applySizingStyles() {
        const { minWidth, maxWidth, minHeight, maxHeight } = this.config;
        const normalizedMinWidth = normalizeSizingValue(minWidth);
        const normalizedMaxWidth = normalizeSizingValue(maxWidth);
        const normalizedMinHeight = normalizeSizingValue(minHeight);
        const normalizedMaxHeight = normalizeSizingValue(maxHeight);
        if (normalizedMinWidth) {
            this.element.style.minWidth = normalizedMinWidth;
        }
        if (normalizedMaxWidth) {
            this.element.style.maxWidth = normalizedMaxWidth;
        }
        if (normalizedMinHeight) {
            this.element.style.minHeight = normalizedMinHeight;
        }
        if (normalizedMaxHeight) {
            this.element.style.maxHeight = normalizedMaxHeight;
        }
    }
    convertPositionToAlignment(position) {
        const [vertical, horizontal] = position.toLowerCase().split(/\s+/);
        const alignmentMap = {
            'top left': {
                align: {
                    horizontal: Direction.Right,
                    vertical: Direction.Bottom,
                },
                targetAlign: {
                    horizontal: Direction.Left,
                    vertical: Direction.Top,
                },
            },
            'top center': {
                align: {
                    horizontal: Direction.Center,
                    vertical: Direction.Bottom,
                },
                targetAlign: {
                    horizontal: Direction.Center,
                    vertical: Direction.Top,
                },
            },
            'top right': {
                align: {
                    horizontal: Direction.Left,
                    vertical: Direction.Bottom,
                },
                targetAlign: {
                    horizontal: Direction.Right,
                    vertical: Direction.Top,
                },
            },
            'top span-right': {
                align: {
                    horizontal: Direction.Left,
                    vertical: Direction.Bottom,
                },
                targetAlign: {
                    horizontal: Direction.Left,
                    vertical: Direction.Top,
                },
            },
            'top span-all': {
                align: {
                    horizontal: Direction.Center,
                    vertical: Direction.Bottom,
                },
                targetAlign: {
                    horizontal: Direction.Center,
                    vertical: Direction.Top,
                },
            },
            'top span-left': {
                align: {
                    horizontal: Direction.Right,
                    vertical: Direction.Bottom,
                },
                targetAlign: {
                    horizontal: Direction.Right,
                    vertical: Direction.Top,
                },
            },
            'center left': {
                align: {
                    horizontal: Direction.Right,
                    vertical: Direction.Center,
                },
                targetAlign: {
                    horizontal: Direction.Left,
                    vertical: Direction.Center,
                },
            },
            'center center': {
                align: {
                    horizontal: Direction.Center,
                    vertical: Direction.Center,
                },
                targetAlign: {
                    horizontal: Direction.Center,
                    vertical: Direction.Center,
                },
            },
            'center right': {
                align: {
                    horizontal: Direction.Left,
                    vertical: Direction.Center,
                },
                targetAlign: {
                    horizontal: Direction.Right,
                    vertical: Direction.Center,
                },
            },
            'center span-right': {
                align: {
                    horizontal: Direction.Left,
                    vertical: Direction.Center,
                },
                targetAlign: {
                    horizontal: Direction.Left,
                    vertical: Direction.Center,
                },
            },
            'center span-left': {
                align: {
                    horizontal: Direction.Right,
                    vertical: Direction.Center,
                },
                targetAlign: {
                    horizontal: Direction.Right,
                    vertical: Direction.Center,
                },
            },
            'center span-all': {
                align: {
                    horizontal: Direction.Center,
                    vertical: Direction.Center,
                },
                targetAlign: {
                    horizontal: Direction.Center,
                    vertical: Direction.Center,
                },
            },
            'bottom left': {
                align: { horizontal: Direction.Right, vertical: Direction.Top },
                targetAlign: {
                    horizontal: Direction.Left,
                    vertical: Direction.Bottom,
                },
            },
            'bottom center': {
                align: {
                    horizontal: Direction.Center,
                    vertical: Direction.Top,
                },
                targetAlign: {
                    horizontal: Direction.Center,
                    vertical: Direction.Bottom,
                },
            },
            'bottom right': {
                align: { horizontal: Direction.Left, vertical: Direction.Top },
                targetAlign: {
                    horizontal: Direction.Right,
                    vertical: Direction.Bottom,
                },
            },
            'bottom span-right': {
                align: { horizontal: Direction.Left, vertical: Direction.Top },
                targetAlign: {
                    horizontal: Direction.Left,
                    vertical: Direction.Bottom,
                },
            },
            'bottom span-all': {
                align: {
                    horizontal: Direction.Center,
                    vertical: Direction.Top,
                },
                targetAlign: {
                    horizontal: Direction.Center,
                    vertical: Direction.Bottom,
                },
            },
            'bottom span-left': {
                align: { horizontal: Direction.Right, vertical: Direction.Top },
                targetAlign: {
                    horizontal: Direction.Right,
                    vertical: Direction.Bottom,
                },
            },
            'span-top left': {
                align: {
                    horizontal: Direction.Right,
                    vertical: Direction.Bottom,
                },
                targetAlign: {
                    horizontal: Direction.Left,
                    vertical: Direction.Bottom,
                },
            },
            'span-top center': {
                align: {
                    horizontal: Direction.Center,
                    vertical: Direction.Bottom,
                },
                targetAlign: {
                    horizontal: Direction.Center,
                    vertical: Direction.Bottom,
                },
            },
            'span-top right': {
                align: {
                    horizontal: Direction.Left,
                    vertical: Direction.Bottom,
                },
                targetAlign: {
                    horizontal: Direction.Right,
                    vertical: Direction.Bottom,
                },
            },
            'span-top span-right': {
                align: {
                    horizontal: Direction.Left,
                    vertical: Direction.Bottom,
                },
                targetAlign: {
                    horizontal: Direction.Left,
                    vertical: Direction.Bottom,
                },
            },
            'span-top span-all': {
                align: {
                    horizontal: Direction.Center,
                    vertical: Direction.Bottom,
                },
                targetAlign: {
                    horizontal: Direction.Center,
                    vertical: Direction.Bottom,
                },
            },
            'span-top span-left': {
                align: {
                    horizontal: Direction.Right,
                    vertical: Direction.Bottom,
                },
                targetAlign: {
                    horizontal: Direction.Right,
                    vertical: Direction.Bottom,
                },
            },
            'span-all span-all': {
                align: {
                    horizontal: Direction.Center,
                    vertical: Direction.Center,
                },
                targetAlign: {
                    horizontal: Direction.Center,
                    vertical: Direction.Center,
                },
            },
            'span-all left': {
                align: {
                    horizontal: Direction.Right,
                    vertical: Direction.Center,
                },
                targetAlign: {
                    horizontal: Direction.Left,
                    vertical: Direction.Center,
                },
            },
            'span-all center': {
                align: {
                    horizontal: Direction.Center,
                    vertical: Direction.Center,
                },
                targetAlign: {
                    horizontal: Direction.Center,
                    vertical: Direction.Center,
                },
            },
            'span-all right': {
                align: {
                    horizontal: Direction.Left,
                    vertical: Direction.Center,
                },
                targetAlign: {
                    horizontal: Direction.Right,
                    vertical: Direction.Center,
                },
            },
            'span-bottom left': {
                align: { horizontal: Direction.Right, vertical: Direction.Top },
                targetAlign: {
                    horizontal: Direction.Left,
                    vertical: Direction.Top,
                },
            },
            'span-bottom center': {
                align: {
                    horizontal: Direction.Center,
                    vertical: Direction.Top,
                },
                targetAlign: {
                    horizontal: Direction.Center,
                    vertical: Direction.Top,
                },
            },
            'span-bottom right': {
                align: { horizontal: Direction.Left, vertical: Direction.Top },
                targetAlign: {
                    horizontal: Direction.Right,
                    vertical: Direction.Top,
                },
            },
            'span-bottom span-right': {
                align: { horizontal: Direction.Left, vertical: Direction.Top },
                targetAlign: {
                    horizontal: Direction.Left,
                    vertical: Direction.Top,
                },
            },
            'span-bottom span-all': {
                align: {
                    horizontal: Direction.Center,
                    vertical: Direction.Top,
                },
                targetAlign: {
                    horizontal: Direction.Center,
                    vertical: Direction.Top,
                },
            },
            'span-bottom span-left': {
                align: { horizontal: Direction.Right, vertical: Direction.Top },
                targetAlign: {
                    horizontal: Direction.Right,
                    vertical: Direction.Top,
                },
            },
        };
        const key = horizontal ? `${vertical} ${horizontal}` : vertical;
        return alignmentMap[key];
    }
    cleanup() {
        if (this.autoPosition) {
            this.autoPosition.stop();
            this.autoPosition = null;
        }
        this.positionUpdater = null;
        if (this.element) {
            this.element?.style?.removeProperty('min-width');
            this.element?.style?.removeProperty('max-width');
            this.element?.style?.removeProperty('min-height');
            this.element?.style?.removeProperty('max-height');
        }
    }
}
