// lightning-base-components/src/lightning/formattedRichTextRenderer/colorShiftUtil.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
export function srgbToLinear(v) {
    v /= 255;
    return v <= 0.04045 ? v / 12.92 : Math.pow((v + 0.055) / 1.055, 2.4);
}
export function getLuminance({ r, g, b }) {
    const linearR = srgbToLinear(r);
    const linearG = srgbToLinear(g);
    const linearB = srgbToLinear(b);
    return (0.2126 * linearR) + (0.7152 * linearG) + (0.0722 * linearB);
}
export function getContrastRatio(color1, color2) {
    const lum1 = getLuminance(color1);
    const lum2 = getLuminance(color2);
    const lighter = Math.max(lum1, lum2);
    const darker = Math.min(lum1, lum2);
    return (lighter + 0.05) / (darker + 0.05);
}
export function hslToRgb(h, s, l) {
    s /= 100;
    l /= 100;
    const c = (1 - Math.abs((2 * l) - 1)) * s, x = c * (1 - Math.abs(((h / 60) % 2) - 1)), m = l - (c / 2);
    let r = 0, g = 0, b = 0;
    if (h >= 0 && h < 60) {
        r = c;
        g = x;
        b = 0;
    }
    else if (h >= 60 && h < 120) {
        r = x;
        g = c;
        b = 0;
    }
    else if (h >= 120 && h < 180) {
        r = 0;
        g = c;
        b = x;
    }
    else if (h >= 180 && h < 240) {
        r = 0;
        g = x;
        b = c;
    }
    else if (h >= 240 && h < 300) {
        r = x;
        g = 0;
        b = c;
    }
    else if (h >= 300 && h < 360) {
        r = c;
        g = 0;
        b = x;
    }
    return {
        r: Math.round((r + m) * 255),
        g: Math.round((g + m) * 255),
        b: Math.round((b + m) * 255)
    };
}
export function rgbToHsl({ r, g, b }) {
    r /= 255;
    g /= 255;
    b /= 255;
    const max = Math.max(r, g, b), min = Math.min(r, g, b);
    let h, s;
    const l = (max + min) / 2;
    if (max === min) {
        h = s = 0;
    }
    else {
        const d = max - min;
        s = l > 0.5 ? d / (2 - max - min) : d / (max + min);
        switch (max) {
            case r:
                h = ((g - b) / d) + (g < b ? 6 : 0);
                break;
            case g:
                h = ((b - r) / d) + 2;
                break;
            case b:
                h = ((r - g) / d) + 4;
                break;
            default:
        }
        h /= 6;
    }
    return { h: h * 360, s: s * 100, l: l * 100 };
}
export function ensureWcagCompliance(bgColor, fgColor) {
    const MIN_CONTRAST = 4.5;
    let currentRatio = getContrastRatio(bgColor, fgColor);
    if (currentRatio >= MIN_CONTRAST) {
        return fgColor;
    }
    const hslColor = rgbToHsl(fgColor);
    let currentLightness = hslColor.l;
    const bgLuminance = getLuminance(bgColor);
    const fgLuminance = getLuminance(fgColor);
    const adjustDirection = fgLuminance < bgLuminance ? -1 : 1;
    while (currentRatio < MIN_CONTRAST && currentLightness >= 0 && currentLightness <= 100) {
        currentLightness += adjustDirection * 1;
        if (currentLightness < 0 || currentLightness > 100) {
            break;
        }
        const newFgColor = hslToRgb(hslColor.h, hslColor.s, currentLightness);
        currentRatio = getContrastRatio(bgColor, newFgColor);
        if (currentRatio >= MIN_CONTRAST) {
            return newFgColor;
        }
    }
    const white = { r: 255, g: 255, b: 255 };
    const black = { r: 0, g: 0, b: 0 };
    return getContrastRatio(bgColor, white) >= getContrastRatio(bgColor, black) ? white : black;
}
