// lightning-base-components/src/lightning/formattedRichTextRenderer/formattedRichTextRenderer.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { api } from 'lwc';
import { setHooks } from 'lwc';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import sanitizeHTML from 'lightning/purifyLib';
import { richTextConfig } from './richTextConfig';
import { updateRawLinkInfo } from 'lightning/routingService';
import { normalizeBoolean, hasOnlyAllowedVideoIframes, } from 'lightning/utilsPrivate';
import { linkTextNodes } from 'lightning/utils';
import { linkTextNodesSSR } from './linkTextNodesSSR';
import { ensureWcagCompliance } from './colorShiftUtil';
import { features } from 'lightning/features';
try {
    setHooks({
        sanitizeHtmlContent(rawHTMLString) {
            return rawHTMLString;
        },
    });
}
catch (e) {
}
function sldsIsV2Enabled() {
    if (!import.meta.env.SSR) {
        return getComputedStyle(document.head).getPropertyValue('--_slds-is-v2-enabled');
    }
    return false;
}
export default class LightningFormattedRichTextRenderer extends LightningShadowBaseClass {
    static validationOptOut = ['class'];
    _themeKey = 'light';
    rendered = false;
    _value = '';
    _disableLinkify = false;
    connected = false;
    richText = null;
    rawTextIfSanitizerThrewError = null;
    serverRenderedContent = null;
    linkingRequired = false;
    @api
    get disableLinkify() {
        return this._disableLinkify;
    }
    set disableLinkify(val) {
        this._disableLinkify = normalizeBoolean(val);
        this.renderRichText();
    }
    @api
    get value() {
        return this._value;
    }
    set value(val) {
        this._value = val === undefined || val === null ? '' : String(val);
        this.renderRichText();
    }
    renderedCallback() {
        if (this.linkingRequired) {
            this.linkRichText();
        }
        if (!this.rendered) {
            this.rendered = true;
            if ((features.enableWcagRichTextLight || features.enableWcagRichTextDark) && sldsIsV2Enabled() && !import.meta.env.SSR) {
                const isDark = window.getComputedStyle(this.template.host)?.colorScheme.includes('dark');
                if (isDark) {
                    this._themeKey = 'dark';
                }
            }
            this.renderRichText();
        }
    }
    connectedCallback() {
        super.connectedCallback();
        this.classList.add('slds-rich-text-editor__output');
        this.connected = true;
        if (!import.meta.env.SSR) {
            this.serverRenderedContent = this.container?.innerHTML;
            if (this.serverRenderedContent) {
                this.richText = this.serverRenderedContent;
            }
        }
    }
    disconnectedCallback() {
        this.removeLinkClickListeners();
        this.connected = false;
    }
    handleClick(event) {
        const anchor = event.currentTarget;
        if (anchor === null) {
            return;
        }
        const target = anchor.target;
        const url = anchor.href;
        updateRawLinkInfo(this, { url, target }).then((linkInfo) => {
            anchor.href = linkInfo.url;
            linkInfo.dispatcher(event);
        });
    }
    renderRichText() {
        if (import.meta.env.SSR) {
            this.richText = this.disableLinkify
                ? this.value
                : linkTextNodesSSR(this.value);
        }
        if (this.rendered) {
            try {
                this.rawTextIfSanitizerThrewError = null;
                if (this.value) {
                    let computedRichTextConfig = richTextConfig;
                    if (hasOnlyAllowedVideoIframes(this.value)) {
                        computedRichTextConfig = {
                            ...richTextConfig,
                            ALLOWED_TAGS: richTextConfig.ALLOWED_TAGS.concat([
                                'iframe',
                            ]),
                            ALLOWED_ATTR: richTextConfig.ALLOWED_ATTR.concat([
                                'allowfullscreen',
                            ]),
                        };
                    }
                    this.richText = sanitizeHTML(this.value, computedRichTextConfig);
                    this.linkingRequired = true;
                    if (sldsIsV2Enabled()) {
                        this.adjustColors();
                    }
                }
                else {
                    this.richText = this.value;
                }
            }
            catch (e) {
                console.warn(e.message);
                this.richText = null;
                this.rawTextIfSanitizerThrewError = this.value;
            }
        }
    }
    linkRichText() {
        if (!this.disableLinkify) {
            linkTextNodes(this.container);
        }
        this.links.forEach((link) => {
            link.addEventListener('click', this.handleClick.bind(this));
        });
        this.linkingRequired = false;
    }
    removeLinkClickListeners() {
        this.links.forEach((link) => {
            link.removeEventListener('click', this.handleClick.bind(this));
        });
    }
    get links() {
        return this.container ? [...this.container.querySelectorAll('a')] : [];
    }
    get container() {
        return this.template.querySelector('span');
    }
    _backgroundColor = null;
    get backgroundColor() {
        if (import.meta.env.SSR) {
            this._backgroundColor = 'rgb(255, 255, 255)';
            return this._backgroundColor;
        }
        if (!import.meta.env.SSR) {
            if (this._backgroundColor === null) {
                this._backgroundColor = window.getComputedStyle(this.template.host).backgroundColor;
                if (this._backgroundColor === 'rgba(0, 0, 0, 0)') {
                    this.template.host.style.backgroundColor = 'inherit';
                    this._backgroundColor = window.getComputedStyle(this.template.host).backgroundColor;
                    this.template.host.style.removeProperty('background-color');
                }
            }
        }
        return this._backgroundColor;
    }
    get backgroundColorRgb() {
        const match = this.backgroundColor.match(/rgb\((\d+),\s*(\d+),\s*(\d+)\)/);
        if (!match) {
            if (this._themeKey === 'light') {
                return { r: 255, g: 255, b: 255 };
            }
            return { r: 0, g: 0, b: 0 };
        }
        return {
            r: parseInt(match[1], 10),
            g: parseInt(match[2], 10),
            b: parseInt(match[3], 10),
        };
    }
    adjustColors() {
        if (this._themeKey === 'light' && !features.enableWcagRichTextLight) {
            return;
        }
        if (this._themeKey === 'dark' && !features.enableWcagRichTextDark) {
            return;
        }
        const colorRgbRegex = /(?<=<[^>]*style="[^"]*)color:\s*rgb\((\d+),\s*(\d+),\s*(\d+)\)/g;
        this.richText = this.richText.replace(colorRgbRegex, (_, rs, gs, bs) => {
            const r = parseInt(rs, 10), g = parseInt(gs, 10), b = parseInt(bs, 10);
            const shifted = ensureWcagCompliance(this.backgroundColorRgb, { r, g, b });
            if ((shifted.r === 255 && shifted.g === 255 && shifted.b === 255)) {
                return '';
            }
            if ((shifted.r === 0 && shifted.g === 0 && shifted.b === 0)) {
                return '';
            }
            return `color:rgb(${shifted.r},${shifted.g},${shifted.b})`;
        });
    }
}
