// lightning-base-components/src/lightning/primitiveCellButton/primitiveCellButton.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { api } from 'lwc';
import { normalizeBoolean } from 'lightning/utilsPrivate';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
export default class PrivateCellButton extends LightningShadowBaseClass {
    static delegatesFocus = true;
    @api
    rowKeyValue;
    @api
    colKeyValue;
    @api
    variant;
    @api
    label;
    @api
    iconName;
    @api
    iconPosition;
    @api
    buttonName;
    @api
    buttonClass;
    @api
    buttonTitle;
    @api
    internalTabIndex;
    @api
    type;
    @api
    alternativeText;
    @api
    iconClass;
    initialRender = true;
    buttonHasFocus = false;
    _connected = false;
    _disabled = false;
    get computedTitle() {
        return this.buttonTitle || this.label;
    }
    @api
    get disabled() {
        return this._disabled;
    }
    set disabled(value) {
        const newValue = normalizeBoolean(value);
        const oldValue = this._disabled;
        if (oldValue === false &&
            newValue === true &&
            this.buttonHasFocus &&
            !import.meta.env.SSR) {
            this.fireCellFalseBlurred();
        }
        this._disabled = newValue;
    }
    get isButtonIconType() {
        return this.type === 'button-icon';
    }
    renderedCallback() {
        if (this.initialRender) {
            this.addListeners();
        }
        this.initialRender = false;
    }
    connectedCallback() {
        super.connectedCallback();
        this._connected = true;
    }
    disconnectedCallback() {
        this._connected = false;
    }
    get buttonCustomElement() {
        if (this._connected) {
            return this.isButtonIconType
                ? this.refs.buttonIcon
                : this.refs.button;
        }
        return undefined;
    }
    addListeners() {
        const buttonCustomElement = this.buttonCustomElement;
        buttonCustomElement.addEventListener('focus', this.handleButtonFocused.bind(this));
        buttonCustomElement.addEventListener('blur', this.handleButtonBlurred.bind(this));
    }
    @api
    focus() {
        if (this.buttonCustomElement) {
            this.buttonCustomElement.focus();
        }
    }
    @api
    click() {
        if (this.buttonCustomElement) {
            this.buttonCustomElement.click();
        }
    }
    handleButtonFocused(event) {
        if (event.target.localName.indexOf('button') > -1) {
            this.buttonHasFocus = true;
        }
    }
    handleButtonBlurred(event) {
        if (event.target.localName.indexOf('button') > -1) {
            this.buttonHasFocus = false;
        }
    }
    fireCellFalseBlurred() {
        const { rowKeyValue, colKeyValue } = this;
        this.dispatchEvent(new CustomEvent('privatecellfalseblurred', {
            composed: true,
            bubbles: true,
            cancelable: true,
            detail: {
                rowKeyValue,
                colKeyValue,
            },
        }));
    }
    handleButtonClick() {
        const { rowKeyValue, colKeyValue } = this;
        setTimeout(() => {
            this.dispatchEvent(new CustomEvent('privatecellbuttonclicked', {
                composed: true,
                bubbles: true,
                cancelable: true,
                detail: {
                    rowKeyValue,
                    colKeyValue,
                },
            }));
        }, 0);
    }
}
