// lightning-base-components/src/lightning/breadcrumbs/__examples__/withOnclick/withOnclick.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { LightningElement } from 'lwc';
export default class DemoBreadcrumbs extends LightningElement {
    handleNavigateToAccount(event) {
        event.preventDefault();
        const caseDiv = this.template.querySelector('.container .case');
        this.hide(caseDiv);
        const accountDiv = this.template.querySelector('.container .account');
        this.show(accountDiv);
    }
    handleNavigateToCase(event) {
        event.preventDefault();
        const accountDiv = this.template.querySelector('.container .account');
        this.hide(accountDiv);
        const caseDiv = this.template.querySelector('.container .case');
        this.show(caseDiv);
    }
    close(event) {
        const name = event.target.value;
        const elmToClose = this.template.querySelector(`.${name}`);
        this.hide(elmToClose);
    }
    show(elm) {
        elm.classList.remove('slds-hide');
        elm.classList.add('slds-show');
    }
    hide(elm) {
        elm.classList.add('slds-hide');
        elm.classList.remove('slds-show');
    }
}
