// lightning-base-components/src/lightning/inputUtils/normalize.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { normalizeString } from 'lightning/utilsPrivate';
export const VARIANT = {
    STANDARD: 'standard',
    LABEL_HIDDEN: 'label-hidden',
    LABEL_STACKED: 'label-stacked',
    LABEL_INLINE: 'label-inline',
};
export function normalizeVariant(value) {
    return normalizeString(value, {
        fallbackValue: VARIANT.STANDARD,
        validValues: [
            VARIANT.STANDARD,
            VARIANT.LABEL_HIDDEN,
            VARIANT.LABEL_STACKED,
            VARIANT.LABEL_INLINE,
        ],
    });
}
export function normalizeInput(value) {
    if (typeof value === 'number' || typeof value === 'string') {
        return String(value);
    }
    return '';
}
export const INPUTMODE_VALID_VALUES = [
    'decimal',
    'email',
    'none',
    'numeric',
    'search',
    'tel',
    'text',
    'url',
];
export function normalizeInputModeSetterValue(inputModeValue) {
    return normalizeString(inputModeValue, {
        fallbackValue: '',
        validValues: INPUTMODE_VALID_VALUES,
    });
}
export function normalizeInputModeGetterValue(overrideValue, inputType) {
    let inputmode;
    if (overrideValue !== undefined) {
        inputmode = overrideValue;
    }
    else if (inputType === 'number') {
        inputmode = 'decimal';
    }
    else if (inputType === 'email') {
        inputmode = 'email';
    }
    else {
        inputmode = '';
    }
    return inputmode;
}
