// lightning-base-components/src/lightning/inputUtils/validity.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import labelBadInput from '@salesforce/label/LightningErrorMessage.validityBadInput';
import labelBadNumericInput from '@salesforce/label/LightningErrorMessage.validityBadNumericInput';
import labelBadEmailInput from '@salesforce/label/LightningErrorMessage.validityBadEmailInput';
import labelPatternMismatch from '@salesforce/label/LightningErrorMessage.validityPatternMismatch';
import labelRangeOverflow from '@salesforce/label/LightningErrorMessage.validityRangeOverflow';
import labelRangeUnderflow from '@salesforce/label/LightningErrorMessage.validityRangeUnderflow';
import labelStepMismatch from '@salesforce/label/LightningErrorMessage.validityStepMismatch';
import labelTooLong from '@salesforce/label/LightningErrorMessage.validityTooLong';
import labelTooShort from '@salesforce/label/LightningErrorMessage.validityTooShort';
import labelTypeMismatch from '@salesforce/label/LightningErrorMessage.validityTypeMismatch';
import labelValueMissing from '@salesforce/label/LightningErrorMessage.validityValueMissing';
import labelSelectAtleastOneValue from '@salesforce/label/LightningErrorMessage.validitySelectAtleastOne';
import { assert, reflectAttribute } from 'lightning/utilsPrivate';
function getConstraintsSortedByPriority() {
    return [
        'customError',
        'badInput',
        'badNumericInput',
        'patternMismatch',
        'rangeOverflow',
        'rangeUnderflow',
        'stepMismatch',
        'tooLong',
        'tooShort',
        'badEmailInput',
        'typeMismatch',
        'valueMissing',
        'selectAtleastOneValue',
    ];
}
const defaultLabels = {
    badInput: labelBadInput,
    customError: labelBadInput,
    badNumericInput: labelBadNumericInput,
    badEmailInput: labelBadEmailInput,
    patternMismatch: labelPatternMismatch,
    rangeOverflow: labelRangeOverflow,
    rangeUnderflow: labelRangeUnderflow,
    stepMismatch: labelStepMismatch,
    tooLong: labelTooLong,
    tooShort: labelTooShort,
    typeMismatch: labelTypeMismatch,
    valueMissing: labelValueMissing,
    selectAtleastOneValue: labelSelectAtleastOneValue,
};
function resolveBestMatch(validity) {
    let validityState;
    if (validity && validity.valid === false) {
        validityState = 'badInput';
        getConstraintsSortedByPriority().some((stateName) => {
            if (validity[stateName] === true) {
                validityState = stateName;
                return true;
            }
            return false;
        });
    }
    return validityState;
}
function computeConstraint(valueProvider, constraint) {
    const provider = valueProvider[constraint];
    if (typeof provider === 'function') {
        return provider();
    }
    if (typeof provider === 'boolean') {
        return provider;
    }
    return false;
}
function newValidityState(constraintsProvider) {
    class ValidityState {
        get valueMissing() {
            return computeConstraint(constraintsProvider, 'valueMissing');
        }
        get typeMismatch() {
            return computeConstraint(constraintsProvider, 'typeMismatch');
        }
        get patternMismatch() {
            return computeConstraint(constraintsProvider, 'patternMismatch');
        }
        get tooLong() {
            return computeConstraint(constraintsProvider, 'tooLong');
        }
        get tooShort() {
            return computeConstraint(constraintsProvider, 'tooShort');
        }
        get rangeUnderflow() {
            return computeConstraint(constraintsProvider, 'rangeUnderflow');
        }
        get rangeOverflow() {
            return computeConstraint(constraintsProvider, 'rangeOverflow');
        }
        get stepMismatch() {
            return computeConstraint(constraintsProvider, 'stepMismatch');
        }
        get customError() {
            return computeConstraint(constraintsProvider, 'customError');
        }
        get badInput() {
            return computeConstraint(constraintsProvider, 'badInput');
        }
        get badNumericInput() {
            return computeConstraint(constraintsProvider, 'badNumericInput');
        }
        get badEmailInput() {
            return computeConstraint(constraintsProvider, 'badEmailInput');
        }
        get selectAtleastOneValue() {
            return computeConstraint(constraintsProvider, 'validitySelectAtleastOneValue');
        }
        get valid() {
            return !(this.valueMissing ||
                this.typeMismatch ||
                this.patternMismatch ||
                this.tooLong ||
                this.tooShort ||
                this.rangeUnderflow ||
                this.rangeOverflow ||
                this.stepMismatch ||
                this.customError ||
                this.badInput ||
                this.badNumericInput ||
                this.badEmailInput ||
                this.selectAtleastOneValue);
        }
    }
    return new ValidityState();
}
export function buildSyntheticValidity(constraintProvider) {
    return Object.freeze(newValidityState(constraintProvider));
}
export function getErrorMessage(validity, labelMap) {
    const key = resolveBestMatch(validity);
    if (key) {
        return labelMap[key] ? labelMap[key] : defaultLabels[key];
    }
    return '';
}
export class FieldConstraintApi {
    _inputComponentProvider;
    _constraintsProvider;
    _constraint;
    _customValidityMessage;
    _inputComponentElement;
    constructor(inputComponentProvider, constraintProviders) {
        assert(typeof inputComponentProvider === 'function');
        this._inputComponentProvider = inputComponentProvider;
        this._constraintsProvider = Object.assign({}, constraintProviders);
        if (!this._constraintsProvider.customError) {
            this._constraintsProvider.customError = () => typeof this._customValidityMessage === 'string' &&
                this._customValidityMessage !== '';
        }
    }
    get validity() {
        if (!this._constraint) {
            this._constraint = buildSyntheticValidity(this._constraintsProvider);
        }
        return this._constraint;
    }
    checkValidity() {
        const isValid = this.validity.valid;
        if (!isValid) {
            if (this.inputComponent) {
                this.inputComponent.dispatchEvent(new CustomEvent('invalid', { cancellable: true }));
            }
        }
        return isValid;
    }
    reportValidity(callback) {
        const valid = this.checkValidity();
        if (this.inputComponent) {
            this.inputComponent.classList.toggle('slds-has-error', !valid);
            reflectAttribute(this.inputComponent, 'invalid', !valid);
            if (callback) {
                callback(this.validationMessage);
            }
        }
        return valid;
    }
    setCustomValidity(message) {
        this._customValidityMessage = message;
    }
    get validationMessage() {
        return getErrorMessage(this.validity, {
            customError: this._customValidityMessage,
            badInput: this.inputComponent.messageWhenBadInput,
            badNumericInput: this.inputComponent.messageWhenBadInput,
            badEmailInput: this.inputComponent.messageWhenTypeMismatch,
            patternMismatch: this.inputComponent.messageWhenPatternMismatch,
            rangeOverflow: this.inputComponent.messageWhenRangeOverflow,
            rangeUnderflow: this.inputComponent.messageWhenRangeUnderflow,
            stepMismatch: this.inputComponent.messageWhenStepMismatch,
            tooShort: this.inputComponent.messageWhenTooShort,
            tooLong: this.inputComponent.messageWhenTooLong,
            typeMismatch: this.inputComponent.messageWhenTypeMismatch,
            valueMissing: this.inputComponent.messageWhenValueMissing,
            selectAtleastOneValue: this.inputComponent.messageWhenValueMissing,
        });
    }
    get inputComponent() {
        if (!this._inputComponentElement) {
            this._inputComponentElement = this._inputComponentProvider();
        }
        return this._inputComponentElement;
    }
}
export class FieldConstraintApiWithProxyInput {
    _inputComponent;
    _overrides;
    _proxyInput;
    _attributes;
    _attributeUpdater;
    _privateConstraintApi;
    constructor(inputComponent, overrides = {}, inputElementName = 'input') {
        if (inputElementName !== 'textarea' && inputElementName !== 'input') {
            throw new Error('FieldConstraintApiWithProxyInput is not supported for this element');
        }
        this._inputComponent = inputComponent;
        this._overrides = overrides;
        this._proxyInput = document.createElement(inputElementName);
    }
    setInputAttributes(attributes) {
        this._attributes = attributes;
        this._attributeUpdater = (attributeNames) => {
            if (!attributes) {
                return;
            }
            if (typeof attributeNames === 'string') {
                this._setAttribute(attributeNames, attributes[attributeNames]());
            }
            else {
                attributeNames.forEach((attributeName) => {
                    this._setAttribute(attributeName, attributes[attributeName]());
                });
            }
        };
        return this._attributeUpdater;
    }
    get validity() {
        return this._constraintApi.validity;
    }
    checkValidity() {
        return this._constraintApi.checkValidity();
    }
    reportValidity(callback) {
        return this._constraintApi.reportValidity(callback);
    }
    setCustomValidity(message) {
        this._constraintApi.setCustomValidity(message);
        this._proxyInput.setCustomValidity(message);
    }
    get validationMessage() {
        return this._constraintApi.validationMessage;
    }
    _setAttribute(attributeName, value) {
        if (value !== null && value !== undefined && value !== false) {
            if (attributeName === 'value') {
                if (this._proxyInput.type === 'file') {
                    return;
                }
                this._proxyInput.value = value;
            }
            else if (typeof value === 'number') {
                reflectAttribute(this._proxyInput, attributeName, value.toString());
            }
            else {
                reflectAttribute(this._proxyInput, attributeName, value);
            }
        }
        else {
            reflectAttribute(this._proxyInput, attributeName);
        }
    }
    get _constraintApi() {
        if (!this._privateConstraintApi) {
            this._updateAllAttributes();
            const computeConstraintWithProxyInput = (constraintName) => {
                const constraintOverride = this._overrides[constraintName];
                const isDisabledOrReadOnly = this._proxyInput.hasAttribute('disabled') ||
                    this._proxyInput.hasAttribute('readonly');
                if (typeof constraintOverride === 'function') {
                    return !isDisabledOrReadOnly && constraintOverride();
                }
                return (!isDisabledOrReadOnly &&
                    this._proxyInput.validity[constraintName]);
            };
            const constraintsProvider = getConstraintsSortedByPriority().reduce((provider, constraint) => {
                provider[constraint] = computeConstraintWithProxyInput.bind(this, constraint);
                return provider;
            }, {});
            this._privateConstraintApi = new FieldConstraintApi(this._inputComponent, constraintsProvider);
        }
        return this._privateConstraintApi;
    }
    _updateAllAttributes() {
        if (this._attributes) {
            Object.entries(this._attributes).forEach(([key, valueFunction]) => {
                this._setAttribute(key, valueFunction());
            });
        }
    }
}
