// lightning-base-components/src/lightning/multiColumnSortingModal/multiColumnSortingModal.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import LightningModal from 'lightning/modal';
import Toast from 'lightning/toast';
import { track, api } from 'lwc';
import modalTitle from '@salesforce/label/LightningMultiColumnSortingModal.modalTitle';
import firstRuleHeading from '@salesforce/label/LightningMultiColumnSortingModal.firstRuleHeading';
import otherRuleHeading from '@salesforce/label/LightningMultiColumnSortingModal.otherRuleHeading';
import addRule from '@salesforce/label/LightningMultiColumnSortingModal.addRule';
import ruleLimit from '@salesforce/label/LightningMultiColumnSortingModal.ruleLimit';
import ruleLimitReached from '@salesforce/label/LightningMultiColumnSortingModal.ruleLimitReached';
import clearButton from '@salesforce/label/LightningMultiColumnSortingModal.clearButton';
import cancelButton from '@salesforce/label/LightningMultiColumnSortingModal.cancelButton';
import applyButton from '@salesforce/label/LightningMultiColumnSortingModal.applyButton';
import duplicateValueValidation from '@salesforce/label/LightningMultiColumnSortingModal.duplicateValueValidation';
import columnMovedUp from '@salesforce/label/LightningMultiColumnSortingModal.columnMovedUp';
import columnMovedDown from '@salesforce/label/LightningMultiColumnSortingModal.columnMovedDown';
import multiColumnSortingToast from '@salesforce/label/LightningDatatable.multiColumnSortingToast';
import ascending from '@salesforce/label/LightningPrimitiveColumnSorter.ascending';
import descending from '@salesforce/label/LightningPrimitiveColumnSorter.descending';
import { features } from 'lightning/features';
import { formatLabel } from 'lightning/utils';
const i18n = {
    modalTitle,
    firstRuleHeading,
    otherRuleHeading,
    addRule,
    ruleLimit,
    ruleLimitReached,
    clearButton,
    cancelButton,
    applyButton,
    duplicateValueValidation,
    multiColumnSortingToast,
    ascending,
    descending,
    columnMovedUp,
    columnMovedDown,
};
const getEmptyRule = () => ({
    selectedColumn: undefined,
    sortOrder: 'asc',
});
const initialRules = [getEmptyRule()];
export default class MultiColumnSortingModal extends LightningModal {
    @api
    dynamicCmp;
    @api
    dynamicCmpEvents;
    @api
    dynamicCmpAttributes;
    @api
    dtInstance;
    size = 'small';
    @track
    rules = initialRules;
    validInputs = true;
    _modalBody;
    _scrollToBottom = false;
    assistiveText = '';
    ruleIndexToSetLast = undefined;
    get i18n() {
        return i18n;
    }
    connectedCallback() {
        const { sortedBy, sortedDirection } = this.dtInstance;
        this.clearRules();
        if (Array.isArray(sortedBy) || Array.isArray(sortedDirection)) {
            if (sortedBy &&
                sortedDirection &&
                sortedBy.length === sortedDirection.length) {
                const length = Math.min(sortedBy.length, 5);
                for (let i = 0; i < length; i += 1) {
                    this.rules[i] = {
                        selectedColumn: sortedBy[i],
                        sortOrder: sortedDirection[i],
                    };
                }
                this.setDuplicates(this.rules);
            }
        }
    }
    renderedCallback() {
        const { comboboxInputs, rules, dynamicCmp, dynamicCmpAttributes, dynamicCmpEvents, refs, } = this;
        if (dynamicCmp) {
            if (dynamicCmpEvents) {
                Object.entries(dynamicCmpEvents).forEach(([key, value]) => {
                    refs.modalBodyContent.addEventListener(key, value);
                });
            }
            if (dynamicCmpAttributes) {
                Object.entries(dynamicCmpAttributes).forEach(([key, value]) => {
                    refs.modalBodyContent[key] = value;
                });
            }
        }
        const duplicateEntryExists = rules.some((rule) => rule.isDupe);
        const lastRuleSelected = rules.length >= 1 &&
            rules[rules.length - 1].selectedColumn !== undefined;
        if (this.validationThroughApplyButton ||
            duplicateEntryExists ||
            lastRuleSelected) {
            comboboxInputs.forEach((combobox, index) => {
                combobox.setCustomValidity(rules[index].isDupe
                    ? this.i18n.duplicateValueValidation
                    : '');
            });
            this.reportValidity(comboboxInputs);
            this.validationThroughApplyButton = false;
        }
        this.handleScrolling();
    }
    get columnSorters() {
        return Array.from(this.template.querySelectorAll('lightning-primitive-column-sorter'));
    }
    get comboboxInputs() {
        return this.columnSorters.map((sorter) => sorter.combobox);
    }
    get modalBody() {
        if (!this._modalBody) {
            this._modalBody = this.template.querySelector('lightning-modal-body').shadowRoot.firstChild;
        }
        return this._modalBody;
    }
    get rulesWithData() {
        const { rules } = this;
        this.setDuplicates(rules);
        return rules.map((rule, index) => ({
            key: `sorter-${index}`,
            ...rule,
            heading: index === 0 ? i18n.firstRuleHeading : i18n.otherRuleHeading,
            unselectedInitialRules: rules.length <= 1 && rules[0].selectedColumn === undefined,
            disabledTrash: rules.length < 2,
        }));
    }
    get processedColumns() {
        const filteredColumns = this.dtInstance.columns.filter((column) => column.label && column.sortable);
        const processedColumns = filteredColumns.map((column) => {
            return {
                ...column,
                value: column.fieldName,
            };
        });
        return processedColumns;
    }
    get ruleLimitText() {
        if (this.rules.length >= 5) {
            return i18n.ruleLimitReached;
        }
        return i18n.ruleLimit;
    }
    get ruleLimitTextStyle() {
        return this.ruleLimitText === i18n.ruleLimitReached
            ? 'color:#8C4B02;'
            : '';
    }
    get isAddRuleButtonDisabled() {
        const unselectedRuleExists = this.rules.some((rule) => rule.selectedColumn === undefined);
        return (unselectedRuleExists || this.rules.length >= 5 || !this.validInputs);
    }
    get isApplyButtonDisabled() {
        const { sortedBy, sortedDirection } = this.dtInstance;
        const isSortable = Array.isArray(sortedBy) &&
            Array.isArray(sortedDirection) &&
            sortedBy.length === sortedDirection.length &&
            sortedBy.length === this.rules.length;
        const isSame = isSortable &&
            sortedBy.every((column, i) => this.rules[i]?.selectedColumn === column &&
                this.rules[i]?.sortOrder === sortedDirection[i]);
        const hasEnoughColumnsSelected = this.rules.filter((rule) => rule.selectedColumn).length >= 1;
        return !hasEnoughColumnsSelected || !this.validInputs || isSame;
    }
    get isClearButtonDisabled() {
        return this.rules[0].selectedColumn === undefined;
    }
    handleAddRule() {
        this._scrollToBottom = true;
        this.rules.push(getEmptyRule());
    }
    handleDeleteRule(event) {
        const ruleIndexToDelete = event.target.index;
        this.rules.splice(ruleIndexToDelete, 1);
        if (this.ruleIndexToSetLast !== undefined &&
            ruleIndexToDelete < this.ruleIndexToSetLast) {
            this.ruleIndexToSetLast--;
        }
        else if (this.ruleIndexToSetLast === ruleIndexToDelete) {
            this.ruleIndexToSetLast = undefined;
        }
        this.setDuplicates(this.rules);
    }
    handleSelectionChange(event) {
        const { index, selectedColumn, sortOrder } = event.detail;
        const existingRule = this.rules[index];
        if (existingRule) {
            const newColumnSelection = existingRule.selectedColumn !== selectedColumn;
            existingRule.selectedColumn = selectedColumn;
            existingRule.sortOrder = sortOrder;
            if (newColumnSelection) {
                this.ruleIndexToSetLast = index;
                this.setDuplicates(this.rules);
            }
        }
    }
    handleRuleMovement(event) {
        const { rules } = this;
        const dir = event.type;
        const ruleIndexToMove = event.target.index;
        const ruleIndexToSwitch = dir === 'moveup' ? ruleIndexToMove - 1 : ruleIndexToMove + 1;
        const tempRule = rules[ruleIndexToMove];
        rules[ruleIndexToMove] = rules[ruleIndexToSwitch];
        rules[ruleIndexToSwitch] = tempRule;
        const column = this.processedColumns.find((col) => col.fieldName === tempRule.selectedColumn);
        const columnLabel = column ? column.label : tempRule.selectedColumn;
        const colLabelFiltered = columnLabel ? columnLabel : 'undefined';
        if (features.enableA11yMultiColSortModal) {
            this.assistiveText = dir === 'moveup'
                ? formatLabel(i18n.columnMovedUp, colLabelFiltered)
                : formatLabel(i18n.columnMovedDown, colLabelFiltered);
        }
        else {
            const dirText = dir === 'moveup' ? 'moved up' : 'moved down';
            this.assistiveText = `Column ${columnLabel} ${dirText}`;
        }
    }
    applyRules() {
        const fieldNames = this.rules.map((rule) => rule.selectedColumn);
        const sortDirections = this.rules.map((rule) => rule.sortOrder);
        this.reportValidity(this.comboboxInputs);
        this.validationThroughApplyButton = true;
        if (this.validInputs) {
            const event = new CustomEvent('sort', {
                detail: {
                    fieldNames,
                    sortDirections,
                    isMultiColumnSort: true,
                },
            });
            this.dtInstance.dispatchEvent(event);
            const toastMessage = this.getToastMessage(fieldNames, sortDirections);
            this.close({
                event,
                success: this.sendToast(toastMessage, 'success'),
                failure: this.sendToast(toastMessage, 'failure'),
            });
        }
    }
    getToastMessage(fieldNames, sortDirections) {
        const capitalizeFirstLetter = (str) => str.charAt(0).toUpperCase() + str.slice(1);
        const message = fieldNames
            .map((sortBy, index) => `${capitalizeFirstLetter(sortBy)} (${sortDirections[index] === 'asc'
            ? i18n.ascending
            : i18n.descending})`)
            .join(', ');
        return formatLabel(i18n.multiColumnSortingToast, message);
    }
    sendToast(toastMessage, variant) {
        return () => {
            Toast.show({
                label: toastMessage,
                mode: 'dismissible',
                variant,
            }, this.dtInstance);
        };
    }
    clearRules() {
        this.rules = [getEmptyRule()];
        this.ruleIndexToSetLast = undefined;
    }
    cancelRules() {
        this.close();
    }
    handleScrolling() {
        const { comboboxInputs } = this;
        let modalBody;
        if (this._scrollToBottom) {
            modalBody = this.modalBody;
            modalBody.scrollTop = modalBody.scrollHeight;
            this._scrollToBottom = false;
        }
        for (let i = 0; i < comboboxInputs.length; i += 1) {
            const combobox = comboboxInputs[i];
            if (combobox.classList.contains('slds-has-error')) {
                const sorter = this.columnSorters[i];
                modalBody = modalBody || this.modalBody;
                modalBody.scrollTop = sorter.offsetTop - modalBody.offsetTop;
                break;
            }
        }
    }
    setDuplicates(rules) {
        const selectedColumns = new Set();
        const ruleIndexToSetLast = this.ruleIndexToSetLast;
        for (let i = 0; i < rules.length; i += 1) {
            if (ruleIndexToSetLast !== i) {
                const { selectedColumn } = rules[i];
                rules[i].isDupe =
                    selectedColumn && selectedColumns.has(selectedColumn);
                selectedColumns.add(selectedColumn);
            }
        }
        if (ruleIndexToSetLast !== undefined) {
            const { selectedColumn } = rules[ruleIndexToSetLast];
            rules[ruleIndexToSetLast].isDupe =
                selectedColumn && selectedColumns.has(selectedColumn);
            selectedColumns.add(selectedColumn);
        }
        return selectedColumns;
    }
    reportValidity(comboboxInputs) {
        comboboxInputs.forEach((combobox) => combobox && combobox.reportValidity());
        this.validInputs = comboboxInputs.every((combobox) => combobox && combobox.validity.valid);
    }
}
