// lightning-base-components/src/lightning/formattedLookup/formattedLookup.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { api, track } from 'lwc';
import { normalizeBoolean } from 'lightning/utilsPrivate';
import { getLinkInfo } from 'lightning/routingService';
import { PrivateLookupItemPickedEvent } from './events';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import { features } from 'lightning/features';
export const EVENTS = {
    PrivateLookupItemPickedEvent,
};
export default class LightningFormattedLookup extends LightningShadowBaseClass {
    @api
    displayValue;
    @api
    tabIndex;
    @api
    get recordId() {
        return this.state.recordId;
    }
    set recordId(value) {
        this.state.recordId = value;
        this.updateLinkData();
    }
    _objectApiName = null;
    @api
    get objectApiName() {
        return this._objectApiName;
    }
    set objectApiName(value) {
        if (features.enableLookupFieldObjApiName) {
            this._objectApiName = value;
            this.updateLinkData();
        }
    }
    @api
    get disabled() {
        return this.state.disabled;
    }
    set disabled(value) {
        this.state.disabled = normalizeBoolean(value);
        this.isNavigable =
            !this.disabled && !!this.dispatcher && !!this.state.url;
    }
    _connected;
    _dataNavigation;
    dispatcher;
    @track
    state = {
        disabled: false,
        recordId: null,
        url: null,
        isNavigable: false,
    };
    constructor() {
        super();
        this._connected = false;
        this.dispatcher = null;
    }
    connectedCallback() {
        super.connectedCallback();
        this._connected = true;
        this.updateLinkData();
        if (this.getAttribute('data-navigation') === 'enable') {
            this._dataNavigation = true;
            this.removeAttribute('data-navigation');
        }
    }
    disconnectedCallback() {
        this._connected = false;
    }
    @api
    focus() {
        if (this.anchor) {
            this.anchor.focus();
        }
    }
    @api
    blur() {
        if (this.anchor) {
            this.anchor.blur();
        }
    }
    @api
    click() {
        if (this.anchor) {
            this.anchor.click();
        }
    }
    get anchor() {
        if (!import.meta.env.SSR && this._connected && this.isNavigable) {
            return this.template.querySelector('a');
        }
        return undefined;
    }
    get isNavigable() {
        return this.state.isNavigable;
    }
    set isNavigable(value) {
        const normalizedValue = normalizeBoolean(value);
        this.state.isNavigable = normalizedValue;
        if (this._connected && this._dataNavigation) {
            if (normalizedValue) {
                this.setAttribute('data-navigation', 'enable');
            }
            else {
                this.removeAttribute('data-navigation');
            }
        }
    }
    updateLinkData() {
        const { recordId } = this.state;
        if (!import.meta.env.SSR && this._connected && recordId) {
            const { objectApiName } = this;
            getLinkInfo(this, {
                stateType: 'standard__recordPage',
                attributes: {
                    recordId,
                    actionName: 'view',
                    ...(features.enableLookupFieldObjApiName &&
                        objectApiName && { objectApiName }),
                },
            }).then((linkInfo) => {
                this.state.url = linkInfo.url;
                this.dispatcher = linkInfo.dispatcher;
                this.isNavigable =
                    !this.disabled && !!this.dispatcher && !!this.state.url;
            });
        }
    }
    handleClick(event) {
        this.dispatchEvent(new PrivateLookupItemPickedEvent({
            recordId: this.recordId,
        }));
        this.dispatcher(event);
    }
}
