// lightning-base-components/src/lightning/modalBody/modalBody.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import { classSet } from 'lightning/utils';
import { getRealDOMId } from 'lightning/utilsPrivate';
import { privateContext } from 'lightning/utilsInternal';
import { findAllTabbableElements, filterTooltips } from 'lightning/focusUtils';
import { modalFocusinEventHandler } from 'lightning/modalUtils';
const { setContext, assertContext } = privateContext;
export default class LightningModalBody extends LightningShadowBaseClass {
    initialRender = true;
    initialSlotRender = true;
    unregisterCallback = null;
    headerPresent = false;
    footerPresent = false;
    handleDefaultSlotChange() {
        if (this.initialSlotRender) {
            this.initialSlotRender = false;
        }
        this.registerWithParent();
    }
    handleScrollableContainerRepositionEvent(event) {
        const { callback: repositionCallback } = event.detail;
        repositionCallback(event.composedPath());
        event.stopPropagation();
    }
    get modalBodyCssClasses() {
        const classes = classSet('slds-modal__content slds-p-around_medium');
        classes.add({
            'slds-modal__content_headless': !this.headerPresent,
            'slds-modal__content_footless': !this.footerPresent,
        });
        return classes.toString();
    }
    get contentElem() {
        return this.template.querySelector('[data-content-container]');
    }
    get modalContentId() {
        return getRealDOMId(this.contentElem);
    }
    get defaultSlotElement() {
        return this.template.querySelector('[data-default-slot]');
    }
    get isDefaultSlotPopulated() {
        const slotElement = this.defaultSlotElement;
        if (slotElement && slotElement.assignedNodes) {
            return slotElement.assignedNodes().length > 0;
        }
        return false;
    }
    get firstTabbableElement() {
        let firstElem = null;
        if (this.isDefaultSlotPopulated) {
            const tabbableElements = findAllTabbableElements(this.defaultSlotElement);
            const filteredElements = filterTooltips(tabbableElements);
            if (filteredElements.length > 0) {
                firstElem = filteredElements[0];
            }
        }
        return firstElem;
    }
    calculateModalUsableHeight(values) {
        const { backdropHeight, paddingTop, paddingBottom } = values;
        const modalUsableHeight = backdropHeight - paddingTop - paddingBottom;
        values.modalUsableHeight = modalUsableHeight;
        return values;
    }
    calculateModalBodyUsableHeight({ modalUsableHeight, headerHeight, footerHeight, }) {
        return modalUsableHeight - headerHeight - footerHeight;
    }
    updateModalBodyHeightValues(values) {
        const updatedValues = this.calculateModalUsableHeight(values);
        const modalBodyUsableHeight = this.calculateModalBodyUsableHeight(updatedValues);
        values.modalBodyUsableHeight = modalBodyUsableHeight;
        values.maxModalBodyHeightStyle = `${modalBodyUsableHeight}px`;
        return values;
    }
    setModalBodyHeight(values, shouldUseFullHeight) {
        const styles = { maxHeight: null, minHeight: null };
        if (!shouldUseFullHeight) {
            const { maxModalBodyHeightStyle, minModalBodyHeightStyle } = this.updateModalBodyHeightValues(values);
            styles.maxHeight = maxModalBodyHeightStyle;
            styles.minHeight = minModalBodyHeightStyle;
        }
        const divElem = this.contentElem;
        Object.assign(divElem.style, styles);
    }
    updateValues(values) {
        const { headerHeight, footerHeight, backdropHeight } = values;
        const MODAL_BODY_MIN_HEIGHT = 80;
        const PAD_TOP_DEFAULT = 48;
        const PAD_BOTTOM_DEFAULT = 80;
        const updatedValues = {
            headerHeight: Math.floor(headerHeight),
            footerHeight: Math.floor(footerHeight),
            backdropHeight: Math.floor(backdropHeight),
            paddingTop: PAD_TOP_DEFAULT,
            paddingBottom: PAD_BOTTOM_DEFAULT,
            modalBodyMinHeight: MODAL_BODY_MIN_HEIGHT,
            minModalBodyHeightStyle: `${MODAL_BODY_MIN_HEIGHT}px`,
            maxModalBodyHeightStyle: null,
            modalBodyUsableHeight: null,
            modalUsableHeight: null,
        };
        return updatedValues;
    }
    handleUpdateModalBodyHeight(values) {
        if (!this.initialRender) {
            const { sizeSetFull, isSmallScreenSize, size, headerHeight, footerHeight, } = values;
            this.headerPresent = headerHeight !== 0;
            this.footerPresent = footerHeight !== 0;
            const shouldUseFullHeight = sizeSetFull && isSmallScreenSize;
            const updatedValues = this.updateValues(values);
            this.setModalBodyHeight(updatedValues, shouldUseFullHeight);
            if (!this.headerPresent) {
                this.setAttribute('data-variant-headless', '');
            }
            if (!this.footerPresent) {
                this.setAttribute('data-variant-footless', '');
            }
            this.setAttribute('data-size', size);
        }
    }
    registerWithParent() {
        assertContext(this);
        const evtRegister = new CustomEvent('privatemodalbodyregister', {
            bubbles: true,
            composed: true,
            detail: {
                bodyId: this.modalContentId,
                bodyIsPopulated: this.isDefaultSlotPopulated,
                defaultSlotHasRendered: !this.initialSlotRender,
                updateBodyCallback: this.handleUpdateModalBodyHeight.bind(this),
                unRegisterCallback: (unregisterCallback) => {
                    this.unregisterCallback = unregisterCallback;
                },
                firstTabbableElemRef: this.firstTabbableElement,
            },
        });
        this.dispatchEvent(evtRegister);
    }
    handleModalBodyFocusin = modalFocusinEventHandler.bind(this);
    initState() {
        this.initialRender = true;
        this.initialSlotRender = true;
        this.unregisterCallback = null;
        this.headerPresent = false;
        this.footerPresent = false;
    }
    constructor() {
        super();
        setContext(this);
    }
    connectedCallback() {
        super.connectedCallback();
        this.initState();
    }
    disconnectedCallback() {
        if (this.unregisterCallback) {
            this.unregisterCallback();
        }
    }
    renderedCallback() {
        if (this.initialRender) {
            this.registerWithParent();
            this.initialRender = false;
        }
    }
}
