// lightning-base-components/src/lightning/primitiveDatatableIeditTypeFactory/primitiveDatatableIeditTypeFactory.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { api } from 'lwc';
import { assert } from 'lightning/utilsPrivate';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import TextTpl from './text.html';
import PhoneTpl from './phone.html';
import EmailTpl from './email.html';
import PercentTpl from './percent.html';
import UrlTpl from './url.html';
import CurrencyTpl from './currency.html';
import NumberTpl from './number.html';
import BooleanTpl from './boolean.html';
import DateLocalTpl from './dateLocal.html';
import DateTpl from './date.html';
import DefaultTpl from './default.html';
const TYPE_TEMPLATE_MAPPINGS = {
    text: TextTpl,
    phone: PhoneTpl,
    email: EmailTpl,
    percent: PercentTpl,
    url: UrlTpl,
    currency: CurrencyTpl,
    number: NumberTpl,
    boolean: BooleanTpl,
    'date-local': DateLocalTpl,
    date: DateTpl,
};
const INVALID_TYPE_MESSAGE = 'column type not supported for inline edit';
export default class LightningPrimitiveDatatableIeditTypeFactory extends LightningShadowBaseClass {
    columnLabel;
    @api
    editedValue;
    @api
    required;
    @api
    typeAttributes;
    @api
    get columnDef() {
        return this._columnDef;
    }
    set columnDef(value) {
        assert(TYPE_TEMPLATE_MAPPINGS.hasOwnProperty(value.type) ||
            value.editableCustomType, INVALID_TYPE_MESSAGE);
        this._columnDef = value;
        this.columnLabel = value.label;
    }
    @api
    get customErrorValidationMessage() {
        return this._customErrorValidationMessage;
    }
    set customErrorValidationMessage(value) {
        this._customErrorValidationMessage = value;
        if (value && import.meta.env.SSR) {
            this.setAndReportInputValidity();
        }
    }
    setAndReportInputValidity() {
        if (this.concreteComponent) {
            this.concreteComponent.setCustomValidity(this.customErrorValidationMessage);
            this.concreteComponent.reportValidity();
        }
    }
    @api
    get isEditableCustomValid() {
        return this.isValidCustomType && this.concreteComponent;
    }
    @api
    get value() {
        if (this.columnDef.type === 'boolean') {
            return this.concreteComponent?.checked;
        }
        return this.concreteComponent?.value;
    }
    @api
    get validity() {
        return this.concreteComponent?.validity;
    }
    @api
    focus() {
        if (this.concreteComponent) {
            this.concreteComponent.focus();
        }
    }
    @api
    showHelpMessageIfInvalid() {
        this.concreteComponent.showHelpMessageIfInvalid();
    }
    get columnType() {
        return this._columnDef.type;
    }
    get concreteComponent() {
        return !import.meta.env.SSR
            ? this.template.querySelector('[data-inputable="true"]')
            : null;
    }
    get customEditTemplate() {
        return this._columnDef.editTemplate;
    }
    get editedDateValue() {
        const dateValue = new Date(this.editedValue);
        if (this.editedValue === null || isNaN(dateValue.getTime())) {
            return '';
        }
        return dateValue.toISOString();
    }
    get isValidCustomType() {
        return (this._columnDef.editableCustomType && this._columnDef.editTemplate);
    }
    get typeAttributeStep() {
        return ((this._columnDef &&
            this._columnDef.typeAttributes &&
            this._columnDef.typeAttributes.step) ||
            '0.01');
    }
    handleComponentFocus() {
        this.dispatchEvent(new CustomEvent('focus'));
    }
    handleComponentBlur() {
        this.dispatchEvent(new CustomEvent('blur'));
    }
    handleComponentChange() {
        this.showHelpMessageIfInvalid();
    }
    render() {
        return this.isValidCustomType
            ? this.customEditTemplate
            : TYPE_TEMPLATE_MAPPINGS[this.columnType] || DefaultTpl;
    }
    renderedCallback() {
        if (this.concreteComponent) {
            this.concreteComponent.addEventListener('focus', this.handleComponentFocus.bind(this));
            this.concreteComponent.addEventListener('blur', this.handleComponentBlur.bind(this));
            this.concreteComponent.addEventListener('change', this.handleComponentChange.bind(this));
            if (this.customErrorValidationMessage) {
                this.setAndReportInputValidity();
            }
        }
    }
}
