// lightning-base-components/src/lightning/primitiveDatatableStatusBar/primitiveDatatableStatusBar.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { api, track } from 'lwc';
import labelSave from '@salesforce/label/LightningDatatable.save';
import labelCancel from '@salesforce/label/LightningDatatable.cancel';
import labelError from '@salesforce/label/LightningDatatable.error';
import { features } from 'lightning/features';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
const i18n = {
    save: labelSave,
    cancel: labelCancel,
    error: labelError,
};
export default class LightningPrimitiveDatatableStatusBar extends LightningShadowBaseClass {
    @track
    privateError = {};
    connectedCallback() {
        super.connectedCallback();
        if (features.enableDTHeightCalcForScrollbar) {
            this.style.display = 'block';
            this.style.height = '3rem';
        }
    }
    @api
    get error() {
        return this.privateError;
    }
    set error(value) {
        this.privateError = value;
        if (this.showError && this.isSaveBtnFocused()) {
            this.handleShowError();
        }
    }
    get i18n() {
        return i18n;
    }
    get showError() {
        const { error } = this;
        if (error && (error.title || error.messages)) {
            return true;
        }
        return false;
    }
    get showErrorBubble() {
        const { error } = this;
        if (error && error.showBubble) {
            return true;
        }
        return false;
    }
    get bubbleOffset() {
        return {
            vertical: -10,
        };
    }
    isSaveBtnFocused() {
        return !import.meta.env.SSR
            ? this.template.querySelector('button.save-btn:focus') !== null
            : false;
    }
    handleCancelButtonClick(event) {
        event.preventDefault();
        event.stopPropagation();
        this.dispatchEvent(new CustomEvent('privatecancel', {
            bubbles: true,
            composed: true,
            cancelable: true,
        }));
    }
    handleSaveButtonClick(event) {
        event.preventDefault();
        event.stopPropagation();
        if (document.activeElement !== event.target) {
            event.target.focus();
        }
        this.dispatchEvent(new CustomEvent('privatesave', {
            bubbles: true,
            composed: true,
            cancelable: true,
        }));
    }
    handleShowError() {
        Promise.resolve().then(() => {
            const trigger = this.refs.tooltip;
            if (trigger) {
                if (this.showErrorBubble) {
                    trigger.showBubble();
                }
                else {
                    trigger.focus();
                }
            }
        });
    }
}
