// lightning-base-components/src/lightning/pillContainer/pillContainer.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import labelContainerLabel from '@salesforce/label/LightningPillContainer.label';
import pillContainerMoreLabel from '@salesforce/label/LightningPillContainer.more';
import { api } from 'lwc';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import LightningPillItem from './pillItem';
import { keyCodes, normalizeBoolean, normalizeString, reflectAttribute, } from 'lightning/utilsPrivate';
import { LightningResizeObserver } from 'lightning/resizeObserver';
import barePillContainer from './barePillContainer.html';
import standardPillContainer from './standardPillContainer.html';
import { relatedTarget as pillRelatedTarget } from 'lightning/pill';
import { privateContext } from 'lightning/utilsInternal';
const { setContext, assertContext } = privateContext;
const PILL_SELECTOR = 'lightning-pill';
const BARE = 'bare';
const STANDARD = 'standard';
const i18n = {
    containerLabel: labelContainerLabel,
};
export default class LightningPillContainer extends LightningShadowBaseClass {
    @api
    label = i18n.containerLabel;
    _variant;
    _singleLine = false;
    _isExpanded = false;
    _isCollapsible = false;
    _focusedIndex = 0;
    _focusedTabIndex = 0;
    _pillsNotFittingCount;
    _pillContainerElementId;
    constructor() {
        super();
        setContext(this);
    }
    connectedCallback() {
        super.connectedCallback();
        this._connected = true;
    }
    disconnectedCallback() {
        this._connected = false;
        if (this._resizeObserver) {
            this._resizeObserver.disconnect();
        }
    }
    render() {
        if (this.variant === BARE) {
            return barePillContainer;
        }
        return standardPillContainer;
    }
    renderedCallback() {
        if (this._resizeObserver) {
            if (!this.computedValueOfIsCollapsible) {
                this._resizeObserver.disconnect();
                this._resizeObserver = undefined;
            }
        }
        else if (this.computedValueOfIsCollapsible) {
            this._resizeObserver = this._setupResizeObserver();
        }
        const groupElm = this.template.querySelector('.slds-listbox_selection-group');
        if (groupElm) {
            groupElm.classList.toggle('slds-is-expanded', this.computedValueOfIsExpanded);
        }
        const ul = this.template.querySelector('ul');
        if (ul && this.pills.length) {
            this.setFocusedItemTabIndex(0);
            if (this.template.querySelector('ul:focus')) {
                this.focus();
            }
        }
    }
    get computedValueOfIsExpanded() {
        let expanded = this._isExpanded;
        if (!this._isCollapsible && !this._isExpanded) {
            expanded = true;
        }
        return expanded;
    }
    get computedValueOfIsCollapsible() {
        let collapsible = this._isCollapsible;
        if (this._isCollapsible && this._isExpanded) {
            collapsible = false;
        }
        return collapsible;
    }
    @api
    get variant() {
        return this._variant || STANDARD;
    }
    set variant(value) {
        this._variant = normalizeString(value, {
            fallbackValue: STANDARD,
            validValues: [STANDARD, BARE],
        });
        reflectAttribute(this, 'variant', this._variant);
    }
    @api
    get singleLine() {
        return this._singleLine;
    }
    set singleLine(value) {
        this._singleLine = normalizeBoolean(value);
        reflectAttribute(this, 'single-line', this._singleLine);
    }
    @api
    get isCollapsible() {
        return this._isCollapsible;
    }
    set isCollapsible(value) {
        this._isCollapsible = normalizeBoolean(value);
    }
    @api
    get isExpanded() {
        return this._isExpanded;
    }
    set isExpanded(value) {
        this._isExpanded = normalizeBoolean(value);
    }
    @api
    items;
    get pills() {
        return Array.isArray(this.items) ? this.items : [];
    }
    get pillViewModels() {
        return this.pills.map((item, index) => {
            const pill = new LightningPillItem(item);
            return {
                pill,
                tabIndex: this._focusedIndex === index ? this._focusedTabIndex : -1,
            };
        });
    }
    get computedWrapperClass() {
        return this.singleLine
            ? 'slds-pill_container'
            : 'slds-listbox_selection-group';
    }
    get computedListboxClass() {
        const singleLineClass = this.singleLine ? ' slds-listbox_inline' : '';
        return `slds-listbox slds-is-relative slds-listbox_horizontal slds-truncate${singleLineClass}`;
    }
    get focusedIndex() {
        if (this._focusedIndex >= this.pills.length) {
            this._focusedIndex = this._deleteLast ? this.pills.length - 1 : 0;
            this._deleteLast = false;
        }
        else if (this._focusedIndex < 0) {
            this._focusedIndex = this.pills.length - 1;
        }
        return this._focusedIndex;
    }
    set focusedIndex(value) {
        this._focusedIndex = value;
    }
    get pillNodes() {
        return this.template.querySelectorAll(PILL_SELECTOR) || [];
    }
    get focusedNode() {
        const pills = this.pillNodes;
        return pills.length <= 0 ? null : pills[this.focusedIndex];
    }
    @api
    focus() {
        const focusedNode = this.focusedNode;
        if (focusedNode) {
            if (focusedNode.isPlainLink) {
                focusedNode.focusLink();
            }
            else {
                focusedNode.focusRemove();
            }
        }
        else {
            const ul = this.template.querySelector('ul');
            if (ul) {
                ul.focus();
            }
        }
    }
    handleRemove(removeEvent) {
        const index = parseInt(removeEvent.detail.name, 10);
        if (typeof index !== 'number' || index < 0) {
            return;
        }
        this.fireEvent(index);
    }
    fireEvent(index) {
        assertContext(this);
        this.switchFocus(index);
        this._deleteLast = index >= this.pills.length - 1;
        this.dispatchEvent(new CustomEvent('itemremove', {
            detail: {
                item: this.pills[index],
                index,
            },
        }));
    }
    setFocusedItemTabIndex(value) {
        const focusedNode = this.focusedNode;
        if (focusedNode) {
            this._focusedTabIndex = value;
        }
    }
    switchFocusedIndex(newValue) {
        this.setFocusedItemTabIndex(-1);
        this.focusedIndex = newValue;
        this.setFocusedItemTabIndex(0);
    }
    switchFocus(newValue) {
        this.switchFocusedIndex(newValue);
        this.focus();
    }
    handleKeyDown(event) {
        if (this.pills.length <= 0) {
            return;
        }
        const index = this.focusedIndex;
        switch (event.keyCode) {
            case keyCodes.left:
            case keyCodes.up:
                this.switchFocus(index - 1);
                event.preventDefault();
                break;
            case keyCodes.right:
            case keyCodes.down:
                this.switchFocus(index + 1);
                event.preventDefault();
                break;
            default:
        }
    }
    handlePillFocus(event) {
        if (!this._hasFocus) {
            this._hasFocus = true;
            this.dispatchEvent(new CustomEvent('focus'));
        }
        const currentTargetIndex = parseInt(event.currentTarget?.name, 10);
        this.focusedIndex = Number.isNaN(currentTargetIndex)
            ? -1
            : currentTargetIndex;
    }
    handlePillBlur(event) {
        const relatedTarget = event.relatedTarget || event.detail[pillRelatedTarget];
        if (!relatedTarget || !this.template.contains(relatedTarget)) {
            this._hasFocus = false;
            this.dispatchEvent(new CustomEvent('blur'));
        }
    }
    handleClick() {
        this.focus();
    }
    handlePillClick(clickEvent) {
        const index = parseInt(clickEvent.currentTarget.name, 10);
        if (index >= 0 && this.focusedIndex !== index) {
            this.switchFocus(index);
        }
        else {
            this.focus();
        }
        clickEvent.stopPropagation();
    }
    handleMoreClick() {
        this.focus();
    }
    get _showMore() {
        return (this.computedValueOfIsCollapsible && !this.computedValueOfIsExpanded);
    }
    get computedPillCountMoreLabel() {
        if (this.computedValueOfIsExpanded ||
            isNaN(this._pillsNotFittingCount) ||
            this._pillsNotFittingCount <= 0) {
            return undefined;
        }
        return pillContainerMoreLabel.replace('{0}', this._pillsNotFittingCount);
    }
    _setupResizeObserver() {
        const pillContainer = this.template.querySelector('ul');
        if (!pillContainer) {
            return null;
        }
        const resizeObserver = new LightningResizeObserver(() => {
            let notFittingCount = 0;
            const pillNodes = this.template.querySelectorAll('li');
            const length = pillNodes.length;
            for (let i = 0; i < length; i++) {
                const node = pillNodes[i];
                if (node.offsetTop > 0) {
                    notFittingCount += 1;
                }
            }
            this._pillsNotFittingCount = notFittingCount;
        });
        resizeObserver.observe(pillContainer);
        return resizeObserver;
    }
}
