// lightning-base-components/src/lightning/primitiveDatatableCell/primitiveDatatableCell.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { api, track } from 'lwc';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import { keyCodes, isRTL } from 'lightning/utilsPrivate';
import { features } from 'lightning/features';
export default class PrimitiveDatatableCell extends LightningShadowBaseClass {
    @api
    rowKeyValue;
    @api
    colKeyValue;
    _hasFocus = 0;
    _mode = 'NAVIGATION';
    _currentInputIndex = 0;
    _internalTabIndex = -1;
    @track
    _actionableElementsCount;
    @api
    get hasFocus() {
        return this._hasFocus;
    }
    get keyboardMode() {
        return this._mode;
    }
    set hasFocus(value) {
        this._hasFocus = value;
        if (value) {
            this.classList.add('slds-has-focus');
        }
        else {
            this.classList.remove('slds-has-focus');
        }
    }
    @api
    setMode(keyboardMode, info) {
        const normalizedInfo = info || { action: 'none' };
        this._mode = keyboardMode;
        if (keyboardMode === 'ACTION') {
            this._internalTabIndex = 0;
            setTimeout(() => {
                this.setFocusToActionableElement(this._currentInputIndex);
            }, 0);
            const actionableElements = this.getActionableElements();
            if (!features.enableDTAutoInlineEditFix && normalizedInfo.action === 'tab') {
                let editActionElement = false;
                actionableElements.some((elem) => {
                    if (elem.getAttribute('data-action-edit')) {
                        editActionElement = elem;
                        return true;
                    }
                    return false;
                });
                if (editActionElement) {
                    editActionElement.click();
                }
            }
            else if (normalizedInfo.action !== 'tab' && actionableElements.length === 1) {
                const elem = actionableElements[0];
                let defaultActions = elem.getAttribute('data-action-triggers');
                defaultActions = defaultActions || '';
                if (defaultActions.indexOf(normalizedInfo.action) !== -1) {
                    actionableElements[this._currentInputIndex].click();
                }
            }
        }
        else {
            this._internalTabIndex = -1;
        }
    }
    @api
    addFocusStyles() {
        this.classList.add('slds-has-focus');
    }
    @api
    removeFocusStyles(setTabIndex) {
        this.classList.remove('slds-has-focus');
        if (setTabIndex) {
            this._internalTabIndex = -1;
        }
    }
    @api
    resetCurrentInputIndex(direction, incomingMode) {
        switch (direction) {
            case -1: {
                const inputs = this.getActionableElements();
                if (incomingMode === 'NAVIGATION') {
                    this._currentInputIndex = 0;
                }
                else {
                    this._currentInputIndex = inputs.length
                        ? inputs.length - 1
                        : 0;
                }
                break;
            }
            case 1:
            case 2:
                this._currentInputIndex = 0;
                break;
            default:
        }
        if (this._mode === 'ACTION' && incomingMode === 'ACTION') {
            this.setFocusToActionableElement(this._currentInputIndex);
        }
    }
    connectedCallback() {
        super.connectedCallback();
        this.addEventListener('focus', this.handleFocus.bind(this));
        this.addEventListener('click', this.handleClick.bind(this));
        this.addEventListener('keydown', this.handleKeydown.bind(this));
    }
    get internalTabIndex() {
        return this._internalTabIndex;
    }
    get canMoveBackward() {
        return this._currentInputIndex > 0;
    }
    get canMoveForward() {
        return this._actionableElementsCount > 1 && this._currentInputIndex < this._actionableElementsCount - 1;
    }
    get canMoveLeft() {
        return isRTL() ? this.canMoveForward : this.canMoveBackward;
    }
    get canMoveRight() {
        return isRTL() ? this.canMoveBackward : this.canMoveForward;
    }
    moveToNextActionableElement() {
        this.setFocusToActionableElement(this._currentInputIndex + 1);
    }
    moveToPrevActionableElement() {
        this.setFocusToActionableElement(this._currentInputIndex - 1);
    }
    handleClick(event) {
        this.addFocusStyles();
        this.fireCellFocusByClickEvent();
    }
    handleKeydown(event) {
        const { keyCode, shiftKey } = event;
        let passThroughEvent = keyCode !== keyCodes.shift;
        if (this._mode === 'ACTION') {
            switch (keyCode) {
                case keyCodes.left:
                    if (this.canMoveLeft) {
                        event.preventDefault();
                        if (isRTL()) {
                            this.moveToNextActionableElement();
                        }
                        else {
                            this.moveToPrevActionableElement();
                        }
                        passThroughEvent = false;
                    }
                    break;
                case keyCodes.right:
                    if (this.canMoveRight) {
                        event.preventDefault();
                        if (isRTL()) {
                            this.moveToPrevActionableElement();
                        }
                        else {
                            this.moveToNextActionableElement();
                        }
                        passThroughEvent = false;
                    }
                    break;
                case keyCodes.tab:
                    if (shiftKey) {
                        if (this.canMoveBackward) {
                            event.preventDefault();
                            this.moveToPrevActionableElement();
                            passThroughEvent = false;
                        }
                    }
                    else {
                        if (this.canMoveForward) {
                            event.preventDefault();
                            this.moveToNextActionableElement();
                            passThroughEvent = false;
                        }
                    }
                    break;
                default:
            }
        }
        else if (this._mode === 'NAVIGATION') {
            if (keyCode === keyCodes.left ||
                keyCode === keyCodes.right ||
                keyCode === keyCodes.up ||
                keyCode === keyCodes.down ||
                keyCode === keyCodes.enter) {
                this.fireCellKeydown(event);
            }
        }
        if (passThroughEvent && this._mode === 'ACTION') {
            this.fireCellKeydown(event);
        }
    }
    getActionableElements() {
        return Array.prototype.slice.call(this.template.querySelectorAll('[data-navigation="enable"]'));
    }
    get resizeElement() {
        return this.template.querySelector('.slds-resizable');
    }
    setFocusToActionableElement(activeInputIndex) {
        const actionableElements = this.getActionableElements();
        this._actionableElementsCount = actionableElements.length;
        if (actionableElements.length > 0) {
            if (activeInputIndex > 0 &&
                activeInputIndex < actionableElements.length) {
                actionableElements[activeInputIndex].focus();
                this._currentInputIndex = activeInputIndex;
            }
            else {
                actionableElements[0].focus();
                this._currentInputIndex = 0;
            }
        }
    }
    handleFocus() {
        if (this._mode === 'ACTION') {
            this.setFocusToActionableElement(this._currentInputIndex);
        }
    }
    fireCellKeydown(keyEvent) {
        const { rowKeyValue, colKeyValue } = this;
        const { keyCode, shiftKey } = keyEvent;
        const event = new CustomEvent('privatecellkeydown', {
            bubbles: true,
            composed: true,
            cancelable: true,
            detail: {
                rowKeyValue,
                colKeyValue,
                keyCode,
                shiftKey,
                keyEvent,
            },
        });
        this.dispatchEvent(event);
    }
    fireCellFocusByClickEvent() {
        const { rowKeyValue, colKeyValue } = this;
        const event = new CustomEvent('privatecellfocusedbyclick', {
            bubbles: true,
            composed: true,
            detail: {
                rowKeyValue,
                colKeyValue,
            },
        });
        this.dispatchEvent(event);
    }
}
