// lightning-base-components/src/lightning/accordionSection/accordionSection.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { api, track } from 'lwc';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import { generateUniqueId } from 'lightning/inputUtils';
import { keyCodes, isHeadingLevelValid, classSetToString, } from 'lightning/utilsPrivate';
import { privateContext } from 'lightning/utilsInternal';
import { classSet } from 'lightning/utils';
import DIR from '@salesforce/i18n/dir';
const { setContext, assertContext } = privateContext;
export default class LightningAccordionSection extends LightningShadowBaseClass {
    static validationOptOut = ['class'];
    @api
    name;
    @api
    label;
    @api
    title;
    @track
    privateIsOpen = false;
    _privateHeadingAriaLevel;
    _slotAssignedElement;
    @api
    get headingLevel() {
        return this._privateHeadingAriaLevel;
    }
    set headingLevel(value) {
        if (isHeadingLevelValid(value)) {
            this._privateHeadingAriaLevel = value;
        }
    }
    pendingFocus = false;
    privateUniqueId = generateUniqueId('lgt-accordion-section');
    constructor() {
        super();
        setContext(this);
    }
    connectedCallback() {
        super.connectedCallback();
        this.setAttribute('role', 'listitem');
        this.classList.add('slds-accordion__list-item');
        this.registerSectionWithParent();
    }
    disconnectedCallback() {
        if (this.privateAccordionSectionObserver) {
            this.privateAccordionSectionObserver.notifySectionDeregister();
        }
    }
    renderedCallback() {
        if (this.privateIsOpen && this.pendingFocus) {
            this.pendingFocus = false;
            this.focusSection();
        }
    }
    get computedAriaExpanded() {
        return this.privateIsOpen.toString();
    }
    get computedAriaHidden() {
        return (!this.privateIsOpen).toString();
    }
    get computedSectionClasses() {
        return classSet('slds-accordion__section')
            .add({
            'slds-is-open': this.privateIsOpen,
        })
            .toString();
    }
    get computedHidden() {
        return this.privateIsOpen ? '' : (!this.privateIsOpen).toString();
    }
    get computedHeaderClass() {
        return classSetToString({
            'slds-accordion__summary-heading': true,
            'slds-m-right_small': this.hostDirection === 'ltr' &&
                this._slotAssignedElement &&
                this._slotAssignedElement !== 'lightning-button-menu',
            'slds-m-left_small': this.hostDirection === 'rtl' && this._slotAssignedElement,
        });
    }
    handleKeyDown(event) {
        switch (event.keyCode) {
            case keyCodes.up:
            case keyCodes.right:
            case keyCodes.down:
            case keyCodes.left:
                event.preventDefault();
                event.stopPropagation();
                this.privateAccordionSectionObserver.notifySectionKeyNav(event.keyCode);
                break;
            default:
                break;
        }
    }
    handleSelectSection() {
        this.pendingFocus = true;
        this.privateAccordionSectionObserver.notifySectionSelect();
    }
    handleSlotChange(event) {
        const slot = event.target;
        const children = slot.assignedElements() || [];
        if (children.length) {
            this._slotAssignedElement = children[0].tagName.toLowerCase();
        }
    }
    get hostDirection() {
        if (import.meta.env.SSR) {
            return DIR || 'ltr';
        }
        const host = this.template.host;
        if (host.parentElement) {
            return (window
                .getComputedStyle(host.parentElement)
                .direction.toLowerCase() || 'ltr');
        }
        return DIR || 'ltr';
    }
    get accordionTitleIcon() {
        if (this.privateIsOpen) {
            return 'utility:switch';
        }
        return this.hostDirection === 'ltr'
            ? 'utility:chevronright'
            : 'utility:chevronleft';
    }
    registerSectionWithParent() {
        assertContext(this);
        const detail = {
            targetId: this.privateUniqueId,
            targetName: this.name,
            openSection: this.openSection.bind(this),
            isOpen: this.isOpen.bind(this),
            closeSection: this.closeSection.bind(this),
            focusSection: this.focusSection.bind(this),
            ackParentAccordion: this.ackParentAccordion.bind(this),
        };
        if (!import.meta.env.SSR) {
            this.dispatchEvent(new CustomEvent('privateaccordionsectionregister', {
                bubbles: true,
                composed: true,
                cancelable: true,
                detail,
            }));
        }
    }
    openSection() {
        this.privateIsOpen = true;
    }
    closeSection() {
        this.privateIsOpen = false;
    }
    focusSection() {
        const sectionButton = !import.meta.env.SSR
            ? this.template.querySelector('button.section-control')
            : null;
        if (sectionButton) {
            sectionButton.blur();
            sectionButton.focus();
        }
    }
    isOpen() {
        return this.privateIsOpen;
    }
    ackParentAccordion(accordionSectionObserver) {
        this.privateAccordionSectionObserver = accordionSectionObserver;
    }
}
