// lightning-base-components/src/lightning/modalHeader/modalHeader.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { api } from 'lwc';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import { getRealDOMId } from 'lightning/utilsPrivate';
import { modalFocusinEventHandler } from 'lightning/modalUtils';
import orgHasWCAGResizeReflowSC from '@salesforce/accessCheck/ComponentAccessibility.orgHasWCAGResizeReflowSC';
const modalHeaderSelector = '.slds-modal__header';
const labelSelector = '[data-label]';
const slotWrapperSelector = '[data-slot-wrapper]';
const defaultSlotSelector = '[data-default-slot]';
export default class LightningModalHeader extends LightningShadowBaseClass {
    initialRender = true;
    initialSlotRender = true;
    unregisterCallback = null;
    defaultSlotPopulated;
    @api
    label = '';
    handleDefaultSlotChange() {
        if (this.initialSlotRender) {
            this.initialSlotRender = false;
        }
        this.registerWithParent();
    }
    get headerHeight() {
        const divElem = this.template.querySelector(modalHeaderSelector);
        const headerRect = divElem ? divElem.getBoundingClientRect() : {};
        const { height } = headerRect;
        const heightValue = height || 0;
        return heightValue;
    }
    get modalTitleElement() {
        return this.template.querySelector(labelSelector);
    }
    get modalLabelId() {
        const labelElem = this.modalTitleElement;
        return getRealDOMId(labelElem);
    }
    get defaultSlotWrapperElement() {
        return this.template.querySelector(slotWrapperSelector);
    }
    get defaultSlotWrapperId() {
        return getRealDOMId(this.defaultSlotWrapperElement);
    }
    get defaultSlotElement() {
        return this.template.querySelector(defaultSlotSelector);
    }
    get isDefaultSlotPopulated() {
        const slotElement = this.defaultSlotElement;
        if (slotElement && slotElement.assignedNodes) {
            const slotPopulated = slotElement.assignedNodes().length > 0;
            this.defaultSlotPopulated = slotPopulated;
            return slotPopulated;
        }
        this.defaultSlotPopulated = false;
        return false;
    }
    get isLabelPopulated() {
        return (this.label && this.label.trim().length > 0) || false;
    }
    registerWithParent() {
        if (orgHasWCAGResizeReflowSC) {
            this.template
                .querySelector(modalHeaderSelector)
                .classList.add('slds-modal__header__reflow_resize');
        }
        const evtRegister = new CustomEvent('privatemodalheaderregister', {
            bubbles: true,
            composed: true,
            detail: {
                labelId: this.modalLabelId,
                labelIsPopulated: this.isLabelPopulated,
                headerHeight: this.headerHeight,
                defaultSlotWrapperId: this.defaultSlotWrapperId,
                defaultSlotIsPopulated: this.isDefaultSlotPopulated,
                defaultSlotHasRendered: !this.initialSlotRender,
                unRegisterCallback: (unregisterCallback) => {
                    this.unregisterCallback = unregisterCallback;
                },
                headerRef: this.modalTitleElement,
            },
        });
        this.dispatchEvent(evtRegister);
    }
    handleModalHeaderFocusin = modalFocusinEventHandler.bind(this);
    initState() {
        this.initialRender = true;
        this.initialSlotRender = true;
        this.unregisterCallback = null;
    }
    connectedCallback() {
        super.connectedCallback();
        this.initState();
    }
    renderedCallback() {
        if (this.initialRender) {
            this.registerWithParent();
            this.initialRender = false;
        }
    }
    disconnectedCallback() {
        if (this.unregisterCallback) {
            this.unregisterCallback();
        }
    }
}
