// lightning-base-components/src/lightning/progressStep/progressStep.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import labelCurrentStage from '@salesforce/label/LightningProgressIndicator.currentStage';
import labelStageComplete from '@salesforce/label/LightningProgressIndicator.stageComplete';
import labelNotStartedStage from '@salesforce/label/LightningProgressIndicator.stageNotStarted';
import labelErrorStage from '@salesforce/label/LightningProgressIndicator.errorStage';
import { api, track } from 'lwc';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import { classSet, formatLabel } from 'lightning/utils';
import { classListMutation, isRTL } from 'lightning/utilsPrivate';
import { privateContext } from 'lightning/utilsInternal';
import { Tooltip, TooltipType, Direction } from 'lightning/tooltipLibrary';
import path from './path.html';
import base from './base.html';
import progressStepStylesheets from './progressStep.css';
const { setContext, assertContext } = privateContext;
const i18n = {
    currentStage: labelCurrentStage,
    stageComplete: labelStageComplete,
    notStartedStage: labelNotStartedStage,
    errorStage: labelErrorStage,
};
const baseIconNameMap = {
    completed: 'utility:success',
    error: 'utility:error',
};
export default class LightningProgressStep extends LightningShadowBaseClass {
    static stylesheets = [progressStepStylesheets];
    static validationOptOut = ['class'];
    @api
    value;
    @track
    _privateLabel;
    _deRegistrationCallback;
    _tooltip;
    @track
    state = {};
    get i18n() {
        return i18n;
    }
    updateInternal(newStatus, newType, newIndex, newActive, shouldFocus) {
        classListMutation(this.classList, this.computeClassSet(newType, newStatus, newActive));
        if (newActive === true && shouldFocus) {
            this.focusPathLink();
        }
        this.state.status = newStatus;
        this.state.type = newType;
        this.state.index = newIndex;
        this.state.active = newActive;
        this.initializeTooltip();
        this.setAttribute('role', this.isPath ? 'presentation' : 'listitem');
    }
    @api
    set label(value) {
        this._privateLabel = value;
        this.initializeTooltip();
    }
    get label() {
        return this._privateLabel;
    }
    computeClassSet(type, status, isActive) {
        const isPath = type === 'path';
        return classSet({
            'slds-progress__item': !isPath,
            'slds-is-completed': !isPath && status === 'completed',
            'slds-has-error': !isPath && status === 'error',
            'slds-is-active': (isPath && isActive === true) ||
                (type === 'base' && status === 'current'),
            'slds-path__item': isPath,
            'slds-is-complete': isPath && status === 'completed',
            'slds-is-current': isPath && (status === 'error' || status === 'current'),
            'slds-is-incomplete': isPath && status === 'incomplete',
            'slds-is-rtl': isPath && isRTL(),
        });
    }
    constructor() {
        super();
        setContext(this);
    }
    connectedCallback() {
        super.connectedCallback();
        assertContext(this);
        if (!import.meta.env.SSR) {
            this.dispatchEvent(new CustomEvent('privateregisterstep', {
                bubbles: true,
                detail: {
                    callback: this.updateInternal.bind(this),
                    stepName: this.value,
                    setDeRegistrationCallback: (deRegistrationCallback) => {
                        this._deRegistrationCallback =
                            deRegistrationCallback;
                    },
                },
            }));
        }
    }
    disconnectedCallback() {
        if (this._deRegistrationCallback) {
            this._deRegistrationCallback();
        }
        if (this._tooltip) {
            this._tooltip.disconnect();
        }
    }
    renderedCallback() {
        this.computeClassSet();
        if (this._tooltip && !this._tooltip.initialized) {
            this._tooltip.initialize();
        }
    }
    get computedBaseStepClass() {
        const classes = classSet('slds-progress__marker');
        if (this.hasIcon) {
            classes
                .add('slds-progress__marker')
                .add('slds-progress__marker_icon')
                .add('slds-icon__container');
        }
        return classes.toString();
    }
    get hasIcon() {
        return (this.state.status === 'completed' || this.state.status === 'error');
    }
    get baseIconName() {
        return baseIconNameMap[this.state.status];
    }
    get isPath() {
        return this.state.type === 'path';
    }
    get ariaSelected() {
        return this.state.active === true ? 'true' : 'false';
    }
    get tabIndex() {
        return this.state.active === true ? 0 : -1;
    }
    get assistiveText() {
        if (this.state.status === 'completed') {
            return formatLabel(this.i18n.stageComplete, this.label);
        }
        else if (this.state.status === 'current') {
            return formatLabel(this.i18n.currentStage, this.label);
        }
        else if (this.state.status === 'incomplete') {
            return formatLabel(this.i18n.notStartedStage, this.label);
        }
        else if (this.state.status === 'error') {
            return formatLabel(this.i18n.errorStage, this.label);
        }
        return this.state.type === 'path' ? '' : this.label;
    }
    handleMouseEnter() {
        this.dispatchEvent(new CustomEvent('stepmouseenter', {
            bubbles: true,
            detail: { index: this.state.index },
        }));
    }
    handleMouseLeave() {
        this.dispatchEvent(new CustomEvent('stepmouseleave', {
            bubbles: true,
            detail: { index: this.state.index },
        }));
    }
    handleFocus() {
        this.dispatchEvent(new CustomEvent('stepfocus', {
            bubbles: true,
            detail: { index: this.state.index },
        }));
    }
    handleBlur() {
        this.dispatchEvent(new CustomEvent('stepblur', {
            bubbles: true,
            detail: { index: this.state.index },
        }));
    }
    focusPathLink() {
        const pathLink = this.template.querySelector('a.slds-path__link');
        if (pathLink) {
            pathLink.focus();
        }
    }
    render() {
        if (this.isPath) {
            return path;
        }
        return base;
    }
    initializeTooltip() {
        if (this._tooltip && this._tooltip instanceof Tooltip) {
            this._tooltip.value = this._privateLabel;
        }
        else if (this._privateLabel && this.state.type && !this.isPath) {
            if (!import.meta.env.SSR) {
                this._tooltip = new Tooltip(this._privateLabel, {
                    root: this,
                    target: () => this.template.querySelector('div'),
                    type: TooltipType.Toggle,
                    align: {
                        horizontal: Direction.Center,
                        vertical: Direction.Bottom,
                    },
                    targetAlign: {
                        horizontal: Direction.Center,
                        vertical: Direction.Top,
                    },
                    disableAriaDescribedBy: true,
                });
                this._tooltip.initialize();
            }
        }
    }
}
