// lightning-base-components/src/lightning/primitiveInputCheckboxButton/primitiveInputCheckboxButton.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { api } from 'lwc';
import { classSetToString, reflectAttribute, normalizeBoolean, isNativeComponent, } from 'lightning/utilsPrivate';
import { privateContext } from 'lightning/utilsInternal';
import { features } from 'lightning/features';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import standardTemplate from './primitiveInputCheckboxButton.html';
import oldTemplate from './primitiveInputCheckboxButtonOld.html';
const { setContext, assertContext } = privateContext;
export default class LightningPrimitiveInputColor extends LightningShadowBaseClass {
    @api
    ariaLabel;
    @api
    ariaInvalid;
    @api
    accessKey;
    @api
    name;
    @api
    required;
    @api
    readOnly;
    @api
    label;
    @api
    ariaKeyShortcuts;
    @api
    ariaDisabled;
    @api
    ariaRoleDescription;
    _checked = false;
    _disabled = false;
    _helpMessage = '';
    rendered = false;
    initialValueSet = false;
    constructor() {
        super();
        if (features.enableAssertContext260) {
            setContext(this);
        }
    }
    @api
    get checked() {
        if (this.initialValueSet) {
            return this.inputElement.checked;
        }
        return this._checked;
    }
    set checked(value) {
        this._checked = normalizeBoolean(value);
        if (this.rendered) {
            this.inputElement.checked = this._checked;
        }
        reflectAttribute(this, 'checked', this.checked);
    }
    @api
    get disabled() {
        return this._disabled;
    }
    set disabled(value) {
        this._disabled = normalizeBoolean(value);
        reflectAttribute(this, 'disabled', this.disabled);
    }
    @api
    get helpMessage() {
        return this._helpMessage;
    }
    set helpMessage(message) {
        this._helpMessage = message;
        reflectAttribute(this, 'invalid', !!message);
    }
    @api
    get inputElement() {
        if (!this.cachedInputElement) {
            let inputElement = this.template.querySelector('input');
            this.cachedInputElement = inputElement;
        }
        return this.cachedInputElement;
    }
    @api
    get ariaDescribedByElements() {
        return this.template.querySelector('[data-help-message]');
    }
    @api
    get ariaErrorMessageElement() {
        return this.template.querySelector('[data-help-message]');
    }
    @api
    get isNativeShadow() {
        return this._isNativeShadow;
    }
    get computedCheckboxButtonClass() {
        return classSetToString({
            'slds-checkbox-button': true,
            'slds-checkbox-button_is-checked': this.checked,
            'slds-checkbox-button_is-disabled': this.disabled,
        });
    }
    get iconName() {
        return this._checked ? 'utility:check' : 'utility:add';
    }
    connectedCallback() {
        super.connectedCallback();
        this._isNativeShadow = isNativeComponent(this);
    }
    disconnectedCallback() {
        this.cachedInputElement = undefined;
        this.initialValueSet = false;
        this.rendered = false;
    }
    renderedCallback() {
        this.rendered = true;
        if (!this.initialValueSet) {
            this.inputElement.checked = this._checked;
            this.initialValueSet = true;
        }
    }
    render() {
        return features.enableUpdatedCheckboxButtonBlueprint
            ? standardTemplate
            : oldTemplate;
    }
    handleBlur() {
        this.dispatchEvent(new CustomEvent('blur'));
    }
    handleFocus() {
        this.dispatchEvent(new CustomEvent('focus'));
    }
    handleChange(event) {
        event.stopPropagation();
        this.dispatchCommitEvent();
        this.dispatchChangeEvent();
    }
    handleClick() {
        if (this.template.activeElement === null) {
            this.template.querySelector("[type='checkbox']").focus();
        }
    }
    dispatchChangeEventWithDetail(detail) {
        this.dispatchEvent(new CustomEvent('change', {
            composed: true,
            detail,
        }));
    }
    dispatchChangeEvent() {
        const detail = {};
        this._checked = this.inputElement.checked;
        reflectAttribute(this, 'checked', this._checked);
        detail.checked = this._checked;
        this.dispatchChangeEventWithDetail(detail);
    }
    dispatchCommitEvent() {
        this.dispatchEvent(new CustomEvent('commit'));
    }
    setInputValue(value) {
        if (features.enableAssertContext260) {
            assertContext(this);
        }
        this.inputElement.value = value;
    }
    normalizeInput(value) {
        if (typeof value === 'number' || typeof value === 'string') {
            return String(value);
        }
        return '';
    }
}
