// lightning-base-components/src/lightning/overlayUtils/overlayUtils.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
export const parent = Symbol('parent');
export const instanceName = Symbol('instanceName');
export const secure = Symbol('secure event');
export const properties = Symbol('properties');
export const requiredProperties = Symbol('requiredProperties');
export function normalizeStyle(style) {
    const props = {};
    if (typeof style === 'string') {
        const regex = /([\w-]*)\s*:\s*([^;]*)/g;
        let match;
        while ((match = regex.exec(style))) {
            if (match[1].startsWith('--')) {
                props[match[1]] = match[2].trim();
            }
            else {
                console.warn(`Overlay 'style' only supports CSS Variables (invalid '${match[1]}' set)`);
            }
        }
    }
    else if (style && typeof style === 'object') {
        Object.keys(style).forEach((property) => {
            if (property.startsWith('--') &&
                typeof style[property] === 'string') {
                props[property] = style[property];
            }
            else {
                console.warn(`Overlay 'style' only supports CSS Variables (invalid '${property}' set)`);
            }
        });
    }
    return Object.keys(props)
        .map((property) => {
        return `${property}:${props[property]}`;
    })
        .join(';');
}
export function normalizeApis(apis) {
    if (apis && typeof apis === 'object') {
        const normalizedApis = { ...apis };
        if (apis.style) {
            normalizedApis.style = normalizeStyle(apis.style);
        }
        return normalizedApis;
    }
    return {};
}
