// lightning-base-components/src/lightning/primitiveDatatableIeditPanel/primitiveDatatableIeditPanel.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { api } from 'lwc';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import labelEdit from '@salesforce/label/LightningDatatable.edit';
import labelUpdateSelectedItems from '@salesforce/label/LightningDatatable.updateSelectedItems';
import labelCancel from '@salesforce/label/LightningDatatable.cancel';
import labelApply from '@salesforce/label/LightningDatatable.apply';
import { getStatCount } from 'lightning/overlayManager';
import { formatLabel } from 'lightning/utils';
import { features } from 'lightning/features';
const i18n = {
    edit: labelEdit,
    updateSelectedItems: labelUpdateSelectedItems,
    cancel: labelCancel,
    apply: labelApply,
};
export default class PrimitiveDatatableIeditPanel extends LightningShadowBaseClass {
    _focusTimeout = null;
    _forceRerenderHackKey = [''];
    _tableInModal = false;
    @api
    customErrorValidationMessage;
    @api
    visible;
    @api
    rowKeyValue;
    @api
    colKeyValue;
    @api
    editedValue;
    @api
    columnDef;
    @api
    isMassEditEnabled = false;
    @api
    numberOfSelectedRows;
    @api
    resolvedTypeAttributes;
    @api
    get isEditableValid() {
        if (!this.columnDef.editableCustomType ||
            this.inputableElement.isEditableCustomValid) {
            return true;
        }
        else if (this.columnDef.editableCustomType) {
            console.error('Editable custom types must define an editTemplate that includes an element with attribute data-inputable set to "true"');
        }
        return false;
    }
    @api
    get isMassEditChecked() {
        return this.isMassEditEnabled && this.refs.massSelectCheckbox.checked;
    }
    @api
    get value() {
        return this.inputableElement ? this.inputableElement.value : null;
    }
    @api
    get validity() {
        return this.inputableElement.validity;
    }
    @api
    focus() {
        const elem = this.inputableElement;
        clearTimeout(this._focusTimeout);
        if (elem) {
            elem.focus();
        }
    }
    @api
    getPositionedElement() {
        return this.refs.section;
    }
    get computedStyle() {
        const styleHash = {
            'z-index': 1000,
            'background-color': 'var(--slds-g-color-surface-1, #fff)',
            'margin-top': '1px',
        };
        styleHash.display = this.visible ? 'block' : 'none';
        return Object.keys(styleHash)
            .map((styleProp) => `${styleProp}:${styleHash[styleProp]}`)
            .join(';');
    }
    get inputableElement() {
        return !import.meta.env.SSR
            ? this.template.querySelector('.dt-type-edit-factory')
            : null;
    }
    get inputKey() {
        return this.rowKeyValue + this.colKeyValue;
    }
    get required() {
        return (this.columnDef.typeAttributes &&
            this.columnDef.typeAttributes.required);
    }
    get massEditCheckboxLabel() {
        return formatLabel(i18n.updateSelectedItems, this.numberOfSelectedRows);
    }
    get dialogAriaLabel() {
        const columnName = this.columnDef.label;
        return `${i18n.edit} ${columnName}`;
    }
    get i18n() {
        return i18n;
    }
    handleFormStartFocus() {
        clearTimeout(this._focusTimeout);
        if (this.isMassEditEnabled) {
            this.focusLastElement();
        }
        else {
            this.triggerEditFinished({
                reason: 'tab-pressed-prev',
            });
        }
    }
    handleFormEndsFocus() {
        clearTimeout(this._focusTimeout);
        if (this.isMassEditEnabled) {
            this.focus();
        }
        else {
            this.triggerEditFinished({
                reason: 'tab-pressed-next',
            });
        }
    }
    handlePanelLostFocus() {
        if (this.visible) {
            const isModalOpen = getStatCount('lwcModalCount') || getStatCount('auraModalCount');
            if (!this.template.activeElement &&
                (!isModalOpen || this._tableInModal)) {
                this.triggerEditFinished({
                    reason: 'lost-focus',
                });
            }
        }
    }
    handleTypeElemFocus() {
        clearTimeout(this._focusTimeout);
    }
    handleTypeElemBlur() {
        if (this.visible && !this.template.activeElement) {
            const blurRowKeyValue = this.rowKeyValue;
            const blurColKeyValue = this.colKeyValue;
            this._focusTimeout = setTimeout(() => {
                if (features.enableDtInlineEditBlurRace &&
                    (this.rowKeyValue !== blurRowKeyValue ||
                        this.colKeyValue !== blurColKeyValue)) {
                    return;
                }
                this.handlePanelLostFocus();
            }, 0);
        }
    }
    handleEditFormSubmit(event) {
        event.preventDefault();
        event.stopPropagation();
        if (!this.isMassEditEnabled) {
            this.processSubmission();
        }
        return false;
    }
    handleCellKeydown(event) {
        const { keyCode } = event;
        if (keyCode === 27) {
            event.preventDefault();
            event.stopPropagation();
            this.cancelEditing();
        }
    }
    handleMassCheckboxChange(event) {
        const customEvent = new CustomEvent('masscheckboxchange', {
            detail: {
                checked: event.detail.checked,
            },
        });
        this.dispatchEvent(customEvent);
    }
    triggerEditFinished(detail) {
        detail.rowKeyValue = detail.rowKeyValue || this.rowKeyValue;
        detail.colKeyValue = detail.colKeyValue || this.colKeyValue;
        const event = new CustomEvent('ieditfinished', {
            detail,
        });
        this.dispatchEvent(event);
    }
    connectedCallback() {
        super.connectedCallback();
        this._tableInModal =
            !import.meta.env.SSR &&
                !!this.template.host.closest('lightning-modal, .uiModal');
    }
    disconnectedCallback() {
        this._tableInModal = false;
    }
    focusLastElement() {
        this.refs.formLastElement.focus();
    }
    processSubmission() {
        if (this.validity.valid) {
            this.triggerEditFinished({ reason: 'submit-action' });
        }
        else {
            this.inputableElement.showHelpMessageIfInvalid();
        }
    }
    cancelEditing() {
        this.triggerEditFinished({
            reason: 'edit-canceled',
        });
    }
}
