// lightning-base-components/src/lightning/menuItem/__examples__/iteration/iteration.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { LightningElement } from 'lwc';
export default class ButtonMenuIteration extends LightningElement {
    myMenuItems = [
        {
            id: 'item-01',
            label: 'Alpha',
            value: 'alpha',
        },
        {
            id: 'item-02',
            label: 'Beta',
            value: 'beta',
        },
        {
            id: 'item-03',
            label: 'Gamma',
            value: 'gamma',
        },
        {
            id: 'item-04',
            label: 'Delta',
            value: 'delta',
        },
        {
            id: 'item-05',
            label: 'Epsilon',
            value: 'epsilon',
        },
    ];
    myComplexMenuItems = [
        {
            id: 'item-01',
            label: 'Alpha',
            value: 'alpha',
            disabled: false,
            prefixIconName: 'utility:bookmark',
            iconName: 'utility:animal_and_nature',
        },
        {
            id: 'item-02',
            label: 'Beta',
            value: 'beta',
            disabled: true,
            iconName: 'utility:company',
        },
        {
            id: 'item-03',
            label: 'Gamma',
            value: 'gamma',
            prefixIconName: 'utility:date_input',
        },
        {
            id: 'item-04',
            label: 'Delta',
            value: 'delta',
            checked: true,
        },
        {
            id: 'item-05',
            label: 'Epsilon',
            value: 'epsilon',
            iconName: 'utility:knowledge_base',
        },
    ];
}
