// lightning-base-components/src/lightning/primitiveHeaderFactory/primitiveHeaderFactory.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import labelChooseARow from '@salesforce/label/LightningDatatable.chooseARow';
import labelChooseARowSelectAll from '@salesforce/label/LightningDatatable.chooseARowSelectAll';
import labelSelectItems from '@salesforce/label/LightningDatatable.selectItems';
import labelSelectOneItem from '@salesforce/label/LightningDatatable.selectOneItem';
import labelSort from '@salesforce/label/LightningDatatable.sort';
import labelSortAsc from '@salesforce/label/LightningDatatable.sortAsc';
import labelSortDesc from '@salesforce/label/LightningDatatable.sortDesc';
import labelSortNone from '@salesforce/label/LightningDatatable.sortNone';
import sortingMenuMultiColumnSort from '@salesforce/label/LightningDatatable.sortingMenuMultiColumnSort';
import sortingMenuRows from '@salesforce/label/LightningDatatable.sortingMenuRows';
import sortingMenuColumns from '@salesforce/label/LightningDatatable.sortingMenuColumns';
import sortingMenuTooltip from '@salesforce/label/LightningDatatable.sortingMenuTooltip';
import PrimitiveDatatableCell from 'lightning/primitiveDatatableCell';
import { api, track } from 'lwc';
import { features } from 'lightning/features';
import { formatLabel } from 'lightning/utils';
import { classListMutation, classSetToString, reflectAttribute, isRTL, } from 'lightning/utilsPrivate';
import selectable from './selectableHeader.html';
import sortable from './sortableHeader.html';
import nonsortable from './nonsortableHeader.html';
import sortingMenu from './sortingMenuHeader.html';
const SELECTABLE_HEADER_TYPE = 'SELECTABLE_CHECKBOX';
const SORTING_MENU_HEADER_TYPE = 'SELECTABLE_BUTTON_MENU';
const i18n = {
    chooseARow: labelChooseARow,
    chooseARowSelectAll: labelChooseARowSelectAll,
    selectItems: labelSelectItems,
    selectOneItem: labelSelectOneItem,
    sort: labelSort,
    sortAsc: labelSortAsc,
    sortDesc: labelSortDesc,
    sortNone: labelSortNone,
    sortingMenuMultiColumnSort,
    sortingMenuRows,
    sortingMenuColumns,
    sortingMenuTooltip,
};
export default class PrimitiveHeaderFactory extends PrimitiveDatatableCell {
    @track
    _def = {};
    @track
    _resizable;
    _sortable = false;
    _hasSetId = false;
    _hideHeader = false;
    _imgSrc = '';
    _wrapTableHeader = false;
    _checked = false;
    @api
    actions;
    @api
    colIndex;
    @api
    dtContextId;
    @api
    headerHeight;
    @api
    headerId;
    @api
    resizestep;
    @api
    showCheckbox = false;
    @api
    sorted;
    @api
    sortedDirection;
    @api
    showActionsMenu;
    @api
    wrapTextMaxLines;
    @api
    get def() {
        return this._def;
    }
    set def(value) {
        this._def = value;
        this.updateElementClasses();
    }
    @api
    get hideHeader() {
        return this._hideHeader;
    }
    set hideHeader(value) {
        this._hideHeader = value;
        this.updateElementClasses();
    }
    @api
    get imgSrc() {
        return this._imgSrc;
    }
    set imgSrc(value) {
        this._imgSrc = value;
    }
    @api
    get wrapTableHeader() {
        return this._wrapTableHeader;
    }
    set wrapTableHeader(value) {
        this._wrapTableHeader = value;
        this.updateElementClasses();
    }
    @api
    get resizable() {
        return this._resizable;
    }
    set resizable(value) {
        this._resizable = value;
        this.updateElementClasses();
    }
    @api
    get sortable() {
        return this._sortable;
    }
    set sortable(value) {
        this._sortable = value;
        this.updateElementClasses();
    }
    @api
    getDomHeight() {
        const child = this.template.querySelector('.slds-hyphenate');
        if (child) {
            return child.offsetHeight;
        }
        return '';
    }
    @api
    getDomWidth() {
        const { child } = this.refs;
        if (child) {
            return child.offsetWidth;
        }
        return '';
    }
    get shouldWrapText() {
        const { _wrapTableHeader } = this;
        return (_wrapTableHeader === 'all' ||
            (_wrapTableHeader === 'by-column' && this._def.wrapText));
    }
    get columnStyles() {
        const outlineStyle = this._sortable ? '' : 'outline:none;';
        const offsetStyle = isRTL() ? `right: ${this._def.offset}px;` : '';
        const widthStyle = this._def.style;
        const heightStyle = this.headerHeight
            ? `height: max(2rem, ${this.headerHeight}px);`
            : '';
        return `${widthStyle}${outlineStyle}${offsetStyle}${heightStyle}`;
    }
    get computedActionClass() {
        return classSetToString({
            'slds-th__action': true,
            'slds-th__action_form': this._def.type !== SELECTABLE_HEADER_TYPE || this.showCheckbox,
            'slds-cell-fixed': !features.enableDatatableVoiceoverFix,
        });
    }
    get computedClass() {
        const { actions } = this;
        return classSetToString({
            'slds-cell-fixed': !features.enableDatatableVoiceoverFix,
            'slds-has-button-menu': actions.customerActions.length > 0 ||
                actions.internalActions.length > 0,
        });
    }
    get computedSortClass() {
        const { sortedDirection } = this;
        return classSetToString({
            'slds-th__action slds-text-link_reset': true,
            'slds-is-sorted': this.sorted,
            'slds-is-sorted_asc': sortedDirection === 'asc',
            'slds-is-sorted_desc': sortedDirection === 'desc',
        });
    }
    get getHeaderLabelClass() {
        return this.shouldWrapText
            ? 'slds-line-clamp slds-hyphenate'
            : 'slds-truncate';
    }
    get getHeaderLabelStyle() {
        return this.shouldWrapText
            ? `--lwc-lineClamp: ${this.wrapTextMaxLines || 'none'}`
            : '';
    }
    get thActionStyles() {
        return this.headerHeight
            ? `height: max(2rem, ${this.headerHeight}px);`
            : '';
    }
    get computedOptionName() {
        return `${this.dtContextId}-options`;
    }
    get hasActions() {
        const { actions } = this;
        return (actions.customerActions.length > 0 ||
            actions.internalActions.length > 0);
    }
    get hasIcon() {
        return this._imgSrc || this.def.iconName;
    }
    get computedColumnHeaderLabel() {
        return this.showCheckbox
            ? this.i18n.chooseARowSelectAll
            : this.i18n.chooseARow;
    }
    get computedSelectAllLabel() {
        const count = this._def?.maxSelectableRows;
        if (count === 1) {
            return this.i18n.selectOneItem;
        }
        else if (count > 1) {
            return formatLabel(this.i18n.selectItems, count);
        }
        return formatLabel(this.i18n.selectItems, count || 0);
    }
    get headerRole() {
        if (this._sortable ||
            (this._resizable && this._def.resizable !== false)) {
            return 'button';
        }
        return false;
    }
    get isRegularHeader() {
        return (this._def.type !== SELECTABLE_HEADER_TYPE &&
            this._def.type !== SORTING_MENU_HEADER_TYPE);
    }
    get isResizable() {
        return this._resizable && this._def.resizable !== false;
    }
    get i18n() {
        return i18n;
    }
    get resizeStep() {
        return this.resizestep;
    }
    get sortedOrderLabel() {
        if (this.sorted) {
            return this.sortedDirection === 'desc'
                ? i18n.sortDesc
                : i18n.sortAsc;
        }
        return i18n.sortNone;
    }
    render() {
        const { type } = this._def;
        if (type === SELECTABLE_HEADER_TYPE) {
            return selectable;
        }
        else if (type === SORTING_MENU_HEADER_TYPE ||
            (type === 'action' && this._def.isLastCol && this.showActionsMenu)) {
            return sortingMenu;
        }
        else if (this._sortable) {
            return sortable;
        }
        return nonsortable;
    }
    renderedCallback() {
        if (!this._hasSetId && this.refs.columnHeader) {
            this.refs.columnHeader.setAttribute('id', this.headerId);
            this._hasSetId = true;
        }
        if (this._def.type === SELECTABLE_HEADER_TYPE) {
            if (this.showCheckbox) {
                this.updateBulkSelectionCheckbox();
            }
        }
    }
    handleSortingClick(event) {
        event.preventDefault();
        if (this._sortable) {
            event.stopPropagation();
            this.fireSortedColumn();
            this.fireCellFocusByClickEvent();
        }
    }
    handleImageError(event) {
        console.warn(`<lightning-datatable> Column with imgSrc="${event.target.src}" failed to load image.`);
        this._imgSrc = '';
    }
    handleSelectAllRows() {
        const { _def, rowKeyValue, colKeyValue } = this;
        let shouldSelectAll;
        if (features.enableDisabledRowsAttribute) {
            shouldSelectAll =
                _def.bulkSelection === 'none' || !_def.enabledRowsSelected;
        }
        else {
            shouldSelectAll = _def.bulkSelection === 'none';
        }
        const actionName = shouldSelectAll
            ? 'selectallrows'
            : 'deselectallrows';
        this._checked = !this._checked;
        reflectAttribute(this, 'checked', this._checked);
        const actionEvent = new CustomEvent(actionName, {
            bubbles: true,
            composed: true,
            detail: {
                rowKeyValue,
                colKeyValue,
            },
        });
        this.dispatchEvent(actionEvent);
    }
    fireSortedColumn() {
        const { columnKey, fieldName } = this._def;
        const event = new CustomEvent('privateupdatecolsort', {
            bubbles: true,
            composed: true,
            detail: {
                columnKey,
                fieldName,
                sortDirection: this.getTargetSortDirection(),
            },
        });
        this.dispatchEvent(event);
    }
    handleSortingMenuSelection(event) {
        event.stopPropagation();
        const selectedItemValue = event.detail.value;
        if (selectedItemValue === 'MultiColumnSorting') {
            const e = new CustomEvent('privateopensortingmodal', {
                bubbles: true,
                composed: true,
            });
            this.dispatchEvent(e);
        }
    }
    updateElementClasses() {
        classListMutation(this.classList, {
            'slds-cell-fixed': features.enableDatatableVoiceoverFix,
            'slds-is-sortable': this._sortable,
            'slds-is-resizable': this._resizable && this._def.resizable !== false,
            'slds-assistive-text': this._hideHeader,
        });
    }
    getTargetSortDirection() {
        if (this.sorted) {
            return this.sortedDirection === 'desc' ? 'asc' : 'desc';
        }
        return this.sortedDirection;
    }
    updateBulkSelectionCheckbox() {
        const { bulkSelection } = this._def;
        const allCheckbox = this.refs.selectAll;
        allCheckbox.indeterminate = bulkSelection === 'some';
        const isChecked = bulkSelection !== 'none';
        this._checked = isChecked;
        allCheckbox.checked = isChecked;
        reflectAttribute(this, 'checked', this._checked);
    }
}
