// lightning-base-components/src/lightning/primitiveInputFile/primitiveInputFile.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { api } from 'lwc';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import { isNativeComponent, reflectAttribute } from 'lightning/utilsPrivate';
import labelRequired from '@salesforce/label/LightningControl.required';
import labelInputFileBodyText from '@salesforce/label/LightningInputFile.bodyText';
import labelInputFileButtonLabel from '@salesforce/label/LightningInputFile.buttonLabel';
const i18n = {
    required: labelRequired,
    inputFileBodyText: labelInputFileBodyText,
    inputFileButtonLabel: labelInputFileButtonLabel,
};
export default class LightningPrimitiveInputFile extends LightningShadowBaseClass {
    _files = null;
    _helpMessage = '';
    @api
    accept;
    @api
    accessKey;
    @api
    ariaInvalid;
    @api
    disabled;
    @api
    label;
    @api
    labelClass;
    @api
    multiple;
    @api
    name;
    @api
    readOnly;
    @api
    required;
    @api
    ariaLabel;
    @api
    ariaKeyShortcuts;
    @api
    ariaDisabled;
    @api
    ariaRoleDescription;
    @api
    get files() {
        return this._files;
    }
    set files(value) {
        if (value && !import.meta.env.SSR) {
            if (value instanceof FileList) {
                this._files = value;
            }
            else {
                const dataTransfer = new DataTransfer();
                if (Array.isArray(value)) {
                    let logError = false;
                    for (const file of value) {
                        if (file instanceof File) {
                            dataTransfer.items.add(file);
                        }
                        else {
                            logError = true;
                        }
                    }
                    if (logError) {
                        console.error('lightning-input type=file only supports FileList, File, or an Array of File objects in the files property');
                    }
                }
                else if (value instanceof File) {
                    dataTransfer.items.add(value);
                }
                else {
                    console.error('lightning-input type=file only supports FileList, File, or an Array of File objects in the files property');
                }
                this._files = dataTransfer.files;
                requestAnimationFrame(() => {
                    if (this.inputElement) {
                        this.inputElement.files = dataTransfer.files;
                    }
                });
            }
        }
    }
    @api
    get inputElement() {
        return this.template.querySelector('input');
    }
    @api
    get ariaDescribedByElements() {
        return this.template.querySelector('[data-help-message]');
    }
    @api
    get ariaErrorMessageElement() {
        return this.template.querySelector('[data-help-message]');
    }
    @api
    get ariaLabelledByElements() {
        return [
            this.template.querySelector('[data-form-label]'),
            this.template.querySelector('[data-file-selector-label]'),
        ];
    }
    @api
    get isNativeShadow() {
        return this._isNativeShadow;
    }
    @api
    get helpMessage() {
        return this._helpMessage;
    }
    set helpMessage(message) {
        this._helpMessage = message;
        reflectAttribute(this, 'invalid', !!message);
    }
    @api
    get variant() {
        return this._variant;
    }
    set variant(variant) {
        this._variant = variant;
        reflectAttribute(this, 'variant', variant);
    }
    get i18n() {
        return i18n;
    }
    handleDropFiles(event) {
        this._files = event.dataTransfer && event.dataTransfer.files;
        this.dispatchEvent(new CustomEvent('change'));
        this.dispatchEvent(new CustomEvent('reportvalidity'));
    }
    handleBlur() {
        this.dispatchEvent(new CustomEvent('blur'));
    }
    handleFocus() {
        this.dispatchEvent(new CustomEvent('focus'));
    }
    handleFileClick() {
        this.inputElement.value = null;
    }
    handleChange() {
        this._files = this.inputElement.files;
        this.dispatchEvent(new CustomEvent('change'));
        this.dispatchEvent(new CustomEvent('reportvalidity'));
    }
    connectedCallback() {
        super.connectedCallback();
        this._isNativeShadow = isNativeComponent(this);
    }
}
