// lightning-base-components/src/lightning/badge/badge.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { api } from 'lwc';
import { classSet } from 'lightning/utils';
import { normalizeString } from 'lightning/utilsPrivate';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
const DEFAULT_POSITION = 'start';
export default class LightningBadge extends LightningShadowBaseClass {
    static validationOptOut = ['class'];
    _iconPosition = DEFAULT_POSITION;
    @api
    label;
    @api
    iconName;
    @api
    iconAlternativeText;
    @api
    get iconPosition() {
        return this._iconPosition;
    }
    set iconPosition(value) {
        this._iconPosition = normalizeString(value, {
            fallbackValue: DEFAULT_POSITION,
            validValues: ['start', 'end'],
        });
    }
    connectedCallback() {
        super.connectedCallback();
        this.classList.add('slds-badge');
    }
    get computedClass() {
        const iconClass = classSet('slds-badge__icon');
        iconClass.add(this.isIconBeforeLabel
            ? 'slds-badge__icon_left'
            : 'slds-badge__icon_right');
        return iconClass.toString();
    }
    get isIconBeforeLabel() {
        return this._iconPosition !== 'end';
    }
}
