// lightning-base-components/src/lightning/mediaUtils/mediaUtils.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
export function processImage(input, options = null) {
    return readInputData(input)
        .then((dataURL) => loadInputDataIntoImage(dataURL))
        .then((image) => resizeAndCompressImage(image, input.type, options));
}
function readInputData(input) {
    return new Promise((resolve, reject) => {
        const reader = new FileReader();
        reader.onloadend = (event) => {
            resolve(event.target.result);
        };
        reader.onerror = () => {
            reject(new Error('Unable to read the input data.'));
        };
        try {
            reader.readAsDataURL(input);
        }
        catch (err) {
            reject(new Error('Unable to read the input data.'));
        }
    });
}
function loadInputDataIntoImage(dataUrl) {
    return new Promise((resolve, reject) => {
        const image = new Image();
        image.onload = function () {
            resolve(image);
        };
        image.onerror = function () {
            reject(new Error('Unable to load the input as an image.'));
        };
        image.src = dataUrl;
    });
}
function getResizeMode(defaultValue, options) {
    let resizeMode = defaultValue;
    if (options && options.resizeMode) {
        resizeMode = `${options.resizeMode}`.toLowerCase().trim();
    }
    if (resizeMode === 'fill' ||
        resizeMode === 'contain' ||
        resizeMode === 'none') {
        return resizeMode;
    }
    throw new Error('Invalid parameter value for resizeMode.');
}
function getResizeStrategy(defaultValue, options) {
    let resizeStrategy = defaultValue;
    if (options && options.resizeStrategy) {
        resizeStrategy = `${options.resizeStrategy}`.toLowerCase().trim();
    }
    if (resizeStrategy === 'reduce' ||
        resizeStrategy === 'enlarge' ||
        resizeStrategy === 'always') {
        return resizeStrategy;
    }
    throw new Error('Invalid parameter value for resizeStrategy.');
}
function getTargetWidth(defaultValue, options) {
    let targetWidth = defaultValue;
    if (options && options.targetWidth) {
        targetWidth = parseFloat(options.targetWidth);
        if (isNaN(targetWidth)) {
            throw new Error('Invalid parameter value for targetWidth.');
        }
    }
    return Math.round(targetWidth);
}
function getTargetHeight(defaultValue, options) {
    let targetHeight = defaultValue;
    if (options && options.targetHeight) {
        targetHeight = parseFloat(options.targetHeight);
        if (isNaN(targetHeight)) {
            throw new Error('Invalid parameter value for targetHeight.');
        }
    }
    return Math.round(targetHeight);
}
function getCompressionQuality(defaultValue, options) {
    let compressionQuality = defaultValue;
    if (options && options.compressionQuality) {
        compressionQuality = parseFloat(options.compressionQuality);
        if (isNaN(compressionQuality)) {
            throw new Error('Invalid parameter value for compressionQuality.');
        }
    }
    return compressionQuality;
}
function getPreserveTransparency(defaultValue, options) {
    if (options &&
        options.preserveTransparency !== undefined &&
        options.preserveTransparency !== null) {
        const strVal = `${options.preserveTransparency}`.toLowerCase().trim();
        if (strVal === `true`) {
            return true;
        }
        else if (strVal === `false`) {
            return false;
        }
        throw new Error('Invalid parameter value for preserveTransparency.');
    }
    return defaultValue;
}
function getImageSmoothingEnabled(defaultValue, options) {
    if (options &&
        options.imageSmoothingEnabled !== undefined &&
        options.imageSmoothingEnabled !== null) {
        const strVal = `${options.imageSmoothingEnabled}`.toLowerCase().trim();
        if (strVal === `true`) {
            return true;
        }
        else if (strVal === `false`) {
            return false;
        }
        throw new Error('Invalid parameter value for imageSmoothingEnabled.');
    }
    return defaultValue;
}
function getBackgroundColor(defaultValue, options) {
    if (options && options.backgroundColor) {
        const strVal = `${options.backgroundColor}`;
        if (CSS.supports('color', strVal)) {
            return strVal;
        }
        throw new Error('Invalid parameter value for backgroundColor.');
    }
    return defaultValue;
}
function computeFinalDimensions(originalWidth, originalHeight, targetWidth, targetHeight, resizeMode, resizeStrategy) {
    let finalWidth = originalWidth;
    let finalHeight = originalHeight;
    if (resizeMode !== 'none' &&
        (resizeStrategy === 'always' ||
            (resizeStrategy === 'reduce' &&
                (originalWidth > targetWidth ||
                    originalHeight > targetHeight)) ||
            (resizeStrategy === 'enlarge' &&
                (originalWidth < targetWidth || originalHeight < targetHeight)))) {
        if (resizeMode === 'fill') {
            finalWidth = targetWidth;
            finalHeight = targetHeight;
        }
        else if (resizeMode === 'contain') {
            const ratio = Math.min(targetWidth / originalWidth, targetHeight / originalHeight);
            finalWidth = Math.round(originalWidth * ratio);
            finalHeight = Math.round(originalHeight * ratio);
        }
    }
    return { finalWidth, finalHeight };
}
function resizeAndCompressImage(image, mimeType, options) {
    return new Promise((resolve, reject) => {
        let resizeMode;
        let resizeStrategy;
        let targetWidth;
        let targetHeight;
        let compressionQuality;
        let imageSmoothingEnabled;
        let preserveTransparency;
        let backgroundColor;
        try {
            resizeMode = getResizeMode('fill', options);
            resizeStrategy = getResizeStrategy('always', options);
            targetWidth = getTargetWidth(image.width, options);
            targetHeight = getTargetHeight(image.height, options);
            compressionQuality = getCompressionQuality(null, options);
            imageSmoothingEnabled = getImageSmoothingEnabled(true, options);
            preserveTransparency = getPreserveTransparency(true, options);
            backgroundColor = getBackgroundColor('white', options);
        }
        catch (error) {
            reject(error);
            return;
        }
        const { finalWidth, finalHeight } = computeFinalDimensions(image.width, image.height, targetWidth, targetHeight, resizeMode, resizeStrategy);
        const canvas = document.createElement('canvas');
        canvas.width = finalWidth;
        canvas.height = finalHeight;
        canvas.imageSmoothingEnabled = imageSmoothingEnabled;
        const ctx = canvas.getContext('2d');
        if (!preserveTransparency) {
            ctx.fillStyle = backgroundColor;
            ctx.fillRect(0, 0, finalWidth, finalHeight);
        }
        ctx.drawImage(image, 0, 0, finalWidth, finalHeight);
        let targetType = 'image/jpeg';
        if (preserveTransparency &&
            (mimeType === 'image/gif' || mimeType === 'image/png')) {
            targetType = 'image/png';
        }
        const resultFunc = function (blob) {
            resolve(blob);
        };
        if (compressionQuality) {
            canvas.toBlob(resultFunc, targetType, compressionQuality);
        }
        else {
            canvas.toBlob(resultFunc, targetType);
        }
    });
}
