// lightning-base-components/src/lightning/baseFormattedText/baseFormattedText.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { api, track } from 'lwc';
import { isTextIgnoreRTL, normalizeBoolean, parseToFormattedLinkifiedParts, } from 'lightning/utilsPrivate';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
export default class BaseFormattedText extends LightningShadowBaseClass {
    @api
    value = '';
    @track
    _linkify = false;
    @api
    get linkify() {
        return this._linkify;
    }
    set linkify(value) {
        this._linkify = normalizeBoolean(value);
    }
    get shouldFormat() {
        return this.linkify && this.isString;
    }
    get isString() {
        return this.value && typeof this.value === 'string';
    }
    get isObject() {
        return typeof this.value === 'object';
    }
    get normalizedValue() {
        if (this.isObject) {
            return '';
        }
        return this.value;
    }
    get formattedParts() {
        if (!this.isString) {
            return [];
        }
        return parseToFormattedLinkifiedParts(this.value, true);
    }
    get ignoreRTL() {
        return isTextIgnoreRTL(this.value);
    }
}
