// lightning-base-components/src/lightning/primitiveTreegridCellToggle/primitiveTreegridCellToggle.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import labelCollapseBranch from '@salesforce/label/LightningPrimitiveCellTree.collapseBranch';
import labelExpandBranch from '@salesforce/label/LightningPrimitiveCellTree.expandBranch';
import { api, track } from 'lwc';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import { classSet, formatLabel } from 'lightning/utils';
import { normalizeString, normalizeBoolean, isRTL } from 'lightning/utilsPrivate';
import { features } from 'lightning/features';
const i18n = {
    collapseBranch: labelCollapseBranch,
    expandBranch: labelExpandBranch,
};
const DEFAULT_SIZE = 'small';
const DEFAULT_VARIANT = 'bare';
export default class PrivateTreeGridCellToggle extends LightningShadowBaseClass {
    @api
    rowKeyValue;
    @api
    colKeyValue;
    @api
    rowToggleIcon;
    @api
    value;
    @track
    _displayValue;
    @track
    _urlLabel;
    _expanded = false;
    _hasChildren = false;
    _size = DEFAULT_SIZE;
    _variant = DEFAULT_VARIANT;
    connectedCallback() {
        super.connectedCallback();
        this.setAttribute('size', this._size);
        this.setAttribute('variant', this._variant);
    }
    _tabindex = 0;
    get computedButtonClass() {
        return classSet('slds-button slds-button_icon slds-m-right_x-small')
            .add({
            'slds-hidden': !this.hasChildren,
        })
            .toString();
    }
    get computedButtonStyle() {
        const { _expanded, rowToggleIcon } = this;
        if (_expanded &&
            rowToggleIcon?.expanded &&
            (rowToggleIcon.expanded.iconName || rowToggleIcon.expanded.iconSrc)) {
            return 'transform: rotate(-90deg);';
        }
        return null;
    }
    @api
    get displayValue() {
        return this._displayValue;
    }
    set displayValue(val) {
        this._displayValue = normalizeString(val) ? val : '';
    }
    @api
    get urlLabel() {
        return this._urlLabel;
    }
    set urlLabel(val) {
        this._urlLabel = normalizeString(val) ? val : '';
    }
    @api
    get tabIndex() {
        return -1;
    }
    set tabIndex(newValue) {
        this.setAttribute('tabindex', newValue);
        this._tabindex = newValue;
    }
    get buttonTabIndex() {
        return this._tabindex;
    }
    @api
    get hasChildren() {
        return this._hasChildren;
    }
    set hasChildren(value) {
        this._hasChildren = normalizeBoolean(value);
    }
    @api
    get isExpanded() {
        return this._expanded;
    }
    set isExpanded(value) {
        this._expanded = normalizeBoolean(value);
    }
    get buttonTitle() {
        let title = this.displayValue || this.urlLabel || this.value;
        if (this.isExpanded) {
            return formatLabel(i18n.collapseBranch, title);
        }
        return formatLabel(i18n.expandBranch, title);
    }
    get iconName() {
        const { _expanded, rowToggleIcon } = this;
        if (rowToggleIcon?.iconName) {
            return rowToggleIcon.iconName;
        }
        else if (_expanded && rowToggleIcon?.expanded?.iconName) {
            return rowToggleIcon.expanded.iconName;
        }
        else if (!_expanded && rowToggleIcon?.collapsed?.iconName) {
            return rowToggleIcon.collapsed.iconName;
        }
        return features.enableTreeGridRTLFix && isRTL()
            ? 'utility:chevronleft'
            : 'utility:chevronright';
    }
    get iconSrc() {
        const { _expanded, rowToggleIcon } = this;
        if (rowToggleIcon?.iconSrc) {
            return rowToggleIcon.iconSrc;
        }
        else if (_expanded && rowToggleIcon?.expanded?.iconSrc) {
            return rowToggleIcon.expanded.iconSrc;
        }
        else if (!_expanded && rowToggleIcon?.collapsed?.iconSrc) {
            return rowToggleIcon.collapsed.iconSrc;
        }
        return null;
    }
    handleChevronClick() {
        const customEvent = new CustomEvent('privatetogglecell', {
            bubbles: true,
            composed: true,
            cancelable: true,
            detail: {
                name: this.rowKeyValue,
                nextState: this.isExpanded ? false : true,
            },
        });
        this.dispatchEvent(customEvent);
    }
    @api
    focus() {
        this.refs.button.focus();
    }
}
