// lightning-base-components/src/lightning/tab/tab.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { api, track } from 'lwc';
import { normalizeBoolean } from 'lightning/utilsPrivate';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import { findAllTabbableElements } from 'lightning/focusUtils';
export default class LightningTab extends LightningShadowBaseClass {
    @track
    _loadContent = false;
    _registered = false;
    _connected;
    _deRegistrationCallback;
    _value;
    _label;
    _title;
    _iconName;
    _iconAlernativeText;
    _endIconName;
    _endIconAlternativeText;
    _showErrorIndicator;
    connectedCallback() {
        super.connectedCallback();
        this._connected = true;
        if (!import.meta.env.SSR &&
            this.template['synthetic']) {
            this.dispatchEvent(new CustomEvent('privatetabregister', {
                cancelable: true,
                bubbles: true,
                composed: true,
                detail: {
                    setDeRegistrationCallback: (deRegistrationCallback) => {
                        this._deRegistrationCallback =
                            deRegistrationCallback;
                    },
                },
            }));
        }
    }
    renderedCallback() {
        if (!this._registered &&
            !(this.template['synthetic'])) {
            this._registered = true;
            this.dispatchEvent(new CustomEvent('privatetabregister', {
                cancelable: true,
                bubbles: true,
                composed: true,
                detail: {
                    setDeRegistrationCallback: (deRegistrationCallback) => {
                        this._deRegistrationCallback =
                            deRegistrationCallback;
                    },
                },
            }));
        }
        this.setAttribute('tabindex', this.firstElementChild &&
            findAllTabbableElements(this.firstElementChild).length > 0
            ? '-1'
            : '0');
    }
    @api
    loadContent() {
        this._loadContent = true;
        if (!import.meta.env.SSR) {
            this.dispatchEvent(new CustomEvent('active'));
        }
    }
    disconnectedCallback() {
        this._connected = false;
        if (this._deRegistrationCallback) {
            this._deRegistrationCallback();
        }
    }
    @api
    get value() {
        return this._value;
    }
    set value(newValue) {
        this._value = String(newValue);
        this._dispatchDataChangeEventIfConnected();
    }
    @api
    get label() {
        return this._label;
    }
    set label(value) {
        this._label = value;
        this._dispatchDataChangeEventIfConnected();
    }
    @api
    get title() {
        return this._title;
    }
    set title(value) {
        this._title = value;
        this._dispatchDataChangeEventIfConnected();
    }
    @api
    get iconName() {
        return this._iconName;
    }
    set iconName(value) {
        this._iconName = value;
        this._dispatchDataChangeEventIfConnected();
    }
    @api
    get iconAssistiveText() {
        return this._iconAlernativeText;
    }
    set iconAssistiveText(value) {
        this._iconAlernativeText = value;
        this._dispatchDataChangeEventIfConnected();
    }
    @api
    get endIconName() {
        return this._endIconName;
    }
    set endIconName(value) {
        this._endIconName = value;
        this._dispatchDataChangeEventIfConnected();
    }
    @api
    get endIconAlternativeText() {
        return this._endIconAlternativeText;
    }
    set endIconAlternativeText(value) {
        this._endIconAlternativeText = value;
        this._dispatchDataChangeEventIfConnected();
    }
    @api
    get showErrorIndicator() {
        return this._showErrorIndicator;
    }
    set showErrorIndicator(value) {
        this._showErrorIndicator = normalizeBoolean(value);
        this._dispatchDataChangeEventIfConnected();
    }
    _dispatchDataChangeEventIfConnected() {
        if (!import.meta.env.SSR && this._connected) {
            this.dispatchEvent(new CustomEvent('privatetabdatachange', {
                cancelable: true,
                bubbles: true,
                composed: true,
            }));
        }
    }
}
