// lightning-base-components/src/lightning/toastContainer/toastContainer.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { api, createElement, track } from 'lwc';
import { classSet } from 'lightning/utils';
import { instanceName, properties, requiredProperties, } from 'lightning/overlayUtils';
import { SHOW_TOAST_EVENT_NAME } from 'lightning/showToastEvent';
import { normalizeNumber, normalizeString } from 'lightning/utilsPrivate';
import { createF6Controller } from 'lightning/f6Controller';
import { getElementWithFocus, returnFocusToElement, } from 'lightning/focusUtils';
import LightningPrimitiveOverlay from 'lightning/primitiveOverlay';
import { privateContext } from 'lightning/utilsInternal';
import { features } from 'lightning/features';
const { setContext, assertContext } = privateContext;
const CONSOLE_ERROR_MESSAGES = {
    MISSING_PROPERTY: 'Unable to show toast, missing toast property\'s "{0}"',
    MISSING_CONFIG: 'Unable to show toast, missing toast config.',
    MISSING_TOAST: 'Unable to show toast, missing toast class reference.',
};
const TOAST_POSITION = {
    TOP_CENTER: 'top-center',
    TOP_LEFT: 'top-left',
    TOP_RIGHT: 'top-right',
    BOTTOM_LEFT: 'bottom-left',
    BOTTOM_CENTER: 'bottom-center',
    BOTTOM_RIGHT: 'bottom-right',
};
const CONTAINER_POSITION = {
    ABSOLUTE: 'absolute',
    FIXED: 'fixed',
};
const DEFAULT_CONTAINER_POSITION = CONTAINER_POSITION.FIXED;
const DEFAULT_MAX_TOASTS_SHOWN = 3;
const DEFAULT_TOAST_POSITION = TOAST_POSITION.TOP_CENTER;
const MAX_NUM_GLOBAL_TOAST_CONTAINER = 1;
const globalContainer = [];
const getGlobalContainer = () => {
    return globalContainer[0];
};
const configure = (toastContainer, config) => {
    if (!toastContainer) {
        return;
    }
    const { containerPosition, maxToasts, toastPosition } = config;
    if (maxToasts) {
        toastContainer.maxToasts = maxToasts;
    }
    if (containerPosition) {
        toastContainer.containerPosition = containerPosition;
    }
    if (toastPosition) {
        toastContainer.toastPosition = toastPosition;
    }
};
const CLOSE_EVENT_NAME = 'close';
export default class ToastContainer extends LightningPrimitiveOverlay {
    static [instanceName] = 'lightning-toast-container';
    static [properties] = ['containerPosition', 'maxToasts', 'toastPosition'];
    _containerPosition = CONTAINER_POSITION.FIXED;
    _maxToasts = DEFAULT_MAX_TOASTS_SHOWN;
    _toastPosition = TOAST_POSITION.TOP_CENTER;
    _displayToasts = [];
    _queue = [];
    @track
    _showToastHandler;
    get containerClass() {
        const toastPosition = this._toastPosition.split('-');
        return classSet(`slds-grid slds-grid_vertical-reverse toast-container ${toastPosition[0]} ${toastPosition[1]}`).add({
            'slds-is-fixed': this.containerPosition === CONTAINER_POSITION.FIXED,
            'slds-is-absolute': this.containerPosition === CONTAINER_POSITION.ABSOLUTE,
        });
    }
    @api
    get containerPosition() {
        return this._containerPosition;
    }
    set containerPosition(value) {
        this._containerPosition = normalizeString(value, {
            fallbackValue: DEFAULT_CONTAINER_POSITION,
            validValues: [
                CONTAINER_POSITION.ABSOLUTE,
                CONTAINER_POSITION.FIXED,
            ],
        });
    }
    @api
    get maxToasts() {
        return this._maxToasts;
    }
    set maxToasts(value) {
        this._maxToasts = normalizeNumber(value, {
            fallbackValue: DEFAULT_MAX_TOASTS_SHOWN,
            minValue: 1,
        });
    }
    get root() {
        return this.template.querySelector('[data-toast-container]');
    }
    @api
    get toastPosition() {
        return this._toastPosition;
    }
    set toastPosition(value) {
        this._toastPosition = normalizeString(value, {
            fallbackValue: DEFAULT_TOAST_POSITION,
            validValues: [
                TOAST_POSITION.TOP_LEFT,
                TOAST_POSITION.TOP_CENTER,
                TOAST_POSITION.TOP_RIGHT,
                TOAST_POSITION.BOTTOM_LEFT,
                TOAST_POSITION.BOTTOM_CENTER,
                TOAST_POSITION.BOTTOM_RIGHT,
            ],
        });
    }
    @api
    close() {
        this.reset();
        super.close();
    }
    constructor() {
        super();
        setContext(this);
    }
    connectedCallback() {
        assertContext(this);
        if (!this.template.synthetic) {
            this.setAttribute('data-render-mode', 'shadow');
        }
        if (globalContainer.length < MAX_NUM_GLOBAL_TOAST_CONTAINER) {
            this._showToastHandler = this.handleShowToast.bind(this);
            document.body.addEventListener(SHOW_TOAST_EVENT_NAME, this._showToastHandler);
            globalContainer.push(this);
            this.setAriaHidden();
            createF6Controller();
        }
    }
    disconnectedCallback() {
        this.close();
    }
    handleShowToast(event) {
        event.stopPropagation();
        const params = this.validateToastParameters(event);
        if (!params) {
            return;
        }
        this._queue.push(params);
        this.showNextToast();
    }
    handleToastClose(closedToast, event) {
        let toastIndex = -1;
        let source;
        this._displayToasts = this._displayToasts.filter(({ toastElement, sourceElement }, index) => {
            if (toastElement === closedToast) {
                toastIndex = index;
                source = sourceElement;
                return false;
            }
            return true;
        });
        this.root.removeChild(closedToast);
        this.showNextToast();
        if (!event.detail.isFocused) {
            return;
        }
        const nextToast = toastIndex > -1 &&
            toastIndex < this._displayToasts.length &&
            this._displayToasts[toastIndex].toastElement;
        if (nextToast && nextToast.isConnected) {
            nextToast.focus();
            return;
        }
        returnFocusToElement(source);
    }
    handleToastRegister(event) {
        const { target: toastElement, detail } = event;
        const { unregisterCallback } = detail;
        const toastCloseHandler = this.handleToastClose.bind(this, toastElement);
        toastElement.addEventListener(CLOSE_EVENT_NAME, toastCloseHandler);
        unregisterCallback(() => {
            toastElement.removeEventListener(CLOSE_EVENT_NAME, toastCloseHandler);
        });
    }
    setAriaHidden() {
        const ariaHiddenValueToSet = !this._displayToasts.length;
        if (this.template.host.ariaHidden !==
            ariaHiddenValueToSet) {
            this.template.host.ariaHidden =
                ariaHiddenValueToSet;
        }
    }
    showNextToast() {
        while (this._queue.length &&
            this._displayToasts.length < this._maxToasts) {
            const { toast, config, sourceElement } = this._queue.shift();
            const { [instanceName]: toastInstanceName } = toast;
            const toastElement = createElement(toastInstanceName, {
                is: toast,
            });
            const allowedProperties = [
                'variant',
                'mode',
                'label',
                'labelLinks',
                'message',
                'messageLinks',
            ];
            Object.entries(config).forEach(([key, value]) => {
                const keyLower = key.toLowerCase();
                const match = keyLower.match(/^on(.+)/);
                if (match) {
                    const eventName = match[1];
                    toastElement.addEventListener(eventName, value);
                }
                else if (allowedProperties.indexOf(key) > -1) {
                    toastElement[key] = value;
                }
            });
            this._displayToasts.push({ toastElement, sourceElement });
            this.root.appendChild(toastElement);
        }
        this.setAriaHidden();
    }
    reset() {
        if (this._showToastHandler) {
            document.body.removeEventListener(SHOW_TOAST_EVENT_NAME, this._showToastHandler);
            this._showToastHandler = null;
        }
        globalContainer.splice(0, globalContainer.length);
        this._containerPosition = CONTAINER_POSITION.FIXED;
        this._maxToasts = DEFAULT_MAX_TOASTS_SHOWN;
        this._toastPosition = TOAST_POSITION.TOP_CENTER;
        this.root.childNodes.forEach((node) => {
            this.root.removeChild(node);
        });
        this._displayToasts = [];
        this._queue = [];
    }
    retrievePlatformShowToastConfig(toastAttributes, detail) {
        const { title, message, messageData, mode, type } = toastAttributes;
        const toast = detail.toast || undefined;
        return {
            config: {
                label: title,
                message,
                messageLinks: messageData,
                mode,
                variant: type,
            },
            toast,
            source: undefined,
        };
    }
    retrieveShowToastConfig(detail) {
        const config = { ...detail };
        const toast = detail.toast || undefined;
        const source = detail.source || undefined;
        delete config.toast;
        delete config.source;
        return { config, toast, source };
    }
    changeShowToastMapping(config) {
        if (!config.label && config.message) {
            config.label = config.message;
            config.message = '';
            if (features.enableMessageToLabelLink && config.messageLinks) {
                config.labelLinks = config.messageLinks;
                delete config.messageLinks;
            }
        }
    }
    validateToastParameters(event) {
        const sourceElement = getElementWithFocus();
        const { toastAttributes, detail } = event;
        let convertedConfig;
        if (toastAttributes && detail) {
            convertedConfig = this.retrievePlatformShowToastConfig(toastAttributes, detail);
        }
        else if (detail) {
            convertedConfig = this.retrieveShowToastConfig(detail);
        }
        if (!convertedConfig) {
            console.error(CONSOLE_ERROR_MESSAGES.MISSING_CONFIG);
            return undefined;
        }
        const { config, toast, source } = convertedConfig;
        if (!toast) {
            console.error('Unable to show toast, missing toast class reference.');
            return undefined;
        }
        if (toastAttributes && detail) {
            this.changeShowToastMapping(config);
        }
        const { [requiredProperties]: toastRequiredProps } = toast;
        if (toastRequiredProps) {
            const missingPropertyName = toastRequiredProps.find((property) => {
                if (!Object.prototype.hasOwnProperty.call(config, property) ||
                    !config[property]) {
                    return property;
                }
                return undefined;
            });
            if (missingPropertyName) {
                console.error(CONSOLE_ERROR_MESSAGES.MISSING_PROPERTY.replace('{0}', missingPropertyName));
                return undefined;
            }
        }
        return { config, toast, source, sourceElement };
    }
    static instance(config = {}) {
        if (!import.meta.env.SSR) {
            if (globalContainer.length < MAX_NUM_GLOBAL_TOAST_CONTAINER) {
                ToastContainer.open(config);
                return getGlobalContainer();
            }
            let toastContainer = getGlobalContainer();
            if (!toastContainer.isConnected) {
                toastContainer.close();
                toastContainer = null;
                ToastContainer.open(config);
                return getGlobalContainer();
            }
            configure(toastContainer, config);
            return toastContainer;
        }
        return {};
    }
}
