// lightning-base-components/src/lightning/accordionUtilsPrivate/accordionSectionList.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { getSectionByName, getSortedSections, getNextSection, } from './accordionUtils';
export class AccordionSectionList {
    privateSectionBag = {};
    privateSortedSections = [];
    add(accordionSectionInterface) {
        if (import.meta.env.SSR) {
            throw new Error('add can only be called in client-side rendering context');
        }
        this.privateSectionBag[accordionSectionInterface.id] =
            accordionSectionInterface;
        this.privateSortedSections = getSortedSections(this.privateSectionBag);
    }
    get sections() {
        return this.privateSortedSections;
    }
    remove(sectionId) {
        if (import.meta.env.SSR) {
            throw new Error('remove can only be called in client-side rendering context');
        }
        delete this.privateSectionBag[sectionId];
        this.privateSortedSections = getSortedSections(this.privateSectionBag);
    }
    getSectionById(sectionId) {
        return this.privateSectionBag[sectionId];
    }
    get openSections() {
        return this.privateSortedSections.filter((section) => section.isOpen());
    }
    getSectionByName(name) {
        return getSectionByName(this.privateSortedSections, name);
    }
    getNextSectionTo(sectionId) {
        return getNextSection(this.privateSortedSections, this.privateSectionBag[sectionId], false);
    }
    getPrevSectionTo(sectionId) {
        return getNextSection(this.privateSortedSections, this.privateSectionBag[sectionId], true);
    }
}
