// lightning-base-components/src/lightning/breadcrumb/breadcrumb.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { api, track } from 'lwc';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import { sanitizeURL } from 'lightning/utilsPrivate';
const DEFAULT_HREF = '#';
export default class LightningBreadcrumb extends LightningShadowBaseClass {
    static validationOptOut = ['class'];
    @api
    label;
    @api
    name;
    @track
    _ariaCurrent;
    connected;
    @api
    get ariaCurrent() {
        return this._ariaCurrent;
    }
    set ariaCurrent(value) {
        this._ariaCurrent = value;
    }
    @track
    _href;
    @api
    get href() {
        return this._href || DEFAULT_HREF;
    }
    set href(value) {
        this._href = sanitizeURL(value);
    }
    @api
    focus() {
        if (this.connected) {
            this.linkElement.focus();
        }
    }
    @api
    blur() {
        if (this.connected) {
            this.linkElement.blur();
        }
    }
    get linkElement() {
        return !import.meta.env.SSR ? this.template.querySelector('a') : null;
    }
    handleClick(e) {
        if (!this._href) {
            e.preventDefault();
        }
    }
    connectedCallback() {
        super.connectedCallback();
        this.connected = true;
        this.classList.add('slds-breadcrumb__item');
        this.setAttribute('role', 'listitem');
    }
    disconnectedCallback() {
        this.connected = false;
    }
}
