// lightning-base-components/src/lightning/formattedEmail/formattedEmail.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { api } from 'lwc';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import labelEmailLabel from '@salesforce/label/LightningFormattedEmail.emailIconLabel';
const i18n = {
    emailLabel: labelEmailLabel,
};
export default class LightningFormattedEmail extends LightningShadowBaseClass {
    @api
    value;
    @api
    label;
    @api
    tabIndex;
    @api
    hideIcon = false;
    _connected = false;
    connectedCallback() {
        super.connectedCallback();
        this._connected = true;
    }
    disconnectedCallback() {
        this._connected = false;
    }
    @api
    focus() {
        if (this.emailAnchor) {
            this.emailAnchor.focus();
        }
    }
    @api
    blur() {
        if (this.emailAnchor) {
            this.emailAnchor.blur();
        }
    }
    @api
    click() {
        const anchor = this.emailAnchor;
        if (anchor && anchor.click) {
            anchor.click();
        }
    }
    get emailAnchor() {
        if (this._connected && this.hasValue) {
            return this.template.querySelector('a');
        }
        return undefined;
    }
    get trimmedValue() {
        return (this.value || '').trim();
    }
    get trimmedLabel() {
        return (this.label || '').trim();
    }
    get hasValue() {
        return !!this.trimmedValue;
    }
    get href() {
        return 'mailto:' + this.trimmedValue;
    }
    get computedLabel() {
        return this.trimmedLabel || this.trimmedValue;
    }
    get i18n() {
        return i18n;
    }
}
