// lightning-base-components/src/lightning/relativeDateTime/relativeDateTime.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { track, api } from 'lwc';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import { assert } from 'lightning/utilsPrivate';
import { privateContext } from 'lightning/utilsInternal';
import { getFormattedRelativeDate, getTimeoutUnitsTillInvalid } from './utils';
const { setContext, assertContext } = privateContext;
export default class LightningRelativeDateTime extends LightningShadowBaseClass {
    @track
    formattedValue = '';
    set value(next) {
        this.privateValue = next;
        this.updateFormattedValue();
    }
    @api
    get value() {
        return this.privateValue;
    }
    privateValue;
    constructor() {
        super();
        setContext(this);
    }
    disconnectedCallback() {
        this.cancelScheduledFormattedValueUpdate();
    }
    renderedCallback() {
        this.scheduleFormattedValueUpdate();
    }
    updateFormattedValue() {
        const oldFormattedValue = this.formattedValue;
        try {
            this.formattedValue = getFormattedRelativeDate(this.value);
        }
        catch (e) {
            const errorMessage = `<lightning-relative-date-time>: Error while formatting ` +
                `"${this.value}": ${e.message}`;
            assert(false, errorMessage);
            this.formattedValue = '';
        }
        return oldFormattedValue !== this.formattedValue;
    }
    scheduleFormattedValueUpdate() {
        assertContext(this);
        this.cancelScheduledFormattedValueUpdate();
        if (this.formattedValue !== '') {
            const units = getTimeoutUnitsTillInvalid(this.value);
            this.formattedValueUpdateTask = setTimeout(this.updateFormattedValue.bind(this), units);
        }
    }
    cancelScheduledFormattedValueUpdate() {
        clearTimeout(this.formattedValueUpdateTask);
    }
}
