// lightning-base-components/src/lightning/select/select.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import labelRequired from '@salesforce/label/LightningControl.required';
import labelHelpTextAlternativeText from '@salesforce/label/LightningInput.helptextAlternativeText';
import { api, track } from 'lwc';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import { classSet, formatLabel } from 'lightning/utils';
import { classListMutation, normalizeBoolean, reflectAttribute, isNativeComponent, valueSetter, } from 'lightning/utilsPrivate';
import { privateContext } from 'lightning/utilsInternal';
import { FieldConstraintApi, InteractingState, isEmptyString, getErrorMessage, normalizeVariant, VARIANT, } from 'lightning/inputUtils';
import AriaObserver from 'lightning/ariaObserver';
import { features } from 'lightning/features';
const { setContext, assertContext } = privateContext;
const i18n = {
    required: labelRequired,
    helpTextAlternativeText: labelHelpTextAlternativeText,
};
export default class LightningSelect extends LightningShadowBaseClass {
    static delegatesFocus = true;
    static validationOptOut = ['class'];
    @api
    label;
    @api
    name;
    @api
    messageWhenValueMissing;
    @api
    accessKey;
    @api
    autocomplete;
    @track
    _options = [];
    _helpMessage = '';
    @track
    _selectedValue;
    @track
    _variant;
    _required = false;
    _disabled = false;
    _multiple = false;
    @track
    _fieldLevelHelp;
    @track
    _size;
    _ariaDescribedBy = '';
    _ariaLabelledBy = '';
    _helpMessageChanged = false;
    ariaObserver = null;
    _isNativeShadow;
    _connected;
    _rendered;
    interactingState;
    customErrorMessage;
    _constraintApi;
    constructor() {
        super();
        setContext(this);
        this.ariaObserver = new AriaObserver(this);
        this._isNativeShadow = isNativeComponent(this);
    }
    connectedCallback() {
        assertContext(this);
        super.connectedCallback();
        if (!this.ariaObserver) {
            this.ariaObserver = new AriaObserver(this);
        }
        this._connected = true;
        this.classList.add('slds-form-element');
        this.updateClassList();
        this.interactingState = new InteractingState();
        this.interactingState.onleave(() => this.reportValidity());
    }
    renderedCallback() {
        assertContext(this);
        if (this.isConnected) {
            this.ariaObserver.sync(this._isNativeShadow);
        }
        if (!this._rendered) {
            this.connectAriaDescribedBy();
            this.connectAriaLabelledBy();
            this._rendered = true;
        }
        if (this.options && this._selectedValue !== undefined) {
            this.selectOptionsByValue(this._selectedValue);
        }
        if (this._helpMessageChanged) {
            this.connectAriaDescribedBy();
            this._helpMessageChanged = false;
        }
    }
    disconnectedCallback() {
        this._connected = false;
        if (this.ariaObserver && this.ariaObserver.disconnect) {
            this.ariaObserver.disconnect();
            this.ariaObserver = null;
        }
    }
    get computedAriaInvalid() {
        return !!this._helpMessage || undefined;
    }
    @api
    get fieldLevelHelp() {
        return this._fieldLevelHelp;
    }
    set fieldLevelHelp(value) {
        this._fieldLevelHelp = value;
    }
    @api
    get variant() {
        return this._variant || VARIANT.STANDARD;
    }
    set variant(value) {
        this._variant = normalizeVariant(value);
        this.updateClassList();
        reflectAttribute(this, 'variant', this._variant);
    }
    @api
    get multiple() {
        return this._multiple;
    }
    set multiple(value) {
        this._multiple = normalizeBoolean(value);
        reflectAttribute(this, 'multiple', this._multiple);
    }
    @api
    get size() {
        if (!this.multiple) {
            return null;
        }
        if (this._size === undefined) {
            return '4';
        }
        return this._size;
    }
    set size(newValue) {
        this._size = newValue;
    }
    @api
    get required() {
        return this._required;
    }
    set required(value) {
        this._required = normalizeBoolean(value);
    }
    @api
    get disabled() {
        return this._disabled;
    }
    set disabled(value) {
        this._disabled = normalizeBoolean(value);
    }
    @api
    get value() {
        return this._selectedValue;
    }
    set value(newValue) {
        this._selectedValue = newValue;
        if (this.isConnected) {
            this.selectOptionsByValue(newValue);
        }
    }
    @api
    get options() {
        return this._options;
    }
    set options(newValue) {
        assertContext(this);
        this._options = newValue;
        if (this.isConnected && newValue) {
            this.selectOptionsByValue(this._selectedValue);
        }
    }
    @api
    get ariaDescribedBy() {
        return this._ariaDescribedBy;
    }
    set ariaDescribedBy(elements) {
        this._ariaDescribedBy = elements;
        this.connectAriaDescribedBy();
    }
    @api
    get ariaLabelledBy() {
        return this._ariaLabelledBy;
    }
    set ariaLabelledBy(references) {
        this._ariaLabelledBy = references;
        this.connectAriaLabelledBy();
    }
    updateClassList() {
        classListMutation(this.classList, {
            'slds-form-element_stacked': this.variant === VARIANT.LABEL_STACKED,
            'slds-form-element_horizontal': this.variant === VARIANT.LABEL_INLINE,
        });
    }
    @api
    focus() {
        if (this.isConnected) {
            this.selectElement.focus();
        }
    }
    @api
    blur() {
        if (this.isConnected) {
            this.selectElement.blur();
        }
    }
    @api
    get validity() {
        return this._constraint.validity;
    }
    @api
    checkValidity() {
        return this._constraint.checkValidity();
    }
    @api
    setCustomValidity(message) {
        this._constraint.setCustomValidity(message);
        this.customErrorMessage = message;
    }
    @api
    reportValidity() {
        return this._constraint.reportValidity((message) => {
            this._helpMessage = message;
            this._helpMessageChanged = true;
        });
    }
    @api
    showHelpMessageIfInvalid() {
        const validity = this.validity;
        if (validity.valid) {
            this._helpMessage = '';
            this.classList.remove('slds-has-error');
        }
        else {
            this.classList.add('slds-has-error');
            this._helpMessage = getErrorMessage(validity, {
                valueMissing: this.messageWhenValueMissing,
                customError: this.customErrorMessage,
            });
            this._helpMessageChanged = true;
        }
    }
    get i18n() {
        return i18n;
    }
    get selectElement() {
        return this.template.querySelector('select');
    }
    get isLabelHidden() {
        return this.variant === VARIANT.LABEL_HIDDEN;
    }
    get computedLabelClass() {
        return classSet('slds-form-element__label')
            .add({ 'slds-assistive-text': this.isLabelHidden })
            .toString();
    }
    get _constraint() {
        if (!this._constraintApi) {
            this._constraintApi = new FieldConstraintApi(() => this, {
                valueMissing: () => {
                    const isMultipleAndNoValue = this.multiple &&
                        (!this.value || this.value.length === 0);
                    return (!this.disabled &&
                        this.required &&
                        (isEmptyString(this.value) ||
                            isMultipleAndNoValue));
                },
            });
        }
        return this._constraintApi;
    }
    handleChange(event) {
        event.preventDefault();
        event.stopPropagation();
        this._selectedValue = this.getSelectedOptionValues();
        this.dispatchEvent(new CustomEvent('change', {
            bubbles: true,
            composed: true,
            detail: {
                value: this._selectedValue,
            },
        }));
    }
    handleFocus() {
        this.interactingState.enter();
        this.dispatchEvent(new CustomEvent('focus'));
    }
    handleBlur() {
        this.interactingState.leave();
        this.dispatchEvent(new CustomEvent('blur'));
    }
    connectAriaDescribedBy() {
        if (!import.meta.env.SSR) {
            this.ariaObserver.connect({
                targetSelector: 'select',
                attribute: 'aria-describedby',
                relatedNodeIds: this.ariaDescribedBy,
                relatedNodes: this.template.querySelector('[data-help-message]'),
            });
        }
    }
    connectAriaLabelledBy() {
        if (!import.meta.env.SSR) {
            this.ariaObserver.connect({
                targetSelector: 'select',
                attribute: 'aria-labelledby',
                relatedNodeIds: this.ariaLabelledBy,
            });
        }
    }
    selectOptionsByValue(optionValue) {
        if (this.multiple) {
            if (Array.isArray(optionValue)) {
                const options = this.template.querySelectorAll('option');
                options.forEach((option) => {
                    option.selected = optionValue.includes(option.value);
                });
            }
        }
        else {
            valueSetter(this.selectElement, [HTMLSelectElement], optionValue);
        }
    }
    getSelectedOptionValues() {
        if (this.multiple) {
            const options = this.template.querySelectorAll('option');
            return Array.prototype.reduce.call(options, (selectedValues, option) => {
                if (option.selected) {
                    selectedValues.push(option.value);
                }
                return selectedValues;
            }, []);
        }
        return this.selectElement.value;
    }
    get helptextAlternativeText() {
        return formatLabel(i18n.helpTextAlternativeText, this.label);
    }
    get isFixStaticErrorDOMEnabled() {
        return features.enableSelectStaticErrDom;
    }
}
