// lightning-base-components/src/lightning/primitiveCellFactory/primitiveCellFactory.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { features } from 'lightning/features';
import PrimitiveDatatableCell from 'lightning/primitiveDatatableCell';
import { api } from 'lwc';
import { classSetToString, isRTL } from 'lightning/utilsPrivate';
import { toDate } from 'lightning/internationalizationLibrary';
import { formatLabel } from 'lightning/utils';
import cellWithStandardLayout from './cellWithStandardLayout.html';
import bareCustomCell from './bareCustomCell.html';
import labelEdit from '@salesforce/label/LightningDatatable.edit';
import labelEditHasError from '@salesforce/label/LightningDatatable.editHasError';
import labelReadOnlyIconAlternativeText from '@salesforce/label/LightningDatatable.readOnlyIconAlternativeText';
import labelTrue from '@salesforce/label/LightningDatatable.true';
import labelFalse from '@salesforce/label/LightningDatatable.false';
const TOOLTIP_ALLOWANCE = 20;
const i18n = {
    edit: labelEdit,
    editHasError: labelEditHasError,
    readOnlyIconAlternativeText: labelReadOnlyIconAlternativeText,
    true: labelTrue,
    false: labelFalse,
};
function isSpreadComputedAlignment(computedAlignment) {
    return (!computedAlignment ||
        computedAlignment === 'left' ||
        (computedAlignment !== 'center' && computedAlignment !== 'right'));
}
export default class PrivateCellFactory extends PrimitiveDatatableCell {
    static validationOptOut = ['class'];
    @api
    types;
    @api
    alignment;
    @api
    value;
    @api
    displayValue;
    @api
    iconName;
    @api
    iconLabel;
    @api
    iconPosition;
    @api
    iconAlternativeText;
    @api
    isCustomBareLayout;
    @api
    editable;
    @api
    isEditHidden;
    @api
    displayReadOnlyIcon;
    @api
    hasError;
    @api
    columnLabel;
    @api
    columnSubType;
    @api
    rowToggleIcon;
    @api
    typeAttribute0;
    @api
    typeAttribute1;
    @api
    typeAttribute2;
    @api
    typeAttribute3;
    @api
    typeAttribute4;
    @api
    typeAttribute5;
    @api
    typeAttribute6;
    @api
    typeAttribute7;
    @api
    typeAttribute8;
    @api
    typeAttribute9;
    @api
    typeAttribute10;
    @api
    typeAttribute21;
    @api
    typeAttribute22;
    @api
    cellTypeAttributes = {};
    @api
    treeAttributes = {};
    @api
    wrapTextMaxLines;
    _isCustomType = false;
    _wrapText = false;
    _rowHasError = false;
    @api
    get isCustomType() {
        return this._isCustomType;
    }
    set isCustomType(value) {
        this._isCustomType = value;
    }
    @api
    get wrapText() {
        return this._wrapText;
    }
    set wrapText(value) {
        if (value) {
            this.classList.add('slds-cell-wrap');
        }
        else {
            this.classList.remove('slds-cell-wrap');
        }
        this._wrapText = value;
    }
    @api
    get columnType() {
        return this._columnType;
    }
    set columnType(value) {
        if (value === 'tree') {
            this.classList.add('slds-no-space');
        }
        this._columnType = value;
    }
    getActionableElements() {
        let result = [];
        const customType = this.refs.customCell;
        if (customType) {
            result = Array.from(customType.getActionableElements());
        }
        const defaultActions = Array.from(this.template.querySelectorAll('[data-navigation="enable"]'));
        return result.concat(defaultActions);
    }
    isType(typeName) {
        return typeName === this.columnType || typeName === this.columnSubType;
    }
    get consolidateTypeAttributes() {
        return features.enableConsolidateTypeAttributes;
    }
    get isText() {
        return this.isType('text');
    }
    get isNumber() {
        return this.isType('number');
    }
    get isCurrency() {
        return this.isType('currency');
    }
    get isPercent() {
        return this.isType('percent');
    }
    get isEmail() {
        return this.isType('email');
    }
    get isDateTime() {
        return this.isType('date');
    }
    get isPhone() {
        return this.isType('phone');
    }
    get isUrl() {
        return this.isType('url');
    }
    get isLocation() {
        return this.isType('location');
    }
    get isReference() {
        return this.isType('reference');
    }
    get isRowNumber() {
        if (this.isType('rowNumber')) {
            let error;
            if (features.enableConsolidateTypeAttributes) {
                error = this.cellTypeAttributes.error;
            }
            else {
                error = this.typeAttribute0;
            }
            this._rowHasError =
                error && error.title && error.messages ? true : false;
            return true;
        }
        return false;
    }
    get isAction() {
        return this.isType('action');
    }
    get isButton() {
        return this.isType('button');
    }
    get isButtonIcon() {
        return this.isType('button-icon');
    }
    get isBoolean() {
        return this.isType('boolean');
    }
    get isDateLocal() {
        return this.isType('date-local');
    }
    isLtrType() {
        const { columnType } = this;
        return (columnType === 'url' ||
            columnType === 'email' ||
            columnType === 'phone');
    }
    get hasTreeData() {
        return this.columnType === 'tree';
    }
    get isEditable() {
        const { types } = this;
        if (types) {
            return this.editable && types.isEditableType(this.columnType);
        }
        return false;
    }
    render() {
        if (this.types && this.isCustomType && this.isCustomBareLayout) {
            return bareCustomCell;
        }
        return cellWithStandardLayout;
    }
    get isSpreadAlignment() {
        return isSpreadComputedAlignment(this.computedAlignment);
    }
    get computedAlignment() {
        const { alignment } = this;
        if (alignment) {
            return alignment;
        }
        const { columnType } = this;
        if (columnType === 'button-icon') {
            return 'center';
        }
        if (columnType === 'currency' ||
            columnType === 'number' ||
            columnType === 'percent') {
            return 'right';
        }
        return alignment;
    }
    get hasLeftIcon() {
        if (this.iconName) {
            const { iconPosition } = this;
            return !iconPosition || iconPosition === 'left';
        }
        return false;
    }
    get hasRightIcon() {
        return this.iconName ? this.iconPosition === 'right' : false;
    }
    get shouldDisplayReadOnlyIcon() {
        return !this.isEditable && this.displayReadOnlyIcon === true;
    }
    get readOnlyIconAlternativeText() {
        return formatLabel(i18n.readOnlyIconAlternativeText, this.columnLabel);
    }
    get computedCellDivClass() {
        const { columnType, _wrapText } = this;
        const isActionType = 'action' === columnType || 'action' === this.columnSubType;
        return classSetToString({
            'slds-truncate': columnType !== 'button-icon' && !isActionType && !_wrapText,
            'slds-hyphenate': _wrapText,
            'slds-line-clamp': _wrapText && this.wrapTextMaxLines,
            'ltr-content-in-rtl': isRTL() && this.isLtrType(),
        });
    }
    get computedWrapperClass() {
        const { columnType, computedAlignment } = this;
        const isActionType = 'action' === columnType || 'action' === this.columnSubType;
        return classSetToString({
            'slds-grid': true,
            'slds-no-space': columnType === 'tree',
            'slds-align_absolute-center': isActionType,
            'slds-grid_align-end': computedAlignment === 'right',
            'slds-grid_align-center': computedAlignment === 'center',
            'slds-grid_align-spread': columnType !== 'tree' &&
                isSpreadComputedAlignment(computedAlignment),
        });
    }
    get computedRowNumberStyle() {
        return this._rowHasError ? '' : `padding-left: ${TOOLTIP_ALLOWANCE}px;`;
    }
    get editIconAssistiveText() {
        const suffix = this.hasError ? ` ${i18n.editHasError}` : '';
        return `${i18n.edit} ${this.columnLabel}${suffix}`;
    }
    get computedEditButtonStyle() {
        return features.enableMaxEditLimit && this.isEditHidden
            ? 'visibility: hidden;'
            : '';
    }
    get urlLabel() {
        if (features.enableConsolidateTypeAttributes) {
            return this.isUrl && this.cellTypeAttributes.label;
        }
        return this.isUrl && this.typeAttribute0;
    }
    get urlTarget() {
        if (features.enableConsolidateTypeAttributes) {
            return this.cellTypeAttributes.target || '_self';
        }
        return this.typeAttribute1 || '_self';
    }
    get urlTooltip() {
        if (features.enableConsolidateTypeAttributes) {
            if (this.cellTypeAttributes.tooltip === '') {
                return '';
            }
            return this.cellTypeAttributes.tooltip || this.value;
        }
        if (this.typeAttribute2 === '') {
            return '';
        }
        return this.typeAttribute2 || this.value;
    }
    get isChecked() {
        return !!this.value;
    }
    get dateValue() {
        return toDate(this.value);
    }
    get computedDateLocalDay() {
        if (features.enableConsolidateTypeAttributes) {
            return this.cellTypeAttributes.day || 'numeric';
        }
        return this.typeAttribute0 || 'numeric';
    }
    get computedDateLocalMonth() {
        if (features.enableConsolidateTypeAttributes) {
            return this.cellTypeAttributes.month || 'short';
        }
        return this.typeAttribute1 || 'short';
    }
    get computedDateLocalYear() {
        if (features.enableConsolidateTypeAttributes) {
            return this.cellTypeAttributes.year || 'numeric';
        }
        return this.typeAttribute2 || 'numeric';
    }
    get computedCssStyles() {
        const { _wrapText, wrapTextMaxLines } = this;
        if (_wrapText && wrapTextMaxLines) {
            return `--lwc-lineClamp: ${wrapTextMaxLines}`;
        }
        else if (_wrapText) {
            return 'overflow-wrap: anywhere';
        }
        return '';
    }
    get booleanCellAssistiveText() {
        return this.value ? i18n.true : i18n.false;
    }
    handleEditButtonClick() {
        const { rowKeyValue, colKeyValue } = this;
        const event = new CustomEvent('privateeditcell', {
            bubbles: true,
            composed: true,
            detail: {
                rowKeyValue,
                colKeyValue,
            },
        });
        this.dispatchEvent(event);
    }
    handleClick() {
        if (!this.classList.contains('slds-has-focus')) {
            this.addFocusStyles();
            this.fireCellFocusByClickEvent();
        }
    }
    fireCellFocusByClickEvent() {
        let needsRefocusOnCellElement = false;
        const { template } = this;
        const { activeElement } = this.template;
        if (activeElement &&
            (activeElement === template.querySelector('div:first-child') ||
                activeElement ===
                    template.querySelector('span.slds-grid:first-child'))) {
            needsRefocusOnCellElement = true;
        }
        const { rowKeyValue, colKeyValue } = this;
        const event = new CustomEvent('privatecellfocusedbyclick', {
            bubbles: true,
            composed: true,
            detail: {
                rowKeyValue,
                colKeyValue,
                needsRefocusOnCellElement,
            },
        });
        this.dispatchEvent(event);
    }
}
