// lightning-base-components/src/lightning/inputAddress/addressFormat.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { addressFormat } from 'lightning/internationalizationLibrary';
const FORMAT_CODE_MAP = {
    A: 'street',
    C: 'city',
    S: 'province',
    Z: 'postalCode',
    K: 'country',
};
export const parseLocaleFormat = function (format) {
    if (isValidFieldFormat(format)) {
        return format
            .toUpperCase()
            .split(/(?=[A-Z])/)
            .map((formatCode) => FORMAT_CODE_MAP[formatCode]);
    }
    return [];
};
export function getInputOrder(config) {
    const { countryCode, hasCountryPicklist, langCode, showCompactAddress } = config;
    let inputOrder = addressFormat.getAddressInputOrderAllField(langCode, countryCode);
    if (hasCountryPicklist) {
        inputOrder = 'K' + inputOrder.replace('K', '');
    }
    const parsed = parseLocaleFormat(inputOrder);
    if (showCompactAddress) {
        const index = parsed.indexOf('street');
        if (index > -1) {
            parsed.splice(index + 1, 0, 'subpremise');
        }
    }
    return parsed;
}
export function getRequiredFields(config) {
    const { countryCode, langCode } = config;
    const requireFields = addressFormat.getAddressRequireFields(langCode, countryCode);
    return parseLocaleFormat(requireFields);
}
function isValidFieldFormat(value) {
    return typeof value === 'string' && /^[ACSZK]+$/i.test(value);
}
