// lightning-base-components/src/lightning/inputAddress/inputAddress.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import labelRequired from '@salesforce/label/LightningControl.required';
import labelHelpTextAlternativeText from '@salesforce/label/LightningInput.helptextAlternativeText';
import labelMailingAddressFieldsUpdateText from '@salesforce/label/LightningInput.mailingAddressFieldsUpdateText';
import locale from '@salesforce/i18n/locale';
import { api, track } from 'lwc';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import { classSet, formatLabel } from 'lightning/utils';
import { assert, normalizeBoolean, classListMutation, reflectAttribute, } from 'lightning/utilsPrivate';
import { privateContext } from 'lightning/utilsInternal';
import { isEmptyString, FieldConstraintApi, InteractingState, normalizeVariant, VARIANT, } from 'lightning/inputUtils';
import { getInputOrder, getRequiredFields } from './addressFormat';
import { getFieldWidth, getFieldWidthClass, getTransformedFieldsMetaForLayout, } from './fieldsLayout';
const { setContext, assertContext } = privateContext;
const FIELD_TYPE = {
    TEXTAREA: 'textarea',
    INPUT: 'input',
    PICKLIST: 'combobox',
    LOOKUP_ADDRESS: 'address',
};
const i18n = {
    required: labelRequired,
    helpTextAlternativeText: labelHelpTextAlternativeText,
    mailingAddressFieldsUpdateText: labelMailingAddressFieldsUpdateText,
};
function getStreetTypeMeta(showCompactAddress, showAddressLookup) {
    if (showCompactAddress) {
        return {
            autocomplete: 'address-line1',
            name: 'street',
            maxlength: 80,
            type: showAddressLookup
                ? FIELD_TYPE.LOOKUP_ADDRESS
                : FIELD_TYPE.INPUT,
        };
    }
    return {
        autocomplete: 'street-address',
        name: 'street',
        maxlength: 255,
        type: FIELD_TYPE.TEXTAREA,
    };
}
export default class LightningInputAddress extends LightningShadowBaseClass {
    static validationOptOut = ['class'];
    static delegatesFocus = true;
    _connected;
    _showCompactAddress;
    _fieldConstraintApis;
    _combinedConstraintApi;
    _messageOnAddressUpdate;
    interactingState;
    @api
    addressLabel;
    @api
    streetLabel;
    @api
    subpremiseLabel;
    @api
    cityLabel;
    @api
    provinceLabel;
    @api
    countryLabel;
    @api
    postalCodeLabel;
    @api
    streetPlaceholder;
    @api
    subpremisePlaceholder;
    @api
    cityPlaceholder;
    @api
    provincePlaceholder;
    @api
    countryPlaceholder;
    @api
    postalCodePlaceholder;
    @api
    addressLookupPlaceholder;
    @api
    addressLookupLabel;
    @api
    provinceOptions;
    @api
    hideProvince;
    @api
    countryOptions;
    @api
    countryDisabled;
    @api
    locale = locale;
    @track
    _showAddressLookup;
    _countryLookupFilter = [];
    @track
    _fieldLevelHelp;
    @track
    _variant;
    @track
    _street = '';
    @track
    _subpremise = '';
    @track
    _city = '';
    @track
    _province = '';
    @track
    _country = '';
    @track
    _postalCode = '';
    @track
    _disabled = false;
    @track
    _readonly = false;
    @track
    _required = false;
    constructor() {
        super();
        setContext(this);
    }
    connectedCallback() {
        assertContext(this);
        super.connectedCallback();
        this._connected = true;
        this.updateHostVariantAndClass();
        this.interactingState = new InteractingState({
            debounceInteraction: true,
        });
        this.interactingState.onenter(() => {
            this.dispatchEvent(new CustomEvent('focus'));
        });
        this.interactingState.onleave(() => {
            this.reportValidity();
            this.dispatchEvent(new CustomEvent('blur'));
        });
    }
    renderedCallback() {
        if (!this.template['synthetic']) {
            const fieldset = this.template.querySelector('fieldset');
            fieldset.setAttribute('tabindex', '-1');
        }
    }
    disconnectedCallback() {
        this._connected = false;
    }
    @api
    get street() {
        return this._street;
    }
    set street(value) {
        this._street = value;
    }
    @api
    get subpremise() {
        return this._subpremise;
    }
    set subpremise(value) {
        this._subpremise = value;
    }
    @api
    get city() {
        return this._city;
    }
    set city(value) {
        this._city = value;
    }
    @api
    get province() {
        return this._province;
    }
    set province(value) {
        this._province = value;
    }
    @api
    get country() {
        return this._country;
    }
    set country(value) {
        this._country = value;
    }
    @api
    get postalCode() {
        return this._postalCode;
    }
    set postalCode(value) {
        this._postalCode = value;
    }
    @api
    get disabled() {
        return this._disabled;
    }
    set disabled(value) {
        this._disabled = normalizeBoolean(value);
    }
    @api
    get showAddressLookup() {
        return this._showAddressLookup;
    }
    set showAddressLookup(value) {
        this._showAddressLookup = normalizeBoolean(value);
    }
    @api
    get countryLookupFilter() {
        return this._countryLookupFilter;
    }
    set countryLookupFilter(value) {
        this._countryLookupFilter = Array.isArray(value) ? value : [];
    }
    @api
    get showCompactAddress() {
        return this._showCompactAddress;
    }
    set showCompactAddress(value) {
        this._showCompactAddress = normalizeBoolean(value);
    }
    @api
    get readOnly() {
        return this._readonly;
    }
    set readOnly(value) {
        this._readonly = normalizeBoolean(value);
    }
    @api
    get required() {
        return this._required;
    }
    set required(value) {
        this._required = normalizeBoolean(value);
    }
    @api
    get variant() {
        return this._variant || VARIANT.STANDARD;
    }
    set variant(value) {
        this._variant = normalizeVariant(value);
        reflectAttribute(this, 'variant', this.variant);
    }
    @api
    get fieldLevelHelp() {
        return this._fieldLevelHelp;
    }
    set fieldLevelHelp(value) {
        this._fieldLevelHelp = value;
    }
    @api
    get validity() {
        return this._combinedConstraint.validity;
    }
    @api
    checkValidity() {
        return this._combinedConstraint.checkValidity();
    }
    @api
    showHelpMessageIfInvalid() {
        this.reportValidity();
    }
    @api
    setCustomValidityForField(message, fieldName) {
        assert(this.fieldsMeta[fieldName] !== undefined, `Invalid 'fieldName': ${fieldName}`);
        this._fieldConstraints[fieldName].setCustomValidity(message);
    }
    @api
    reportValidity() {
        const valid = this.checkValidity();
        if (!this._connected) {
            return valid;
        }
        this.inputOrder.forEach((field) => {
            this._reportValidityForField(field);
        });
        return valid;
    }
    @api
    focus() {
        this.template.querySelector('[data-field]').focus();
    }
    @api
    blur() {
        Array.prototype.forEach.call(this.template.querySelectorAll('[data-field]'), (field) => field.blur());
    }
    get searchAddressButtonDisabled() {
        return this.disabled || this.readOnly;
    }
    get addressLabelRequired() {
        return this.addressLabel && this.required;
    }
    get i18n() {
        return i18n;
    }
    get isLabelHidden() {
        return this.variant === VARIANT.LABEL_HIDDEN;
    }
    updateHostVariantAndClass() {
        const isStackedType = this.variant === VARIANT.LABEL_STACKED ||
            this.variant === VARIANT.LABEL_HIDDEN;
        classListMutation(this.classList, {
            'slds-form-element': true,
            'slds-form-element_stacked': isStackedType,
            'slds-form-element_horizontal': this.variant === VARIANT.LABEL_INLINE,
        });
        reflectAttribute(this, 'variant', this.variant);
    }
    get showDedicatedLookupAddressField() {
        return this._showAddressLookup && !this._showCompactAddress;
    }
    get computedFieldsetClass() {
        return classSet('slds-form-element slds-form-element_compound slds-form-element_address')
            .add({
            'slds-form-element_stacked': this.variant === VARIANT.LABEL_STACKED,
            'slds-form-element_horizontal': this.variant === VARIANT.LABEL_INLINE,
        })
            .toString();
    }
    get computedLegendClass() {
        return classSet('slds-form-element__label slds-form-element__legend')
            .add({ 'slds-assistive-text': this.isLabelHidden })
            .toString();
    }
    get fieldsTypeMeta() {
        return {
            street: getStreetTypeMeta(this.showCompactAddress, this.showAddressLookup),
            subpremise: {
                autocomplete: 'address-line2',
                name: 'subpremise',
                maxlength: 80,
                type: FIELD_TYPE.INPUT,
            },
            city: {
                autocomplete: 'address-level2',
                name: 'city',
                maxlength: 40,
                type: FIELD_TYPE.INPUT,
            },
            province: {
                autocomplete: 'address-level1',
                name: 'province',
                maxlength: 80,
                type: this.isProvincePicklistEnabled()
                    ? FIELD_TYPE.PICKLIST
                    : FIELD_TYPE.INPUT,
            },
            postalCode: {
                autocomplete: 'postal-code',
                name: 'postalCode',
                maxlength: 20,
                type: FIELD_TYPE.INPUT,
            },
            country: {
                autocomplete: 'country',
                name: 'country',
                maxlength: 80,
                type: this.isCountryPicklistEnabled()
                    ? FIELD_TYPE.PICKLIST
                    : FIELD_TYPE.INPUT,
            },
        };
    }
    get inputOrder() {
        const hasCountryPicklist = this.fieldsTypeMeta.country.type === FIELD_TYPE.PICKLIST;
        const [langCode, countryCode] = this.locale.split('-');
        return getInputOrder({
            countryCode,
            hasCountryPicklist,
            langCode,
            showAddressLookup: this._showAddressLookup,
            showCompactAddress: this._showCompactAddress,
        });
    }
    get requiredFields() {
        const [langCode, countryCode] = this.locale.split('-');
        return getRequiredFields({
            countryCode,
            langCode,
            showAddressLookup: this._showAddressLookup,
            showCompactAddress: this._showCompactAddress,
        });
    }
    get fieldsMeta() {
        const fieldsMeta = {};
        this.inputOrder.forEach((name) => {
            fieldsMeta[name] = Object.assign({}, this.fieldsTypeMeta[name], getFieldWidth(name));
        });
        this.requiredFields.forEach((name) => {
            if (this.countryDisabled && name === 'country') {
                fieldsMeta[name].required = false;
            }
            else if (!this._hasOptions(name)) {
                fieldsMeta[name].required = false;
            }
            else {
                fieldsMeta[name].required = true;
            }
        });
        return fieldsMeta;
    }
    get domFieldsMeta() {
        const shouldHideProvince = this.hideProvince && !this.fieldsMeta?.province?.required;
        let rowKey = 0;
        const out = getTransformedFieldsMetaForLayout(this.fieldsMeta, this.inputOrder).map((row) => {
            const rowHasProvinceField = row.some(({ name }) => name === 'province');
            const rowList = row.map((field) => {
                const { name, type, required, maxlength, autocomplete } = field;
                if (rowHasProvinceField && shouldHideProvince) {
                    field.width = name === 'province' ? 0 : 6;
                }
                const widthClass = getFieldWidthClass(field);
                const value = this.getFieldValue(name);
                const suppressNoneOption = this.required &&
                    !!required &&
                    type === FIELD_TYPE.PICKLIST &&
                    value &&
                    value !== ''
                    ? true
                    : false;
                return {
                    isInput: type === FIELD_TYPE.INPUT,
                    isPicklist: type === FIELD_TYPE.PICKLIST,
                    isTextArea: type === FIELD_TYPE.TEXTAREA,
                    isLookupAddress: type === FIELD_TYPE.LOOKUP_ADDRESS,
                    value,
                    options: this.getFieldOptions(name),
                    required: this.required && !!required,
                    disabled: this.getFieldDisabled(name) || this.disabled,
                    classnames: `slds-form-element ${widthClass} ${shouldHideProvince && name === 'province'
                        ? 'slds-hide'
                        : ''}`,
                    placeholder: this.getFieldPlaceholder(name),
                    label: this.getFieldLabel(name),
                    suppressNoneOption,
                    autocomplete,
                    maxlength,
                    name,
                };
            });
            rowList.name = rowKey++;
            return rowList;
        });
        return out;
    }
    isCountryPicklistEnabled() {
        return Array.isArray(this.countryOptions);
    }
    isProvincePicklistEnabled() {
        return Array.isArray(this.provinceOptions);
    }
    handleAddress(evt) {
        const address = evt.detail.address || {};
        this.street = address.street || '';
        this.subpremise = address.subpremise || '';
        this.city = address.city || '';
        let state = address.state;
        if (this.isProvincePicklistEnabled() &&
            address.stateCode &&
            this.provinceOptions.some((option) => option.value === address.stateCode)) {
            state = address.stateCode;
        }
        this.province = state || '';
        const country = this.isCountryPicklistEnabled()
            ? address.countryCode
            : address.country;
        this.country = country || '';
        this.postalCode = address.postalCode || '';
        this._messageOnAddressUpdate = formatLabel(this.i18n.mailingAddressFieldsUpdateText, this.country);
        this.dispatchAddressChangeEvent();
    }
    handleFocus() {
        this.interactingState.enter();
    }
    handleBlur(event) {
        this.interactingState.leave();
        const field = event.target.dataset.field;
        this._reportValidityForField(field);
    }
    handleChange(event) {
        event.stopPropagation();
        const fieldName = event.target.dataset.field;
        const value = event.detail.value;
        if (this.getFieldValue(fieldName) === value) {
            return;
        }
        this[fieldName] = value;
        this.dispatchAddressChangeEvent();
    }
    dispatchAddressChangeEvent() {
        this.dispatchEvent(new CustomEvent('change', {
            composed: true,
            bubbles: true,
            detail: {
                street: this.street,
                subpremise: this.subpremise,
                city: this.city,
                province: this.province,
                country: this.country,
                postalCode: this.postalCode,
                validity: this.validity,
            },
        }));
    }
    getFieldValue(fieldName) {
        assertContext(this);
        return this[fieldName];
    }
    getFieldOptions(fieldName) {
        return this[`${fieldName}Options`];
    }
    getFieldLabel(fieldName) {
        return this[`${fieldName}Label`];
    }
    getFieldPlaceholder(fieldName) {
        return this[`${fieldName}Placeholder`];
    }
    getFieldDisabled(fieldName) {
        return !!this[`${fieldName}Disabled`];
    }
    getFieldElement(fieldName) {
        return this.template.querySelector(`[data-field="${fieldName}"]`);
    }
    get _fieldConstraints() {
        if (!this._fieldConstraintApis) {
            this._fieldConstraintApis = [
                'street',
                'city',
                'province',
                'country',
                'postalCode',
            ].reduce((constraints, field) => {
                constraints[field] = new FieldConstraintApi(() => this.getFieldElement(field), {
                    valueMissing: () => !this.disabled &&
                        this.required &&
                        this.requiredFields.indexOf(field) >= 0 &&
                        (field !== 'country' ||
                            !this.countryDisabled) &&
                        (field !== 'province' ||
                            this._hasOptions(field) ||
                            !this.isProvincePicklistEnabled()) &&
                        isEmptyString(this[field]),
                });
                return constraints;
            }, {});
        }
        return this._fieldConstraintApis;
    }
    get _combinedConstraint() {
        if (!this._combinedConstraintApi) {
            this._combinedConstraintApi = new FieldConstraintApi(() => this, {
                customError: () => Object.values(this._fieldConstraints).some((constraint) => constraint.validity.customError),
                valueMissing: () => Object.values(this._fieldConstraints).some((constraint) => constraint.validity.valueMissing),
            });
        }
        return this._combinedConstraintApi;
    }
    _reportValidityForField(field) {
        if (this._fieldConstraints[field]) {
            this._fieldConstraints[field].reportValidity((helpMessage) => {
                const fieldElement = this.getFieldElement(field);
                fieldElement.setCustomValidity(helpMessage);
                fieldElement.reportValidity();
            });
        }
    }
    get helptextAlternativeText() {
        return formatLabel(i18n.helpTextAlternativeText, this.addressLabel);
    }
    _hasOptions(field) {
        const fieldData = this.fieldsTypeMeta[field];
        const comboboxFields = {
            province: 'provinceOptions',
            country: 'countryOptions',
        };
        const options = comboboxFields[fieldData.name];
        if (options && fieldData.type === 'combobox') {
            return this[options].length !== 0;
        }
        return true;
    }
}
