// lightning-base-components/src/lightning/i18nService/i18nService.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import commonDigits from '@salesforce/i18n/common.digits';
import commonCalendarData from '@salesforce/i18n/common.calendarData';
import defaultCalendar from '@salesforce/i18n/defaultCalendar';
import defaultNumberingSystem from '@salesforce/i18n/defaultNumberingSystem';
import calendarData from '@salesforce/i18n/calendarData';
import decimalSeparator from '@salesforce/i18n/number.decimalSeparator';
import groupingSeparator from '@salesforce/i18n/number.groupingSeparator';
import percentSign from '@salesforce/i18n/number.percentSign';
import plusSign from '@salesforce/i18n/number.plusSign';
import minusSign from '@salesforce/i18n/number.minusSign';
import exponentialSign from '@salesforce/i18n/number.exponentialSign';
import superscriptingExponentSign from '@salesforce/i18n/number.superscriptingExponentSign';
import perMilleSign from '@salesforce/i18n/number.perMilleSign';
import infinity from '@salesforce/i18n/number.infinity';
import nan from '@salesforce/i18n/number.nan';
import currencySymbol from '@salesforce/i18n/number.currencySymbol';
function extractProperties(value) {
    const valueType = typeof value;
    if (valueType === 'string' ||
        valueType === 'number' ||
        valueType === 'boolean' ||
        valueType === 'undefined') {
        return value;
    }
    if (valueType === 'object') {
        if (Array.isArray(value)) {
            const newValue = [];
            for (const arrayValue of value) {
                newValue.push(extractProperties(arrayValue));
            }
            return newValue;
        }
        else {
            const newValue = {};
            const propNames = Object.getOwnPropertyNames(value).sort();
            for (const propName of propNames) {
                const propValue = extractProperties(value[propName]);
                if (propValue !== undefined) {
                    newValue[propName] = propValue;
                }
            }
            return newValue;
        }
    }
    return undefined;
}
class Cache {
    constructor() {
        this.map = new Map();
    }
    has(key) {
        const newKey = this.normalize(key);
        return this.map.has(newKey);
    }
    get(key) {
        const newKey = this.normalize(key);
        return this.map.get(newKey);
    }
    set(key, value) {
        const newKey = this.normalize(key);
        this.map.set(newKey, value);
        return this;
    }
    clear() {
        return this.map.clear();
    }
    normalize(key) {
        const newKey = extractProperties(key);
        return JSON.stringify(newKey);
    }
}
const dateTimeFormatCache = new Map();
function clearDateTimeFormatCache() {
    dateTimeFormatCache.clear();
}
function getDateTimeFormat(locale = 'en-US', options = {}) {
    if (!dateTimeFormatCache.has(locale)) {
        dateTimeFormatCache.set(locale, new Cache());
    }
    if (!dateTimeFormatCache.get(locale).has(options)) {
        const instance = new Intl.DateTimeFormat(locale, options);
        dateTimeFormatCache.get(locale).set(options, instance);
    }
    return dateTimeFormatCache.get(locale).get(options);
}
function escapeRegex(value) {
    return value.replace(/[-/\\^$*+?.()|[\]{}]/g, '\\$&');
}
const intlCalendarMap = {
    gregory: 'gregorian',
};
function intlCalendarToCldr(calendar) {
    return Object.prototype.hasOwnProperty.call(intlCalendarMap, calendar)
        ? intlCalendarMap[calendar]
        : calendar;
}
const eraDateRegex = /^([-]?[0-9]+)-([0-9]{1,2})-([0-9]{1,2})$/;
function parseEraDate(value) {
    if (value && eraDateRegex.test(value)) {
        const matches = value.match(eraDateRegex);
        return {
            year: parseInt(matches[0], 10),
            month: parseInt(matches[0], 10),
            day: parseInt(matches[0], 10),
        };
    }
    return undefined;
}
function getGregorianYear(eraIndex, year, month, day, calendarType = 'gregorian', calendarData) {
    if (calendarType === 'gregorian') {
        return eraIndex === 0 ? -year : year;
    }
    else if (calendarData && calendarData.calendarSystem === 'solar' && calendarData.eras) {
        const eras = Object.keys(calendarData.eras)
            .sort()
            .map((k) => calendarData.eras[k]);
        if (eraIndex >= eras.length || eraIndex < 0) {
            eraIndex = eras.length - 1;
        }
        const era = eras[eraIndex];
        const yearAdjusted = year > 0 ? year - 1 : year < 0 ? year + 1 : 0;
        if (era._end) {
            const eraEnd = parseEraDate(era._end);
            if (!eraEnd)
                return undefined;
            return eraEnd.year - yearAdjusted;
        }
        else if (era._start) {
            const eraStart = parseEraDate(era._start);
            if (!eraStart)
                return undefined;
            return eraStart.year + yearAdjusted;
        }
    }
    else if (calendarData && calendarData.calendarSystem !== 'solar' && calendarData.eras) {
        return year;
    }
    return undefined;
}
function stringToDigits(value, localeDigits) {
    const digits = [];
    [...value].forEach((char) => {
        const digit = localeDigits.indexOf(char);
        if (digit >= 0) {
            digits.push(digit);
        }
    });
    return digits.reverse();
}
function parseDigits(value, digits) {
    return stringToDigits(value, Array.from(digits)).reduce((int, cur, idx) => (int += cur * 10 ** idx), 0);
}
function getDigitsRegexPatternInternal(digits, extras = '') {
    if (digits.length > 1 &&
        digits.reduce((sum, digit, idx, arr) => idx === 0
            ? true
            : sum && digit.codePointAt(0) - arr[idx - 1].codePointAt(0) === 1, false)) {
        return '[' + [digits[0], digits[digits.length - 1]].join('-') + escapeRegex(extras) + ']';
    }
    return '[' + digits.join('') + escapeRegex(extras) + ']';
}
function getDigitsRegexPattern(digits, extras = '') {
    return getDigitsRegexPatternInternal(Array.from(digits), extras);
}
let utcFormat;
function getUtcFormat() {
    if (!utcFormat) {
        utcFormat = new Intl.DateTimeFormat('en-US', {
            timeZone: 'UTC',
            hourCycle: 'h23',
            year: 'numeric',
            month: 'numeric',
            day: 'numeric',
            hour: 'numeric',
            minute: 'numeric',
        });
    }
    return utcFormat;
}
const EN_US_DATE_REGEXP = /(\d+).(\d+).(\d+),?\s+(\d+).(\d+)(.(\d+))?/;
const EN_US_DATE_INDEX_DAY = 1;
const EN_US_DATE_INDEX_HOUR = 3;
const EN_US_DATE_INDEX_MINUTE = 4;
const TIMEZONE_DATEFORMATS = new Map();
function parseEnUsDate(value) {
    const dateString = value.replace(/[\u200E\u200F]/g, '');
    return [].slice.call(EN_US_DATE_REGEXP.exec(dateString), 1).map(Math.floor);
}
function diffMinutes(date1, date2) {
    let day = date1[EN_US_DATE_INDEX_DAY] - date2[EN_US_DATE_INDEX_DAY];
    const hour = date1[EN_US_DATE_INDEX_HOUR] - date2[EN_US_DATE_INDEX_HOUR];
    const min = date1[EN_US_DATE_INDEX_MINUTE] - date2[EN_US_DATE_INDEX_MINUTE];
    if (day > 15) {
        day = -1;
    }
    if (day < -15) {
        day = 1;
    }
    return 60 * (24 * day + hour) + min;
}
function getTimeZoneOffset(timeZone, date) {
    if (!TIMEZONE_DATEFORMATS.has(timeZone)) {
        TIMEZONE_DATEFORMATS.set(timeZone, new Intl.DateTimeFormat('en-US', {
            timeZone: timeZone,
            hourCycle: 'h23',
            year: 'numeric',
            month: 'numeric',
            day: 'numeric',
            hour: 'numeric',
            minute: 'numeric',
        }));
    }
    const dateFormat = TIMEZONE_DATEFORMATS.get(timeZone);
    return diffMinutes(parseEnUsDate(getUtcFormat().format(date)), parseEnUsDate(dateFormat.format(date)));
}
function isValidWeek(week) {
    return (Number.isInteger(week) && week >= 1 && week <= 53);
}
function isValidDayOfWeek(dayOfWeek) {
    return (Number.isInteger(dayOfWeek) &&
        dayOfWeek >= 1 &&
        dayOfWeek <= 7);
}
function isValidMonth(month) {
    return is31DayMonth(month) || is30DayMonth(month) || month === 2;
}
function is31DayMonth(month) {
    return (month === 1 ||
        month === 3 ||
        month === 5 ||
        month === 7 ||
        month === 8 ||
        month === 10 ||
        month === 12);
}
function is30DayMonth(month) {
    return month === 4 || month === 6 || month === 9 || month === 11;
}
function isValidOrdinal(ordinal, isLeapYear) {
    return (Number.isInteger(ordinal) &&
        ordinal >= 1 &&
        ((isLeapYear && ordinal <= 366) ||
            (!isLeapYear && ordinal <= 365)));
}
function isValidDate(month, day, isLeapYear) {
    return (isValidMonth(month) &&
        Number.isInteger(day) &&
        day >= 1 &&
        ((day <= 30 && is30DayMonth(month)) ||
            (day <= 31 && is31DayMonth(month)) ||
            (isLeapYear && month === 2 && day <= 29) ||
            (!isLeapYear && month === 2 && day <= 28)));
}
class Calendar {
    constructor(timeZone) {
        this.year = 0;
        this.month = 1;
        this.day = 1;
        this.hour = 0;
        this.minute = 0;
        this.second = 0;
        this.millisecond = 0;
        this.timeZone = timeZone;
    }
    clear() {
        this.year = 0;
        this.month = 1;
        this.day = 1;
        this.hour = 0;
        this.minute = 0;
        this.second = 0;
        this.millisecond = 0;
    }
    getDate(smallYearOffset = 2000) {
        if (this.year < 100 && this.year >= 0) {
            this.year += smallYearOffset;
        }
        const localDate = new Date(this.year, this.month - 1, this.day, this.hour, this.minute, this.second, this.millisecond);
        if (this.timeZone) {
            const localOffset = localDate.getTimezoneOffset();
            const timeZoneOffset = getTimeZoneOffset(this.timeZone, localDate);
            const offset = localOffset - timeZoneOffset;
            if (Math.floor(offset) !== 0) {
                const adjustedDate = new Date(localDate.getTime() - offset * 60 * 1000);
                return adjustedDate;
            }
        }
        return localDate;
    }
}
function tokenizeDateTimePattern(value) {
    const tokenValues = [...'GyYQqM<wWdDFEecabBhHKkmsSAxXOvVXxZz'];
    const tokens = [];
    let isEscapedText = false;
    let isPrevSingleQuote = false;
    [...value].forEach((char) => {
        const tokenValue = tokenValues.indexOf(char) >= 0 ? char : undefined;
        let isCharText = false;
        if ((tokenValue !== undefined || char !== "'") && isPrevSingleQuote) {
            isEscapedText = !isEscapedText;
        }
        if (tokenValue !== undefined && !isEscapedText) {
            if (tokens.length > 0 && tokens[tokens.length - 1].type === tokenValue) {
                let tokenLength = tokens[tokens.length - 1].length;
                tokenLength = tokenLength === undefined ? 1 : tokenLength + 1;
                tokens[tokens.length - 1].length = tokenLength;
            }
            else if (tokenValue !== undefined) {
                tokens.push({
                    type: tokenValue,
                    length: 1,
                });
            }
            isPrevSingleQuote = false;
        }
        else if (char === "'" && !isPrevSingleQuote) {
            isPrevSingleQuote = true;
        }
        else if (char === "'" && isPrevSingleQuote) {
            isCharText = true;
            isPrevSingleQuote = false;
        }
        else {
            isCharText = true;
            isPrevSingleQuote = false;
        }
        if (isCharText) {
            if (tokens.length > 0 && tokens[tokens.length - 1].type === undefined) {
                let tokenText = tokens[tokens.length - 1].text;
                tokenText = tokenText === undefined ? char : tokenText + char;
                tokens[tokens.length - 1].text = tokenText;
            }
            else {
                tokens.push({
                    text: char,
                });
            }
        }
    });
    return tokens;
}
function getSymbols(token, calendarData) {
    switch (token) {
        case 'a':
        case 'aa':
        case 'aaa':
            return calendarData.dayPeriods.format.abbreviated;
        case 'aaaa':
            return calendarData.dayPeriods.format.wide;
        case 'aaaaa':
            return calendarData.dayPeriods.format.narrow;
        case 'G':
            return calendarData.eras.eraAbbr;
        case 'GGGG':
            return calendarData.eras.eraNames;
        case 'GGGGG':
            return calendarData.eras.eraNarrow;
        case 'MMM':
            return calendarData.months.format.abbreviated;
        case 'MMMM':
            return calendarData.months.format.wide;
        case 'MMMMM':
            return calendarData.months.format.narrow;
        case 'EEEE':
            return calendarData.days.format.wide;
        case 'EEEEE':
            return calendarData.days.format.narrow;
        case 'EEEEEE':
            return calendarData.days.format.abbreviated;
        default:
            return {};
    }
}
function getTokenDigits(token) {
    switch (token.type) {
        case 'y':
        case 'd':
        case 'K':
        case 'H':
        case 'h':
        case 'k':
        case 'm':
        case 's':
        case 'S':
        case 'A':
        case 'Q':
            return token.length === 2 ? 2 : 0;
        case 'M':
        case 'e':
        case 'c':
            switch (token.length) {
                case 1:
                    return 0;
                case 2:
                    return 2;
                default:
                    return -1;
            }
    }
    return -1;
}
const regexps = new Map();
function getExpression$1(pattern, digits, calData) {
    const tokens = tokenizeDateTimePattern(pattern);
    const groups = [];
    const digitsRange = getDigitsRegexPattern(digits);
    let exprValue = '^[\\s]*';
    for (let i = 0; i < tokens.length; i++) {
        const token = tokens[i];
        if (token.text !== undefined && token.text !== '') {
            exprValue += '(' + escapeRegex(token.text) + ')';
            groups.push(token);
        }
        else if (token.type !== undefined) {
            const tokenString = token.type.repeat(token.length);
            const digits = getTokenDigits(token);
            if (digits === -1) {
                if (tokenString === 'z') {
                    exprValue += '([^\\s]+)';
                }
                else {
                    const symbols = getSymbols(tokenString, calData);
                    exprValue +=
                        '(' +
                            Object.keys(symbols)
                                .map((k) => escapeRegex(symbols[k]))
                                .join('|') +
                            ')';
                }
                groups.push(token);
            }
            else if (digits === 0) {
                exprValue += '(' + digitsRange + '+)';
                groups.push(token);
            }
            else if (digits > 0) {
                exprValue += '(' + digitsRange + '{1,' + digits + '})';
                groups.push(token);
            }
        }
    }
    exprValue += '[\\s]*$';
    if (!regexps.has(exprValue)) {
        regexps.set(exprValue, new RegExp(exprValue));
    }
    return { regexp: regexps.get(exprValue), groups };
}
function getKeyByValue(parent, value) {
    for (const key in parent) {
        if (Object.prototype.hasOwnProperty.call(parent, key) && parent[key] === value) {
            return key;
        }
    }
    return undefined;
}
function getKeyFromSymbol(symbol, token, data) {
    const tokenString = token.type.repeat(token.length);
    switch (token.type) {
        case 'G':
            switch (tokenString) {
                case 'G':
                    return getKeyByValue(data.eras.eraNames, symbol);
                case 'GGGG':
                    return getKeyByValue(data.eras.eraNarrow, symbol);
                case 'GGGGG':
                    return getKeyByValue(data.eras.eraAbbr, symbol);
            }
            break;
        case 'M':
        case 'L':
            switch (tokenString) {
                case 'MMM':
                    return getKeyByValue(data.months.format.abbreviated, symbol);
                case 'MMMM':
                    return getKeyByValue(data.months.format.wide, symbol);
                case 'MMMMM':
                    return getKeyByValue(data.months.format.narrow, symbol);
                case 'LLL':
                    return getKeyByValue(data.months['stand-alone'].short, symbol);
                case 'LLLL':
                    return getKeyByValue(data.months['stand-alone'].wide, symbol);
                case 'LLLLL':
                    return getKeyByValue(data.months['stand-alone'].narrow, symbol);
            }
            break;
        case 'a':
        case 'b':
        case 'B':
            if (token.length >= 1 && token.length <= 3)
                return getKeyByValue(data.dayPeriods.format.abbreviated, symbol);
            if (token.length === 4)
                return getKeyByValue(data.dayPeriods.format.wide, symbol);
            if (token.length === 5)
                return getKeyByValue(data.dayPeriods.format.narrow, symbol);
            break;
        case 'E':
        case 'e':
        case 'c':
            switch (tokenString) {
                case 'E':
                case 'EE':
                case 'EEE':
                case 'eee':
                case 'ccc':
                    return getKeyByValue(data.days.format.abbreviated, symbol);
                case 'EEEE':
                case 'eeee':
                case 'cccc':
                    return getKeyByValue(data.days.format.wide, symbol);
                case 'EEEEE':
                case 'eeeee':
                case 'ccccc':
                    return getKeyByValue(data.days.format.abbreviated, symbol);
                case 'EEEEEE':
                case 'eeeeee':
                case 'cccccc':
                    return getKeyByValue(data.days.format.short, symbol);
            }
            break;
    }
    return undefined;
}
const MONTHS_KEYS = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12'];
class DateTimeParseImpl {
    constructor(options) {
        const data = {
            commonDigits,
            commonCalendarData,
            defaultCalendar,
            defaultNumberingSystem,
            calendarData,
        };
        this.calendarType = options.calendar
            ? intlCalendarToCldr(options.calendar)
            : data.defaultCalendar;
        const numSys = options.numberingSystem
            ? options.numberingSystem
            : data.defaultNumberingSystem;
        const timeZone = options.timeZone ? options.timeZone : undefined;
        options.numberingSystem = numSys;
        options.timeZone = timeZone;
        this.options = options;
        this.digits = data.commonDigits[numSys];
        this.commonCalendarData = data.commonCalendarData[this.calendarType];
        this.calendarData = data.calendarData[this.calendarType];
        this.expr = getExpression$1(this.options.pattern, this.digits, this.calendarData);
    }
    parse(value) {
        const matches = value.match(this.expr.regexp);
        if (!matches || matches.length <= this.expr.groups.length) {
            throw new Error("Invalid date: '" + value + "' for pattern: '" + this.options.pattern + "'");
        }
        const cal = new Calendar(this.options.timeZone);
        cal.clear();
        let era;
        let dayPeriod;
        this.expr.groups.forEach((group, groupIndex) => {
            if (group.type === undefined) {
                return;
            }
            const match = matches[groupIndex + 1];
            if (getTokenDigits(group) >= 0) {
                switch (group.type) {
                    case 'y':
                        cal.year = parseDigits(match, this.digits);
                        break;
                    case 'M':
                        cal.month = parseDigits(match, this.digits);
                        break;
                    case 'd':
                        cal.day = parseDigits(match, this.digits);
                        break;
                    case 'K':
                        cal.hour = parseDigits(match, this.digits);
                        break;
                    case 'H':
                        cal.hour = parseDigits(match, this.digits);
                        break;
                    case 'h':
                        cal.hour = parseDigits(match, this.digits);
                        break;
                    case 'k':
                        cal.hour = parseDigits(match, this.digits);
                        break;
                    case 'm':
                        cal.minute = parseDigits(match, this.digits);
                        break;
                    case 's':
                        cal.second = parseDigits(match, this.digits);
                        break;
                    case 'A':
                        cal.millisecond = parseDigits(match, this.digits);
                        break;
                }
            }
            else {
                const key = getKeyFromSymbol(match, group, this.calendarData);
                if (key) {
                    switch (group.type) {
                        case 'G':
                            era = parseInt(key, 10);
                            break;
                        case 'M':
                        case 'L':
                            cal.month = MONTHS_KEYS.indexOf(key) + 1;
                            break;
                        case 'a':
                        case 'b':
                        case 'B':
                            dayPeriod = key;
                            break;
                    }
                }
            }
        });
        if (era !== undefined && era >= 0) {
            const adjustedYear = getGregorianYear(era, cal.year, cal.month, cal.day, this.calendarType, this.commonCalendarData);
            if (adjustedYear !== undefined)
                cal.year = adjustedYear;
        }
        if (dayPeriod !== undefined) {
            switch (dayPeriod) {
                case 'am':
                case 'morning1':
                case 'morning2':
                    break;
                case 'pm':
                case 'afternoon1':
                case 'afternoon2':
                case 'evening1':
                case 'night1':
                    if (cal.hour >= 0 && cal.hour < 12)
                        cal.hour += 12;
                    break;
                case 'midnight':
                    if (cal.hour === 12)
                        cal.hour = 0;
                    break;
            }
        }
        return cal.getDate();
    }
}
const cldrParserCache = new Cache();
function clearDateTimeCLDRParserCache() {
    cldrParserCache.clear();
}
function getDateTimeCLDRParser(options) {
    if (!cldrParserCache.has(options)) {
        cldrParserCache.set(options, new DateTimeParseImpl(options));
    }
    return cldrParserCache.get(options);
}
const intlRelativeTimeFormatCache = new Map();
function clearRelativeTimeFormatCache() {
    intlRelativeTimeFormatCache.clear();
}
function getRelativeTimeFormat(locale = 'en-US', options = {}) {
    if (!intlRelativeTimeFormatCache.has(locale)) {
        intlRelativeTimeFormatCache.set(locale, new Cache());
    }
    if (!intlRelativeTimeFormatCache.get(locale).has(options)) {
        intlRelativeTimeFormatCache
            .get(locale)
            .set(options, new Intl.RelativeTimeFormat(locale, options));
    }
    return intlRelativeTimeFormatCache.get(locale).get(options);
}
const numberFormatCache = new Map();
function clearNumberFormatCache() {
    numberFormatCache.clear();
}
function getNumberFormat(locale = 'en-US', options = {}) {
    if (!numberFormatCache.has(locale)) {
        numberFormatCache.set(locale, new Cache());
    }
    if (!numberFormatCache.get(locale).has(options)) {
        numberFormatCache.get(locale).set(options, new Intl.NumberFormat(locale, options));
    }
    return numberFormatCache.get(locale).get(options);
}
function getExpressions(pattern, digits, symbols, currencySymbol, lenient) {
    const patterns = pattern.split(';');
    if (patterns.length >= 2) {
        return {
            positive: getExpression(patterns[0], false, digits, symbols, currencySymbol, lenient),
            negative: getExpression(patterns[1], true, digits, symbols, currencySymbol, lenient),
        };
    }
    return {
        positive: getExpression(pattern, false, digits, symbols, currencySymbol, lenient),
    };
}
function getExpression(pattern, isNegative = false, digits, symbols, currencySymbol, lenient) {
    const groups = [];
    let fractionalZeros = 0;
    let integerZeros = 0;
    const integerSeparatorIntervals = [];
    const exponentialPosition = pattern.indexOf('E');
    const decimalPattern = exponentialPosition > 0 ? pattern.slice(0, exponentialPosition) : pattern.slice(0);
    const decimalPosition = decimalPattern.indexOf('.');
    const integerPattern = decimalPosition >= 0 ? decimalPattern.slice(0, decimalPosition) : decimalPattern.slice(0);
    let integerSeparatorInterval = 0;
    let integerDigitsStarted = false;
    let hasPlusSign = false;
    let hasMinusSign = false;
    const integerPatternChars = [];
    [...integerPattern].forEach((char) => integerPatternChars.push(char));
    integerPatternChars.reverse();
    [...integerPatternChars].forEach((char) => {
        switch (char) {
            case '.':
                break;
            case '+':
                hasPlusSign = true;
                groups.push({ token: 'plusSign' });
                break;
            case '-':
                hasMinusSign = true;
                groups.push({ token: 'minusSign' });
                break;
            case '%':
                groups.push({ token: 'percentSign' });
                break;
            case '¤':
                groups.push({ token: 'currencySign' });
                break;
            case '0':
                integerZeros++;
            case '#':
                integerSeparatorInterval++;
                if (!integerDigitsStarted) {
                    groups.push({ token: 'integer' });
                    integerDigitsStarted = true;
                }
                break;
            case ',':
                integerSeparatorIntervals.push(integerSeparatorInterval);
                integerSeparatorInterval = 0;
                break;
            default:
                groups.push({ token: 'string', value: char });
        }
    });
    groups.reverse();
    if (decimalPosition >= 0) {
        groups.push({ token: 'decimalSign' });
        const fractionalPattern = decimalPattern.slice(decimalPosition);
        let fractionalDigitsStarted = false;
        [...fractionalPattern].forEach((char) => {
            switch (char) {
                case '.':
                    break;
                case '+':
                    hasPlusSign = true;
                    groups.push({ token: 'plusSign' });
                    break;
                case '-':
                    hasMinusSign = true;
                    groups.push({ token: 'minusSign' });
                    break;
                case '%':
                    groups.push({ token: 'percentSign' });
                    break;
                case '¤':
                    groups.push({ token: 'currencySign' });
                    break;
                case '0':
                    fractionalZeros++;
                case '#':
                    if (!fractionalDigitsStarted) {
                        groups.push({ token: 'fraction' });
                        fractionalDigitsStarted = true;
                    }
                    break;
                case ',':
                    break;
                default:
                    groups.push({ token: 'string', value: char });
            }
        });
    }
    let exprValue = '';
    for (let i = 0; i < groups.length; i++) {
        const group = groups[i];
        switch (group.token) {
            case 'integer':
                exprValue +=
                    '(' +
                        getDigitsRegexPattern(digits, integerSeparatorIntervals.length > 0 ? symbols.group : '') +
                        '+)';
                break;
            case 'fraction':
                exprValue +=
                    '(' +
                        getDigitsRegexPattern(digits, integerSeparatorIntervals.length > 0 ? symbols.group : '') +
                        '+)';
                if (lenient || (decimalPosition > 0 && fractionalZeros <= 0)) {
                    exprValue += '?';
                }
                break;
            case 'decimalSign':
                exprValue += '(' + escapeRegex(symbols.decimal) + ')';
                if (lenient || (decimalPosition > 0 && fractionalZeros <= 0)) {
                    exprValue += '?';
                }
                break;
            case 'plusSign':
                exprValue += '(' + escapeRegex(symbols.plusSign) + ')?';
                break;
            case 'minusSign':
                exprValue += '(' + escapeRegex(symbols.minusSign) + ')?';
                break;
            case 'percentSign':
                exprValue += '(' + escapeRegex(symbols.percentSign) + ')?';
                break;
            case 'currencySign':
                exprValue += '(' + escapeRegex(currencySymbol) + ')?';
                break;
            case 'string':
                if (group.value !== undefined) {
                    if (/^\s+$/.test(group.value)) {
                        exprValue += '(\\s+)' + (lenient ? '?' : '');
                    }
                    else {
                        exprValue += '(' + escapeRegex(group.value) + ')';
                    }
                }
                break;
        }
    }
    if (!isNegative) {
        if (!hasPlusSign && !hasMinusSign) {
            exprValue =
                '(' +
                    escapeRegex(symbols.plusSign) +
                    '|' +
                    escapeRegex(symbols.minusSign) +
                    ')?' +
                    exprValue;
            groups.unshift({ token: 'plusOrMinusSign' });
        }
        else if (!hasPlusSign) {
            exprValue = '(' + escapeRegex(symbols.plusSign) + ')?' + exprValue;
            groups.unshift({ token: 'plusSign' });
        }
        else if (!hasMinusSign) {
            exprValue = '(' + escapeRegex(symbols.minusSign) + ')?' + exprValue;
            groups.unshift({ token: 'minusSign' });
        }
    }
    const expr = new RegExp(exprValue);
    return {
        expr,
        groups,
        isNegative,
        minIntegerDigits: integerZeros,
        minFractionalDigits: fractionalZeros,
    };
}
function parseNumber(value, useNegativeExpr = false, positiveExpr, negativeExpr, digits, symbols) {
    if (!useNegativeExpr && negativeExpr) {
        const result = parseNumber(value, true, positiveExpr, negativeExpr, digits, symbols);
        if (!isNaN(result))
            return result;
    }
    const expr = useNegativeExpr && negativeExpr ? negativeExpr : positiveExpr;
    const matches = value.match(expr.expr);
    if (!matches) {
        return NaN;
    }
    let hasMinusSign = false;
    let hasPlusSign = false;
    let isPercentPattern = false;
    let integerPart = '';
    let fractionPart = '';
    const groupRegex = new RegExp(escapeRegex(symbols.group), 'g');
    for (let i = 0; i < expr.groups.length; i++) {
        const group = expr.groups[i];
        isPercentPattern = isPercentPattern || group.token === 'percentSign';
        const match = matches[i + 1];
        if (!match)
            continue;
        switch (group.token) {
            case 'integer':
                integerPart += parseDigits(match.replace(groupRegex, ''), digits);
                break;
            case 'fraction':
                fractionPart += parseDigits(match.replace(groupRegex, ''), digits);
                break;
            case 'plusSign':
                hasPlusSign = true;
                break;
            case 'minusSign':
                hasMinusSign = true;
                break;
            case 'plusOrMinusSign':
                switch (match) {
                    case symbols.plusSign:
                        hasPlusSign = true;
                        break;
                    case symbols.minusSign:
                        hasMinusSign = true;
                        break;
                }
                break;
        }
    }
    if (hasMinusSign && hasPlusSign) {
        throw new Error('String has both plus and minus sign');
    }
    let result = NaN;
    integerPart.replace(/\b0+\B/, '');
    if (isPercentPattern) {
        fractionPart = integerPart.slice(-2).padStart(2, '0') + fractionPart;
        integerPart = integerPart.slice(0, -2).padStart(1, '0');
        let num = (hasMinusSign ? '-' : '') + integerPart;
        if (fractionPart !== '') {
            num += '.' + fractionPart;
        }
        result = Number.parseFloat(num);
    }
    else {
        let num = (hasMinusSign ? '-' : '') + integerPart;
        if (expr.minFractionalDigits === undefined ||
            (expr.minFractionalDigits !== undefined &&
                expr.minFractionalDigits <= 0 &&
                fractionPart === '')) {
            result = Number.parseInt(num);
        }
        else {
            if (fractionPart !== '') {
                num += '.' + fractionPart;
            }
            result = Number.parseFloat(num);
        }
    }
    return useNegativeExpr && negativeExpr ? -result : result;
}
class NumberParseImpl {
    constructor(options) {
        const data = {
            commonDigits,
            defaultNumberingSystem,
            currencySymbol,
            symbols: {
                decimal: decimalSeparator,
                group: groupingSeparator,
                percentSign: percentSign,
                plusSign: plusSign,
                minusSign: minusSign,
                exponential: exponentialSign,
                superScriptingExponent: superscriptingExponentSign,
                perMille: perMilleSign,
                infinity: infinity,
                nan: nan,
            },
        };
        options.numberingSystem = options.numberingSystem
            ? options.numberingSystem
            : data.defaultNumberingSystem;
        this.options = options;
        this.digits = data.commonDigits[options.numberingSystem];
        this.symbols = data.symbols;
        this.currencySymbol = data.currencySymbol;
        this.lenient = !(options.lenient === false);
        const exprs = getExpressions(this.options.pattern, this.digits, this.symbols, this.currencySymbol, this.lenient);
        this.positiveExpr = exprs.positive;
        this.negativeExpr = exprs.negative;
    }
    parse(value) {
        return parseNumber(value, true, this.positiveExpr, this.negativeExpr, this.digits, this.symbols);
    }
}
const NumberParseCache = new Cache();
function clearNumberParserCache() {
    NumberParseCache.clear();
}
function getNumberParser(options) {
    if (!NumberParseCache.has(options)) {
        NumberParseCache.set(options, new NumberParseImpl(options));
    }
    return NumberParseCache.get(options);
}
const EXPR = /^[\s]*(((-|\+)?([0-9]{4})[-]?|--)(([0-9]{2})|([0-9]{2})[-]?([0-9]{2})|(W([0-9]{2})([-]?([0-9]))?)|[-]?([0-9]{3})))(T([0-9]{2})([:]?([0-9]{2})([:]?([0-9]{2})(\.([0-9]{3}))?)?)?(Z|(-|\+)([0-9]{2})([:]?([0-9]{2})?)?)?)?[\s]*$/;
const INDEX_DATE = 1;
const INDEX_YEAR_OR_NONE = INDEX_DATE + 1;
const INDEX_YEAR_SIGN = INDEX_YEAR_OR_NONE + 1;
const INDEX_YEAR = INDEX_YEAR_OR_NONE + 2;
const INDEX_MONTH_MONTHDATE_WEEK_ORDINAL = INDEX_YEAR + 1;
const INDEX_MONTH = INDEX_MONTH_MONTHDATE_WEEK_ORDINAL + 1;
const INDEX_MONTHDATE = INDEX_MONTH_MONTHDATE_WEEK_ORDINAL + 2;
const INDEX_WEEK = INDEX_MONTH_MONTHDATE_WEEK_ORDINAL + 4;
const INDEX_ORDINAL = INDEX_MONTH_MONTHDATE_WEEK_ORDINAL + 8;
const INDEX_TIME = INDEX_MONTH_MONTHDATE_WEEK_ORDINAL + 9;
const INDEX_MINUTE = INDEX_TIME + 2;
const INDEX_SECOND = INDEX_TIME + 4;
const INDEX_MILLISECOND = INDEX_TIME + 6;
const INDEX_TZ = INDEX_TIME + 8;
const INDEX_TZ_MINUTE = INDEX_TIME + 11;
const TIME_ONLY_EXPR = /^[\s]*(([0-9]{2})([:]?([0-9]{2})([:]?([0-9]{2})(\.([0-9]{3}))?)?)?(Z|(-|\+)([0-9]{2})([:]?([0-9]{2})?)?)?)[\s]*$/;
const TIME_ONLY_INDEX_TIME = 1;
const TIME_ONLY_INDEX_MINUTE = TIME_ONLY_INDEX_TIME + 2;
const TIME_ONLY_INDEX_SECOND = TIME_ONLY_INDEX_TIME + 4;
const TIME_ONLY_INDEX_MILLISECOND = TIME_ONLY_INDEX_TIME + 6;
const TIME_ONLY_INDEX_TZ = TIME_ONLY_INDEX_TIME + 8;
const TIME_ONLY_INDEX_TZ_MINUTE = TIME_ONLY_INDEX_TIME + 11;
function parseDateTimeIsoString(value) {
    if (value && EXPR.test(value)) {
        return parseDateTimeString(value);
    }
    else if (value && TIME_ONLY_EXPR.test(value)) {
        return parseTimeOnlyString(value);
    }
    throw new Error("Unparseable date '" + value + "'");
}
function parseTimeOnlyString(value) {
    const matches = value.match(TIME_ONLY_EXPR);
    if (matches) {
        let hour = 0;
        let minute = 0;
        let second = 0;
        let millisecond = 0;
        if (matches[TIME_ONLY_INDEX_TIME] !== undefined) {
            hour = Number(matches[TIME_ONLY_INDEX_TIME + 1]);
            if (matches[TIME_ONLY_INDEX_MINUTE]) {
                minute = Number(matches[TIME_ONLY_INDEX_MINUTE + 1]);
            }
            if (matches[TIME_ONLY_INDEX_SECOND] !== undefined) {
                second = Number(matches[TIME_ONLY_INDEX_SECOND + 1]);
            }
            if (matches[TIME_ONLY_INDEX_MILLISECOND] !== undefined) {
                millisecond = Number(matches[TIME_ONLY_INDEX_MILLISECOND + 1]);
            }
            if (matches[TIME_ONLY_INDEX_TZ] !== undefined) {
                if (matches[TIME_ONLY_INDEX_TZ] !== 'Z') {
                    let offset = Number(matches[TIME_ONLY_INDEX_TZ + 2]) * 60;
                    if (matches[TIME_ONLY_INDEX_TZ_MINUTE] !== undefined) {
                        offset += Number(matches[TIME_ONLY_INDEX_TZ_MINUTE + 1]);
                    }
                    if (matches[TIME_ONLY_INDEX_TZ + 1] === '+') {
                        offset = -offset;
                    }
                    minute += offset;
                }
            }
        }
        const date = new Date();
        const timestamp = Date.UTC(date.getUTCFullYear(), date.getUTCMonth(), date.getUTCDate(), hour, minute, second, millisecond);
        const parsedDate = new Date(timestamp);
        return parsedDate;
    }
    throw new Error("Unparseable date '" + value + "'");
}
function parseDateTimeString(value) {
    const matches = value.match(EXPR);
    if (matches) {
        let year;
        let isLeapYear;
        let isFullDate = true;
        let isDateExtendedFormat = false;
        if (matches[INDEX_YEAR]) {
            year = Number(matches[INDEX_YEAR]);
            if (matches[INDEX_YEAR_SIGN] && matches[INDEX_YEAR_SIGN] === '-') {
                year = -year;
            }
            isDateExtendedFormat =
                isDateExtendedFormat || matches[INDEX_YEAR_OR_NONE].endsWith('-');
        }
        let month = 1;
        let day = 1;
        if (year !== undefined && matches[INDEX_MONTH_MONTHDATE_WEEK_ORDINAL]) {
            isLeapYear = new Date(Date.UTC(year, 1, 29, 0, 0, 0, 0)).getUTCDate() === 29;
            isDateExtendedFormat =
                isDateExtendedFormat ||
                    matches[INDEX_MONTH_MONTHDATE_WEEK_ORDINAL].indexOf('-') >= 0;
            if (matches[INDEX_MONTH]) {
                if (!matches[INDEX_YEAR_OR_NONE].endsWith('-')) {
                    throw new Error("Unparseable date '" + value + "'");
                }
                month = Number(matches[INDEX_MONTH]);
                day = 1;
                isFullDate = false;
            }
            else if (matches[INDEX_MONTHDATE]) {
                month = Number(matches[INDEX_MONTHDATE]);
                day = Number(matches[INDEX_MONTHDATE + 1]);
            }
            else if (matches[INDEX_WEEK]) {
                const week = Number(matches[INDEX_WEEK + 1]);
                let dayOfWeek = 1;
                if (matches[INDEX_WEEK + 2]) {
                    dayOfWeek = Number(matches[INDEX_WEEK + 3]);
                }
                else {
                    isFullDate = false;
                }
                if (!isValidWeek(week) || !isValidDayOfWeek(dayOfWeek)) {
                    throw new Error("Unparseable date '" + value + "'");
                }
                const jan1 = new Date(Date.UTC(year, 0, 1, 0, 0, 0, 0));
                const jan1Day = jan1.getUTCDay();
                const week1Day1Year = jan1Day === 2 || jan1Day === 3 || jan1Day === 4 ? year - 1 : year;
                const week1Day1Month = jan1Day === 2 || jan1Day === 3 || jan1Day === 4 ? 12 : 1;
                const week1Day1Date = jan1Day === 2 || jan1Day === 3 || jan1Day === 4
                    ? 33 - jan1Day
                    : (9 - jan1Day) % 7;
                const weekDate = new Date(Date.UTC(week1Day1Year, week1Day1Month - 1, week1Day1Date + 7 * (week - 1) + (dayOfWeek - 1), 0, 0, 0, 0));
                year = weekDate.getUTCFullYear();
                month = weekDate.getUTCMonth() + 1;
                day = weekDate.getUTCDate();
            }
            else if (matches[INDEX_ORDINAL]) {
                const ordinal = Number(matches[INDEX_ORDINAL]);
                if (!isValidOrdinal(ordinal, isLeapYear)) {
                    throw new Error("Unparseable date '" + value + "'");
                }
                const ordinalDate = new Date(Date.UTC(year, 0, ordinal, 0, 0, 0, 0));
                month = ordinalDate.getUTCMonth() + 1;
                day = ordinalDate.getUTCDate();
            }
        }
        else if (year === undefined &&
            matches[INDEX_MONTH_MONTHDATE_WEEK_ORDINAL] &&
            matches[INDEX_MONTHDATE]) {
            year = 2000;
            month = Number(matches[INDEX_MONTHDATE]);
            day = Number(matches[INDEX_MONTHDATE + 1]);
            isLeapYear = new Date(Date.UTC(year, 1, 29, 0, 0, 0, 0)).getUTCDate() === 29;
            isFullDate = false;
            isDateExtendedFormat =
                isDateExtendedFormat ||
                    matches[INDEX_MONTH_MONTHDATE_WEEK_ORDINAL].indexOf('-') >= 0;
        }
        else {
            throw new Error("Unparseable date '" + value + "'");
        }
        if (!isValidDate(month, day, isLeapYear)) {
            throw new Error("Unparseable date '" + value + "'");
        }
        let hour = 0;
        let minute = 0;
        let second = 0;
        let millisecond = 0;
        let isTimeExtendedFormat = false;
        let isHourOnly = true;
        if (matches[INDEX_TIME] !== undefined) {
            if (!isFullDate) {
                throw new Error("Unparseable date '" + value + "'");
            }
            hour = Number(matches[INDEX_TIME + 1]);
            if (matches[INDEX_MINUTE]) {
                minute = Number(matches[INDEX_MINUTE + 1]);
                isTimeExtendedFormat =
                    isTimeExtendedFormat || matches[INDEX_MINUTE].startsWith(':');
                isHourOnly = false;
            }
            if (matches[INDEX_SECOND] !== undefined) {
                second = Number(matches[INDEX_SECOND + 1]);
                isTimeExtendedFormat =
                    isTimeExtendedFormat || matches[INDEX_SECOND].startsWith(':');
                isHourOnly = false;
            }
            if (matches[INDEX_MILLISECOND] !== undefined) {
                millisecond = Number(matches[INDEX_MILLISECOND + 1]);
            }
            if (matches[INDEX_TZ] !== undefined) {
                if (matches[INDEX_TZ] !== 'Z') {
                    let offset = Number(matches[INDEX_TZ + 2]) * 60;
                    if (matches[INDEX_TZ_MINUTE] !== undefined) {
                        offset += Number(matches[INDEX_TZ_MINUTE + 1]);
                    }
                    if (matches[INDEX_TZ + 1] === '+') {
                        offset = -offset;
                    }
                    minute += offset;
                }
            }
            if (isDateExtendedFormat !== isTimeExtendedFormat && !isHourOnly) {
                throw new Error("Unparseable date '" + value + "'");
            }
        }
        const timestamp = Date.UTC(year, month - 1, day, hour, minute, second, millisecond);
        return new Date(timestamp);
    }
    throw new Error("Unparseable date '" + value + "'");
}
class DateTimeIso8601Parse {
    constructor() { }
    parse(value) {
        return parseDateTimeIsoString(value);
    }
}
const iso8601Parser = new DateTimeIso8601Parse();
function getDateTimeISO8601Parser() {
    return iso8601Parser;
}
function clearCache() {
    clearDateTimeFormatCache();
    clearDateTimeCLDRParserCache();
    clearRelativeTimeFormatCache();
    clearNumberFormatCache();
    clearNumberParserCache();
}
export { clearCache, getDateTimeCLDRParser, getDateTimeFormat, getDateTimeISO8601Parser, getNumberFormat, getNumberParser, getRelativeTimeFormat };
