// lightning-base-components/src/lightning/baseCombobox/baseCombobox.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import labelAriaSelectedOptions from '@salesforce/label/LightningCombobox.ariaSelectedOptions';
import labelDeselectOptionKeyboard from '@salesforce/label/LightningCombobox.deselectOptionKeyboard';
import labelLoadingText from '@salesforce/label/LightningCombobox.loadingText';
import labelPillCloseButtonAlternativeText from '@salesforce/label/LightningCombobox.pillCloseButtonAlternativeText';
import labelPlaceholder from '@salesforce/label/LightningCombobox.placeholder';
import labelDefaultGroupHeader from '@salesforce/label/LightningCombobox.defaultGroupHeader';
import { api, track } from 'lwc';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import { handleKeyDownOnInput } from './keyboard';
import { BaseComboboxEvents } from './baseComboboxEvents';
import { classSet, formatLabel } from 'lightning/utils';
import { privateContext } from 'lightning/utilsInternal';
const { setContext } = privateContext;
import { assert, computeAriaInvalid, getRealDOMId, normalizeBoolean, normalizeString, reflectAttribute, synchronizeAttrs, } from 'lightning/utilsPrivate';
import { AutoPosition, Direction } from 'lightning/positionLibrary';
import { VARIANT } from 'lightning/inputUtils';
import { features } from 'lightning/features';
import AriaObserver from 'lightning/ariaObserver';
const i18n = {
    ariaSelectedOptions: labelAriaSelectedOptions,
    deselectOptionKeyboard: labelDeselectOptionKeyboard,
    pillCloseButtonAlternativeText: labelPillCloseButtonAlternativeText,
    loadingText: labelLoadingText,
    placeholder: labelPlaceholder,
    defaultGroupHeader: labelDefaultGroupHeader,
};
const SMALL_MIN_HEIGHT = '2.25rem';
const MEDIUM_MIN_HEIGHT = '6.75rem';
const INPUT_ELEMENT_SELECTOR = '.slds-combobox__input';
const INPUT_PART_ELEMENT_SELECTOR = '[part~="input-text"]';
const FAUX_INPUT_BUTTON_ELEMENT_SELECTOR = `${INPUT_ELEMENT_SELECTOR}.slds-input_faux`;
const VARIANT_LOOKUP = 'lookup';
const VIEWPORT_HEIGHT_SMALL = 834;
const ARIA_CONTROLS = 'aria-controls';
const ARIA_DETAILS = 'aria-details';
const ARIA_LABELLEDBY = 'aria-labelledby';
const ARIA_DESCRIBEDBY = 'aria-describedby';
const ARIA_LABEL = 'aria-label';
const ARIA_ACTIVEDESCENDANT = 'aria-activedescendant';
const ARIA_INVALID = 'aria-invalid';
const ARIA_ERRORMESSAGE = 'aria-errormessage';
export default class LightningBaseCombobox extends LightningShadowBaseClass {
    static delegatesFocus = true;
    static formAssociated = true;
    static validationOptOut = ['class'];
    @api
    get autocomplete() {
        return this._autocomplete;
    }
    set autocomplete(value) {
        this._autocomplete = value || 'off';
    }
    @api
    inputText = '';
    @api
    inputIconName = 'utility:down';
    @api
    inputIconSize = 'x-small';
    @api
    inputIconAlternativeText;
    @api
    inputMaxlength;
    @api
    showInputActivityIndicator = false;
    @api
    required = false;
    @api
    placeholder = i18n.placeholder;
    @api
    inputLabel;
    @api
    customRole = 'combobox';
    @api
    customAriaHasPopup = 'listbox';
    @api
    accessKey;
    @api
    ariaInvalid;
    @api
    name;
    @api
    inputPill;
    @api
    attributionLogoUrl;
    @api
    attributionLogoAssistiveText;
    @api
    helpTextElement;
    @api
    dropdownTriggerClass = '';
    _showDropdownActivityIndicator = false;
    @track
    _items = [];
    _disabled = false;
    _dropdownVisible = false;
    _hasDropdownOpened = false;
    _highlightedOptionElementId = null;
    _internals;
    _variant;
    _dropdownHeight = 'standard';
    _readonly = false;
    _logoLoaded = false;
    _inputDescribedBy = [];
    _inputAriaControls;
    _inputAriaDetails;
    _activeElementDomId;
    _autocomplete = 'off';
    _dropdownAlignment = 'left';
    originDisableDefaultHighlight;
    privateDisableDefaultHighlight;
    _editingMode = false;
    _connected = false;
    ariaObserver = null;
    _inputElement = null;
    constructor() {
        super();
        if (features.enableAssertContext260) {
            setContext(this);
        }
        this._events = new BaseComboboxEvents(this);
        if (features.enableComboboxElementInternals && !import.meta.env.SSR) {
            this._internals = this.attachInternals();
        }
        this.setupAriaObserverIfAbsent();
    }
    renderedCallback() {
        this.inputElement = null;
        this.dispatchEvent(new CustomEvent('ready', {
            detail: {
                id: this.inputId,
                name: this.name,
            },
        }));
        this.syncAriaObserver();
        this.synchronizeA11y();
        reflectAttribute(this, 'data-input-pill', this.hasInputPill);
        reflectAttribute(this.template.querySelector('lightning-primitive-icon'), 'data-input-pill-close-primicon', this.isCloseVisible);
        reflectAttribute(this.template.querySelector('slds-combobox__input-entity-icon'), 'data-input-pill-search-icon', this.hasInputPill);
    }
    connectedCallback() {
        super.connectedCallback();
        this.setupAriaObserverIfAbsent();
        if (!import.meta.env.SSR) {
            this.overrideDropdownAlignment();
        }
        this.classList.add('slds-combobox_container');
        this._connected = true;
        this._keyboardInterface = this.dropdownKeyboardInterface();
    }
    disconnectedCallback() {
        this._connected = false;
        this._listBoxElementCache = undefined;
        this.removeAriaObserver();
    }
    setupAriaObserverIfAbsent() {
        if (!this.ariaObserver) {
            this.ariaObserver = new AriaObserver(this);
        }
    }
    removeAriaObserver() {
        if (this.ariaObserver && this.ariaObserver.disconnect) {
            this.ariaObserver.disconnect();
            this.ariaObserver = null;
        }
    }
    @api
    get dropdownAlignment() {
        return this._dropdownAlignment;
    }
    set dropdownAlignment(value) {
        this._dropdownAlignment = value;
    }
    @api
    get inputControlsElement() {
        return this._inputAriaControls;
    }
    set inputControlsElement(el) {
        this._inputAriaControls = el;
        this.connectAriaControls();
    }
    @api
    get inputDetailsElements() {
        return this._inputAriaDetails;
    }
    set inputDetailsElements(element) {
        this._inputAriaDetails = element;
        this.connectAriaAttribute({
            attribute: ARIA_DETAILS,
            relatedNodeIds: element,
        });
    }
    @api
    get internals() {
        return this._internals;
    }
    syncAriaObserver() {
        this.setupAriaObserverIfAbsent();
        const ariaObserverSyncExists = this.ariaObserver && this.ariaObserver.sync;
        if (this.isConnected && ariaObserverSyncExists) {
            this.ariaObserver.sync();
        }
    }
    connectAriaControls() {
        this.connectAriaAttribute({
            attribute: ARIA_CONTROLS,
            relatedNodeIds: this._inputAriaControls,
            relatedNodes: this.dropdownElement,
        });
    }
    connectAriaAttribute(params) {
        this.setupAriaObserverIfAbsent();
        if (this.ariaObserver && this.ariaObserver.connect) {
            const relatedNodeIds = Array.isArray(params.relatedNodeIds)
                ? params.relatedNodeIds.map((el) => getRealDOMId(el)).join(' ')
                : getRealDOMId(params.relatedNodeIds);
            this.ariaObserver.connect({
                ...params,
                targetSelector: INPUT_ELEMENT_SELECTOR,
                relatedNodeIds,
            });
        }
    }
    @api
    get ariaErrorMessageElement() {
        return this._ariaErrorMessageElement;
    }
    set ariaErrorMessageElement(element) {
        this._ariaErrorMessageElement = element;
        this.connectAriaErrorMessageElement();
    }
    @api
    get inputDescribedByElements() {
        return this._inputDescribedBy;
    }
    set inputDescribedByElements(elements) {
        this._inputDescribedBy = elements;
        this.connectAriaDescribedByElements();
    }
    connectAriaDescribedByElements() {
        this.connectAriaAttribute({
            attribute: ARIA_DESCRIBEDBY,
            relatedNodeIds: this._inputDescribedBy,
            relatedNodes: this.helpTextElement,
        });
    }
    connectAriaErrorMessageElement() {
        this.connectAriaAttribute({
            attribute: ARIA_ERRORMESSAGE,
            relatedNodeIds: this._ariaErrorMessageElement,
        });
    }
    @api
    get inputLabelledByElement() {
        return this._inputLabelledBy;
    }
    set inputLabelledByElement(element) {
        this._inputLabelledBy = element;
        this.connectAriaAttribute({
            attribute: ARIA_LABELLEDBY,
            relatedNodeIds: element,
        });
    }
    @api
    get rootAriaNode() {
        return this.ariaObserver.root;
    }
    set rootAriaNode(root) {
        this.setupAriaObserverIfAbsent();
        if (this.ariaObserver) {
            this.ariaObserver.root = root;
        }
    }
    get inputId() {
        return getRealDOMId(this.inputElement);
    }
    @api
    get dropdownHeight() {
        return this._dropdownHeight;
    }
    set dropdownHeight(height) {
        this._dropdownHeight = normalizeString(height, {
            fallbackValue: 'standard',
            validValues: ['standard', 'small'],
        });
    }
    @api
    get showDropdownActivityIndicator() {
        return this._showDropdownActivityIndicator;
    }
    set showDropdownActivityIndicator(value) {
        this._showDropdownActivityIndicator = normalizeBoolean(value);
        if (this._connected) {
            if (this._showDropdownActivityIndicator) {
                if (this._shouldOpenDropDown) {
                    this.openDropdownIfNotEmpty();
                }
            }
            else if (this._dropdownVisible && this.isDropdownEmpty) {
                this.closeDropdown();
            }
        }
    }
    @api
    ariaDisabled;
    @api
    get computedAriaDisabled() {
        return this.ariaDisabled === 'true' || this.disabled ? true : false;
    }
    @api
    get disabled() {
        return this._disabled;
    }
    set disabled(value) {
        this._disabled = normalizeBoolean(value);
        reflectAttribute(this, 'disabled', this._disabled);
        if (this._disabled && this._dropdownVisible) {
            this.closeDropdown();
        }
    }
    @api
    get readOnly() {
        return this._readonly;
    }
    set readOnly(value) {
        this._readonly = normalizeBoolean(value);
        if (features.enableComboboxElementInternals) {
            reflectAttribute(this, 'readonly', this._readonly);
        }
        if (this._readonly && this._dropdownVisible) {
            this.closeDropdown();
        }
    }
    @api
    get variant() {
        return this._variant || VARIANT.STANDARD;
    }
    set variant(value) {
        this._variant = normalizeString(value, {
            fallbackValue: VARIANT.STANDARD,
            validValues: [
                VARIANT.STANDARD,
                VARIANT.LABEL_INLINE,
                VARIANT.LABEL_STACKED,
                VARIANT.LABEL_HIDDEN,
                VARIANT_LOOKUP,
            ],
        });
    }
    @api
    get items() {
        return this._unprocessedItems;
    }
    set items(items = []) {
        this._unprocessedItems = items;
        this.updateItems(items);
        if (this._connected) {
            if (this._hasDropdownOpened) {
                this.updateItems(items);
                if (this._dropdownVisible) {
                    if (this.isDropdownEmpty) {
                        this.closeDropdown();
                    }
                    else {
                        this.highlightDefaultItem();
                        this.startDropdownAutoPositioning();
                    }
                }
            }
            if (this._shouldOpenDropDown) {
                this.openDropdownIfNotEmpty();
            }
        }
    }
    @api
    get disableDefaultHighlight() {
        return this.originDisableDefaultHighlight;
    }
    set disableDefaultHighlight(value) {
        this.originDisableDefaultHighlight = value;
        this.privateDisableDefaultHighlight = normalizeBoolean(value);
    }
    @api
    highlightInputText() {
        if (this._connected) {
            requestAnimationFrame(() => {
                const { inputElement } = this;
                inputElement.setSelectionRange(0, inputElement.value.length);
            });
        }
    }
    get showAttribution() {
        return this.attributionLogoUrl;
    }
    @api
    focus() {
        if (this._connected) {
            this.inputElement.focus();
        }
    }
    @api
    focusAndOpenDropdownIfNotEmpty() {
        if (this._connected) {
            if (!this._inputHasFocus) {
                this.focus();
            }
            this.openDropdownIfNotEmpty();
        }
    }
    @api
    blur() {
        if (this._connected) {
            this.inputElement.blur();
        }
    }
    @api
    setValidity(flags, message) {
        if (this._internals) {
            this._internals.setValidity(flags, message, this.inputElement);
        }
    }
    synchronizeA11y() {
        const input = this.inputElement;
        if (!input) {
            return;
        }
        synchronizeAttrs(input, {
            [ARIA_LABEL]: this.inputLabel,
            [ARIA_INVALID]: this.computedAriaInvalid,
            [ARIA_ACTIVEDESCENDANT]: this.computedAriaActiveDescendant,
        });
        if (this.isConnected) {
            this.connectAriaControls();
        }
    }
    itemId(index) {
        return this.inputId + '-' + index;
    }
    itemIndexFromId(id) {
        return parseInt(id.substring(id.lastIndexOf('-') + 1), 10);
    }
    processItem(item) {
        const itemCopy = {};
        itemCopy.type = item.type;
        itemCopy.iconName = item.iconName;
        itemCopy.iconSize = item.iconSize;
        itemCopy.iconAlternativeText = item.iconAlternativeText;
        itemCopy.rightIconName = item.rightIconName;
        itemCopy.rightIconSize = item.rightIconSize;
        itemCopy.rightIconAlternativeText = item.rightIconAlternativeText;
        itemCopy.text = item.text;
        itemCopy.subText = item.subText;
        itemCopy.value = item.value;
        itemCopy.checked = item.checked || false;
        itemCopy.selectable =
            ['option-card', 'option-inline'].indexOf(item.type) >= 0;
        if (itemCopy.selectable) {
            itemCopy.index = this._selectableItems;
            itemCopy.id = this.itemId(itemCopy.index);
            this._selectableItems += 1;
            if (item.highlight) {
                this._highlightedItemIndex = itemCopy.index;
            }
        }
        return itemCopy;
    }
    get clearSelectionText() {
        return formatLabel(this.i18n.pillCloseButtonAlternativeText, this.inputLabel);
    }
    get isAutocompleteEnabled() {
        return this.autocomplete && this.autocomplete !== 'off';
    }
    get isStandardOrLabelVariant() {
        return [
            VARIANT.STANDARD,
            VARIANT.LABEL_INLINE,
            VARIANT.LABEL_STACKED,
            VARIANT.LABEL_HIDDEN,
        ].includes(this.variant);
    }
    get isDefaultReadOnlyTypes() {
        return this.isStandardOrLabelVariant || this.hasInputPill;
    }
    get isUserInputDisabled() {
        return this.isStandardOrLabelVariant && !this.isAutocompleteEnabled;
    }
    get _inputAriaReadOnly() {
        if (this._dropdownVisible && this.isStandardOrLabelVariant) {
            return true;
        }
        return (this._readonly ||
            (!this.isAutocompleteEnabled && this.isDefaultReadOnlyTypes));
    }
    get computedAriaActiveDescendant() {
        return this.isUserInputDisabled || !this._editingMode
            ? this._activeElementDomId
            : undefined;
    }
    get computedAriaAutocomplete() {
        if (this.hasInputPill) {
            return null;
        }
        return this._inputAriaReadOnly ? 'none' : 'list';
    }
    get computedPlaceholder() {
        if (features.enableComboboxPlaceholderA11y) {
            return this.placeholder;
        }
        return this.hasInputPill ? this.inputPill.label : this.placeholder;
    }
    get computedInputValue() {
        return this.hasInputPill ? this.inputPill.label : this.inputText;
    }
    handleListboxScroll(event) {
        event.stopPropagation();
        const listbox = event.target;
        const height = listbox.getBoundingClientRect().height;
        const maxScroll = listbox.scrollHeight - height;
        const buffer = 20;
        const bottomReached = listbox.scrollTop + buffer >= maxScroll;
        if (bottomReached) {
            this._events.dispatchEndReached();
        }
    }
    get listboxElement() {
        if (!this._listBoxElementCache) {
            this._listBoxElementCache =
                this.template.querySelector('[role="listbox"]');
        }
        return this._listBoxElementCache;
    }
    get computedUniqueElementId() {
        return this.inputId;
    }
    get dropdownElement() {
        return !import.meta.env.SSR
            ? this.template.querySelector('[data-dropdown-element]')
            : null;
    }
    get i18n() {
        return i18n;
    }
    get computedDropdownTriggerClass() {
        const dropdownTriggerClass = features.enableReadOnlyTimepickerFix
            ? this.dropdownTriggerClass || ''
            : '';
        return classSet(`slds-combobox slds-dropdown-trigger slds-dropdown-trigger_click ${dropdownTriggerClass}`)
            .add({ 'slds-is-open': this._dropdownVisible })
            .toString();
    }
    get computedDropdownClass() {
        const alignment = this.dropdownAlignment;
        let dropdownLengthClass = '';
        if (this._dropdownVisible) {
            if (this.dropdownHeight === 'standard') {
                if (window.innerHeight <= VIEWPORT_HEIGHT_SMALL) {
                    dropdownLengthClass = 'slds-dropdown_length-with-icon-7';
                }
                else {
                    dropdownLengthClass = 'slds-dropdown_length-with-icon-10';
                }
            }
            else if (this.dropdownHeight === 'small') {
                dropdownLengthClass = 'slds-dropdown_length-with-icon-5';
            }
        }
        return classSet(`slds-listbox slds-listbox_vertical slds-dropdown slds-dropdown_fluid ${dropdownLengthClass}`)
            .add({
            'slds-dropdown_left': alignment === 'left' || alignment === 'auto',
            'slds-dropdown_center': alignment === 'center',
            'slds-dropdown_right': alignment === 'right',
            'slds-dropdown_bottom': alignment === 'bottom-center',
            'slds-dropdown_bottom slds-dropdown_right slds-dropdown_bottom-right': alignment === 'bottom-right',
            'slds-dropdown_bottom slds-dropdown_left slds-dropdown_bottom-left': alignment === 'bottom-left',
        })
            .toString();
    }
    get computedInputClass() {
        const classes = classSet('slds-combobox__input');
        classes.add({
            'slds-input_faux': this.isUserInputDisabled,
            'fix-slds-input_faux': this.isUserInputDisabled,
            'slds-input': !this.isUserInputDisabled,
            'slds-is-disabled': this.disabled,
            'slds-combobox__input-value': this.computedInputValue,
            'slds-input-has-icon_group-right': !this.computedInputValue && this.showInputActivityIndicator,
        });
        return classes.toString();
    }
    get _shouldOpenDropDown() {
        return (!this.dropdownDisabled &&
            this._inputHasFocus &&
            this._requestedDropdownOpen);
    }
    get dropdownDisabled() {
        return this.readOnly || this.disabled;
    }
    handleOptionClick(event) {
        if (event.target.hasAttribute('aria-selected')) {
            event.stopPropagation();
            event.preventDefault();
            this.selectOptionAndCloseDropdown(event.target);
        }
    }
    handleOptionMouseEnter(event) {
        if (event.target.hasAttribute('aria-selected')) {
            this._editingMode = false;
            this.highlightOption(event.target);
        }
    }
    handleDropdownMouseLeave() {
        this.removeHighlight();
        if (!this._inputHasFocus) {
            this.closeDropdown();
        }
    }
    handleTriggerClick(event) {
        event.stopPropagation();
        this.allowBlur();
        if (this.dropdownDisabled) {
            return;
        }
        if (!this.hasInputPill) {
            if (this._inputAriaReadOnly) {
                if (this._dropdownVisible) {
                    this.closeDropdown();
                }
                else {
                    this.openDropdownIfNotEmpty();
                }
            }
            else {
                this.openDropdownIfNotEmpty();
            }
            this.inputElement.focus();
        }
    }
    handlePillKeyDown(event) {
        if (this.dropdownDisabled) {
            return;
        }
        if (event.key === 'Delete' ||
            event.key === 'Del' ||
            event.key === 'Backspace') {
            this.handlePillRemove();
        }
    }
    handleInputKeyDown(event) {
        if (this.dropdownDisabled) {
            return;
        }
        if (event.isComposing ||
            event.keyCode === 229 ||
            event.ctrlKey ||
            event.altKey ||
            event.metaKey) {
            return;
        }
        if (this.hasInputPill) {
            this.handlePillKeyDown(event);
        }
        else {
            handleKeyDownOnInput({
                event,
                currentIndex: this.getCurrentHighlightedOptionIndex(),
                dropdownInterface: this._keyboardInterface,
            });
        }
    }
    handleInputClick() {
        this.dispatchEvent(new CustomEvent('inputclick'));
    }
    handleTextChange(event) {
        const inputValue = event.target.value;
        if (this.variant === VARIANT.STANDARD) {
            let selectedItem;
            const itemByValue = this._unprocessedItems.find((item) => item.value === inputValue);
            selectedItem = itemByValue;
            if (!itemByValue) {
                const itemByLabel = this._unprocessedItems.find((item) => item.value && item.text === inputValue);
                selectedItem = itemByLabel;
            }
            if (selectedItem) {
                this._events.dispatchSelect(selectedItem.value);
            }
        }
        else {
            this._events.dispatchTextChange(inputValue);
        }
    }
    handleFocus() {
        this._inputHasFocus = true;
        this._events.dispatchFocus();
    }
    handleInput(event) {
        if (!this.hasInputPill) {
            this._events.dispatchTextInput(event.target.value);
        }
    }
    handleBlur() {
        this._inputHasFocus = false;
        if (this._cancelBlur) {
            return;
        }
        this.closeDropdown();
        this._events.dispatchBlur();
    }
    handleDropdownMouseDown(event) {
        const mainButton = 0;
        if (event.button === mainButton) {
            this.cancelBlur();
        }
    }
    handleDropdownMouseUp() {
        this.allowBlur();
    }
    highlightOption(option) {
        this.removeHighlight();
        if (option) {
            option.highlight();
            this._highlightedOptionElement = option;
            this._highlightedOptionElementId =
                option.getAttribute('data-item-id');
            this._activeElementDomId = option.id;
        }
        this.synchronizeA11y();
    }
    highlightOptionAndScrollIntoView(optionElement) {
        if (this._selectableItems.length === 0 || !optionElement) {
            return;
        }
        this.highlightOption(optionElement);
        scrollIntoViewIfNeeded(optionElement, this.listboxElement);
    }
    removeHighlight() {
        const option = this._highlightedOptionElement;
        if (option) {
            option.removeHighlight();
            this._highlightedOptionElement = null;
            this._highlightedOptionElementId = null;
            this._activeElementDomId = null;
        }
    }
    selectOptionAndCloseDropdown(optionElement) {
        this.closeDropdown();
        this.inputElement.focus();
        const selectedValue = optionElement.getAttribute('data-value');
        this._events.dispatchSelect(selectedValue);
    }
    handleInputSelect(event) {
        event.stopPropagation();
    }
    openDropdownIfNotEmpty() {
        if (this._dropdownVisible) {
            return;
        }
        const noOptions = !Array.isArray(this.items) || this.items.length === 0;
        if (noOptions && !this._requestedDropdownOpen) {
            this._events.dispatchDropdownOpenRequest();
        }
        if (this.isDropdownEmpty) {
            this._requestedDropdownOpen = true;
            return;
        }
        if (!this._hasDropdownOpened) {
            if (this._unprocessedItems) {
                this.updateItems(this._unprocessedItems);
            }
            this._hasDropdownOpened = true;
        }
        this._requestedDropdownOpen = false;
        this._dropdownVisible = true;
        this.startDropdownAutoPositioning();
        this.highlightDefaultItem();
        this._events.dispatchDropdownOpen();
    }
    @api
    closeDropdown() {
        if (!this._dropdownVisible) {
            return;
        }
        this.stopDropdownPositioning();
        this.removeHighlight();
        this._dropdownVisible = false;
    }
    findOptionElementByIndex(index) {
        return this.template.querySelector(`[data-item-id="${this.itemId(index)}"]`);
    }
    allowBlur() {
        this._cancelBlur = false;
    }
    cancelBlur() {
        this._cancelBlur = true;
    }
    getCurrentHighlightedOptionIndex() {
        if (this._highlightedOptionElementId &&
            this._highlightedOptionElementId.length > 0) {
            return this.itemIndexFromId(this._highlightedOptionElementId);
        }
        return -1;
    }
    get inputElement() {
        if (!import.meta.env.SSR) {
            this._inputElement =
                this._inputElement ||
                    this.template.querySelector(INPUT_ELEMENT_SELECTOR);
            return this._inputElement;
        }
        return null;
    }
    set inputElement(val) {
        this._inputElement = val;
    }
    startDropdownAutoPositioning() {
        if (this.dropdownAlignment !== 'auto') {
            return;
        }
        if (!this._autoPosition) {
            this._autoPosition = new AutoPosition(this);
        }
        const selectorToUse = this.isUserInputDisabled
            ? FAUX_INPUT_BUTTON_ELEMENT_SELECTOR
            : INPUT_PART_ELEMENT_SELECTOR;
        this._autoPosition.start({
            target: () => this.template.querySelector(selectorToUse),
            element: () => this.template.querySelector('div.slds-dropdown'),
            align: {
                horizontal: Direction.Left,
                vertical: Direction.Top,
            },
            targetAlign: {
                horizontal: Direction.Left,
                vertical: Direction.Bottom,
            },
            autoFlip: true,
            alignWidth: true,
            autoShrinkHeight: true,
            minHeight: this._selectableItems < 3
                ? SMALL_MIN_HEIGHT
                : MEDIUM_MIN_HEIGHT,
            keepInViewport: true,
        });
    }
    stopDropdownPositioning() {
        if (this._autoPosition) {
            this._autoPosition.stop();
        }
    }
    get hasInputPill() {
        return this.inputPill && Object.keys(this.inputPill).length > 0;
    }
    get isCloseVisible() {
        if (this.disabled) {
            return false;
        }
        return this.hasInputPill;
    }
    handlePillRemove() {
        this.inputElement.focus();
        this._events.dispatchPillRemove(this.inputPill);
    }
    get computedFormElementClass() {
        const hasIcon = this.hasInputPill && this.inputPill.iconName;
        return classSet('slds-combobox__form-element slds-input-has-icon')
            .add({
            'slds-input-has-icon_right': !hasIcon,
            'slds-input-has-icon_left-right': hasIcon,
        })
            .toString();
    }
    get computedAriaExpanded() {
        return this._dropdownVisible ? 'true' : 'false';
    }
    updateItems(items) {
        if (!items) {
            return;
        }
        assert(Array.isArray(items), '"items" must be an array');
        this._selectableItems = 0;
        this._highlightedItemIndex = 0;
        this._items = items.map((item) => {
            if (item.items) {
                const groupCopy = {
                    label: item.label ||
                        (item.items.length ? this.i18n.defaultGroupHeader : undefined),
                };
                groupCopy.items = item.items.map((groupItem) => {
                    return this.processItem(groupItem);
                });
                return groupCopy;
            }
            return this.processItem(item);
        });
    }
    highlightDefaultItem() {
        this.removeHighlight();
        if (!this.privateDisableDefaultHighlight && !this._editingMode) {
            requestAnimationFrame(() => {
                this.highlightOptionAndScrollIntoView(this.findOptionElementByIndex(this._highlightedItemIndex));
            });
        }
    }
    get isDropdownEmpty() {
        return (!this.showDropdownActivityIndicator &&
            (!Array.isArray(this.items) || this.items.length === 0));
    }
    dropdownKeyboardInterface() {
        const that = this;
        return {
            getTotalOptions() {
                return that._selectableItems;
            },
            selectByIndex(index) {
                that.selectOptionAndCloseDropdown(that.findOptionElementByIndex(index));
            },
            highlightOptionWithIndex(index) {
                that.highlightOptionAndScrollIntoView(that.findOptionElementByIndex(index));
            },
            isInputReadOnly() {
                return that._inputAriaReadOnly;
            },
            shouldPreventInputDeletion() {
                return (that.isAutocompleteEnabled && that.isDefaultReadOnlyTypes);
            },
            highlightOptionWithText(currentIndex, text) {
                for (let index = currentIndex + 1; index < that._items.length; index++) {
                    const option = that._items[index];
                    if (option.selectable &&
                        option.text &&
                        option.text
                            .toLowerCase()
                            .indexOf(text.toLowerCase()) === 0) {
                        that.highlightOptionAndScrollIntoView(that.findOptionElementByIndex(index));
                        return;
                    }
                }
                for (let index = 0; index < currentIndex; index++) {
                    const option = that._items[index];
                    if (option.selectable &&
                        option.text &&
                        option.text
                            .toLowerCase()
                            .indexOf(text.toLowerCase()) === 0) {
                        that.highlightOptionAndScrollIntoView(that.findOptionElementByIndex(index));
                        return;
                    }
                }
            },
            isDropdownVisible() {
                return that._dropdownVisible;
            },
            openDropdownIfNotEmpty() {
                that.openDropdownIfNotEmpty();
            },
            closeDropdown() {
                that.closeDropdown();
            },
            setEditingMode(isEditing) {
                that._editingMode = isEditing;
                if (isEditing) {
                    that._activeElementDomId = null;
                }
                that.synchronizeA11y();
            },
            getEditingMode() {
                return that._editingMode;
            },
        };
    }
    get computedAriaInvalid() {
        return computeAriaInvalid(this.ariaInvalid, true);
    }
    isShadowRoot(node) {
        return node && node.nodeType === 11;
    }
    parentNodeContainsClass(host, className) {
        let element = host;
        while (element?.parentNode) {
            element = this.isShadowRoot(element.parentNode)
                ? element.parentNode.host
                : element.parentNode;
            if (features.enableBaseComboboxNullCheck && !element) {
                break;
            }
            if (element.classList && element.classList.contains(className)) {
                return true;
            }
        }
        return false;
    }
    overrideDropdownAlignment() {
        const isModal = this.parentNodeContainsClass(this.template.host, 'slds-modal');
        if (isModal) {
            this._dropdownAlignment = 'auto';
        }
    }
}
function scrollIntoViewIfNeeded(element, scrollingParent) {
    const parentRect = scrollingParent.getBoundingClientRect();
    const findMeRect = element.getBoundingClientRect();
    if (findMeRect.top < parentRect.top) {
        if (element.offsetTop + findMeRect.height < parentRect.height) {
            scrollingParent.scrollTop = 0;
        }
        else {
            scrollingParent.scrollTop = element.offsetTop;
        }
    }
    else if (findMeRect.bottom > parentRect.bottom) {
        scrollingParent.scrollTop += findMeRect.bottom - parentRect.bottom;
    }
}
