// lightning-base-components/src/lightning/overlayManager/overlayManager.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
const TYPE_MODAL = 'lightning-modal';
const TYPE_OVERLAY = 'lightning-overlay';
const TYPE_ALERT = 'lightning-alert';
const TYPE_CONFIRM = 'lightning-confirm';
const TYPE_PROMPT = 'lightning-prompt';
const TYPE_PRIMITIVE_OVERLAY = 'lightning-primitive-overlay';
export const TYPE_TOAST_CONTAINER = 'lightning-toast-container';
export const LWC_OVERLAY_ENGINE = 'lwc';
export const LWC_OVERLAY_STARTING_ZINDEX = 9000;
export const LWC_TOAST_CONTAINER_STARTING_ZINDEX = 10000;
export const LWC_ZINDEX_INCREMENT = 2;
export const LWC_ZINDEX_OFFSET = 1;
export const LWC_OVERLAY_TYPES = Object.freeze({
    [TYPE_PRIMITIVE_OVERLAY]: { pos: 'absolute' },
    [TYPE_MODAL]: { pos: 'absolute' },
    [TYPE_OVERLAY]: { pos: 'absolute' },
    [TYPE_ALERT]: { pos: 'absolute' },
    [TYPE_CONFIRM]: { pos: 'absolute' },
    [TYPE_PROMPT]: { pos: 'absolute' },
    [TYPE_TOAST_CONTAINER]: {
        pos: 'fixed',
        maxNum: 1,
        statName: 'lwcToastContainerCount',
    },
});
export const AURA_OVERLAY_ENGINE = 'aura';
export const AURA_STARTING_ZINDEX = 9001;
export const AURA_ZINDEX_INCREMENT = 2;
export const AURA_OVERLAY_TYPES = {
    'ui:modal': { pos: 'absolute' },
};
const state = {
    stack: [],
    callback: null,
    mixedOverlayPresent: false,
    auraHighestZindex: 9001,
    auraOverlayActive: false,
    auraModalCount: 0,
    auraAllOverlayCount: 0,
    lwcModalCount: 0,
    lwcToastContainerCount: 0,
    lwcOverlayCount: 0,
    lwcAllOverlayCount: 0,
};
function triggerLwcContainerCallback() {
    if (state.callback && state.lwcAllOverlayCount > 0) {
        state.callback(state);
    }
}
export function normalizeOverlayDetails(overlayEngine, overlayType, overlayDetails) {
    const overlayObject = {};
    const missingRequiredValues = !overlayEngine || !overlayType;
    const reqdValuesNotCorrectType = (overlayEngine !== LWC_OVERLAY_ENGINE ||
        overlayEngine !== AURA_OVERLAY_ENGINE) &&
        typeof overlayType !== 'string';
    if (missingRequiredValues || reqdValuesNotCorrectType) {
        return null;
    }
    overlayObject.overlayEngine = overlayEngine;
    overlayObject.overlayType = overlayType;
    if (overlayDetails) {
        const { zIndex, id, index } = overlayDetails;
        const cleanedDetails = {};
        if (zIndex !== undefined || zIndex !== null) {
            const zIndexNumber = parseInt(zIndex, 10);
            const isZIndexANumber = !Number.isNaN(zIndexNumber);
            if (isZIndexANumber) {
                cleanedDetails.zIndex = zIndexNumber;
            }
        }
        if (id && typeof id === 'string') {
            cleanedDetails.id = id;
        }
        if (index !== undefined ||
            (index !== null && typeof index === 'number')) {
            cleanedDetails.index = index;
        }
        overlayObject.overlayDetails = cleanedDetails;
    }
    return overlayObject;
}
function getOverlayStackStats() {
    const stats = {
        hasLwc: false,
        lwcOverlays: 0,
        lwcAllOverlays: 0,
        lwcModals: 0,
        lwcToastContainers: 0,
        hasAura: false,
        auraModals: 0,
        auraOverlays: 0,
        auraAllOverlays: 0,
        auraOverlayActive: false,
        auraHighestZIndex: null,
    };
    const stackLength = state.stack.length;
    state.stack.forEach(({ overlayEngine, overlayType, overlayDetails }, stackIndex) => {
        const isLwc = overlayEngine === LWC_OVERLAY_ENGINE;
        const isAura = overlayEngine === AURA_OVERLAY_ENGINE;
        if (isLwc) {
            if (!stats.hasLwc) {
                stats.hasLwc = true;
            }
            const typeExists = !!LWC_OVERLAY_TYPES[overlayType];
            if (overlayType && typeExists) {
                const isModalType = overlayType === TYPE_MODAL;
                const isToastContainerType = overlayType === TYPE_TOAST_CONTAINER;
                if (isModalType) {
                    stats.lwcModals++;
                }
                else if (isToastContainerType) {
                    stats.lwcToastContainers++;
                }
                else {
                    stats.lwcOverlays++;
                }
                stats.lwcAllOverlays++;
            }
        }
        if (isAura) {
            if (!stats.hasAura) {
                stats.hasAura = true;
            }
            if (overlayType) {
                const isModalType = !!AURA_OVERLAY_TYPES[overlayType];
                if (isModalType) {
                    stats.auraModals++;
                }
                else {
                    stats.auraOverlays++;
                }
                stats.auraAllOverlays++;
            }
            if (overlayDetails && overlayDetails.zIndex) {
                const { zIndex } = overlayDetails;
                const { auraHighestZIndex } = stats;
                const zIndexIsNumber = !isNaN(zIndex);
                if (zIndexIsNumber &&
                    (!auraHighestZIndex || zIndex > auraHighestZIndex)) {
                    stats.auraHighestZIndex = zIndex;
                }
            }
            if (stackLength === stackIndex + 1) {
                stats.auraOverlayActive = true;
            }
        }
    });
    return stats;
}
function updateStateWithStackStats(stats) {
    if (!stats) {
        return;
    }
    if (stats.hasAura) {
        state.auraModalCount = stats.auraModals;
        state.auraOverlayCount = stats.auraOverlays;
        state.auraAllOverlayCount = stats.auraAllOverlays;
        state.auraOverlayActive = stats.auraOverlayActive;
        if (stats.auraHighestZIndex) {
            state.auraHighestZindex = stats.auraHighestZIndex;
        }
    }
    else {
        initAuraOverlayState();
    }
    if (stats.hasLwc) {
        state.lwcModalCount = stats.lwcModals;
        state.lwcToastContainerCount = stats.lwcToastContainers;
        state.lwcOverlayCount = stats.lwcOverlays;
        state.lwcAllOverlayCount = stats.lwcAllOverlays;
    }
    else {
        initLwcOverlayState();
    }
    state.mixedOverlayPresent = stats.hasLwc && stats.hasAura;
}
function updateSharedOverlayState() {
    if (state.stack.length === 0) {
        initAuraOverlayState();
        initLwcOverlayState();
    }
    else {
        const stats = getOverlayStackStats();
        updateStateWithStackStats(stats);
    }
}
function cleanOverlayDetails(details) {
    const cleanedDetails = {};
    const detailsKeys = {
        id: 'string',
        zIndex: 'number',
        index: 'number',
    };
    if (!details || typeof details !== 'object') {
        return cleanedDetails;
    }
    Object.entries(details).forEach(([key, value]) => {
        if (detailsKeys[key] && typeof value === detailsKeys[key]) {
            cleanedDetails[key] = value;
        }
    });
    return cleanedDetails;
}
export function addOverlayToSharedState(overlayObject) {
    const { overlayEngine, overlayType, overlayDetails: initialOverlayDetails, } = overlayObject;
    const missingValues = !overlayObject || !overlayEngine || !overlayType;
    const standardEngines = overlayEngine === LWC_OVERLAY_ENGINE ||
        overlayEngine === AURA_OVERLAY_ENGINE;
    if (missingValues || !standardEngines) {
        return;
    }
    const overlayDetails = cleanOverlayDetails(initialOverlayDetails);
    state.stack.push({ overlayEngine, overlayType, overlayDetails });
    updateSharedOverlayState();
    triggerLwcContainerCallback();
}
export function removeOverlayFromSharedState(overlayObject) {
    const stackLength = state.stack.length;
    const { overlayEngine, overlayDetails } = overlayObject;
    const missingValues = !overlayObject || !overlayEngine;
    const standardEngines = overlayEngine === LWC_OVERLAY_ENGINE ||
        overlayEngine === AURA_OVERLAY_ENGINE;
    let overlayId = null;
    let indexToRemove = null;
    if (stackLength === 0 || missingValues || !standardEngines) {
        return;
    }
    if (overlayDetails && overlayDetails.id) {
        overlayId = overlayDetails.id;
    }
    if (overlayId) {
        for (let index = stackLength - 1; index > -1; index--) {
            if (state.stack[index]) {
                const { overlayDetails: { id: oId }, } = state.stack[index];
                if (oId && oId === overlayId) {
                    indexToRemove = index;
                    break;
                }
            }
        }
    }
    if (indexToRemove !== null) {
        state.stack.splice(indexToRemove, 1);
        updateSharedOverlayState();
        triggerLwcContainerCallback();
    }
}
export function subscribeOverlay(shouldCall, callback) {
    state.callback = callback;
    if (shouldCall) {
        triggerLwcContainerCallback();
    }
}
export function getStatCount(statName) {
    return !state || !state[statName] ? null : state[statName];
}
function initLwcOverlayState() {
    state.lwcModalCount = 0;
    state.lwcToastContainerCount = 0;
    state.lwcOverlayCount = 0;
    state.lwcAllOverlayCount = 0;
}
function initAuraOverlayState() {
    state.auraModalCount = 0;
    state.auraOverlayCount = 0;
    state.auraAllOverlayCount = 0;
    state.auraHighestZindex = AURA_STARTING_ZINDEX;
    state.auraOverlayActive = false;
    state.auraOverlayPresent = false;
}
export function isLwcModalActive() {
    let isLwcModal = false;
    if (state.stack?.length > 0) {
        const top = state.stack[state.stack.length - 1];
        isLwcModal =
            top?.overlayEngine === LWC_OVERLAY_ENGINE &&
                top?.overlayType === TYPE_MODAL;
    }
    return isLwcModal;
}
