// lightning-base-components/src/lightning/buttonStateful/buttonStateful.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { api, track } from 'lwc';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import { classSet } from 'lightning/utils';
import { normalizeBoolean, normalizeString as normalize, buttonGroupOrderClass, } from 'lightning/utilsPrivate';
import { reflectAttribute } from 'lightning/utilsPrivate';
import { features } from 'lightning/features';
import { privateContext } from 'lightning/utilsInternal';
const { setContext, assertContext } = privateContext;
const DEFAULT_VARIANT = 'neutral';
export default class LightningButtonStateful extends LightningShadowBaseClass {
    static delegatesFocus = true;
    _originalVariant = DEFAULT_VARIANT;
    _privateVariant = DEFAULT_VARIANT;
    @track
    state = {
        isClicked: false,
    };
    @api
    disabled;
    @api
    iconNameWhenOn;
    @api
    iconNameWhenOff;
    @api
    iconNameWhenHover;
    @api
    labelWhenOff;
    @api
    labelWhenOn;
    @api
    labelWhenHover;
    @api
    groupOrder = '';
    @api
    get variant() {
        return this._originalVariant;
    }
    set variant(value) {
        this._originalVariant = value;
        this._privateVariant = this.normalizeVariant(value);
        reflectAttribute(this, 'variant', this._privateVariant);
    }
    @api
    get selected() {
        return this.state.selected || false;
    }
    set selected(value) {
        this.state.selected = normalizeBoolean(value);
    }
    constructor() {
        super();
        setContext(this);
    }
    renderedCallback() {
        this.template.host.style.pointerEvents = this.disabled ? 'none' : '';
    }
    @api
    focus() {
        if (!import.meta.env.SSR) {
            this.template.querySelector('button').focus();
        }
    }
    get computedButtonClass() {
        const classes = classSet('slds-button slds-button_stateful')
            .add({
            'slds-button_neutral': this._privateVariant === 'neutral',
            'slds-button_brand': this._privateVariant === 'brand',
            'slds-button_inverse': this._privateVariant === 'inverse',
            'slds-button_destructive': this._privateVariant === 'destructive',
            'slds-button_success': this._privateVariant === 'success',
        })
            .add({
            'slds-not-selected': !this.selected,
            'slds-is-selected': this.selected && !this.state.isClicked,
            'slds-is-selected-clicked': this.selected && this.state.isClicked,
        })
            .add(buttonGroupOrderClass(this.groupOrder));
        return classes.toString();
    }
    normalizeVariant(value) {
        if (features.enableAssertContext260Round2) {
            assertContext(this);
        }
        return normalize(value, {
            fallbackValue: DEFAULT_VARIANT,
            validValues: [
                'neutral',
                'brand',
                'inverse',
                'destructive',
                'success',
                'text',
            ],
        });
    }
    get privateLabelWhenOn() {
        let outputVal = this.labelWhenOn;
        if (this.isValidLabel(outputVal)) {
            return outputVal;
        }
        outputVal = '';
        console.warn(`<lightning-button-stateful> The "labelWhenOn" attribute value is required to show the label when selected has a value of true`);
        return outputVal;
    }
    get privateLabelWhenOff() {
        let outputVal = this.labelWhenOff;
        if (this.isValidLabel(outputVal)) {
            return outputVal;
        }
        outputVal = '';
        console.warn(`<lightning-button-stateful> The "labelWhenOff" attribute value is required to show the label when selected has a value of false`);
        return outputVal;
    }
    get privateLabelWhenHover() {
        const outputVal = this.labelWhenHover;
        if (this.isValidLabel(outputVal)) {
            return outputVal;
        }
        return this.privateLabelWhenOn;
    }
    get privateIconNameWhenHover() {
        if (this.iconNameWhenHover) {
            return this.iconNameWhenHover;
        }
        return this.iconNameWhenOn;
    }
    handleButtonClick() {
        this.state.isClicked = true;
    }
    handleButtonBlur() {
        this.state.isClicked = false;
        this.dispatchEvent(new CustomEvent('blur'));
    }
    handleButtonFocus() {
        this.dispatchEvent(new CustomEvent('focus'));
    }
    get computedAriaPressed() {
        return String(this.state.selected);
    }
    isValidLabel(labelVal) {
        if (typeof labelVal !== 'string' || labelVal.length === 0) {
            return false;
        }
        return true;
    }
}
LightningButtonStateful['interopMap'] = {
    props: {
        selected: 'state',
    },
};
