// lightning-base-components/src/lightning/calendar/calendar.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import labelAriaLabelMonth from '@salesforce/label/LightningDateTimePicker.ariaLabelMonth';
import labelNextMonth from '@salesforce/label/LightningDateTimePicker.nextMonth';
import labelPreviousMonth from '@salesforce/label/LightningDateTimePicker.previousMonth';
import labelToday from '@salesforce/label/LightningDateTimePicker.today';
import labelYearSelector from '@salesforce/label/LightningDateTimePicker.yearSelector';
import { api, track } from 'lwc';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import { classSet } from 'lightning/utils';
import { generateUniqueId } from 'lightning/inputUtils';
import firstDayOfWeek from '@salesforce/i18n/firstDayOfWeek';
import { handleKeyDownOnCalendar } from './keyboard';
import salesforceLanguage from '@salesforce/i18n/lang';
import showJapaneseCalendar from '@salesforce/i18n/showJapaneseCalendar';
import { toOtherCalendar, toLocalizedDigits, getCurrentDateString, isBefore, isAfter, parseDateTime, getISODateString, getNameOfWeekdays, getMonthNames, } from 'lightning/internationalizationLibrary';
import { STANDARD_DATE_FORMAT, TIME_SEPARATOR } from 'lightning/iso8601Utils';
import { features } from 'lightning/features';
const i18n = {
    ariaLabelMonth: labelAriaLabelMonth,
    nextMonth: labelNextMonth,
    previousMonth: labelPreviousMonth,
    today: labelToday,
    yearSelector: labelYearSelector,
};
const WEEKS_PER_MONTH = 6;
const DAYS_PER_WEEK = 7;
const calendarCache = {};
const JAPANESE_CALENDAR_YEARS = [
    { key: 'R', year: 2019, label: '\u4ee4\u548c' },
    { key: 'H', year: 1989, label: '\u5e73\u6210' },
    { key: 'S', year: 1926, label: '\u662d\u548c' },
    { key: 'T', year: 1912, label: '\u5927\u6b63' },
    { key: 'M', year: 1868, label: '\u660e\u6cbb' },
];
export default class LightningCalendar extends LightningShadowBaseClass {
    @track
    calendarYear = null;
    @track
    calendarMonth = null;
    @api
    min;
    @api
    max;
    @api
    get value() {
        return this.selectedDate;
    }
    set value(newValue) {
        const dateOnlyString = typeof newValue === 'string'
            ? newValue.split(TIME_SEPARATOR)[0]
            : newValue;
        if (dateOnlyString !== this.selectedDate) {
            this.selectedDate = dateOnlyString;
            if (!this._connected) {
                return;
            }
            const newDate = this.parseDate(dateOnlyString);
            if (!newDate) {
                this.selectedDate = null;
                this.renderToday();
            }
            else {
                this.selectDate(newDate);
            }
        }
    }
    constructor() {
        super();
        this.uniqueId = generateUniqueId();
    }
    renderedCallback() {
        this.dispatchEvent(new CustomEvent('ready'));
    }
    connectedCallback() {
        super.connectedCallback();
        this._connected = true;
        this.todayDate = getCurrentDateString();
        const renderDate = this.getSelectedDate() || this.getTodaysDate();
        this.renderCalendar(renderDate);
        this.keyboardInterface = this.calendarKeyboardInterface();
    }
    disconnectedCallback() {
        this._connected = false;
    }
    @api
    focus() {
        requestAnimationFrame(() => {
            const dateElement = this.getFocusableDateCell();
            if (dateElement) {
                dateElement.focus();
            }
        });
    }
    get i18n() {
        return i18n;
    }
    get computedAriaLabel() {
        const renderedMonth = this.getCalendarDate().getMonth();
        return i18n.ariaLabelMonth + getMonthNames()[renderedMonth].fullName;
    }
    get computedMonthTitle() {
        const renderedMonth = this.getCalendarDate().getMonth();
        return getMonthNames()[renderedMonth].fullName;
    }
    get computedNextDisabled() {
        if (!features.enableCalendarMinMaxNavigation) {
            return false;
        }
        const maxDate = this.parseDate(this.max);
        const calendarDate = this.getCalendarDate();
        const nextMonth = calendarDate.getMonth() + 1;
        const currentYear = calendarDate.getFullYear();
        const nextMonthStart = new Date(currentYear, nextMonth, 1);
        return isBefore(maxDate, nextMonthStart, 'day');
    }
    get computedPrevDisabled() {
        if (!features.enableCalendarMinMaxNavigation) {
            return false;
        }
        const minDate = this.parseDate(this.min);
        const calendarDate = this.getCalendarDate();
        const currentMonth = calendarDate.getMonth();
        const currentYear = calendarDate.getFullYear();
        const prevMonthEnd = new Date(currentYear, currentMonth, 0);
        return isAfter(minDate, prevMonthEnd, 'day');
    }
    get computedWeekdayLabels() {
        const nameOfWeekdays = getNameOfWeekdays();
        const firstDay = this.getFirstDayOfWeek();
        const computedWeekdayLabels = [];
        for (let i = firstDay; i < nameOfWeekdays.length; i++) {
            computedWeekdayLabels.push(nameOfWeekdays[i]);
        }
        for (let i = 0; i < firstDay; i++) {
            computedWeekdayLabels.push(nameOfWeekdays[i]);
        }
        return computedWeekdayLabels;
    }
    get computedSelectElementId() {
        return this.uniqueId + '-select';
    }
    get computedWeekdaysElementId() {
        return this.uniqueId + '-weekdays';
    }
    get computedMonthTitleId() {
        return this.uniqueId + '-month';
    }
    get computedYearList() {
        const sampleDate = new Date();
        const currentYear = sampleDate.getFullYear();
        const minDate = this.parseDate(this.min);
        const maxDate = this.parseDate(this.max);
        const minYear = minDate ? minDate.getFullYear() : currentYear - 100;
        const maxYear = maxDate ? maxDate.getFullYear() : currentYear + 100;
        const yearList = [];
        for (let year = minYear; year <= maxYear; year++) {
            yearList.push({
                label: this.getYearDisplayValue(sampleDate, year),
                value: year,
            });
        }
        return yearList;
    }
    get monthIndex() {
        return this.getCalendarDate().getMonth();
    }
    get hostDirection() {
        let computedStyle;
        if (!import.meta.env.SSR) {
            computedStyle = window.getComputedStyle(this.template.host);
        }
        return computedStyle ? computedStyle.direction.toLowerCase() : 'ltr';
    }
    get previousMonthIconName() {
        return this.hostDirection === 'ltr' ? 'utility:left' : 'utility:right';
    }
    get nextMonthIconName() {
        return this.hostDirection === 'ltr' ? 'utility:right' : 'utility:left';
    }
    get isCalendarGridCellAriaFixEnabled() {
        return features.enableCalendarGridCellAriaFix;
    }
    get computedGridAriaLabelledby() {
        return features.enableCalendarGridCellAriaFix
            ? 'month-title year-selected'
            : null;
    }
    getJapaneseCalendarYear(year) {
        const JAPANESE = 'ja';
        const displayValue = (japaneseYearData, useKey) => {
            return useKey ? japaneseYearData.key : japaneseYearData.label;
        };
        const useShortName = salesforceLanguage !== JAPANESE;
        for (let i = 0; i < JAPANESE_CALENDAR_YEARS.length; i++) {
            if (year >= JAPANESE_CALENDAR_YEARS[i].year) {
                let jpYear = displayValue(JAPANESE_CALENDAR_YEARS[i], useShortName) +
                    (year - JAPANESE_CALENDAR_YEARS[i].year + 1);
                if (year === JAPANESE_CALENDAR_YEARS[i].year &&
                    i < JAPANESE_CALENDAR_YEARS.length - 1) {
                    const previousJapaneseYearData = JAPANESE_CALENDAR_YEARS[i + 1];
                    jpYear +=
                        '/' +
                            displayValue(previousJapaneseYearData, useShortName) +
                            (year - previousJapaneseYearData.year + 1);
                }
                return jpYear;
            }
        }
        return null;
    }
    getYearDisplayValue(date, yearValue) {
        date.setFullYear(yearValue);
        let displayValue = toLocalizedDigits(String(toOtherCalendar(date).getFullYear()));
        if (showJapaneseCalendar) {
            const jpYear = this.getJapaneseCalendarYear(yearValue);
            if (jpYear) {
                displayValue += ' (' + jpYear + ')';
            }
        }
        return displayValue;
    }
    get computedMonth() {
        if (!this._connected) {
            return [];
        }
        const selectedDate = this.getSelectedDate();
        const renderDate = this.getCalendarDate();
        const cacheKey = this.getCalendarCacheKey(renderDate, selectedDate);
        if (cacheKey in calendarCache) {
            return calendarCache[cacheKey];
        }
        const todayDate = this.getTodaysDate();
        const focusableDate = this.getInitialFocusDate(todayDate, selectedDate, renderDate);
        const calendarDates = {
            selectedDate,
            renderDate,
            focusableDate,
            todayDate,
            minDate: this.parseDate(this.min),
            maxDate: this.parseDate(this.max),
        };
        const monthCells = [];
        const date = this.getCalendarStartDate(renderDate);
        for (let week = 0; week < WEEKS_PER_MONTH; week++) {
            const weekCells = {
                id: week,
                days: [],
            };
            for (let weekday = 0; weekday < DAYS_PER_WEEK; weekday++) {
                const dayCell = this.getDateCellAttributes(date, calendarDates);
                weekCells.days.push(dayCell);
                date.setDate(date.getDate() + 1);
            }
            monthCells.push(weekCells);
        }
        calendarCache[cacheKey] = monthCells;
        return monthCells;
    }
    getDateCellAttributes(date, calendarDates) {
        const isBetweenMinAndMax = this.isBetween(date, calendarDates.minDate, calendarDates.maxDate);
        const isInAdjacentMonth = !this.dateInCalendar(date, calendarDates.renderDate) ||
            !isBetweenMinAndMax;
        const ariaDisabled = features.enableCalendarMinMaxNavigation
            ? !isBetweenMinAndMax
            : null;
        const isSelected = this.isSame(date, calendarDates.selectedDate);
        const isToday = this.isSame(date, calendarDates.todayDate);
        const ariaCurrent = isToday ? 'date' : false;
        const tabIndex = this.isSame(date, calendarDates.focusableDate)
            ? '0'
            : false;
        const className = classSet()
            .add({
            'slds-is-today': isToday,
            'slds-is-selected': isSelected,
            'slds-day_adjacent-month': isInAdjacentMonth,
        })
            .toString() || undefined;
        return {
            dayInMonth: toLocalizedDigits(String(date.getDate())),
            dateValue: this.formatDate(date),
            isSelected: isSelected ? 'true' : 'false',
            ariaDisabled,
            className,
            tabIndex,
            ariaCurrent,
        };
    }
    dispatchSelectEvent() {
        this.dispatchEvent(new CustomEvent('select', {
            composed: true,
            bubbles: true,
            cancelable: true,
            detail: { value: this.selectedDate },
        }));
    }
    dateInCalendar(date, calendarDate) {
        const renderedCalendar = calendarDate || this.getCalendarDate();
        return (date.getMonth() === renderedCalendar.getMonth() &&
            date.getFullYear() === renderedCalendar.getFullYear());
    }
    getInitialFocusDate(todayDate, selectedDate, renderedDate) {
        if (selectedDate && this.dateInCalendar(selectedDate, renderedDate)) {
            return selectedDate;
        }
        if (this.dateInCalendar(todayDate, renderedDate)) {
            return todayDate;
        }
        return new Date(renderedDate.getFullYear(), renderedDate.getMonth(), 1);
    }
    getTodaysDate() {
        if (this.todayDate) {
            return this.parseDate(this.todayDate);
        }
        return new Date();
    }
    getSelectedDate() {
        return this.parseDate(this.selectedDate);
    }
    getCalendarDate() {
        if (this.calendarYear) {
            return new Date(this.calendarYear, this.calendarMonth, 1);
        }
        return this.getTodaysDate();
    }
    getCalendarStartDate(renderedDate) {
        const firstDayOfMonth = new Date(renderedDate.getFullYear(), renderedDate.getMonth(), 1);
        return this.getStartOfWeek(firstDayOfMonth);
    }
    getStartOfWeek(dayInWeek) {
        const firstDay = this.getFirstDayOfWeek();
        let startDay = dayInWeek.getDay();
        while (startDay !== firstDay) {
            dayInWeek.setDate(dayInWeek.getDate() - 1);
            startDay = dayInWeek.getDay();
        }
        return dayInWeek;
    }
    getFirstDayOfWeek() {
        return firstDayOfWeek - 1;
    }
    selectDate(newDate) {
        if (this.dateInCalendar(newDate)) {
            const dateElement = this.getElementByDate(this.formatDate(newDate));
            this.selectDateInCalendar(dateElement);
        }
        else {
            this.renderCalendar(newDate);
        }
    }
    selectDateInCalendar(dateElement) {
        const tdElement = dateElement.parentElement;
        this.selectedDate = tdElement.getAttribute('data-value');
        this.removeCurrentlySelectedDateAttributes();
        this.addSelectedDateAttributes(dateElement);
    }
    selectDateInCalendarAndDispatchSelect(dateElement) {
        this.selectDateInCalendar(dateElement);
        this.dispatchSelectEvent();
    }
    getElementByDate(dateString) {
        const tdElement = this.template.querySelector(`td[data-value='${dateString}']`);
        return tdElement
            ? tdElement.querySelector(`span[role="button"]`)
            : null;
    }
    getFocusableDateCell() {
        return this.template.querySelector(`span[role="button"][tabIndex="0"]`);
    }
    unfocusDateCell(element) {
        if (element) {
            element.removeAttribute('tabIndex');
        }
    }
    focusDateCell(element) {
        if (element) {
            element.setAttribute('tabIndex', 0);
            element.focus();
        }
    }
    focusElementByDate(date) {
        requestAnimationFrame(() => {
            const element = this.getElementByDate(this.formatDate(date));
            if (element) {
                this.unfocusDateCell(this.getFocusableDateCell());
                this.focusDateCell(element);
            }
        });
    }
    renderCalendar(newDate) {
        this.calendarMonth = newDate.getMonth();
        this.calendarYear = newDate.getFullYear();
    }
    renderToday() {
        const todaysDate = this.getTodaysDate();
        if (this.dateInCalendar(todaysDate)) {
            this.removeCurrentlySelectedDateAttributes();
            this.unfocusDateCell(this.getFocusableDateCell());
            const todayElement = this.getElementByDate(this.todayDate);
            this.focusDateCell(todayElement);
        }
        else {
            this.renderCalendar(todaysDate);
        }
    }
    removeCurrentlySelectedDateAttributes() {
        const currentlySelectedElement = this.template.querySelector(`td[class*='slds-is-selected'] span[role="button"]`);
        if (currentlySelectedElement) {
            currentlySelectedElement.parentElement.classList.remove('slds-is-selected');
            currentlySelectedElement.parentElement.setAttribute('aria-selected', 'false');
        }
        this.unfocusDateCell(this.getFocusableDateCell());
    }
    addSelectedDateAttributes(dateElement) {
        this.focusDateCell(dateElement);
        dateElement.parentElement.classList.add('slds-is-selected');
        dateElement.parentElement.setAttribute('aria-selected', 'true');
    }
    handleCalendarKeyDown(event) {
        const dateString = event.target.parentElement.getAttribute('data-value');
        handleKeyDownOnCalendar(event, this.parseDate(dateString), this.keyboardInterface);
    }
    handleDateClick(event) {
        event.stopPropagation();
        this.selectDateInCalendarAndDispatchSelect(event.target);
    }
    handleTodayClick(event) {
        event.stopPropagation();
        this.selectedDate = this.todayDate;
        this.selectDate(this.getTodaysDate());
        this.dispatchSelectEvent();
    }
    handleYearSelectClick(event) {
        event.stopPropagation();
    }
    handleYearChange(event) {
        event.stopPropagation();
        const newYearValue = event.target.value;
        if (this.calendarYear !== newYearValue) {
            this.calendarYear = newYearValue;
        }
    }
    goToNextMonth(event) {
        event.stopPropagation();
        const calendarDate = this.getCalendarDate();
        calendarDate.setMonth(calendarDate.getMonth() + 1);
        this.renderCalendar(calendarDate);
        if (this.computedNextDisabled && this.refs.calendar) {
            this.refs.calendar.focus();
        }
    }
    goToPreviousMonth(event) {
        event.stopPropagation();
        const calendarDate = this.getCalendarDate();
        calendarDate.setMonth(calendarDate.getMonth() - 1);
        this.renderCalendar(calendarDate);
        if (this.computedPrevDisabled && this.refs.calendar) {
            this.refs.calendar.focus();
        }
    }
    calendarKeyboardInterface() {
        const that = this;
        return {
            focusDate(newDate) {
                if (!that.dateInCalendar(newDate)) {
                    if (features.enableCalendarMinMaxNavigation) {
                        const calendarDate = that.getCalendarDate();
                        const isBeforeCurrentDate = isBefore(newDate, calendarDate, 'day');
                        const isAfterCurrentDate = isAfter(newDate, calendarDate, 'day');
                        if ((isBeforeCurrentDate &&
                            that.computedPrevDisabled) ||
                            (isAfterCurrentDate && that.computedNextDisabled)) {
                            return;
                        }
                    }
                    that.renderCalendar(newDate);
                }
                that.focusElementByDate(newDate);
            },
            getStartOfWeek(dayInWeek) {
                return that.getStartOfWeek(dayInWeek);
            },
            selectDate(dateElement) {
                that.selectDateInCalendarAndDispatchSelect(dateElement);
            },
        };
    }
    formatDate(date) {
        return getISODateString(date);
    }
    parseDate(dateString) {
        return parseDateTime(dateString, STANDARD_DATE_FORMAT, true);
    }
    isSame(date1, date2) {
        if (!date1 || !date2) {
            return false;
        }
        return (date1.getFullYear() === date2.getFullYear() &&
            date1.getMonth() === date2.getMonth() &&
            date1.getDate() === date2.getDate());
    }
    isBetween(date, date1, date2) {
        let isBeforeEndDate = true;
        let isAfterStartDate = true;
        if (date2) {
            isBeforeEndDate =
                isBefore(date, date2, 'day') || this.isSame(date, date2);
        }
        if (date1) {
            isAfterStartDate =
                isAfter(date, date1, 'day') || this.isSame(date, date1);
        }
        return isBeforeEndDate && isAfterStartDate;
    }
    getCalendarCacheKey(renderDate, selectedDate) {
        let key = renderDate.getFullYear() + '-' + renderDate.getMonth();
        key += this.min ? 'min' + this.min : '';
        key += this.max ? 'max' + this.max : '';
        key += features.enableCalendarMinMaxNavigation
            ? 'enableCalendarMinMaxNavigation'
            : '';
        if (selectedDate && this.dateInCalendar(selectedDate, renderDate)) {
            key += '_' + selectedDate.getDate();
        }
        return key;
    }
}
