// lightning-base-components/src/lightning/configProvider/defaultConfig.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import localizationService from './defaultLocalizationConfig';
function getConfigFromAura($A) {
    return {
        getLocalizationService() {
            return $A.localizationService;
        },
        getPathPrefix() {
            return $A.getContext().getPathPrefix();
        },
        getToken(name) {
            return $A.getToken(name);
        },
    };
}
function createStandAloneConfig() {
    return {
        getLocalizationService() {
            return localizationService;
        },
        getPathPrefix() {
            return '';
        },
        getToken() {
            return undefined;
        },
        getOneConfig() {
            return {
                densitySetting: '',
            };
        },
    };
}
export function getDefaultConfig() {
    if (!import.meta.env.SSR &&
        window.$A !== undefined &&
        window.$A.localizationService) {
        return getConfigFromAura(window.$A);
    }
    return createStandAloneConfig();
}
