// lightning-base-components/src/lightning/modalFooter/modalFooter.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import { classSet } from 'lightning/utils';
import { modalFocusinEventHandler } from 'lightning/modalUtils';
import orgHasWCAGResizeReflowSC from '@salesforce/accessCheck/ComponentAccessibility.orgHasWCAGResizeReflowSC';
const footerClass = 'slds-modal__footer';
const hideClass = 'slds-hide';
const footerSelector = `.${footerClass}`;
const footerSlotSelector = '[data-footer-slot]';
export default class LightningModalFooter extends LightningShadowBaseClass {
    initialRender = true;
    initialSlotRender = true;
    hideFooter = false;
    unregisterCallback = null;
    handleDefaultSlotChange() {
        if (this.initialSlotRender) {
            this.initialSlotRender = false;
        }
        this.registerWithParent();
        this.hideFooter = !this.isDefaultSlotPopulated;
    }
    get footerCssClasses() {
        const classes = classSet(footerClass);
        classes.add({
            [hideClass]: this.hideFooter,
        });
        classes.add({
            'slds-modal__footer__reflow_resize': !!orgHasWCAGResizeReflowSC,
        });
        return classes.toString();
    }
    get footerHeight() {
        const divElem = this.template.querySelector(footerSelector);
        const footerRect = divElem ? divElem.getBoundingClientRect() : {};
        const { height } = footerRect;
        return height || 0;
    }
    get defaultSlotElement() {
        return this.template.querySelector(footerSlotSelector);
    }
    get isDefaultSlotPopulated() {
        const slotElement = this.defaultSlotElement;
        if (slotElement && slotElement.assignedNodes) {
            return slotElement.assignedNodes().length > 0;
        }
        return true;
    }
    registerWithParent() {
        const evtRegister = new CustomEvent('privatemodalfooterregister', {
            bubbles: true,
            composed: true,
            detail: {
                footerHeight: this.footerHeight,
                defaultSlotIsPopulated: this.isDefaultSlotPopulated,
                defaultSlotHasRendered: !this.initialSlotRender,
                unRegisterCallback: (unregisterCallback) => {
                    this.unregisterCallback = unregisterCallback;
                },
            },
        });
        this.dispatchEvent(evtRegister);
    }
    handleModalFooterFocusin = modalFocusinEventHandler.bind(this);
    initState() {
        this.initialRender = true;
        this.initialSlotRender = true;
        this.hideFooter = false;
        this.unregisterCallback = null;
    }
    connectedCallback() {
        super.connectedCallback();
        this.initState();
    }
    disconnectedCallback() {
        if (this.unregisterCallback) {
            this.unregisterCallback();
        }
    }
    renderedCallback() {
        if (this.initialRender) {
            this.registerWithParent();
            this.initialRender = false;
        }
        this.hideFooter = !this.isDefaultSlotPopulated;
    }
}
