// lightning-base-components/src/lightning/datatable/autoWidthStrategy.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { clamp } from './utils';
import { getColumnWidth, getTotalWidthsData } from './widthManagerShared';
const MIN_MAX_THRESHOLD = 0.5;
const TRUNCATION_ALLOWANCE = 20;
function getTotalColumnWidth(columnWidths) {
    let total = 0;
    for (let i = 0, { length } = columnWidths; i < length; i += 1) {
        total += columnWidths[i];
    }
    return total;
}
function getReducedWidth(currentWidth, widthToRemove, minColumnWidth) {
    return Math.max(currentWidth - widthToRemove, minColumnWidth);
}
function getEnlargedWidth(currentWidth, widthToAdd, maxColumnWidth) {
    return Math.min(currentWidth + widthToAdd, maxColumnWidth);
}
function getExpectedTableWidth(availableWidth, widthsData) {
    const minTotalFlexibleWidth = widthsData.totalFlexibleColumns * widthsData.minColumnWidth;
    const minExpectedTableWidth = minTotalFlexibleWidth +
        widthsData.totalFixedWidth +
        widthsData.totalResizedWidth;
    return widthsData.totalFlexibleColumns === 0
        ? minExpectedTableWidth
        : Math.max(minExpectedTableWidth, availableWidth);
}
export class AutoWidthStrategy {
    columnWidthRatios = [];
    columnWidthData = {};
    columnWidthsDistribution = {};
    constructor(minColumnWidth, maxColumnWidth, wrapTextMaxLines = 3) {
        this.columnWidthData = {
            minColumnWidth,
            maxColumnWidth,
            wrapTextMaxLines,
        };
        this.columnWidthsDistribution.colsWithMinWidth = [];
        this.columnWidthsDistribution.colsWithMaxWidth = [];
    }
    set minColumnWidth(value) {
        this.columnWidthData.minColumnWidth = value;
    }
    set maxColumnWidth(value) {
        this.columnWidthData.maxColumnWidth = value;
    }
    set wrapTextMaxLines(value) {
        this.columnWidthData.wrapTextMaxLines = value || 3;
    }
    getAdjustedColumnWidths(datatableInterface, columns, recomputeAutoWidthRatios) {
        const widthsData = getTotalWidthsData(this.columnWidthData, columns);
        const availableWidth = datatableInterface.getAvailableWidthFromDom();
        const expectedTableWidth = getExpectedTableWidth(availableWidth, widthsData, columns);
        this.resetColumnWidthsDistribution();
        if (recomputeAutoWidthRatios || this.columnWidthRatios.length === 0) {
            this.calculateColumnWidthRatios(datatableInterface, columns, widthsData);
        }
        if (recomputeAutoWidthRatios &&
            this.columnWidthRatios.length !== columns.length) {
            return { columnWidths: [] };
        }
        const columnWidths = this.distributeWidthFromRatios(expectedTableWidth, columns, widthsData);
        const columnWidthsSum = getTotalColumnWidth(columnWidths);
        if (expectedTableWidth > columnWidthsSum) {
            this.redistributeExtraWidth(expectedTableWidth, columnWidths, columns);
        }
        else if (expectedTableWidth < columnWidthsSum) {
            this.redistributeDeficitWidth(expectedTableWidth, columnWidths, columns);
        }
        return { columnWidths };
    }
    getRatios(cellWidths, totalWidth) {
        const { length } = cellWidths;
        const ratios = Array(length);
        for (let i = 0; i < length; i += 1) {
            ratios[i] = (100 * cellWidths[i]) / totalWidth;
        }
        return ratios;
    }
    calculateColumnWidthRatios(datatableInterface, columns, widthsData) {
        const { wrapTextMaxLines } = widthsData;
        const dataCellWidths = datatableInterface.getDataCellWidths();
        for (let colIndex = 0, { length } = dataCellWidths; colIndex < length; colIndex += 1) {
            const col = columns[colIndex];
            if (col) {
                if (col.wrapText) {
                    const width = dataCellWidths[colIndex];
                    dataCellWidths[colIndex] = width / wrapTextMaxLines;
                }
                else {
                    const { fixedWidth } = col;
                    if (fixedWidth && isFinite(fixedWidth)) {
                        dataCellWidths[colIndex] = fixedWidth;
                    }
                }
            }
        }
        this.setColumnWidthRatios(datatableInterface, dataCellWidths, widthsData);
    }
    setColumnWidthRatios(datatableInterface, dataCellWidths, widthsData) {
        this.columnWidthRatios = [];
        const tableScrollWidth = datatableInterface.getTableElementWidth();
        if (tableScrollWidth > 0) {
            const { totalFixedWidth, totalResizedWidth } = widthsData;
            if (dataCellWidths.length === 0) {
                const headerCellWidths = datatableInterface.getHeaderCellWidths();
                if (headerCellWidths.length > 0) {
                    const totalHeaderWidth = headerCellWidths.reduce((total, width) => {
                        total += width;
                        return total;
                    }, 0);
                    const totalFlexibleWidth = totalHeaderWidth - totalFixedWidth - totalResizedWidth;
                    this.columnWidthRatios = this.getRatios(headerCellWidths, totalFlexibleWidth);
                }
            }
            else {
                const totalCellWidth = dataCellWidths.reduce((total, width) => {
                    total += width;
                    return total;
                }, 0);
                const totalFlexibleWidth = Math.min(tableScrollWidth, totalCellWidth) -
                    totalFixedWidth -
                    totalResizedWidth;
                this.columnWidthRatios = this.getRatios(dataCellWidths, totalFlexibleWidth);
            }
        }
    }
    resetColumnWidthsDistribution() {
        this.columnWidthsDistribution.colsWithMinWidth = [];
        this.columnWidthsDistribution.colsWithMaxWidth = [];
    }
    distributeWidthFromRatios(availableWidth, columns, widthsData) {
        const columnWidths = [];
        for (let colIndex = 0; colIndex < columns.length; colIndex += 1) {
            const col = columns[colIndex];
            const width = getColumnWidth(col) ||
                this.getColumnWidthFromRatio(availableWidth, widthsData, colIndex);
            columnWidths[colIndex] = width;
        }
        return columnWidths;
    }
    getColumnWidthFromRatio(availableWidth, widthsData, colIndex) {
        const { totalFixedWidth, totalResizedWidth, minColumnWidth, maxColumnWidth, } = widthsData;
        const widthRatio = this.columnWidthRatios[colIndex];
        const totalFlexibleWidth = availableWidth - totalFixedWidth - totalResizedWidth;
        const calculatedWidth = Math.floor((totalFlexibleWidth * widthRatio) / 100) +
            TRUNCATION_ALLOWANCE;
        const { colsWithMinWidth, colsWithMaxWidth } = this.columnWidthsDistribution;
        const minWidthThreshold = minColumnWidth + Math.ceil(MIN_MAX_THRESHOLD * minColumnWidth);
        if (calculatedWidth < minWidthThreshold) {
            colsWithMinWidth.push(colIndex);
        }
        const maxWidthThreshold = maxColumnWidth - Math.ceil(MIN_MAX_THRESHOLD * maxColumnWidth);
        if (calculatedWidth > maxWidthThreshold) {
            colsWithMaxWidth.push(colIndex);
        }
        return clamp(calculatedWidth, minColumnWidth, maxColumnWidth);
    }
    redistributeExtraWidth(expectedTableWidth, columnWidths, columns) {
        const { colsWithMinWidth } = this.columnWidthsDistribution;
        const { length: colsWithMinWidthLength } = colsWithMinWidth;
        const widthsData = getTotalWidthsData(this.columnWidthData, columns);
        const { maxColumnWidth, totalResizedColumns, totalFixedColumns } = widthsData;
        let columnWidthsSum = getTotalColumnWidth(columnWidths);
        let extraSpace = 0;
        let extraWidthPerColumn = 0;
        let totalColsToDistribute = 0;
        if (colsWithMinWidthLength > 0) {
            extraSpace = expectedTableWidth - columnWidthsSum;
            totalColsToDistribute = colsWithMinWidthLength;
            extraWidthPerColumn = Math.floor(extraSpace / totalColsToDistribute);
            for (let i = 0; i < colsWithMinWidthLength; i += 1) {
                const colIndex = colsWithMinWidth[i];
                const currentWidth = columnWidths[colIndex];
                const newWidth = getEnlargedWidth(currentWidth, extraWidthPerColumn, maxColumnWidth);
                columnWidthsSum += newWidth - currentWidth;
                columnWidths[colIndex] = newWidth;
            }
        }
        extraSpace = expectedTableWidth - columnWidthsSum;
        if (extraSpace > 0) {
            const totalFixedWidthColumns = totalResizedColumns + totalFixedColumns;
            const { length: colCount } = columns;
            totalColsToDistribute = colCount - totalFixedWidthColumns;
            extraWidthPerColumn = Math.floor(extraSpace / totalColsToDistribute);
            for (let colIndex = 0; colIndex < colCount; colIndex += 1) {
                const col = columns[colIndex];
                const currentWidth = columnWidths[colIndex];
                if (!getColumnWidth(col)) {
                    const newWidth = getEnlargedWidth(currentWidth, extraWidthPerColumn, maxColumnWidth);
                    columnWidthsSum += newWidth - currentWidth;
                    columnWidths[colIndex] = newWidth;
                }
            }
        }
    }
    redistributeDeficitWidth(expectedTableWidth, columnWidths, columns) {
        const { colsWithMaxWidth } = this.columnWidthsDistribution;
        const { length: colsWithMaxWidthLength } = colsWithMaxWidth;
        const widthsData = getTotalWidthsData(this.columnWidthData, columns);
        const { minColumnWidth, totalResizedColumns, totalFixedColumns } = widthsData;
        let columnWidthsSum = getTotalColumnWidth(columnWidths);
        let extraSpace = expectedTableWidth - columnWidthsSum;
        let totalColsToDistribute = 0;
        let extraWidthPerColumn = 0;
        if (colsWithMaxWidthLength > 0) {
            totalColsToDistribute = colsWithMaxWidthLength;
            extraWidthPerColumn = Math.floor(Math.abs(extraSpace) / totalColsToDistribute);
            for (let i = 0, { length } = colsWithMaxWidth; i < length; i += 1) {
                const colIndex = colsWithMaxWidth[i];
                const currentWidth = columnWidths[colIndex];
                const newWidth = getReducedWidth(currentWidth, extraWidthPerColumn, minColumnWidth);
                columnWidthsSum -= currentWidth - newWidth;
                columnWidths[colIndex] = newWidth;
            }
        }
        extraSpace = expectedTableWidth - columnWidthsSum;
        const totalFixedWidthColumns = totalResizedColumns + totalFixedColumns;
        if (extraSpace < 0) {
            const { colsWithMinWidth } = this.columnWidthsDistribution;
            totalColsToDistribute =
                columns.length -
                    (totalFixedWidthColumns + colsWithMinWidth.length);
            extraWidthPerColumn = Math.ceil(Math.abs(extraSpace) / totalColsToDistribute);
            for (let colIndex = 0; colIndex < columns.length; colIndex += 1) {
                const col = columns[colIndex];
                const currentWidth = columnWidths[colIndex];
                if (!colsWithMinWidth.includes(colIndex) &&
                    !getColumnWidth(col)) {
                    const newWidth = getReducedWidth(currentWidth, extraWidthPerColumn, minColumnWidth);
                    columnWidthsSum -= currentWidth - newWidth;
                    columnWidths[colIndex] = newWidth;
                }
            }
        }
    }
}
