// lightning-base-components/src/lightning/datatable/columnResizer.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { isRTL, normalizeBoolean } from 'lightning/utilsPrivate';
import { clamp, getScrollerY, normalizeNumberAttribute } from './utils';
export function getResizerDefaultState() {
    return {
        columnWidths: [],
        columnWidthsMode: 'fixed',
        maxColumnWidth: 1000,
        minColumnWidth: 50,
        resizeColumnDisabled: false,
        resizeStep: 10,
        tableWidth: 0,
    };
}
export const RESIZER_DEFAULT_STATE = getResizerDefaultState();
export function setResizeColumnDisabled(widthsData, value) {
    widthsData.resizeColumnDisabled = normalizeBoolean(value);
}
export function setResizeStep(widthsData, value) {
    widthsData.resizeStep = normalizeNumberAttribute('resizeStep', value, 'non-negative', RESIZER_DEFAULT_STATE.resizeStep);
}
export function setMinColumnWidth(columns, widthsData, value) {
    widthsData.minColumnWidth = normalizeNumberAttribute('minColumnWidth', value, 'non-negative', RESIZER_DEFAULT_STATE.minColumnWidth);
    updateColumnWidthsMetadata(columns, widthsData);
}
export function setMaxColumnWidth(columns, widthsData, value) {
    widthsData.maxColumnWidth = normalizeNumberAttribute('maxColumnWidth', value, 'non-negative', RESIZER_DEFAULT_STATE.maxColumnWidth);
    updateColumnWidthsMetadata(columns, widthsData);
}
export function resizeColumn(columns, widthsData, colIndex, width) {
    const col = columns[colIndex];
    const { columnWidths } = widthsData;
    const currentWidth = columnWidths[colIndex];
    const { minWidth, maxWidth } = col;
    const newWidth = clamp(width, minWidth, maxWidth);
    if (currentWidth !== newWidth) {
        const newDelta = newWidth - currentWidth;
        columnWidths[colIndex] = newWidth;
        widthsData.tableWidth += newDelta;
        col.columnWidth = newWidth;
        col.style = newWidth ? `width: ${newWidth}px;` : '';
        if (isRTL()) {
            for (let i = colIndex + 1; i < columns.length; i += 1) {
                columns[i].offset += newDelta;
            }
        }
        col.isResized = true;
    }
}
export function resizeColumnWithDelta(columns, widthsData, colIndex, delta) {
    const currentWidth = widthsData.columnWidths[colIndex];
    resizeColumn(columns, widthsData, colIndex, currentWidth + delta);
}
export function updateColumnWidthsMetadata(columns, widthsData) {
    const { maxColumnWidth, minColumnWidth } = widthsData;
    for (let colIndex = 0; colIndex < columns.length; colIndex += 1) {
        const col = columns[colIndex];
        if (!col.internal) {
            col.maxWidth = maxColumnWidth;
            col.minWidth = minColumnWidth;
        }
        const { initialWidth } = col;
        if (initialWidth) {
            const { minWidth: min, maxWidth: max } = col;
            col.initialWidth = clamp(initialWidth, min, max);
        }
    }
}
export function getCustomerColumnWidths(columns, widthsData) {
    const widths = [];
    const { columnWidths } = widthsData;
    for (let colIndex = 0; colIndex < columns.length; colIndex += 1) {
        if (columns[colIndex].internal !== true) {
            widths.push(columnWidths[colIndex]);
        }
    }
    return widths;
}
export function isTableRenderedVisible(template, refs) {
    const scrollerY = getScrollerY(template, refs);
    return (scrollerY &&
        !!(scrollerY.offsetParent ||
            scrollerY.offsetHeight ||
            scrollerY.offsetWidth));
}
