// lightning-base-components/src/lightning/datatable/columnWidthManager.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { RESIZER_DEFAULT_STATE, isTableRenderedVisible } from './columnResizer';
import { AutoWidthStrategy } from './autoWidthStrategy';
import { FixedWidthStrategy } from './fixedWidthStrategy';
import { getScrollerX } from './utils';
import { getRowNumberColumnIndex, getAdjustedRowNumberColumnWidth, } from './rowNumber';
import { getColumnWidth } from './widthManagerShared';
import { isRTL } from 'lightning/utilsPrivate';
import { features } from 'lightning/features';
const AUTO_WIDTH_MODE = 'auto';
const FIXED_WIDTH_MODE = 'fixed';
export function updateColumnWidths(columns, widthsData, adjustedWidths) {
    const { columnWidths } = adjustedWidths;
    const { length: colCount } = columns;
    if (columnWidths.length !== colCount) {
        return;
    }
    let columnWidthsSum = 0;
    const rtl = isRTL();
    for (let colIndex = 0; colIndex < colCount; colIndex += 1) {
        const col = columns[colIndex];
        const newWidth = columnWidths[colIndex];
        widthsData.columnWidths[colIndex] = newWidth;
        col.columnWidth = newWidth;
        col.style = newWidth ? `width: ${newWidth}px;` : '';
        if (rtl) {
            col.offset = columnWidthsSum;
        }
        columnWidthsSum += newWidth;
    }
    widthsData.tableWidth = columnWidthsSum;
}
export class ColumnWidthManager {
    columnWidthMode = FIXED_WIDTH_MODE;
    widthStrategies;
    isResizingUpdateQueued = false;
    isAutoResizingUpdateQueued = false;
    constructor(widthsData) {
        const minColumnWidth = widthsData.minColumnWidth || RESIZER_DEFAULT_STATE.minColumnWidth;
        const maxColumnWidth = widthsData.maxColumnWidth || RESIZER_DEFAULT_STATE.maxColumnWidth;
        const fixedWidthStrategy = new FixedWidthStrategy(minColumnWidth, maxColumnWidth);
        const autoWidthStrategy = new AutoWidthStrategy(minColumnWidth, maxColumnWidth);
        this.widthStrategies = {
            [FIXED_WIDTH_MODE]: fixedWidthStrategy,
            [AUTO_WIDTH_MODE]: autoWidthStrategy,
        };
    }
    get columnWidthStrategy() {
        return this.widthStrategies[this.columnWidthMode];
    }
    set minColumnWidth(value) {
        const { widthStrategies } = this;
        const keys = Object.keys(widthStrategies);
        for (let i = 0, { length } = keys; i < length; i += 1) {
            const strategy = keys[i];
            widthStrategies[strategy].minColumnWidth = value;
        }
    }
    set maxColumnWidth(value) {
        const { widthStrategies } = this;
        const keys = Object.keys(widthStrategies);
        for (let i = 0, { length } = keys; i < length; i += 1) {
            const strategy = keys[i];
            widthStrategies[strategy].maxColumnWidth = value;
        }
    }
    set wrapTextMaxLines(value) {
        this.widthStrategies[AUTO_WIDTH_MODE].wrapTextMaxLines = value;
    }
    shouldFireResizeEvent(previousWidthsData, columns) {
        return (this.columnWidthMode === FIXED_WIDTH_MODE &&
            previousWidthsData.columnWidths.length !== columns.length);
    }
    adjustColumnsSize(template, refs, columns, widthsData) {
        const { columnWidthStrategy, isAutoResizingUpdateQueued, isResizingUpdateQueued, } = this;
        if (isResizingUpdateQueued) {
            let adjustedWidths;
            if (this.shouldResizeWithUpdate(template, refs, widthsData)) {
                adjustedWidths = columnWidthStrategy.getAdjustedColumnWidths(this.getDatatableInterface(template, refs), columns, isAutoResizingUpdateQueued);
            }
            else {
                adjustedWidths = {
                    columnWidths: widthsData.columnWidths,
                };
            }
            updateColumnWidths(columns, widthsData, adjustedWidths);
        }
        this.isAutoResizingUpdateQueued = false;
        this.isResizingUpdateQueued = false;
    }
    adjustColumnsSizeAfterResize(dt, columns, widthsData) {
        const adjustedWidths = this.columnWidthStrategy.getAdjustedColumnWidths(this.getDatatableInterface(dt.template, dt.refs), columns);
        updateColumnWidths(columns, widthsData, adjustedWidths);
    }
    handleDataChange(columns) {
        if (this.columnWidthMode === AUTO_WIDTH_MODE) {
            this.isResizingUpdateQueued = true;
            this.setAutoResizingUpdate(columns);
        }
    }
    handleColumnsChange(columns) {
        if (columns.length > 0) {
            this.isResizingUpdateQueued = true;
            this.setAutoResizingUpdate(columns);
        }
    }
    handleWidthModeChange(columns) {
        if (columns.length > 0) {
            this.isResizingUpdateQueued = true;
            this.setAutoResizingUpdate(columns);
        }
    }
    handleRowNumberOffsetChange(state, widthsData) {
        const colIndex = getRowNumberColumnIndex(state);
        if (colIndex !== -1) {
            const { columns } = state;
            const col = columns[colIndex];
            const newWidth = getAdjustedRowNumberColumnWidth(state);
            if (col.initialWidth !== newWidth) {
                col.initialWidth = Math.max(newWidth, col.minWidth);
                if (widthsData.columnWidths.length > 0) {
                    this.isResizingUpdateQueued = true;
                }
            }
        }
    }
    handleCheckboxColumnChange(previousValue, newValue, columns) {
        if (columns.length > 0 && previousValue !== newValue) {
            this.isResizingUpdateQueued = true;
        }
    }
    handleActionsColumnChange(previousValue, newValue, columns) {
        if (columns.length > 0 && previousValue !== newValue) {
            this.isResizingUpdateQueued = true;
        }
    }
    handleRowNumberColumnChange(previousValue, newValue, columns) {
        if (columns.length > 0 && previousValue !== newValue) {
            this.isResizingUpdateQueued = true;
            this.setAutoResizingUpdate(columns);
        }
    }
    setAutoResizingUpdate(columns) {
        if (this.columnWidthMode === AUTO_WIDTH_MODE) {
            this.isAutoResizingUpdateQueued = true;
        }
        for (let colIndex = 0; colIndex < columns.length; colIndex += 1) {
            const col = columns[colIndex];
            if (!getColumnWidth(col)) {
                col.columnWidth = null;
                col.style = '';
            }
        }
    }
    hasDataChanged(previousData, newData) {
        const { length: previousDataLength } = previousData;
        const { length: newDataLength } = newData;
        return (newDataLength > 0 &&
            (previousDataLength === 0 || previousDataLength !== newDataLength));
    }
    shouldResizeWithUpdate(template, refs, widthsData) {
        if (widthsData.columnWidths.length > 0) {
            return isTableRenderedVisible(template, refs);
        }
        return true;
    }
    getDatatableInterface(template, refs) {
        return {
            getAvailableWidthFromDom() {
                const scrollerX = getScrollerX(template, refs);
                if (features.enableDatatableUseClientWidth) {
                    return scrollerX.clientWidth;
                }
                return scrollerX.offsetWidth;
            },
            getDataCellWidths() {
                const firstRow = refs.tableBody.firstChild;
                if (firstRow) {
                    const cellElements = firstRow.children;
                    const { length } = cellElements;
                    const result = Array(length);
                    for (let i = 0; i < length; i += 1) {
                        result[i] = cellElements[i].offsetWidth;
                    }
                    return result;
                }
                return [];
            },
            getHeaderCellWidths() {
                const headerCells = refs.headerRow.children;
                const result = [];
                for (let i = 0; i < headerCells.length; i += 1) {
                    const headerElement = headerCells[i]
                        .firstChild;
                    const headerDomWidth = headerElement.getDomWidth();
                    if (headerDomWidth) {
                        result.push(headerDomWidth);
                    }
                }
                return result;
            },
            getTableElementWidth() {
                const elem = (refs?.tableElement ||
                    template.querySelector('.slds-table'));
                return elem.offsetWidth;
            },
        };
    }
}
