// lightning-base-components/src/lightning/datatable/columns.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import labelRowNumber from '@salesforce/label/LightningDatatable.rowNumber';
import rowActionsDefaultAriaLabel from '@salesforce/label/LightningDatatable.rowActionsDefaultAriaLabel';
import { normalizeBoolean } from 'lightning/utilsPrivate';
import { isObjectLike } from './utils';
import { getCellByKeys, HEADER_ROW_KEY } from './indexes';
import { ROW_NUMBER_INITIAL_WIDTH } from './rowNumber';
import { SELECTABLE_HEADER_TYPE, SORTING_MENU_HEADER_TYPE, } from './rowSelectionShared';
import { features } from 'lightning/features';
const i18n = {
    rowActionsDefaultAriaLabel,
    rowNumber: labelRowNumber,
};
const ROW_NUMBER_COLUMN = {
    type: 'rowNumber',
    ariaLabel: i18n.rowNumber,
    initialWidth: ROW_NUMBER_INITIAL_WIDTH,
    internal: true,
    minWidth: 52,
    maxWidth: 1000,
    resizable: false,
    sortable: false,
    tabIndex: -1,
};
const SELECTABLE_COLUMN = {
    type: SELECTABLE_HEADER_TYPE,
    fixedWidth: 32,
    internal: true,
    tabIndex: -1,
};
const SORTING_MENU_COLUMN = {
    type: SORTING_MENU_HEADER_TYPE,
    fixedWidth: 32,
    internal: true,
    tabIndex: -1,
    resizable: false,
};
export function setColumns(state, rawColumns, types) {
    state.hadTreeDataTypePreviously = state.treeColumn !== undefined;
    state.treeColumn = undefined;
    const { columns: oldColumns, indexes } = state;
    for (let i = 0, { length } = oldColumns; i < length; i += 1) {
        const cellKeyValue = `${HEADER_ROW_KEY}-${oldColumns[i].colKeyValue}`;
        indexes[cellKeyValue] = undefined;
    }
    const { length: colCount } = rawColumns;
    if (colCount === 0) {
        state.columns = [];
        state.headerIndexes = {};
        return;
    }
    let lastActionColumnIndex = -1;
    let normColCount = colCount;
    let { showRowNumberColumn } = state;
    const { showActionsMenu } = state;
    const notHideCheckboxColumn = !state.hideCheckboxColumn;
    if (!showRowNumberColumn) {
        for (let colIndex = 0; colIndex < colCount; colIndex += 1) {
            if (rawColumns[colIndex].editable) {
                showRowNumberColumn = true;
                state.showRowNumberColumn = true;
                break;
            }
        }
    }
    if (showRowNumberColumn) {
        normColCount += 1;
    }
    if (notHideCheckboxColumn) {
        normColCount += 1;
    }
    rawColumns.forEach((column, index) => {
        if (column.type === 'action') {
            lastActionColumnIndex = index;
        }
    });
    const isActionColumnInEnd = lastActionColumnIndex === rawColumns.length - 1;
    if (showActionsMenu && !isActionColumnInEnd) {
        normColCount += 1;
    }
    const normalizedColumns = Array(normColCount);
    const headerIndexes = {};
    state.columns = normalizedColumns;
    state.headerIndexes = headerIndexes;
    let firstColumnForReaders = 0;
    if (showRowNumberColumn) {
        normalizedColumns[firstColumnForReaders] = Object.assign({}, ROW_NUMBER_COLUMN);
        firstColumnForReaders += 1;
    }
    if (notHideCheckboxColumn) {
        normalizedColumns[firstColumnForReaders] = Object.assign({}, SELECTABLE_COLUMN);
        firstColumnForReaders += 1;
    }
    if (showActionsMenu && !isActionColumnInEnd) {
        normalizedColumns[normColCount - 1] = Object.assign({}, SORTING_MENU_COLUMN);
    }
    for (let colIndex = 0; colIndex < colCount; colIndex += 1) {
        normalizedColumns[colIndex + firstColumnForReaders] =
            rawColumns[colIndex];
    }
    const { defaultSortDirection, sortedBy, sortedDirection } = state;
    const isSortedByArray = Array.isArray(sortedBy);
    const seenColumnKeys = new Set();
    let hasHighlightStyling = false;
    for (let colIndex = 0; colIndex < normColCount; colIndex += 1) {
        const refCol = normalizedColumns[colIndex];
        const { columnKey } = refCol;
        if (columnKey && seenColumnKeys.has(columnKey)) {
            console.error(`The "columnKey" column property must be unique. Found a duplicate of columnKey "${columnKey}".`);
        }
        seenColumnKeys.add(columnKey);
        const { type: refColumnType } = refCol;
        const colName = columnKey || refCol.fieldName;
        const normCol = {
            ariaLabel: '',
            cellAttributes: {},
            resizable: true,
            type: 'text',
            typeAttributes: {},
            subType: '',
            role: 'columnheader',
            scope: 'col',
        };
        if (refColumnType === 'action') {
            normCol.ariaLabel = i18n.rowActionsDefaultAriaLabel;
            normCol.fixedWidth = 50;
            normCol.resizable = false;
        }
        else if (refColumnType === 'tree') {
            normCol.subType = 'text';
        }
        else if (refColumnType === SELECTABLE_HEADER_TYPE) {
            normCol.role = 'gridcell';
            normCol.scope = null;
        }
        Object.assign(normCol, refCol);
        normalizedColumns[colIndex] = normCol;
        let { type: columnType } = normCol;
        normCol.ariaLabel = normCol.label || normCol.ariaLabel || null;
        normCol.editableCustomType =
            types.isStandardCellLayoutForCustomType(columnType);
        if (!normCol.internal) {
            if (!types.isValidType(columnType)) {
                columnType = 'text';
                normCol.type = columnType;
            }
            if (types.isEditableType(columnType)) {
                const { editable } = normCol;
                if (!(editable && editable.fieldName)) {
                    normCol.editable = normalizeBoolean(editable);
                }
                normCol.editTemplate =
                    types.getCustomTypeEditTemplate(columnType);
            }
            else {
                normCol.editable = false;
                normCol.editTemplate = undefined;
            }
            const { sortable } = normCol;
            if (sortable &&
                (isSortedByArray
                    ? sortedBy.includes(colName)
                    : sortedBy === colName)) {
                normCol.sorted = true;
                normCol.sortAriaLabel = isSortedByArray
                    ? sortedDirection[sortedBy.indexOf(colName)] === 'desc'
                        ? 'descending'
                        : 'ascending'
                    : sortedDirection === 'desc'
                        ? 'descending'
                        : 'ascending';
                normCol.sortedDirection = isSortedByArray
                    ? sortedDirection[sortedBy.indexOf(colName)]
                    : sortedDirection;
            }
            else {
                normCol.sorted = false;
                normCol.sortAriaLabel = sortable ? 'other' : null;
                normCol.sortedDirection = defaultSortDirection;
            }
            const { cellAttributes } = normCol;
            if (features.enableDatatableHoverStyling &&
                !hasHighlightStyling &&
                cellAttributes?.highlightClass) {
                hasHighlightStyling = true;
            }
        }
        const colKeyValue = normCol.colKeyValue ||
            `${colName || colIndex}-${columnType}-${colIndex}`;
        normCol.colKeyValue = colKeyValue;
        normCol.isLastCol = colIndex === normColCount - 1;
        normCol.isScopeCol = colIndex === firstColumnForReaders;
        normCol.name = colName;
        normCol.tabIndex = -1;
        headerIndexes[colKeyValue] = colIndex;
        if (normCol.type === 'tree') {
            if (state.treeColumn === undefined) {
                state.treeColumn = normCol;
            }
            const { typeAttributes } = normCol;
            const typeAttributesOverrides = {};
            if (!types.isValidTypeForTree(typeAttributes.subType)) {
                typeAttributesOverrides.subType = 'text';
            }
            if (!typeAttributes.subTypeAttributes) {
                typeAttributesOverrides.subTypeAttributes = {};
            }
            normCol.typeAttributes = Object.assign({}, typeAttributes, typeAttributesOverrides);
        }
        const { actions } = normCol;
        normCol.actions = {
            menuAlignment: normCol.isLastCol ||
                normalizedColumns[colIndex + 1].type === 'action' ||
                normalizedColumns[colIndex + 1].fixedWidth < 100
                ? 'auto-right'
                : 'auto-left',
            customerActions: Array.isArray(actions) ? actions : [],
            internalActions: [],
        };
    }
    state.hasHighlightStyling = hasHighlightStyling;
}
export function getTypeAttributesValues(column) {
    const { typeAttributes } = column;
    return isObjectLike(typeAttributes) ? typeAttributes : {};
}
export function getSubTypeAttributesValues(column) {
    const { subTypeAttributes } = column.typeAttributes;
    return isObjectLike(subTypeAttributes) ? subTypeAttributes : {};
}
export function getCellAttributesValues(column) {
    const { cellAttributes } = column;
    return (isObjectLike(cellAttributes) ? cellAttributes : {});
}
export function getUserColumnIndex(state, colKeyValue) {
    const { columns } = state;
    const headerCell = getCellByKeys(state, HEADER_ROW_KEY, colKeyValue);
    const colIndex = headerCell && headerCell.colIndex;
    if (colIndex === undefined ||
        colIndex === -1 ||
        columns[colIndex].internal) {
        return -1;
    }
    let internalColumns = 0;
    for (let i = 0; i < colIndex; i += 1) {
        if (columns[i].internal) {
            internalColumns += 1;
        }
    }
    return colIndex - internalColumns;
}
export function getStateColumnIndex(state, colKeyValue) {
    return state.headerIndexes[colKeyValue];
}
