// lightning-base-components/src/lightning/datatable/datatable.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { api, track, unwrap } from 'lwc';
import divTemplate from './templates/div/div.html';
import tableTemplate from './templates/table/table.html';
import labelAriaLiveActionMode from '@salesforce/label/LightningDatatable.ariaLiveActionMode';
import labelAriaLiveNavigationMode from '@salesforce/label/LightningDatatable.ariaLiveNavigationMode';
import { features } from 'lightning/features';
import { EVENTS as FORMATTED_LOOKUP_EVENTS } from 'lightning/formattedLookup';
import { generateUniqueId } from 'lightning/inputUtils';
import MultiColumnSortingModal from 'lightning/multiColumnSortingModal';
import { getLinkInfo, hasLinkProvider, updateRawLinkInfo, } from 'lightning/routingService';
import { classSetToString, isSafari, normalizeBoolean, normalizeString, synchronizeAttrs, } from 'lightning/utilsPrivate';
import { privateContext } from 'lightning/utilsInternal';
import { updateObservers, disconnectObservers } from './resizeObserver';
import { ColumnWidthManager } from './columnWidthManager';
import { getDefaultState } from './state';
import { setColumns } from './columns';
import { getCustomerColumnWidths, getResizerDefaultState, resizeColumnWithDelta, setMaxColumnWidth, setMinColumnWidth, setResizeColumnDisabled, setResizeStep, updateColumnWidthsMetadata, } from './columnResizer';
import { setErrors } from './errors';
import { setKeyField, recomputeCellStyles, updateCells, updateCellClassForRoleBasedMode, updateRowsAndCells, } from './rows';
import { handleCellButtonClick, handleLoadDynamicActions, handleRowActionTriggered, } from './rowLevelActions';
import { setRowNumberOffset } from './rowNumber';
import { handleDeselectAllRows, handleDeselectRow, handleRowSelectionChange, handleSelectAllRows, handleSelectRow, setMaxRowSelection, setDisabledRowsKeys, setSelectedRowsKeys, syncSelectedRowsKeys, updateBulkSelectionState, setSingleRowSelectionMode, updateRowDisabledStatesForEditLimit, } from './rowSelection';
import { getCurrentSelectionLength, getRowsKeys, getSelectedRowsKeys, isInputTypeCheckbox, } from './rowSelectionShared';
import { handleHeaderActionMenuClosed, handleHeaderActionMenuOpening, handleHeaderActionTriggered, updateHeaderInternalActions, } from './headerActions';
import { getCellByKeys, getRowByKey } from './indexes';
import { handleLoadMoreCheck, handlePrefetch, setEnableInfiniteLoading, setLoading, getLoadMoreOffset, setLoadMoreOffset, } from './infiniteLoading';
import { cancelInlineEdit, closeInlineEdit, getDirtyValues, handleEditCell, handleInlineEditFinish, handleInlineEditPanelScroll, handleMassCheckboxChange, isInlineEditTriggered, openInlineEditOnActiveCell, setDirtyValues, trimDirtyValuesForEditLimit, } from './inlineEdit';
import { addFocusStylesToActiveCell, datatableHasFocus, FOCUS_CLASS, getActiveCellElement, getCellElementChild, getCellElementFromEventTarget, handleDatatableFocusIn, handleDatatableFocusOut, handleKeydownOnCell, handleKeydownOnTable, handlePrimitiveDatatableCellClick, handlePrimitiveDatatableCellFocus, handlePrimitiveDatatableCellKeydown, isActiveCell, isCellElement, KEYBOARD_ACTION_MODE, KEYBOARD_NAVIGATION_MODE, reactToKeyboardOnRow, refocusCellElement, scrollToCell, setBlurActiveCell, setCellClickedForFocus, setCellToFocusFromPrev, setFocusActiveCell, syncActiveCell, unsetRowNavigationMode, updateActiveCellTabIndexAfterSync, updateCellTabIndex, updateCellToFocusFromPrev, updateTabIndexActiveCell, updateTabIndexActiveRow, updateRowNavigationMode, updateRowTabIndex, } from './keyboard';
import { setVirtualize, RenderManager } from './renderManager';
import { setDefaultSortDirection, setSortedBy, setSortedDirection, updateSorting, } from './sort';
import { cleanData, getGridContainerFromScrollerY, getScrollerX, getScrollerY, isLoadMore, isObjectLike, isRenderModeInline, isRenderModeRoleBased, resetHighlightedRowClasses, setHighlightedRowClasses, unsetHighlightedRowClasses, } from './utils';
import { setWrapTextMaxLines } from './wrapText';
import { handleVariableRowHeights, resetRowHeights, findFirstVisibleIndex, addTableHeight, } from './virtualization';
import DatatableTypes from './types';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
const { PrivateLookupItemPickedEvent } = FORMATTED_LOOKUP_EVENTS;
const i18n = {
    ariaLiveNavigationMode: labelAriaLiveNavigationMode,
    ariaLiveActionMode: labelAriaLiveActionMode,
};
const INIT_HEADER_HEIGHT = 32;
const { setContext, assertContext } = privateContext;
export default class LightningDatatable extends LightningShadowBaseClass {
    static customTypes;
    @track
    state = getDefaultState();
    @track
    widthsData = getResizerDefaultState();
    _actionsMinHeightStyle = '';
    @track
    _columnWidthManager;
    _customerDisabledRows = null;
    _customerSelectedRows = null;
    _datatableId = generateUniqueId('lgt-datatable');
    _draftValues = [];
    _headerHeight = 0;
    _isResizing = false;
    @track
    _positionRelationship;
    _privateTypes = {};
    _rawColumns = [];
    @track
    _renderConfig;
    @track
    _renderManager;
    _renderMode = 'default';
    _renderedTemplate = tableTemplate;
    _shouldResetFocus = false;
    _suppressBottomBar = false;
    _rowToggleIcon = {};
    @track
    _widthObserver;
    _hasSetInitialCustomTypes = false;
    _shouldResetCustomTypes = false;
    _recalculateHeaderHeight = false;
    _maxRowSelectionShouldFireEvent = false;
    _rowSelectionEventPending;
    _lastScrollX;
    _lastScrollY;
    _heightObserver;
    _hasRendered;
    _suppressScroll;
    getWrapperHeight;
    @api
    ariaLabel = null;
    @api
    ariaLabelledBy = null;
    ariaDescribedBy = null;
    @api
    get columnWidthsMode() {
        return this.widthsData.columnWidthsMode;
    }
    set columnWidthsMode(value) {
        const { widthsData } = this;
        const normalizedValue = normalizeString(value, {
            fallbackValue: 'fixed',
            validValues: ['fixed', 'auto'],
        });
        this._columnWidthManager.columnWidthMode = normalizedValue;
        if (widthsData.columnWidthsMode !== normalizedValue) {
            this._columnWidthManager.handleWidthModeChange(this.state.columns);
        }
        widthsData.columnWidthsMode = normalizedValue;
    }
    @api
    get columns() {
        return this._rawColumns;
    }
    set columns(value) {
        const _rawColumns = Array.isArray(value) ? value : [];
        this._rawColumns = _rawColumns;
        this.updateColumns(_rawColumns);
        this._columnWidthManager.handleColumnsChange(this.state.columns);
    }
    @api
    get data() {
        return this.state.data;
    }
    set data(value) {
        const { _customerDisabledRows, _customerSelectedRows, state } = this;
        const { data: previousData, columns } = state;
        const data = cleanData(value);
        const isInfiniteLoading = isLoadMore(previousData, data);
        state.data = data;
        if (!isInfiniteLoading && data.length && columns?.length) {
            this._columnWidthManager.handleDataChange(columns);
        }
        if (typeof state.keyField === 'string' && this.isConnected) {
            this.updateRows();
        }
        if (isInfiniteLoading) {
            addTableHeight(state, previousData, data);
        }
        else {
            state.heightCache = {};
            state.offsets = [0];
            state.offsetRanges = [{ start: 0, end: 0 }];
            state.firstVisibleIndex = 0;
            state.tableHeight = state.rowHeight * data.length;
        }
        if (_customerSelectedRows) {
            this.setSelectedRows(_customerSelectedRows);
        }
        if (features.enableDisabledRowsAttribute && _customerDisabledRows) {
            setDisabledRowsKeys(state, _customerDisabledRows);
        }
    }
    @api
    get defaultSortDirection() {
        return this.state.defaultSortDirection;
    }
    set defaultSortDirection(value) {
        const { state } = this;
        setDefaultSortDirection(state, value);
        updateSorting(state);
    }
    @api
    get disabledRows() {
        return getRowsKeys(this.state.disabledRowsKeys);
    }
    set disabledRows(value) {
        this._customerDisabledRows = value;
        setDisabledRowsKeys(this.state, value);
    }
    @api
    get draftValues() {
        return getDirtyValues(this.state);
    }
    set draftValues(value) {
        const { state } = this;
        this._draftValues = value;
        setDirtyValues(state, value);
        if (this.isConnected && typeof state.keyField === 'string') {
            this.updateRowsAndCells();
        }
        updateActiveCellTabIndexAfterSync(state);
    }
    @api
    get enableInfiniteLoading() {
        return this.state.enableInfiniteLoading;
    }
    set enableInfiniteLoading(value) {
        const { state } = this;
        setEnableInfiniteLoading(state, value);
        if (this.isConnected) {
            handlePrefetch.call(this);
        }
    }
    @api
    get errors() {
        return this.state.errors;
    }
    set errors(value) {
        const { state } = this;
        setErrors(state, value);
        if (this.isConnected) {
            updateCells(state);
        }
    }
    @api
    get hideBorders() {
        return this.state.hideBorders;
    }
    set hideBorders(value) {
        this.state.hideBorders = normalizeBoolean(value);
    }
    @api
    get hideCheckboxColumn() {
        return this.state.hideCheckboxColumn;
    }
    set hideCheckboxColumn(value) {
        const { _rawColumns, state } = this;
        const normalizedValue = normalizeBoolean(value);
        this._columnWidthManager.handleCheckboxColumnChange(state.hideCheckboxColumn, normalizedValue, state.columns);
        state.hideCheckboxColumn = normalizedValue;
        if (this.isConnected) {
            this.updateColumns(_rawColumns);
        }
        else if (_rawColumns) {
            state.hadTreeDataTypePreviously = false;
            state.treeColumn = undefined;
            setColumns(state, _rawColumns, this._privateTypes);
        }
    }
    @api
    get showActionsMenu() {
        return this.state.showActionsMenu;
    }
    set showActionsMenu(value) {
        const { state } = this;
        const normalizedValue = normalizeBoolean(value);
        this._columnWidthManager.handleActionsColumnChange(state.showActionsMenu, normalizedValue, state.columns);
        state.showActionsMenu = normalizedValue;
        this.updateColumns(this._rawColumns);
    }
    @api
    get hideTableHeader() {
        return this.state.hideTableHeader;
    }
    set hideTableHeader(value) {
        this.state.hideTableHeader = normalizeBoolean(value);
    }
    @api
    get wrapTableHeader() {
        return this.state.wrapTableHeader;
    }
    set wrapTableHeader(value) {
        const prevWrapTableHeader = this.state.wrapTableHeader;
        const wrapTableHeader = normalizeString(value, {
            fallbackValue: 'none',
            validValues: ['all', 'by-column', 'none'],
        });
        this.state.wrapTableHeader = wrapTableHeader;
        if (wrapTableHeader !== prevWrapTableHeader) {
            this._recalculateHeaderHeight = true;
        }
    }
    @api
    get isLoading() {
        return this.state.isLoading;
    }
    set isLoading(value) {
        setLoading(this.state, value);
    }
    @api
    get keyField() {
        return this.state.keyField;
    }
    set keyField(value) {
        const { state } = this;
        setKeyField(state, value);
        setDirtyValues(state, this._draftValues);
        if (this.isConnected) {
            this.updateRows();
        }
    }
    @api
    get loadMoreOffset() {
        return getLoadMoreOffset(this.state);
    }
    set loadMoreOffset(value) {
        setLoadMoreOffset(this.state, value);
    }
    @api
    get maxColumnWidth() {
        return this.widthsData.maxColumnWidth;
    }
    set maxColumnWidth(value) {
        const { widthsData } = this;
        setMaxColumnWidth(this.state.columns, widthsData, value);
        this._columnWidthManager.maxColumnWidth = widthsData.maxColumnWidth;
    }
    @api
    get maxRowSelection() {
        return this.state.maxRowSelection;
    }
    set maxRowSelection(value) {
        const { state, _customerSelectedRows } = this;
        const previousSelectionLength = getCurrentSelectionLength(state);
        setMaxRowSelection(state, value, this._datatableId);
        if (previousSelectionLength > 0) {
            this.fireSelectedRowsChange(this.getSelectedRows());
        }
        else if (!this.isConnected &&
            state.data.length &&
            _customerSelectedRows) {
            this._maxRowSelectionShouldFireEvent = true;
        }
    }
    @api
    get maxEditLimit() {
        return this.state.maxEditLimit;
    }
    set maxEditLimit(value) {
        const { state } = this;
        const normalizedValue = parseInt(value, 10);
        state.maxEditLimit =
            isNaN(normalizedValue) || normalizedValue < 0
                ? null
                : normalizedValue;
        if (features.enableMaxEditLimit &&
            this.isConnected &&
            this.hasValidKeyField) {
            this.updateRowsAndCells();
        }
        trimDirtyValuesForEditLimit(state);
        updateRowDisabledStatesForEditLimit(this.state, this._privateTypes);
    }
    @api
    get singleRowSelectionMode() {
        return this.state.singleRowSelectionMode;
    }
    set singleRowSelectionMode(value) {
        const { state } = this;
        if (!state.hasCalledUpdateRowsAndCells) {
            this.updateRowsAndCells();
        }
        setSingleRowSelectionMode(state, value, this._datatableId);
    }
    @api
    get minColumnWidth() {
        return this.widthsData.minColumnWidth;
    }
    set minColumnWidth(value) {
        const { widthsData } = this;
        setMinColumnWidth(this.state.columns, widthsData, value);
        this._columnWidthManager.minColumnWidth = widthsData.minColumnWidth;
    }
    @api
    get renderConfig() {
        return this._renderConfig;
    }
    set renderConfig(value) {
        if (isObjectLike(value)) {
            this._renderManager.configure(this.state, this.getWrapperHeight, value);
            this._renderManager.updateViewportRendering(this.state, this.gridContainer, true);
            this._renderConfig = value;
        }
    }
    @api
    get renderMode() {
        return this._renderMode;
    }
    set renderMode(value) {
        const { _renderConfig, state } = this;
        const renderMode = normalizeString(value, {
            fallbackValue: 'default',
            validValues: [
                'default',
                'inline',
                'role-based',
                'role-based-inline',
            ],
        });
        const renderModeInline = isRenderModeInline(renderMode);
        const renderModeRoleBased = isRenderModeRoleBased(renderMode);
        this._renderMode = renderMode;
        state.hasRenderedFirstTime = false;
        state.hasRenderedTable = false;
        state.renderModeInline = renderModeInline;
        state.renderModeRoleBased = renderModeRoleBased;
        if (this.isConnected) {
            this.updateRowsAndCells();
        }
        if (_renderConfig) {
            setVirtualize(state, _renderConfig.virtualize);
        }
        updateCellClassForRoleBasedMode(state);
    }
    @api
    get resizeColumnDisabled() {
        return this.widthsData.resizeColumnDisabled;
    }
    set resizeColumnDisabled(value) {
        setResizeColumnDisabled(this.widthsData, value);
    }
    @api
    get resizeStep() {
        return this.widthsData.resizeStep;
    }
    set resizeStep(value) {
        setResizeStep(this.widthsData, value);
    }
    @api
    get rowNumberOffset() {
        return this.state.rowNumberOffset;
    }
    set rowNumberOffset(value) {
        const { state } = this;
        setRowNumberOffset(state, value);
        this._columnWidthManager.handleRowNumberOffsetChange(state, this.widthsData);
    }
    @api
    get selectedRows() {
        if (features.enableDisabledRowsAttribute) {
            return getRowsKeys(this.state.selectedRowsKeys);
        }
        return getSelectedRowsKeys(this.state);
    }
    set selectedRows(value) {
        this._customerSelectedRows = value;
        this.setSelectedRows(value);
    }
    @api
    get showRowNumberColumn() {
        return this.state.showRowNumberColumn;
    }
    set showRowNumberColumn(value) {
        const { state, _rawColumns } = this;
        const { length: colCount } = _rawColumns;
        if (state.showRowNumberColumn) {
            for (let colIndex = 0; colIndex < colCount; colIndex += 1) {
                if (_rawColumns[colIndex].editable) {
                    return;
                }
            }
        }
        this._columnWidthManager.handleRowNumberColumnChange(state.rowNumberOffset, value, state.columns);
        state.showRowNumberColumn = normalizeBoolean(value);
        if (this.isConnected) {
            this.updateColumns(_rawColumns);
        }
        else if (_rawColumns) {
            state.hadTreeDataTypePreviously = false;
            state.treeColumn = undefined;
            setColumns(state, _rawColumns, this._privateTypes);
        }
    }
    @api
    get sortedBy() {
        return this.state.sortedBy;
    }
    set sortedBy(value) {
        const { state } = this;
        setSortedBy(state, value);
        updateSorting(state);
    }
    @api
    get sortedDirection() {
        return this.state.sortedDirection;
    }
    set sortedDirection(value) {
        const { state } = this;
        setSortedDirection(state, value);
        updateSorting(state);
    }
    @api
    get suppressBottomBar() {
        return this._suppressBottomBar;
    }
    set suppressBottomBar(value) {
        this._suppressBottomBar = normalizeBoolean(value);
    }
    @api
    get rowToggleIcon() {
        return this._rowToggleIcon || {};
    }
    set rowToggleIcon(value) {
        if (isObjectLike(value)) {
            this._rowToggleIcon = value;
        }
        else {
            this._rowToggleIcon = {};
        }
    }
    @api
    get wrapTextMaxLines() {
        return this.state.wrapTextMaxLines;
    }
    set wrapTextMaxLines(value) {
        const { state } = this;
        setWrapTextMaxLines(state, value);
        this._columnWidthManager.wrapTextMaxLines = state.wrapTextMaxLines;
        if (this.isConnected) {
            updateCells(state);
        }
    }
    @api
    focus() {
        const { state } = this;
        syncActiveCell(state);
        setFocusActiveCell(state, this.template, this.refs);
    }
    @api
    getSelectedRows() {
        const { state } = this;
        if (!state.hasCalledUpdateRowsAndCells) {
            this.updateRowsAndCells();
        }
        let data;
        if (!import.meta.env.SSR) {
            data = unwrap(state.data);
        }
        else {
            data = state.data;
        }
        const { rows } = state;
        const selectedRows = [];
        for (let i = 0, { length: rowCount } = rows; i < rowCount; i += 1) {
            const row = rows[i];
            if (row.isSelected) {
                selectedRows.push(data[i]);
            }
        }
        return selectedRows;
    }
    @api
    openInlineEdit() {
        if (!this.state.hasCalledUpdateRowsAndCells) {
            this.updateRowsAndCells();
        }
        openInlineEditOnActiveCell(this);
    }
    @api
    scrollToTop() {
        const scrollerY = getScrollerY(this.template, this.validRefs);
        if (scrollerY) {
            scrollerY.scrollTop = 0;
        }
    }
    get consolidateTypeAttributes() {
        return features.enableConsolidateTypeAttributes;
    }
    get gridContainer() {
        return !import.meta.env.SSR
            ? getGridContainerFromScrollerY(getScrollerY(this.template, this.validRefs))
            : null;
    }
    get computedTableContainerClass() {
        return classSetToString({
            'slds-table_header-fixed_container': !this.state.hideTableHeader,
            'slds-scrollable_x': !this._isResizing,
        });
    }
    get computedTableClass() {
        const { state } = this;
        const headerType = state.hideTableHeader ? 'hidden' : 'fixed';
        return classSetToString({
            'slds-table': true,
            [`slds-table_header-${headerType}`]: true,
            'slds-table_bordered': !(state.hideBorders && state.hideTableHeader),
            'slds-table_edit': true,
            'slds-table_resizable-cols': !this.widthsData.resizeColumnDisabled,
            'slds-tree slds-table_tree': !!state.treeColumn,
        });
    }
    get computedTableRole() {
        return this.state.treeColumn ? 'treegrid' : 'grid';
    }
    get computedTableStyle() {
        const tableLayout = this._columnWidthManager.isAutoResizingUpdateQueued
            ? 'auto'
            : 'fixed';
        let style = `table-layout: ${tableLayout};`;
        const { tableWidth } = this.widthsData;
        if (tableWidth) {
            style += `width: ${tableWidth}px;`;
        }
        return style;
    }
    get computedTbodyStyle() {
        const { state } = this;
        const { rowNumberOffset } = state;
        let style = '';
        if (state.showRowNumberColumn && rowNumberOffset >= 0) {
            const { firstVisibleIndex, bufferSize } = state;
            const firstRenderedRow = Math.max(firstVisibleIndex - bufferSize, 0);
            const rowNumber = firstRenderedRow + rowNumberOffset;
            style += `counter-reset: row-number ${rowNumber};`;
        }
        if (state.virtualize) {
            const { tableHeight } = state;
            style += `position:relative;height:${tableHeight}px`;
        }
        return style;
    }
    get computedScrollerYStyle() {
        const minHeight = this._actionsMinHeightStyle
            ? `${this._actionsMinHeightStyle};`
            : '';
        if (this._columnWidthManager.isAutoResizingUpdateQueued) {
            return `${minHeight}overflow-x: auto;`;
        }
        const { tableWidth } = this.widthsData;
        return tableWidth
            ? `${minHeight}width: ${tableWidth}px;`
            : `${minHeight}`;
    }
    get computedScrollerXStyle() {
        let style;
        if (features.enableDTHeightCalcForScrollbar) {
            style = this.showStatusBar
                ? 'height: calc(100% - 3rem);'
                : 'height: 100%;';
        }
        else {
            style = 'height: 100%;';
            if (this.showStatusBar) {
                style += 'padding-bottom: 3rem;';
            }
        }
        if (this._columnWidthManager.isAutoResizingUpdateQueued) {
            style += 'overflow-x: auto;';
        }
        if (this.state.wrapTableHeader !== 'none') {
            style += `padding-top: max(2rem, ${this._headerHeight}px);`;
        }
        return style;
    }
    get computedCheckboxColumnHeaderId() {
        return `${this._datatableId}-check-column-header`;
    }
    get computedAriaLiveClassForNavMode() {
        const isNavMode = this.state.keyboardMode === KEYBOARD_NAVIGATION_MODE;
        return classSetToString({
            'slds-hide': !isNavMode,
            'slds-assistive-text': isNavMode,
        });
    }
    get computedAriaLiveClassForActionMode() {
        const isActionMode = this.state.keyboardMode === KEYBOARD_ACTION_MODE;
        return classSetToString({
            'slds-hide': !isActionMode,
            'slds-assistive-text': isActionMode,
        });
    }
    get ariaRowCount() {
        const { state } = this;
        if (!state.hasCalledUpdateRowsAndCells) {
            this.updateRowsAndCells();
        }
        return state.rows.length + 1;
    }
    get hasValidKeyField() {
        if (typeof this.state.keyField === 'string') {
            return true;
        }
        console.error(`The "keyField" is a required attribute of lightning:datatable.`);
        return false;
    }
    get hasResizebleColumns() {
        return !this.widthsData.resizeColumnDisabled;
    }
    get validRefs() {
        return this._hasRendered ? this.refs : null;
    }
    get renderedRows() {
        const { state } = this;
        if (!state.hasCalledUpdateRowsAndCells) {
            this.updateRowsAndCells();
        }
        const { virtualize, rows, renderedRowCount } = state;
        if (virtualize) {
            const { firstIndex, lastIndex } = this._renderManager.getRenderedRange(state);
            this._shouldResetFocus = true;
            return rows.slice(firstIndex, lastIndex);
        }
        if (state.enableViewportRendering) {
            return rows.slice(0, renderedRowCount);
        }
        return rows;
    }
    get showSelectAllCheckbox() {
        return isInputTypeCheckbox(this.state);
    }
    get showStatusBar() {
        return !this._suppressBottomBar && isInlineEditTriggered(this.state);
    }
    get tableError() {
        return this.state.errors.table;
    }
    get i18n() {
        return i18n;
    }
    constructor() {
        super();
        this._columnWidthManager = new ColumnWidthManager(this.widthsData);
        this._privateTypes = new DatatableTypes(this.constructor.customTypes);
        this._hasSetInitialCustomTypes =
            !!this._privateTypes.privateCustomTypes.size;
        this._renderManager = new RenderManager();
        if (!import.meta.env.SSR) {
            this.getWrapperHeight = () => {
                const scrollerX = getScrollerX(this.template, this.refs);
                const paddingTop = this.state.hideTableHeader
                    ? 0
                    : parseFloat(getComputedStyle(scrollerX).paddingTop) ||
                        INIT_HEADER_HEIGHT;
                return scrollerX.offsetHeight - paddingTop;
            };
        }
        setContext(this);
    }
    connectedCallback() {
        super.connectedCallback();
        const { _customerDisabledRows, _customerSelectedRows, state, template, } = this;
        this.addEventListener('rowselection', (event) => {
            this.handleRowSelectionChange(event);
        });
        if (!import.meta.env.SSR) {
            template.addEventListener('selectallrows', (event) => {
                this.handleSelectionCellClick(event);
            });
            template.addEventListener('deselectallrows', (event) => {
                this.handleSelectionCellClick(event);
            });
            template.addEventListener('selectrow', (event) => {
                this.handleSelectionCellClick(event);
            });
            template.addEventListener('deselectrow', (event) => {
                this.handleSelectionCellClick(event);
            });
            template.addEventListener('resizecol', (event) => {
                this.handleResizeColumn(event);
            });
            template.addEventListener('privateupdatecolsort', (event) => {
                this.handleUpdateColumnSort(event);
            });
            template.addEventListener('privatecellkeydown', (event) => {
                this.handleKeydownOnCell(event);
            });
            template.addEventListener('privatecellfocusedbyclick', (event) => {
                this.handleCellFocusByClick(event);
            });
            template.addEventListener('privatecellfalseblurred', (event) => {
                this.handleFalseCellBlur(event);
            });
            template.addEventListener('click', (event) => {
                this.handlePrimitiveDatatableCellClick(event);
            });
            template.addEventListener('focus', (event) => {
                this.handlePrimitiveDatatableCellFocus(event);
            });
            template.addEventListener('mousedown', () => {
                this._suppressScroll = true;
            });
            template.addEventListener('keydown', (event) => {
                this.handlePrimitiveDatatableCellKeydown(event);
            });
            template.addEventListener('privatecellactiontriggered', (event) => {
                this.handleRowActionTriggered(event);
            });
            template.addEventListener('privatecellactionmenuopening', (event) => {
                this.handleLoadDynamicActions(event);
            });
            template.addEventListener('privatecellbuttonclicked', (event) => {
                this.handleCellButtonClick(event);
            });
            template.addEventListener('privatecellheaderactionmenuopening', (event) => {
                this.handleHeaderActionMenuOpening(event);
            });
            template.addEventListener('privatecellheaderactionmenuclosed', event => {
                this.handleHeaderActionMenuClosed(event);
            });
            template.addEventListener('privatecellheaderactiontriggered', (event) => {
                this.handleHeaderActionTriggered(event);
            });
            template.addEventListener('privateopensortingmodal', (event) => {
                this.handleOpenSortingModal(event);
            });
            template.addEventListener('privateeditcell', (event) => {
                this.handlePrivateDatatableEditCell(event);
            });
        }
        this.updateColumnsAndRows();
        if (state.enableViewportRendering || state.virtualize) {
            state.renderedRowCount = 1;
        }
        if (_customerSelectedRows) {
            this.setSelectedRows(_customerSelectedRows);
            const selectionLength = getCurrentSelectionLength(this.state);
            if (selectionLength > 0 && this._maxRowSelectionShouldFireEvent) {
                this.fireSelectedRowsChange(this.getSelectedRows());
                this._maxRowSelectionShouldFireEvent = false;
            }
        }
        if (features.enableDisabledRowsAttribute && _customerDisabledRows) {
            setDisabledRowsKeys(state, _customerDisabledRows);
        }
    }
    render() {
        const requireTableTemplate = this._columnWidthManager.isAutoResizingUpdateQueued;
        if (!this._hasRendered) {
            this._hasRendered = true;
        }
        const templateToRender = this.state.renderModeRoleBased && !requireTableTemplate
            ? divTemplate
            : tableTemplate;
        if (templateToRender !== this._renderedTemplate) {
            this._shouldResetCustomTypes =
                !this._privateTypes.privateCustomTypes.size;
            disconnectObservers(this);
            this._renderManager.disconnectResizeObserver();
            this._renderedTemplate = templateToRender;
        }
        return templateToRender;
    }
    renderedCallback() {
        const { _columnWidthManager, gridContainer, refs, state, template, widthsData, } = this;
        if (!this._hasSetInitialCustomTypes) {
            this.setCustomTypes();
            this._hasSetInitialCustomTypes = true;
            if (this._privateTypes.privateCustomTypes.size &&
                widthsData.columnWidthsMode === 'auto') {
                return;
            }
        }
        synchronizeAttrs(gridContainer, {
            'aria-label': this.ariaLabel,
            'aria-labelledby': this.ariaLabelledBy,
            'aria-describedby': this.ariaDescribedBy,
        });
        const { renderModeInline } = state;
        if (_columnWidthManager.isResizingUpdateQueued) {
            const { columns } = state;
            const fireResizeEvent = _columnWidthManager.shouldFireResizeEvent(widthsData, columns);
            _columnWidthManager.adjustColumnsSize(template, refs, columns, widthsData);
            state.shouldResetHeights = true;
            if (state.renderModeRoleBased) {
                recomputeCellStyles(state);
            }
            if (fireResizeEvent) {
                this.fireOnResize(false);
            }
            this._recalculateHeaderHeight = true;
            this.updateTableAndScrollerStyleOnRender();
        }
        updateObservers(this);
        this._customerSelectedRows = null;
        state.cellToFocusNext = undefined;
        if (this._shouldResetFocus) {
            this._shouldResetFocus = false;
            const { activeCell } = state;
            if (activeCell &&
                activeCell.hasFocus &&
                !state.inlineEdit.isPanelVisible) {
                const cellElement = getActiveCellElement(template, state);
                const cellElementChild = getCellElementChild(cellElement);
                if (!renderModeInline &&
                    cellElement &&
                    cellElementChild &&
                    !cellElementChild.classList.contains(FOCUS_CLASS)) {
                    setFocusActiveCell(state, template, refs, null, null, false);
                }
            }
        }
        const { hasRenderedFirstTime, hasRenderedTable } = state;
        const virtualize = !!state.virtualize;
        const { length: rowCount } = state.rows;
        state.hasRenderedFirstTime = true;
        state.hasRenderedTable = true;
        let shouldUpdateWrapperHeight = false;
        if (virtualize || state.enableViewportRendering) {
            const { _renderManager } = this;
            if (rowCount) {
                if (virtualize) {
                    this.updateVirtualizedRowHeights();
                }
            }
            if (!hasRenderedFirstTime) {
                const outerContainer = refs?.outerContainer ||
                    template.querySelector('div.dt-outer-container');
                _renderManager.connectResizeObserver(outerContainer);
            }
            if (!_renderManager.hasWrapperHeight()) {
                shouldUpdateWrapperHeight = true;
            }
        }
        requestAnimationFrame(() => {
            if (shouldUpdateWrapperHeight) {
                const { _renderManager } = this;
                _renderManager.updateWrapperHeight(this.getWrapperHeight);
                if (state.renderedRowCount) {
                    _renderManager.updateViewportRendering(state, gridContainer, true);
                }
            }
            this.updateHeaderHeight();
            if (rowCount && renderModeInline && !hasRenderedTable) {
                this.updateCheckboxCellsAriaAttrs();
                if (hasLinkProvider(template)) {
                    this.updateLookupCellsLinkInfo();
                    this.updateUrlCellsLinkInfo();
                }
            }
        });
    }
    updateHeaderHeight() {
        if (this.state.wrapTableHeader !== 'none' &&
            this._recalculateHeaderHeight) {
            const headerCells = this.refs.headerRow?.children || [];
            let headerTextHeight = 0;
            for (let i = 0; i < headerCells.length; i += 1) {
                const headerElement = headerCells[i].firstChild;
                const headerCellHeight = headerElement.getDomHeight();
                if (headerCellHeight > headerTextHeight) {
                    headerTextHeight = headerCellHeight;
                }
            }
            this._headerHeight = headerTextHeight + 16;
            this._recalculateHeaderHeight = false;
        }
        else if (this._recalculateHeaderHeight) {
            this._headerHeight = 0;
            this._recalculateHeaderHeight = false;
        }
    }
    updateTableAndScrollerStyleOnRender() {
        const scrollYEle = getScrollerY(this.template, this.refs);
        const tableElement = getGridContainerFromScrollerY(scrollYEle);
        if (tableElement) {
            tableElement.style = this.computedTableStyle;
        }
        if (scrollYEle) {
            scrollYEle.style = this.computedScrollerYStyle;
        }
    }
    disconnectedCallback() {
        disconnectObservers(this);
        this._renderManager.disconnectResizeObserver();
    }
    handleCellButtonClick = handleCellButtonClick;
    handlePrivateDatatableEditCell = handleEditCell;
    handleHeaderActionMenuClosed = handleHeaderActionMenuClosed;
    handleHeaderActionMenuOpening = handleHeaderActionMenuOpening;
    handleHeaderActionTriggered = handleHeaderActionTriggered;
    handleKeydownOnCell = handleKeydownOnCell;
    handleLoadDynamicActions = handleLoadDynamicActions;
    handlePrimitiveDatatableCellClick = handlePrimitiveDatatableCellClick;
    handlePrimitiveDatatableCellFocus = handlePrimitiveDatatableCellFocus;
    handlePrimitiveDatatableCellKeydown = handlePrimitiveDatatableCellKeydown;
    handleRowSelectionChange = handleRowSelectionChange;
    handleRowActionTriggered = handleRowActionTriggered;
    handleCustomTypesChange() {
        if (this._shouldResetCustomTypes) {
            this.setCustomTypes();
        }
        else {
            this._shouldResetCustomTypes = true;
        }
    }
    handleTableKeydown = handleKeydownOnTable;
    handleKeydownOnDataRow(event) {
        const { state } = this;
        if (state.keyboardMode === KEYBOARD_NAVIGATION_MODE && state.rowMode) {
            event.stopPropagation();
            const { currentTarget: tr } = event;
            reactToKeyboardOnRow(this, {
                target: tr,
                detail: {
                    keyCode: event.keyCode,
                    keyEvent: event,
                    rowExpanded: tr.getAttribute('aria-expanded') === 'true',
                    rowHasChildren: !!tr.getAttribute('aria-expanded'),
                    rowKeyValue: tr.getAttribute('data-row-key-value'),
                    rowLevel: tr.getAttribute('aria-level'),
                },
            });
        }
    }
    handleHorizontalScroll(event) {
        assertContext(this);
        handleInlineEditPanelScroll.call(this, event);
    }
    handleVerticalScroll(event) {
        assertContext(this);
        const { state } = this;
        if (state.enableInfiniteLoading) {
            handleLoadMoreCheck.call(this, event);
        }
        handleInlineEditPanelScroll.call(this, event);
        if (state.virtualize) {
            state.firstVisibleIndex = findFirstVisibleIndex(this.state, event.target.scrollTop);
        }
        else if (state.enableViewportRendering) {
            this._renderManager.handleScroll(state, event);
        }
    }
    handleTableCellClick(event) {
        const { target } = event;
        if (isCellElement(target)) {
            const cellElement = target;
            const { template, state } = this;
            state.hasReceivedFocus = true;
            const { rowKeyValue } = cellElement.parentElement
                .dataset;
            const { colKeyValue } = cellElement.dataset;
            if (state.rowMode ||
                !isActiveCell(state, rowKeyValue, colKeyValue)) {
                const { activeCell } = state;
                if (activeCell && state.rowMode) {
                    unsetRowNavigationMode(state);
                    updateRowTabIndex(state, activeCell.rowIndex, -1);
                }
                this.setActiveCell(rowKeyValue, colKeyValue);
            }
            if (!datatableHasFocus(state, template)) {
                setCellClickedForFocus(state);
            }
            scrollToCell(state, template, this.refs, cellElement);
        }
    }
    handleUpdateColumnSort(event) {
        event.stopPropagation();
        const { fieldName, columnKey, sortDirection } = event.detail;
        this.fireSortedColumnChange(fieldName, columnKey, sortDirection);
    }
    handleOpenSortingModal(event) {
        event.stopPropagation();
        MultiColumnSortingModal.open({
            dtInstance: this,
        }).then((response) => {
            if (response && response.event.defaultPrevented) {
                response.failure();
            }
            else if (response) {
                response.success();
            }
        });
    }
    handleResizeColumn(event) {
        event.stopPropagation();
        const { state } = this;
        const { colIndex, widthDelta: delta } = event.detail;
        if (delta !== 0) {
            resizeColumnWithDelta(state.columns, this.widthsData, colIndex, delta);
            if (state.renderModeRoleBased) {
                recomputeCellStyles(state);
            }
            this.fireOnResize(true);
            this.fixHeaderForSafari();
            this._recalculateHeaderHeight = true;
        }
    }
    handleResizeStart(event) {
        event.stopPropagation();
        this._isResizing = true;
    }
    handleResizeEnd(event) {
        event.stopPropagation();
        this._isResizing = false;
        this.state.shouldResetHeights = true;
    }
    handleSelectionCellClick(event) {
        this.handleCellFocusByClick(event);
        const { type } = event;
        if (type === 'selectrow') {
            handleSelectRow.call(this, event);
        }
        else if (type === 'deselectrow') {
            handleDeselectRow.call(this, event);
        }
        else if (type === 'selectallrows') {
            handleSelectAllRows.call(this, event);
        }
        else if (type === 'deselectallrows') {
            handleDeselectAllRows.call(this, event);
        }
        updateRowDisabledStatesForEditLimit(this.state, this._privateTypes);
    }
    handleCellFocusByClick(event) {
        event.stopPropagation();
        const { state } = this;
        state.hasReceivedFocus = true;
        const { rowKeyValue, colKeyValue, needsRefocusOnCellElement } = event.detail;
        if (!isActiveCell(state, rowKeyValue, colKeyValue)) {
            const { activeCell } = state;
            if (activeCell && state.rowMode) {
                unsetRowNavigationMode(state);
                updateRowTabIndex(state, activeCell.rowIndex, -1);
            }
            this.setActiveCell(rowKeyValue, colKeyValue);
            refocusCellElement(state, this.template, needsRefocusOnCellElement);
        }
        const cellElement = getCellElementFromEventTarget(event.target);
        scrollToCell(state, this.template, this.refs, cellElement);
    }
    handleFalseCellBlur(event) {
        event.stopPropagation();
        const { state } = this;
        const { rowKeyValue, colKeyValue } = event.detail;
        if (!isActiveCell(state, rowKeyValue, colKeyValue)) {
            this.setActiveCell(rowKeyValue, colKeyValue);
        }
        setFocusActiveCell(state, this.template, this.refs);
    }
    handleTableFocusIn = handleDatatableFocusIn;
    handleTableFocusOut(event) {
        handleDatatableFocusOut.call(this, event);
        const { activeCell } = this.state;
        if (activeCell) {
            activeCell.hasFocus = this._shouldResetFocus;
        }
    }
    handleInlineEditFinish(event) {
        assertContext(this);
        handleInlineEditFinish.call(this, event);
    }
    handleMassCheckboxChange = handleMassCheckboxChange;
    handleInlineEditSave(event) {
        event.stopPropagation();
        event.preventDefault();
        closeInlineEdit(this);
        const { draftValues } = this;
        this.dispatchEvent(new CustomEvent('save', {
            detail: {
                draftValues,
            },
        }));
    }
    handleInlineEditCancel(event) {
        event.stopPropagation();
        event.preventDefault();
        closeInlineEdit(this);
        const customerEvent = new CustomEvent('cancel', {
            cancelable: true,
        });
        this.dispatchEvent(customerEvent);
        if (!customerEvent.defaultPrevented) {
            cancelInlineEdit(this);
        }
        updateActiveCellTabIndexAfterSync(this.state);
    }
    handleMouseOver(event) {
        const { state } = this;
        if (!features.enableDatatableHoverStyling ||
            !state.hasHighlightStyling) {
            return;
        }
        event.stopPropagation();
        const tr = event.target.closest('[role="row"]');
        if (tr) {
            const rowKeyValue = tr.getAttribute('data-row-key-value');
            setHighlightedRowClasses(state, rowKeyValue);
            state.hoveredRowKeyValue = rowKeyValue;
        }
    }
    handleMouseOut(event) {
        const { state } = this;
        if (!features.enableDatatableHoverStyling ||
            !state.hasHighlightStyling) {
            return;
        }
        event.stopPropagation();
        const tr = event.target.closest('[role="row"]');
        if (tr) {
            const rowKeyValue = tr.getAttribute('data-row-key-value');
            unsetHighlightedRowClasses(state, rowKeyValue, true);
            state.hoveredRowKeyValue = undefined;
        }
    }
    handlePrivateLookupItemPicked(event) {
        event.stopPropagation();
        event.preventDefault();
        const { recordId } = event.detail;
        const { rowIndex, rowKeyValue } = this.computeRowLookupItemPickedInformation(event.currentTarget);
        this.dispatchEvent(new CustomEvent(FORMATTED_LOOKUP_EVENTS.PrivateLookupItemPickedEvent.NAME, {
            composed: true,
            cancelable: true,
            bubbles: true,
            detail: {
                recordId,
                rowIndex,
                rowKeyValue,
            },
        }));
    }
    handleInlinedCheckboxClick(event) {
        const { state } = this;
        event.stopPropagation();
        const { currentTarget: checkboxElement } = event;
        const { parentElement: checkboxContainer } = checkboxElement;
        const { parentElement: cellElement } = checkboxContainer;
        const { colKeyValue } = cellElement.dataset;
        const { rowKeyValue } = cellElement.parentElement.dataset;
        const row = getRowByKey(state, rowKeyValue);
        this.fireInlinedCheckboxSelection(checkboxElement, row.isSelected, {
            rowKeyValue,
            colKeyValue,
            isMultiple: event.shiftKey,
        });
    }
    handleInlinedCheckboxContainerClick(event) {
        const { state } = this;
        const { currentTarget: checkboxContainer } = event;
        const { parentElement: cellElement } = checkboxContainer;
        const { rowKeyValue } = cellElement.parentElement.dataset;
        const row = getRowByKey(state, rowKeyValue);
        if (row.isDisabled) {
            return;
        }
        event.preventDefault();
        const { colKeyValue } = cellElement.dataset;
        const { firstElementChild: checkboxElement } = checkboxContainer;
        this.fireInlinedCheckboxSelection(checkboxElement, row.isSelected, {
            rowKeyValue,
            colKeyValue,
            isMultiple: event.shiftKey,
        });
    }
    handleInlinedCheckboxContainerMouseDown(event) {
        if (this.state.renderModeInline && event.shiftKey) {
            event.preventDefault();
        }
    }
    handleInlinedEditButtonClick(event) {
        const { state } = this;
        const cellElement = state.renderModeInline
            ? getCellElementFromEventTarget(event.target)
            : undefined;
        if (cellElement) {
            state.hasReceivedFocus = true;
            const cellElementChild = getCellElementChild(cellElement);
            cellElementChild.dispatchEvent(new CustomEvent('privateeditcell', {
                bubbles: true,
                composed: true,
                detail: {
                    rowKeyValue: cellElement.parentElement
                        .dataset.rowKeyValue,
                    colKeyValue: cellElement.dataset.colKeyValue,
                },
            }));
        }
    }
    fireInlinedCheckboxSelection(checkboxElement, isSelected, detail) {
        const actionName = isSelected ? 'deselectrow' : 'selectrow';
        checkboxElement.dispatchEvent(new CustomEvent(actionName, {
            bubbles: true,
            composed: true,
            detail,
        }));
    }
    fireSelectedRowsChange(selectedRows, config = {}) {
        if (!import.meta.env.SSR) {
            this.dispatchEvent(new CustomEvent('rowselection', {
                detail: { selectedRows, config },
            }));
        }
    }
    fireSortedColumnChange(fieldName, columnKey, sortDirection) {
        const sortEvent = new CustomEvent('sort', {
            detail: {
                fieldNames: [fieldName],
                sortDirections: [sortDirection],
                isMultiColumnSort: false,
                fieldName,
                columnKey,
                sortDirection,
            },
        });
        this.dispatchEvent(sortEvent);
    }
    fireOnResize(isUserTriggered) {
        this.dispatchEvent(new CustomEvent('resize', {
            detail: {
                columnWidths: getCustomerColumnWidths(this.state.columns, this.widthsData),
                isUserTriggered: !!isUserTriggered,
            },
        }));
    }
    updateRows() {
        this.updateRowsBeforeIndexes();
        this.updateRowsAndCells();
        this.updateRowsAfterIndexes();
    }
    updateRowsBeforeIndexes() {
        if (!import.meta.env.SSR) {
            setCellToFocusFromPrev(this.state, this.template);
        }
    }
    updateRowsAfterIndexes() {
        const { state, widthsData } = this;
        if (state.enableViewportRendering || state.virtualize) {
            this._renderManager.updateViewportRendering(state, this.gridContainer, !!state.virtualize);
        }
        this._columnWidthManager.handleRowNumberOffsetChange(state, widthsData);
        updateCellToFocusFromPrev(state);
        syncSelectedRowsKeys(state, this.getSelectedRows()).ifChanged(() => {
            if (!this._rowSelectionEventPending) {
                this._rowSelectionEventPending = true;
                Promise.resolve().then(() => {
                    if (this._rowSelectionEventPending) {
                        this.fireSelectedRowsChange(this.getSelectedRows());
                        this._rowSelectionEventPending = false;
                    }
                });
            }
        });
        syncActiveCell(state);
        if (state.keyboardMode === KEYBOARD_NAVIGATION_MODE) {
            updateTabIndexActiveCell(state);
            updateTabIndexActiveRow(state);
        }
        if (!import.meta.env.SSR && state.activeCell && state.cellToFocusNext) {
            setFocusActiveCell(state, this.template, this.refs);
        }
    }
    updateColumns(columns) {
        setColumns(this.state, columns, this._privateTypes);
        if (this.isConnected) {
            this.updateColumnsBeforeIndexes();
            this.updateRowsAndCells();
            this.updateColumnsAfterIndexes();
        }
    }
    updateColumnsBeforeIndexes() {
        const { state } = this;
        if (!import.meta.env.SSR) {
            setCellToFocusFromPrev(state, this.template);
        }
        updateRowNavigationMode(state);
        setDirtyValues(state, this._draftValues);
        updateHeaderInternalActions(state);
    }
    updateColumnsAfterIndexes() {
        const { state, widthsData } = this;
        updateBulkSelectionState(state);
        this._columnWidthManager.handleRowNumberOffsetChange(state, widthsData);
        updateColumnWidthsMetadata(state.columns, widthsData);
        updateCellToFocusFromPrev(state);
        if (state.data.length > 0 &&
            state.columns.length !== widthsData.columnWidths.length) {
            syncActiveCell(state);
        }
        if (state.keyboardMode === KEYBOARD_NAVIGATION_MODE) {
            updateTabIndexActiveCell(state);
            updateTabIndexActiveRow(state);
        }
        if (!import.meta.env.SSR && state.activeCell && state.cellToFocusNext) {
            setFocusActiveCell(state, this.template, this.refs);
        }
    }
    updateColumnsAndRows() {
        this.updateColumnsBeforeIndexes(this._rawColumns);
        this.updateRowsBeforeIndexes();
        this.updateRowsAndCells();
        this.updateRowsAfterIndexes();
        this.updateColumnsAfterIndexes();
    }
    updateRowsAndCells() {
        const { state } = this;
        state.hasRenderedTable = false;
        updateRowsAndCells(state, this._privateTypes, this._datatableId);
    }
    updateVirtualizedRowHeights() {
        if (features.enableAssertContext260) {
            assertContext(this);
        }
        const state = this.state;
        const virtualizedRows = state.virtualize && this.renderedRows.length;
        if (state.shouldResetHeights) {
            resetRowHeights(state);
            state.shouldResetHeights = false;
        }
        else if (virtualizedRows) {
            handleVariableRowHeights(this.template, state, this.renderedRows, this.refs);
        }
    }
    setSelectedRows(value) {
        setSelectedRowsKeys(this.state, value);
        this.handleRowSelectionChange();
        updateRowDisabledStatesForEditLimit(this.state, this._privateTypes);
    }
    setActiveCell(rowKeyValue, colKeyValue) {
        const { state, template } = this;
        const newActiveCell = getCellByKeys(state, rowKeyValue, colKeyValue);
        setBlurActiveCell(state, template);
        state.activeCell = newActiveCell;
        addFocusStylesToActiveCell(state, template);
        updateCellTabIndex(state, newActiveCell.rowIndex, newActiveCell.colIndex, 0);
        resetHighlightedRowClasses(state, rowKeyValue, false);
    }
    setCustomTypes() {
        const customTypesSlot = this.template.querySelector('slot[name="customdatatypes"]');
        if (customTypesSlot) {
            const assignedNodes = customTypesSlot.assignedNodes();
            if (assignedNodes.length) {
                let provider = assignedNodes[0];
                if (provider?.tagName === 'SLOT') {
                    const innerAssignedNodes = provider.assignedNodes();
                    provider = (innerAssignedNodes.length ? innerAssignedNodes[0] : {});
                }
                if (typeof provider.getDataTypes === 'function') {
                    this._privateTypes = new DatatableTypes(provider.getDataTypes());
                    this.updateColumns(this._rawColumns);
                    this._columnWidthManager.handleColumnsChange(this.state.columns, this.validRefs);
                }
            }
            else {
                this._shouldResetCustomTypes = true;
            }
        }
    }
    getViewableRect() {
        const { refs, template } = this;
        const scrollerY = getScrollerY(template, refs);
        const scrollerYRect = scrollerY.getBoundingClientRect();
        const scrollerX = getScrollerX(template, refs);
        const scrollerXRect = scrollerX.getBoundingClientRect();
        return {
            left: scrollerXRect.left,
            right: scrollerXRect.right,
            top: scrollerYRect.top,
            bottom: scrollerYRect.bottom,
        };
    }
    fixHeaderForSafari() {
        if (isSafari) {
            const thead = this.template.querySelector('thead');
            if (thead) {
                thead.style.display = 'none';
                requestAnimationFrame(() => {
                    thead.style.display = '';
                });
            }
        }
    }
    computeRowLookupItemPickedInformation(currentTarget) {
        const { dataset } = currentTarget;
        const rowIndex = Number.parseInt(dataset.rowNumber, 10) - 1;
        const { rowKeyValue } = dataset;
        return { rowIndex, rowKeyValue };
    }
    updateCheckboxCellsAriaAttrs() {
        const { computedCheckboxColumnHeaderId, state, template } = this;
        const { checkboxCells } = state;
        const { length: cellCount } = checkboxCells;
        const checkboxElements = cellCount
            ? template.querySelectorAll('[data-inline-type="primitive-cell-checkbox"]')
            : [];
        let scopedIdPostfix;
        for (let i = 0, { length } = checkboxElements; i < length; i += 1) {
            const cell = checkboxCells[i];
            const checkboxElem = checkboxElements[i];
            const { firstElementChild: inputElem } = checkboxElem;
            if (scopedIdPostfix === undefined) {
                const { lastElementChild: labelElem } = checkboxElem;
                const { id } = labelElem;
                scopedIdPostfix = id.slice(id.lastIndexOf('-'));
            }
            inputElem.setAttribute('aria-labelledby', `${cell.checkboxLabelId}${scopedIdPostfix} ${computedCheckboxColumnHeaderId}`);
        }
    }
    updateLookupCellsLinkInfo() {
        assertContext(this);
        const { state, template } = this;
        const { lookupCells } = state;
        const { length: cellCount } = lookupCells;
        const lookupElements = cellCount
            ? template.querySelectorAll('[data-inline-type="formatted-lookup"]')
            : [];
        for (let i = 0, { length } = lookupElements; i < length; i += 1) {
            const cell = lookupCells[i];
            const lookupElem = lookupElements[i];
            const { value: recordId } = cell;
            cell._handleLookupAnchorClick = () => {
                lookupElem.dispatchEvent(new PrivateLookupItemPickedEvent({ recordId }));
            };
            if (lookupElem.getAttribute('data-navigation') === 'enable') {
                lookupElem.removeAttribute('data-navigation');
            }
            if (recordId) {
                getLinkInfo(lookupElem, {
                    stateType: 'standard__recordPage',
                    attributes: {
                        recordId,
                        actionName: 'view',
                    },
                }).then((linkInfo) => {
                    const { dispatcher, url } = linkInfo;
                    const lookupIsNavigable = !!(dispatcher && url);
                    cell.lookupIsNavigable = lookupIsNavigable;
                    cell.lookupLink = url;
                    cell._handleLookupAnchorClick = (event) => {
                        lookupElem.dispatchEvent(new PrivateLookupItemPickedEvent({ recordId }));
                        dispatcher.call(lookupElem, event);
                    };
                    if (lookupIsNavigable) {
                        lookupElem.setAttribute('data-navigation', 'enable');
                    }
                    else {
                        lookupElem.removeAttribute('data-navigation');
                    }
                });
            }
        }
    }
    updateUrlCellsLinkInfo() {
        if (features.enableAssertContext260Round2) {
            assertContext(this);
        }
        const { state, template } = this;
        const { shownUrlCells } = state;
        const { length: cellCount } = shownUrlCells;
        const urlElements = cellCount
            ? template.querySelectorAll('[data-inline-type="formatted-url"]')
            : [];
        for (let i = 0, { length } = urlElements; i < length; i += 1) {
            const cell = shownUrlCells[i];
            const urlElem = urlElements[i];
            updateRawLinkInfo(urlElem, {
                url: cell.urlLink,
                target: cell.urlTarget,
            }).then((linkInfo) => {
                const { dispatcher, url } = linkInfo;
                cell._handleUrlAnchorClick = (event) => {
                    dispatcher.call(urlElem, event);
                };
                cell.urlLink = url;
            });
        }
    }
}
