// lightning-base-components/src/lightning/datatable/fixedWidthStrategy.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { getColumnWidth, getTotalWidthsData, } from './widthManagerShared';
function getExpectedTableWidth(availableWidth, widthsData) {
    const { totalFixedWidth, totalResizedWidth, totalFlexibleColumns, minColumnWidth, } = widthsData;
    const minTotalFlexibleWidth = totalFlexibleColumns * minColumnWidth;
    const minExpectedTableWidth = minTotalFlexibleWidth + totalFixedWidth + totalResizedWidth;
    return widthsData.totalFlexibleColumns === 0
        ? minExpectedTableWidth
        : Math.max(minExpectedTableWidth, availableWidth);
}
export class FixedWidthStrategy {
    widthsData = {};
    constructor(minColumnWidth, maxColumnWidth) {
        this.widthsData = { minColumnWidth, maxColumnWidth };
    }
    set minColumnWidth(value) {
        this.widthsData.minColumnWidth = value;
    }
    set maxColumnWidth(value) {
        this.widthsData.maxColumnWidth = value;
    }
    getAdjustedColumnWidths(datatableInterface, _columns) {
        const totalWidthsData = getTotalWidthsData(this.widthsData, _columns);
        const availableWidth = datatableInterface.getAvailableWidthFromDom();
        const expectedTableWidth = getExpectedTableWidth(availableWidth, totalWidthsData);
        const expectedFlexibleColumnWidth = this.getFlexibleColumnWidth(totalWidthsData, expectedTableWidth);
        const columnWidths = [];
        for (let colIndex = 0; colIndex < _columns.length; colIndex += 1) {
            const width = getColumnWidth(_columns[colIndex]) ||
                expectedFlexibleColumnWidth;
            columnWidths[colIndex] = width;
        }
        return { columnWidths };
    }
    getFlexibleColumnWidth(widthsData, totalTableWidth) {
        const { totalFixedWidth, totalResizedWidth, totalFlexibleColumns, minColumnWidth, maxColumnWidth, } = widthsData;
        const totalFlexibleWidth = totalTableWidth - totalFixedWidth - totalResizedWidth;
        const avgFlexibleColumnWidth = Math.floor(totalFlexibleWidth / totalFlexibleColumns);
        const allowedSpace = Math.max(avgFlexibleColumnWidth, minColumnWidth);
        return Math.min(maxColumnWidth, allowedSpace);
    }
}
