// lightning-base-components/src/lightning/datatable/headerActions.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { unwrap } from 'lwc';
import { getUserColumnIndex } from './columns';
import { getInternalActions, handleTriggeredAction } from './wrapText';
const ACTION_REM_HEIGHT = 2.125;
const DIVIDER_REM_HEIGHT = 1.0625;
export function updateHeaderInternalActions(state) {
    const { columns } = state;
    const { length: colCount } = columns;
    for (let colIndex = 0; colIndex < colCount; colIndex += 1) {
        const col = columns[colIndex];
        col.actions.internalActions = getInternalActions(col);
    }
}
export function handleHeaderActionTriggered(event) {
    event.stopPropagation();
    const { action, actionType, colKeyValue } = event.detail;
    if (actionType !== 'customer') {
        handleTriggeredAction(this, action, colKeyValue);
    }
    dispatchHeaderActionEvent(this, action, colKeyValue);
}
export function handleHeaderActionMenuOpening(event) {
    event.stopPropagation();
    event.preventDefault();
    const actionsHeight = event.detail.actionsCount * ACTION_REM_HEIGHT;
    const dividersHeight = event.detail.dividersCount * DIVIDER_REM_HEIGHT;
    const wrapperHeight = 1;
    const totalRemHeight = actionsHeight + dividersHeight + wrapperHeight;
    this._actionsMinHeightStyle = `min-height: ${totalRemHeight}rem`;
    event.detail.saveContainerPosition(this.getViewableRect());
}
export function handleHeaderActionMenuClosed() {
    this._actionsMinHeightStyle = '';
}
function dispatchHeaderActionEvent(dt, action, colKeyValue) {
    const userColumnIndex = getUserColumnIndex(dt.state, colKeyValue);
    const customerColumnDefinition = dt.columns[userColumnIndex];
    dt.dispatchEvent(new CustomEvent('headeraction', {
        detail: {
            action: unwrap(action),
            columnDefinition: unwrap(customerColumnDefinition),
        },
    }));
}
