// lightning-base-components/src/lightning/datatable/infiniteLoading.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { normalizeBoolean } from 'lightning/utilsPrivate';
import { getScrollOffsetFromTableEnd, getScrollerY, isNonNegativeInteger, } from './utils';
const SCROLL_ALLOWANCE = 2;
export const DEFAULT_LOAD_MORE_OFFSET = 20;
export function setLoading(state, value) {
    state.isLoading = normalizeBoolean(value);
}
export function setEnableInfiniteLoading(state, value) {
    state.enableInfiniteLoading = normalizeBoolean(value);
}
export function getLoadMoreOffset(state) {
    return state.loadMoreOffset;
}
export function setLoadMoreOffset(state, value) {
    if (!isNonNegativeInteger(value)) {
        console.warn(`The "loadMoreOffset" value passed into lightning:datatable is incorrect. "loadMoreOffset" value should be an integer >= 0.`);
    }
    state.loadMoreOffset = isNonNegativeInteger(value)
        ? parseInt(value, 10)
        : DEFAULT_LOAD_MORE_OFFSET;
}
export function handleLoadMoreCheck(event) {
    const { state } = this;
    if (state.isLoading) {
        return;
    }
    const contentContainer = event.target
        .firstChild;
    if (!contentContainer) {
        return;
    }
    const offset = getScrollOffsetFromTableEnd(contentContainer);
    const threshold = getLoadMoreOffset(state);
    if (offset < threshold) {
        this.dispatchEvent(new CustomEvent('loadmore'));
    }
}
export function handlePrefetch() {
    const { state, template, validRefs } = this;
    if (state.isLoading ||
        !state.enableInfiniteLoading ||
        (state.enableViewportRendering &&
            this._renderManager &&
            !this._renderManager.hasWrapperHeight()) ||
        !state.rows.length) {
        return;
    }
    const scrollerY = !import.meta.env.SSR
        ? getScrollerY(template, validRefs)
        : null;
    if (scrollerY && isScrollerVisible(scrollerY) && !isScrollable(scrollerY)) {
        this.dispatchEvent(new CustomEvent('loadmore'));
    }
}
function isScrollable(scroller) {
    return scroller.scrollHeight > scroller.clientHeight + SCROLL_ALLOWANCE;
}
function isScrollerVisible(scroller) {
    return !!(scroller.offsetParent &&
        scroller.offsetHeight > 1 &&
        scroller.offsetWidth > 1);
}
