// lightning-base-components/src/lightning/datatable/inlineEdit.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { startPositioning, stopPositioning, Direction as DIRECTION, } from 'lightning/positionLibrary';
import { setFocusActiveCell, reactToTabBackward, reactToTabForward, getActiveCellElement, isActiveCellEditable, NAVIGATION_DIR, getCellElementFromEventTarget, } from './keyboard';
import { updateRowsAndCells } from './rows';
import { getStateColumnIndex } from './columns';
import { resetErrors } from './errors';
import { setAriaSelectedOnCell, unsetAriaSelectedOnCell } from './rowSelection';
import { getCurrentSelectionLength, getRowsKeys, getSelectedRowsKeys, isSelectedRow, } from './rowSelectionShared';
import { isObjectLike } from './utils';
import { hasOwnProperties, isUndefinedOrNull } from 'lightning/utilsPrivate';
import { getCellByKeys, getUserRowByKey, isValidCell } from './indexes';
import { features } from 'lightning/features';
const IEDIT_PANEL_SELECTOR = '[data-iedit-panel="true"]';
const HIDE_PANEL_THRESHOLD = 5;
export function handleEditCell(event) {
    const cellElement = getCellElementFromEventTarget(event.target);
    openInlineEdit(this, cellElement);
}
export function handleInlineEditFinish(event) {
    stopPanelPositioning(this);
    const { reason, rowKeyValue, colKeyValue } = event.detail;
    processInlineEditFinish(this, reason, rowKeyValue, colKeyValue);
    this._draftValues = this.draftValues;
}
export function handleMassCheckboxChange(event) {
    const state = this.state;
    if (event.detail.checked) {
        setAriaSelectedOnAllSelectedRows(state);
    }
    else {
        const { inlineEdit: inlineEditState } = state;
        unsetAriaSelectedOnAllSelectedRows(this.state);
        setAriaSelectedOnCell(state, inlineEditState.rowKeyValue, inlineEditState.colKeyValue);
    }
}
export function handleInlineEditPanelScroll(event) {
    const { state } = this;
    const { inlineEdit: inlineEditState } = state;
    const { isPanelVisible } = inlineEditState;
    if (!isPanelVisible) {
        return;
    }
    let delta = 0;
    const { target: scroller } = event;
    if (scroller.classList.contains('slds-scrollable_y')) {
        const scrollY = scroller.scrollTop;
        if (this._lastScrollY == null) {
            this._lastScrollY = scrollY;
        }
        else if (scroller.scrollHeight > scroller.offsetHeight) {
            delta = Math.abs(this._lastScrollY - scrollY);
        }
    }
    else {
        const scrollX = scroller.scrollLeft;
        if (this._lastScrollX == null) {
            this._lastScrollX = scrollX;
        }
        else if (scroller.scrollWidth > scroller.offsetWidth) {
            delta = Math.abs(this._lastScrollX - scrollX);
        }
    }
    if (delta > HIDE_PANEL_THRESHOLD) {
        const { rowKeyValue, colKeyValue } = inlineEditState;
        this._lastScrollX = null;
        this._lastScrollY = null;
        stopPanelPositioning(this);
        processInlineEditFinish(this, 'lost-focus', rowKeyValue, colKeyValue);
    }
    else {
        repositionPanel(this);
    }
}
function dispatchCellChangeEvent(dt, changes) {
    dt.dispatchEvent(new CustomEvent('cellchange', {
        detail: {
            draftValues: getResolvedCellChanges(dt.state, changes),
        },
    }));
}
export function isInlineEditTriggered(state) {
    return hasOwnProperties(state.inlineEdit.dirtyValues);
}
export function cancelInlineEdit(dt) {
    const { state, _privateTypes } = dt;
    resetErrors(state);
    state.inlineEdit.dirtyValues = {};
    dt._draftValues = [];
    updateRowsAndCells(state, _privateTypes);
}
export function closeInlineEdit(dt) {
    const { inlineEdit: inlineEditState } = dt.state;
    if (inlineEditState.isPanelVisible) {
        processInlineEditFinish(dt, 'lost-focus', inlineEditState.rowKeyValue, inlineEditState.colKeyValue);
    }
}
function processInlineEditFinish(dt, reason, rowKeyValue, colKeyValue) {
    const { state, template, refs } = dt;
    const { inlineEdit: inlineEditState } = state;
    const shouldSaveData = isValidCell(state, rowKeyValue, colKeyValue) &&
        reason !== 'edit-canceled' &&
        !(reason === 'lost-focus' && inlineEditState.massEditEnabled);
    if (shouldSaveData) {
        const panel = template.querySelector(IEDIT_PANEL_SELECTOR);
        const isValidEditValue = panel.validity.valid;
        const currentValue = getCellByKeys(state, rowKeyValue, colKeyValue).value;
        const { isMassEditChecked: updateAllSelectedRows, value: editValue } = panel;
        if (isValidEditValue &&
            (editValue !== currentValue || updateAllSelectedRows)) {
            const changes = {
                [rowKeyValue]: {
                    [colKeyValue]: editValue,
                },
            };
            if (updateAllSelectedRows) {
                let selectedRowsKeys;
                if (features.enableDisabledRowsAttribute) {
                    selectedRowsKeys = getRowsKeys(state.selectedRowsKeys);
                }
                else {
                    selectedRowsKeys = getSelectedRowsKeys(state);
                }
                for (let i = 0; i < selectedRowsKeys.length; i += 1) {
                    const otherRowKeyValue = selectedRowsKeys[i];
                    changes[otherRowKeyValue] = {
                        [colKeyValue]: editValue,
                    };
                }
            }
            updateDirtyValues(state, changes);
            dispatchCellChangeEvent(dt, changes);
            updateRowsAndCells(state, dt._privateTypes);
        }
    }
    if (reason !== 'lost-focus') {
        if (reason === 'tab-pressed-next') {
            reactToTabForward(state, template, refs);
        }
        else if (reason === 'tab-pressed-prev') {
            reactToTabBackward(state, template, refs);
        }
        else {
            setFocusActiveCell(state, template, refs, NAVIGATION_DIR.USE_CURRENT);
        }
    }
    unsetAriaSelectedOnAllSelectedRows(state);
    unsetAriaSelectedOnCell(state, rowKeyValue, colKeyValue);
    inlineEditState.isPanelVisible = false;
}
function openInlineEdit(dt, cellElement) {
    startPanelPositioning(dt, cellElement);
    const { state, template, _privateTypes: types } = dt;
    const { inlineEdit: inlineEditState } = state;
    if (inlineEditState.isPanelVisible) {
        processInlineEditFinish(dt, 'lost-focus', inlineEditState.rowKeyValue, inlineEditState.colKeyValue);
    }
    const { rowKeyValue } = cellElement.parentElement.dataset;
    const { colKeyValue } = cellElement.dataset;
    const cell = getCellByKeys(state, rowKeyValue, colKeyValue);
    const { colIndex } = cell;
    const col = state.columns[colIndex];
    inlineEditState.isPanelVisible = true;
    inlineEditState.rowKeyValue = rowKeyValue;
    inlineEditState.colKeyValue = colKeyValue;
    inlineEditState.columnDef = col;
    inlineEditState.editedValue = cell.value;
    inlineEditState.customErrorValidationMessage =
        cell.customErrorValidationMessage;
    inlineEditState.massEditSelectedRows = getCurrentSelectionLength(state);
    inlineEditState.massEditEnabled =
        inlineEditState.massEditSelectedRows > 1 &&
            isSelectedRow(state, rowKeyValue);
    const { typeAttributes } = col || {};
    if (typeAttributes) {
        inlineEditState.resolvedTypeAttributes = resolveNestedTypeAttributes(state, rowKeyValue, colIndex, types, typeAttributes);
    }
    setAriaSelectedOnCell(state, rowKeyValue, colKeyValue);
    setTimeout(() => {
        const panel = template.querySelector('lightning-primitive-datatable-iedit-panel');
        if (!panel.isEditableValid) {
            processInlineEditFinish(dt, 'edit-canceled', inlineEditState.rowKeyValue, inlineEditState.colKeyValue);
        }
        else {
            panel.focus();
        }
    }, 0);
}
export function openInlineEditOnActiveCell(dt) {
    const { state } = dt;
    if (state.data && state.data.length > 0) {
        if (!isActiveCellEditable(state)) {
            const firstEditableCell = getFirstEditableCell(dt);
            if (firstEditableCell) {
                state.activeCell = firstEditableCell;
                setFocusAndOpenInlineEdit(dt);
            }
        }
        else {
            setFocusAndOpenInlineEdit(dt);
        }
    }
}
async function setFocusAndOpenInlineEdit(dt) {
    const { state, template, refs } = dt;
    await setFocusActiveCell(state, template, refs, NAVIGATION_DIR.USE_CURRENT);
    const cellElement = getActiveCellElement(template, state);
    openInlineEdit(dt, cellElement);
}
function startPanelPositioning(dt, cellElement) {
    requestAnimationFrame(() => {
        stopPanelPositioning(dt);
        dt._positionRelationship = startPositioning(dt, {
            target: cellElement,
            element() {
                const panel = dt.template.querySelector(IEDIT_PANEL_SELECTOR);
                return panel.getPositionedElement();
            },
            autoFlip: true,
            flipInOverlay: true,
            align: {
                horizontal: DIRECTION.Left,
                vertical: DIRECTION.Top,
            },
            targetAlign: {
                horizontal: DIRECTION.Left,
                vertical: DIRECTION.Top,
            },
        });
    });
}
function stopPanelPositioning(dt) {
    const { _positionRelationship } = dt;
    if (_positionRelationship) {
        stopPositioning(_positionRelationship);
        dt._positionRelationship = undefined;
    }
}
function repositionPanel(dt) {
    requestAnimationFrame(() => {
        const { _positionRelationship } = dt;
        if (_positionRelationship) {
            _positionRelationship.reposition();
        }
    });
}
export function getDirtyValues(state) {
    return getResolvedCellChanges(state, state.inlineEdit.dirtyValues);
}
export function setDirtyValues(state, values) {
    const { columns, keyField, maxEditLimit } = state;
    let dirtyValues = Array.isArray(values) ? values : [];
    if (features.enableMaxEditLimit &&
        !isUndefinedOrNull(maxEditLimit) &&
        dirtyValues.length > maxEditLimit) {
        console.warn(`The number of edited rows (${dirtyValues.length}) exceeds the maxEditLimit (${maxEditLimit}). Only the first ${maxEditLimit} rows will be processed.`);
        dirtyValues = dirtyValues.slice(0, maxEditLimit);
    }
    const result = {};
    for (let dirtyIndex = 0; dirtyIndex < dirtyValues.length; dirtyIndex += 1) {
        const rowValues = dirtyValues[dirtyIndex];
        const colKeys = Object.keys(rowValues);
        const colChanges = {};
        for (let colKeysIndex = 0, { length: colKeysLength } = colKeys; colKeysIndex < colKeysLength; colKeysIndex += 1) {
            const externalColKeyValue = colKeys[colKeysIndex];
            for (let colIndex = 0; colIndex < columns.length; colIndex += 1) {
                const col = columns[colIndex];
                if (col.columnKey === externalColKeyValue ||
                    (!col.columnKey && col.fieldName === externalColKeyValue)) {
                    colChanges[col.colKeyValue] =
                        rowValues[externalColKeyValue];
                    break;
                }
            }
        }
        delete colChanges[keyField];
        result[rowValues[keyField]] = colChanges;
    }
    state.inlineEdit.dirtyValues = result;
}
export function trimDirtyValuesForEditLimit(state) {
    if (!features.enableMaxEditLimit) {
        return;
    }
    const { maxEditLimit } = state;
    if (isUndefinedOrNull(maxEditLimit)) {
        return;
    }
    const { dirtyValues } = state.inlineEdit;
    const rowKeys = Object.keys(dirtyValues);
    if (rowKeys.length > maxEditLimit) {
        console.warn(`The number of edited rows (${rowKeys.length}) exceeds the maxEditLimit (${maxEditLimit}). Only the first ${maxEditLimit} rows will be processed.`);
        const trimmedDirtyValues = {};
        for (let i = 0; i < maxEditLimit; i += 1) {
            const rowKey = rowKeys[i];
            trimmedDirtyValues[rowKey] = dirtyValues[rowKey];
        }
        state.inlineEdit.dirtyValues = trimmedDirtyValues;
    }
}
function updateDirtyValues(state, changes) {
    const dirtyValues = state.inlineEdit.dirtyValues;
    const rowKeys = Object.keys(changes);
    for (let rowKeyIndex = 0; rowKeyIndex < rowKeys.length; rowKeyIndex += 1) {
        const rowKeyValue = rowKeys[rowKeyIndex];
        if (!Object.prototype.hasOwnProperty.call(dirtyValues, rowKeyValue)) {
            dirtyValues[rowKeyValue] = {};
        }
        Object.assign(dirtyValues[rowKeyValue], changes[rowKeyValue]);
    }
}
function getResolvedCellChanges(state, rowChanges) {
    const { columns, keyField } = state;
    const result = [];
    const changeRowKeys = Object.keys(rowChanges);
    for (let changeRowIndex = 0, { length: changeRowCount } = changeRowKeys; changeRowIndex < changeRowCount; changeRowIndex += 1) {
        const rowKeyValue = changeRowKeys[changeRowIndex];
        const colChanges = rowChanges[rowKeyValue];
        const changeColKeys = Object.keys(colChanges);
        const { length: changeColCount } = changeColKeys;
        if (changeColCount) {
            const cellChanges = {};
            for (let changeColIndex = 0; changeColIndex < changeColCount; changeColIndex += 1) {
                const colKeyValue = changeColKeys[changeColIndex];
                const colIndex = getStateColumnIndex(state, colKeyValue);
                if (colIndex !== -1) {
                    const { name: colName } = columns[colIndex];
                    cellChanges[colName] = colChanges[colKeyValue];
                }
            }
            cellChanges[keyField] = rowKeyValue;
            result.push(cellChanges);
        }
    }
    return result;
}
export function resolveNestedTypeAttributes(state, rowKeyValue, colIndex, types, typeAttributes) {
    const col = state.columns[colIndex];
    const typeDesc = typeAttributes && types.getType(col.type);
    const attributeNames = typeDesc && typeDesc.typeAttributes;
    if (attributeNames) {
        const resolvedTypeAttributes = {};
        const { length: attributeNamesLength } = attributeNames;
        const _rowData = attributeNamesLength
            ? getUserRowByKey(state, rowKeyValue)
            : undefined;
        if (_rowData) {
            for (let i = 0; i < attributeNamesLength; i += 1) {
                const attrName = attributeNames[i];
                const attrValue = typeAttributes[attrName];
                if (attrValue) {
                    resolvedTypeAttributes[attrName] =
                        resolveNestedTypeAttributesHelper(_rowData, attrValue);
                }
            }
        }
        return resolvedTypeAttributes;
    }
    return typeAttributes;
}
function resolveNestedTypeAttributesHelper(rowData, attrValue) {
    let resolvedTypeAttributes = {};
    if (!isObjectLike(attrValue)) {
        return attrValue === undefined ? resolvedTypeAttributes : attrValue;
    }
    if (Array.isArray(attrValue)) {
        return attrValue.map((value) => resolveNestedTypeAttributesHelper(rowData, value));
    }
    const typeAttributesNames = Object.keys(attrValue);
    for (let i = 0, { length } = typeAttributesNames; i < length; i += 1) {
        const name = typeAttributesNames[i];
        const value = attrValue[name];
        if (value !== undefined) {
            if (name === 'fieldName') {
                resolvedTypeAttributes = rowData[value];
            }
            else if (isObjectLike(value)) {
                const { fieldName } = value;
                if (fieldName) {
                    resolvedTypeAttributes[name] = rowData[fieldName];
                }
                else {
                    resolvedTypeAttributes[name] =
                        resolveNestedTypeAttributesHelper(rowData, value);
                }
            }
            else {
                resolvedTypeAttributes[name] = value;
            }
        }
    }
    return resolvedTypeAttributes;
}
function getFirstEditableCell(dt) {
    const { state } = dt;
    const { columns } = state;
    const editableColumns = [];
    for (let i = 0, { length } = columns; i < length; i += 1) {
        const col = columns[i];
        if (col.editable) {
            editableColumns.push(col);
        }
    }
    const { length: editableColumnsLength } = editableColumns;
    if (editableColumnsLength > 0) {
        const { rows } = state;
        for (let rowIndex = 0; rowIndex < rows.length; rowIndex += 1) {
            const row = rows[rowIndex];
            const { key: rowKeyValue } = row;
            for (let i = 0; i < editableColumnsLength; i += 1) {
                const { colKeyValue } = editableColumns[i];
                const cell = getCellByKeys(state, rowKeyValue, colKeyValue);
                if (cell && cell.editable) {
                    return cell;
                }
            }
        }
    }
    return undefined;
}
function setAriaSelectedOnAllSelectedRows(state) {
    const { colKeyValue } = state.inlineEdit;
    const { selectedRowsKeys } = state;
    const keys = Object.keys(selectedRowsKeys);
    for (let i = 0, { length } = keys; i < length; i += 1) {
        const rowKeyValue = keys[i];
        if (selectedRowsKeys[rowKeyValue]) {
            setAriaSelectedOnCell(state, rowKeyValue, colKeyValue);
        }
    }
}
function unsetAriaSelectedOnAllSelectedRows(state) {
    const { colKeyValue } = state.inlineEdit;
    const { selectedRowsKeys } = state;
    const keys = Object.keys(selectedRowsKeys);
    for (let i = 0, { length } = keys; i < length; i += 1) {
        const rowKeyValue = keys[i];
        if (selectedRowsKeys[rowKeyValue]) {
            unsetAriaSelectedOnCell(state, rowKeyValue, colKeyValue);
        }
    }
}
