// lightning-base-components/src/lightning/datatable/keyboard.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { getShadowActiveElements, isRTL, keyCodes, } from 'lightning/utilsPrivate';
import { getCellByKeys, getCellFromIndexes, getRowIndexByKey, HEADER_ROW_KEY, isValidCell, } from './indexes';
import { fireRowToggleEvent } from './tree';
import { findFirstVisibleIndex } from './virtualization';
import { getColDataSelector, getRowDataSelector, getScrollerX, getScrollerY, resetHighlightedRowClasses, unsetHighlightedRowClasses, } from './utils';
import { features } from 'lightning/features';
export const FOCUS_CLASS = 'slds-has-focus';
export const KEYBOARD_NAVIGATION_MODE = 'NAVIGATION';
export const KEYBOARD_ACTION_MODE = 'ACTION';
const TOP_MARGIN = 80;
const BOTTOM_MARGIN = 80;
const SCROLL_OFFSET = 20;
const ARROW_RIGHT = 39;
const ARROW_LEFT = 37;
const ARROW_DOWN = 40;
const ARROW_UP = 38;
const ENTER = 13;
const ESCAPE = 27;
const TAB = 9;
const SPACE = 32;
export const NAVIGATION_DIR = isRTL()
    ? {
        RIGHT: -1,
        LEFT: 1,
        USE_CURRENT: 0,
        RESET: 2,
        TAB_FORWARD: -1,
        TAB_BACKWARD: 1,
    }
    : {
        RIGHT: 1,
        LEFT: -1,
        USE_CURRENT: 0,
        RESET: 2,
        TAB_FORWARD: 1,
        TAB_BACKWARD: -1,
    };
export function handleKeydownOnCell(event) {
    event.stopPropagation();
    reactToKeyboardInActionMode(this, event);
}
export function handleKeydownOnTable(event) {
    if (isCellElement(event.target)) {
        reactToKeyboardInNavMode(this, event);
    }
}
function reactToKeyboardInActionMode(dt, event) {
    const { refs, state, template } = dt;
    switch (event.detail.keyCode) {
        case ARROW_LEFT:
            reactToArrowLeft(state, template, refs, event);
            break;
        case ARROW_RIGHT:
            reactToArrowRight(state, template, refs, event);
            break;
        case ARROW_UP:
            reactToArrowUp(state, template, refs, event);
            break;
        case ARROW_DOWN:
            reactToArrowDown(state, template, refs, event);
            break;
        case ENTER:
        case SPACE:
            reactToEnter(state, template, refs, event);
            break;
        case ESCAPE:
            reactToEscape(state, template, refs, event);
            break;
        case TAB:
            reactToTab(state, template, refs, event);
            break;
        default:
            break;
    }
}
function reactToKeyboardInNavMode(dt, event) {
    const { refs, state, template } = dt;
    const { activeCell } = state;
    if (!activeCell) {
        return;
    }
    const { colKeyValue, rowKeyValue } = activeCell;
    const { keyCode, shiftKey } = event;
    const syntheticEvent = {
        detail: {
            colKeyValue,
            rowKeyValue,
            keyCode,
            shiftKey,
        },
        preventDefault: () => { },
        stopPropagation: () => { },
    };
    switch (event.keyCode) {
        case ARROW_LEFT:
            event.preventDefault();
            reactToArrowLeft(state, template, refs, syntheticEvent);
            break;
        case ARROW_RIGHT:
            event.preventDefault();
            reactToArrowRight(state, template, refs, syntheticEvent);
            break;
        case ARROW_UP:
            event.preventDefault();
            reactToArrowUp(state, template, refs, syntheticEvent);
            break;
        case ARROW_DOWN:
            event.preventDefault();
            reactToArrowDown(state, template, refs, syntheticEvent);
            break;
        case ENTER:
        case SPACE:
            event.preventDefault();
            reactToEnter(state, template, refs, syntheticEvent);
            break;
        case ESCAPE:
            syntheticEvent.detail.keyEvent = event;
            reactToEscape(state, template, refs, syntheticEvent);
            break;
        case TAB:
            reactToTab(state, template, refs, syntheticEvent);
            break;
        default:
            break;
    }
}
function reactToArrowLeft(state, template, refs, event) {
    const { rowKeyValue, colKeyValue } = event.detail;
    const { colIndex } = getCellByKeys(state, rowKeyValue, colKeyValue);
    if (colIndex === 0 && canBeRowNavigationMode(state)) {
        setBlurActiveCell(state, template);
        setRowNavigationMode(state);
        setFocusActiveRow(state, template, refs);
    }
    else {
        const nextColIndex = getNextIndexLeft(state, colIndex);
        if (nextColIndex === undefined) {
            return;
        }
        setBlurActiveCell(state, template);
        state.activeCell = getCellByKeys(state, rowKeyValue, state.columns[nextColIndex].colKeyValue);
        setFocusActiveCell(state, template, refs, NAVIGATION_DIR.LEFT);
    }
}
function reactToArrowRight(state, template, refs, event) {
    const { rowKeyValue, colKeyValue } = event.detail;
    const { colIndex } = getCellByKeys(state, rowKeyValue, colKeyValue);
    const nextColIndex = getNextIndexRight(state, colIndex);
    if (nextColIndex === undefined) {
        return;
    }
    setBlurActiveCell(state, template);
    state.activeCell = getCellByKeys(state, rowKeyValue, state.columns[nextColIndex].colKeyValue);
    setFocusActiveCell(state, template, refs, NAVIGATION_DIR.RIGHT);
}
function reactToArrowUp(state, template, refs, event) {
    const { rowKeyValue, colKeyValue, keyEvent } = event.detail;
    const { rowIndex } = getCellByKeys(state, rowKeyValue, colKeyValue);
    const nextRowIndex = getNextIndexUp(state, rowIndex);
    if (nextRowIndex === undefined ||
        (state.hideTableHeader && nextRowIndex === -1)) {
        return;
    }
    if (keyEvent) {
        keyEvent.stopPropagation();
    }
    setBlurActiveCell(state, template);
    state.activeCell = getCellByKeys(state, nextRowIndex === -1 ? HEADER_ROW_KEY : state.rows[nextRowIndex].key, colKeyValue);
    setFocusActiveCell(state, template, refs, NAVIGATION_DIR.USE_CURRENT);
}
function reactToArrowDown(state, template, refs, event) {
    const { rowKeyValue, colKeyValue, keyEvent } = event.detail;
    const { rowIndex } = getCellByKeys(state, rowKeyValue, colKeyValue);
    const nextRowIndex = getNextIndexDown(state, rowIndex);
    if (nextRowIndex === undefined ||
        (state.hideTableHeader && nextRowIndex === -1)) {
        return;
    }
    if (keyEvent) {
        keyEvent.stopPropagation();
    }
    setBlurActiveCell(state, template);
    state.activeCell = getCellByKeys(state, nextRowIndex === -1 ? HEADER_ROW_KEY : state.rows[nextRowIndex].key, colKeyValue);
    setFocusActiveCell(state, template, refs, NAVIGATION_DIR.USE_CURRENT);
}
function reactToEnter(state, template, refs, event) {
    if (state.keyboardMode === KEYBOARD_NAVIGATION_MODE) {
        state.keyboardMode = KEYBOARD_ACTION_MODE;
        const { keyCode, keyEvent } = event.detail;
        const { rowIndex, colIndex } = state.activeCell;
        if (keyEvent) {
            keyEvent.preventDefault();
        }
        const info = { action: undefined };
        if (keyCode === SPACE) {
            info.action = 'space';
        }
        else if (keyCode === ENTER) {
            info.action = 'enter';
        }
        setModeActiveCell(state, template, info);
        updateCellTabIndex(state, rowIndex, colIndex, -1);
    }
}
function reactToEscape(state, template, refs, event) {
    if (state.keyboardMode === KEYBOARD_ACTION_MODE) {
        event.detail.keyEvent.stopPropagation();
        state.keyboardMode = KEYBOARD_NAVIGATION_MODE;
        setModeActiveCell(state, template);
        setFocusActiveCell(state, template, refs, NAVIGATION_DIR.RESET);
    }
}
function reactToTab(state, template, refs, event) {
    event.preventDefault();
    event.stopPropagation();
    const { detail } = event;
    const direction = getTabDirection(detail.shiftKey);
    const isExitCell = isActiveCellAnExitCell(state, direction);
    if (state.keyboardMode === KEYBOARD_ACTION_MODE) {
        if (isExitCell === false) {
            const { keyEvent } = detail;
            if (keyEvent) {
                keyEvent.preventDefault();
            }
            if (direction === 'BACKWARD') {
                reactToTabBackward(state, template, refs);
            }
            else {
                reactToTabForward(state, template, refs);
            }
        }
        else {
            state.keyboardMode = KEYBOARD_NAVIGATION_MODE;
            setModeActiveCell(state, template);
            state.isExitingActionMode = true;
        }
    }
    else {
        state.isExitingActionMode = true;
    }
}
export function reactToTabForward(state, template, refs) {
    const { nextColIndex, nextRowIndex } = getNextIndexOnTab(state, 'FORWARD');
    setBlurActiveCell(state, template);
    state.activeCell = getCellByKeys(state, nextRowIndex === -1 ? HEADER_ROW_KEY : state.rows[nextRowIndex].key, state.columns[nextColIndex].colKeyValue);
    setFocusActiveCell(state, template, refs, NAVIGATION_DIR.TAB_FORWARD, {
        action: 'tab',
    });
}
export function reactToTabBackward(state, template, refs) {
    const { nextColIndex, nextRowIndex } = getNextIndexOnTab(state, 'BACKWARD');
    setBlurActiveCell(state, template);
    state.activeCell = getCellByKeys(state, nextRowIndex === -1 ? HEADER_ROW_KEY : state.rows[nextRowIndex].key, state.columns[nextColIndex].colKeyValue);
    setFocusActiveCell(state, template, refs, NAVIGATION_DIR.TAB_BACKWARD, {
        action: 'tab',
    });
}
function getTabDirection(shiftKey) {
    return shiftKey ? 'BACKWARD' : 'FORWARD';
}
function getNextIndexOnTab(state, direction = 'FORWARD') {
    const { rowIndex, colIndex } = state.activeCell;
    return direction === 'BACKWARD'
        ? getNextIndexOnTabBackward(state, rowIndex, colIndex)
        : getNextIndexOnTabForward(state, rowIndex, colIndex);
}
function getNextIndexOnTabForward(state, rowIndex, colIndex) {
    const columnsCount = state.columns.length;
    if (columnsCount > colIndex + 1) {
        return {
            nextRowIndex: rowIndex,
            nextColIndex: colIndex + 1,
        };
    }
    return {
        nextRowIndex: getNextIndexDownWrapped(state, rowIndex),
        nextColIndex: 0,
    };
}
function getNextIndexOnTabBackward(state, rowIndex, colIndex) {
    const columnsCount = state.columns.length;
    if (colIndex > 0) {
        return {
            nextRowIndex: rowIndex,
            nextColIndex: colIndex - 1,
        };
    }
    return {
        nextRowIndex: getNextIndexUpWrapped(state, rowIndex),
        nextColIndex: columnsCount - 1,
    };
}
export function reactToKeyboardOnRow(dt, event) {
    const { refs, state, template } = dt;
    const { localName } = event.target;
    if (localName &&
        localName.indexOf('tr') !== -1 &&
        isRowNavigationMode(state)) {
        switch (event.detail.keyCode) {
            case ARROW_LEFT:
                reactToArrowLeftOnRow(dt, state, template, refs, event);
                break;
            case ARROW_RIGHT:
                reactToArrowRightOnRow(dt, state, template, refs, event);
                break;
            case ARROW_UP:
                reactToArrowUpOnRow(state, template, refs, event);
                break;
            case ARROW_DOWN:
                reactToArrowDownOnRow(state, template, refs, event);
                break;
            default:
                break;
        }
    }
}
function reactToArrowLeftOnRow(dt, state, template, refs, event) {
    const { rowKeyValue, rowHasChildren, rowExpanded, rowLevel } = event.detail;
    if (rowHasChildren && rowExpanded) {
        fireRowToggleEvent(dt, rowKeyValue, rowExpanded);
    }
    else if (rowLevel > 1) {
        const { treeColumn } = state;
        if (treeColumn) {
            const { colKeyValue } = treeColumn;
            const { rowIndex } = getCellByKeys(state, rowKeyValue, colKeyValue);
            const parentIndex = getRowParentIndex(state, rowLevel, rowIndex);
            if (parentIndex !== -1) {
                setBlurActiveRow(state, template);
                state.activeCell = getCellByKeys(state, state.rows[parentIndex].key, colKeyValue);
                setFocusActiveRow(state, template, refs);
            }
        }
    }
}
function reactToArrowRightOnRow(dt, state, template, refs, event) {
    const { rowKeyValue, rowHasChildren, rowExpanded } = event.detail;
    if (rowHasChildren && !rowExpanded) {
        fireRowToggleEvent(dt, rowKeyValue, rowExpanded);
    }
    else {
        setBlurActiveRow(state, template);
        unsetRowNavigationMode(state);
        if (features.enableTreeGridSecondArrKeyUpdate) {
            const rowIndex = getRowIndexByKey(state, rowKeyValue);
            state.activeCell = getCellByKeys(state, state.rows[rowIndex].key, state.columns[0].colKeyValue);
        }
        setFocusActiveCell(state, template, refs, NAVIGATION_DIR.USE_CURRENT);
    }
}
function reactToArrowUpOnRow(state, template, refs, event) {
    const { rowKeyValue, keyEvent } = event.detail;
    const { treeColumn } = state;
    keyEvent.stopPropagation();
    keyEvent.preventDefault();
    if (treeColumn) {
        const { colKeyValue } = treeColumn;
        const { rowIndex } = getCellByKeys(state, rowKeyValue, colKeyValue);
        const prevRowIndex = getNextIndexUpWrapped(state, rowIndex);
        if (prevRowIndex !== -1) {
            setBlurActiveRow(state, template);
            state.activeCell = getCellByKeys(state, state.rows[prevRowIndex].key, colKeyValue);
            setFocusActiveRow(state, template, refs);
        }
    }
}
function reactToArrowDownOnRow(state, template, refs, event) {
    const { rowKeyValue, keyEvent } = event.detail;
    const { treeColumn } = state;
    keyEvent.stopPropagation();
    keyEvent.preventDefault();
    if (treeColumn) {
        const { colKeyValue } = treeColumn;
        const { rowIndex } = getCellByKeys(state, rowKeyValue, colKeyValue);
        const nextRowIndex = getNextIndexDownWrapped(state, rowIndex);
        if (nextRowIndex !== -1) {
            setBlurActiveRow(state, template);
            state.activeCell = getCellByKeys(state, state.rows[nextRowIndex].key, colKeyValue);
            setFocusActiveRow(state, template, refs);
        }
    }
}
function getDefaultActiveCell(state) {
    const { columns } = state;
    const { length: colCount } = columns;
    if (colCount) {
        let colIndex = 0;
        let existCustomerColumn = false;
        for (let i = 0; i < colCount; i += 1) {
            colIndex = i;
            if (columns[i].internal !== true) {
                existCustomerColumn = true;
                break;
            }
        }
        if (!existCustomerColumn) {
            colIndex = 0;
        }
        const { rows } = state;
        return getCellByKeys(state, rows.length > 0 ? rows[0].key : HEADER_ROW_KEY, columns[colIndex].colKeyValue);
    }
    return undefined;
}
function setDefaultActiveCell(state) {
    state.activeCell = getDefaultActiveCell(state);
}
export function getActiveCellElement(template, state) {
    const { activeCell } = state;
    return activeCell
        ? getCellElement(template, activeCell.rowKeyValue, activeCell.colKeyValue)
        : undefined;
}
export function isActiveCell(state, rowKeyValue, colKeyValue) {
    const { activeCell } = state;
    if (activeCell) {
        const { rowKeyValue: currentRowKeyValue, colKeyValue: currentColKeyValue, } = activeCell;
        return (currentRowKeyValue.toString() === rowKeyValue.toString() &&
            currentColKeyValue === colKeyValue);
    }
    return false;
}
export function syncActiveCell(state) {
    const { activeCell } = state;
    if (!state.hasReceivedFocus) {
        if (activeCell && stillValidActiveCell(state)) {
            const updatedActiveCell = state.activeCell;
            const { rowIndex, colIndex } = updatedActiveCell;
            updateCellTabIndex(state, rowIndex, colIndex, -1);
        }
        setDefaultActiveCell(state);
    }
    else if (!activeCell || !stillValidActiveCell(state)) {
        if (activeCell && state.cellToFocusNext) {
            setNextActiveCellFromPrev(state);
        }
        else {
            setDefaultActiveCell(state);
        }
    }
}
function setNextActiveCellFromPrev(state) {
    const { rowIndex, colIndex } = state.cellToFocusNext;
    const { columns, rows } = state;
    const { length: rowCount } = rows;
    let nextRowIndex = rowIndex;
    if (nextRowIndex > rowCount - 1) {
        nextRowIndex = rowCount - 1;
    }
    const { length: colCount } = columns;
    let nextColIndex = colIndex;
    if (nextColIndex > colCount - 1) {
        nextColIndex = colCount - 1;
    }
    const nextActiveCell = getCellFromIndexes(state, nextRowIndex, nextColIndex);
    if (nextActiveCell) {
        state.activeCell = nextActiveCell;
    }
    else {
        setDefaultActiveCell(state);
    }
    state.keyboardMode = KEYBOARD_NAVIGATION_MODE;
}
export function isActiveCellAnExitCell(state, direction) {
    const { rowIndex, colIndex } = state.activeCell;
    const { nextRowIndex, nextColIndex } = getNextIndexOnTab(state, direction);
    if ((rowIndex === -1 &&
        colIndex === 0 &&
        nextRowIndex !== -1 &&
        nextColIndex !== 0) ||
        (rowIndex !== -1 && nextRowIndex === -1 && nextColIndex === 0)) {
        return true;
    }
    return false;
}
function setModeActiveCell(state, template, info) {
    const cellElement = getActiveCellElement(template, state);
    const cellElementChild = getCellElementChild(cellElement);
    if (cellElementChild) {
        const { keyboardMode } = state;
        if (typeof cellElementChild.setMode ===
            'function') {
            cellElementChild.setMode(keyboardMode, info);
        }
        else {
            setMode(state.activeCell, cellElementChild, keyboardMode, info);
        }
    }
}
function stillValidActiveCell(state) {
    const { activeCell, indexes } = state;
    if (activeCell) {
        const indexedCell = indexes[activeCell.cellKeyValue];
        if (indexedCell) {
            state.activeCell = indexedCell;
        }
        return indexedCell !== undefined;
    }
    return false;
}
export function getCellElement(template, rowKeyValue, colKeyValue) {
    const cellElement = template.querySelector(`${getRowDataSelector(rowKeyValue)} ${getColDataSelector(colKeyValue)}`);
    return (cellElement || undefined);
}
export function getCellElementChild(cellElement) {
    const firstElementChild = cellElement
        ? cellElement.firstElementChild
        : undefined;
    return firstElementChild;
}
export function getCellElementFromEventTarget(target) {
    const cellElement = target.closest('[role="gridcell"],[role="columnheader"],[role="rowheader"]');
    return (cellElement || undefined);
}
function canMoveBackward(cell) {
    return cell.currentInputIndex > 0;
}
function canMoveForward(cell) {
    const { actionableElementsCount } = cell;
    return (actionableElementsCount > 1 &&
        cell.currentInputIndex < actionableElementsCount - 1);
}
function canMoveLeft(cell) {
    return isRTL() ? canMoveForward(cell) : canMoveBackward(cell);
}
function canMoveRight(cell) {
    return isRTL() ? canMoveBackward(cell) : canMoveForward(cell);
}
function fireCellKeydown(cell, cellElementChild, keyEvent) {
    const { rowKeyValue, colKeyValue } = cell;
    const { keyCode, shiftKey } = keyEvent;
    cellElementChild.dispatchEvent(new CustomEvent('privatecellkeydown', {
        bubbles: true,
        composed: true,
        cancelable: true,
        detail: {
            rowKeyValue,
            colKeyValue,
            keyCode,
            shiftKey,
            keyEvent,
        },
    }));
}
function fireCellFocusByClickEvent(cell, cellElementChild) {
    const { rowKeyValue, colKeyValue } = cell;
    cellElementChild.dispatchEvent(new CustomEvent('privatecellfocusedbyclick', {
        bubbles: true,
        composed: true,
        detail: {
            rowKeyValue,
            colKeyValue,
        },
    }));
}
export function handlePrimitiveDatatableCellClick(event) {
    const { state } = this;
    const cellElement = state.renderModeInline
        ? getCellElementFromEventTarget(event.target)
        : undefined;
    if (cellElement) {
        const cellElementChild = getCellElementChild(cellElement);
        const cell = getCellByKeys(state, cellElement.parentElement.dataset.rowKeyValue, cellElement.dataset.colKeyValue);
        fireCellFocusByClickEvent(cell, cellElementChild);
    }
}
export function handlePrimitiveDatatableCellFocus(event) {
    const { state } = this;
    const cellElement = state.renderModeInline
        ? getCellElementFromEventTarget(event.target)
        : undefined;
    if (cellElement) {
        const cellElementChild = getCellElementChild(cellElement);
        const cell = getCellByKeys(this.state, cellElement.parentElement.dataset.rowKeyValue, cellElement.dataset.colKeyValue);
        if (cell.keyboardMode === KEYBOARD_ACTION_MODE) {
            setFocusToActionableElement(cell, cellElementChild, cell.currentInputIndex);
        }
    }
}
export function handlePrimitiveDatatableCellKeydown(event) {
    const { state } = this;
    const { target } = event;
    const cellElement = state.renderModeInline &&
        target.tagName !== 'LIGHTNING-PRIMITIVE-CELL-FACTORY' &&
        !target.getAttribute('role')
        ? getCellElementFromEventTarget(event.target)
        : undefined;
    if (cellElement === undefined) {
        return;
    }
    const cellElementChild = getCellElementChild(cellElement);
    const cell = getCellByKeys(state, cellElement.parentElement.dataset.rowKeyValue, cellElement.dataset.colKeyValue);
    const { keyboardMode } = cell;
    const { keyCode, shiftKey } = event;
    if (keyboardMode === KEYBOARD_ACTION_MODE) {
        let passThroughEvent = keyCode !== keyCodes.shift;
        if (keyCode === keyCodes.left) {
            if (canMoveLeft(cell)) {
                event.preventDefault();
                if (isRTL()) {
                    moveToNextActionableElement(cell, cellElementChild);
                }
                else {
                    moveToPrevActionableElement(cell, cellElementChild);
                }
                passThroughEvent = false;
            }
        }
        else if (keyCode === keyCodes.right) {
            if (canMoveRight(cell)) {
                event.preventDefault();
                if (isRTL()) {
                    moveToPrevActionableElement(cell, cellElementChild);
                }
                else {
                    moveToNextActionableElement(cell, cellElementChild);
                }
                passThroughEvent = false;
            }
        }
        else if (keyCode === keyCodes.tab) {
            if (shiftKey) {
                if (canMoveBackward(cell)) {
                    event.preventDefault();
                    moveToPrevActionableElement(cell, cellElementChild);
                    passThroughEvent = false;
                }
            }
            else if (canMoveForward(cell)) {
                event.preventDefault();
                moveToNextActionableElement(cell, cellElementChild);
                passThroughEvent = false;
            }
        }
        if (passThroughEvent) {
            fireCellKeydown(cell, cellElementChild, event);
        }
    }
    else if (keyboardMode === KEYBOARD_NAVIGATION_MODE) {
        if (keyCode === keyCodes.left ||
            keyCode === keyCodes.right ||
            keyCode === keyCodes.up ||
            keyCode === keyCodes.down ||
            keyCode === keyCodes.enter) {
            fireCellKeydown(cell, cellElementChild, event);
        }
    }
}
function getActionableElements(cellElementChild) {
    let result = [];
    const customType = cellElementChild.querySelector('lightning-primitive-custom-cell');
    if (customType) {
        result = Array.from(customType.getActionableElements());
    }
    const defaultActions = Array.from(cellElementChild.querySelectorAll('[data-navigation="enable"]'));
    return result.concat(defaultActions);
}
function moveToNextActionableElement(cell, cellElementChild) {
    setFocusToActionableElement(cell, cellElementChild, cell.currentInputIndex + 1);
}
function moveToPrevActionableElement(cell, cellElementChild) {
    setFocusToActionableElement(cell, cellElementChild, cell.currentInputIndex - 1);
}
function resetCurrentInputIndex(cell, cellElementChild, direction, incomingMode) {
    if (direction === -1) {
        const { length } = getActionableElements(cellElementChild);
        cell.currentInputIndex = length ? length - 1 : 0;
    }
    else if (direction === 1 || direction === 2) {
        cell.currentInputIndex = 0;
    }
    if (incomingMode === KEYBOARD_ACTION_MODE &&
        cell.keyboardMode === KEYBOARD_ACTION_MODE) {
        setFocusToActionableElement(cell, cellElementChild, cell.currentInputIndex);
    }
}
function setFocusToActionableElement(cell, cellElementChild, activeInputIndex) {
    const actionableElements = getActionableElements(cellElementChild);
    const { length: actionableElementsLength } = actionableElements;
    cell.actionableElementsCount = actionableElementsLength;
    if (actionableElementsLength) {
        if (activeInputIndex > 0 &&
            activeInputIndex < actionableElementsLength) {
            actionableElements[activeInputIndex].focus();
            cell.currentInputIndex = activeInputIndex;
        }
        else {
            actionableElements[0].focus();
            cell.currentInputIndex = 0;
        }
    }
}
function setMode(cell, cellElementChild, keyboardMode, info) {
    const normalizedInfo = info || { action: 'none' };
    cell.keyboardMode = keyboardMode;
    if (keyboardMode === KEYBOARD_ACTION_MODE) {
        cell.internalTabIndex = 0;
        setTimeout(() => {
            setFocusToActionableElement(cell, cellElementChild, cell.currentInputIndex);
        }, 0);
        const actionableElements = getActionableElements(cellElementChild);
        const { length: actionableElementsLength } = actionableElements;
        if (!features.enableDTAutoInlineEditFix && normalizedInfo.action === 'tab') {
            for (let i = 0; i < actionableElementsLength; i += 1) {
                const elem = actionableElements[i];
                if (elem.getAttribute('data-action-edit')) {
                    elem.click();
                    break;
                }
            }
        }
        else if (normalizedInfo.action !== 'tab' && actionableElementsLength === 1) {
            const elem = actionableElements[0];
            const defaultActions = elem.getAttribute('data-action-triggers') || '';
            if (defaultActions.indexOf(normalizedInfo.action) !== -1) {
                elem.click();
            }
        }
    }
    else {
        cell.internalTabIndex = -1;
    }
}
export function setFocusActiveCell(state, template, refs, direction, info, shouldScroll = true) {
    const { keyboardMode } = state;
    const { rowIndex, colIndex } = state.activeCell;
    let cellElement = getActiveCellElement(template, state);
    state.activeCell.hasFocus = !(info && isActiveCellValid(state));
    updateCellTabIndex(state, rowIndex, colIndex);
    let promise;
    if (shouldScroll && !cellElement && isActiveCellValid(state)) {
        scrollToRow(state, template, refs, rowIndex);
        promise = waitForActiveCellRender(template, state);
    }
    else {
        promise = Promise.resolve();
    }
    const rowKeyValue = cellElement && cellElement.parentElement.dataset.rowKeyValue;
    return promise.then(() => {
        return new Promise((resolve) => {
            setTimeout(() => {
                if (!cellElement ||
                    !isValidCell(state, rowKeyValue, cellElement.dataset.colKeyValue)) {
                    cellElement = getActiveCellElement(template, state);
                }
                if (cellElement) {
                    const cellElementChild = getCellElementChild(cellElement);
                    const { activeCell } = state;
                    if (cellElementChild) {
                        if (direction) {
                            if (typeof cellElementChild.resetCurrentInputIndex === 'function') {
                                cellElementChild.resetCurrentInputIndex(direction, keyboardMode);
                            }
                            else {
                                resetCurrentInputIndex(activeCell, cellElement, direction, keyboardMode);
                            }
                        }
                        if (typeof cellElementChild
                            .addFocusStyles === 'function') {
                            cellElementChild.addFocusStyles();
                        }
                    }
                    cellElement.classList.add(FOCUS_CLASS);
                    cellElement.focus({ preventScroll: !shouldScroll });
                    if (cellElementChild) {
                        if (typeof cellElementChild
                            .setMode === 'function') {
                            cellElementChild.setMode(keyboardMode, info);
                        }
                        else {
                            setMode(activeCell, cellElementChild, keyboardMode, info);
                        }
                    }
                    resetHighlightedRowClasses(state, rowKeyValue);
                }
                resolve();
            }, 0);
        });
    });
}
function waitForActiveCellRender(template, state) {
    return new Promise((resolve) => {
        const start = Date.now();
        const interval = setInterval(() => {
            if (getActiveCellElement(template, state) ||
                start - Date.now() > 1000) {
                clearInterval(interval);
                resolve();
            }
        }, 0);
    });
}
export function setBlurActiveCell(state, template) {
    const { activeCell } = state;
    if (activeCell) {
        const { rowIndex, colIndex } = activeCell;
        state.activeCell.hasFocus = false;
        let cellElement = getActiveCellElement(template, state);
        const rowKeyValue = cellElement && cellElement.parentElement.dataset.rowKeyValue;
        setTimeout(() => {
            if (!cellElement ||
                !isValidCell(state, rowKeyValue, cellElement.dataset.colKeyValue)) {
                cellElement = getActiveCellElement(template, state);
            }
            if (cellElement) {
                const { activeElement } = document;
                const cellElementChild = getCellElementChild(cellElement);
                if (cellElementChild) {
                    if (typeof cellElementChild
                        .removeFocusStyles === 'function') {
                        if (activeElement === cellElementChild) {
                            cellElementChild.blur();
                        }
                        cellElementChild.removeFocusStyles(true);
                    }
                    else {
                        state.activeCell.internalTabIndex = -1;
                        if (cellElement.contains(activeElement)) {
                            cellElement.blur();
                        }
                    }
                }
                cellElement.classList.remove(FOCUS_CLASS);
            }
        }, 0);
        updateCellTabIndex(state, rowIndex, colIndex, -1);
    }
}
export function setCellToFocusFromPrev(state, template) {
    const { activeCell } = state;
    if (activeCell &&
        !state.cellToFocusNext &&
        datatableHasFocus(state, template)) {
        const { length: rowCount } = state.rows;
        const lastIndex = rowCount - 1;
        const { length: colCount } = state.columns;
        let { rowIndex, colIndex } = activeCell;
        colIndex = 0;
        if (rowIndex === lastIndex) {
            rowIndex = lastIndex;
            colIndex = colCount ? colCount - 1 : 0;
        }
        state.cellToFocusNext = {
            rowIndex,
            colIndex,
        };
    }
}
export function updateCellToFocusFromPrev(state) {
    if (state.activeCell &&
        state.cellToFocusNext &&
        stillValidActiveCell(state)) {
        state.cellToFocusNext = null;
    }
}
export function addFocusStylesToActiveCell(state, template) {
    state.activeCell.hasFocus = true;
    const cellElement = getActiveCellElement(template, state);
    if (cellElement) {
        cellElement.classList.add(FOCUS_CLASS);
    }
}
function setFocusActiveRow(state, template, refs) {
    const { activeCell } = state;
    if (!activeCell) {
        return;
    }
    const { rowIndex, rowKeyValue } = activeCell;
    const rowElement = getActiveCellRowElement(template, state);
    updateRowTabIndex(state, rowIndex);
    setTimeout(() => {
        rowElement.focus({ preventScroll: true });
        updateScrollTop(state, template, refs, rowElement);
        resetHighlightedRowClasses(state, rowKeyValue);
    }, 0);
}
function setBlurActiveRow(state, template) {
    const { activeCell } = state;
    if (activeCell) {
        setTimeout(() => {
            const rowElement = getActiveCellRowElement(template, state);
            if (document.activeElement === rowElement) {
                rowElement.blur();
            }
        }, 0);
        updateRowTabIndex(state, activeCell.rowIndex, -1);
    }
}
export function refocusCellElement(state, template, needsRefocusOnCellElement) {
    if (needsRefocusOnCellElement) {
        const cellElement = getActiveCellElement(template, state);
        if (cellElement) {
            cellElement.focus();
        }
        setTimeout(() => {
            setCellClickedForFocus(state);
        }, 0);
    }
    else if (!datatableHasFocus(state, template)) {
        setCellClickedForFocus(state);
    }
}
export function datatableHasFocus(state, template) {
    return state.cellClicked || isFocusInside(template);
}
function isFocusInside(currentTarget) {
    const activeElements = getShadowActiveElements();
    for (let i = 0, { length } = activeElements; i < length; i += 1) {
        if (currentTarget.contains(activeElements[i])) {
            return true;
        }
    }
    return false;
}
export function handleDatatableFocusIn(event) {
    const { state } = this;
    state.hasReceivedFocus = true;
    state.isExitingActionMode = false;
    if (isFocusInside(event.currentTarget) && isCellElement(event.target)) {
        const { activeCell } = state;
        if (activeCell) {
            const { rowKeyValue } = activeCell;
            resetHighlightedRowClasses(state, rowKeyValue);
        }
        if (activeCell && !state.rowMode) {
            activeCell.hasFocus = true;
            const cellElement = getActiveCellElement(this.template, state);
            if (cellElement) {
                const cellElementChild = getCellElementChild(cellElement);
                if (cellElementChild) {
                    if (typeof cellElementChild
                        .addFocusStyles === 'function') {
                        cellElementChild.addFocusStyles();
                    }
                    cellElement.tabIndex = 0;
                }
                cellElement.classList.add(FOCUS_CLASS);
                if (!this._suppressScroll) {
                    scrollToCell(state, this.template, this.refs, cellElement);
                }
                else {
                    this._suppressScroll = false;
                }
            }
        }
        resetCellClickedForFocus(state);
    }
}
export function scrollToCell(state, template, refs, cellElement) {
    const isLastCol = state.activeCell?.colIndex === state.columns?.length - 1;
    if (isLastCol) {
        const scrollerX = getScrollerX(template, refs);
        scrollerX.scrollLeft = features.enablePositionLibFlipCalcWithViewportRight && isRTL() ? -scrollerX.scrollWidth : scrollerX.scrollWidth;
    }
    else {
        const outerContainer = refs?.outerContainer ||
            template.querySelector('div.dt-outer-container');
        const cellElementPosition = cellElement.getBoundingClientRect();
        const outerContainerPosition = outerContainer.getBoundingClientRect();
        if (cellElementPosition.left < outerContainerPosition.left ||
            cellElementPosition.right > outerContainerPosition.right) {
            cellElement.scrollIntoView({
                block: 'nearest',
                inline: 'nearest',
            });
        }
    }
}
export function handleDatatableFocusOut(event) {
    const { state } = this;
    const { currentTarget, relatedTarget } = event;
    const containsRelatedTarget = relatedTarget && currentTarget.contains(relatedTarget);
    if (!relatedTarget ||
        (relatedTarget && !containsRelatedTarget) ||
        (relatedTarget && containsRelatedTarget && state.isExitingActionMode)) {
        unsetHighlightedRowClasses(state, state.focusedRowKeyValue);
        state.focusedRowKeyValue = undefined;
        if (state.activeCell && !state.rowMode) {
            const cellElement = getActiveCellElement(this.template, state);
            if (cellElement) {
                const cellElementChild = getCellElementChild(cellElement);
                if (cellElementChild &&
                    typeof cellElementChild
                        .removeFocusStyles === 'function') {
                    cellElementChild.removeFocusStyles();
                }
                cellElement.classList.remove(FOCUS_CLASS);
            }
        }
    }
}
export function setCellClickedForFocus(state) {
    state.cellClicked = true;
}
function resetCellClickedForFocus(state) {
    state.cellClicked = false;
}
export function updateCellTabIndex(state, rowIndex, colIndex, tabIndex = 0) {
    if (isHeaderRow(rowIndex)) {
        state.columns[colIndex].tabIndex = tabIndex;
    }
    else {
        state.rows[rowIndex].cells[colIndex].tabIndex = tabIndex;
    }
}
export function updateRowTabIndex(state, rowIndex, tabIndex = 0) {
    if (!isHeaderRow(rowIndex)) {
        state.rows[rowIndex].tabIndex = tabIndex;
    }
}
export function updateTabIndexActiveCell(state, index = 0) {
    if (state.activeCell && !stillValidActiveCell(state)) {
        syncActiveCell(state);
    }
    updateActiveCellTabIndexAfterSync(state, index);
}
export function updateTabIndexActiveRow(state, index = 0) {
    if (state.activeCell && !stillValidActiveCell(state)) {
        syncActiveCell(state);
    }
    updateActiveRowTabIndexAfterSync(state, index);
}
function getNextIndexUp(state, rowIndex) {
    return rowIndex === -1 ? undefined : rowIndex - 1;
}
function getNextIndexDown(state, rowIndex) {
    const rowsCount = state.rows.length;
    return rowIndex + 1 < rowsCount ? rowIndex + 1 : undefined;
}
function getNextColumnIndex(columnsCount, colIndex) {
    return columnsCount > colIndex + 1 ? colIndex + 1 : undefined;
}
function getPrevColumnIndex(colIndex) {
    return colIndex > 0 ? colIndex - 1 : undefined;
}
function getNextIndexRight(state, colIndex) {
    return isRTL()
        ? getPrevColumnIndex(colIndex)
        : getNextColumnIndex(state.columns.length, colIndex);
}
function getNextIndexLeft(state, colIndex) {
    return isRTL()
        ? getNextColumnIndex(state.columns.length, colIndex)
        : getPrevColumnIndex(colIndex);
}
function getNextIndexUpWrapped(state, rowIndex) {
    const { length: rowCount } = state.rows;
    return rowIndex === -1 ? rowCount - 1 : rowIndex - 1;
}
function getNextIndexDownWrapped(state, rowIndex) {
    const { length: rowCount } = state.rows;
    return rowIndex + 1 < rowCount ? rowIndex + 1 : -1;
}
function canBeRowNavigationMode(state) {
    return (state.keyboardMode === KEYBOARD_NAVIGATION_MODE && state.treeColumn);
}
function isRowNavigationMode(state) {
    return (state.keyboardMode === KEYBOARD_NAVIGATION_MODE &&
        state.rowMode === true);
}
function setRowNavigationMode(state) {
    if (state.treeColumn && state.keyboardMode === KEYBOARD_NAVIGATION_MODE) {
        state.rowMode = true;
    }
}
export function unsetRowNavigationMode(state) {
    state.rowMode = false;
}
export function updateRowNavigationMode(state) {
    if (!state.treeColumn) {
        state.rowMode = false;
    }
    else if (state.rowMode === false && !state.hadTreeDataTypePreviously) {
        state.rowMode = true;
    }
}
export function isCellElement(target) {
    const role = target.getAttribute('role');
    return (role === 'gridcell' || role === 'columnheader' || role === 'rowheader');
}
function isHeaderRow(rowIndex) {
    return rowIndex === -1;
}
function getActiveCellRowElement(template, state) {
    let result;
    const { activeCell } = state;
    if (activeCell) {
        result = template.querySelector(getRowDataSelector(activeCell.rowKeyValue));
    }
    return result || undefined;
}
export function getRowParentIndex(state, rowLevel, rowIndex) {
    const parentIndex = rowIndex - 1;
    const { rows } = state;
    for (let i = parentIndex; i >= 0; i--) {
        if (rows[i].level === rowLevel - 1) {
            return i;
        }
    }
    return -1;
}
function updateScrollTop(state, template, refs, element) {
    const scrollableY = getScrollerY(template, refs);
    const scrollingParent = scrollableY.parentElement;
    const parentRect = scrollingParent.getBoundingClientRect();
    const findMeRect = element.getBoundingClientRect();
    if (findMeRect.top < parentRect.top + TOP_MARGIN) {
        scrollableY.scrollTop -= SCROLL_OFFSET;
    }
    else if (findMeRect.bottom > parentRect.bottom - BOTTOM_MARGIN) {
        scrollableY.scrollTop += SCROLL_OFFSET;
    }
    findFirstVisibleIndex(state, scrollableY.scrollTop);
}
function scrollToRow(state, template, refs, rowIndex) {
    const { firstVisibleIndex, bufferSize, renderedRowCount, rowHeight } = state;
    let scrollTop = rowIndex * rowHeight;
    if (firstVisibleIndex > rowIndex) {
        const rowsInViewport = renderedRowCount - (2 * bufferSize);
        scrollTop = Math.max(scrollTop - (rowsInViewport * rowHeight), 0);
    }
    const scrollableY = getScrollerY(template, refs);
    scrollableY.scrollTop = scrollTop;
    findFirstVisibleIndex(state, scrollTop);
}
export function isActiveCellEditable(state) {
    const { activeCell } = state;
    return activeCell ? activeCell.editable : false;
}
function isActiveCellValid(state) {
    const { activeCell } = state;
    if (activeCell) {
        return isValidCell(state, activeCell.rowKeyValue, activeCell.colKeyValue);
    }
    return false;
}
export function updateActiveCellTabIndexAfterSync(state, tabIndex = 0) {
    const { activeCell } = state;
    if (activeCell && !isRowNavigationMode(state)) {
        updateCellTabIndex(state, activeCell.rowIndex, activeCell.colIndex, tabIndex);
    }
}
export function updateActiveRowTabIndexAfterSync(state, tabIndex = 0) {
    const { activeCell } = state;
    if (activeCell && isRowNavigationMode(state)) {
        updateRowTabIndex(state, activeCell.rowIndex, tabIndex);
    }
}
