// lightning-base-components/src/lightning/datatable/renderManager.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { getScrollOffsetFromTableEnd, normalizeNumberAttribute } from './utils';
import { LightningResizeObserver } from 'lightning/resizeObserver';
import { normalizeBoolean, normalizeString } from 'lightning/utilsPrivate';
import { resetRowHeights } from './virtualization';
export const DEFAULT_ROW_HEIGHT = 30.5;
const DEFAULT_BUFFER_SIZE = 5;
const ROW_THRESHOLD = 10;
const DEFAULT_SCROLL_THRESHOLD = ROW_THRESHOLD * DEFAULT_ROW_HEIGHT;
const VERTICAL_VIRTUALIZATION = 'vertical';
export function setViewportRendering(state, value) {
    state.enableViewportRendering = normalizeBoolean(value);
}
export function setVirtualize(state, value) {
    if (state.renderModeRoleBased) {
        state.virtualize = normalizeString(value, {
            fallbackValue: '',
            validValues: [VERTICAL_VIRTUALIZATION],
        });
    }
    else {
        state.virtualize = '';
    }
}
function getDefaultPreviousInfo() {
    return {
        renderedRowCount: 0,
        totalRowCount: 0,
        firstRowKey: '',
    };
}
export class RenderManager {
    threshold;
    wrapperHeight;
    previousCache;
    _heightResizeObserver;
    _resizeObserverConnected;
    constructor() {
        this.threshold = DEFAULT_SCROLL_THRESHOLD;
        this.wrapperHeight = 0;
        this.previousCache = getDefaultPreviousInfo();
    }
    configure(state, getWrapperHeight, config) {
        const { rowHeight, bufferSize } = config;
        setVirtualize(state, config.virtualize);
        setViewportRendering(state, config.viewportRendering);
        if (state.virtualize || state.enableViewportRendering) {
            this.initializeResizeObserver(state, getWrapperHeight);
        }
        const computedBufferSize = typeof bufferSize != 'undefined' ? bufferSize : DEFAULT_BUFFER_SIZE;
        state.bufferSize = normalizeNumberAttribute('bufferSize', computedBufferSize, 'non-negative', DEFAULT_BUFFER_SIZE);
        if (typeof rowHeight === 'number') {
            state.rowHeight = rowHeight;
            this.threshold = ROW_THRESHOLD * rowHeight;
        }
    }
    updateViewportRendering(state, gridContainer, forceUpdate) {
        if (this.hasDataChanged(state.rows) || forceUpdate) {
            this.updateRenderedRows(state, this.getRowCountWithBuffer(state));
        }
        else if (this.previousCache.totalRowCount < state.rows.length &&
            this.isWithinThreshold(gridContainer)) {
            this.updateRenderedRows(state, this.previousCache.renderedRowCount +
                this.getRowCountWithBuffer(state));
        }
        else {
            this.updateRenderedRows(state, this.previousCache.renderedRowCount);
        }
    }
    handleScroll(state, event) {
        const { rows, renderedRowCount } = state;
        if (this.isWithinThreshold(event.target.firstChild) &&
            renderedRowCount < rows.length) {
            this.updateRenderedRows(state, renderedRowCount + this.getRowCountWithBuffer(state));
        }
    }
    getRenderedRange(state) {
        const { firstVisibleIndex, bufferSize, renderedRowCount, rows, rowHeight, heightCache, } = state;
        const firstIndex = Math.max(firstVisibleIndex - bufferSize, 0);
        let lastIndex = firstVisibleIndex - bufferSize + renderedRowCount;
        const firstVisibleKey = rows[firstVisibleIndex] && rows[firstVisibleIndex].key;
        if (heightCache[firstVisibleKey]) {
            let i = firstVisibleIndex;
            let currentHeight = 0;
            let knownRowHeight = rows[i] && heightCache[rows[i].key];
            while (knownRowHeight && currentHeight < this.wrapperHeight) {
                currentHeight += knownRowHeight;
                i += 1;
                knownRowHeight = rows[i] && heightCache[rows[i].key];
            }
            if (currentHeight >= this.wrapperHeight) {
                lastIndex = i + bufferSize;
            }
            else {
                const extraRowEstimate = Math.ceil((this.wrapperHeight - currentHeight) / rowHeight);
                lastIndex = i + extraRowEstimate + bufferSize;
            }
        }
        if (renderedRowCount === rows.length &&
            lastIndex <= rows.length - bufferSize) {
            lastIndex = renderedRowCount;
        }
        return { firstIndex, lastIndex };
    }
    updateRenderedRows(state, rowCount) {
        const rows = state.rows;
        const totalRows = rows.length;
        const normalizedRowCount = rowCount
            ? Math.min(rowCount, totalRows)
            : totalRows;
        if (state.enableViewportRendering) {
            if (state.renderedRowCount < normalizedRowCount) {
                state.renderedRowCount = normalizedRowCount;
                this.previousCache.renderedRowCount = normalizedRowCount;
            }
        }
        else {
            state.renderedRowCount = normalizedRowCount;
            this.previousCache.renderedRowCount = normalizedRowCount;
        }
        this.previousCache.totalRowCount = totalRows;
        if (rows.length > 0) {
            this.previousCache.firstRowKey = rows[0].key;
        }
    }
    updateWrapperHeight(getWrapperHeight) {
        this.wrapperHeight = getWrapperHeight();
    }
    initializeResizeObserver(state, getWrapperHeight) {
        if (!this._heightResizeObserver) {
            this._heightResizeObserver = new LightningResizeObserver(() => {
                if (this._resizeObserverConnected) {
                    this.updateWrapperHeight(getWrapperHeight);
                    const rowCountWithBuffer = this.getRowCountWithBuffer(state);
                    resetRowHeights(state);
                    if (rowCountWithBuffer > state.renderedRowCount ||
                        state.virtualize) {
                        this.updateRenderedRows(state, rowCountWithBuffer);
                    }
                }
            });
        }
    }
    connectResizeObserver(resizeTarget) {
        const { _heightResizeObserver } = this;
        if (_heightResizeObserver) {
            _heightResizeObserver.observe(resizeTarget);
            this._resizeObserverConnected = true;
        }
    }
    disconnectResizeObserver() {
        const { _heightResizeObserver } = this;
        if (_heightResizeObserver) {
            this._resizeObserverConnected = false;
            _heightResizeObserver.disconnect();
        }
    }
    getRowCountInViewport(state) {
        return Math.ceil(this.wrapperHeight / state.rowHeight);
    }
    getRowCountWithBuffer(state) {
        const multiplier = state.virtualize ? 2 : 1;
        return (this.getRowCountInViewport(state) + (state.bufferSize * multiplier));
    }
    isWithinThreshold(target) {
        const offset = getScrollOffsetFromTableEnd(target);
        return offset <= this.threshold;
    }
    hasDataChanged(rows) {
        return (rows.length > 0 && this.previousCache.firstRowKey !== rows[0].key);
    }
    hasWrapperHeight() {
        return !!this.wrapperHeight;
    }
}
