// lightning-base-components/src/lightning/datatable/resizeObserver.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { LightningResizeObserver } from 'lightning/resizeObserver';
import { isTableRenderedVisible } from './columnResizer';
import { handlePrefetch } from './infiniteLoading';
import { recomputeCellStyles } from './rows';
function widthObserverCallback() {
    const { _columnWidthManager, state, widthsData } = this;
    _columnWidthManager.adjustColumnsSizeAfterResize(this, state.columns, widthsData);
    this._recalculateHeaderHeight = true;
    if (state.renderModeRoleBased) {
        recomputeCellStyles(state);
    }
    state.shouldResetHeights = true;
    requestAnimationFrame(() => {
        this.updateHeaderHeight();
    });
}
const OBSERVER_MAP = {
    tableHeight: {
        observerProp: '_heightObserver',
        callback: handlePrefetch,
        targetRefs: ['scrollerY', 'tableElement'],
    },
    width: {
        observerProp: '_widthObserver',
        callback: widthObserverCallback,
        targetRefs: ['widthObserver'],
    },
};
export class LightningDatatableResizeObserver {
    _connected = false;
    _resizeObserver;
    _resizeSensor;
    constructor(dt, observerType) {
        const { callback, targetRefs } = OBSERVER_MAP[observerType];
        const resizeCallback = callback.bind(dt);
        this._resizeObserver = new LightningResizeObserver(() => {
            const { refs, template } = dt;
            if (this._connected && isTableRenderedVisible(template, refs)) {
                resizeCallback();
            }
        });
        if (dt.refs) {
            for (let i = 0; i < targetRefs.length; i += 1) {
                const resizeTarget = dt.refs[targetRefs[i]];
                this._resizeObserver.observe(resizeTarget);
            }
        }
        this._connected = true;
    }
    observe(targetElement) {
        if (this._resizeObserver) {
            this._resizeObserver.observe(targetElement);
        }
        else if (this._resizeSensor) {
            this._resizeSensor.reattach(targetElement);
        }
        this._connected = true;
    }
    disconnect() {
        if (this._resizeObserver) {
            this._resizeObserver.disconnect();
        }
        else if (this._resizeSensor) {
            this._resizeSensor.detach();
            this._resizeSensor = undefined;
        }
        this._connected = false;
    }
    isConnected() {
        return this._connected;
    }
}
function createOrConnectObserver(dt, observerType) {
    const { observerProp, targetRefs } = OBSERVER_MAP[observerType];
    const observer = dt[observerProp];
    if (!observer) {
        dt[observerProp] = new LightningDatatableResizeObserver(dt, observerType);
    }
    else if (!observer.isConnected() && dt.refs) {
        for (let i = 0; i < targetRefs.length; i += 1) {
            const resizeTarget = dt.refs[targetRefs[i]];
            observer.observe(resizeTarget);
        }
    }
}
export function updateObservers(dt) {
    const observerTypes = Object.keys(OBSERVER_MAP);
    const { length: observerCount } = observerTypes;
    for (let i = 0; i < observerCount; i += 1) {
        createOrConnectObserver(dt, observerTypes[i]);
    }
}
export function disconnectObservers(dt) {
    const observerTypes = Object.keys(OBSERVER_MAP);
    const { length: observerCount } = observerTypes;
    for (let i = 0; i < observerCount; i += 1) {
        const { observerProp } = OBSERVER_MAP[observerTypes[i]];
        if (dt[observerProp]) {
            dt[observerProp].disconnect();
        }
    }
}
