// lightning-base-components/src/lightning/datatable/rowNumber.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { normalizeNumberAttribute } from './utils';
import labelRowLevelErrorAssistiveText from '@salesforce/label/LightningDatatable.rowLevelErrorAssistiveText';
import { formatLabel } from 'lightning/utils';
const CHAR_WIDTH = 10;
const ROW_NUMBER_PADDING = 12;
const ROW_NUMBER_COLUMN_TYPE = 'rowNumber';
export const ROW_NUMBER_INITIAL_WIDTH = 52;
export const TOOLTIP_ALLOWANCE = 20;
const i18n = {
    rowLevelErrorAssistiveText: labelRowLevelErrorAssistiveText,
};
export function setRowNumberOffset(state, value) {
    state.rowNumberOffset = normalizeNumberAttribute('rowNumberOffset', value, 'non-negative', 0);
}
export function getAdjustedRowNumberColumnWidth(state) {
    const numberOfChars = String(state.rows.length + state.rowNumberOffset).length;
    return Math.max((CHAR_WIDTH * numberOfChars) +
        ROW_NUMBER_PADDING +
        TOOLTIP_ALLOWANCE, ROW_NUMBER_INITIAL_WIDTH);
}
export function getRowNumberColumnIndex(state) {
    if (state.showRowNumberColumn) {
        const { columns } = state;
        for (let i = 0, { length } = columns; i < length; i += 1) {
            if (columns[i].type === ROW_NUMBER_COLUMN_TYPE) {
                return i;
            }
        }
    }
    return -1;
}
export function getRowNumberError(rowErrors, rowTitle) {
    const { cells, fieldNames, messages, title } = rowErrors || {};
    let numberOfErrors = 0;
    if (cells) {
        numberOfErrors = Object.keys(cells).length;
    }
    else if (Array.isArray(fieldNames)) {
        numberOfErrors = fieldNames.length;
    }
    else if (typeof fieldNames === 'object') {
        numberOfErrors = Object.keys(fieldNames).length;
    }
    const alternativeText = formatLabel(i18n.rowLevelErrorAssistiveText, rowTitle || '', numberOfErrors);
    return { title, messages, alternativeText };
}
export function getRowNumberErrorColumnDef(rowErrors, rowTitle) {
    return {
        type: ROW_NUMBER_COLUMN_TYPE,
        typeAttributes: {
            error: getRowNumberError(rowErrors, rowTitle),
        },
    };
}
