// lightning-base-components/src/lightning/datatable/rows.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { features } from 'lightning/features';
import { assert, isRTL, isUndefinedOrNull } from 'lightning/utilsPrivate';
import { getTypeAttributesValues, getSubTypeAttributesValues, getCellAttributesValues, } from './columns';
import { HEADER_ROW_INDEX, HEADER_ROW_KEY } from './indexes';
import { setInliningProperties, updateInlineCell, updateInlineClassAndStyle, updateInlineCellValue, } from './rowsInlining';
import { getRowNumberError } from './rowNumber';
import { getCurrentSelectionLength, isInputTypeCheckbox, isRowDisabledForEditLimit, SELECTABLE_HEADER_TYPE, SORTING_MENU_HEADER_TYPE, SINGLE_ROW_SELECTION_MODES, setDisabledRows, } from './rowSelectionShared';
import { classSet, isObjectLike } from './utils';
const CELL_EDIT_CLASS = 'slds-cell-edit';
const HAS_ERROR_CLASS = 'slds-has-error';
const IS_EDITED_CLASS = 'slds-is-edited';
const ROLE_BASED_CELL_CLASS = 'cell';
const TREE__ITEM_CLASS = 'slds-tree__item';
function resolveAttributeValue(attrValue, rowData) {
    if (isObjectLike(attrValue)) {
        const { fieldName } = attrValue;
        if (fieldName) {
            return rowData[fieldName];
        }
    }
    return attrValue;
}
export function setKeyField(state, value) {
    if (typeof value === 'string') {
        state.keyField = value;
    }
    else {
        assert(false, `The "keyField" value expected in lightning:datatable must be type String.`);
        state.keyField = undefined;
    }
}
export function updateRowsAndCells(state, types, datatableId) {
    const { columns, data, keyField, maxRowSelection, renderModeInline, rowHeight, virtualize, } = state;
    const { privateCustomTypes } = types;
    const rtl = isRTL();
    const { length: colCount } = columns;
    const isInputCheckbox = isInputTypeCheckbox(state);
    const currentSelectionLength = isInputCheckbox
        ? getCurrentSelectionLength(state)
        : 1;
    const inputType = isInputCheckbox
        ? SINGLE_ROW_SELECTION_MODES.CHECKBOX
        : SINGLE_ROW_SELECTION_MODES.RADIO;
    let scopeCol;
    let treeColTypeAttrs;
    const indexes = {};
    for (let colIndex = 0; colIndex < colCount; colIndex += 1) {
        const col = columns[colIndex];
        const { colKeyValue, type: columnType } = col;
        const cellKeyValue = `${HEADER_ROW_KEY}-${colKeyValue}`;
        const headerCell = {
            cellKeyValue,
            colIndex,
            colKeyValue,
            focused: false,
            key: colKeyValue,
            rowIndex: HEADER_ROW_INDEX,
            rowKeyValue: HEADER_ROW_KEY,
        };
        indexes[cellKeyValue] = headerCell;
        if (!scopeCol && col.isScopeCol && types.isValidType(columnType)) {
            scopeCol = col;
        }
        if (columnType === 'tree' && treeColTypeAttrs === undefined) {
            treeColTypeAttrs = getTypeAttributesValues(col);
        }
    }
    const checkboxCells = [];
    const lookupCells = [];
    const shownUrlCells = [];
    const rows = [];
    state.checkboxCells = checkboxCells;
    state.indexes = indexes;
    state.lookupCells = lookupCells;
    state.shownUrlCells = shownUrlCells;
    state.rows = rows;
    data.forEach((rowData, rowIndex) => {
        const rowDataKeyField = rowData[keyField];
        const rowKeyValue = isUndefinedOrNull(rowDataKeyField)
            ? `row-${rowIndex}`
            : rowDataKeyField;
        state.indexes[rowKeyValue] = { rowIndex };
        const isRowDisabled = features.enableDisabledRowsAttribute &&
            !!state.disabledRowsKeys[rowKeyValue];
        const isRowSelected = !!state.selectedRowsKeys[rowKeyValue];
        const ariaSelected = isRowSelected ? 'true' : false;
        const rowNumber = rowIndex + 1;
        const scopeColValue = scopeCol
            ? rowData[scopeCol.fieldName]
            : undefined;
        const cells = Array(colCount);
        const checkboxHidden = state.hideCheckboxColumn;
        const row = {
            ariaRowIndex: rowIndex + 2,
            ariaSelected: checkboxHidden ? null : ariaSelected,
            cells,
            classnames: `slds-hint-parent${isRowSelected ? ' slds-is-selected' : ''}`,
            inputType,
            isSelected: isRowSelected,
            isDisabled: features.enableMaxEditLimit
                ? isRowDisabled
                : isRowDisabled ||
                    (!isRowSelected &&
                        isInputCheckbox &&
                        currentSelectionLength === maxRowSelection),
            key: rowKeyValue,
            rowIndex,
            rowNumber,
            style: virtualize
                ? `position:absolute;top:${rowIndex * rowHeight}px;`
                : '',
            tabIndex: -1,
            level: undefined,
            posInSet: undefined,
            setSize: undefined,
            isExpanded: undefined,
            hasChildren: undefined,
        };
        if (treeColTypeAttrs) {
            const hasChildren = !!resolveAttributeValue(treeColTypeAttrs.hasChildren, rowData);
            row.hasChildren = hasChildren;
            row.isExpanded = hasChildren
                ? `${!!resolveAttributeValue(treeColTypeAttrs.isExpanded, rowData)}`
                : undefined;
            row.level =
                resolveAttributeValue(treeColTypeAttrs.level, rowData) || 1;
            row.posInSet =
                resolveAttributeValue(treeColTypeAttrs.posInSet, rowData) || 1;
            row.setSize =
                resolveAttributeValue(treeColTypeAttrs.setSize, rowData) || 1;
        }
        indexes[rowKeyValue] = row;
        rows.push(row);
        for (let colIndex = 0; colIndex < colCount; colIndex += 1) {
            const col = columns[colIndex];
            const { colKeyValue, isScopeCol, label: dataLabel, type: columnType, typeAttributes, } = col;
            const cellKeyValue = `${rowKeyValue}-${colKeyValue}`;
            const columnSubType = typeAttributes
                ? typeAttributes.subType
                : undefined;
            const hasTreeData = columnType === 'tree';
            const privateType = privateCustomTypes &&
                privateCustomTypes.get(hasTreeData ? columnSubType : columnType);
            const standardCellLayout = privateType
                ? privateType.standardCellLayout
                : false;
            const displayReadOnlyIcon = !!col.displayReadOnlyIcon;
            const editable = !!resolveAttributeValue(col.editable, rowData);
            const isEditable = editable && types.isEditableType(columnType);
            const isValidType = types.isValidType(columnType);
            const isCheckbox = columnType === SELECTABLE_HEADER_TYPE;
            const isSortingMenu = columnType === SORTING_MENU_HEADER_TYPE;
            const isCustom = privateType !== undefined;
            const isCustomBareLayout = isCustom && !standardCellLayout;
            const isCustomStandardLayout = isCustom && standardCellLayout;
            const isDataType = isValidType && !isScopeCol;
            const isDataTypeScope = isValidType && isScopeCol;
            const cell = {
                ariaReadOnly: !editable,
                cellKeyValue,
                class: '',
                colIndex,
                colKeyValue,
                columnSubType,
                columnType,
                computedClass: '',
                computedStyle: '',
                dataLabel,
                displayReadOnlyIcon,
                displayValue: rowData.displayValue || '',
                describedBy: null,
                editable,
                hasError: undefined,
                hasTreeData,
                isCheckbox,
                isCustom,
                isCustomBareLayout,
                isCustomStandardLayout,
                isDataType,
                isDataTypeScope,
                isEditable,
                isRTL: rtl,
                isSortingMenu,
                rowIndex,
                rowKeyValue,
                rowNumber,
                scopeColValue,
                style: '',
                tabIndex: -1,
                value: undefined,
                wrapText: undefined,
                wrapTextMaxLines: 0,
            };
            if (features.enableConsolidateTypeAttributes) {
                cell.typeAttributes = {};
                cell.treeAttributes = {};
            }
            indexes[cellKeyValue] = cell;
            cells[colIndex] = cell;
            if (!col.internal) {
                let attributeNames;
                let attributeValues;
                if (columnSubType) {
                    attributeNames = types.getAttributesNames(columnSubType);
                    attributeValues = getSubTypeAttributesValues(col);
                }
                else {
                    const typeDesc = types.getType(columnType);
                    attributeNames = typeDesc && typeDesc.typeAttributes;
                    attributeValues = getTypeAttributesValues(col);
                }
                const attributeNamesLength = attributeNames
                    ? attributeNames.length
                    : 0;
                for (let i = 0; i < attributeNamesLength; i += 1) {
                    const attrName = attributeNames[i];
                    const attrValue = attributeValues[attrName];
                    const resolvedValue = resolveAttributeValue(attrValue, rowData);
                    if (features.enableConsolidateTypeAttributes) {
                        cell.typeAttributes[attrName] = resolvedValue;
                    }
                    else {
                        cell[`typeAttribute${i}`] = resolvedValue;
                    }
                }
                const cellAttributes = getCellAttributesValues(col);
                const keys = Object.keys(cellAttributes);
                for (let i = 0; i < keys.length; i += 1) {
                    const attrName = keys[i];
                    const attrValue = cellAttributes[attrName];
                    cell[attrName] = resolveAttributeValue(attrValue, rowData);
                }
                if (hasTreeData) {
                    if (features.enableConsolidateTypeAttributes) {
                        cell.treeAttributes.hasChildren = row.hasChildren;
                        cell.treeAttributes.isExpanded = row.isExpanded === 'true';
                    }
                    else {
                        cell.typeAttribute21 = row.hasChildren;
                        cell.typeAttribute22 = row.isExpanded === 'true';
                    }
                }
            }
            state.indexes[row.key][colKeyValue] = [rowIndex, colIndex];
            cells[colIndex] = cell;
            if (renderModeInline) {
                setInliningProperties(state, cell, isInputCheckbox, datatableId);
            }
            updateCell(state, rowIndex, colIndex, rowData);
            if (cell.showUrlLink) {
                shownUrlCells.push(cell);
            }
        }
    });
    if (features.enableMaxEditLimit) {
        setDisabledRows(state);
    }
    state.hasCalledUpdateRowsAndCells = true;
}
export function updateCells(state, types) {
    if (!state.hasCalledUpdateRowsAndCells) {
        updateRowsAndCells(state, types);
        return;
    }
    const { data, rows } = state;
    for (let rowIndex = 0; rowIndex < rows.length; rowIndex += 1) {
        const rowData = data[rowIndex];
        for (let colIndex = 0; colIndex < rows[rowIndex].cells.length; colIndex += 1) {
            updateCell(state, rowIndex, colIndex, rowData);
        }
    }
}
export function updateCell(state, rowIndex, colIndex, rowData) {
    const { renderModeRoleBased } = state;
    const { rows: errorsRows } = state.errors;
    const row = state.rows[rowIndex];
    const { key: rowKeyValue } = row;
    const dirtyRowData = state.inlineEdit.dirtyValues[rowKeyValue];
    const rowErrors = errorsRows && errorsRows[rowKeyValue];
    const cellErrors = rowErrors && rowErrors.cells;
    const col = state.columns[colIndex];
    const cell = row.cells[colIndex];
    const { colKeyValue, columnKey, fieldName, wrapText } = col;
    let errorFieldNames = (rowErrors && rowErrors.fieldNames) || [];
    let customErrorValidationMessage = null;
    if (!Array.isArray(errorFieldNames)) {
        errorFieldNames = Object.keys(errorFieldNames);
        customErrorValidationMessage = rowErrors.fieldNames[fieldName];
    }
    const dirtyValue = dirtyRowData && dirtyRowData[colKeyValue];
    const hasError = columnKey
        ? !!(cellErrors && cellErrors[columnKey])
        : !!(errorFieldNames && errorFieldNames.includes(fieldName));
    const value = dirtyValue === undefined ? rowData[fieldName] : dirtyValue;
    const wrapTextMaxLines = wrapText ? state.wrapTextMaxLines : undefined;
    cell.hasError = hasError;
    cell.wrapText = wrapText;
    cell.wrapTextMaxLines = wrapTextMaxLines;
    cell.value = value;
    cell.customErrorValidationMessage = customErrorValidationMessage;
    if (features.enableMaxEditLimit) {
        cell.isEditHidden = false;
        if (cell.isEditable && isRowDisabledForEditLimit(state, rowKeyValue)) {
            cell.isEditHidden = true;
        }
    }
    if (col.internal && col.type === 'rowNumber') {
        if (features.enableConsolidateTypeAttributes) {
            cell.typeAttributes.error =
                getRowNumberError(rowErrors, cell.scopeColValue);
        }
        else {
            cell.typeAttribute0 = getRowNumberError(rowErrors, cell.scopeColValue);
        }
    }
    const cellAttributes = getCellAttributesValues(col);
    const cellClassAttr = resolveAttributeValue(cellAttributes.class, rowData) || '';
    let cellClass = '';
    if (!features.enableDatatableHoverStyling) {
        cellClass = cellClassAttr;
    }
    if (cell.hasTreeData) {
        cellClass += (cellClass.length ? ' ' : '') + TREE__ITEM_CLASS;
    }
    if (cell.displayReadOnlyIcon || cell.editable) {
        cellClass += (cellClass.length ? ' ' : '') + CELL_EDIT_CLASS;
    }
    if (hasError) {
        cellClass += (cellClass.length ? ' ' : '') + HAS_ERROR_CLASS;
    }
    if (dirtyValue !== undefined) {
        cellClass += (cellClass.length ? ' ' : '') + IS_EDITED_CLASS;
        cell.describedBy = 'unsaved-cell-notification';
    }
    if (renderModeRoleBased) {
        cellClass += (cellClass.length ? ' ' : '') + ROLE_BASED_CELL_CLASS;
    }
    if (!cell.isCustom && features.enableDatatablePositionOverride) {
        cellClass += (cellClass.length ? ' ' : '') + 'fix-safari_perf';
    }
    if (features.enableDatatableHoverStyling) {
        cell.defaultClass = cellClass;
    }
    else {
        cell.class = cellClass;
    }
    cell.style = computeCellStyles(cell, col, renderModeRoleBased, cell.style);
    if (cell.isInlined) {
        updateInlineCell(cell);
        updateInlineClassAndStyle(cell, cellClass);
        updateInlineCellValue(cell, value);
    }
    if (features.enableDatatableHoverStyling &&
        state.hasHighlightStyling &&
        (state.hoveredRowKeyValue === rowKeyValue ||
            state.focusedRowKeyValue === rowKeyValue)) {
        cell.computedClass = `${cell.highlightClass} ${cell.defaultClass}`;
    }
    else if (features.enableDatatableHoverStyling) {
        cell.computedClass = `${cell.class} ${cell.defaultClass}`;
    }
    else {
        cell.computedClass = cell.class;
    }
}
function computeCellStyles(cell, col, renderModeRoleBased, initValue = '') {
    let cellStyle = initValue;
    if (cellStyle && !cellStyle.endsWith(';')) {
        cellStyle += ';';
    }
    if (!cellStyle && cell.isCustomBareLayout) {
        cellStyle = 'padding: 0;';
    }
    if (renderModeRoleBased) {
        const { columnWidth } = col;
        if (columnWidth > 0) {
            cellStyle += `width:${columnWidth}px;`;
        }
    }
    return cellStyle;
}
export function recomputeCellStyles(state) {
    const { columns } = state;
    state.rows.forEach((row) => {
        row.cells.forEach((cell, colIndex) => {
            const colData = columns[colIndex];
            const cellAttrs = getCellAttributesValues(colData);
            const style = resolveAttributeValue(cellAttrs.style, row) || '';
            cell.style = computeCellStyles(cell, colData, true, style);
        });
    });
}
export function updateCellClassForRoleBasedMode(state) {
    if (state.renderModeRoleBased && state.rows) {
        state.rows.forEach((row) => {
            row.cells.forEach((cell) => {
                const classProp = features.enableDatatableHoverStyling ? 'defaultClass' : 'class';
                const classes = classSet(cell[classProp]);
                classes.add('cell');
                cell[classProp] = classes.toString();
            });
        });
    }
}
