// lightning-base-components/src/lightning/datatable/rowsInlining.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { features } from 'lightning/features';
import labelEdit from '@salesforce/label/LightningDatatable.edit';
import labelEditHasError from '@salesforce/label/LightningDatatable.editHasError';
import labelEmailLabel from '@salesforce/label/LightningFormattedEmail.emailIconLabel';
import labelFalse from '@salesforce/label/LightningDatatable.false';
import labelReadOnlyIconAlternativeText from '@salesforce/label/LightningDatatable.readOnlyIconAlternativeText';
import labelSelectItem from '@salesforce/label/LightningDatatable.selectItem';
import labelTrue from '@salesforce/label/LightningDatatable.true';
import { dateTimeFormat, numberFormat, toDate, } from 'lightning/internationalizationLibrary';
import { isTextIgnoreRTL, isValidPhone, parseToFormattedLinkifiedParts, toFormattedLocation, toFormattedNumber, toFormattedDate, toNorthAmericanPhoneNumber, } from 'lightning/utilsPrivate';
import { formatLabel, formatUrl } from 'lightning/utils';
import { TOOLTIP_ALLOWANCE } from './rowNumber';
import { getDefaultState } from './state';
const CELL_WRAP = 'slds-cell-wrap';
const GRID_CLASS = 'slds-grid';
const GRID_ALIGN_END_CLASS = 'slds-grid_align-end';
const GRID_ALIGN_SPREAD_CLASS = 'slds-grid_align-spread';
const HYPHENATE_CLASS = 'slds-hyphenate';
const LINE_CLAMP_CLASS = 'slds-line-clamp';
const LTR_CONTENT_IN_RTL_CLASS = 'ltr-content-in-rtl';
const NO_SPACE_CLASS = 'slds-no-space';
const TRUNCATE_CLASS = 'slds-truncate';
const { keyboardMode: DEFAULT_KEYBOARD_MODE } = getDefaultState();
const i18n = {
    edit: labelEdit,
    editHasError: labelEditHasError,
    emailLabel: labelEmailLabel,
    false: labelFalse,
    readOnlyIconAlternativeText: labelReadOnlyIconAlternativeText,
    selectItem: labelSelectItem,
    true: labelTrue,
};
const noopHandler = () => { };
const BASE_INLINING_PROPS = {
    isInlined: true,
    actionableElementsCount: 0,
    computedCellDivClass: '',
    computedCellDivStyle: '',
    computedWrapperClass: '',
    currentInputIndex: 0,
    editIconAssistiveText: '',
    hasFocus: false,
    hasLeftIcon: false,
    hasRightIcon: false,
    internalTabIndex: -1,
    isRTL: false,
    keyboardMode: DEFAULT_KEYBOARD_MODE,
    readOnlyIconAlternativeText: '',
    shouldDisplayReadOnlyIcon: false,
};
export function setInliningProperties(state, cell, isInputTypeCheckbox, datatableId) {
    const { columnSubType, columnType, isCheckbox, isCustom } = cell;
    if (columnType === 'boolean' || columnSubType === 'boolean') {
        Object.assign(cell, BASE_INLINING_PROPS, {
            isBoolean: true,
            isChecked: false,
            booleanCellAssistiveText: '',
        });
    }
    else if (isCheckbox && isInputTypeCheckbox) {
        const { rowNumber } = cell;
        Object.assign(cell, BASE_INLINING_PROPS, {
            checkboxId: `${datatableId}-check-id-${rowNumber}`,
            checkboxLabelId: `check-button-label-${rowNumber}`,
            checkboxName: `${datatableId}-options-${rowNumber}`,
            checkboxSelectItemAssistiveText: `${i18n.selectItem} ${rowNumber}`,
        });
        state.checkboxCells.push(cell);
    }
    else if (isCheckbox) {
        Object.assign(cell, { isInlined: false });
    }
    else if (isCustom) {
        Object.assign(cell, BASE_INLINING_PROPS);
    }
    else if (columnType === 'currency' || columnSubType === 'currency') {
        Object.assign(cell, BASE_INLINING_PROPS, {
            isCurrency: true,
            currencyValue: '',
        });
    }
    else if (columnType === 'date-local' || columnSubType === 'date-local') {
        Object.assign(cell, BASE_INLINING_PROPS, {
            isDateLocal: true,
            dateLocalValue: '',
        });
    }
    else if (columnType === 'date' || columnSubType === 'date') {
        Object.assign(cell, BASE_INLINING_PROPS, {
            isDateTime: true,
            dateTimeValue: '',
        });
    }
    else if (columnType === 'email' || columnSubType === 'email') {
        Object.assign(cell, BASE_INLINING_PROPS, {
            isEmail: true,
            emailIconAssistiveText: '',
            emailIconHidden: false,
            emailLabel: '',
            showEmailLink: false,
        });
    }
    else if (columnType === 'location' || columnSubType === 'location') {
        Object.assign(cell, BASE_INLINING_PROPS, {
            isLocation: true,
            locationValue: '',
        });
    }
    else if (columnType === 'reference' || columnSubType === 'reference') {
        Object.assign(cell, BASE_INLINING_PROPS, {
            isLookup: true,
            _handleLookupAnchorClick: noopHandler,
            handleLookupAnchorClick: (event) => {
                cell._handleLookupAnchorClick(event);
            },
            lookupIsNavigable: false,
            lookupLink: '',
            lookupValue: features.enableConsolidateTypeAttributes
                ? cell.typeAttributes?.displayValue
                : cell.typeAttribute0,
        });
        state.lookupCells.push(cell);
    }
    else if (columnType === 'number' || columnSubType === 'number') {
        Object.assign(cell, BASE_INLINING_PROPS, {
            isNumber: true,
            numberValue: '',
        });
    }
    else if (columnType === 'percent' || columnSubType === 'percent') {
        Object.assign(cell, BASE_INLINING_PROPS, {
            isPercent: true,
            percentValue: '',
        });
    }
    else if (columnType === 'phone' || columnSubType === 'phone') {
        Object.assign(cell, BASE_INLINING_PROPS, {
            isPhone: true,
            phoneLabel: '',
            phoneLink: '',
            showPhoneLink: false,
        });
    }
    else if (columnType === 'rowNumber' || columnSubType === 'rowNumber') {
        Object.assign(cell, BASE_INLINING_PROPS, {
            isRowNumber: true,
            computedRowNumberStyle: '',
            rowHasError: false,
        });
    }
    else if (columnType === 'text' || columnSubType === 'text') {
        Object.assign(cell, BASE_INLINING_PROPS, {
            isText: true,
            ignoreRTL: false,
            shouldFormatText: false,
            textParts: undefined,
            textValue: undefined,
        });
    }
    else if (columnType === 'url' || columnSubType === 'url') {
        Object.assign(cell, BASE_INLINING_PROPS, {
            isURL: true,
            _handleUrlAnchorClick: undefined,
            handlUrlAnchorClick: undefined,
            showUrlLink: false,
            urlLabel: '',
            urlLink: '',
            urlTarget: '',
            urlTooltip: '',
        });
    }
}
export function updateInlineCell(cell) {
    if (cell.isEditable) {
        cell.editIconAssistiveText = `${i18n.edit} ${cell.dataLabel}${cell.hasError ? ` ${i18n.editHasError}` : ''}`;
        cell.editLabelId = `edit-${cell.cellKeyValue}`;
        if (features.enableMaxEditLimit) {
            cell.computedEditButtonStyle = cell.isEditHidden
                ? 'visibility: hidden;'
                : '';
        }
    }
    else {
        cell.shouldDisplayReadOnlyIcon = cell.displayReadOnlyIcon;
        if (cell.displayReadOnlyIcon) {
            cell.readOnlyIconAlternativeText = formatLabel(i18n.readOnlyIconAlternativeText, cell.dataLabel);
        }
    }
    if (cell.iconName) {
        const { iconPosition } = cell;
        if (iconPosition === 'right') {
            cell.hasRightIcon = true;
        }
        else if (!iconPosition || iconPosition === 'left') {
            cell.hasLeftIcon = true;
        }
    }
    if (cell.isRowNumber) {
        const error = features.enableConsolidateTypeAttributes
            ? cell.typeAttributes?.error
            : cell.typeAttribute0;
        const rowHasError = !!(error && error.title && error.messages);
        cell.computedRowNumberStyle = rowHasError
            ? ''
            : `padding-left: ${TOOLTIP_ALLOWANCE}px;`;
        cell.rowHasError = rowHasError;
    }
}
export function updateInlineClassAndStyle(cell, cellClass) {
    let computedCellDivClass = '';
    let computedCellDivStyle = '';
    const { alignment, columnType, hasTreeData, wrapText } = cell;
    if (wrapText) {
        cellClass += (cellClass.length ? ' ' : '') + CELL_WRAP;
        computedCellDivClass += HYPHENATE_CLASS;
        const { wrapTextMaxLines } = cell;
        if (wrapTextMaxLines) {
            computedCellDivClass += ` ${LINE_CLAMP_CLASS}`;
            computedCellDivStyle = `--lwc-lineClamp: ${wrapTextMaxLines}`;
        }
        else {
            computedCellDivStyle = 'overflow-wrap: anywhere;';
        }
    }
    else {
        computedCellDivClass += TRUNCATE_CLASS;
    }
    if (cell.isRTL &&
        (columnType === 'url' ||
            columnType === 'email' ||
            columnType === 'phone')) {
        computedCellDivClass += ` ${LTR_CONTENT_IN_RTL_CLASS}`;
    }
    let computedWrapperClass = GRID_CLASS;
    if (alignment === 'right' ||
        columnType === 'currency' ||
        columnType === 'number' ||
        columnType === 'percent') {
        computedWrapperClass += ` ${GRID_ALIGN_END_CLASS}`;
    }
    else if (!hasTreeData &&
        (!alignment ||
            alignment === 'left' ||
            (alignment !== 'center' && alignment !== 'right'))) {
        computedWrapperClass += ` ${GRID_ALIGN_SPREAD_CLASS}`;
    }
    if (hasTreeData) {
        computedWrapperClass += ` ${NO_SPACE_CLASS}`;
    }
    if (features.enableDatatableHoverStyling) {
        cell.defaultClass = cellClass;
    }
    else {
        cell.class = cellClass;
    }
    cell.computedCellDivClass = computedCellDivClass;
    cell.computedCellDivStyle = computedCellDivStyle;
    cell.computedWrapperClass = computedWrapperClass;
}
export function updateInlineCellValue(cell, value) {
    const typeAttributes = features.enableConsolidateTypeAttributes
        ? cell.typeAttributes || {}
        : cell;
    const getTypeAttribute = (attrName, index) => {
        if (features.enableConsolidateTypeAttributes) {
            return typeAttributes[attrName];
        }
        return typeAttributes[`typeAttribute${index}`];
    };
    if (cell.isBoolean) {
        cell.isChecked = !!value;
        cell.booleanCellAssistiveText = value ? i18n.true : i18n.false;
    }
    else if (cell.isCurrency) {
        cell.currencyValue = toFormattedNumber({
            value,
            formatStyle: 'currency',
            currencyCode: getTypeAttribute('currencyCode', 0),
            currencyDisplayAs: getTypeAttribute('currencyDisplayAs', 1),
            minimumIntegerDigits: getTypeAttribute('minimumIntegerDigits', 2),
            minimumFractionDigits: getTypeAttribute('minimumFractionDigits', 3),
            maximumFractionDigits: getTypeAttribute('maximumFractionDigits', 4),
            minimumSignificantDigits: getTypeAttribute('minimumSignificantDigits', 5),
            maximumSignificantDigits: getTypeAttribute('maximumSignificantDigits', 6),
        }, numberFormat);
    }
    else if (cell.isDateLocal) {
        cell.dateLocalValue = toFormattedDate({
            value,
            day: getTypeAttribute('day', 0) || 'numeric',
            month: getTypeAttribute('month', 1) || 'short',
            year: getTypeAttribute('year', 2) || 'numeric',
            timeZone: 'UTC',
        }, dateTimeFormat);
    }
    else if (cell.isDateTime) {
        cell.dateTimeValue = toFormattedDate({
            value: toDate(value),
            dateStyle: 'medium',
            day: getTypeAttribute('day', 0),
            era: getTypeAttribute('era', 1),
            hour: getTypeAttribute('hour', 2),
            hour12: getTypeAttribute('hour12', 3),
            minute: getTypeAttribute('minute', 4),
            month: getTypeAttribute('month', 5),
            second: getTypeAttribute('second', 6),
            timeZone: getTypeAttribute('timeZone', 7),
            timeZoneName: getTypeAttribute('timeZoneName', 8),
            weekday: getTypeAttribute('weekday', 9),
            year: getTypeAttribute('year', 10),
        }, dateTimeFormat);
    }
    else if (cell.isEmail) {
        let emailIconAssistiveText = '';
        const emailLabel = value ? value.trim() : '';
        const showEmailLink = !!emailLabel;
        if (showEmailLink) {
            emailIconAssistiveText = i18n.emailLabel;
        }
        cell.emailIconAssistiveText = emailIconAssistiveText;
        cell.emailLabel = emailLabel;
        cell.showEmailLink = showEmailLink;
    }
    else if (cell.isLocation) {
        cell.locationValue = toFormattedLocation(value);
    }
    else if (cell.isNumber) {
        cell.numberValue = toFormattedNumber({
            value,
            minimumIntegerDigits: getTypeAttribute('minimumIntegerDigits', 0),
            minimumFractionDigits: getTypeAttribute('minimumFractionDigits', 1),
            maximumFractionDigits: getTypeAttribute('maximumFractionDigits', 2),
            minimumSignificantDigits: getTypeAttribute('minimumSignificantDigits', 3),
            maximumSignificantDigits: getTypeAttribute('maximumSignificantDigits', 4),
        }, numberFormat);
    }
    else if (cell.isPercent) {
        cell.percentValue = toFormattedNumber({
            value,
            formatStyle: 'percent',
            minimumIntegerDigits: getTypeAttribute('minimumIntegerDigits', 0),
            minimumFractionDigits: getTypeAttribute('minimumFractionDigits', 1),
            maximumFractionDigits: getTypeAttribute('maximumFractionDigits', 2),
            minimumSignificantDigits: getTypeAttribute('minimumSignificantDigits', 3),
            maximumSignificantDigits: getTypeAttribute('maximumSignificantDigits', 4),
        }, numberFormat);
    }
    else if (cell.isPhone) {
        let phoneLabel = '';
        let phoneLink = '';
        const showPhoneLink = isValidPhone(value);
        if (showPhoneLink) {
            phoneLabel = toNorthAmericanPhoneNumber(value);
            phoneLink = `tel:${value}`;
        }
        cell.phoneLabel = phoneLabel;
        cell.phoneLink = phoneLink;
        cell.showPhoneLink = showPhoneLink;
    }
    else if (cell.isText) {
        const linkify = getTypeAttribute('linkify', 0);
        const shouldFormatText = linkify
            ? typeof value === 'string' && value !== ''
            : false;
        cell.ignoreRTL = cell.isRTL && isTextIgnoreRTL(value);
        cell.shouldFormatText = shouldFormatText;
        cell.textParts = shouldFormatText
            ? parseToFormattedLinkifiedParts(value, true)
            : [];
        cell.textValue = typeof value === 'object' ? '' : value;
    }
    else if (cell.isURL) {
        let _handleUrlAnchorClick;
        let handleUrlAnchorClick;
        let urlLabel = '';
        let urlLink = '';
        let urlTarget = '';
        let urlTooltip = '';
        const showUrlLink = value !== null && value !== undefined && value !== '';
        if (showUrlLink) {
            _handleUrlAnchorClick = noopHandler;
            handleUrlAnchorClick = (event) => {
                cell._handleUrlAnchorClick(event);
            };
            urlLink = formatUrl(value);
            const labelAttr = getTypeAttribute('label', 0);
            urlLabel =
                labelAttr !== null &&
                    labelAttr !== undefined &&
                    labelAttr !== ''
                    ? labelAttr
                    : urlLink;
            urlTarget = getTypeAttribute('target', 1) || '_self';
            const tooltipAttr = getTypeAttribute('tooltip', 2);
            urlTooltip = tooltipAttr === '' ? '' : tooltipAttr || value;
        }
        cell._handleUrlAnchorClick = _handleUrlAnchorClick;
        cell.handleUrlAnchorClick = handleUrlAnchorClick;
        cell.showUrlLink = showUrlLink;
        cell.urlLabel = urlLabel;
        cell.urlLink = urlLink;
        cell.urlTarget = urlTarget;
        cell.urlTooltip = urlTooltip;
    }
}
