// lightning-base-components/src/lightning/datatable/rowSelection.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { getCellByKeys, getRowByKey, getRowIndexByKey, HEADER_ROW_KEY, } from './indexes';
import { isNonNegativeInteger } from './utils';
import { getCurrentSelectionLength, getRemainingEditSlots, getRowsKeys, getSelectedRowsKeys, isInputTypeCheckbox, SELECTABLE_HEADER_TYPE, SINGLE_ROW_SELECTION_MODES, setDisabledRows, isDisabledRow, getCurrentEditedRowsCount, } from './rowSelectionShared';
import { normalizeString, isUndefinedOrNull } from 'lightning/utilsPrivate';
import { setInliningProperties } from './rowsInlining';
import { features } from 'lightning/features';
import { updateCells } from './rows';
const MAX_ROW_SELECTION_DEFAULT = undefined;
const ROWS_ACTION = {
    SELECT_ALL_ROWS: 'selectAllRows',
    DESELECT_ALL_ROWS: 'deselectAllRows',
    ROW_SELECT: 'rowSelect',
    ROW_DESELECT: 'rowDeselect',
};
function isNewEditSelection(state, rowKeyValue) {
    if (!features.enableMaxEditLimit) {
        return false;
    }
    const wasAlreadySelected = !!state.selectedRowsKeys[rowKeyValue];
    const isAlreadyEdited = !!state.inlineEdit?.dirtyValues?.[rowKeyValue];
    return !wasAlreadySelected && !isAlreadyEdited;
}
export function handleSelectAllRows(event) {
    event.stopPropagation();
    markAllRowsSelected(this.state);
    this.fireSelectedRowsChange(this.getSelectedRows(), {
        action: ROWS_ACTION.SELECT_ALL_ROWS,
    });
}
export function handleDeselectAllRows(event) {
    event.stopPropagation();
    markAllRowsDeselected(this.state, false);
    this.fireSelectedRowsChange(this.getSelectedRows(), {
        action: ROWS_ACTION.DESELECT_ALL_ROWS,
    });
}
export function handleSelectRow(event) {
    event.stopPropagation();
    const { state } = this;
    const { rowKeyValue, isMultiple } = event.detail;
    const fromRowKey = isMultiple
        ? getLastRowSelection(state) || rowKeyValue
        : rowKeyValue;
    markSelectedRowsInterval(state, fromRowKey, rowKeyValue);
    state.lastSelectedRowKey = rowKeyValue;
    this.fireSelectedRowsChange(this.getSelectedRows(), {
        action: ROWS_ACTION.ROW_SELECT,
        value: rowKeyValue,
    });
}
export function handleDeselectRow(event) {
    event.stopPropagation();
    const { state } = this;
    const { rowKeyValue, isMultiple } = event.detail;
    const fromRowKey = isMultiple
        ? getLastRowSelection(state) || rowKeyValue
        : rowKeyValue;
    markDeselectedRowsInterval(state, fromRowKey, rowKeyValue);
    state.lastSelectedRowKey = rowKeyValue;
    this.fireSelectedRowsChange(this.getSelectedRows(), {
        action: ROWS_ACTION.ROW_DESELECT,
        value: rowKeyValue,
    });
}
export function handleRowSelectionChange() {
    updateBulkSelectionState(this.state);
}
export function markAllRowsSelected(state) {
    const { enableDisabledRowsAttribute } = features;
    const { disabledRowsKeys, maxRowSelection, rows, selectedRowsKeys } = state;
    const newSelectedRowsKeys = {};
    let numRowsSelected = getDisabledAndSelectedRowCount(state);
    const remainingEditSlots = getRemainingEditSlots(state);
    let newSelectionsCount = 0;
    for (let i = 0, { length: rowCount } = rows; i < rowCount; i += 1) {
        const row = rows[i];
        const selectedRowsCounter = enableDisabledRowsAttribute
            ? numRowsSelected
            : i;
        if (enableDisabledRowsAttribute && disabledRowsKeys[row.key]) {
            if (selectedRowsKeys[row.key]) {
                newSelectedRowsKeys[row.key] = true;
            }
        }
        else if ((maxRowSelection === undefined ||
            selectedRowsCounter < maxRowSelection) &&
            (!isNewEditSelection(state, row.key) ||
                newSelectionsCount < remainingEditSlots)) {
            newSelectedRowsKeys[row.key] = true;
            setRowSelectedAttributes(row, true);
            if (enableDisabledRowsAttribute) {
                numRowsSelected += 1;
            }
            if (isNewEditSelection(state, row.key)) {
                newSelectionsCount += 1;
            }
        }
        else {
            row.isDisabled = true;
            setRowSelectedAttributes(row, false);
        }
    }
    state.selectedRowsKeys = newSelectedRowsKeys;
}
export function markAllRowsDeselected(state, deselectDisabledRows = true) {
    const { enableDisabledRowsAttribute } = features;
    const { disabledRowsKeys, rows, selectedRowsKeys } = state;
    const newSelectedRowsKeys = {};
    for (let i = 0, { length: rowCount } = rows; i < rowCount; i += 1) {
        const row = rows[i];
        if (!enableDisabledRowsAttribute) {
            row.isDisabled = false;
            setRowSelectedAttributes(row, false);
        }
        else {
            const shouldEnable = !disabledRowsKeys[row.key];
            if (shouldEnable || deselectDisabledRows) {
                if (shouldEnable) {
                    row.isDisabled = false;
                }
                setRowSelectedAttributes(row, false);
            }
            else if (!shouldEnable && selectedRowsKeys[row.key]) {
                newSelectedRowsKeys[row.key] = true;
            }
        }
    }
    state.selectedRowsKeys = newSelectedRowsKeys;
}
function markSelectedRowsInterval(state, startRowKey, endRowKey) {
    const { enableDisabledRowsAttribute } = features;
    const { disabledRowsKeys, rows } = state;
    const { start, end } = getRowIntervalIndexes(state, startRowKey, endRowKey);
    const total = state.maxRowSelection || rows.length;
    let rowIndex = start;
    let maxSelectionReached;
    while (rowIndex <= end && !maxSelectionReached) {
        const { key } = rows[rowIndex];
        if (!enableDisabledRowsAttribute || !disabledRowsKeys[key]) {
            markRowSelected(state, key);
        }
        maxSelectionReached = getCurrentSelectionLength(state) >= total;
        rowIndex += 1;
    }
}
function markDeselectedRowsInterval(state, startRowKey, endRowKey) {
    const { enableDisabledRowsAttribute } = features;
    const { disabledRowsKeys, rows, selectedRowsKeys } = state;
    const { start, end } = getRowIntervalIndexes(state, startRowKey, endRowKey);
    for (let rowIndex = start; rowIndex <= end; rowIndex += 1) {
        const row = rows[rowIndex];
        if (!enableDisabledRowsAttribute || !disabledRowsKeys[row.key]) {
            setRowSelectedAttributes(row, false);
            selectedRowsKeys[row.key] = false;
        }
        if (getCurrentSelectionLength(state) === state.maxRowSelection - 1) {
            markDeselectedRowEnabled(state);
        }
    }
}
export function markRowSelected(state, rowKeyValue) {
    const row = getRowByKey(state, rowKeyValue);
    const prevSelectionLength = getCurrentSelectionLength(state);
    const total = state.maxRowSelection || state.rows.length;
    setRowSelectedAttributes(row, true);
    let { selectedRowsKeys } = state;
    const { singleRowSelectionMode } = state;
    if (total > 1 ||
        (total === 1 &&
            singleRowSelectionMode === SINGLE_ROW_SELECTION_MODES.CHECKBOX)) {
        if (!selectedRowsKeys[row.key]) {
            selectedRowsKeys[row.key] = true;
            if (prevSelectionLength + 1 === total) {
                markDeselectedRowDisabled(state);
            }
        }
    }
    else {
        if (prevSelectionLength === 1) {
            const prevSelectedRow = getRowByKey(state, Object.keys(selectedRowsKeys)[0]);
            setRowSelectedAttributes(prevSelectedRow, false);
            selectedRowsKeys = {};
            state.selectedRowsKeys = selectedRowsKeys;
        }
        selectedRowsKeys[row.key] = true;
    }
}
function markRowsSelectedByKeys(state, keys) {
    for (let i = 0, { length } = keys; i < length; i += 1) {
        const row = getRowByKey(state, keys[i]);
        setRowSelectedAttributes(row, true);
    }
}
function markRowsDeselectedByKeys(state, keys) {
    for (let i = 0, { length } = keys; i < length; i += 1) {
        const row = getRowByKey(state, keys[i]);
        setRowSelectedAttributes(row, false);
    }
}
export function markDeselectedRowDisabled(state) {
    if (features.enableMaxEditLimit) {
        setDisabledRows(state);
    }
    else {
        const { rows, selectedRowsKeys } = state;
        for (let i = 0, { length: rowCount } = rows; i < rowCount; i += 1) {
            const row = rows[i];
            if (!selectedRowsKeys[row.key]) {
                row.isDisabled = true;
            }
        }
    }
}
export function markDeselectedRowEnabled(state) {
    const { enableDisabledRowsAttribute, enableMaxEditLimit } = features;
    if (enableMaxEditLimit) {
        setDisabledRows(state);
    }
    else {
        const { rows, selectedRowsKeys, disabledRowsKeys } = state;
        for (let i = 0, { length: rowCount } = rows; i < rowCount; i += 1) {
            const row = rows[i];
            if (enableDisabledRowsAttribute) {
                if (!selectedRowsKeys[row.key] && !disabledRowsKeys[row.key]) {
                    row.isDisabled = false;
                }
            }
            else if (!selectedRowsKeys[row.key]) {
                row.isDisabled = false;
            }
        }
    }
}
export function setDisabledRowsKeys(state, value) {
    if (Array.isArray(value)) {
        const disabledRows = filterValidKeys(state, value);
        updateDisabledRowsByKeys(state, disabledRows);
    }
    else {
        console.error(`The "disabled-rows" passed into "lightning-datatable" must be an Array with the keys of the disabled rows. We instead received ${value}`);
        markAllRowsEnabled(state);
    }
    updateBulkSelectionState(state);
    if (isSelectedRadioDisabled(state)) {
        markDeselectedRowDisabled(state);
    }
}
function updateDisabledRowsByKeys(state, newKeysArr) {
    const newKeysObj = {};
    const { disabledRowsKeys: oldKeysObj } = state;
    for (let i = 0, { length } = newKeysArr; i < length; i += 1) {
        const rowKeyValue = newKeysArr[i];
        newKeysObj[rowKeyValue] = true;
        if (!oldKeysObj[rowKeyValue]) {
            const row = getRowByKey(state, rowKeyValue);
            if (row) {
                row.isDisabled = true;
            }
        }
    }
    const oldKeysArr = Object.keys(oldKeysObj);
    for (let i = 0, { length } = oldKeysArr; i < length; i += 1) {
        const rowKeyValue = oldKeysArr[i];
        if (oldKeysObj[rowKeyValue] && !newKeysObj[rowKeyValue]) {
            const row = getRowByKey(state, rowKeyValue);
            if (row) {
                row.isDisabled = false;
            }
        }
    }
    state.disabledRowsKeys = newKeysObj;
}
function markAllRowsEnabled(state) {
    const { rows } = state;
    for (let i = 0, { length: rowCount } = rows; i < rowCount; i += 1) {
        const row = rows[i];
        row.isDisabled = false;
    }
    state.disabledRowsKeys = {};
}
function getDisabledRowCount(state) {
    if (!features.enableDisabledRowsAttribute) {
        return 0;
    }
    const { disabledRowsKeys } = state;
    const keys = Object.keys(disabledRowsKeys);
    let disabledCount = 0;
    for (let i = 0; i < keys.length; i += 1) {
        const key = keys[i];
        if (disabledRowsKeys[key]) {
            disabledCount += 1;
        }
    }
    return disabledCount;
}
function getDisabledAndSelectedRowCount(state) {
    if (!features.enableDisabledRowsAttribute) {
        return 0;
    }
    const { disabledRowsKeys, selectedRowsKeys } = state;
    const keys = Object.keys(disabledRowsKeys);
    let disabledAndSelectedCount = 0;
    for (let i = 0; i < keys.length; i += 1) {
        const key = keys[i];
        if (disabledRowsKeys[key] && selectedRowsKeys[key]) {
            disabledAndSelectedCount += 1;
        }
    }
    return disabledAndSelectedCount;
}
function allRowsDisabled(state) {
    if (!features.enableDisabledRowsAttribute) {
        return false;
    }
    const { rows } = state;
    const numRows = rows.length;
    const numDisabledRows = getDisabledRowCount(state);
    return numRows === numDisabledRows;
}
function isSelectedRadioDisabled(state) {
    if (!features.enableDisabledRowsAttribute) {
        return false;
    }
    return (!isInputTypeCheckbox(state) &&
        getDisabledAndSelectedRowCount(state) === 1);
}
export function setSelectedRowsKeys(state, value) {
    if (Array.isArray(value)) {
        const { maxRowSelection } = state;
        const previousSelectionLength = getCurrentSelectionLength(state);
        let selectedRows = filterValidKeys(state, value);
        if (selectedRows.length > maxRowSelection) {
            console.warn(`The number of keys in selectedRows for lightning:datatable exceeds the limit defined by maxRowSelection.`);
            selectedRows = selectedRows.slice(0, maxRowSelection);
        }
        if (features.enableMaxEditLimit &&
            !isUndefinedOrNull(state.maxEditLimit)) {
            const currentEditedRowsCount = getCurrentEditedRowsCount(state);
            const maxEditableRows = state.maxEditLimit;
            if (currentEditedRowsCount + selectedRows.length >
                maxEditableRows) {
                console.warn(`The number of selected rows would exceed the limit defined by maxEditLimit (${maxEditableRows}).`);
                selectedRows = selectedRows.slice(0, Math.max(0, maxEditableRows - currentEditedRowsCount));
            }
        }
        const selectedRowsKeys = normalizeSelectedRowsKey(selectedRows);
        const selectionOperations = getSelectedDiff(state, selectedRows);
        const deselectionOperations = getDeselectedDiff(state, selectedRowsKeys);
        markRowsSelectedByKeys(state, selectionOperations);
        markRowsDeselectedByKeys(state, deselectionOperations);
        state.selectedRowsKeys = selectedRowsKeys;
        const { length: selectedRowsCount } = selectedRows;
        if ((selectedRowsCount === maxRowSelection &&
            isInputTypeCheckbox(state)) ||
            isSelectedRadioDisabled(state)) {
            markDeselectedRowDisabled(state);
        }
        else if (selectedRowsCount < maxRowSelection &&
            previousSelectionLength === maxRowSelection) {
            markDeselectedRowEnabled(state);
        }
    }
    else {
        console.error(`The "selectedRows" passed into "lightning:datatable" must be an Array with the keys of the selected rows. We receive instead ${value}`);
        markAllRowsDeselected(state);
    }
}
export function syncSelectedRowsKeys(state, selectedRows) {
    let changed = false;
    const { keyField, selectedRowsKeys, maxRowSelection } = state;
    const { length: selectedRowCount } = selectedRows;
    if (Object.keys(selectedRowsKeys).length !== selectedRowCount) {
        changed = true;
        state.selectedRowsKeys = updateSelectedRowsKeysFromSelectedRows(selectedRows, keyField);
    }
    else {
        for (let i = 0; i < selectedRowCount; i += 1) {
            const row = selectedRows[i];
            if (!selectedRowsKeys[row[keyField]]) {
                changed = true;
                state.selectedRowsKeys = updateSelectedRowsKeysFromSelectedRows(selectedRows, keyField);
                break;
            }
        }
    }
    if (changed) {
        const total = maxRowSelection || state.rows.length;
        if (isInputTypeCheckbox(state)) {
            if (selectedRowCount < total) {
                markDeselectedRowEnabled(state);
            }
            else {
                markDeselectedRowDisabled(state);
            }
        }
    }
    updateBulkSelectionState(state);
    return {
        ifChanged(callback) {
            if (changed && typeof callback === 'function') {
                callback(selectedRows);
            }
        },
    };
}
function updateSelectedRowsKeysFromSelectedRows(selectedRows, keyField) {
    const selectedRowsKeys = {};
    for (let i = 0, { length } = selectedRows; i < length; i += 1) {
        const row = selectedRows[i];
        selectedRowsKeys[row[keyField]] = true;
    }
    return selectedRowsKeys;
}
function normalizeSelectedRowsKey(keys) {
    const selectedRowsKeys = {};
    for (let i = 0, { length } = keys; i < length; i += 1) {
        selectedRowsKeys[keys[i]] = true;
    }
    return selectedRowsKeys;
}
function filterValidKeys(state, keys) {
    const filtered = [];
    const { rows } = state;
    if (rows.length) {
        const { indexes } = state;
        for (let i = 0, { length } = keys; i < length; i += 1) {
            const key = keys[i];
            if (indexes[key]) {
                filtered.push(key);
            }
        }
    }
    return filtered;
}
function getLastRowSelection(state) {
    const { lastSelectedRowKey } = state;
    const keyIsValid = lastSelectedRowKey !== undefined &&
        getRowByKey(state, lastSelectedRowKey) !== undefined;
    return keyIsValid ? lastSelectedRowKey : undefined;
}
export function inputTypeNeedsToChange(previousMaxRowSelection, newMaxRowSelection) {
    return (previousMaxRowSelection === 0 ||
        newMaxRowSelection === 0 ||
        (previousMaxRowSelection === 1 &&
            (newMaxRowSelection > 1 || newMaxRowSelection === undefined)) ||
        (newMaxRowSelection === 1 &&
            (previousMaxRowSelection > 1 ||
                previousMaxRowSelection === undefined)));
}
export function updateRowSelectionInputType(state, datatableId) {
    const isCheckbox = isInputTypeCheckbox(state);
    const type = isCheckbox
        ? SINGLE_ROW_SELECTION_MODES.CHECKBOX
        : SINGLE_ROW_SELECTION_MODES.RADIO;
    const { rows } = state;
    let resetInliningProps = false;
    let checkboxColIndex;
    if (state.renderModeInline && !state.hideCheckboxColumn) {
        resetInliningProps = true;
        checkboxColIndex = state.showRowNumberColumn ? 1 : 0;
    }
    if (features.enableMaxEditLimit) {
        setDisabledRows(state);
    }
    else {
        for (let i = 0, { length: rowCount } = rows; i < rowCount; i += 1) {
            const row = rows[i];
            row.inputType = type;
            row.isDisabled = isDisabledRow(state, row.key);
            if (resetInliningProps) {
                setInliningProperties(state, row.cells[checkboxColIndex], isCheckbox, datatableId);
            }
        }
    }
}
export function setMaxRowSelection(state, value, datatableId) {
    let previousSelectedRowsKeys;
    if (features.enableDisabledRowsAttribute) {
        previousSelectedRowsKeys = getRowsKeys(state.selectedRowsKeys);
    }
    else {
        previousSelectedRowsKeys = getSelectedRowsKeys(state);
    }
    markAllRowsDeselected(state);
    if (isNonNegativeInteger(value)) {
        const previousMaxRowSelection = state.maxRowSelection;
        const newMaxRowSelection = Number(value);
        state.maxRowSelection = newMaxRowSelection;
        const numberOfRows = Math.min(previousSelectedRowsKeys.length, newMaxRowSelection);
        for (let i = 0; i < numberOfRows; i += 1) {
            markRowSelected(state, previousSelectedRowsKeys[i]);
        }
        if (inputTypeNeedsToChange(previousMaxRowSelection, newMaxRowSelection)) {
            updateRowSelectionInputType(state, datatableId);
            updateBulkSelectionState(state);
            if (isSelectedRadioDisabled(state)) {
                markDeselectedRowDisabled(state);
            }
        }
    }
    else {
        state.maxRowSelection = MAX_ROW_SELECTION_DEFAULT;
        if (value !== null && value !== undefined) {
            console.error(`The maxRowSelection value passed into lightning:datatable should be a positive integer. We receive instead (${value}).`);
        }
    }
}
export function setSingleRowSelectionMode(state, value, datatableId) {
    const resolvedValue = normalizeString(value, {
        fallbackValue: SINGLE_ROW_SELECTION_MODES.RADIO,
        validValues: [
            SINGLE_ROW_SELECTION_MODES.CHECKBOX,
            SINGLE_ROW_SELECTION_MODES.RADIO,
        ],
    });
    if (resolvedValue !== state.singleRowSelectionMode) {
        state.singleRowSelectionMode = value;
        updateRowSelectionInputType(state, datatableId);
        updateBulkSelectionState(state);
    }
}
export function updateBulkSelectionState(state) {
    const columns = state.columns || {};
    for (let colIndex = 0; colIndex < columns.length; colIndex += 1) {
        const col = columns[colIndex];
        if (col.type === SELECTABLE_HEADER_TYPE) {
            const { maxRowSelection, rows: { length: rowCount }, } = state;
            const total = getMaxSelectableRows(state);
            const maxSelectableForLabel = getMaxSelectableRows(state);
            const selected = getCurrentSelectionLength(state);
            const selectedAndDisabled = getDisabledAndSelectedRowCount(state);
            const updatedCol = Object.assign({}, col);
            if (selected) {
                if (selected - selectedAndDisabled) {
                    updatedCol.enabledRowsSelected = true;
                }
                else {
                    updatedCol.enabledRowsSelected = false;
                }
                updatedCol.bulkSelection = selected === total ? 'all' : 'some';
            }
            else {
                updatedCol.bulkSelection = 'none';
                updatedCol.enabledRowsSelected = false;
            }
            updatedCol.isBulkSelectionDisabled =
                maxRowSelection === 0 ||
                    rowCount === 0 ||
                    allRowsDisabled(state);
            updatedCol.maxSelectableRows = maxSelectableForLabel;
            columns[colIndex] = updatedCol;
            return;
        }
    }
}
function getRowIntervalIndexes(state, startRowKey, endRowKey) {
    const start = startRowKey === HEADER_ROW_KEY
        ? 0
        : getRowIndexByKey(state, startRowKey);
    const end = getRowIndexByKey(state, endRowKey);
    return {
        start: Math.min(start, end),
        end: Math.max(start, end),
    };
}
export function setAriaSelectedOnCell(state, rowKeyValue, colKeyValue) {
    const cell = getCellByKeys(state, rowKeyValue, colKeyValue);
    if (cell) {
        cell.ariaSelected = 'true';
    }
}
export function unsetAriaSelectedOnCell(state, rowKeyValue, colKeyValue) {
    const cell = getCellByKeys(state, rowKeyValue, colKeyValue);
    if (cell) {
        cell.ariaSelected = false;
    }
}
function setRowSelectedAttributes(row, selectedValue) {
    row.ariaSelected = selectedValue;
    row.classnames = `slds-hint-parent${selectedValue ? ' slds-is-selected' : ''}`;
    row.isSelected = selectedValue;
}
function getSelectedDiff(state, selectedRows) {
    const filtered = [];
    const { selectedRowsKeys } = state;
    for (let i = 0, { length } = selectedRows; i < length; i += 1) {
        const rowKeyValue = selectedRows[i];
        if (!selectedRowsKeys[rowKeyValue]) {
            filtered.push(rowKeyValue);
        }
    }
    return filtered;
}
function getDeselectedDiff(state, value) {
    const filtered = [];
    const { selectedRowsKeys } = state;
    const keys = Object.keys(selectedRowsKeys);
    for (let i = 0, { length } = keys; i < length; i += 1) {
        const rowKeyValue = keys[i];
        if (selectedRowsKeys[rowKeyValue] && !value[rowKeyValue]) {
            filtered.push(rowKeyValue);
        }
    }
    return filtered;
}
export function updateRowDisabledStatesForEditLimit(state, types) {
    if (!features.enableMaxEditLimit ||
        isUndefinedOrNull(state.maxEditLimit) ||
        !state.rows?.length) {
        return;
    }
    setDisabledRows(state);
    updateCells(state, types);
}
export function getMaxSelectableRows(state) {
    const { maxRowSelection, rows: { length: rowCount }, } = state;
    let maxSelectable = rowCount;
    if (features.enableDisabledRowsAttribute) {
        const disabledCount = getDisabledRowCount(state);
        const disabledAndSelectedCount = getDisabledAndSelectedRowCount(state);
        maxSelectable = rowCount - disabledCount + disabledAndSelectedCount;
    }
    if (!isUndefinedOrNull(maxRowSelection)) {
        maxSelectable = Math.min(maxSelectable, maxRowSelection);
    }
    if (features.enableMaxEditLimit && !isUndefinedOrNull(state.maxEditLimit)) {
        maxSelectable = Math.min(maxSelectable, state.maxEditLimit);
    }
    return Math.max(0, maxSelectable);
}
