// lightning-base-components/src/lightning/datatable/rowSelectionShared.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { features } from 'lightning/features';
import { isUndefinedOrNull } from 'lightning/utilsPrivate';
export const SELECTABLE_HEADER_TYPE = 'SELECTABLE_CHECKBOX';
export const SORTING_MENU_HEADER_TYPE = 'SELECTABLE_BUTTON_MENU';
export const SINGLE_ROW_SELECTION_MODES = {
    CHECKBOX: 'checkbox',
    RADIO: 'radio',
};
const DISABLE_SCENARIOS = {
    DISABLE_UNSELECTED: 'disable-unselected',
    DISABLE_UNEDITED: 'disable-unedited',
    DISABLE_UNSELECTED_AND_UNEDITED: 'disable-unselected-and-unedited',
    NONE: 'none',
};
export function isSelectedRow(state, rowKeyValue) {
    return !!state.selectedRowsKeys[rowKeyValue];
}
function countSelectedButNotEditedRows(state) {
    let count = 0;
    if (state.selectedRowsKeys) {
        const selectedRowsKeys = Object.keys(state.selectedRowsKeys);
        for (let i = 0; i < selectedRowsKeys.length; i += 1) {
            const selectedRowKey = selectedRowsKeys[i];
            if (state.selectedRowsKeys[selectedRowKey] &&
                !state.inlineEdit?.dirtyValues?.[selectedRowKey]) {
                count += 1;
            }
        }
    }
    return count;
}
export function getCurrentEditedRowsCount(state) {
    return Object.keys(state.inlineEdit?.dirtyValues || {}).length;
}
export function getRemainingEditSlots(state) {
    if (!features.enableMaxEditLimit) {
        return Infinity;
    }
    const { maxEditLimit } = state;
    if (isUndefinedOrNull(maxEditLimit)) {
        return Infinity;
    }
    return Math.max(0, maxEditLimit - getCurrentEditedRowsCount(state));
}
export function isDisabledRow(state, rowKeyValue) {
    const { disabledRowsKeys } = state;
    if (features.enableDisabledRowsAttribute && disabledRowsKeys[rowKeyValue]) {
        return true;
    }
    if (!state.selectedRowsKeys[rowKeyValue]) {
        const { maxRowSelection } = state;
        return (isInputTypeCheckbox(state) &&
            getCurrentSelectionLength(state) === maxRowSelection);
    }
    return false;
}
export function isRowDisabledForEditLimit(state, rowKeyValue) {
    const { disabledRowsKeys } = state;
    if (features.enableDisabledRowsAttribute && disabledRowsKeys[rowKeyValue]) {
        return false;
    }
    if (state.inlineEdit?.dirtyValues?.[rowKeyValue]) {
        return false;
    }
    if (state.selectedRowsKeys && state.selectedRowsKeys[rowKeyValue]) {
        return false;
    }
    const maxEditLimitReached = isMaxEditLimitMet(state);
    return maxEditLimitReached;
}
function isMaxEditLimitMet(state) {
    const { maxEditLimit } = state;
    if (isUndefinedOrNull(maxEditLimit)) {
        return false;
    }
    const currentEditedRowsCount = getCurrentEditedRowsCount(state);
    const selectedButNotEditedCount = countSelectedButNotEditedRows(state);
    const totalCurrentlyEditable = currentEditedRowsCount + selectedButNotEditedCount;
    return totalCurrentlyEditable >= maxEditLimit;
}
export function getCurrentSelectionLength(state) {
    let filteredLength = 0;
    const { selectedRowsKeys } = state;
    const keys = Object.keys(selectedRowsKeys);
    for (let i = 0, { length } = keys; i < length; i += 1) {
        if (selectedRowsKeys[keys[i]]) {
            filteredLength += 1;
        }
    }
    return filteredLength;
}
export function getRowsKeys(rowKeysObj) {
    const filtered = [];
    const keys = Object.keys(rowKeysObj);
    for (let i = 0, { length } = keys; i < length; i += 1) {
        const rowKeyValue = keys[i];
        if (rowKeysObj[rowKeyValue]) {
            filtered.push(rowKeyValue);
        }
    }
    return filtered;
}
export function getSelectedRowsKeys(state) {
    return getRowsKeys(state.selectedRowsKeys);
}
export function isInputTypeCheckbox(state) {
    return (state.maxRowSelection !== 1 ||
        state.singleRowSelectionMode === SINGLE_ROW_SELECTION_MODES.CHECKBOX);
}
export function setDisabledRows(state) {
    const { rows, disabledRowsKeys } = state;
    if (!rows || !rows.length) {
        return;
    }
    const isCheckbox = isInputTypeCheckbox(state);
    const inputType = isCheckbox
        ? SINGLE_ROW_SELECTION_MODES.CHECKBOX
        : SINGLE_ROW_SELECTION_MODES.RADIO;
    const disableScenario = calculateDisableScenario(state);
    for (let i = 0, { length: rowCount } = rows; i < rowCount; i += 1) {
        const row = rows[i];
        const { key: rowKeyValue } = row;
        row.inputType = inputType;
        if (features.enableDisabledRowsAttribute &&
            disabledRowsKeys[rowKeyValue]) {
            row.isDisabled = true;
        }
        else {
            row.isDisabled = shouldDisableRowForScenario(disableScenario, state, rowKeyValue);
        }
    }
}
function calculateDisableScenario(state) {
    const { maxRowSelection, maxEditLimit } = state;
    const currentSelectionCount = getCurrentSelectionLength(state);
    const maxRowSelectionReached = isInputTypeCheckbox(state) && currentSelectionCount === maxRowSelection;
    if (features.enableMaxEditLimit && !isUndefinedOrNull(maxEditLimit)) {
        const maxEditLimitReached = isMaxEditLimitMet(state);
        if (maxRowSelectionReached) {
            return DISABLE_SCENARIOS.DISABLE_UNSELECTED;
        }
        else if (maxEditLimitReached) {
            return DISABLE_SCENARIOS.DISABLE_UNEDITED;
        }
    }
    else if (maxRowSelectionReached) {
        return DISABLE_SCENARIOS.DISABLE_UNSELECTED;
    }
    return DISABLE_SCENARIOS.NONE;
}
function shouldDisableRowForScenario(disableScenario, state, rowKeyValue) {
    const { selectedRowsKeys } = state;
    const isSelected = !!selectedRowsKeys[rowKeyValue];
    const isEdited = !!state.inlineEdit?.dirtyValues?.[rowKeyValue];
    switch (disableScenario) {
        case DISABLE_SCENARIOS.DISABLE_UNSELECTED: {
            return !isSelected;
        }
        case DISABLE_SCENARIOS.DISABLE_UNEDITED: {
            return !(isEdited || isSelected);
        }
        case DISABLE_SCENARIOS.NONE:
        default: {
            return false;
        }
    }
}
