// lightning-base-components/src/lightning/datatable/state.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
export function getDefaultState() {
    return {
        columns: [],
        hadTreeDataTypePreviously: false,
        hideCheckboxColumn: false,
        showActionsMenu: false,
        treeColumn: undefined,
        data: [],
        hasCalledUpdateRowsAndCells: false,
        keyField: undefined,
        rows: [],
        checkboxCells: [],
        lookupCells: [],
        shownUrlCells: [],
        hasHighlightStyling: false,
        hoveredRowKeyValue: undefined,
        focusedRowKeyValue: undefined,
        disabledRowsKeys: {},
        selectedRowsKeys: {},
        lastSelectedRowKey: undefined,
        maxRowSelection: undefined,
        singleRowSelectionMode: undefined,
        maxEditLimit: null,
        hideBorders: false,
        hideTableHeader: false,
        wrapTableHeader: 'none',
        headerIndexes: {},
        indexes: {},
        keyboardMode: 'NAVIGATION',
        rowMode: false,
        activeCell: undefined,
        tabindex: 0,
        cellToFocusNext: null,
        cellClicked: false,
        normalized: false,
        wrapTextMaxLines: undefined,
        sortedBy: undefined,
        sortedDirection: undefined,
        defaultSortDirection: 'asc',
        showRowNumberColumn: false,
        rowNumberOffset: 0,
        enableInfiniteLoading: false,
        loadMoreOffset: 20,
        isLoading: false,
        renderModeInline: false,
        renderModeRoleBased: false,
        enableViewportRendering: undefined,
        virtualize: '',
        bufferSize: 5,
        rowHeight: 30.5,
        renderedRowCount: 0,
        firstVisibleIndex: 0,
        heightCache: {},
        offsets: [0],
        offsetRanges: [],
        firstRowOffset: 0,
        tableHeight: 0,
        shouldResetHeights: false,
        inlineEdit: {
            rowKeyValue: undefined,
            colKeyValue: undefined,
            columnDef: {},
            dirtyValues: {},
            editedValue: undefined,
            isPanelVisible: false,
            massEditEnabled: false,
            massEditSelectedRows: undefined,
            resolvedAttributeTypes: {},
        },
        errors: {
            rows: {},
            table: {},
        },
        hasReceivedFocus: false,
        hasRenderedFirstTime: false,
    };
}
