// lightning-base-components/src/lightning/datatable/types.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { assert } from 'lightning/utilsPrivate';
const STANDARD_TYPE_NAME_ENTRIES = [
    ['text', ['linkify']],
    ['boolean', []],
    [
        'number',
        [
            'minimumIntegerDigits',
            'minimumFractionDigits',
            'maximumFractionDigits',
            'minimumSignificantDigits',
            'maximumSignificantDigits',
        ],
    ],
    [
        'currency',
        [
            'currencyCode',
            'currencyDisplayAs',
            'minimumIntegerDigits',
            'minimumFractionDigits',
            'maximumFractionDigits',
            'minimumSignificantDigits',
            'maximumSignificantDigits',
        ],
    ],
    [
        'percent',
        [
            'minimumIntegerDigits',
            'minimumFractionDigits',
            'maximumFractionDigits',
            'minimumSignificantDigits',
            'maximumSignificantDigits',
        ],
    ],
    ['email', []],
    [
        'date',
        [
            'day',
            'era',
            'hour',
            'hour12',
            'minute',
            'month',
            'second',
            'timeZone',
            'timeZoneName',
            'weekday',
            'year',
        ],
    ],
    ['date-local', ['day', 'month', 'year']],
    ['phone', []],
    ['url', ['label', 'target', 'tooltip']],
    ['location', []],
    ['reference', ['displayValue']],
    ['rowNumber', ['error']],
    ['action', ['menuAlignment', 'rowActions']],
    [
        'button',
        [
            'variant',
            'label',
            'iconName',
            'iconPosition',
            'disabled',
            'name',
            'class',
            'title',
        ],
    ],
    [
        'button-icon',
        [
            'variant',
            'alternativeText',
            'iconName',
            'iconClass',
            'disabled',
            'name',
            'class',
            'title',
        ],
    ],
    [
        'tree',
        [
            'hasChildren',
            'isExpanded',
            'level',
            'setSize',
            'posInSet',
            'subType',
        ],
    ],
];
const STANDARD_TYPE_VALUE_ENTRIES = Array(STANDARD_TYPE_NAME_ENTRIES.length);
for (let i = 0, { length } = STANDARD_TYPE_NAME_ENTRIES; i < length; i += 1) {
    const { 0: typeName, 1: typeAttributes } = STANDARD_TYPE_NAME_ENTRIES[i];
    STANDARD_TYPE_VALUE_ENTRIES[i] = [
        typeName,
        {
            typeAttributes,
            type: 'standard',
        },
    ];
}
const STANDARD_TYPE_NAMES = new Map(STANDARD_TYPE_NAME_ENTRIES);
const STANDARD_TYPE_VALUES = new Map(STANDARD_TYPE_VALUE_ENTRIES);
const TREE_SUPPORTED_TYPES = new Set([
    'text',
    'url',
    'date',
    'number',
    'currency',
    'percent',
    'button',
    'button-icon',
    'reference',
]);
const EDITABLE_STANDARD_TYPES = new Set([
    'text',
    'percent',
    'phone',
    'email',
    'url',
    'currency',
    'number',
    'boolean',
    'date-local',
    'date',
]);
export default class DatatableTypes {
    privateCustomTypes = new Map();
    constructor(types) {
        if (typeof types === 'object' && types !== null) {
            const { privateCustomTypes } = this;
            const keys = Object.keys(types);
            for (let i = 0, { length } = keys; i < length; i += 1) {
                const key = keys[i];
                const { template, editTemplate, typeAttributes = [], standardCellLayout = false, } = types[key];
                privateCustomTypes.set(key, {
                    template,
                    editTemplate,
                    typeAttributes,
                    standardCellLayout: standardCellLayout === true,
                    type: 'custom',
                });
            }
        }
    }
    getAttributesNames(typeName) {
        assert(this.isValidType(typeName), `You are trying to access an invalid type (${typeName})`);
        if (!this.isCustomType(typeName)) {
            return STANDARD_TYPE_NAMES.get(typeName) || [];
        }
        const typeData = this.privateCustomTypes.get(typeName);
        return typeData.typeAttributes || [];
    }
    getType(typeName) {
        const { privateCustomTypes } = this;
        const customType = privateCustomTypes.get === Map.prototype.get &&
            privateCustomTypes.get(typeName);
        return customType || STANDARD_TYPE_VALUES.get(typeName);
    }
    getCustomTypeEditTemplate(typeName) {
        const privateType = this.privateCustomTypes.get(typeName);
        return privateType ? privateType.editTemplate : undefined;
    }
    isValidType(typeName) {
        return (this.privateCustomTypes.has(typeName) ||
            STANDARD_TYPE_NAMES.has(typeName));
    }
    isEditableType(typeName) {
        if (EDITABLE_STANDARD_TYPES.has(typeName)) {
            return true;
        }
        const privateType = this.privateCustomTypes.get(typeName);
        return privateType ? privateType.standardCellLayout : false;
    }
    isCustomType(typeName) {
        return this.privateCustomTypes.has(typeName);
    }
    isStandardCellLayoutForCustomType(typeName) {
        const privateType = this.privateCustomTypes.get(typeName);
        return privateType ? privateType.standardCellLayout : false;
    }
    isValidTypeForTree(typeName) {
        return (TREE_SUPPORTED_TYPES.has(typeName) ||
            this.isStandardCellLayoutForCustomType(typeName));
    }
}
