// lightning-base-components/src/lightning/datatable/utils.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { getRowByKey } from './indexes';
import { classSetToString } from 'lightning/utilsPrivate';
import { features } from 'lightning/features';
const CLASS_SET_PROTOTYPE = {
    add(className) {
        if (typeof className === 'string') {
            this[className] = true;
        }
        else {
            Object.assign(this, className);
        }
        return this;
    },
    invert() {
        const keys = Object.keys(this);
        for (let i = 0, { length } = keys; i < length; i += 1) {
            const key = keys[i];
            this[key] = !this[key];
        }
        return this;
    },
    toString() {
        return classSetToString(this);
    },
};
const NON_NEGATIVE_INTEGER_REGEXP = /^\d+$/;
const POSITIVE_INTEGER_REGEXP = /^[0-9]*[1-9][0-9]*$/;
const RENDER_MODE_INLINE_REGEXP = /\binline\b/;
const RENDER_MODE_ROLE_BASED_REGEXP = /\brole-based\b/;
export function escapeDoubleQuotes(value) {
    if (typeof value == 'string') {
        return value.replace(/"/g, '\\"');
    }
    return value;
}
export const isObjectLike = function (value) {
    return typeof value === 'object' && value !== null;
};
export function classSet(config) {
    return (typeof config === 'string'
        ? { __proto__: CLASS_SET_PROTOTYPE, [config]: true }
        : Object.assign({ __proto__: CLASS_SET_PROTOTYPE }, config));
}
export function clamp(num, min, max) {
    return num <= min ? min : num >= max ? max : num;
}
export function getColDataSelector(colKeyValue) {
    return `[data-col-key-value="${escapeDoubleQuotes(colKeyValue)}"]`;
}
export function getRowDataSelector(rowKeyValue) {
    return `[data-row-key-value="${escapeDoubleQuotes(rowKeyValue)}"]`;
}
export function getGridContainerFromScrollerY(scrollerY) {
    if (!scrollerY) {
        return null;
    }
    const { firstElementChild: gridContainer } = scrollerY;
    return gridContainer;
}
export function getScrollerX(template, refs) {
    return ((refs && refs.scrollerX) ||
        template.querySelector('.slds-scrollable_x'));
}
export function getScrollerY(template, refs) {
    return ((refs && refs.scrollerY) ||
        template.querySelector('.slds-scrollable_y'));
}
export function isPositiveInteger(value) {
    return POSITIVE_INTEGER_REGEXP.test(value);
}
export function isNonNegativeInteger(value) {
    return NON_NEGATIVE_INTEGER_REGEXP.test(value);
}
export function isRenderModeInline(renderMode) {
    return RENDER_MODE_INLINE_REGEXP.test(renderMode);
}
export function isRenderModeRoleBased(renderMode) {
    return RENDER_MODE_ROLE_BASED_REGEXP.test(renderMode);
}
export function normalizeNumberAttribute(attrName, value, numberType, fallback) {
    let warningMessage;
    if (numberType === 'positive') {
        if (isPositiveInteger(value)) {
            return Number.parseInt(value, 10);
        }
        warningMessage = `The attribute "${attrName}" value passed in is incorrect. "${attrName}" value should be an integer > 0.`;
    }
    else if (numberType === 'non-negative') {
        if (isNonNegativeInteger(value)) {
            return Number.parseInt(value, 10);
        }
        warningMessage = `The attribute "${attrName}" value passed in is incorrect. "${attrName}" value should be an integer >= 0.`;
    }
    else {
        warningMessage =
            'Invalid number type during normalization of number attribute';
    }
    console.warn(warningMessage);
    return fallback;
}
export function getScrollOffsetFromTableEnd(el) {
    return (el.scrollHeight -
        el.parentNode.scrollTop -
        el.parentNode.clientHeight);
}
export function cleanData(value) {
    return Array.isArray(value) ? value : [];
}
export function isLoadMore(prevData, newData) {
    const cleanPrevData = cleanData(prevData);
    const cleanNewData = cleanData(newData);
    try {
        return cleanPrevData.length && cleanNewData.length
            ? JSON.stringify(cleanPrevData[0]) ===
                JSON.stringify(cleanNewData[0])
            : false;
    }
    catch (e) {
        console.warn(`Invalid data attribute, could not stringify`);
        return false;
    }
}
export function setHighlightedRowClasses(state, rowKeyValue) {
    if (!features.enableDatatableHoverStyling || !state.hasHighlightStyling) {
        return;
    }
    const row = getRowByKey(state, rowKeyValue);
    const cells = row?.cells || [];
    for (let i = 0; i < cells.length; i++) {
        const cell = row.cells[i];
        cell.computedClass = `${cell.highlightClass || cell.class} ${cell.defaultClass}`;
    }
}
export function unsetHighlightedRowClasses(state, rowKeyValue, isHoverEvent) {
    if (!features.enableDatatableHoverStyling ||
        !state.hasHighlightStyling ||
        !rowKeyValue) {
        return;
    }
    const otherRowKeyValue = isHoverEvent
        ? state.focusedRowKeyValue
        : state.hoveredRowKeyValue;
    if (otherRowKeyValue !== rowKeyValue) {
        const row = getRowByKey(state, rowKeyValue);
        const cells = row?.cells || [];
        for (let i = 0; i < cells.length; i++) {
            const cell = row.cells[i];
            cell.computedClass = `${cell.class} ${cell.defaultClass}`;
        }
    }
}
export function resetHighlightedRowClasses(state, rowKeyValue) {
    if (!features.enableDatatableHoverStyling || !state.hasHighlightStyling) {
        return;
    }
    const currentRowKeyValue = state.focusedRowKeyValue;
    if (currentRowKeyValue !== rowKeyValue) {
        if (currentRowKeyValue) {
            unsetHighlightedRowClasses(state, currentRowKeyValue);
        }
        if (rowKeyValue) {
            setHighlightedRowClasses(state, rowKeyValue);
        }
    }
    state.focusedRowKeyValue = rowKeyValue.toString();
}
