// lightning-base-components/src/lightning/datatable/virtualization.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
export function addTableHeight(state, prevData, newData) {
    const extraRowCount = newData.length - prevData.length;
    state.tableHeight += extraRowCount * state.rowHeight;
}
export function resetRowHeights(state) {
    const { firstVisibleIndex } = state;
    const firstRenderedIndex = Math.max(firstVisibleIndex - state.bufferSize, 0);
    const newOffsets = [0];
    const firstRenderedOffset = state.offsets[firstRenderedIndex];
    newOffsets[firstRenderedIndex] = firstRenderedOffset;
    state.heightCache = {};
    state.offsets = newOffsets;
    state.offsetRanges = [{ start: 0, end: 0 }];
    if (firstRenderedIndex > 0) {
        state.offsetRanges.push({
            start: firstRenderedIndex,
            end: firstRenderedIndex,
        });
    }
    if (state.virtualize && state.rows.length) {
        const remainingRows = state.rows.length - firstRenderedIndex;
        const remainingRowsHeight = state.rowHeight * remainingRows;
        state.tableHeight = firstRenderedOffset + remainingRowsHeight;
    }
}
export function handleVariableRowHeights(template, state, renderedRows, refs) {
    const currentRange = {
        start: renderedRows[0].rowIndex,
        end: renderedRows[renderedRows.length - 1].rowIndex + 1,
    };
    let adjustFromIndex;
    let adjustmentValue = 0;
    const offsetRangeIndex = findOffsetRangeIndex(state.offsetRanges, currentRange.start);
    const rowElements = template.querySelectorAll('[data-rowgroup-body] [role="row"]');
    for (let i = 0; i < renderedRows.length; i++) {
        const row = renderedRows[i];
        if (!state.heightCache[row.key] && state.heightCache[row.key] !== 0) {
            const rowElement = rowElements[i];
            if (rowElement) {
                let height = rowElement.getBoundingClientRect().height;
                if (row.rowIndex && row.rowIndex === currentRange.start) {
                    height++;
                }
                state.heightCache[row.key] = height;
                setOffset(state, row, offsetRangeIndex, height);
                adjustmentValue += height - state.rowHeight;
                adjustFromIndex = row.rowIndex + 2;
            }
        }
    }
    state.tableHeight += adjustmentValue;
    updateOffsetRanges(state, offsetRangeIndex, currentRange, adjustFromIndex, adjustmentValue);
    updateVirtualizeStyles(template, state, renderedRows, refs);
}
export function findFirstVisibleIndex(state, scrollTop) {
    const { offsetRanges, offsets, rowHeight } = state;
    let start = 0;
    let end = offsetRanges.length - 1;
    while (start <= end) {
        const mid = Math.floor((start + end) / 2);
        const prevRange = offsetRanges[mid];
        const nextRange = offsetRanges[mid + 1];
        const startOffset = offsets[prevRange.start];
        const endOffset = offsets[prevRange.end];
        const scrollTopAfterPrevStart = startOffset <= scrollTop;
        const scrollTopInPrevRange = scrollTopAfterPrevStart && scrollTop <= endOffset;
        const scrollTopBeforeNextRange = !nextRange || scrollTop < offsets[nextRange.start];
        const scrollTopBetweenRanges = scrollTopAfterPrevStart && scrollTopBeforeNextRange;
        if (scrollTopInPrevRange) {
            return searchForOffset(state, prevRange, scrollTop);
        }
        else if (scrollTopBetweenRanges) {
            const diff = scrollTop - endOffset;
            const extraRows = Math.floor(diff / rowHeight);
            state.firstRowOffset = diff % rowHeight;
            return prevRange.end + extraRows;
        }
        if (scrollTop < startOffset) {
            end = mid - 1;
        }
        else {
            start = mid + 1;
        }
    }
    return -1;
}
function searchForOffset(state, range, scrollTop) {
    let { start, end } = range;
    const offsets = state.offsets;
    while (start <= end) {
        const mid = Math.floor((start + end) / 2);
        const currentOffsetUnderScrollTop = offsets[mid] <= scrollTop;
        const nextOffsetUnderScrollTop = offsets[mid + 1] && offsets[mid + 1] <= scrollTop;
        if (currentOffsetUnderScrollTop && !nextOffsetUnderScrollTop) {
            state.firstRowOffset = scrollTop - offsets[mid];
            return mid;
        }
        else if (currentOffsetUnderScrollTop) {
            start = mid + 1;
        }
        else {
            end = mid - 1;
        }
    }
    return -1;
}
function findOffsetRangeIndex(offsetRanges, rowIndex) {
    let start = 0;
    let end = offsetRanges.length - 1;
    while (start <= end) {
        const mid = Math.floor((start + end) / 2);
        const currentRangeStartsBeforeRowIndex = offsetRanges[mid].start <= rowIndex;
        const nextRangeStartsBeforeRowIndex = offsetRanges[mid + 1] && offsetRanges[mid + 1].start <= rowIndex;
        if (currentRangeStartsBeforeRowIndex &&
            !nextRangeStartsBeforeRowIndex) {
            return mid;
        }
        else if (currentRangeStartsBeforeRowIndex) {
            start = mid + 1;
        }
        else {
            end = mid - 1;
        }
    }
    return -1;
}
function setOffset(state, row, offsetRangeIndex, height) {
    const currentRange = state.offsetRanges[offsetRangeIndex];
    let currentOffset = state.offsets[row.rowIndex];
    if (!currentOffset) {
        currentOffset = 0;
        if (currentRange) {
            const baseOffset = state.offsets[currentRange.end];
            const estimatedRowOffsets = (row.rowIndex - currentRange.end) * state.rowHeight;
            currentOffset = baseOffset + estimatedRowOffsets;
            state.offsets[row.rowIndex] = currentOffset;
        }
    }
    state.offsets[row.rowIndex + 1] = currentOffset + height;
}
function updateOffsetRanges(state, offsetRangeIndex, currentRange, adjustFromIndex, adjustmentValue) {
    const { offsets, offsetRanges } = state;
    const prevRange = offsetRanges[offsetRangeIndex];
    const nextRange = offsetRanges[offsetRangeIndex + 1];
    const overlapsWithPrevRange = prevRange && checkOverlap(prevRange, currentRange);
    const overlapsWithNextRange = nextRange && checkOverlap(currentRange, nextRange);
    if (overlapsWithNextRange && adjustmentValue) {
        for (let i = adjustFromIndex; i <= nextRange.end; i++) {
            offsets[i] += adjustmentValue;
        }
    }
    if (overlapsWithPrevRange && overlapsWithNextRange) {
        nextRange.start = prevRange.start;
        nextRange.end = Math.max(currentRange.end, nextRange.end);
        state.offsetRanges.splice(offsetRangeIndex, 1);
    }
    else if (overlapsWithPrevRange) {
        prevRange.end = Math.max(prevRange.end, currentRange.end);
    }
    else if (overlapsWithNextRange) {
        nextRange.start = currentRange.start;
        nextRange.end = Math.max(currentRange.end, nextRange.end);
        offsetRangeIndex += 1;
    }
    else {
        state.offsetRanges.splice(offsetRangeIndex + 1, 0, currentRange);
        offsetRangeIndex += 1;
    }
    if (offsetRangeIndex >= 0) {
        for (let i = offsetRangeIndex + 1; i < offsetRanges.length; i++) {
            const { start: rangeStart, end: rangeEnd } = offsetRanges[i];
            for (let j = rangeStart; j <= rangeEnd; j++) {
                offsets[j] += adjustmentValue;
            }
        }
    }
}
function checkOverlap(range1, range2) {
    return ((range1.start <= range2.start && range2.start <= range1.end) ||
        (range1.start <= range2.end && range2.end <= range1.end));
}
function updateVirtualizeStyles(template, state, renderedRows, refs) {
    const scrollerY = refs?.scrollerY ||
        template.querySelector('.slds-scrollable_y');
    scrollerY.scrollTop =
        state.offsets[state.firstVisibleIndex] + state.firstRowOffset;
    const addOffset = !!renderedRows[0].rowIndex;
    renderedRows.forEach((row, index) => {
        let top = state.offsets[row.rowIndex];
        if (addOffset && !index) {
            top += 1;
        }
        row.style = `position:absolute;top:${top}px;`;
    });
}
