// lightning-base-components/src/lightning/datatable/widthManagerShared.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
export function getTotalWidthsData(columnWidthMetaData, columns) {
    const totalWidthsData = {
        totalFixedWidth: 0,
        totalFixedColumns: 0,
        totalResizedWidth: 0,
        totalResizedColumns: 0,
        totalFlexibleColumns: 0,
        minColumnWidth: columnWidthMetaData.minColumnWidth,
        maxColumnWidth: columnWidthMetaData.maxColumnWidth,
        wrapTextMaxLines: columnWidthMetaData.wrapTextMaxLines,
    };
    for (let colIndex = 0; colIndex < columns.length; colIndex += 1) {
        const col = columns[colIndex];
        const { fixedWidth, initialWidth } = col;
        if (fixedWidth && isFinite(fixedWidth)) {
            totalWidthsData.totalFixedWidth += fixedWidth;
            totalWidthsData.totalFixedColumns += 1;
        }
        else if (col.isResized) {
            totalWidthsData.totalResizedWidth += col.columnWidth;
            totalWidthsData.totalResizedColumns += 1;
        }
        else if (initialWidth && isFinite(initialWidth)) {
            totalWidthsData.totalResizedWidth += initialWidth;
            totalWidthsData.totalResizedColumns += 1;
        }
        else {
            totalWidthsData.totalFlexibleColumns += 1;
        }
    }
    return totalWidthsData;
}
export function getColumnWidth(column) {
    const { fixedWidth, initialWidth } = column;
    if (fixedWidth && isFinite(fixedWidth)) {
        return fixedWidth;
    }
    const normalizedInitialWidth = isFinite(initialWidth) ? initialWidth : null;
    return column.isResized
        ? column.columnWidth || normalizedInitialWidth
        : normalizedInitialWidth;
}
