// lightning-base-components/src/lightning/datatable/wrapText.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import labelClipText from '@salesforce/label/LightningDatatable.clipText';
import labelWrapText from '@salesforce/label/LightningDatatable.wrapText';
import { normalizeBoolean } from 'lightning/utilsPrivate';
import { getStateColumnIndex } from './columns';
import { updateCell } from './rows';
import { normalizeNumberAttribute } from './utils';
const NON_WRAPPABLE_TYPES = new Set([
    'action',
    'boolean',
    'button',
    'button-icon',
    'date-local',
    'rowNumber',
]);
const i18n = {
    clipText: labelClipText,
    wrapText: labelWrapText,
};
export function setWrapTextState(col) {
    if (!NON_WRAPPABLE_TYPES.has(col.type)) {
        col.wrapText = normalizeBoolean(col.wrapText);
    }
    else {
        col.wrapText = false;
    }
}
export function setWrapTextMaxLines(state, value) {
    state.wrapTextMaxLines = normalizeNumberAttribute('wrapTextMaxLines', value, 'positive', undefined);
    state.shouldResetHeights = true;
}
function updateWrapTextAndMaxLinesValuesInCells(state, colIndex) {
    const { data, rows } = state;
    for (let i = 0; i < rows.length; i += 1) {
        updateCell(state, i, colIndex, data[i]);
    }
}
export function getInternalActions(col) {
    setWrapTextState(col);
    if (col.hideDefaultActions || NON_WRAPPABLE_TYPES.has(col.type)) {
        return [];
    }
    const isTextWrapped = col.wrapText || false;
    const { clipText, wrapText } = i18n;
    return [
        {
            label: wrapText,
            title: wrapText,
            checked: isTextWrapped,
            name: 'wrapText',
        },
        {
            label: clipText,
            title: clipText,
            checked: !isTextWrapped,
            name: 'clipText',
        },
    ];
}
export function handleTriggeredAction(dt, action, colKeyValue) {
    const { name } = action;
    const isWrapText = name === 'wrapText';
    if (isWrapText || name === 'clipText') {
        const { state } = dt;
        const colIndex = getStateColumnIndex(state, colKeyValue);
        const col = state.columns[colIndex];
        if (col.wrapText !== isWrapText) {
            state.shouldResetHeights = true;
            col.wrapText = isWrapText;
            updateSelectedOptionInHeaderActions(state, colKeyValue);
        }
    }
    dt._recalculateHeaderHeight = true;
}
function updateSelectedOptionInHeaderActions(state, colKeyValue) {
    const colIndex = getStateColumnIndex(state, colKeyValue);
    if (colIndex === -1) {
        return;
    }
    const { columns } = state;
    const col = columns[colIndex];
    const { internalActions } = col.actions;
    for (let i = 0; i < internalActions.length; i += 1) {
        const action = internalActions[i];
        const { name } = action;
        if (name === 'wrapText') {
            action.checked = col.wrapText;
        }
        else if (name === 'clipText') {
            action.checked = !col.wrapText;
        }
    }
    updateWrapTextAndMaxLinesValuesInCells(state, colIndex);
    col.actions = Object.assign({}, col.actions);
}
