// lightning-base-components/src/lightning/carousel/carousel.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import labelAutoPlay from '@salesforce/label/LightningCarousel.autoPlay';
import labelTabString from '@salesforce/label/LightningCarousel.tabString';
import { api, track } from 'lwc';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import { keyCodes, normalizeBoolean, isRTL } from 'lightning/utilsPrivate';
import { privateContext } from 'lightning/utilsInternal';
import { features } from 'lightning/features';
const INDICATOR_ACTION = 'slds-carousel__indicator-action';
const SLDS_ACTIVE = 'slds-is-active';
const PAUSE_ICON = 'utility:pause';
const PLAY_ICON = 'utility:right';
const FALSE_STRING = 'false';
const TRUE_STRING = 'true';
const SWIPE_DISTANCE_THRESHOLD = 20;
const DIRECTION_LEFT = 'left';
const DIRECTION_RIGHT = 'right';
const i18n = {
    autoPlay: labelAutoPlay,
    tabString: labelTabString,
};
const { setContext, assertContext } = privateContext;
const objectToElementMap = new WeakMap();
export default class LightningCarousel extends LightningShadowBaseClass {
    _disableAutoScroll;
    _disableAutoRefresh;
    _handlePrivateImageRegister;
    hasSyncChronizedA11y;
    indicatorActionsElements;
    get i18n() {
        return i18n;
    }
    @api
    get disableAutoScroll() {
        return this._disableAutoScroll || false;
    }
    set disableAutoScroll(value) {
        this._disableAutoScroll = normalizeBoolean(value);
    }
    @api
    get disableAutoRefresh() {
        return this._disableAutoRefresh || false;
    }
    set disableAutoRefresh(value) {
        this._disableAutoRefresh = normalizeBoolean(value);
    }
    @api
    scrollDuration = 5;
    @track
    paginationItems = [];
    @track
    autoScrollIcon = PAUSE_ICON;
    @track
    ariaPressed = FALSE_STRING;
    @track
    carouselPanelsStyle;
    togglePlayString = i18n.autoPlay;
    initialRender = true;
    activeIndexItem = 0;
    carouselItems = [];
    @track
    autoScrollTimeOut;
    swipeXStart = 0;
    handlePrivateImageRegister(event) {
        event.stopPropagation();
        if (this.paginationItems.length === 6) {
            return;
        }
        const arrayOfImagesInDOM = Array.from(this.querySelectorAll('lightning-carousel-image'));
        const indexInDOM = arrayOfImagesInDOM.findIndex((element) => element === event.target);
        this.unselectCurrentItem();
        const { detail, target } = event;
        const { contentId, guid } = detail;
        const { description } = target;
        const paginationItemDetail = {
            key: guid,
            id: `pagination-item-${guid}`,
            tabTitle: description ? description + ' ' + i18n.tabString : null,
            className: INDICATOR_ACTION,
            tabIndex: '-1',
            contentId,
            ariaSelected: FALSE_STRING,
        };
        objectToElementMap.set(detail, target);
        this.paginationItems.splice(indexInDOM, 0, paginationItemDetail);
        this.carouselItems.splice(indexInDOM, 0, detail);
        this.selectNewItem({
            itemIndex: 0,
            transitionEffectApplied: false,
        });
    }
    handlePrivateImageRegisterDEPRECATED(event) {
        const target = event.target, item = event.detail, currentIndex = this.carouselItems.length, isItemActive = currentIndex === this.activeIndexItem, paginationItemDetail = {
            key: item.guid,
            id: `pagination-item-${currentIndex}`,
            tabTitle: target.description
                ? target.description + ' ' + i18n.tabString
                : null,
            className: isItemActive
                ? INDICATOR_ACTION + ' ' + SLDS_ACTIVE
                : INDICATOR_ACTION,
            tabIndex: isItemActive ? '0' : '-1',
            contentId: event.detail.contentId,
            ariaSelected: isItemActive ? TRUE_STRING : FALSE_STRING,
        };
        event.stopPropagation();
        if (currentIndex > 5) {
            return;
        }
        if (isItemActive) {
            item.callbacks.select();
        }
        this.paginationItems.push(paginationItemDetail);
        this.carouselItems.push(item);
    }
    handleSlotChange() {
        this.handleItemRemovals();
    }
    handleItemRemovals() {
        const elementsInDOM = Array.from(this.querySelectorAll('lightning-carousel-image'));
        this.paginationItems = this.paginationItems.filter((_item, index) => {
            const element = objectToElementMap.get(this.carouselItems[index]);
            return elementsInDOM.includes(element);
        });
        this.carouselItems = this.carouselItems.filter((item) => {
            const element = objectToElementMap.get(item);
            return elementsInDOM.includes(element);
        });
        this.selectNewItem({
            itemIndex: 0,
            transitionEffectApplied: false,
        });
    }
    constructor() {
        super();
        setContext(this);
    }
    connectedCallback() {
        super.connectedCallback();
        assertContext(this);
        this.setAttribute('data-handles-touch', true);
        this._handlePrivateImageRegister = this.handlePrivateImageRegister.bind(this);
        this.addEventListener('privateimageregister', this._handlePrivateImageRegister);
    }
    renderedCallback() {
        if (this.initialRender) {
            if (!this.disableAutoScroll) {
                this.setAutoScroll();
            }
        }
        this.synchronizeA11y();
        this.initialRender = false;
    }
    disconnectedCallback() {
        this.cancelAutoScrollTimeOut();
        this.removeEventListener('privateimageregister', this._handlePrivateImageRegister);
    }
    synchronizeA11y() {
        if (this.paginationItems.length === 0 || this.carouselItems.length === 0) {
            return;
        }
        if (!this.hasSyncChronizedA11y) {
            const paginationElements = this.template.querySelectorAll('.slds-carousel__indicators a');
            paginationElements.forEach((element, index) => {
                element.setAttribute('aria-controls', this.paginationItems[index].contentId);
                this.carouselItems[index].callbacks.setLabelledBy(element.id);
                this.hasSyncChronizedA11y = true;
            });
        }
    }
    setAutoScroll() {
        assertContext(this);
        const scrollDuration = parseInt(this.scrollDuration, 10) * 1000;
        const carouselItemsLength = this.carouselItems.length;
        if (this.activeIndexItem === carouselItemsLength - 1 &&
            this.disableAutoRefresh) {
            this.autoScrollOff();
            return;
        }
        this.cancelAutoScrollTimeOut();
        this.autoScrollTimeOut = setTimeout(this.startAutoScroll.bind(this), scrollDuration);
    }
    startAutoScroll() {
        this.selectNextSibling();
        this.setAutoScroll();
    }
    cancelAutoScrollTimeOut() {
        clearTimeout(this.autoScrollTimeOut);
    }
    toggleAutoScroll() {
        const ariaPressed = this.ariaPressed;
        if (ariaPressed === FALSE_STRING) {
            this.autoScrollOff();
        }
        else {
            this.autoScrollOn();
        }
    }
    autoScrollOn() {
        const carouselItemsLength = this.carouselItems.length;
        if (!this.disableAutoScroll) {
            if (this.activeIndexItem === carouselItemsLength - 1 &&
                this.disableAutoRefresh) {
                this.unselectCurrentItem();
                this.selectNewItem({
                    itemIndex: 0,
                    transitionEffectApplied: true,
                });
            }
            this.autoScrollIcon = PAUSE_ICON;
            this.ariaPressed = FALSE_STRING;
            this.setAutoScroll();
        }
    }
    autoScrollOff() {
        if (!this.disableAutoScroll) {
            this.ariaPressed = TRUE_STRING;
            this.autoScrollIcon = PLAY_ICON;
            this.cancelAutoScrollTimeOut();
        }
    }
    onItemSelect(event) {
        const currentTarget = event.currentTarget, itemIndex = currentTarget.getAttribute('data-index');
        this.autoScrollOff();
        if (this.activeIndexItem !== itemIndex) {
            this.unselectCurrentItem();
            this.selectNewItem({
                itemIndex: itemIndex,
                transitionEffectApplied: true,
            });
            this.activeIndexItem = parseInt(itemIndex, 10);
        }
    }
    onItemLinkSelect(event) {
        event.preventDefault();
    }
    unselectCurrentItem() {
        if (this.paginationItems.length === 0 || this.carouselItems.length === 0) {
            return;
        }
        const activePaginationItem = this.paginationItems[this.activeIndexItem];
        activePaginationItem.tabIndex = '-1';
        activePaginationItem.ariaSelected = FALSE_STRING;
        activePaginationItem.className = INDICATOR_ACTION;
        this.carouselItems[this.activeIndexItem].callbacks.unselect();
    }
    selectNewItem({ itemIndex, transitionEffectApplied, }) {
        if (this.paginationItems.length === 0 ||
            this.carouselItems.length === 0) {
            return;
        }
        const activePaginationItem = this.paginationItems[itemIndex];
        if (!this.carouselItems[itemIndex].callbacks.isSelected()) {
            activePaginationItem.tabIndex = '0';
            activePaginationItem.ariaSelected = TRUE_STRING;
            activePaginationItem.className =
                INDICATOR_ACTION + ' ' + SLDS_ACTIVE;
            if (transitionEffectApplied) {
                const rtl = isRTL();
                const translateValue = features.enableCarouselRTLFix && rtl
                    ? itemIndex * 100
                    : -(itemIndex * 100);
                this.carouselPanelsStyle = `transform:translateX(${translateValue}%);`;
            }
            else {
                this.carouselPanelsStyle = `transform:translateX(0); transition-duration:0s;`;
            }
            this.carouselItems[itemIndex].callbacks.select();
            this.activeIndexItem = itemIndex;
        }
    }
    keyDownHandler(event) {
        const key = event.keyCode;
        let indicatorActionsElements = this.indicatorActionsElements;
        const rtl = features.enableCarouselRTLFix && isRTL();
        if (key === keyCodes.right || key === keyCodes.left) {
            event.preventDefault();
            event.stopPropagation();
            this.autoScrollOff();
            if ((!rtl && key === keyCodes.right) ||
                (rtl && key === keyCodes.left)) {
                this.selectNextSibling();
            }
            else {
                this.selectPreviousSibling();
            }
        }
        if (!indicatorActionsElements) {
            indicatorActionsElements = this.template.querySelectorAll('.slds-carousel__indicator-action');
            this.indicatorActionsElements = indicatorActionsElements;
        }
        indicatorActionsElements[this.activeIndexItem].focus();
    }
    selectNextSibling() {
        const carouselItemsLength = this.carouselItems.length;
        let itemIndex = this.activeIndexItem + 1;
        if (this.activeIndexItem === carouselItemsLength - 1) {
            if (this.disableAutoRefresh) {
                this.autoScrollOff();
                return;
            }
            itemIndex = 0;
        }
        this.unselectCurrentItem();
        this.selectNewItem({
            itemIndex,
            transitionEffectApplied: true,
        });
    }
    selectPreviousSibling() {
        const carouselItemsLength = this.carouselItems.length;
        let itemIndex = this.activeIndexItem - 1;
        if (this.activeIndexItem === 0) {
            if (this.disableAutoRefresh) {
                this.autoScrollOff();
                return;
            }
            itemIndex = carouselItemsLength - 1;
        }
        this.unselectCurrentItem();
        this.selectNewItem({
            itemIndex,
            transitionEffectApplied: true,
        });
    }
    handleTouchStart({ changedTouches }) {
        this.swipeXStart = (changedTouches && changedTouches[0].clientX) || 0;
    }
    handleTouchEnd(event) {
        const { changedTouches } = event;
        const swipeXEnd = (changedTouches && changedTouches[0].clientX) || 0;
        const dx = swipeXEnd - this.swipeXStart;
        if (Math.abs(dx) > SWIPE_DISTANCE_THRESHOLD) {
            const rtl = features.enableCarouselRTLFix && isRTL();
            const direction = Math.sign(dx) === 1 ? DIRECTION_LEFT : DIRECTION_RIGHT;
            if ((!rtl && direction === DIRECTION_RIGHT) ||
                (rtl && direction === DIRECTION_LEFT)) {
                this.selectNextSibling();
            }
            else {
                this.selectPreviousSibling();
            }
            event.preventDefault();
        }
        this.swipeXStart = 0;
    }
}
