// lightning-base-components/src/lightning/card/card.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { api, track } from 'lwc';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import { classSet } from 'lightning/utils';
import { isNarrow, isBase } from './utils';
import { isHeadingLevelValid } from 'lightning/utilsPrivate';
import orgHasA11yResizeReflowPhase2 from '@salesforce/accessCheck/A11yResizeReflow.orgHasA11yResizeReflowPhase2';
import resizeReflowStyles from './card.lbc.resizereflow.css';
import emptyStylesheet from './card.lbc.empty.css';
export default class LightningCard extends LightningShadowBaseClass {
    static stylesheets = [
        (orgHasA11yResizeReflowPhase2 && resizeReflowStyles) ||
            emptyStylesheet,
    ];
    @api
    title;
    @api
    iconName;
    _privateVariant = 'base';
    set variant(value) {
        if (isNarrow(value) || isBase(value)) {
            this._privateVariant = value;
        }
        else {
            this._privateVariant = 'base';
        }
    }
    @api
    get variant() {
        return this._privateVariant;
    }
    _hasTitle = true;
    _showFooter = true;
    renderedCallback() {
        if (this.footerSlot) {
            this._showFooter = this.footerSlot.assignedElements().length !== 0;
        }
        if (this.titleSlot) {
            const hasSlotTitle = this.titleSlot.assignedElements().length !== 0;
            this._hasTitle = hasSlotTitle || this.hasStringTitle;
        }
    }
    @track
    _privateHeadingAriaLevel;
    @api
    get headingLevel() {
        return this._privateHeadingAriaLevel;
    }
    set headingLevel(value) {
        if (isHeadingLevelValid(value)) {
            this._privateHeadingAriaLevel = value;
        }
    }
    get titleSlot() {
        return this.template.querySelector('slot[name=title]');
    }
    get footerSlot() {
        return this.template.querySelector('slot[name=footer]');
    }
    get computedWrapperClassNames() {
        return classSet('slds-card').add({
            'slds-card_narrow': isNarrow(this._privateVariant),
        });
    }
    get hasIcon() {
        return !!this.iconName;
    }
    get hasStringTitle() {
        return !!this.title;
    }
    get isResizeReflowEnabled() {
        return orgHasA11yResizeReflowPhase2 ? '' : null;
    }
    @track
    privateHeaderLabel;
    @api
    get label() {
        if (!this._hasTitle || this.hideHeader) {
            return this.privateHeaderLabel;
        }
        return null;
    }
    set label(value) {
        this.privateHeaderLabel = value;
    }
    get computedHidden() {
        if (!this.label && this.hideHeader) {
            console.warn('A `lightning-card` with `hide-header` requires `label` to be set.');
        }
        return !this.hideHeader;
    }
    hiddenHeader = false;
    @api
    get hideHeader() {
        return this.hiddenHeader;
    }
    set hideHeader(value) {
        this.hiddenHeader = value;
    }
}
