// lightning-base-components/src/lightning/ariaObserver/ariaObserver.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { guid, isNativeComponent } from 'lightning/utilsPrivate';
import { features } from 'lightning/features';
import { setAriaActiveDescendant, setAriaDescribedBy, setAriaLabelledBy, setAriaControls, setAriaDetails, setAriaErrorMessage, } from './polyfill.js';
const SUPPORTED_ATTRIBUTES_ARRAY = [
    [
        'aria-controls',
        {
            ariaReflection: 'ariaControlsElements',
            polyfill: setAriaControls,
        },
    ],
    [
        'aria-labelledby',
        {
            ariaReflection: 'ariaLabelledByElements',
            polyfill: setAriaLabelledBy,
        },
    ],
    [
        'aria-describedby',
        {
            ariaReflection: 'ariaDescribedByElements',
            polyfill: setAriaDescribedBy,
        },
    ],
    [
        'aria-errormessage',
        {
            ariaReflection: 'ariaErrorMessageElements',
            polyfill: setAriaErrorMessage,
        },
    ],
    [
        'aria-activedescendant',
        {
            ariaReflection: 'ariaActiveDescendantElement',
            polyfill: setAriaActiveDescendant,
        },
    ],
    [
        'aria-details',
        {
            ariaReflection: 'ariaDetailsElements',
            polyfill: setAriaDetails,
        },
    ],
];
const SUPPORTED_ATTRIBUTES_MAP = new Map(SUPPORTED_ATTRIBUTES_ARRAY);
let BROWSER_SUPPORTS_ARIA_ELEMENT_REFLECTION;
if (!import.meta.env.SSR) {
    BROWSER_SUPPORTS_ARIA_ELEMENT_REFLECTION = [
        ...SUPPORTED_ATTRIBUTES_MAP.values(),
    ].every(({ ariaReflection }) => ariaReflection in Element.prototype);
}
function extractElements(root, ids) {
    if (typeof ids !== 'string' || ids === '') {
        return [];
    }
    return splitIds(ids)
        .map((id) => root.querySelector(`[id*="${id}"]`))
        .filter((el) => !!el);
}
function splitIds(ids) {
    return ids ? (ids + '').trim().split(/\s+/) : [];
}
export default class AriaObserver {
    component;
    template;
    isNativeShadow;
    state;
    liveIds;
    guid;
    placeholderContainer;
    ariaReflectionMap;
    nodeMap;
    _root;
    mo;
    constructor(component) {
        this.component = component;
        this.template = component.template;
        this.isNativeShadow = isNativeComponent(component);
        this.state = {};
        this.liveIds = {};
        this.guid = guid();
        this.placeholderContainer = null;
        this.ariaReflectionMap = new Map();
        this.nodeMap = new Map();
    }
    connectLiveIdRef(refs, callback) {
        const selector = (refs + '')
            .trim()
            .split(/\s+/)
            .map((ref) => `[id*="${ref}"]`)
            .join(',');
        let liveId;
        if (this.liveIds[refs]) {
            this.liveIds[refs].callbacks.push(callback);
        }
        else {
            liveId = { refs, selector, callbacks: [callback] };
            this.liveIds[refs] = liveId;
        }
    }
    connect({ attribute, targetSelector, targetNode, relatedNodeIds, relatedNodes, }) {
        this.state[attribute] = this.state[attribute] || {};
        const attrState = this.state[attribute];
        attrState.targetNode = targetNode;
        attrState.targetSelector = targetSelector;
        attrState.relatedNodes = (!Array.isArray(relatedNodes) ? [relatedNodes] : relatedNodes).filter(Boolean);
        attrState.relatedNodeIds = Array.isArray(relatedNodeIds)
            ? relatedNodeIds.join(' ')
            : relatedNodeIds;
        if (this.component.isConnected) {
            this.privateUpdate(attribute);
        }
    }
    sync(isNativeShadow) {
        if (isNativeShadow != null) {
            this.isNativeShadow = isNativeShadow;
        }
        if (!this.component.isConnected) {
            throw new Error(`Invalid sync invocation. It can only be invoked during renderedCallback().`);
        }
        if (!this.root) {
            this.root =
                this.template && this.template.host
                    ? this.template.host.getRootNode()
                    : null;
        }
        this.privateUpdateLiveIds();
        for (const attrName in this.state) {
            if (Object.prototype.hasOwnProperty.call(this.state, attrName)) {
                this.privateUpdate(attrName);
            }
        }
    }
    get privateIsMoRequired() {
        return this.isNativeShadow || Object.keys(this.liveIds).length !== 0;
    }
    get root() {
        return this._root;
    }
    set root(root) {
        this._root = root;
        if (this._root && this.privateIsMoRequired) {
            this.privateCreateMutationObserver();
        }
    }
    privateUpdate(attrName) {
        const { targetSelector, targetNode = this.template.querySelector(targetSelector), relatedNodeIds, relatedNodes, } = this.state[attrName];
        if (!targetNode) {
            return;
        }
        const attribute = SUPPORTED_ATTRIBUTES_MAP.get(attrName);
        if (!attribute) {
            throw new Error(`${attrName} is not supported by AriaObserver. Supported attributes: ${Array.from(SUPPORTED_ATTRIBUTES_MAP.keys())}`);
        }
        if (this.isNativeShadow) {
            const allRelatedNodes = [
                ...relatedNodes,
                ...extractElements(this.root, relatedNodeIds),
            ];
            if (BROWSER_SUPPORTS_ARIA_ELEMENT_REFLECTION) {
                if (features.enableMobileGroupedCbAriaUpdates &&
                    attrName === 'aria-activedescendant') {
                    const elem = relatedNodes?.length > 0 ? relatedNodes[0] : null;
                    targetNode[attribute.ariaReflection] = elem;
                }
                else {
                    targetNode[attribute.ariaReflection] = relatedNodes;
                }
            }
            else {
                attribute.polyfill(targetNode, allRelatedNodes, attrName);
            }
            const rootNode = targetNode.getRootNode();
            for (let i = 0; i < allRelatedNodes.length; i++) {
                this.privateAddNestedReferences(rootNode, allRelatedNodes[i]);
            }
        }
        else {
            const attributeValue = [
                ...splitIds(relatedNodeIds),
                ...relatedNodes.map((n) => n.id),
            ].join(' ');
            if (attributeValue) {
                targetNode.setAttribute(attrName, attributeValue);
            }
            else {
                targetNode.removeAttribute(attrName);
            }
        }
    }
    privateExtractCorrectElements(selector = '', elements) {
        const selectors = selector.split(/\s/g);
        const matchSelectors = `(${selectors.join('|')})`;
        const regex = new RegExp(`^${matchSelectors}(-[0-9]+)$`);
        return [...elements].filter((element) => {
            return regex.test(element.id);
        });
    }
    privateCreateMutationObserver() {
        this.disconnect();
        this.mo = new MutationObserver(() => {
            if (!this.component.isConnected) {
                return;
            }
            this.sync();
        });
        this.mo.observe(this.root, {
            characterData: true,
            childList: true,
            subtree: true,
        });
    }
    privateExtractIds(elements) {
        return elements
            .map((el) => {
            return el.getAttribute('id');
        })
            .join(' ');
    }
    privateUpdateLiveIds() {
        const root = this.template && this.template.host
            ? this.template.host.getRootNode()
            : null;
        if (!root) {
            return;
        }
        for (const liveId in this.liveIds) {
            if (Object.prototype.hasOwnProperty.call(this.liveIds, liveId)) {
                const thisId = this.liveIds[liveId];
                if (!thisId.elements || !thisId.elements.length) {
                    const splitRefIds = splitIds(liveId);
                    try {
                        const refElements = Array.from(root.querySelectorAll(thisId.selector));
                        thisId.elements = refElements.sort((a, b) => {
                            const idA = a
                                .getAttribute('id')
                                ?.replace(/-[0-9]+$/g, '');
                            const idB = b
                                .getAttribute('id')
                                ?.replace(/-[0-9]+$/g, '');
                            return (splitRefIds.indexOf(idA) -
                                splitRefIds.indexOf(idB));
                        });
                    }
                    catch (error) {
                        if (error instanceof DOMException) {
                            console.warn(`lightning/ariaObserver could not find elements to connect live IDs due to an invalid id selector.\n\n${error.message}`);
                        }
                        return;
                    }
                }
                const newThisId = this.privateExtractCorrectElements(thisId.refs, thisId.elements);
                const newIds = this.privateExtractIds(newThisId);
                if (newIds.length && newIds !== thisId.ids) {
                    for (let i = 0; i < thisId.callbacks.length; i++) {
                        thisId.callbacks[i](newIds);
                    }
                    thisId.ids = newIds;
                }
            }
        }
    }
    privateAddNestedReferences(rootNode, relatedNode) {
        for (let i = 0; i < SUPPORTED_ATTRIBUTES_ARRAY.length; i++) {
            const [attrName, { polyfill }] = SUPPORTED_ATTRIBUTES_ARRAY[i];
            const attrValue = relatedNode.getAttribute(attrName);
            if (attrValue) {
                let mirroredNode = this.nodeMap.get(relatedNode);
                if (!mirroredNode) {
                    let ariaReflection = this.ariaReflectionMap.get(rootNode);
                    if (!ariaReflection) {
                        ariaReflection = rootNode.querySelector('.aria-element-reflection-mirror');
                        this.ariaReflectionMap.set(rootNode, ariaReflection);
                    }
                    if (ariaReflection) {
                        mirroredNode = ariaReflection.querySelector(`[${attrName}='${attrValue}']`);
                        this.nodeMap.set(relatedNode, mirroredNode);
                    }
                }
                if (mirroredNode) {
                    const nestedReferenceNodes = extractElements(this.root, attrValue);
                    polyfill(mirroredNode, nestedReferenceNodes, attrName);
                }
            }
        }
    }
    disconnect() {
        if (this.mo) {
            this.mo.disconnect();
            this.mo = undefined;
        }
    }
}
