// lightning-base-components/src/lightning/lookupAddress/location.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { features } from 'lightning/features';
export const DEFAULT_LOCATION = { lat: 37.790091, lng: -122.396848 };
const GEOLOCATION_TIMEOUT = 10000;
const getCurrentPosition = !import.meta.env.SSR && navigator?.geolocation?.getCurrentPosition
    ? navigator.geolocation.getCurrentPosition.bind(navigator.geolocation)
    : (success, error) => error?.();
function attemptGeolocation(enableHighAccuracy) {
    return new Promise((resolve, reject) => {
        getCurrentPosition((position) => {
            resolve({
                lat: position.coords.latitude,
                lng: position.coords.longitude,
            });
        }, reject, { enableHighAccuracy, timeout: GEOLOCATION_TIMEOUT });
    });
}
function getLocationWithTimeout() {
    return Promise.any([
        attemptGeolocation(false),
        attemptGeolocation(true),
    ]).catch((error) => {
        console.warn(`Geolocation error (${error?.code}): ${error?.message || 'Unable to determine location'}. Defaulting to lat: ${DEFAULT_LOCATION.lat}, lng: ${DEFAULT_LOCATION.lng}`);
        return DEFAULT_LOCATION;
    });
}
function getLocationLegacy() {
    return new Promise((resolve) => {
        getCurrentPosition((position) => {
            resolve({
                lat: position.coords.latitude,
                lng: position.coords.longitude,
            });
        }, () => resolve(DEFAULT_LOCATION));
    });
}
export function getLocation() {
    if (features.enableGeolocationTimeout) {
        return getLocationWithTimeout();
    }
    return getLocationLegacy();
}
