// lightning-base-components/src/lightning/treeGrid/treeGrid.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { api, track } from 'lwc';
import LightningDatatable from 'lightning/datatable';
import { normalizeColumns, normalizeData } from './normalizer';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import { arraysEqual } from 'lightning/utilsPrivate';
import { privateContext } from 'lightning/utilsInternal';
const ROWS_ACTION = {
    SELECT_ALL_ROWS: 'selectAllRows',
    DESELECT_ALL_ROWS: 'deselectAllRows',
    ROW_SELECT: 'rowSelect',
    ROW_DESELECT: 'rowDeselect',
};
const { setContext, assertContext } = privateContext;
export default class LightningTreeGrid extends LightningShadowBaseClass {
    @track
    _rawColumns;
    @track
    _rawData;
    @track
    _keyField;
    _publicExpandedRows = [];
    @track
    _ariaLabel;
    @track
    _ogDatatableCustomTypes;
    @api
    columnWidthsMode;
    @api
    defaultSortDirection = 'asc';
    @api
    hideCheckboxColumn = false;
    @api
    hideTableHeader = false;
    @api
    isLoading = false;
    _selectedRowKeys = [];
    @api
    hideBorders = false;
    @api
    get keyField() {
        return this._keyField;
    }
    set keyField(value) {
        this._keyField = value;
        this.flattenData();
    }
    @api
    maxColumnWidth = 1000;
    @api
    minColumnWidth = 50;
    @api
    resizeColumnDisabled = false;
    @api
    rowNumberOffset = 0;
    @api
    disabledRows = [];
    @api
    showRowNumberColumn = false;
    @api
    sortedBy;
    @api
    sortedDirection;
    @api
    rowToggleIcon;
    @track
    _columns;
    @track
    _data;
    @track
    _expandedRows = [];
    @track
    _selectedRows = [];
    constructor() {
        super();
        setContext(this);
        this.template.addEventListener('privatetogglecell', this.handleToggle);
        this.template.addEventListener('toggleallheader', this.handleToggleAll);
    }
    connectedCallback() {
        super.connectedCallback();
        this._selectedRowKeys = this.selectedRows.slice();
        if (this.constructor.customTypes) {
            this._ogDatatableCustomTypes = LightningDatatable.customTypes;
            LightningDatatable.customTypes = this.constructor.customTypes;
        }
    }
    renderedCallback() {
        if (this.constructor.customTypes) {
            LightningDatatable.customTypes = this._ogDatatableCustomTypes;
        }
    }
    @api
    get columns() {
        return this._rawColumns;
    }
    set columns(value) {
        this._rawColumns = value;
        this._columns = normalizeColumns(value);
    }
    @api
    get data() {
        return this._rawData;
    }
    set data(value) {
        this._rawData = value;
        this.flattenData();
    }
    @api
    get expandedRows() {
        if (!arraysEqual(this._expandedRows, this._publicExpandedRows)) {
            this._publicExpandedRows = Object.assign([], this._expandedRows);
        }
        return this._publicExpandedRows;
    }
    set expandedRows(value) {
        this._publicExpandedRows = Object.assign([], value);
        this._expandedRows = Object.assign([], value);
        this.flattenData();
        this._selectedRows = this._selectedRowKeys.slice();
    }
    get normalizedColumns() {
        return this._columns;
    }
    get normalizedData() {
        return this._data;
    }
    getSelectedRowKeys(rowKeys) {
        assertContext(this);
        const selRowKeys = [];
        rowKeys.forEach((row) => {
            selRowKeys.push(row[this.keyField]);
        });
        return selRowKeys;
    }
    @api
    getSelectedRows() {
        return this.template.querySelector('lightning-datatable').getSelectedRows();
    }
    @api
    getCurrentExpandedRows() {
        return this.expandedRows;
    }
    @api
    expandAll() {
        assertContext(this);
        this.toggleAllRows(this.data, true);
        this._selectedRows = this._selectedRowKeys.slice();
    }
    @api
    collapseAll() {
        assertContext(this);
        this.toggleAllRows(this.data, false);
        this._selectedRows = this._selectedRowKeys.slice();
    }
    @api
    get ariaLabel() {
        return this._ariaLabel;
    }
    set ariaLabel(value) {
        this._ariaLabel = value;
    }
    handleRowDeSelection(event) {
        event.stopPropagation();
        const { rowKeyValue } = event.detail;
        const index = this._selectedRowKeys.indexOf(rowKeyValue);
        this._selectedRowKeys.splice(index, 1);
    }
    handleToggle = (event) => {
        assertContext(this);
        event.stopPropagation();
        const { name, nextState } = event.detail;
        this.toggleRow(this.data, name, nextState);
        this._selectedRows = this._selectedRowKeys.slice();
    };
    @api
    get selectedRows() {
        return this._selectedRows;
    }
    set selectedRows(value) {
        this._selectedRows = value;
        if (Array.isArray(value)) {
            this._selectedRowKeys = value.slice();
        }
    }
    handleToggleAll = (event) => {
        assertContext(this);
        event.stopPropagation();
        const { nextState } = event.detail;
        this.toggleAllRows(this.data, nextState);
        this._selectedRows = this._selectedRowKeys.slice();
    };
    handleRowSelection(event) {
        event.stopPropagation();
        if (event.detail.config) {
            switch (event.detail.config.action) {
                case ROWS_ACTION.ROW_SELECT:
                    this._selectedRowKeys.push(event.detail.config.value);
                    break;
                case ROWS_ACTION.ROW_DESELECT: {
                    const index = this._selectedRowKeys.indexOf(event.detail.config.value);
                    this._selectedRowKeys.splice(index, 1);
                    break;
                }
                case ROWS_ACTION.SELECT_ALL_ROWS:
                    this._selectedRowKeys = this.getSelectedRowKeys(event.detail.selectedRows);
                    break;
                case ROWS_ACTION.DESELECT_ALL_ROWS:
                    this._selectedRowKeys = this.getSelectedRowKeys(event.detail.selectedRows);
                    break;
                default:
                    break;
            }
            event.detail.config.selectedRowKeys = this._selectedRowKeys.slice();
        }
        this.fireSelectedRowsChange(event.detail);
    }
    handleHeaderAction(event) {
        event.stopPropagation();
        this.fireHeaderAction(event.detail);
    }
    handleRowAction(event) {
        event.stopPropagation();
        this.fireRowAction(event.detail);
    }
    handleSort(event) {
        event.stopPropagation();
        this.dispatchEvent(new CustomEvent('sort', {
            detail: event.detail,
        }));
    }
    fireRowToggleChange(name, isExpanded, hasChildrenContent, row) {
        const event = new CustomEvent('toggle', {
            detail: { name, isExpanded, hasChildrenContent, row },
        });
        this.dispatchEvent(event);
    }
    fireToggleAllChange(isExpanded) {
        const event = new CustomEvent('toggleall', {
            detail: { isExpanded },
        });
        this.dispatchEvent(event);
    }
    fireSelectedRowsChange(eventDetails) {
        const event = new CustomEvent('rowselection', {
            detail: eventDetails,
        });
        this.dispatchEvent(event);
    }
    fireHeaderAction(eventDetails) {
        const event = new CustomEvent('headeraction', {
            detail: eventDetails,
        });
        this.dispatchEvent(event);
    }
    fireRowAction(eventDetails) {
        const event = new CustomEvent('rowaction', {
            detail: eventDetails,
        });
        this.dispatchEvent(event);
    }
    flattenData() {
        if (this.keyField) {
            this._data = normalizeData(this.data, this.expandedRows, this.keyField);
        }
    }
    updateExpandedRows(name, isExpanded) {
        const itemPosition = this._expandedRows.indexOf(name);
        if (itemPosition > -1 && isExpanded === false) {
            this._expandedRows.splice(itemPosition, 1);
        }
        else if (itemPosition === -1 && isExpanded) {
            this._expandedRows.push(name);
        }
    }
    hasChildrenContent(row) {
        let hasChildrenContent = false;
        if (Object.prototype.hasOwnProperty.call(row, '_children') && Array.isArray(row._children) && row._children.length > 0) {
            hasChildrenContent = true;
        }
        return hasChildrenContent;
    }
    toggleRow(data, name, isExpanded) {
        data.forEach((row) => {
            const hasChildrenContent = this.hasChildrenContent(row);
            if (row[this.keyField] === name) {
                this.updateExpandedRows(name, isExpanded);
                this.fireRowToggleChange(name, isExpanded, hasChildrenContent, row);
            }
            else if (hasChildrenContent) {
                this.toggleRow(row._children, name, isExpanded);
            }
        });
        this.flattenData();
    }
    _toggleAllRecursionCounter = 1;
    toggleAllRows(data, isExpanded, rowsToToggle = []) {
        if (isExpanded) {
            data.forEach((row) => {
                const hasChildrenContent = this.hasChildrenContent(row);
                if (hasChildrenContent) {
                    rowsToToggle.push(row[this.keyField]);
                    this._toggleAllRecursionCounter++;
                    this.toggleAllRows(row._children, isExpanded, rowsToToggle);
                }
            });
        }
        if (--this._toggleAllRecursionCounter === 0) {
            this._toggleAllRecursionCounter = 1;
            this._expandedRows = rowsToToggle;
            this.fireToggleAllChange(isExpanded);
            this.flattenData();
        }
    }
}
