// lightning-base-components/src/lightning/verticalNavigation/verticalNavigation.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { api, track } from 'lwc';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import { classSet } from 'lightning/utils';
import { assert, normalizeBoolean, normalizeString, } from 'lightning/utilsPrivate';
import { privateContext } from 'lightning/utilsInternal';
import subPage from '@salesforce/label/LightningVerticalNavigation.subPage';
const ALLOWED_CHILDREN = [
    'LIGHTNING-VERTICAL-NAVIGATION-ITEM',
    'LIGHTNING-VERTICAL-NAVIGATION-ITEM-BADGE',
    'LIGHTNING-VERTICAL-NAVIGATION-ITEM-ICON',
];
const { setContext, assertContext } = privateContext;
export default class LightningVerticalNavigation extends LightningShadowBaseClass {
    @track
    _compact;
    @track
    _shaded;
    @track
    _selectedItem;
    @api
    get compact() {
        return this._compact || false;
    }
    set compact(compact) {
        this._compact = normalizeBoolean(compact);
    }
    @api
    get shaded() {
        return this._shaded || false;
    }
    set shaded(shaded) {
        this._shaded = normalizeBoolean(shaded);
    }
    @api
    get selectedItem() {
        return this._selectedItem || '';
    }
    set selectedItem(selectedItem) {
        this.selectNavigationItem(normalizeString(selectedItem, { toLowerCase: false }));
    }
    get ariaLabel() {
        return this.privateAriaLabel || subPage;
    }
    set ariaLabel(ariaLabel) {
        this.privateAriaLabel = ariaLabel;
    }
    get computedClass() {
        const classes = classSet('slds-nav-vertical');
        if (this.shaded) {
            classes.add('slds-nav-vertical_shade');
        }
        if (this.compact) {
            classes.add('slds-nav-vertical_compact');
        }
        return classes.toString();
    }
    constructor() {
        super();
        setContext(this);
    }
    connectedCallback() {
        super.connectedCallback();
        this.addItemEventListeners();
    }
    addItemEventListeners() {
        assertContext(this);
        if (!import.meta.env.SSR) {
            this.addEventListener('privateitemregister', this.handleItemRegister.bind(this));
            this.addEventListener('privateitemselect', this.handleItemSelect.bind(this));
        }
    }
    removeItemEventListeners() {
        assertContext(this);
        if (!import.meta.env.SSR) {
            this.removeEventListener('privateitemregister', this.handleItemRegister.bind(this));
            this.removeEventListener('privateitemselect', this.handleItemSelect.bind(this));
        }
    }
    verticalNavigationItems = [];
    handleItemRegister(event) {
        event.stopPropagation();
        const target = event.target, callbacks = event.detail.callbacks, itemName = event.detail.name, isItemSelected = this._selectedItem === itemName;
        assert(target.nodeType in ALLOWED_CHILDREN, 'Attempt was made to register unsupported type.');
        if (target.nodeType in ALLOWED_CHILDREN) {
            const navigationItem = {
                name: itemName,
                callbacks,
            };
            this.verticalNavigationItems.push(navigationItem);
        }
        if (isItemSelected) {
            callbacks.select();
        }
    }
    handleItemSelect(event) {
        event.stopPropagation();
        this.selectNavigationItem(event.detail.name);
    }
    selectNavigationItem(itemName) {
        if (!import.meta.env.SSR) {
            const beforeselectevent = new CustomEvent('beforeselect', {
                cancelable: true,
                detail: {
                    name: itemName,
                },
            });
            this.dispatchEvent(beforeselectevent);
            if (!beforeselectevent.defaultPrevented) {
                this.verticalNavigationItems.forEach((navigationItem) => {
                    if (navigationItem.name === itemName) {
                        navigationItem.callbacks.select();
                    }
                    else {
                        navigationItem.callbacks.deselect();
                    }
                });
                this._selectedItem = itemName;
                this.dispatchEvent(new CustomEvent('select', {
                    detail: {
                        name: itemName,
                    },
                }));
            }
        }
    }
    disconnectedCallback() {
        this.removeItemEventListeners();
    }
}
