// lightning-base-components/src/lightning/formattedUrl/formattedUrl.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { api, track } from 'lwc';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import { formatUrl } from 'lightning/utils';
import { updateRawLinkInfo } from 'lightning/routingService';
import { features } from 'lightning/features';
import { privateContext } from 'lightning/utilsInternal';
const { setContext, assertContext } = privateContext;
export default class LightningFormattedUrl extends LightningShadowBaseClass {
    @api
    target;
    @api
    tooltip;
    @api
    label;
    @api
    tabIndex;
    @track
    _url;
    @track
    _value;
    _connected = false;
    _dispatcher = () => { };
    @api
    get value() {
        return this._value;
    }
    set value(value) {
        this._value = value;
        if (this._connected) {
            this.updateLinkInfo(value);
        }
    }
    constructor() {
        super();
        setContext(this);
    }
    connectedCallback() {
        super.connectedCallback();
        this._connected = true;
        this.updateLinkInfo(this.value);
    }
    disconnectedCallback() {
        this._connected = false;
    }
    @api
    focus() {
        if (this.urlAnchor) {
            this.urlAnchor.focus();
        }
    }
    @api
    blur() {
        if (this.urlAnchor) {
            this.urlAnchor.blur();
        }
    }
    @api
    click() {
        const anchor = this.urlAnchor;
        if (anchor && anchor.click) {
            anchor.click();
        }
    }
    get urlAnchor() {
        if (this._connected && this.hasValue) {
            return this.template.querySelector('a');
        }
        return undefined;
    }
    handleClick(event) {
        if (features.enableAssertContext260Round2) {
            assertContext(this);
        }
        this._dispatcher(event);
    }
    updateLinkInfo(url) {
        updateRawLinkInfo(this, {
            url: formatUrl(url),
            target: this.target,
        }).then((linkInfo) => {
            this._url = linkInfo.url;
            this._dispatcher = linkInfo.dispatcher;
        });
    }
    get computedLabel() {
        const { label, computedUrl } = this;
        return label != null && label !== '' ? label : computedUrl;
    }
    get computedUrl() {
        return this._url || formatUrl(this.value);
    }
    get hasValue() {
        const url = this.value;
        return url != null && url !== '';
    }
}
