// lightning-base-components/src/lightning/formattedAddress/formattedAddress.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import locale from '@salesforce/i18n/locale';
import { api, track } from 'lwc';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import { normalizeBoolean, normalizeString } from 'lightning/utilsPrivate';
import { addressFormat } from 'lightning/internationalizationLibrary';
const MAP_HOST = 'https://www.google.com/maps?q=';
const VARIANT_TRUNCATE = 'truncate';
const VARIANT_PLAIN = 'plain';
export default class LightningFormattedAddress extends LightningShadowBaseClass {
    @track
    state = {
        street: '',
        city: '',
        province: '',
        country: '',
        postalCode: '',
        latitude: '',
        longitude: '',
        disabled: false,
    };
    @track
    href;
    @api
    locale = locale;
    _variant;
    @api
    get street() {
        return this.state.street;
    }
    set street(value) {
        this.state.street = value;
    }
    @api
    get city() {
        return this.state.city;
    }
    set city(value) {
        this.state.city = value;
    }
    @api
    get province() {
        return this.state.province;
    }
    set province(value) {
        this.state.province = value;
    }
    @api
    get country() {
        return this.state.country;
    }
    set country(value) {
        this.state.country = value;
    }
    @api
    get postalCode() {
        return this.state.postalCode;
    }
    set postalCode(value) {
        this.state.postalCode = value;
    }
    @api
    get latitude() {
        return this.state.latitude;
    }
    set latitude(value) {
        this.state.latitude = value;
    }
    @api
    get longitude() {
        return this.state.longitude;
    }
    set longitude(value) {
        this.state.longitude = value;
    }
    @api
    get disabled() {
        return this.state.disabled;
    }
    set disabled(value) {
        this.state.disabled = normalizeBoolean(value);
    }
    @api
    get variant() {
        return this._variant || VARIANT_TRUNCATE;
    }
    set variant(value) {
        this._variant = normalizeString(value, {
            fallbackValue: VARIANT_TRUNCATE,
            validValues: [VARIANT_TRUNCATE, VARIANT_PLAIN],
        });
    }
    @track
    _showStaticMap;
    @api
    get showStaticMap() {
        return this._showStaticMap || false;
    }
    set showStaticMap(value) {
        this._showStaticMap = normalizeBoolean(value);
    }
    @api
    click() {
        if (this.addressLink) {
            this.addressLink.click();
        }
    }
    get showPlainText() {
        return this.hasValue && this.isPlainText;
    }
    get showMapLink() {
        return this.hasValue && this.isMapLink;
    }
    get isPlainText() {
        return this.disabled;
    }
    get isMapLink() {
        return !this.disabled;
    }
    get hasValue() {
        return !!(this.street ||
            this.city ||
            this.province ||
            this.country ||
            this.postalCode);
    }
    get address() {
        const [langCode, countryCode] = this.locale.split('-');
        return (addressFormat.formatAddressAllFields(langCode, countryCode, {
            address: this.street,
            city: this.city,
            state: this.province,
            country: this.country,
            zipCode: this.postalCode,
        }) || '');
    }
    get addressLines() {
        return this.address.split('\n');
    }
    get mapQuery() {
        const { address, latitude, longitude } = this;
        return latitude && longitude ? `${latitude},${longitude}` : address;
    }
    get computedAriaLabel() {
        return this.address;
    }
    get internalTabIndex() {
        return this.getAttribute('tabindex');
    }
    get mapUrl() {
        return encodeURI(MAP_HOST) + encodeURIComponent(this.mapQuery);
    }
    get addressLink() {
        return this.template.querySelector('a');
    }
    get computedAddressLineClass() {
        return this.variant === VARIANT_TRUNCATE ? 'slds-truncate' : null;
    }
}
