// lightning-base-components/src/lightning/overlayContainer/overlayContainer.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import { api, createElement } from 'lwc';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
import { parent, instanceName, normalizeApis } from 'lightning/overlayUtils';
import { guid } from 'lightning/utilsPrivate';
import { addOverlayToSharedState, removeOverlayFromSharedState, normalizeOverlayDetails, subscribeOverlay, getStatCount, TYPE_TOAST_CONTAINER, LWC_OVERLAY_STARTING_ZINDEX, LWC_TOAST_CONTAINER_STARTING_ZINDEX, LWC_OVERLAY_TYPES, LWC_ZINDEX_INCREMENT, LWC_ZINDEX_OFFSET, LWC_OVERLAY_ENGINE, } from 'lightning/overlayManager';
export default class LightningOverlayContainer extends LightningShadowBaseClass {
    stack = [];
    afterFirstRender = false;
    get root() {
        return this.template.querySelector('[data-overlay-container]');
    }
    @api
    push(scope, apis) {
        apis = apis || {};
        if (typeof apis !== 'object') {
            throw new Error('Invalid .open() or .open({}) argument.');
        }
        if ('innerHTML' in apis || 'outerHTML' in apis) {
            throw new Error('Illegal api values');
        }
        const { [instanceName]: overlayInstanceName, [parent]: overlayParent } = scope;
        if (!this.isOverlayAllowedToAdd(overlayInstanceName)) {
            return Promise.reject();
        }
        const element = createElement(overlayInstanceName, { is: scope });
        const normalizedApis = normalizeApis(apis);
        Object.entries(normalizedApis).forEach(([key, value]) => {
            const keyLower = key.toLowerCase();
            const match = keyLower.match(/^on(.+)/);
            if (match) {
                const eventName = match[1];
                element.addEventListener(eventName, value);
            }
            else {
                element[key] = value;
            }
        });
        let promise = null;
        const id = guid();
        if (overlayParent) {
            const parentElement = createElement(overlayParent[instanceName], {
                is: overlayParent,
            });
            if (overlayInstanceName === 'lightning-overlay') {
                parentElement.focusTrapEnabled = apis.enableFocusTrap;
            }
            this.root.appendChild(parentElement);
            parentElement.defaultSlot.appendChild(element);
            promise = new Promise((resolve) => {
                this.stack.push({ resolve, element, parentElement, id });
            });
        }
        else {
            this.root.appendChild(element);
            promise = new Promise((resolve) => {
                this.stack.push({ resolve, element, id });
            });
        }
        const overlayDetails = normalizeOverlayDetails(LWC_OVERLAY_ENGINE, overlayInstanceName, { id });
        addOverlayToSharedState(overlayDetails);
        return promise;
    }
    findElementInStack(element) {
        return this.stack.find((item) => item.element === element);
    }
    findElementIndexInStack(element) {
        return this.stack.indexOf(((item) => item.element === element));
    }
    @api
    pop(element, result, promise) {
        const stackItemToRemove = this.findElementInStack(element);
        if (stackItemToRemove && !stackItemToRemove.isClosing) {
            stackItemToRemove.isClosing = true;
            const { element: elem, parentElement, resolve, id, } = stackItemToRemove;
            resolve(result);
            if (promise) {
                promise.then(() => {
                    this.removeOverlay(elem, parentElement, id);
                });
            }
            else {
                this.removeOverlay(elem, parentElement, id);
            }
        }
        else {
            console.info('The LightningOverlay extended component is closed or in the process of closing. this.close() was previously called.');
        }
        return null;
    }
    removeOverlay(element, parentElement, id) {
        const overlayType = element && element.tagName.toLowerCase();
        if (parentElement) {
            const defaultSlot = parentElement.defaultSlot;
            if (parentElement.parentNode) {
                parentElement.parentNode.removeChild(parentElement);
            }
            if (defaultSlot && defaultSlot.firstElementChild) {
                defaultSlot.removeChild(defaultSlot.firstElementChild);
            }
        }
        else if (element) {
            if (element.parentNode) {
                element.parentNode.removeChild(element);
            }
            else {
                element.remove();
            }
        }
        const indexToRemove = this.findElementIndexInStack(element);
        if (indexToRemove > -1) {
            this.stack.splice(indexToRemove, 1);
        }
        const overlayDetails = normalizeOverlayDetails(LWC_OVERLAY_ENGINE, overlayType, { id });
        removeOverlayFromSharedState(overlayDetails);
    }
    connectedCallback() {
        subscribeOverlay(this.afterFirstRender, (state) => {
            const { mixedOverlayPresent, auraHighestZindex, lwcAllOverlayCount, stack, } = state;
            if (lwcAllOverlayCount === 0) {
                return;
            }
            let lwcZIndexToSet = LWC_OVERLAY_STARTING_ZINDEX;
            let lwcOverlayCount = 0;
            const lwcToastContainerZIndexToSet = LWC_TOAST_CONTAINER_STARTING_ZINDEX;
            const toastOverlayStack = [];
            stack.forEach((overlayObject, stackIndex) => {
                const { overlayEngine, overlayType, overlayDetails } = overlayObject;
                let overlayElem;
                const positionToSet = (LWC_OVERLAY_TYPES[overlayType] &&
                    LWC_OVERLAY_TYPES[overlayType].pos) ||
                    'fixed';
                const isToastContainer = overlayType === TYPE_TOAST_CONTAINER;
                if (!mixedOverlayPresent) {
                    overlayElem = this.root.childNodes[lwcOverlayCount];
                    if (overlayElem) {
                        const zIndexToSet = isToastContainer
                            ? lwcToastContainerZIndexToSet
                            : lwcZIndexToSet;
                        overlayElem.style.zIndex = zIndexToSet;
                        overlayElem.style.position = positionToSet;
                        const updateDetailsObject = {
                            zIndex: zIndexToSet,
                        };
                        const updatedOverlayObject = normalizeOverlayDetails(overlayEngine, overlayType, Object.assign({}, overlayDetails, updateDetailsObject));
                        if (updatedOverlayObject !== null) {
                            state.stack[stackIndex] = updatedOverlayObject;
                        }
                        lwcOverlayCount += 1;
                    }
                }
                else if (overlayEngine === LWC_OVERLAY_ENGINE) {
                    overlayElem = this.root.childNodes[lwcOverlayCount];
                    const existingZindexValueIsSet = !!(overlayDetails && overlayDetails.zIndex);
                    const existingZindexValue = overlayDetails && overlayDetails.zIndex;
                    let lwcZIndex;
                    if (!existingZindexValueIsSet ||
                        Number.isNaN(existingZindexValue)) {
                        if (isToastContainer) {
                            lwcZIndex = lwcToastContainerZIndexToSet;
                        }
                        else {
                            lwcZIndexToSet =
                                stackIndex !== 0
                                    ? auraHighestZindex + LWC_ZINDEX_OFFSET
                                    : lwcZIndexToSet;
                            lwcZIndex = lwcZIndexToSet;
                        }
                        const updateDetailsObject = {
                            zIndex: lwcZIndex,
                        };
                        const updatedOverlayObj = normalizeOverlayDetails(overlayEngine, overlayType, Object.assign({}, overlayDetails, updateDetailsObject));
                        if (updatedOverlayObj !== null) {
                            state.stack[stackIndex] = updatedOverlayObj;
                        }
                    }
                    else if (isToastContainer) {
                        lwcZIndex =
                            existingZindexValue >=
                                LWC_TOAST_CONTAINER_STARTING_ZINDEX
                                ? existingZindexValue
                                : LWC_TOAST_CONTAINER_STARTING_ZINDEX;
                    }
                    else {
                        lwcZIndexToSet =
                            existingZindexValue >= LWC_OVERLAY_STARTING_ZINDEX
                                ? existingZindexValue
                                : LWC_OVERLAY_STARTING_ZINDEX;
                        lwcZIndex = lwcZIndexToSet;
                    }
                    if (overlayElem) {
                        overlayElem.style.zIndex = lwcZIndex;
                        overlayElem.style.position = positionToSet;
                        lwcOverlayCount += 1;
                    }
                    if (isToastContainer) {
                        toastOverlayStack.push({ overlayElem, stackIndex });
                    }
                }
                else {
                    lwcZIndexToSet += LWC_ZINDEX_INCREMENT;
                }
            });
        });
        this.afterFirstRender = true;
    }
    isOverlayAllowedToAdd(overlayType) {
        const isApprovedOverlayType = !!LWC_OVERLAY_TYPES[overlayType];
        if (!isApprovedOverlayType) {
            return false;
        }
        const { maxNum, statName } = LWC_OVERLAY_TYPES[overlayType];
        const maxNumNotSet = typeof maxNum === 'undefined';
        const count = (statName && getStatCount(statName)) || 0;
        return maxNumNotSet || maxNum > count;
    }
}
