// lightning-base-components/src/lightning/internationalizationLibrary/number/utils.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import locale from '@salesforce/i18n/locale';
import { getNumberFormat } from 'lightning/i18nService';
export const STYLE = {
    DECIMAL: 'decimal',
    CURRENCY: 'currency',
    PERCENT: 'percent',
};
const CURRENCY_DISPLAY = {
    CODE: 'code',
    SYMBOL: 'symbol',
    NAME: 'name',
};
const SAFE_NUM_LENGTH = 15;
const SAFE_NUM_REGEXP = /0+$/;
function toNumber(value) {
    return parseInt(value, 10);
}
function getFractionSkeleton(skeleton) {
    const format = skeleton.split(';')[0];
    return format.split('.')[1] || '';
}
function normalizeFractionDigits(skeleton, options) {
    const fractionSkeleton = getFractionSkeleton(skeleton);
    const defaultMinimum = fractionSkeleton.replace(/[^0]/g, '').length;
    const defaultMaximum = fractionSkeleton.replace(/[^0#]/g, '').length;
    const minDigits = toNumber(options.minimumFractionDigits);
    const maxDigits = toNumber(options.maximumFractionDigits);
    if (!isFinite(options.minimumFractionDigits)) {
        options.minimumFractionDigits = Math.min(defaultMinimum, isNaN(maxDigits) ? defaultMinimum : maxDigits);
    }
    if (!isFinite(options.maximumFractionDigits) || maxDigits < minDigits) {
        options.maximumFractionDigits = Math.max(defaultMaximum, isNaN(minDigits) ? defaultMaximum : minDigits);
    }
}
function getFractionPart(options) {
    const { minimumFractionDigits, maximumFractionDigits } = options;
    return ('.' +
        '0'.repeat(minimumFractionDigits) +
        '#'.repeat(maximumFractionDigits - minimumFractionDigits));
}
export function updateFractionPart(skeleton, options) {
    normalizeFractionDigits(skeleton, options);
    const fractionPart = getFractionPart(options);
    return addFractionsToPattern(skeleton, fractionPart);
}
function addFractionsToPattern(pattern, fractionPart) {
    if (!fractionPart) {
        return pattern;
    }
    if (pattern.indexOf(';') > 0) {
        const [positivePattern, negativePattern] = pattern.split(';');
        return `${addFractionsToPattern(positivePattern, fractionPart)};${addFractionsToPattern(negativePattern, fractionPart)}`;
    }
    if (pattern.indexOf('.') > 0) {
        return pattern.replace(/\.(0|#)*/, fractionPart);
    }
    const position = Math.max(pattern.lastIndexOf('0'), pattern.lastIndexOf('#')) + 1;
    return [
        pattern.slice(0, position),
        fractionPart,
        pattern.slice(position),
    ].join('');
}
function getGroupingCount(skeleton) {
    const match = skeleton.match(/,[#0]*\./);
    return match ? match[0].length - 2 : 0;
}
export function updateIntegerPart(skeleton, options) {
    const minimumIntegerDigits = options.minimumIntegerDigits;
    const groupingCount = getGroupingCount(skeleton);
    if (!minimumIntegerDigits) {
        return skeleton;
    }
    if (minimumIntegerDigits <= groupingCount) {
        return skeleton.replace(/,[#0]*\./, ',' +
            '#'.repeat(groupingCount - minimumIntegerDigits) +
            '0'.repeat(minimumIntegerDigits) +
            '.');
    }
    return skeleton.replace(/[#0]*\./, '0'.repeat(minimumIntegerDigits - groupingCount) +
        ',' +
        '0'.repeat(groupingCount) +
        '.');
}
function getBestMatchCurrencySymbol(code, currencyDisplay) {
    const opts = {
        style: STYLE.CURRENCY,
        currency: code,
        minimumFractionDigits: 0,
    };
    if (currencyDisplay) {
        opts.currencyDisplay = currencyDisplay;
    }
    const nf = getNumberFormat(locale, opts);
    return nf.format(2).replace(/2/g, '');
}
export function getCurrency(options) {
    const currencyDisplay = options.currencyDisplay || CURRENCY_DISPLAY.SYMBOL;
    if (currencyDisplay === CURRENCY_DISPLAY.SYMBOL ||
        currencyDisplay === CURRENCY_DISPLAY.NAME) {
        return getBestMatchCurrencySymbol(options.currency, currencyDisplay);
    }
    return options.currency;
}
export function updateCurrencySymbol(skeleton, currencyCode, options) {
    const symbol = String.fromCharCode(164);
    if (options.currencyDisplay === CURRENCY_DISPLAY.NAME) {
        return skeleton.replace(symbol, '') + currencyCode;
    }
    return skeleton.replace(symbol, currencyCode);
}
export function exceedsSafeLength(value) {
    const numberAsString = value.toString();
    const [intPart, fractionPart] = numberAsString.split('.');
    const digitCount = intPart.length +
        (fractionPart ? fractionPart.replace(SAFE_NUM_REGEXP, '').length : 0);
    return digitCount >= SAFE_NUM_LENGTH;
}
