// lightning-base-components/src/lightning/datatable/__examples__/withAllColumnTypes/generateData.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
function mulberry32(a) {
    return function () {
        let t = (a += 0x6d2b79f5);
        t = Math.imul(t ^ (t >>> 15), t | 1);
        t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
        return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
    };
}
const getRandom = mulberry32(12345);
export default function generateData({ amountOfRecords }) {
    return [...Array(amountOfRecords)].map((_, index) => {
        return {
            name: `Name (${index})`,
            boolean: getRandom() < 0.5,
            website: 'www.salesforce.com',
            phone: `${Math.floor((getRandom() * 9000000000)) + 1000000000}`,
            currency: Math.floor(getRandom() * 100),
            date: new Date(Date.now() + (86400000 * Math.ceil(getRandom() * 20))),
            dateLocal: new Date(Date.now()),
            email: 'test@salesforce.com',
            location: {
                latitude: (getRandom() * 90).toFixed(2),
                longitude: (getRandom() * 180).toFixed(2),
            },
            number: Math.floor(getRandom() * 100),
            percent: getRandom(),
            iconName: 'action:approval',
        };
    });
}
