// lightning-base-components/src/lightning/treeItem/treeItem.js
/*
 * Copyright (c) 2025, Salesforce, Inc.,
 * All rights reserved.
 * For full license text, see the LICENSE.txt file
 */
import labelCollapseBranch from '@salesforce/label/LightningTree.collapseBranch';
import labelExpandBranch from '@salesforce/label/LightningTree.expandBranch';
import { api, track } from 'lwc';
import { classSet } from 'lightning/utils';
import { isRTL, keyCodes, sanitizeURL } from 'lightning/utilsPrivate';
import LightningShadowBaseClass from 'lightning/shadowBaseClassPrivate';
const i18n = {
    collapseBranch: labelCollapseBranch,
    expandBranch: labelExpandBranch,
};
export default class LightningTreeItem extends LightningShadowBaseClass {
    static validationOptOut = ['class'];
    @track
    _children = [];
    @track
    _tabindexes = {};
    @track
    _selected = {};
    _focusedChild = null;
    _endIconProps;
    @api
    isRoot = false;
    @api
    label;
    @api
    metatext;
    @api
    nodeRef;
    @api
    isExpanded;
    @api
    isDisabled = false;
    @api
    nodename;
    @api
    nodeKey;
    @api
    isLeaf;
    @api
    selected;
    @api
    get endIcon() {
        return this._endIconProps;
    }
    set endIcon(value) {
        if (!value) {
            this._endIconProps = undefined;
            return;
        }
        const size = value.iconName.includes('utility') ? 'x-small' : 'small';
        this._endIconProps = Object.assign({ size }, value);
    }
    @api
    get childItems() {
        return this._children;
    }
    set childItems(value) {
        this._children = value;
        const childLen = this._children.length;
        for (let i = 0; i < childLen; i++) {
            this.setSelectedAttribute(i, 'false');
        }
    }
    @api
    get focusedChild() {
        return this._focusedChild;
    }
    set focusedChild(value) {
        this._focusedChild = value;
    }
    @api
    get href() {
        return this._href;
    }
    set href(value) {
        this._href = sanitizeURL(value);
    }
    _href;
    setSelectedAttribute(childNum, value) {
        this._selected[childNum] = value;
    }
    connectedCallback() {
        super.connectedCallback();
        if (!import.meta.env.SSR) {
            this.dispatchEvent(new CustomEvent('privateregisteritem', {
                composed: true,
                bubbles: true,
                detail: {
                    focusCallback: this.makeChildFocusable.bind(this),
                    unfocusCallback: this.makeChildUnfocusable.bind(this),
                    key: this.nodeKey,
                },
            }));
        }
        this.addEventListener('keydown', this.handleKeydown.bind(this));
    }
    renderedCallback() {
        if (typeof this.focusedChild === 'number') {
            const child = this.getNthChildItem(this.focusedChild + 1);
            if (child) {
                child.tabIndex = '0';
            }
        }
    }
    get buttonLabel() {
        if (this.nodeRef?.expanded) {
            return i18n.collapseBranch;
        }
        return i18n.expandBranch;
    }
    get showExpanded() {
        if (!this.nodeRef) {
            return false;
        }
        return !this.isDisabled && this.nodeRef.expanded;
    }
    get computedButtonClass() {
        return classSet('slds-button slds-button_icon slds-m-right_x-small ')
            .add({
            'slds-hidden': this.isLeaf || this.isDisabled,
        })
            .toString();
    }
    get computedTreeItemContentClass() {
        return this.endIcon
            ? 'content-with-end-icon'
            : 'slds-has-flexi-truncate';
    }
    get computedIconName() {
        return isRTL() ? 'utility:chevronleft' : 'utility:chevronright';
    }
    get children() {
        return this._children.map((child, idx) => {
            return {
                node: child,
                tabindex: this._tabindexes[idx],
                selected: this._selected[idx],
            };
        });
    }
    preventDefaultAndStopPropagation(event) {
        event.preventDefault();
        event.stopPropagation();
    }
    handleClick(event) {
        if (!this.isDisabled) {
            if (this.href === 'javascript:void(0)') {
                event.preventDefault();
            }
            let target = 'anchor';
            if (event.target.tagName === 'BUTTON' ||
                event.target.tagName === 'LIGHTNING-PRIMITIVE-ICON') {
                target = 'chevron';
            }
            const customEvent = new CustomEvent('privateitemclick', {
                bubbles: true,
                composed: true,
                cancelable: true,
                detail: {
                    name: this.nodename,
                    key: this.nodeKey,
                    target,
                },
            });
            this.dispatchEvent(customEvent);
        }
    }
    handleKeydown(event) {
        switch (event.keyCode) {
            case keyCodes.space:
            case keyCodes.enter:
                this.preventDefaultAndStopPropagation(event);
                this.template.querySelector('.slds-tree__item a').click();
                break;
            case keyCodes.up:
            case keyCodes.down:
            case keyCodes.right:
            case keyCodes.left:
            case keyCodes.home:
            case keyCodes.end:
                this.preventDefaultAndStopPropagation(event);
                this.dispatchEvent(new CustomEvent('privateitemkeydown', {
                    bubbles: true,
                    composed: true,
                    cancelable: true,
                    detail: {
                        key: this.nodeKey,
                        keyCode: event.keyCode,
                    },
                }));
                break;
            default:
                break;
        }
    }
    makeChildFocusable(childKey, shouldFocus, shouldSelect) {
        const child = this.getImmediateChildItem(childKey);
        if (child) {
            if (child.tabIndex !== '0') {
                child.tabIndex = '0';
            }
            if (shouldFocus) {
                child.focus();
            }
            if (shouldSelect) {
                child.ariaSelected = true;
            }
        }
    }
    makeChildUnfocusable() {
        this.ariaSelected = 'false';
        this.removeAttribute('tabindex');
    }
    getImmediateChildItem(key) {
        return this.template.querySelector("lightning-tree-item[data-key='" + key + "']");
    }
    getNthChildItem(n) {
        return this.template.querySelector('lightning-tree-item:nth-of-type(' + n + ')');
    }
}
